// @mtcute/core/highlevel/methods/stickers/_utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const normalizeFileToDocument = require("../files/normalize-file-to-document.cjs");
const sticker = require("../../types/media/sticker.cjs");
async function _normalizeInputStickerSetItem(client, sticker$1, params) {
  return {
    _: "inputStickerSetItem",
    document: await normalizeFileToDocument._normalizeFileToDocument(client, sticker$1.file, params ?? {}),
    emoji: sticker$1.emojis,
    maskCoords: sticker$1.maskPosition ? {
      _: "maskCoords",
      n: sticker.MASK_POSITION_POINT_TO_TL[sticker$1.maskPosition.point],
      x: sticker$1.maskPosition.x,
      y: sticker$1.maskPosition.y,
      zoom: sticker$1.maskPosition.scale
    } : void 0
  };
}
exports._normalizeInputStickerSetItem = _normalizeInputStickerSetItem;
