// @mtcute/core/highlevel/methods/stickers/create-sticker-set.js
import { _normalizeFileToDocument } from "../files/normalize-file-to-document.js";
import { resolveUser } from "../users/resolve-peer.js";
import { _normalizeInputStickerSetItem } from "./_utils.js";
import { StickerSet as StickerSet$2 } from "../../types/misc/sticker-set.js";
async function createStickerSet(client, params) {
  const owner = await resolveUser(client, params.owner);
  const inputStickers = [];
  let i = 0;
  for (const sticker of params.stickers) {
    const progressCallback = params.progressCallback?.bind(null, i);
    inputStickers.push(await _normalizeInputStickerSetItem(client, sticker, { progressCallback }));
    i += 1;
  }
  const res = await client.call({
    _: "stickers.createStickerSet",
    masks: params.type === "mask",
    emojis: params.type === "emoji",
    textColor: params.adaptive,
    userId: owner,
    title: params.title,
    shortName: params.shortName,
    stickers: inputStickers,
    thumb: params.thumb ? await _normalizeFileToDocument(client, params.thumb, {}) : void 0
  });
  return new StickerSet$2(res);
}
export {
  createStickerSet
};
