// @mtcute/core/highlevel/methods/stickers/create-sticker-set.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const normalizeFileToDocument = require("../files/normalize-file-to-document.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const _utils = require("./_utils.cjs");
const stickerSet = require("../../types/misc/sticker-set.cjs");
async function createStickerSet(client, params) {
  const owner = await resolvePeer.resolveUser(client, params.owner);
  const inputStickers = [];
  let i = 0;
  for (const sticker of params.stickers) {
    const progressCallback = params.progressCallback?.bind(null, i);
    inputStickers.push(await _utils._normalizeInputStickerSetItem(client, sticker, { progressCallback }));
    i += 1;
  }
  const res = await client.call({
    _: "stickers.createStickerSet",
    masks: params.type === "mask",
    emojis: params.type === "emoji",
    textColor: params.adaptive,
    userId: owner,
    title: params.title,
    shortName: params.shortName,
    stickers: inputStickers,
    thumb: params.thumb ? await normalizeFileToDocument._normalizeFileToDocument(client, params.thumb, {}) : void 0
  });
  return new stickerSet.StickerSet(res);
}
exports.createStickerSet = createStickerSet;
