// @mtcute/core/client.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const base = require("./highlevel/base.cjs");
const checkPassword = require("./highlevel/methods/auth/check-password.cjs");
const getPasswordHint = require("./highlevel/methods/auth/get-password-hint.cjs");
const logOut = require("./highlevel/methods/auth/log-out.cjs");
const recoverPassword = require("./highlevel/methods/auth/recover-password.cjs");
const resendCode = require("./highlevel/methods/auth/resend-code.cjs");
const sendCode = require("./highlevel/methods/auth/send-code.cjs");
const sendRecoveryCode = require("./highlevel/methods/auth/send-recovery-code.cjs");
const signInBot = require("./highlevel/methods/auth/sign-in-bot.cjs");
const signInQr = require("./highlevel/methods/auth/sign-in-qr.cjs");
const signIn = require("./highlevel/methods/auth/sign-in.cjs");
const startTest = require("./highlevel/methods/auth/start-test.cjs");
const start = require("./highlevel/methods/auth/start.cjs");
const utils$1 = require("./highlevel/methods/auth/utils.cjs");
const answerBotGuestChatQuery = require("./highlevel/methods/bots/answer-bot-guest-chat-query.cjs");
const answerCallbackQuery = require("./highlevel/methods/bots/answer-callback-query.cjs");
const answerInlineQuery = require("./highlevel/methods/bots/answer-inline-query.cjs");
const answerPreCheckoutQuery = require("./highlevel/methods/bots/answer-pre-checkout-query.cjs");
const deleteMyCommands = require("./highlevel/methods/bots/delete-my-commands.cjs");
const getBotInfo = require("./highlevel/methods/bots/get-bot-info.cjs");
const getBotMenuButton = require("./highlevel/methods/bots/get-bot-menu-button.cjs");
const getCallbackAnswer = require("./highlevel/methods/bots/get-callback-answer.cjs");
const getGameHighScores = require("./highlevel/methods/bots/get-game-high-scores.cjs");
const getMyCommands = require("./highlevel/methods/bots/get-my-commands.cjs");
const openWebview = require("./highlevel/methods/bots/open-webview.cjs");
const prepareInlineMessage = require("./highlevel/methods/bots/prepare-inline-message.cjs");
const setBotInfo = require("./highlevel/methods/bots/set-bot-info.cjs");
const setBotMenuButton = require("./highlevel/methods/bots/set-bot-menu-button.cjs");
const setGameScore = require("./highlevel/methods/bots/set-game-score.cjs");
const setMyCommands = require("./highlevel/methods/bots/set-my-commands.cjs");
const setMyDefaultRights = require("./highlevel/methods/bots/set-my-default-rights.cjs");
const toggleEmojiStatusPermission = require("./highlevel/methods/bots/toggle-emoji-status-permission.cjs");
const addChatMembers = require("./highlevel/methods/chats/add-chat-members.cjs");
const archiveChats = require("./highlevel/methods/chats/archive-chats.cjs");
const banChatMember = require("./highlevel/methods/chats/ban-chat-member.cjs");
const createChannel = require("./highlevel/methods/chats/create-channel.cjs");
const createGroup = require("./highlevel/methods/chats/create-group.cjs");
const createSupergroup = require("./highlevel/methods/chats/create-supergroup.cjs");
const deleteChannel = require("./highlevel/methods/chats/delete-channel.cjs");
const deleteChatPhoto = require("./highlevel/methods/chats/delete-chat-photo.cjs");
const deleteGroup = require("./highlevel/methods/chats/delete-group.cjs");
const deleteHistory = require("./highlevel/methods/chats/delete-history.cjs");
const deleteUserHistory = require("./highlevel/methods/chats/delete-user-history.cjs");
const editAdminRights = require("./highlevel/methods/chats/edit-admin-rights.cjs");
const editChatMemberRank = require("./highlevel/methods/chats/edit-chat-member-rank.cjs");
const getChatEventLog = require("./highlevel/methods/chats/get-chat-event-log.cjs");
const getChatMember = require("./highlevel/methods/chats/get-chat-member.cjs");
const getChatMembers = require("./highlevel/methods/chats/get-chat-members.cjs");
const getChatPreview = require("./highlevel/methods/chats/get-chat-preview.cjs");
const getChat = require("./highlevel/methods/chats/get-chat.cjs");
const getChats = require("./highlevel/methods/chats/get-chats.cjs");
const getCreatorAfterLeave = require("./highlevel/methods/chats/get-creator-after-leave.cjs");
const getFullChat = require("./highlevel/methods/chats/get-full-chat.cjs");
const getFullUser = require("./highlevel/methods/chats/get-full-user.cjs");
const getMessageAuthor = require("./highlevel/methods/chats/get-message-author.cjs");
const getPeer = require("./highlevel/methods/chats/get-peer.cjs");
const getPeers = require("./highlevel/methods/chats/get-peers.cjs");
const getSimilarChannels = require("./highlevel/methods/chats/get-similar-channels.cjs");
const getUser = require("./highlevel/methods/chats/get-user.cjs");
const iterChatEventLog = require("./highlevel/methods/chats/iter-chat-event-log.cjs");
const iterChatMembers = require("./highlevel/methods/chats/iter-chat-members.cjs");
const joinChat = require("./highlevel/methods/chats/join-chat.cjs");
const kickChatMember = require("./highlevel/methods/chats/kick-chat-member.cjs");
const leaveChat = require("./highlevel/methods/chats/leave-chat.cjs");
const markChatUnread = require("./highlevel/methods/chats/mark-chat-unread.cjs");
const openChat = require("./highlevel/methods/chats/open-chat.cjs");
const reorderUsernames = require("./highlevel/methods/chats/reorder-usernames.cjs");
const restrictChatMember = require("./highlevel/methods/chats/restrict-chat-member.cjs");
const saveDraft = require("./highlevel/methods/chats/save-draft.cjs");
const setChatColor = require("./highlevel/methods/chats/set-chat-color.cjs");
const setChatDefaultPermissions = require("./highlevel/methods/chats/set-chat-default-permissions.cjs");
const setChatDescription = require("./highlevel/methods/chats/set-chat-description.cjs");
const setChatPhoto = require("./highlevel/methods/chats/set-chat-photo.cjs");
const setChatTitle = require("./highlevel/methods/chats/set-chat-title.cjs");
const setChatTtl = require("./highlevel/methods/chats/set-chat-ttl.cjs");
const setChatUsername = require("./highlevel/methods/chats/set-chat-username.cjs");
const setSlowMode = require("./highlevel/methods/chats/set-slow-mode.cjs");
const toggleContentProtection = require("./highlevel/methods/chats/toggle-content-protection.cjs");
const toggleFragmentUsername = require("./highlevel/methods/chats/toggle-fragment-username.cjs");
const toggleJoinRequests = require("./highlevel/methods/chats/toggle-join-requests.cjs");
const toggleJoinToSend = require("./highlevel/methods/chats/toggle-join-to-send.cjs");
const transferChatOwnership = require("./highlevel/methods/chats/transfer-chat-ownership.cjs");
const unarchiveChats = require("./highlevel/methods/chats/unarchive-chats.cjs");
const unbanChatMember = require("./highlevel/methods/chats/unban-chat-member.cjs");
const addContact = require("./highlevel/methods/contacts/add-contact.cjs");
const deleteContacts = require("./highlevel/methods/contacts/delete-contacts.cjs");
const getContacts = require("./highlevel/methods/contacts/get-contacts.cjs");
const importContacts = require("./highlevel/methods/contacts/import-contacts.cjs");
const setContactNote = require("./highlevel/methods/contacts/set-contact-note.cjs");
const createFolder = require("./highlevel/methods/dialogs/create-folder.cjs");
const deleteFolder = require("./highlevel/methods/dialogs/delete-folder.cjs");
const editFolder = require("./highlevel/methods/dialogs/edit-folder.cjs");
const findDialogs = require("./highlevel/methods/dialogs/find-dialogs.cjs");
const findFolder = require("./highlevel/methods/dialogs/find-folder.cjs");
const getChatlistPreview = require("./highlevel/methods/dialogs/get-chatlist-preview.cjs");
const getFolders = require("./highlevel/methods/dialogs/get-folders.cjs");
const getPeerDialogs = require("./highlevel/methods/dialogs/get-peer-dialogs.cjs");
const iterDialogs = require("./highlevel/methods/dialogs/iter-dialogs.cjs");
const joinChatlist = require("./highlevel/methods/dialogs/join-chatlist.cjs");
const setFoldersOrder = require("./highlevel/methods/dialogs/set-folders-order.cjs");
const downloadBuffer = require("./highlevel/methods/files/download-buffer.cjs");
const downloadChunk = require("./highlevel/methods/files/download-chunk.cjs");
const downloadIterable = require("./highlevel/methods/files/download-iterable.cjs");
const downloadStream = require("./highlevel/methods/files/download-stream.cjs");
const normalizeInputFile = require("./highlevel/methods/files/normalize-input-file.cjs");
const normalizeInputMedia = require("./highlevel/methods/files/normalize-input-media.cjs");
const uploadFile = require("./highlevel/methods/files/upload-file.cjs");
const uploadMedia = require("./highlevel/methods/files/upload-media.cjs");
const createForumTopic = require("./highlevel/methods/forums/create-forum-topic.cjs");
const deleteForumTopicHistory = require("./highlevel/methods/forums/delete-forum-topic-history.cjs");
const editForumTopic = require("./highlevel/methods/forums/edit-forum-topic.cjs");
const getForumTopicsById = require("./highlevel/methods/forums/get-forum-topics-by-id.cjs");
const getForumTopics = require("./highlevel/methods/forums/get-forum-topics.cjs");
const iterForumTopics = require("./highlevel/methods/forums/iter-forum-topics.cjs");
const reorderPinnedForumTopics = require("./highlevel/methods/forums/reorder-pinned-forum-topics.cjs");
const toggleForumTopicClosed = require("./highlevel/methods/forums/toggle-forum-topic-closed.cjs");
const toggleForumTopicPinned = require("./highlevel/methods/forums/toggle-forum-topic-pinned.cjs");
const toggleForum = require("./highlevel/methods/forums/toggle-forum.cjs");
const toggleGeneralTopicHidden = require("./highlevel/methods/forums/toggle-general-topic-hidden.cjs");
const _normalizeInputStarGift = require("./highlevel/methods/gifts/_normalize-input-star-gift.cjs");
const acceptStarGift = require("./highlevel/methods/gifts/accept-star-gift.cjs");
const buyResaleGift = require("./highlevel/methods/gifts/buy-resale-gift.cjs");
const getResaleStarGifts = require("./highlevel/methods/gifts/get-resale-star-gifts.cjs");
const getSavedStarGiftsById = require("./highlevel/methods/gifts/get-saved-star-gifts-by-id.cjs");
const getSavedStarGifts = require("./highlevel/methods/gifts/get-saved-star-gifts.cjs");
const getStarGiftOptions = require("./highlevel/methods/gifts/get-star-gift-options.cjs");
const getStarGiftUpgradeOptions = require("./highlevel/methods/gifts/get-star-gift-upgrade-options.cjs");
const getStarGiftValue = require("./highlevel/methods/gifts/get-star-gift-value.cjs");
const getStarGiftWithdrawalUrl = require("./highlevel/methods/gifts/get-star-gift-withdrawal-url.cjs");
const getUniqueStarGift = require("./highlevel/methods/gifts/get-unique-star-gift.cjs");
const iterSavedStarGifts = require("./highlevel/methods/gifts/iter-saved-star-gifts.cjs");
const pinStarGift = require("./highlevel/methods/gifts/pin-star-gift.cjs");
const prepayStarGiftUpgrade = require("./highlevel/methods/gifts/prepay-star-gift-upgrade.cjs");
const resolveStarGiftOffer = require("./highlevel/methods/gifts/resolve-star-gift-offer.cjs");
const sendStarGiftOffer = require("./highlevel/methods/gifts/send-star-gift-offer.cjs");
const sendStarGift = require("./highlevel/methods/gifts/send-star-gift.cjs");
const setGiftResalePrice = require("./highlevel/methods/gifts/set-gift-resale-price.cjs");
const transferStarGift = require("./highlevel/methods/gifts/transfer-star-gift.cjs");
const upgradeStarGift = require("./highlevel/methods/gifts/upgrade-star-gift.cjs");
const createInviteLink = require("./highlevel/methods/invite-links/create-invite-link.cjs");
const editInviteLink = require("./highlevel/methods/invite-links/edit-invite-link.cjs");
const exportInviteLink = require("./highlevel/methods/invite-links/export-invite-link.cjs");
const getInviteLinkMembers = require("./highlevel/methods/invite-links/get-invite-link-members.cjs");
const getInviteLink = require("./highlevel/methods/invite-links/get-invite-link.cjs");
const getInviteLinks = require("./highlevel/methods/invite-links/get-invite-links.cjs");
const getPrimaryInviteLink = require("./highlevel/methods/invite-links/get-primary-invite-link.cjs");
const hideAllJoinRequests = require("./highlevel/methods/invite-links/hide-all-join-requests.cjs");
const hideJoinRequest = require("./highlevel/methods/invite-links/hide-join-request.cjs");
const iterInviteLinkMembers = require("./highlevel/methods/invite-links/iter-invite-link-members.cjs");
const iterInviteLinks = require("./highlevel/methods/invite-links/iter-invite-links.cjs");
const revokeInviteLink = require("./highlevel/methods/invite-links/revoke-invite-link.cjs");
const appendTodoList = require("./highlevel/methods/messages/append-todo-list.cjs");
const closePoll = require("./highlevel/methods/messages/close-poll.cjs");
const createRichStreamingDraft = require("./highlevel/methods/messages/create-rich-streaming-draft.cjs");
const createStreamingDraft = require("./highlevel/methods/messages/create-streaming-draft.cjs");
const deleteMessages = require("./highlevel/methods/messages/delete-messages.cjs");
const deleteScheduledMessages = require("./highlevel/methods/messages/delete-scheduled-messages.cjs");
const editInlineMessage = require("./highlevel/methods/messages/edit-inline-message.cjs");
const editMessage = require("./highlevel/methods/messages/edit-message.cjs");
const forwardMessages = require("./highlevel/methods/messages/forward-messages.cjs");
const getAllScheduledMessages = require("./highlevel/methods/messages/get-all-scheduled-messages.cjs");
const getAvailableEffects = require("./highlevel/methods/messages/get-available-effects.cjs");
const getCallbackQueryMessage = require("./highlevel/methods/messages/get-callback-query-message.cjs");
const getDiscussionMessage = require("./highlevel/methods/messages/get-discussion-message.cjs");
const getFactCheck = require("./highlevel/methods/messages/get-fact-check.cjs");
const getHistory = require("./highlevel/methods/messages/get-history.cjs");
const getMessageByLink = require("./highlevel/methods/messages/get-message-by-link.cjs");
const getMessageGroup = require("./highlevel/methods/messages/get-message-group.cjs");
const getMessageReactions = require("./highlevel/methods/messages/get-message-reactions.cjs");
const getMessagesUnsafe = require("./highlevel/methods/messages/get-messages-unsafe.cjs");
const getMessages = require("./highlevel/methods/messages/get-messages.cjs");
const getReactionUsers = require("./highlevel/methods/messages/get-reaction-users.cjs");
const getReplyTo = require("./highlevel/methods/messages/get-reply-to.cjs");
const getScheduledMessages = require("./highlevel/methods/messages/get-scheduled-messages.cjs");
const getWebPagePreview = require("./highlevel/methods/messages/get-web-page-preview.cjs");
const iterHistory = require("./highlevel/methods/messages/iter-history.cjs");
const iterReactionUsers = require("./highlevel/methods/messages/iter-reaction-users.cjs");
const iterSearchGlobal = require("./highlevel/methods/messages/iter-search-global.cjs");
const iterSearchMessages = require("./highlevel/methods/messages/iter-search-messages.cjs");
const pinMessage = require("./highlevel/methods/messages/pin-message.cjs");
const readHistory = require("./highlevel/methods/messages/read-history.cjs");
const readReactions = require("./highlevel/methods/messages/read-reactions.cjs");
const searchGlobal = require("./highlevel/methods/messages/search-global.cjs");
const searchHashtag = require("./highlevel/methods/messages/search-hashtag.cjs");
const searchMessages = require("./highlevel/methods/messages/search-messages.cjs");
const searchPostsGlobal = require("./highlevel/methods/messages/search-posts-global.cjs");
const sendAnswer = require("./highlevel/methods/messages/send-answer.cjs");
const sendComment = require("./highlevel/methods/messages/send-comment.cjs");
const sendCopyGroup = require("./highlevel/methods/messages/send-copy-group.cjs");
const sendCopy = require("./highlevel/methods/messages/send-copy.cjs");
const sendMediaGroup = require("./highlevel/methods/messages/send-media-group.cjs");
const sendMedia = require("./highlevel/methods/messages/send-media.cjs");
const sendPaidReaction = require("./highlevel/methods/messages/send-paid-reaction.cjs");
const sendQuote = require("./highlevel/methods/messages/send-quote.cjs");
const sendReaction = require("./highlevel/methods/messages/send-reaction.cjs");
const sendReply = require("./highlevel/methods/messages/send-reply.cjs");
const sendRichMessage = require("./highlevel/methods/messages/send-rich-message.cjs");
const sendScheduled = require("./highlevel/methods/messages/send-scheduled.cjs");
const sendText = require("./highlevel/methods/messages/send-text.cjs");
const sendTyping = require("./highlevel/methods/messages/send-typing.cjs");
const sendVote = require("./highlevel/methods/messages/send-vote.cjs");
const setTyping = require("./highlevel/methods/messages/set-typing.cjs");
const toggleTodoCompleted = require("./highlevel/methods/messages/toggle-todo-completed.cjs");
const translateMessage = require("./highlevel/methods/messages/translate-message.cjs");
const translateText = require("./highlevel/methods/messages/translate-text.cjs");
const unpinAllMessages = require("./highlevel/methods/messages/unpin-all-messages.cjs");
const unpinMessage = require("./highlevel/methods/messages/unpin-message.cjs");
const getCollectibleInfo = require("./highlevel/methods/misc/get-collectible-info.cjs");
const initTakeoutSession = require("./highlevel/methods/misc/init-takeout-session.cjs");
const normalizePrivacyRules = require("./highlevel/methods/misc/normalize-privacy-rules.cjs");
const withParams = require("./highlevel/methods/misc/with-params.cjs");
const changeCloudPassword = require("./highlevel/methods/password/change-cloud-password.cjs");
const enableCloudPassword = require("./highlevel/methods/password/enable-cloud-password.cjs");
const passwordEmail = require("./highlevel/methods/password/password-email.cjs");
const removeCloudPassword = require("./highlevel/methods/password/remove-cloud-password.cjs");
const applyBoost = require("./highlevel/methods/premium/apply-boost.cjs");
const canApplyBoost = require("./highlevel/methods/premium/can-apply-boost.cjs");
const createBusinessChatLink = require("./highlevel/methods/premium/create-business-chat-link.cjs");
const editBusinessChatLink = require("./highlevel/methods/premium/edit-business-chat-link.cjs");
const getBoostStats = require("./highlevel/methods/premium/get-boost-stats.cjs");
const getBoosts = require("./highlevel/methods/premium/get-boosts.cjs");
const getBusinessChatLinks = require("./highlevel/methods/premium/get-business-chat-links.cjs");
const getBusinessConnection = require("./highlevel/methods/premium/get-business-connection.cjs");
const getMyBoostSlots = require("./highlevel/methods/premium/get-my-boost-slots.cjs");
const getStarsTransactions = require("./highlevel/methods/premium/get-stars-transactions.cjs");
const iterBoosters = require("./highlevel/methods/premium/iter-boosters.cjs");
const iterStarsTransactions = require("./highlevel/methods/premium/iter-stars-transactions.cjs");
const setBusinessIntro = require("./highlevel/methods/premium/set-business-intro.cjs");
const setBusinessWorkHours = require("./highlevel/methods/premium/set-business-work-hours.cjs");
const addStickerToSet = require("./highlevel/methods/stickers/add-sticker-to-set.cjs");
const createStickerSet = require("./highlevel/methods/stickers/create-sticker-set.cjs");
const deleteStickerFromSet = require("./highlevel/methods/stickers/delete-sticker-from-set.cjs");
const getCustomEmojis = require("./highlevel/methods/stickers/get-custom-emojis.cjs");
const getInstalledStickers = require("./highlevel/methods/stickers/get-installed-stickers.cjs");
const getMyStickerSets = require("./highlevel/methods/stickers/get-my-sticker-sets.cjs");
const getStickerSet = require("./highlevel/methods/stickers/get-sticker-set.cjs");
const moveStickerInSet = require("./highlevel/methods/stickers/move-sticker-in-set.cjs");
const replaceStickerInSet = require("./highlevel/methods/stickers/replace-sticker-in-set.cjs");
const setChatStickerSet = require("./highlevel/methods/stickers/set-chat-sticker-set.cjs");
const setStickerSetThumb = require("./highlevel/methods/stickers/set-sticker-set-thumb.cjs");
const canSendStory = require("./highlevel/methods/stories/can-send-story.cjs");
const deleteStories = require("./highlevel/methods/stories/delete-stories.cjs");
const editStory = require("./highlevel/methods/stories/edit-story.cjs");
const getAllStories = require("./highlevel/methods/stories/get-all-stories.cjs");
const getPeerStories = require("./highlevel/methods/stories/get-peer-stories.cjs");
const getProfileStories = require("./highlevel/methods/stories/get-profile-stories.cjs");
const getStoriesById = require("./highlevel/methods/stories/get-stories-by-id.cjs");
const getStoriesInteractions = require("./highlevel/methods/stories/get-stories-interactions.cjs");
const getStoryLink = require("./highlevel/methods/stories/get-story-link.cjs");
const getStoryViewers = require("./highlevel/methods/stories/get-story-viewers.cjs");
const hideMyStoriesViews = require("./highlevel/methods/stories/hide-my-stories-views.cjs");
const incrementStoriesViews = require("./highlevel/methods/stories/increment-stories-views.cjs");
const iterAllStories = require("./highlevel/methods/stories/iter-all-stories.cjs");
const iterProfileStories = require("./highlevel/methods/stories/iter-profile-stories.cjs");
const iterStoryViewers = require("./highlevel/methods/stories/iter-story-viewers.cjs");
const readStories = require("./highlevel/methods/stories/read-stories.cjs");
const sendStoryReaction = require("./highlevel/methods/stories/send-story-reaction.cjs");
const sendStory = require("./highlevel/methods/stories/send-story.cjs");
const togglePeerStoriesArchived = require("./highlevel/methods/stories/toggle-peer-stories-archived.cjs");
const toggleStoriesPinned = require("./highlevel/methods/stories/toggle-stories-pinned.cjs");
const blockUser = require("./highlevel/methods/users/block-user.cjs");
const deleteProfilePhotos = require("./highlevel/methods/users/delete-profile-photos.cjs");
const editCloseFriends = require("./highlevel/methods/users/edit-close-friends.cjs");
const getCommonChats = require("./highlevel/methods/users/get-common-chats.cjs");
const getGlobalTtl = require("./highlevel/methods/users/get-global-ttl.cjs");
const getMe = require("./highlevel/methods/users/get-me.cjs");
const getMyUsername = require("./highlevel/methods/users/get-my-username.cjs");
const getPeerSettings = require("./highlevel/methods/users/get-peer-settings.cjs");
const getProfilePhoto = require("./highlevel/methods/users/get-profile-photo.cjs");
const getProfilePhotos = require("./highlevel/methods/users/get-profile-photos.cjs");
const getSavedMusic = require("./highlevel/methods/users/get-saved-music.cjs");
const getUsers = require("./highlevel/methods/users/get-users.cjs");
const isPeerAvailable = require("./highlevel/methods/users/is-peer-available.cjs");
const iterProfilePhotos = require("./highlevel/methods/users/iter-profile-photos.cjs");
const resolvePeerMany = require("./highlevel/methods/users/resolve-peer-many.cjs");
const resolvePeer = require("./highlevel/methods/users/resolve-peer.cjs");
const resolvePhoneNumber = require("./highlevel/methods/users/resolve-phone-number.cjs");
const saveMusicToProfile = require("./highlevel/methods/users/save-music-to-profile.cjs");
const setEmojiStatus = require("./highlevel/methods/users/set-emoji-status.cjs");
const setGlobalTtl = require("./highlevel/methods/users/set-global-ttl.cjs");
const setMyBirthday = require("./highlevel/methods/users/set-my-birthday.cjs");
const setMyProfilePhoto = require("./highlevel/methods/users/set-my-profile-photo.cjs");
const setMyUsername = require("./highlevel/methods/users/set-my-username.cjs");
const setOnline = require("./highlevel/methods/users/set-online.cjs");
const unblockUser = require("./highlevel/methods/users/unblock-user.cjs");
const updateProfile = require("./highlevel/methods/users/update-profile.cjs");
const conversation = require("./highlevel/types/conversation.cjs");
const parsed = require("./highlevel/updates/parsed.cjs");
const errors = require("./types/errors.cjs");
class TelegramClient {
  _client;
  constructor(opts) {
    this.onRawUpdate = new utils.Emitter();
    this.onUpdate = new utils.Emitter();
    this.onNewMessage = new utils.Emitter();
    this.onEditMessage = new utils.Emitter();
    this.onMessageGroup = new utils.Emitter();
    this.onDeleteMessage = new utils.Emitter();
    this.onChatMemberUpdate = new utils.Emitter();
    this.onInlineQuery = new utils.Emitter();
    this.onBotGuestChatQuery = new utils.Emitter();
    this.onChosenInlineResult = new utils.Emitter();
    this.onCallbackQuery = new utils.Emitter();
    this.onInlineCallbackQuery = new utils.Emitter();
    this.onBusinessCallbackQuery = new utils.Emitter();
    this.onPollUpdate = new utils.Emitter();
    this.onPollVote = new utils.Emitter();
    this.onUserStatusUpdate = new utils.Emitter();
    this.onUserTyping = new utils.Emitter();
    this.onHistoryRead = new utils.Emitter();
    this.onBotStopped = new utils.Emitter();
    this.onBotChatJoinRequest = new utils.Emitter();
    this.onChatJoinRequest = new utils.Emitter();
    this.onPreCheckoutQuery = new utils.Emitter();
    this.onStoryUpdate = new utils.Emitter();
    this.onDeleteStory = new utils.Emitter();
    this.onBotReactionUpdate = new utils.Emitter();
    this.onBotReactionCountUpdate = new utils.Emitter();
    this.onBusinessConnectionUpdate = new utils.Emitter();
    this.onNewBusinessMessage = new utils.Emitter();
    this.onEditBusinessMessage = new utils.Emitter();
    this.onBusinessMessageGroup = new utils.Emitter();
    this.onDeleteBusinessMessage = new utils.Emitter();
    if ("client" in opts) {
      this._client = opts.client;
    } else {
      if (!opts.storage || typeof opts.storage === "string" || !opts.transport || !opts.crypto || !opts.platform) {
        throw new errors.MtUnsupportedError(
          "You need to explicitly provide storage, transport, crypto and platform for @mtcute/core"
        );
      }
      this._client = new base.BaseTelegramClient(opts);
    }
    Object.defineProperty(this, "log", { value: this._client.log });
    Object.defineProperty(this, "storage", { value: this._client.storage });
    Object.defineProperty(this, "timers", { value: this._client.timers });
    Object.defineProperty(this, "stopSignal", { value: this._client.stopSignal });
    Object.defineProperty(this, "appConfig", { value: this._client.appConfig });
    Object.defineProperty(this, "platform", { value: this._client.platform });
    Object.defineProperty(this, "onServerUpdate", { value: this._client.onServerUpdate });
    Object.defineProperty(this, "onRawUpdate", { value: this._client.onRawUpdate });
    Object.defineProperty(this, "onConnectionState", { value: this._client.onConnectionState });
    Object.defineProperty(this, "onError", { value: this._client.onError });
    if (!opts.disableUpdates) {
      const skipConversationUpdates = opts.skipConversationUpdates ?? true;
      const { messageGroupingInterval } = opts.updates ?? {};
      this._client.onRawUpdate.add(
        parsed.makeParsedUpdateHandler({
          messageGroupingInterval,
          onUpdate: (update) => {
            if (conversation.Conversation.handleUpdate(this, update) && skipConversationUpdates) return;
            this.onUpdate.emit(update);
            switch (update.name) {
              case "new_message":
                this.onNewMessage.emit(update.data);
                break;
              case "edit_message":
                this.onEditMessage.emit(update.data);
                break;
              case "message_group":
                this.onMessageGroup.emit(update.data);
                break;
              case "delete_message":
                this.onDeleteMessage.emit(update.data);
                break;
              case "chat_member":
                this.onChatMemberUpdate.emit(update.data);
                break;
              case "inline_query":
                this.onInlineQuery.emit(update.data);
                break;
              case "bot_guest_chat_query":
                this.onBotGuestChatQuery.emit(update.data);
                break;
              case "chosen_inline_result":
                this.onChosenInlineResult.emit(update.data);
                break;
              case "callback_query":
                this.onCallbackQuery.emit(update.data);
                break;
              case "inline_callback_query":
                this.onInlineCallbackQuery.emit(update.data);
                break;
              case "business_callback_query":
                this.onBusinessCallbackQuery.emit(update.data);
                break;
              case "poll":
                this.onPollUpdate.emit(update.data);
                break;
              case "poll_vote":
                this.onPollVote.emit(update.data);
                break;
              case "user_status":
                this.onUserStatusUpdate.emit(update.data);
                break;
              case "user_typing":
                this.onUserTyping.emit(update.data);
                break;
              case "history_read":
                this.onHistoryRead.emit(update.data);
                break;
              case "bot_stopped":
                this.onBotStopped.emit(update.data);
                break;
              case "bot_chat_join_request":
                this.onBotChatJoinRequest.emit(update.data);
                break;
              case "chat_join_request":
                this.onChatJoinRequest.emit(update.data);
                break;
              case "pre_checkout_query":
                this.onPreCheckoutQuery.emit(update.data);
                break;
              case "story":
                this.onStoryUpdate.emit(update.data);
                break;
              case "delete_story":
                this.onDeleteStory.emit(update.data);
                break;
              case "bot_reaction":
                this.onBotReactionUpdate.emit(update.data);
                break;
              case "bot_reaction_count":
                this.onBotReactionCountUpdate.emit(update.data);
                break;
              case "business_connection":
                this.onBusinessConnectionUpdate.emit(update.data);
                break;
              case "new_business_message":
                this.onNewBusinessMessage.emit(update.data);
                break;
              case "edit_business_message":
                this.onEditBusinessMessage.emit(update.data);
                break;
              case "business_message_group":
                this.onBusinessMessageGroup.emit(update.data);
                break;
              case "delete_business_message":
                this.onDeleteBusinessMessage.emit(update.data);
                break;
            }
          }
        })
      );
    }
  }
  withParams(params) {
    return withParams.withParams(this, params);
  }
}
TelegramClient.prototype.checkPassword = function(...args) {
  return checkPassword.checkPassword(this._client, ...args);
};
TelegramClient.prototype.getPasswordHint = function(...args) {
  return getPasswordHint.getPasswordHint(this._client, ...args);
};
TelegramClient.prototype.logOut = function(...args) {
  return logOut.logOut(this._client, ...args);
};
TelegramClient.prototype.recoverPassword = function(...args) {
  return recoverPassword.recoverPassword(this._client, ...args);
};
TelegramClient.prototype.resendCode = function(...args) {
  return resendCode.resendCode(this._client, ...args);
};
TelegramClient.prototype.sendCode = function(...args) {
  return sendCode.sendCode(this._client, ...args);
};
TelegramClient.prototype.sendRecoveryCode = function(...args) {
  return sendRecoveryCode.sendRecoveryCode(this._client, ...args);
};
TelegramClient.prototype.signInBot = function(...args) {
  return signInBot.signInBot(this._client, ...args);
};
TelegramClient.prototype.signInQr = function(...args) {
  return signInQr.signInQr(this._client, ...args);
};
TelegramClient.prototype.signIn = function(...args) {
  return signIn.signIn(this._client, ...args);
};
TelegramClient.prototype.startTest = function(...args) {
  return startTest.startTest(this._client, ...args);
};
TelegramClient.prototype.start = function(...args) {
  return start.start(this._client, ...args);
};
TelegramClient.prototype.isSelfPeer = function(...args) {
  return utils$1.isSelfPeer(this._client, ...args);
};
TelegramClient.prototype.answerBotGuestChatQuery = function(...args) {
  return answerBotGuestChatQuery.answerBotGuestChatQuery(this._client, ...args);
};
TelegramClient.prototype.answerCallbackQuery = function(...args) {
  return answerCallbackQuery.answerCallbackQuery(this._client, ...args);
};
TelegramClient.prototype.answerInlineQuery = function(...args) {
  return answerInlineQuery.answerInlineQuery(this._client, ...args);
};
TelegramClient.prototype.answerPreCheckoutQuery = function(...args) {
  return answerPreCheckoutQuery.answerPreCheckoutQuery(this._client, ...args);
};
TelegramClient.prototype.deleteMyCommands = function(...args) {
  return deleteMyCommands.deleteMyCommands(this._client, ...args);
};
TelegramClient.prototype.getBotInfo = function(...args) {
  return getBotInfo.getBotInfo(this._client, ...args);
};
TelegramClient.prototype.getBotMenuButton = function(...args) {
  return getBotMenuButton.getBotMenuButton(this._client, ...args);
};
TelegramClient.prototype.getCallbackAnswer = function(...args) {
  return getCallbackAnswer.getCallbackAnswer(this._client, ...args);
};
TelegramClient.prototype.getGameHighScores = function(...args) {
  return getGameHighScores.getGameHighScores(this._client, ...args);
};
TelegramClient.prototype.getInlineGameHighScores = function(...args) {
  return getGameHighScores.getInlineGameHighScores(this._client, ...args);
};
TelegramClient.prototype.getMyCommands = function(...args) {
  return getMyCommands.getMyCommands(this._client, ...args);
};
TelegramClient.prototype.openWebview = function(...args) {
  return openWebview.openWebview(this._client, ...args);
};
TelegramClient.prototype.closeWebview = function(...args) {
  return openWebview.closeWebview(this._client, ...args);
};
TelegramClient.prototype.prepareInlineMessage = function(...args) {
  return prepareInlineMessage.prepareInlineMessage(this._client, ...args);
};
TelegramClient.prototype.setBotInfo = function(...args) {
  return setBotInfo.setBotInfo(this._client, ...args);
};
TelegramClient.prototype.setBotMenuButton = function(...args) {
  return setBotMenuButton.setBotMenuButton(this._client, ...args);
};
TelegramClient.prototype.setGameScore = function(...args) {
  return setGameScore.setGameScore(this._client, ...args);
};
TelegramClient.prototype.setInlineGameScore = function(...args) {
  return setGameScore.setInlineGameScore(this._client, ...args);
};
TelegramClient.prototype.setMyCommands = function(...args) {
  return setMyCommands.setMyCommands(this._client, ...args);
};
TelegramClient.prototype.setMyDefaultRights = function(...args) {
  return setMyDefaultRights.setMyDefaultRights(this._client, ...args);
};
TelegramClient.prototype.toggleEmojiStatusPermission = function(...args) {
  return toggleEmojiStatusPermission.toggleEmojiStatusPermission(this._client, ...args);
};
TelegramClient.prototype.addChatMembers = function(...args) {
  return addChatMembers.addChatMembers(this._client, ...args);
};
TelegramClient.prototype.archiveChats = function(...args) {
  return archiveChats.archiveChats(this._client, ...args);
};
TelegramClient.prototype.banChatMember = function(...args) {
  return banChatMember.banChatMember(this._client, ...args);
};
TelegramClient.prototype.createChannel = function(...args) {
  return createChannel.createChannel(this._client, ...args);
};
TelegramClient.prototype.createGroup = function(...args) {
  return createGroup.createGroup(this._client, ...args);
};
TelegramClient.prototype.createSupergroup = function(...args) {
  return createSupergroup.createSupergroup(this._client, ...args);
};
TelegramClient.prototype.deleteChannel = function(...args) {
  return deleteChannel.deleteChannel(this._client, ...args);
};
TelegramClient.prototype.deleteSupergroup = function(...args) {
  return deleteChannel.deleteChannel(this._client, ...args);
};
TelegramClient.prototype.deleteChatPhoto = function(...args) {
  return deleteChatPhoto.deleteChatPhoto(this._client, ...args);
};
TelegramClient.prototype.deleteGroup = function(...args) {
  return deleteGroup.deleteGroup(this._client, ...args);
};
TelegramClient.prototype.deleteHistory = function(...args) {
  return deleteHistory.deleteHistory(this._client, ...args);
};
TelegramClient.prototype.deleteUserHistory = function(...args) {
  return deleteUserHistory.deleteUserHistory(this._client, ...args);
};
TelegramClient.prototype.editAdminRights = function(...args) {
  return editAdminRights.editAdminRights(this._client, ...args);
};
TelegramClient.prototype.editChatMemberRank = function(...args) {
  return editChatMemberRank.editChatMemberRank(this._client, ...args);
};
TelegramClient.prototype.getChatEventLog = function(...args) {
  return getChatEventLog.getChatEventLog(this._client, ...args);
};
TelegramClient.prototype.getChatMember = function(...args) {
  return getChatMember.getChatMember(this._client, ...args);
};
TelegramClient.prototype.getChatMembers = function(...args) {
  return getChatMembers.getChatMembers(this._client, ...args);
};
TelegramClient.prototype.getChatPreview = function(...args) {
  return getChatPreview.getChatPreview(this._client, ...args);
};
TelegramClient.prototype.getChat = function(...args) {
  return getChat.getChat(this._client, ...args);
};
TelegramClient.prototype.getChats = function(...args) {
  return getChats.getChats(this._client, ...args);
};
TelegramClient.prototype.getCreatorAfterLeave = function(...args) {
  return getCreatorAfterLeave.getCreatorAfterLeave(this._client, ...args);
};
TelegramClient.prototype.getFullChat = function(...args) {
  return getFullChat.getFullChat(this._client, ...args);
};
TelegramClient.prototype.getFullUser = function(...args) {
  return getFullUser.getFullUser(this._client, ...args);
};
TelegramClient.prototype.getMessageAuthor = function(...args) {
  return getMessageAuthor.getMessageAuthor(this._client, ...args);
};
TelegramClient.prototype.getPeer = function(...args) {
  return getPeer.getPeer(this._client, ...args);
};
TelegramClient.prototype.getPeers = function(...args) {
  return getPeers.getPeers(this._client, ...args);
};
TelegramClient.prototype.getSimilarChannels = function(...args) {
  return getSimilarChannels.getSimilarChannels(this._client, ...args);
};
TelegramClient.prototype.getUser = function(...args) {
  return getUser.getUser(this._client, ...args);
};
TelegramClient.prototype.iterChatEventLog = function(...args) {
  return iterChatEventLog.iterChatEventLog(this._client, ...args);
};
TelegramClient.prototype.iterChatMembers = function(...args) {
  return iterChatMembers.iterChatMembers(this._client, ...args);
};
TelegramClient.prototype.joinChat = function(...args) {
  return joinChat.joinChat(this._client, ...args);
};
TelegramClient.prototype.kickChatMember = function(...args) {
  return kickChatMember.kickChatMember(this._client, ...args);
};
TelegramClient.prototype.leaveChat = function(...args) {
  return leaveChat.leaveChat(this._client, ...args);
};
TelegramClient.prototype.markChatUnread = function(...args) {
  return markChatUnread.markChatUnread(this._client, ...args);
};
TelegramClient.prototype.openChat = function(...args) {
  return openChat.openChat(this._client, ...args);
};
TelegramClient.prototype.closeChat = function(...args) {
  return openChat.closeChat(this._client, ...args);
};
TelegramClient.prototype.reorderUsernames = function(...args) {
  return reorderUsernames.reorderUsernames(this._client, ...args);
};
TelegramClient.prototype.restrictChatMember = function(...args) {
  return restrictChatMember.restrictChatMember(this._client, ...args);
};
TelegramClient.prototype.saveDraft = function(...args) {
  return saveDraft.saveDraft(this._client, ...args);
};
TelegramClient.prototype.setChatColor = function(...args) {
  return setChatColor.setChatColor(this._client, ...args);
};
TelegramClient.prototype.setChatDefaultPermissions = function(...args) {
  return setChatDefaultPermissions.setChatDefaultPermissions(this._client, ...args);
};
TelegramClient.prototype.setChatDescription = function(...args) {
  return setChatDescription.setChatDescription(this._client, ...args);
};
TelegramClient.prototype.setChatPhoto = function(...args) {
  return setChatPhoto.setChatPhoto(this._client, ...args);
};
TelegramClient.prototype.setChatTitle = function(...args) {
  return setChatTitle.setChatTitle(this._client, ...args);
};
TelegramClient.prototype.setChatTtl = function(...args) {
  return setChatTtl.setChatTtl(this._client, ...args);
};
TelegramClient.prototype.setChatUsername = function(...args) {
  return setChatUsername.setChatUsername(this._client, ...args);
};
TelegramClient.prototype.setSlowMode = function(...args) {
  return setSlowMode.setSlowMode(this._client, ...args);
};
TelegramClient.prototype.toggleContentProtection = function(...args) {
  return toggleContentProtection.toggleContentProtection(this._client, ...args);
};
TelegramClient.prototype.toggleFragmentUsername = function(...args) {
  return toggleFragmentUsername.toggleFragmentUsername(this._client, ...args);
};
TelegramClient.prototype.toggleJoinRequests = function(...args) {
  return toggleJoinRequests.toggleJoinRequests(this._client, ...args);
};
TelegramClient.prototype.toggleJoinToSend = function(...args) {
  return toggleJoinToSend.toggleJoinToSend(this._client, ...args);
};
TelegramClient.prototype.transferChatOwnership = function(...args) {
  return transferChatOwnership.transferChatOwnership(this._client, ...args);
};
TelegramClient.prototype.unarchiveChats = function(...args) {
  return unarchiveChats.unarchiveChats(this._client, ...args);
};
TelegramClient.prototype.unbanChatMember = function(...args) {
  return unbanChatMember.unbanChatMember(this._client, ...args);
};
TelegramClient.prototype.unrestrictChatMember = function(...args) {
  return unbanChatMember.unbanChatMember(this._client, ...args);
};
TelegramClient.prototype.addContact = function(...args) {
  return addContact.addContact(this._client, ...args);
};
TelegramClient.prototype.deleteContacts = function(...args) {
  return deleteContacts.deleteContacts(this._client, ...args);
};
TelegramClient.prototype.getContacts = function(...args) {
  return getContacts.getContacts(this._client, ...args);
};
TelegramClient.prototype.importContacts = function(...args) {
  return importContacts.importContacts(this._client, ...args);
};
TelegramClient.prototype.setContactNote = function(...args) {
  return setContactNote.setContactNote(this._client, ...args);
};
TelegramClient.prototype.createFolder = function(...args) {
  return createFolder.createFolder(this._client, ...args);
};
TelegramClient.prototype.deleteFolder = function(...args) {
  return deleteFolder.deleteFolder(this._client, ...args);
};
TelegramClient.prototype.editFolder = function(...args) {
  return editFolder.editFolder(this._client, ...args);
};
TelegramClient.prototype.findDialogs = function(...args) {
  return findDialogs.findDialogs(this._client, ...args);
};
TelegramClient.prototype.findFolder = function(...args) {
  return findFolder.findFolder(this._client, ...args);
};
TelegramClient.prototype.getChatlistPreview = function(...args) {
  return getChatlistPreview.getChatlistPreview(this._client, ...args);
};
TelegramClient.prototype.getFolders = function(...args) {
  return getFolders.getFolders(this._client, ...args);
};
TelegramClient.prototype.getPeerDialogs = function(...args) {
  return getPeerDialogs.getPeerDialogs(this._client, ...args);
};
TelegramClient.prototype.iterDialogs = function(...args) {
  return iterDialogs.iterDialogs(this._client, ...args);
};
TelegramClient.prototype.joinChatlist = function(...args) {
  return joinChatlist.joinChatlist(this._client, ...args);
};
TelegramClient.prototype.setFoldersOrder = function(...args) {
  return setFoldersOrder.setFoldersOrder(this._client, ...args);
};
TelegramClient.prototype.downloadAsBuffer = function(...args) {
  return downloadBuffer.downloadAsBuffer(this._client, ...args);
};
TelegramClient.prototype.downloadChunk = function(...args) {
  return downloadChunk.downloadChunk(this._client, ...args);
};
TelegramClient.prototype._normalizeFileDownloadLocation = function(...args) {
  return downloadIterable._normalizeFileDownloadLocation(this._client, ...args);
};
TelegramClient.prototype.downloadAsIterable = function(...args) {
  return downloadIterable.downloadAsIterable(this._client, ...args);
};
TelegramClient.prototype.downloadAsStream = function(...args) {
  return downloadStream.downloadAsStream(this._client, ...args);
};
TelegramClient.prototype._normalizeInputFile = function(...args) {
  return normalizeInputFile._normalizeInputFile(this._client, ...args);
};
TelegramClient.prototype._normalizeInputMedia = function(...args) {
  return normalizeInputMedia._normalizeInputMedia(this._client, ...args);
};
TelegramClient.prototype.uploadFile = function(...args) {
  return uploadFile.uploadFile(this._client, ...args);
};
TelegramClient.prototype.uploadMedia = function(...args) {
  return uploadMedia.uploadMedia(this._client, ...args);
};
TelegramClient.prototype.createForumTopic = function(...args) {
  return createForumTopic.createForumTopic(this._client, ...args);
};
TelegramClient.prototype.deleteForumTopicHistory = function(...args) {
  return deleteForumTopicHistory.deleteForumTopicHistory(this._client, ...args);
};
TelegramClient.prototype.editForumTopic = function(...args) {
  return editForumTopic.editForumTopic(this._client, ...args);
};
TelegramClient.prototype.getForumTopicsById = function(...args) {
  return getForumTopicsById.getForumTopicsById(this._client, ...args);
};
TelegramClient.prototype.getForumTopics = function(...args) {
  return getForumTopics.getForumTopics(this._client, ...args);
};
TelegramClient.prototype.iterForumTopics = function(...args) {
  return iterForumTopics.iterForumTopics(this._client, ...args);
};
TelegramClient.prototype.reorderPinnedForumTopics = function(...args) {
  return reorderPinnedForumTopics.reorderPinnedForumTopics(this._client, ...args);
};
TelegramClient.prototype.toggleForumTopicClosed = function(...args) {
  return toggleForumTopicClosed.toggleForumTopicClosed(this._client, ...args);
};
TelegramClient.prototype.toggleForumTopicPinned = function(...args) {
  return toggleForumTopicPinned.toggleForumTopicPinned(this._client, ...args);
};
TelegramClient.prototype.updateForumSettings = function(...args) {
  return toggleForum.updateForumSettings(this._client, ...args);
};
TelegramClient.prototype.toggleGeneralTopicHidden = function(...args) {
  return toggleGeneralTopicHidden.toggleGeneralTopicHidden(this._client, ...args);
};
TelegramClient.prototype._normalizeInputStarGift = function(...args) {
  return _normalizeInputStarGift._normalizeInputStarGift(this._client, ...args);
};
TelegramClient.prototype.acceptStarGift = function(...args) {
  return acceptStarGift.acceptStarGift(this._client, ...args);
};
TelegramClient.prototype.buyResaleGift = function(...args) {
  return buyResaleGift.buyResaleGift(this._client, ...args);
};
TelegramClient.prototype.getStarGiftResaleOptions = function(...args) {
  return getResaleStarGifts.getStarGiftResaleOptions(this._client, ...args);
};
TelegramClient.prototype.getSavedStarGiftsById = function(...args) {
  return getSavedStarGiftsById.getSavedStarGiftsById(this._client, ...args);
};
TelegramClient.prototype.getSavedStarGifts = function(...args) {
  return getSavedStarGifts.getSavedStarGifts(this._client, ...args);
};
TelegramClient.prototype.getStarGiftOptions = function(...args) {
  return getStarGiftOptions.getStarGiftOptions(this._client, ...args);
};
TelegramClient.prototype.getStarGiftUpgradeOptions = function(...args) {
  return getStarGiftUpgradeOptions.getStarGiftUpgradeOptions(this._client, ...args);
};
TelegramClient.prototype.getStarGiftValue = function(...args) {
  return getStarGiftValue.getStarGiftValue(this._client, ...args);
};
TelegramClient.prototype.getStarGiftWithdrawalUrl = function(...args) {
  return getStarGiftWithdrawalUrl.getStarGiftWithdrawalUrl(this._client, ...args);
};
TelegramClient.prototype.getUniqueStarGift = function(...args) {
  return getUniqueStarGift.getUniqueStarGift(this._client, ...args);
};
TelegramClient.prototype.iterSavedStarGifts = function(...args) {
  return iterSavedStarGifts.iterSavedStarGifts(this._client, ...args);
};
TelegramClient.prototype.togglePinnedStarGifts = function(...args) {
  return pinStarGift.togglePinnedStarGifts(this._client, ...args);
};
TelegramClient.prototype.prepayStarGiftUpgrade = function(...args) {
  return prepayStarGiftUpgrade.prepayStarGiftUpgrade(this._client, ...args);
};
TelegramClient.prototype.resolveStarGiftOffer = function(...args) {
  return resolveStarGiftOffer.resolveStarGiftOffer(this._client, ...args);
};
TelegramClient.prototype.sendStarGiftOffer = function(...args) {
  return sendStarGiftOffer.sendStarGiftOffer(this._client, ...args);
};
TelegramClient.prototype.sendStarGift = function(...args) {
  return sendStarGift.sendStarGift(this._client, ...args);
};
TelegramClient.prototype.setResaleStarGiftPrice = function(...args) {
  return setGiftResalePrice.setResaleStarGiftPrice(this._client, ...args);
};
TelegramClient.prototype.transferStarGift = function(...args) {
  return transferStarGift.transferStarGift(this._client, ...args);
};
TelegramClient.prototype.upgradeStarGift = function(...args) {
  return upgradeStarGift.upgradeStarGift(this._client, ...args);
};
TelegramClient.prototype.createInviteLink = function(...args) {
  return createInviteLink.createInviteLink(this._client, ...args);
};
TelegramClient.prototype.editInviteLink = function(...args) {
  return editInviteLink.editInviteLink(this._client, ...args);
};
TelegramClient.prototype.exportInviteLink = function(...args) {
  return exportInviteLink.exportInviteLink(this._client, ...args);
};
TelegramClient.prototype.getInviteLinkMembers = function(...args) {
  return getInviteLinkMembers.getInviteLinkMembers(this._client, ...args);
};
TelegramClient.prototype.getInviteLink = function(...args) {
  return getInviteLink.getInviteLink(this._client, ...args);
};
TelegramClient.prototype.getInviteLinks = function(...args) {
  return getInviteLinks.getInviteLinks(this._client, ...args);
};
TelegramClient.prototype.getPrimaryInviteLink = function(...args) {
  return getPrimaryInviteLink.getPrimaryInviteLink(this._client, ...args);
};
TelegramClient.prototype.hideAllJoinRequests = function(...args) {
  return hideAllJoinRequests.hideAllJoinRequests(this._client, ...args);
};
TelegramClient.prototype.hideJoinRequest = function(...args) {
  return hideJoinRequest.hideJoinRequest(this._client, ...args);
};
TelegramClient.prototype.iterInviteLinkMembers = function(...args) {
  return iterInviteLinkMembers.iterInviteLinkMembers(this._client, ...args);
};
TelegramClient.prototype.iterInviteLinks = function(...args) {
  return iterInviteLinks.iterInviteLinks(this._client, ...args);
};
TelegramClient.prototype.revokeInviteLink = function(...args) {
  return revokeInviteLink.revokeInviteLink(this._client, ...args);
};
TelegramClient.prototype.appendTodoList = function(...args) {
  return appendTodoList.appendTodoList(this._client, ...args);
};
TelegramClient.prototype.closePoll = function(...args) {
  return closePoll.closePoll(this._client, ...args);
};
TelegramClient.prototype.createRichStreamingDraft = function(...args) {
  return createRichStreamingDraft.createRichStreamingDraft(this._client, ...args);
};
TelegramClient.prototype.createStreamingDraft = function(...args) {
  return createStreamingDraft.createStreamingDraft(this._client, ...args);
};
TelegramClient.prototype.deleteMessagesById = function(...args) {
  return deleteMessages.deleteMessagesById(this._client, ...args);
};
TelegramClient.prototype.deleteMessages = function(...args) {
  return deleteMessages.deleteMessages(this._client, ...args);
};
TelegramClient.prototype.deleteScheduledMessages = function(...args) {
  return deleteScheduledMessages.deleteScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.editInlineMessage = function(...args) {
  return editInlineMessage.editInlineMessage(this._client, ...args);
};
TelegramClient.prototype.editMessage = function(...args) {
  return editMessage.editMessage(this._client, ...args);
};
TelegramClient.prototype.forwardMessagesById = function(...args) {
  return forwardMessages.forwardMessagesById(this._client, ...args);
};
TelegramClient.prototype.forwardMessages = function(...args) {
  return forwardMessages.forwardMessages(this._client, ...args);
};
TelegramClient.prototype.getAllScheduledMessages = function(...args) {
  return getAllScheduledMessages.getAllScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.getAvailableMessageEffects = function(...args) {
  return getAvailableEffects.getAvailableMessageEffects(this._client, ...args);
};
TelegramClient.prototype.getCallbackQueryMessage = function(...args) {
  return getCallbackQueryMessage.getCallbackQueryMessage(this._client, ...args);
};
TelegramClient.prototype.getDiscussionMessage = function(...args) {
  return getDiscussionMessage.getDiscussionMessage(this._client, ...args);
};
TelegramClient.prototype.getFactCheck = function(...args) {
  return getFactCheck.getFactCheck(this._client, ...args);
};
TelegramClient.prototype.getHistory = function(...args) {
  return getHistory.getHistory(this._client, ...args);
};
TelegramClient.prototype.getMessageByLink = function(...args) {
  return getMessageByLink.getMessageByLink(this._client, ...args);
};
TelegramClient.prototype.getMessageGroup = function(...args) {
  return getMessageGroup.getMessageGroup(this._client, ...args);
};
TelegramClient.prototype.getMessageReactionsById = function(...args) {
  return getMessageReactions.getMessageReactionsById(this._client, ...args);
};
TelegramClient.prototype.getMessageReactions = function(...args) {
  return getMessageReactions.getMessageReactions(this._client, ...args);
};
TelegramClient.prototype.getMessagesUnsafe = function(...args) {
  return getMessagesUnsafe.getMessagesUnsafe(this._client, ...args);
};
TelegramClient.prototype.getMessages = function(...args) {
  return getMessages.getMessages(this._client, ...args);
};
TelegramClient.prototype.getReactionUsers = function(...args) {
  return getReactionUsers.getReactionUsers(this._client, ...args);
};
TelegramClient.prototype.getReplyTo = function(...args) {
  return getReplyTo.getReplyTo(this._client, ...args);
};
TelegramClient.prototype.getScheduledMessages = function(...args) {
  return getScheduledMessages.getScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.getWebPagePreview = function(...args) {
  return getWebPagePreview.getWebPagePreview(this._client, ...args);
};
TelegramClient.prototype.iterHistory = function(...args) {
  return iterHistory.iterHistory(this._client, ...args);
};
TelegramClient.prototype.iterReactionUsers = function(...args) {
  return iterReactionUsers.iterReactionUsers(this._client, ...args);
};
TelegramClient.prototype.iterSearchGlobal = function(...args) {
  return iterSearchGlobal.iterSearchGlobal(this._client, ...args);
};
TelegramClient.prototype.iterSearchMessages = function(...args) {
  return iterSearchMessages.iterSearchMessages(this._client, ...args);
};
TelegramClient.prototype.pinMessage = function(...args) {
  return pinMessage.pinMessage(this._client, ...args);
};
TelegramClient.prototype.readHistory = function(...args) {
  return readHistory.readHistory(this._client, ...args);
};
TelegramClient.prototype.readReactions = function(...args) {
  return readReactions.readReactions(this._client, ...args);
};
TelegramClient.prototype.searchGlobal = function(...args) {
  return searchGlobal.searchGlobal(this._client, ...args);
};
TelegramClient.prototype.searchHashtag = function(...args) {
  return searchHashtag.searchHashtag(this._client, ...args);
};
TelegramClient.prototype.iterSearchHashtag = function(...args) {
  return searchHashtag.iterSearchHashtag(this._client, ...args);
};
TelegramClient.prototype.searchMessages = function(...args) {
  return searchMessages.searchMessages(this._client, ...args);
};
TelegramClient.prototype.searchPostsGlobal = function(...args) {
  return searchPostsGlobal.searchPostsGlobal(this._client, ...args);
};
TelegramClient.prototype.answerText = function(...args) {
  return sendAnswer.answerText(this._client, ...args);
};
TelegramClient.prototype.answerMedia = function(...args) {
  return sendAnswer.answerMedia(this._client, ...args);
};
TelegramClient.prototype.answerMediaGroup = function(...args) {
  return sendAnswer.answerMediaGroup(this._client, ...args);
};
TelegramClient.prototype.commentText = function(...args) {
  return sendComment.commentText(this._client, ...args);
};
TelegramClient.prototype.commentMedia = function(...args) {
  return sendComment.commentMedia(this._client, ...args);
};
TelegramClient.prototype.commentMediaGroup = function(...args) {
  return sendComment.commentMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendCopyGroup = function(...args) {
  return sendCopyGroup.sendCopyGroup(this._client, ...args);
};
TelegramClient.prototype.sendCopy = function(...args) {
  return sendCopy.sendCopy(this._client, ...args);
};
TelegramClient.prototype.sendMediaGroup = function(...args) {
  return sendMediaGroup.sendMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendMedia = function(...args) {
  return sendMedia.sendMedia(this._client, ...args);
};
TelegramClient.prototype.sendPaidReaction = function(...args) {
  return sendPaidReaction.sendPaidReaction(this._client, ...args);
};
TelegramClient.prototype.quoteWithText = function(...args) {
  return sendQuote.quoteWithText(this._client, ...args);
};
TelegramClient.prototype.quoteWithMedia = function(...args) {
  return sendQuote.quoteWithMedia(this._client, ...args);
};
TelegramClient.prototype.quoteWithMediaGroup = function(...args) {
  return sendQuote.quoteWithMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendReaction = function(...args) {
  return sendReaction.sendReaction(this._client, ...args);
};
TelegramClient.prototype.replyText = function(...args) {
  return sendReply.replyText(this._client, ...args);
};
TelegramClient.prototype.replyMedia = function(...args) {
  return sendReply.replyMedia(this._client, ...args);
};
TelegramClient.prototype.replyMediaGroup = function(...args) {
  return sendReply.replyMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendRichMessage = function(...args) {
  return sendRichMessage.sendRichMessage(this._client, ...args);
};
TelegramClient.prototype.sendScheduled = function(...args) {
  return sendScheduled.sendScheduled(this._client, ...args);
};
TelegramClient.prototype.sendText = function(...args) {
  return sendText.sendText(this._client, ...args);
};
TelegramClient.prototype.sendTyping = function(...args) {
  return sendTyping.sendTyping(this._client, ...args);
};
TelegramClient.prototype.sendVote = function(...args) {
  return sendVote.sendVote(this._client, ...args);
};
TelegramClient.prototype.setTyping = function(...args) {
  return setTyping.setTyping(this._client, ...args);
};
TelegramClient.prototype.toggleTodoCompleted = function(...args) {
  return toggleTodoCompleted.toggleTodoCompleted(this._client, ...args);
};
TelegramClient.prototype.translateMessage = function(...args) {
  return translateMessage.translateMessage(this._client, ...args);
};
TelegramClient.prototype.translateText = function(...args) {
  return translateText.translateText(this._client, ...args);
};
TelegramClient.prototype.unpinAllMessages = function(...args) {
  return unpinAllMessages.unpinAllMessages(this._client, ...args);
};
TelegramClient.prototype.unpinMessage = function(...args) {
  return unpinMessage.unpinMessage(this._client, ...args);
};
TelegramClient.prototype.getCollectibleInfo = function(...args) {
  return getCollectibleInfo.getCollectibleInfo(this._client, ...args);
};
TelegramClient.prototype.initTakeoutSession = function(...args) {
  return initTakeoutSession.initTakeoutSession(this._client, ...args);
};
TelegramClient.prototype._normalizePrivacyRules = function(...args) {
  return normalizePrivacyRules._normalizePrivacyRules(this._client, ...args);
};
TelegramClient.prototype.changeCloudPassword = function(...args) {
  return changeCloudPassword.changeCloudPassword(this._client, ...args);
};
TelegramClient.prototype.enableCloudPassword = function(...args) {
  return enableCloudPassword.enableCloudPassword(this._client, ...args);
};
TelegramClient.prototype.verifyPasswordEmail = function(...args) {
  return passwordEmail.verifyPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.resendPasswordEmail = function(...args) {
  return passwordEmail.resendPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.cancelPasswordEmail = function(...args) {
  return passwordEmail.cancelPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.removeCloudPassword = function(...args) {
  return removeCloudPassword.removeCloudPassword(this._client, ...args);
};
TelegramClient.prototype.applyBoost = function(...args) {
  return applyBoost.applyBoost(this._client, ...args);
};
TelegramClient.prototype.canApplyBoost = function(...args) {
  return canApplyBoost.canApplyBoost(this._client, ...args);
};
TelegramClient.prototype.createBusinessChatLink = function(...args) {
  return createBusinessChatLink.createBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.editBusinessChatLink = function(...args) {
  return editBusinessChatLink.editBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.deleteBusinessChatLink = function(...args) {
  return editBusinessChatLink.deleteBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.getBoostStats = function(...args) {
  return getBoostStats.getBoostStats(this._client, ...args);
};
TelegramClient.prototype.getBoosts = function(...args) {
  return getBoosts.getBoosts(this._client, ...args);
};
TelegramClient.prototype.getBusinessChatLinks = function(...args) {
  return getBusinessChatLinks.getBusinessChatLinks(this._client, ...args);
};
TelegramClient.prototype.getBusinessConnection = function(...args) {
  return getBusinessConnection.getBusinessConnection(this._client, ...args);
};
TelegramClient.prototype.getMyBoostSlots = function(...args) {
  return getMyBoostSlots.getMyBoostSlots(this._client, ...args);
};
TelegramClient.prototype.getStarsTransactions = function(...args) {
  return getStarsTransactions.getStarsTransactions(this._client, ...args);
};
TelegramClient.prototype.iterBoosters = function(...args) {
  return iterBoosters.iterBoosters(this._client, ...args);
};
TelegramClient.prototype.iterStarsTransactions = function(...args) {
  return iterStarsTransactions.iterStarsTransactions(this._client, ...args);
};
TelegramClient.prototype.setBusinessIntro = function(...args) {
  return setBusinessIntro.setBusinessIntro(this._client, ...args);
};
TelegramClient.prototype.setBusinessWorkHours = function(...args) {
  return setBusinessWorkHours.setBusinessWorkHours(this._client, ...args);
};
TelegramClient.prototype.addStickerToSet = function(...args) {
  return addStickerToSet.addStickerToSet(this._client, ...args);
};
TelegramClient.prototype.createStickerSet = function(...args) {
  return createStickerSet.createStickerSet(this._client, ...args);
};
TelegramClient.prototype.deleteStickerFromSet = function(...args) {
  return deleteStickerFromSet.deleteStickerFromSet(this._client, ...args);
};
TelegramClient.prototype.getCustomEmojis = function(...args) {
  return getCustomEmojis.getCustomEmojis(this._client, ...args);
};
TelegramClient.prototype.getCustomEmojisFromMessages = function(...args) {
  return getCustomEmojis.getCustomEmojisFromMessages(this._client, ...args);
};
TelegramClient.prototype.getInstalledStickers = function(...args) {
  return getInstalledStickers.getInstalledStickers(this._client, ...args);
};
TelegramClient.prototype.getMyStickerSets = function(...args) {
  return getMyStickerSets.getMyStickerSets(this._client, ...args);
};
TelegramClient.prototype.getStickerSet = function(...args) {
  return getStickerSet.getStickerSet(this._client, ...args);
};
TelegramClient.prototype.moveStickerInSet = function(...args) {
  return moveStickerInSet.moveStickerInSet(this._client, ...args);
};
TelegramClient.prototype.replaceStickerInSet = function(...args) {
  return replaceStickerInSet.replaceStickerInSet(this._client, ...args);
};
TelegramClient.prototype.setChatStickerSet = function(...args) {
  return setChatStickerSet.setChatStickerSet(this._client, ...args);
};
TelegramClient.prototype.setStickerSetThumb = function(...args) {
  return setStickerSetThumb.setStickerSetThumb(this._client, ...args);
};
TelegramClient.prototype.canSendStory = function(...args) {
  return canSendStory.canSendStory(this._client, ...args);
};
TelegramClient.prototype.deleteStories = function(...args) {
  return deleteStories.deleteStories(this._client, ...args);
};
TelegramClient.prototype.editStory = function(...args) {
  return editStory.editStory(this._client, ...args);
};
TelegramClient.prototype.getAllStories = function(...args) {
  return getAllStories.getAllStories(this._client, ...args);
};
TelegramClient.prototype.getPeerStories = function(...args) {
  return getPeerStories.getPeerStories(this._client, ...args);
};
TelegramClient.prototype.getProfileStories = function(...args) {
  return getProfileStories.getProfileStories(this._client, ...args);
};
TelegramClient.prototype.getStoriesById = function(...args) {
  return getStoriesById.getStoriesById(this._client, ...args);
};
TelegramClient.prototype.getStoriesInteractions = function(...args) {
  return getStoriesInteractions.getStoriesInteractions(this._client, ...args);
};
TelegramClient.prototype.getStoryLink = function(...args) {
  return getStoryLink.getStoryLink(this._client, ...args);
};
TelegramClient.prototype.getStoryViewers = function(...args) {
  return getStoryViewers.getStoryViewers(this._client, ...args);
};
TelegramClient.prototype.hideMyStoriesViews = function(...args) {
  return hideMyStoriesViews.hideMyStoriesViews(this._client, ...args);
};
TelegramClient.prototype.incrementStoriesViews = function(...args) {
  return incrementStoriesViews.incrementStoriesViews(this._client, ...args);
};
TelegramClient.prototype.iterAllStories = function(...args) {
  return iterAllStories.iterAllStories(this._client, ...args);
};
TelegramClient.prototype.iterProfileStories = function(...args) {
  return iterProfileStories.iterProfileStories(this._client, ...args);
};
TelegramClient.prototype.iterStoryViewers = function(...args) {
  return iterStoryViewers.iterStoryViewers(this._client, ...args);
};
TelegramClient.prototype.readStories = function(...args) {
  return readStories.readStories(this._client, ...args);
};
TelegramClient.prototype.sendStoryReaction = function(...args) {
  return sendStoryReaction.sendStoryReaction(this._client, ...args);
};
TelegramClient.prototype.sendStory = function(...args) {
  return sendStory.sendStory(this._client, ...args);
};
TelegramClient.prototype.togglePeerStoriesArchived = function(...args) {
  return togglePeerStoriesArchived.togglePeerStoriesArchived(this._client, ...args);
};
TelegramClient.prototype.toggleStoriesPinned = function(...args) {
  return toggleStoriesPinned.toggleStoriesPinned(this._client, ...args);
};
TelegramClient.prototype.blockUser = function(...args) {
  return blockUser.blockUser(this._client, ...args);
};
TelegramClient.prototype.deleteProfilePhotos = function(...args) {
  return deleteProfilePhotos.deleteProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.editCloseFriendsRaw = function(...args) {
  return editCloseFriends.editCloseFriendsRaw(this._client, ...args);
};
TelegramClient.prototype.editCloseFriends = function(...args) {
  return editCloseFriends.editCloseFriends(this._client, ...args);
};
TelegramClient.prototype.getCommonChats = function(...args) {
  return getCommonChats.getCommonChats(this._client, ...args);
};
TelegramClient.prototype.getGlobalTtl = function(...args) {
  return getGlobalTtl.getGlobalTtl(this._client, ...args);
};
TelegramClient.prototype.getMe = function(...args) {
  return getMe.getMe(this._client, ...args);
};
TelegramClient.prototype.getMyUsername = function(...args) {
  return getMyUsername.getMyUsername(this._client, ...args);
};
TelegramClient.prototype.getPeerSettings = function(...args) {
  return getPeerSettings.getPeerSettings(this._client, ...args);
};
TelegramClient.prototype.getProfilePhoto = function(...args) {
  return getProfilePhoto.getProfilePhoto(this._client, ...args);
};
TelegramClient.prototype.getProfilePhotos = function(...args) {
  return getProfilePhotos.getProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.getSavedMusic = function(...args) {
  return getSavedMusic.getSavedMusic(this._client, ...args);
};
TelegramClient.prototype.getUsers = function(...args) {
  return getUsers.getUsers(this._client, ...args);
};
TelegramClient.prototype.isPeerAvailable = function(...args) {
  return isPeerAvailable.isPeerAvailable(this._client, ...args);
};
TelegramClient.prototype.iterProfilePhotos = function(...args) {
  return iterProfilePhotos.iterProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.resolvePeerMany = function(...args) {
  return resolvePeerMany.resolvePeerMany(this._client, ...args);
};
TelegramClient.prototype.resolvePeer = function(...args) {
  return resolvePeer.resolvePeer(this._client, ...args);
};
TelegramClient.prototype.resolveUser = function(...args) {
  return resolvePeer.resolveUser(this._client, ...args);
};
TelegramClient.prototype.resolveChannel = function(...args) {
  return resolvePeer.resolveChannel(this._client, ...args);
};
TelegramClient.prototype.resolvePhoneNumber = function(...args) {
  return resolvePhoneNumber.resolvePhoneNumber(this._client, ...args);
};
TelegramClient.prototype.saveMusicToProfile = function(...args) {
  return saveMusicToProfile.saveMusicToProfile(this._client, ...args);
};
TelegramClient.prototype.unsaveMusicFromProfile = function(...args) {
  return saveMusicToProfile.unsaveMusicFromProfile(this._client, ...args);
};
TelegramClient.prototype.setEmojiStatus = function(...args) {
  return setEmojiStatus.setEmojiStatus(this._client, ...args);
};
TelegramClient.prototype.setGlobalTtl = function(...args) {
  return setGlobalTtl.setGlobalTtl(this._client, ...args);
};
TelegramClient.prototype.setMyBirthday = function(...args) {
  return setMyBirthday.setMyBirthday(this._client, ...args);
};
TelegramClient.prototype.setMyProfilePhoto = function(...args) {
  return setMyProfilePhoto.setMyProfilePhoto(this._client, ...args);
};
TelegramClient.prototype.setMyUsername = function(...args) {
  return setMyUsername.setMyUsername(this._client, ...args);
};
TelegramClient.prototype.sendOnline = function(...args) {
  return setOnline.sendOnline(this._client, ...args);
};
TelegramClient.prototype.setOnline = function(...args) {
  return setOnline.setOnline(this._client, ...args);
};
TelegramClient.prototype.unblockUser = function(...args) {
  return unblockUser.unblockUser(this._client, ...args);
};
TelegramClient.prototype.updateProfile = function(...args) {
  return updateProfile.updateProfile(this._client, ...args);
};
TelegramClient.prototype.prepare = function(...args) {
  return this._client.prepare(...args);
};
TelegramClient.prototype.connect = function(...args) {
  return this._client.connect(...args);
};
TelegramClient.prototype.disconnect = function(...args) {
  return this._client.disconnect(...args);
};
TelegramClient.prototype.destroy = function(...args) {
  return this._client.destroy(...args);
};
TelegramClient.prototype.notifyLoggedIn = function(...args) {
  return this._client.notifyLoggedIn(...args);
};
TelegramClient.prototype.notifyLoggedOut = function(...args) {
  return this._client.notifyLoggedOut(...args);
};
TelegramClient.prototype.notifyChannelOpened = function(...args) {
  return this._client.notifyChannelOpened(...args);
};
TelegramClient.prototype.notifyChannelClosed = function(...args) {
  return this._client.notifyChannelClosed(...args);
};
TelegramClient.prototype.startUpdatesLoop = function(...args) {
  return this._client.startUpdatesLoop(...args);
};
TelegramClient.prototype.stopUpdatesLoop = function(...args) {
  return this._client.stopUpdatesLoop(...args);
};
TelegramClient.prototype.call = function(...args) {
  return this._client.call(...args);
};
TelegramClient.prototype.importSession = function(...args) {
  return this._client.importSession(...args);
};
TelegramClient.prototype.exportSession = function(...args) {
  return this._client.exportSession(...args);
};
TelegramClient.prototype.handleClientUpdate = function(...args) {
  return this._client.handleClientUpdate(...args);
};
TelegramClient.prototype.getApiCredentials = function(...args) {
  return this._client.getApiCredentials(...args);
};
TelegramClient.prototype.getPoolSize = function(...args) {
  return this._client.getPoolSize(...args);
};
TelegramClient.prototype.getPrimaryDcId = function(...args) {
  return this._client.getPrimaryDcId(...args);
};
TelegramClient.prototype.computeSrpParams = function(...args) {
  return this._client.computeSrpParams(...args);
};
TelegramClient.prototype.computeNewPasswordHash = function(...args) {
  return this._client.computeNewPasswordHash(...args);
};
TelegramClient.prototype.changePrimaryDc = function(...args) {
  return this._client.changePrimaryDc(...args);
};
TelegramClient.prototype.getMtprotoMessageId = function(...args) {
  return this._client.getMtprotoMessageId(...args);
};
TelegramClient.prototype.recreateDc = function(...args) {
  return this._client.recreateDc(...args);
};
exports.BaseTelegramClient = base.BaseTelegramClient;
exports.TelegramClient = TelegramClient;
