// @mtcute/core/client.js
import { Emitter } from "@fuman/utils";
import { BaseTelegramClient } from "./highlevel/base.js";
import { checkPassword } from "./highlevel/methods/auth/check-password.js";
import { getPasswordHint } from "./highlevel/methods/auth/get-password-hint.js";
import { logOut } from "./highlevel/methods/auth/log-out.js";
import { recoverPassword } from "./highlevel/methods/auth/recover-password.js";
import { resendCode } from "./highlevel/methods/auth/resend-code.js";
import { sendCode } from "./highlevel/methods/auth/send-code.js";
import { sendRecoveryCode } from "./highlevel/methods/auth/send-recovery-code.js";
import { signInBot } from "./highlevel/methods/auth/sign-in-bot.js";
import { signInQr } from "./highlevel/methods/auth/sign-in-qr.js";
import { signIn } from "./highlevel/methods/auth/sign-in.js";
import { startTest } from "./highlevel/methods/auth/start-test.js";
import { start } from "./highlevel/methods/auth/start.js";
import { isSelfPeer } from "./highlevel/methods/auth/utils.js";
import { answerBotGuestChatQuery } from "./highlevel/methods/bots/answer-bot-guest-chat-query.js";
import { answerCallbackQuery } from "./highlevel/methods/bots/answer-callback-query.js";
import { answerInlineQuery } from "./highlevel/methods/bots/answer-inline-query.js";
import { answerPreCheckoutQuery } from "./highlevel/methods/bots/answer-pre-checkout-query.js";
import { deleteMyCommands } from "./highlevel/methods/bots/delete-my-commands.js";
import { getBotInfo } from "./highlevel/methods/bots/get-bot-info.js";
import { getBotMenuButton } from "./highlevel/methods/bots/get-bot-menu-button.js";
import { getCallbackAnswer } from "./highlevel/methods/bots/get-callback-answer.js";
import { getGameHighScores, getInlineGameHighScores } from "./highlevel/methods/bots/get-game-high-scores.js";
import { getMyCommands } from "./highlevel/methods/bots/get-my-commands.js";
import { openWebview, closeWebview } from "./highlevel/methods/bots/open-webview.js";
import { prepareInlineMessage } from "./highlevel/methods/bots/prepare-inline-message.js";
import { setBotInfo } from "./highlevel/methods/bots/set-bot-info.js";
import { setBotMenuButton } from "./highlevel/methods/bots/set-bot-menu-button.js";
import { setGameScore, setInlineGameScore } from "./highlevel/methods/bots/set-game-score.js";
import { setMyCommands } from "./highlevel/methods/bots/set-my-commands.js";
import { setMyDefaultRights } from "./highlevel/methods/bots/set-my-default-rights.js";
import { toggleEmojiStatusPermission } from "./highlevel/methods/bots/toggle-emoji-status-permission.js";
import { addChatMembers } from "./highlevel/methods/chats/add-chat-members.js";
import { archiveChats } from "./highlevel/methods/chats/archive-chats.js";
import { banChatMember } from "./highlevel/methods/chats/ban-chat-member.js";
import { createChannel } from "./highlevel/methods/chats/create-channel.js";
import { createGroup } from "./highlevel/methods/chats/create-group.js";
import { createSupergroup } from "./highlevel/methods/chats/create-supergroup.js";
import { deleteChannel } from "./highlevel/methods/chats/delete-channel.js";
import { deleteChatPhoto } from "./highlevel/methods/chats/delete-chat-photo.js";
import { deleteGroup } from "./highlevel/methods/chats/delete-group.js";
import { deleteHistory } from "./highlevel/methods/chats/delete-history.js";
import { deleteUserHistory } from "./highlevel/methods/chats/delete-user-history.js";
import { editAdminRights } from "./highlevel/methods/chats/edit-admin-rights.js";
import { editChatMemberRank } from "./highlevel/methods/chats/edit-chat-member-rank.js";
import { getChatEventLog } from "./highlevel/methods/chats/get-chat-event-log.js";
import { getChatMember } from "./highlevel/methods/chats/get-chat-member.js";
import { getChatMembers } from "./highlevel/methods/chats/get-chat-members.js";
import { getChatPreview } from "./highlevel/methods/chats/get-chat-preview.js";
import { getChat } from "./highlevel/methods/chats/get-chat.js";
import { getChats } from "./highlevel/methods/chats/get-chats.js";
import { getCreatorAfterLeave } from "./highlevel/methods/chats/get-creator-after-leave.js";
import { getFullChat } from "./highlevel/methods/chats/get-full-chat.js";
import { getFullUser } from "./highlevel/methods/chats/get-full-user.js";
import { getMessageAuthor } from "./highlevel/methods/chats/get-message-author.js";
import { getPeer } from "./highlevel/methods/chats/get-peer.js";
import { getPeers } from "./highlevel/methods/chats/get-peers.js";
import { getSimilarChannels } from "./highlevel/methods/chats/get-similar-channels.js";
import { getUser } from "./highlevel/methods/chats/get-user.js";
import { iterChatEventLog } from "./highlevel/methods/chats/iter-chat-event-log.js";
import { iterChatMembers } from "./highlevel/methods/chats/iter-chat-members.js";
import { joinChat } from "./highlevel/methods/chats/join-chat.js";
import { kickChatMember } from "./highlevel/methods/chats/kick-chat-member.js";
import { leaveChat } from "./highlevel/methods/chats/leave-chat.js";
import { markChatUnread } from "./highlevel/methods/chats/mark-chat-unread.js";
import { openChat, closeChat } from "./highlevel/methods/chats/open-chat.js";
import { reorderUsernames } from "./highlevel/methods/chats/reorder-usernames.js";
import { restrictChatMember } from "./highlevel/methods/chats/restrict-chat-member.js";
import { saveDraft } from "./highlevel/methods/chats/save-draft.js";
import { setChatColor } from "./highlevel/methods/chats/set-chat-color.js";
import { setChatDefaultPermissions } from "./highlevel/methods/chats/set-chat-default-permissions.js";
import { setChatDescription } from "./highlevel/methods/chats/set-chat-description.js";
import { setChatPhoto } from "./highlevel/methods/chats/set-chat-photo.js";
import { setChatTitle } from "./highlevel/methods/chats/set-chat-title.js";
import { setChatTtl } from "./highlevel/methods/chats/set-chat-ttl.js";
import { setChatUsername } from "./highlevel/methods/chats/set-chat-username.js";
import { setSlowMode } from "./highlevel/methods/chats/set-slow-mode.js";
import { toggleContentProtection } from "./highlevel/methods/chats/toggle-content-protection.js";
import { toggleFragmentUsername } from "./highlevel/methods/chats/toggle-fragment-username.js";
import { toggleJoinRequests } from "./highlevel/methods/chats/toggle-join-requests.js";
import { toggleJoinToSend } from "./highlevel/methods/chats/toggle-join-to-send.js";
import { transferChatOwnership } from "./highlevel/methods/chats/transfer-chat-ownership.js";
import { unarchiveChats } from "./highlevel/methods/chats/unarchive-chats.js";
import { unbanChatMember } from "./highlevel/methods/chats/unban-chat-member.js";
import { addContact } from "./highlevel/methods/contacts/add-contact.js";
import { deleteContacts } from "./highlevel/methods/contacts/delete-contacts.js";
import { getContacts } from "./highlevel/methods/contacts/get-contacts.js";
import { importContacts } from "./highlevel/methods/contacts/import-contacts.js";
import { setContactNote } from "./highlevel/methods/contacts/set-contact-note.js";
import { createFolder } from "./highlevel/methods/dialogs/create-folder.js";
import { deleteFolder } from "./highlevel/methods/dialogs/delete-folder.js";
import { editFolder } from "./highlevel/methods/dialogs/edit-folder.js";
import { findDialogs } from "./highlevel/methods/dialogs/find-dialogs.js";
import { findFolder } from "./highlevel/methods/dialogs/find-folder.js";
import { getChatlistPreview } from "./highlevel/methods/dialogs/get-chatlist-preview.js";
import { getFolders } from "./highlevel/methods/dialogs/get-folders.js";
import { getPeerDialogs } from "./highlevel/methods/dialogs/get-peer-dialogs.js";
import { iterDialogs } from "./highlevel/methods/dialogs/iter-dialogs.js";
import { joinChatlist } from "./highlevel/methods/dialogs/join-chatlist.js";
import { setFoldersOrder } from "./highlevel/methods/dialogs/set-folders-order.js";
import { downloadAsBuffer } from "./highlevel/methods/files/download-buffer.js";
import { downloadChunk } from "./highlevel/methods/files/download-chunk.js";
import { _normalizeFileDownloadLocation, downloadAsIterable } from "./highlevel/methods/files/download-iterable.js";
import { downloadAsStream } from "./highlevel/methods/files/download-stream.js";
import { _normalizeInputFile } from "./highlevel/methods/files/normalize-input-file.js";
import { _normalizeInputMedia } from "./highlevel/methods/files/normalize-input-media.js";
import { uploadFile } from "./highlevel/methods/files/upload-file.js";
import { uploadMedia } from "./highlevel/methods/files/upload-media.js";
import { createForumTopic } from "./highlevel/methods/forums/create-forum-topic.js";
import { deleteForumTopicHistory } from "./highlevel/methods/forums/delete-forum-topic-history.js";
import { editForumTopic } from "./highlevel/methods/forums/edit-forum-topic.js";
import { getForumTopicsById } from "./highlevel/methods/forums/get-forum-topics-by-id.js";
import { getForumTopics } from "./highlevel/methods/forums/get-forum-topics.js";
import { iterForumTopics } from "./highlevel/methods/forums/iter-forum-topics.js";
import { reorderPinnedForumTopics } from "./highlevel/methods/forums/reorder-pinned-forum-topics.js";
import { toggleForumTopicClosed } from "./highlevel/methods/forums/toggle-forum-topic-closed.js";
import { toggleForumTopicPinned } from "./highlevel/methods/forums/toggle-forum-topic-pinned.js";
import { updateForumSettings } from "./highlevel/methods/forums/toggle-forum.js";
import { toggleGeneralTopicHidden } from "./highlevel/methods/forums/toggle-general-topic-hidden.js";
import { _normalizeInputStarGift } from "./highlevel/methods/gifts/_normalize-input-star-gift.js";
import { acceptStarGift } from "./highlevel/methods/gifts/accept-star-gift.js";
import { buyResaleGift } from "./highlevel/methods/gifts/buy-resale-gift.js";
import { getStarGiftResaleOptions } from "./highlevel/methods/gifts/get-resale-star-gifts.js";
import { getSavedStarGiftsById } from "./highlevel/methods/gifts/get-saved-star-gifts-by-id.js";
import { getSavedStarGifts } from "./highlevel/methods/gifts/get-saved-star-gifts.js";
import { getStarGiftOptions } from "./highlevel/methods/gifts/get-star-gift-options.js";
import { getStarGiftUpgradeOptions } from "./highlevel/methods/gifts/get-star-gift-upgrade-options.js";
import { getStarGiftValue } from "./highlevel/methods/gifts/get-star-gift-value.js";
import { getStarGiftWithdrawalUrl } from "./highlevel/methods/gifts/get-star-gift-withdrawal-url.js";
import { getUniqueStarGift } from "./highlevel/methods/gifts/get-unique-star-gift.js";
import { iterSavedStarGifts } from "./highlevel/methods/gifts/iter-saved-star-gifts.js";
import { togglePinnedStarGifts } from "./highlevel/methods/gifts/pin-star-gift.js";
import { prepayStarGiftUpgrade } from "./highlevel/methods/gifts/prepay-star-gift-upgrade.js";
import { resolveStarGiftOffer } from "./highlevel/methods/gifts/resolve-star-gift-offer.js";
import { sendStarGiftOffer } from "./highlevel/methods/gifts/send-star-gift-offer.js";
import { sendStarGift } from "./highlevel/methods/gifts/send-star-gift.js";
import { setResaleStarGiftPrice } from "./highlevel/methods/gifts/set-gift-resale-price.js";
import { transferStarGift } from "./highlevel/methods/gifts/transfer-star-gift.js";
import { upgradeStarGift } from "./highlevel/methods/gifts/upgrade-star-gift.js";
import { createInviteLink } from "./highlevel/methods/invite-links/create-invite-link.js";
import { editInviteLink } from "./highlevel/methods/invite-links/edit-invite-link.js";
import { exportInviteLink } from "./highlevel/methods/invite-links/export-invite-link.js";
import { getInviteLinkMembers } from "./highlevel/methods/invite-links/get-invite-link-members.js";
import { getInviteLink } from "./highlevel/methods/invite-links/get-invite-link.js";
import { getInviteLinks } from "./highlevel/methods/invite-links/get-invite-links.js";
import { getPrimaryInviteLink } from "./highlevel/methods/invite-links/get-primary-invite-link.js";
import { hideAllJoinRequests } from "./highlevel/methods/invite-links/hide-all-join-requests.js";
import { hideJoinRequest } from "./highlevel/methods/invite-links/hide-join-request.js";
import { iterInviteLinkMembers } from "./highlevel/methods/invite-links/iter-invite-link-members.js";
import { iterInviteLinks } from "./highlevel/methods/invite-links/iter-invite-links.js";
import { revokeInviteLink } from "./highlevel/methods/invite-links/revoke-invite-link.js";
import { appendTodoList } from "./highlevel/methods/messages/append-todo-list.js";
import { closePoll } from "./highlevel/methods/messages/close-poll.js";
import { createRichStreamingDraft } from "./highlevel/methods/messages/create-rich-streaming-draft.js";
import { createStreamingDraft } from "./highlevel/methods/messages/create-streaming-draft.js";
import { deleteMessagesById, deleteMessages } from "./highlevel/methods/messages/delete-messages.js";
import { deleteScheduledMessages } from "./highlevel/methods/messages/delete-scheduled-messages.js";
import { editInlineMessage } from "./highlevel/methods/messages/edit-inline-message.js";
import { editMessage } from "./highlevel/methods/messages/edit-message.js";
import { forwardMessagesById, forwardMessages } from "./highlevel/methods/messages/forward-messages.js";
import { getAllScheduledMessages } from "./highlevel/methods/messages/get-all-scheduled-messages.js";
import { getAvailableMessageEffects } from "./highlevel/methods/messages/get-available-effects.js";
import { getCallbackQueryMessage } from "./highlevel/methods/messages/get-callback-query-message.js";
import { getDiscussionMessage } from "./highlevel/methods/messages/get-discussion-message.js";
import { getFactCheck } from "./highlevel/methods/messages/get-fact-check.js";
import { getHistory } from "./highlevel/methods/messages/get-history.js";
import { getMessageByLink } from "./highlevel/methods/messages/get-message-by-link.js";
import { getMessageGroup } from "./highlevel/methods/messages/get-message-group.js";
import { getMessageReactionsById, getMessageReactions } from "./highlevel/methods/messages/get-message-reactions.js";
import { getMessagesUnsafe } from "./highlevel/methods/messages/get-messages-unsafe.js";
import { getMessages } from "./highlevel/methods/messages/get-messages.js";
import { getReactionUsers } from "./highlevel/methods/messages/get-reaction-users.js";
import { getReplyTo } from "./highlevel/methods/messages/get-reply-to.js";
import { getScheduledMessages } from "./highlevel/methods/messages/get-scheduled-messages.js";
import { getWebPagePreview } from "./highlevel/methods/messages/get-web-page-preview.js";
import { iterHistory } from "./highlevel/methods/messages/iter-history.js";
import { iterReactionUsers } from "./highlevel/methods/messages/iter-reaction-users.js";
import { iterSearchGlobal } from "./highlevel/methods/messages/iter-search-global.js";
import { iterSearchMessages } from "./highlevel/methods/messages/iter-search-messages.js";
import { pinMessage } from "./highlevel/methods/messages/pin-message.js";
import { readHistory } from "./highlevel/methods/messages/read-history.js";
import { readReactions } from "./highlevel/methods/messages/read-reactions.js";
import { searchGlobal } from "./highlevel/methods/messages/search-global.js";
import { searchHashtag, iterSearchHashtag } from "./highlevel/methods/messages/search-hashtag.js";
import { searchMessages } from "./highlevel/methods/messages/search-messages.js";
import { searchPostsGlobal } from "./highlevel/methods/messages/search-posts-global.js";
import { answerText, answerMedia, answerMediaGroup } from "./highlevel/methods/messages/send-answer.js";
import { commentText, commentMedia, commentMediaGroup } from "./highlevel/methods/messages/send-comment.js";
import { sendCopyGroup } from "./highlevel/methods/messages/send-copy-group.js";
import { sendCopy } from "./highlevel/methods/messages/send-copy.js";
import { sendMediaGroup } from "./highlevel/methods/messages/send-media-group.js";
import { sendMedia } from "./highlevel/methods/messages/send-media.js";
import { sendPaidReaction } from "./highlevel/methods/messages/send-paid-reaction.js";
import { quoteWithText, quoteWithMedia, quoteWithMediaGroup } from "./highlevel/methods/messages/send-quote.js";
import { sendReaction } from "./highlevel/methods/messages/send-reaction.js";
import { replyText, replyMedia, replyMediaGroup } from "./highlevel/methods/messages/send-reply.js";
import { sendRichMessage } from "./highlevel/methods/messages/send-rich-message.js";
import { sendScheduled } from "./highlevel/methods/messages/send-scheduled.js";
import { sendText } from "./highlevel/methods/messages/send-text.js";
import { sendTyping } from "./highlevel/methods/messages/send-typing.js";
import { sendVote } from "./highlevel/methods/messages/send-vote.js";
import { setTyping } from "./highlevel/methods/messages/set-typing.js";
import { toggleTodoCompleted } from "./highlevel/methods/messages/toggle-todo-completed.js";
import { translateMessage } from "./highlevel/methods/messages/translate-message.js";
import { translateText } from "./highlevel/methods/messages/translate-text.js";
import { unpinAllMessages } from "./highlevel/methods/messages/unpin-all-messages.js";
import { unpinMessage } from "./highlevel/methods/messages/unpin-message.js";
import { getCollectibleInfo } from "./highlevel/methods/misc/get-collectible-info.js";
import { initTakeoutSession } from "./highlevel/methods/misc/init-takeout-session.js";
import { _normalizePrivacyRules } from "./highlevel/methods/misc/normalize-privacy-rules.js";
import { withParams } from "./highlevel/methods/misc/with-params.js";
import { changeCloudPassword } from "./highlevel/methods/password/change-cloud-password.js";
import { enableCloudPassword } from "./highlevel/methods/password/enable-cloud-password.js";
import { verifyPasswordEmail, resendPasswordEmail, cancelPasswordEmail } from "./highlevel/methods/password/password-email.js";
import { removeCloudPassword } from "./highlevel/methods/password/remove-cloud-password.js";
import { applyBoost } from "./highlevel/methods/premium/apply-boost.js";
import { canApplyBoost } from "./highlevel/methods/premium/can-apply-boost.js";
import { createBusinessChatLink } from "./highlevel/methods/premium/create-business-chat-link.js";
import { editBusinessChatLink, deleteBusinessChatLink } from "./highlevel/methods/premium/edit-business-chat-link.js";
import { getBoostStats } from "./highlevel/methods/premium/get-boost-stats.js";
import { getBoosts } from "./highlevel/methods/premium/get-boosts.js";
import { getBusinessChatLinks } from "./highlevel/methods/premium/get-business-chat-links.js";
import { getBusinessConnection } from "./highlevel/methods/premium/get-business-connection.js";
import { getMyBoostSlots } from "./highlevel/methods/premium/get-my-boost-slots.js";
import { getStarsTransactions } from "./highlevel/methods/premium/get-stars-transactions.js";
import { iterBoosters } from "./highlevel/methods/premium/iter-boosters.js";
import { iterStarsTransactions } from "./highlevel/methods/premium/iter-stars-transactions.js";
import { setBusinessIntro } from "./highlevel/methods/premium/set-business-intro.js";
import { setBusinessWorkHours } from "./highlevel/methods/premium/set-business-work-hours.js";
import { addStickerToSet } from "./highlevel/methods/stickers/add-sticker-to-set.js";
import { createStickerSet } from "./highlevel/methods/stickers/create-sticker-set.js";
import { deleteStickerFromSet } from "./highlevel/methods/stickers/delete-sticker-from-set.js";
import { getCustomEmojis, getCustomEmojisFromMessages } from "./highlevel/methods/stickers/get-custom-emojis.js";
import { getInstalledStickers } from "./highlevel/methods/stickers/get-installed-stickers.js";
import { getMyStickerSets } from "./highlevel/methods/stickers/get-my-sticker-sets.js";
import { getStickerSet } from "./highlevel/methods/stickers/get-sticker-set.js";
import { moveStickerInSet } from "./highlevel/methods/stickers/move-sticker-in-set.js";
import { replaceStickerInSet } from "./highlevel/methods/stickers/replace-sticker-in-set.js";
import { setChatStickerSet } from "./highlevel/methods/stickers/set-chat-sticker-set.js";
import { setStickerSetThumb } from "./highlevel/methods/stickers/set-sticker-set-thumb.js";
import { canSendStory } from "./highlevel/methods/stories/can-send-story.js";
import { deleteStories } from "./highlevel/methods/stories/delete-stories.js";
import { editStory } from "./highlevel/methods/stories/edit-story.js";
import { getAllStories } from "./highlevel/methods/stories/get-all-stories.js";
import { getPeerStories } from "./highlevel/methods/stories/get-peer-stories.js";
import { getProfileStories } from "./highlevel/methods/stories/get-profile-stories.js";
import { getStoriesById } from "./highlevel/methods/stories/get-stories-by-id.js";
import { getStoriesInteractions } from "./highlevel/methods/stories/get-stories-interactions.js";
import { getStoryLink } from "./highlevel/methods/stories/get-story-link.js";
import { getStoryViewers } from "./highlevel/methods/stories/get-story-viewers.js";
import { hideMyStoriesViews } from "./highlevel/methods/stories/hide-my-stories-views.js";
import { incrementStoriesViews } from "./highlevel/methods/stories/increment-stories-views.js";
import { iterAllStories } from "./highlevel/methods/stories/iter-all-stories.js";
import { iterProfileStories } from "./highlevel/methods/stories/iter-profile-stories.js";
import { iterStoryViewers } from "./highlevel/methods/stories/iter-story-viewers.js";
import { readStories } from "./highlevel/methods/stories/read-stories.js";
import { sendStoryReaction } from "./highlevel/methods/stories/send-story-reaction.js";
import { sendStory } from "./highlevel/methods/stories/send-story.js";
import { togglePeerStoriesArchived } from "./highlevel/methods/stories/toggle-peer-stories-archived.js";
import { toggleStoriesPinned } from "./highlevel/methods/stories/toggle-stories-pinned.js";
import { blockUser } from "./highlevel/methods/users/block-user.js";
import { deleteProfilePhotos } from "./highlevel/methods/users/delete-profile-photos.js";
import { editCloseFriendsRaw, editCloseFriends } from "./highlevel/methods/users/edit-close-friends.js";
import { getCommonChats } from "./highlevel/methods/users/get-common-chats.js";
import { getGlobalTtl } from "./highlevel/methods/users/get-global-ttl.js";
import { getMe } from "./highlevel/methods/users/get-me.js";
import { getMyUsername } from "./highlevel/methods/users/get-my-username.js";
import { getPeerSettings } from "./highlevel/methods/users/get-peer-settings.js";
import { getProfilePhoto } from "./highlevel/methods/users/get-profile-photo.js";
import { getProfilePhotos } from "./highlevel/methods/users/get-profile-photos.js";
import { getSavedMusic } from "./highlevel/methods/users/get-saved-music.js";
import { getUsers } from "./highlevel/methods/users/get-users.js";
import { isPeerAvailable } from "./highlevel/methods/users/is-peer-available.js";
import { iterProfilePhotos } from "./highlevel/methods/users/iter-profile-photos.js";
import { resolvePeerMany } from "./highlevel/methods/users/resolve-peer-many.js";
import { resolvePeer, resolveUser, resolveChannel } from "./highlevel/methods/users/resolve-peer.js";
import { resolvePhoneNumber } from "./highlevel/methods/users/resolve-phone-number.js";
import { saveMusicToProfile, unsaveMusicFromProfile } from "./highlevel/methods/users/save-music-to-profile.js";
import { setEmojiStatus } from "./highlevel/methods/users/set-emoji-status.js";
import { setGlobalTtl } from "./highlevel/methods/users/set-global-ttl.js";
import { setMyBirthday } from "./highlevel/methods/users/set-my-birthday.js";
import { setMyProfilePhoto } from "./highlevel/methods/users/set-my-profile-photo.js";
import { setMyUsername } from "./highlevel/methods/users/set-my-username.js";
import { sendOnline, setOnline } from "./highlevel/methods/users/set-online.js";
import { unblockUser } from "./highlevel/methods/users/unblock-user.js";
import { updateProfile } from "./highlevel/methods/users/update-profile.js";
import { Conversation } from "./highlevel/types/conversation.js";
import { makeParsedUpdateHandler } from "./highlevel/updates/parsed.js";
import { MtUnsupportedError } from "./types/errors.js";
class TelegramClient {
  _client;
  constructor(opts) {
    this.onRawUpdate = new Emitter();
    this.onUpdate = new Emitter();
    this.onNewMessage = new Emitter();
    this.onEditMessage = new Emitter();
    this.onMessageGroup = new Emitter();
    this.onDeleteMessage = new Emitter();
    this.onChatMemberUpdate = new Emitter();
    this.onInlineQuery = new Emitter();
    this.onBotGuestChatQuery = new Emitter();
    this.onChosenInlineResult = new Emitter();
    this.onCallbackQuery = new Emitter();
    this.onInlineCallbackQuery = new Emitter();
    this.onBusinessCallbackQuery = new Emitter();
    this.onPollUpdate = new Emitter();
    this.onPollVote = new Emitter();
    this.onUserStatusUpdate = new Emitter();
    this.onUserTyping = new Emitter();
    this.onHistoryRead = new Emitter();
    this.onBotStopped = new Emitter();
    this.onBotChatJoinRequest = new Emitter();
    this.onChatJoinRequest = new Emitter();
    this.onPreCheckoutQuery = new Emitter();
    this.onStoryUpdate = new Emitter();
    this.onDeleteStory = new Emitter();
    this.onBotReactionUpdate = new Emitter();
    this.onBotReactionCountUpdate = new Emitter();
    this.onBusinessConnectionUpdate = new Emitter();
    this.onNewBusinessMessage = new Emitter();
    this.onEditBusinessMessage = new Emitter();
    this.onBusinessMessageGroup = new Emitter();
    this.onDeleteBusinessMessage = new Emitter();
    if ("client" in opts) {
      this._client = opts.client;
    } else {
      if (!opts.storage || typeof opts.storage === "string" || !opts.transport || !opts.crypto || !opts.platform) {
        throw new MtUnsupportedError(
          "You need to explicitly provide storage, transport, crypto and platform for @mtcute/core"
        );
      }
      this._client = new BaseTelegramClient(opts);
    }
    Object.defineProperty(this, "log", { value: this._client.log });
    Object.defineProperty(this, "storage", { value: this._client.storage });
    Object.defineProperty(this, "timers", { value: this._client.timers });
    Object.defineProperty(this, "stopSignal", { value: this._client.stopSignal });
    Object.defineProperty(this, "appConfig", { value: this._client.appConfig });
    Object.defineProperty(this, "platform", { value: this._client.platform });
    Object.defineProperty(this, "onServerUpdate", { value: this._client.onServerUpdate });
    Object.defineProperty(this, "onRawUpdate", { value: this._client.onRawUpdate });
    Object.defineProperty(this, "onConnectionState", { value: this._client.onConnectionState });
    Object.defineProperty(this, "onError", { value: this._client.onError });
    if (!opts.disableUpdates) {
      const skipConversationUpdates = opts.skipConversationUpdates ?? true;
      const { messageGroupingInterval } = opts.updates ?? {};
      this._client.onRawUpdate.add(
        makeParsedUpdateHandler({
          messageGroupingInterval,
          onUpdate: (update) => {
            if (Conversation.handleUpdate(this, update) && skipConversationUpdates) return;
            this.onUpdate.emit(update);
            switch (update.name) {
              case "new_message":
                this.onNewMessage.emit(update.data);
                break;
              case "edit_message":
                this.onEditMessage.emit(update.data);
                break;
              case "message_group":
                this.onMessageGroup.emit(update.data);
                break;
              case "delete_message":
                this.onDeleteMessage.emit(update.data);
                break;
              case "chat_member":
                this.onChatMemberUpdate.emit(update.data);
                break;
              case "inline_query":
                this.onInlineQuery.emit(update.data);
                break;
              case "bot_guest_chat_query":
                this.onBotGuestChatQuery.emit(update.data);
                break;
              case "chosen_inline_result":
                this.onChosenInlineResult.emit(update.data);
                break;
              case "callback_query":
                this.onCallbackQuery.emit(update.data);
                break;
              case "inline_callback_query":
                this.onInlineCallbackQuery.emit(update.data);
                break;
              case "business_callback_query":
                this.onBusinessCallbackQuery.emit(update.data);
                break;
              case "poll":
                this.onPollUpdate.emit(update.data);
                break;
              case "poll_vote":
                this.onPollVote.emit(update.data);
                break;
              case "user_status":
                this.onUserStatusUpdate.emit(update.data);
                break;
              case "user_typing":
                this.onUserTyping.emit(update.data);
                break;
              case "history_read":
                this.onHistoryRead.emit(update.data);
                break;
              case "bot_stopped":
                this.onBotStopped.emit(update.data);
                break;
              case "bot_chat_join_request":
                this.onBotChatJoinRequest.emit(update.data);
                break;
              case "chat_join_request":
                this.onChatJoinRequest.emit(update.data);
                break;
              case "pre_checkout_query":
                this.onPreCheckoutQuery.emit(update.data);
                break;
              case "story":
                this.onStoryUpdate.emit(update.data);
                break;
              case "delete_story":
                this.onDeleteStory.emit(update.data);
                break;
              case "bot_reaction":
                this.onBotReactionUpdate.emit(update.data);
                break;
              case "bot_reaction_count":
                this.onBotReactionCountUpdate.emit(update.data);
                break;
              case "business_connection":
                this.onBusinessConnectionUpdate.emit(update.data);
                break;
              case "new_business_message":
                this.onNewBusinessMessage.emit(update.data);
                break;
              case "edit_business_message":
                this.onEditBusinessMessage.emit(update.data);
                break;
              case "business_message_group":
                this.onBusinessMessageGroup.emit(update.data);
                break;
              case "delete_business_message":
                this.onDeleteBusinessMessage.emit(update.data);
                break;
            }
          }
        })
      );
    }
  }
  withParams(params) {
    return withParams(this, params);
  }
}
TelegramClient.prototype.checkPassword = function(...args) {
  return checkPassword(this._client, ...args);
};
TelegramClient.prototype.getPasswordHint = function(...args) {
  return getPasswordHint(this._client, ...args);
};
TelegramClient.prototype.logOut = function(...args) {
  return logOut(this._client, ...args);
};
TelegramClient.prototype.recoverPassword = function(...args) {
  return recoverPassword(this._client, ...args);
};
TelegramClient.prototype.resendCode = function(...args) {
  return resendCode(this._client, ...args);
};
TelegramClient.prototype.sendCode = function(...args) {
  return sendCode(this._client, ...args);
};
TelegramClient.prototype.sendRecoveryCode = function(...args) {
  return sendRecoveryCode(this._client, ...args);
};
TelegramClient.prototype.signInBot = function(...args) {
  return signInBot(this._client, ...args);
};
TelegramClient.prototype.signInQr = function(...args) {
  return signInQr(this._client, ...args);
};
TelegramClient.prototype.signIn = function(...args) {
  return signIn(this._client, ...args);
};
TelegramClient.prototype.startTest = function(...args) {
  return startTest(this._client, ...args);
};
TelegramClient.prototype.start = function(...args) {
  return start(this._client, ...args);
};
TelegramClient.prototype.isSelfPeer = function(...args) {
  return isSelfPeer(this._client, ...args);
};
TelegramClient.prototype.answerBotGuestChatQuery = function(...args) {
  return answerBotGuestChatQuery(this._client, ...args);
};
TelegramClient.prototype.answerCallbackQuery = function(...args) {
  return answerCallbackQuery(this._client, ...args);
};
TelegramClient.prototype.answerInlineQuery = function(...args) {
  return answerInlineQuery(this._client, ...args);
};
TelegramClient.prototype.answerPreCheckoutQuery = function(...args) {
  return answerPreCheckoutQuery(this._client, ...args);
};
TelegramClient.prototype.deleteMyCommands = function(...args) {
  return deleteMyCommands(this._client, ...args);
};
TelegramClient.prototype.getBotInfo = function(...args) {
  return getBotInfo(this._client, ...args);
};
TelegramClient.prototype.getBotMenuButton = function(...args) {
  return getBotMenuButton(this._client, ...args);
};
TelegramClient.prototype.getCallbackAnswer = function(...args) {
  return getCallbackAnswer(this._client, ...args);
};
TelegramClient.prototype.getGameHighScores = function(...args) {
  return getGameHighScores(this._client, ...args);
};
TelegramClient.prototype.getInlineGameHighScores = function(...args) {
  return getInlineGameHighScores(this._client, ...args);
};
TelegramClient.prototype.getMyCommands = function(...args) {
  return getMyCommands(this._client, ...args);
};
TelegramClient.prototype.openWebview = function(...args) {
  return openWebview(this._client, ...args);
};
TelegramClient.prototype.closeWebview = function(...args) {
  return closeWebview(this._client, ...args);
};
TelegramClient.prototype.prepareInlineMessage = function(...args) {
  return prepareInlineMessage(this._client, ...args);
};
TelegramClient.prototype.setBotInfo = function(...args) {
  return setBotInfo(this._client, ...args);
};
TelegramClient.prototype.setBotMenuButton = function(...args) {
  return setBotMenuButton(this._client, ...args);
};
TelegramClient.prototype.setGameScore = function(...args) {
  return setGameScore(this._client, ...args);
};
TelegramClient.prototype.setInlineGameScore = function(...args) {
  return setInlineGameScore(this._client, ...args);
};
TelegramClient.prototype.setMyCommands = function(...args) {
  return setMyCommands(this._client, ...args);
};
TelegramClient.prototype.setMyDefaultRights = function(...args) {
  return setMyDefaultRights(this._client, ...args);
};
TelegramClient.prototype.toggleEmojiStatusPermission = function(...args) {
  return toggleEmojiStatusPermission(this._client, ...args);
};
TelegramClient.prototype.addChatMembers = function(...args) {
  return addChatMembers(this._client, ...args);
};
TelegramClient.prototype.archiveChats = function(...args) {
  return archiveChats(this._client, ...args);
};
TelegramClient.prototype.banChatMember = function(...args) {
  return banChatMember(this._client, ...args);
};
TelegramClient.prototype.createChannel = function(...args) {
  return createChannel(this._client, ...args);
};
TelegramClient.prototype.createGroup = function(...args) {
  return createGroup(this._client, ...args);
};
TelegramClient.prototype.createSupergroup = function(...args) {
  return createSupergroup(this._client, ...args);
};
TelegramClient.prototype.deleteChannel = function(...args) {
  return deleteChannel(this._client, ...args);
};
TelegramClient.prototype.deleteSupergroup = function(...args) {
  return deleteChannel(this._client, ...args);
};
TelegramClient.prototype.deleteChatPhoto = function(...args) {
  return deleteChatPhoto(this._client, ...args);
};
TelegramClient.prototype.deleteGroup = function(...args) {
  return deleteGroup(this._client, ...args);
};
TelegramClient.prototype.deleteHistory = function(...args) {
  return deleteHistory(this._client, ...args);
};
TelegramClient.prototype.deleteUserHistory = function(...args) {
  return deleteUserHistory(this._client, ...args);
};
TelegramClient.prototype.editAdminRights = function(...args) {
  return editAdminRights(this._client, ...args);
};
TelegramClient.prototype.editChatMemberRank = function(...args) {
  return editChatMemberRank(this._client, ...args);
};
TelegramClient.prototype.getChatEventLog = function(...args) {
  return getChatEventLog(this._client, ...args);
};
TelegramClient.prototype.getChatMember = function(...args) {
  return getChatMember(this._client, ...args);
};
TelegramClient.prototype.getChatMembers = function(...args) {
  return getChatMembers(this._client, ...args);
};
TelegramClient.prototype.getChatPreview = function(...args) {
  return getChatPreview(this._client, ...args);
};
TelegramClient.prototype.getChat = function(...args) {
  return getChat(this._client, ...args);
};
TelegramClient.prototype.getChats = function(...args) {
  return getChats(this._client, ...args);
};
TelegramClient.prototype.getCreatorAfterLeave = function(...args) {
  return getCreatorAfterLeave(this._client, ...args);
};
TelegramClient.prototype.getFullChat = function(...args) {
  return getFullChat(this._client, ...args);
};
TelegramClient.prototype.getFullUser = function(...args) {
  return getFullUser(this._client, ...args);
};
TelegramClient.prototype.getMessageAuthor = function(...args) {
  return getMessageAuthor(this._client, ...args);
};
TelegramClient.prototype.getPeer = function(...args) {
  return getPeer(this._client, ...args);
};
TelegramClient.prototype.getPeers = function(...args) {
  return getPeers(this._client, ...args);
};
TelegramClient.prototype.getSimilarChannels = function(...args) {
  return getSimilarChannels(this._client, ...args);
};
TelegramClient.prototype.getUser = function(...args) {
  return getUser(this._client, ...args);
};
TelegramClient.prototype.iterChatEventLog = function(...args) {
  return iterChatEventLog(this._client, ...args);
};
TelegramClient.prototype.iterChatMembers = function(...args) {
  return iterChatMembers(this._client, ...args);
};
TelegramClient.prototype.joinChat = function(...args) {
  return joinChat(this._client, ...args);
};
TelegramClient.prototype.kickChatMember = function(...args) {
  return kickChatMember(this._client, ...args);
};
TelegramClient.prototype.leaveChat = function(...args) {
  return leaveChat(this._client, ...args);
};
TelegramClient.prototype.markChatUnread = function(...args) {
  return markChatUnread(this._client, ...args);
};
TelegramClient.prototype.openChat = function(...args) {
  return openChat(this._client, ...args);
};
TelegramClient.prototype.closeChat = function(...args) {
  return closeChat(this._client, ...args);
};
TelegramClient.prototype.reorderUsernames = function(...args) {
  return reorderUsernames(this._client, ...args);
};
TelegramClient.prototype.restrictChatMember = function(...args) {
  return restrictChatMember(this._client, ...args);
};
TelegramClient.prototype.saveDraft = function(...args) {
  return saveDraft(this._client, ...args);
};
TelegramClient.prototype.setChatColor = function(...args) {
  return setChatColor(this._client, ...args);
};
TelegramClient.prototype.setChatDefaultPermissions = function(...args) {
  return setChatDefaultPermissions(this._client, ...args);
};
TelegramClient.prototype.setChatDescription = function(...args) {
  return setChatDescription(this._client, ...args);
};
TelegramClient.prototype.setChatPhoto = function(...args) {
  return setChatPhoto(this._client, ...args);
};
TelegramClient.prototype.setChatTitle = function(...args) {
  return setChatTitle(this._client, ...args);
};
TelegramClient.prototype.setChatTtl = function(...args) {
  return setChatTtl(this._client, ...args);
};
TelegramClient.prototype.setChatUsername = function(...args) {
  return setChatUsername(this._client, ...args);
};
TelegramClient.prototype.setSlowMode = function(...args) {
  return setSlowMode(this._client, ...args);
};
TelegramClient.prototype.toggleContentProtection = function(...args) {
  return toggleContentProtection(this._client, ...args);
};
TelegramClient.prototype.toggleFragmentUsername = function(...args) {
  return toggleFragmentUsername(this._client, ...args);
};
TelegramClient.prototype.toggleJoinRequests = function(...args) {
  return toggleJoinRequests(this._client, ...args);
};
TelegramClient.prototype.toggleJoinToSend = function(...args) {
  return toggleJoinToSend(this._client, ...args);
};
TelegramClient.prototype.transferChatOwnership = function(...args) {
  return transferChatOwnership(this._client, ...args);
};
TelegramClient.prototype.unarchiveChats = function(...args) {
  return unarchiveChats(this._client, ...args);
};
TelegramClient.prototype.unbanChatMember = function(...args) {
  return unbanChatMember(this._client, ...args);
};
TelegramClient.prototype.unrestrictChatMember = function(...args) {
  return unbanChatMember(this._client, ...args);
};
TelegramClient.prototype.addContact = function(...args) {
  return addContact(this._client, ...args);
};
TelegramClient.prototype.deleteContacts = function(...args) {
  return deleteContacts(this._client, ...args);
};
TelegramClient.prototype.getContacts = function(...args) {
  return getContacts(this._client, ...args);
};
TelegramClient.prototype.importContacts = function(...args) {
  return importContacts(this._client, ...args);
};
TelegramClient.prototype.setContactNote = function(...args) {
  return setContactNote(this._client, ...args);
};
TelegramClient.prototype.createFolder = function(...args) {
  return createFolder(this._client, ...args);
};
TelegramClient.prototype.deleteFolder = function(...args) {
  return deleteFolder(this._client, ...args);
};
TelegramClient.prototype.editFolder = function(...args) {
  return editFolder(this._client, ...args);
};
TelegramClient.prototype.findDialogs = function(...args) {
  return findDialogs(this._client, ...args);
};
TelegramClient.prototype.findFolder = function(...args) {
  return findFolder(this._client, ...args);
};
TelegramClient.prototype.getChatlistPreview = function(...args) {
  return getChatlistPreview(this._client, ...args);
};
TelegramClient.prototype.getFolders = function(...args) {
  return getFolders(this._client, ...args);
};
TelegramClient.prototype.getPeerDialogs = function(...args) {
  return getPeerDialogs(this._client, ...args);
};
TelegramClient.prototype.iterDialogs = function(...args) {
  return iterDialogs(this._client, ...args);
};
TelegramClient.prototype.joinChatlist = function(...args) {
  return joinChatlist(this._client, ...args);
};
TelegramClient.prototype.setFoldersOrder = function(...args) {
  return setFoldersOrder(this._client, ...args);
};
TelegramClient.prototype.downloadAsBuffer = function(...args) {
  return downloadAsBuffer(this._client, ...args);
};
TelegramClient.prototype.downloadChunk = function(...args) {
  return downloadChunk(this._client, ...args);
};
TelegramClient.prototype._normalizeFileDownloadLocation = function(...args) {
  return _normalizeFileDownloadLocation(this._client, ...args);
};
TelegramClient.prototype.downloadAsIterable = function(...args) {
  return downloadAsIterable(this._client, ...args);
};
TelegramClient.prototype.downloadAsStream = function(...args) {
  return downloadAsStream(this._client, ...args);
};
TelegramClient.prototype._normalizeInputFile = function(...args) {
  return _normalizeInputFile(this._client, ...args);
};
TelegramClient.prototype._normalizeInputMedia = function(...args) {
  return _normalizeInputMedia(this._client, ...args);
};
TelegramClient.prototype.uploadFile = function(...args) {
  return uploadFile(this._client, ...args);
};
TelegramClient.prototype.uploadMedia = function(...args) {
  return uploadMedia(this._client, ...args);
};
TelegramClient.prototype.createForumTopic = function(...args) {
  return createForumTopic(this._client, ...args);
};
TelegramClient.prototype.deleteForumTopicHistory = function(...args) {
  return deleteForumTopicHistory(this._client, ...args);
};
TelegramClient.prototype.editForumTopic = function(...args) {
  return editForumTopic(this._client, ...args);
};
TelegramClient.prototype.getForumTopicsById = function(...args) {
  return getForumTopicsById(this._client, ...args);
};
TelegramClient.prototype.getForumTopics = function(...args) {
  return getForumTopics(this._client, ...args);
};
TelegramClient.prototype.iterForumTopics = function(...args) {
  return iterForumTopics(this._client, ...args);
};
TelegramClient.prototype.reorderPinnedForumTopics = function(...args) {
  return reorderPinnedForumTopics(this._client, ...args);
};
TelegramClient.prototype.toggleForumTopicClosed = function(...args) {
  return toggleForumTopicClosed(this._client, ...args);
};
TelegramClient.prototype.toggleForumTopicPinned = function(...args) {
  return toggleForumTopicPinned(this._client, ...args);
};
TelegramClient.prototype.updateForumSettings = function(...args) {
  return updateForumSettings(this._client, ...args);
};
TelegramClient.prototype.toggleGeneralTopicHidden = function(...args) {
  return toggleGeneralTopicHidden(this._client, ...args);
};
TelegramClient.prototype._normalizeInputStarGift = function(...args) {
  return _normalizeInputStarGift(this._client, ...args);
};
TelegramClient.prototype.acceptStarGift = function(...args) {
  return acceptStarGift(this._client, ...args);
};
TelegramClient.prototype.buyResaleGift = function(...args) {
  return buyResaleGift(this._client, ...args);
};
TelegramClient.prototype.getStarGiftResaleOptions = function(...args) {
  return getStarGiftResaleOptions(this._client, ...args);
};
TelegramClient.prototype.getSavedStarGiftsById = function(...args) {
  return getSavedStarGiftsById(this._client, ...args);
};
TelegramClient.prototype.getSavedStarGifts = function(...args) {
  return getSavedStarGifts(this._client, ...args);
};
TelegramClient.prototype.getStarGiftOptions = function(...args) {
  return getStarGiftOptions(this._client, ...args);
};
TelegramClient.prototype.getStarGiftUpgradeOptions = function(...args) {
  return getStarGiftUpgradeOptions(this._client, ...args);
};
TelegramClient.prototype.getStarGiftValue = function(...args) {
  return getStarGiftValue(this._client, ...args);
};
TelegramClient.prototype.getStarGiftWithdrawalUrl = function(...args) {
  return getStarGiftWithdrawalUrl(this._client, ...args);
};
TelegramClient.prototype.getUniqueStarGift = function(...args) {
  return getUniqueStarGift(this._client, ...args);
};
TelegramClient.prototype.iterSavedStarGifts = function(...args) {
  return iterSavedStarGifts(this._client, ...args);
};
TelegramClient.prototype.togglePinnedStarGifts = function(...args) {
  return togglePinnedStarGifts(this._client, ...args);
};
TelegramClient.prototype.prepayStarGiftUpgrade = function(...args) {
  return prepayStarGiftUpgrade(this._client, ...args);
};
TelegramClient.prototype.resolveStarGiftOffer = function(...args) {
  return resolveStarGiftOffer(this._client, ...args);
};
TelegramClient.prototype.sendStarGiftOffer = function(...args) {
  return sendStarGiftOffer(this._client, ...args);
};
TelegramClient.prototype.sendStarGift = function(...args) {
  return sendStarGift(this._client, ...args);
};
TelegramClient.prototype.setResaleStarGiftPrice = function(...args) {
  return setResaleStarGiftPrice(this._client, ...args);
};
TelegramClient.prototype.transferStarGift = function(...args) {
  return transferStarGift(this._client, ...args);
};
TelegramClient.prototype.upgradeStarGift = function(...args) {
  return upgradeStarGift(this._client, ...args);
};
TelegramClient.prototype.createInviteLink = function(...args) {
  return createInviteLink(this._client, ...args);
};
TelegramClient.prototype.editInviteLink = function(...args) {
  return editInviteLink(this._client, ...args);
};
TelegramClient.prototype.exportInviteLink = function(...args) {
  return exportInviteLink(this._client, ...args);
};
TelegramClient.prototype.getInviteLinkMembers = function(...args) {
  return getInviteLinkMembers(this._client, ...args);
};
TelegramClient.prototype.getInviteLink = function(...args) {
  return getInviteLink(this._client, ...args);
};
TelegramClient.prototype.getInviteLinks = function(...args) {
  return getInviteLinks(this._client, ...args);
};
TelegramClient.prototype.getPrimaryInviteLink = function(...args) {
  return getPrimaryInviteLink(this._client, ...args);
};
TelegramClient.prototype.hideAllJoinRequests = function(...args) {
  return hideAllJoinRequests(this._client, ...args);
};
TelegramClient.prototype.hideJoinRequest = function(...args) {
  return hideJoinRequest(this._client, ...args);
};
TelegramClient.prototype.iterInviteLinkMembers = function(...args) {
  return iterInviteLinkMembers(this._client, ...args);
};
TelegramClient.prototype.iterInviteLinks = function(...args) {
  return iterInviteLinks(this._client, ...args);
};
TelegramClient.prototype.revokeInviteLink = function(...args) {
  return revokeInviteLink(this._client, ...args);
};
TelegramClient.prototype.appendTodoList = function(...args) {
  return appendTodoList(this._client, ...args);
};
TelegramClient.prototype.closePoll = function(...args) {
  return closePoll(this._client, ...args);
};
TelegramClient.prototype.createRichStreamingDraft = function(...args) {
  return createRichStreamingDraft(this._client, ...args);
};
TelegramClient.prototype.createStreamingDraft = function(...args) {
  return createStreamingDraft(this._client, ...args);
};
TelegramClient.prototype.deleteMessagesById = function(...args) {
  return deleteMessagesById(this._client, ...args);
};
TelegramClient.prototype.deleteMessages = function(...args) {
  return deleteMessages(this._client, ...args);
};
TelegramClient.prototype.deleteScheduledMessages = function(...args) {
  return deleteScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.editInlineMessage = function(...args) {
  return editInlineMessage(this._client, ...args);
};
TelegramClient.prototype.editMessage = function(...args) {
  return editMessage(this._client, ...args);
};
TelegramClient.prototype.forwardMessagesById = function(...args) {
  return forwardMessagesById(this._client, ...args);
};
TelegramClient.prototype.forwardMessages = function(...args) {
  return forwardMessages(this._client, ...args);
};
TelegramClient.prototype.getAllScheduledMessages = function(...args) {
  return getAllScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.getAvailableMessageEffects = function(...args) {
  return getAvailableMessageEffects(this._client, ...args);
};
TelegramClient.prototype.getCallbackQueryMessage = function(...args) {
  return getCallbackQueryMessage(this._client, ...args);
};
TelegramClient.prototype.getDiscussionMessage = function(...args) {
  return getDiscussionMessage(this._client, ...args);
};
TelegramClient.prototype.getFactCheck = function(...args) {
  return getFactCheck(this._client, ...args);
};
TelegramClient.prototype.getHistory = function(...args) {
  return getHistory(this._client, ...args);
};
TelegramClient.prototype.getMessageByLink = function(...args) {
  return getMessageByLink(this._client, ...args);
};
TelegramClient.prototype.getMessageGroup = function(...args) {
  return getMessageGroup(this._client, ...args);
};
TelegramClient.prototype.getMessageReactionsById = function(...args) {
  return getMessageReactionsById(this._client, ...args);
};
TelegramClient.prototype.getMessageReactions = function(...args) {
  return getMessageReactions(this._client, ...args);
};
TelegramClient.prototype.getMessagesUnsafe = function(...args) {
  return getMessagesUnsafe(this._client, ...args);
};
TelegramClient.prototype.getMessages = function(...args) {
  return getMessages(this._client, ...args);
};
TelegramClient.prototype.getReactionUsers = function(...args) {
  return getReactionUsers(this._client, ...args);
};
TelegramClient.prototype.getReplyTo = function(...args) {
  return getReplyTo(this._client, ...args);
};
TelegramClient.prototype.getScheduledMessages = function(...args) {
  return getScheduledMessages(this._client, ...args);
};
TelegramClient.prototype.getWebPagePreview = function(...args) {
  return getWebPagePreview(this._client, ...args);
};
TelegramClient.prototype.iterHistory = function(...args) {
  return iterHistory(this._client, ...args);
};
TelegramClient.prototype.iterReactionUsers = function(...args) {
  return iterReactionUsers(this._client, ...args);
};
TelegramClient.prototype.iterSearchGlobal = function(...args) {
  return iterSearchGlobal(this._client, ...args);
};
TelegramClient.prototype.iterSearchMessages = function(...args) {
  return iterSearchMessages(this._client, ...args);
};
TelegramClient.prototype.pinMessage = function(...args) {
  return pinMessage(this._client, ...args);
};
TelegramClient.prototype.readHistory = function(...args) {
  return readHistory(this._client, ...args);
};
TelegramClient.prototype.readReactions = function(...args) {
  return readReactions(this._client, ...args);
};
TelegramClient.prototype.searchGlobal = function(...args) {
  return searchGlobal(this._client, ...args);
};
TelegramClient.prototype.searchHashtag = function(...args) {
  return searchHashtag(this._client, ...args);
};
TelegramClient.prototype.iterSearchHashtag = function(...args) {
  return iterSearchHashtag(this._client, ...args);
};
TelegramClient.prototype.searchMessages = function(...args) {
  return searchMessages(this._client, ...args);
};
TelegramClient.prototype.searchPostsGlobal = function(...args) {
  return searchPostsGlobal(this._client, ...args);
};
TelegramClient.prototype.answerText = function(...args) {
  return answerText(this._client, ...args);
};
TelegramClient.prototype.answerMedia = function(...args) {
  return answerMedia(this._client, ...args);
};
TelegramClient.prototype.answerMediaGroup = function(...args) {
  return answerMediaGroup(this._client, ...args);
};
TelegramClient.prototype.commentText = function(...args) {
  return commentText(this._client, ...args);
};
TelegramClient.prototype.commentMedia = function(...args) {
  return commentMedia(this._client, ...args);
};
TelegramClient.prototype.commentMediaGroup = function(...args) {
  return commentMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendCopyGroup = function(...args) {
  return sendCopyGroup(this._client, ...args);
};
TelegramClient.prototype.sendCopy = function(...args) {
  return sendCopy(this._client, ...args);
};
TelegramClient.prototype.sendMediaGroup = function(...args) {
  return sendMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendMedia = function(...args) {
  return sendMedia(this._client, ...args);
};
TelegramClient.prototype.sendPaidReaction = function(...args) {
  return sendPaidReaction(this._client, ...args);
};
TelegramClient.prototype.quoteWithText = function(...args) {
  return quoteWithText(this._client, ...args);
};
TelegramClient.prototype.quoteWithMedia = function(...args) {
  return quoteWithMedia(this._client, ...args);
};
TelegramClient.prototype.quoteWithMediaGroup = function(...args) {
  return quoteWithMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendReaction = function(...args) {
  return sendReaction(this._client, ...args);
};
TelegramClient.prototype.replyText = function(...args) {
  return replyText(this._client, ...args);
};
TelegramClient.prototype.replyMedia = function(...args) {
  return replyMedia(this._client, ...args);
};
TelegramClient.prototype.replyMediaGroup = function(...args) {
  return replyMediaGroup(this._client, ...args);
};
TelegramClient.prototype.sendRichMessage = function(...args) {
  return sendRichMessage(this._client, ...args);
};
TelegramClient.prototype.sendScheduled = function(...args) {
  return sendScheduled(this._client, ...args);
};
TelegramClient.prototype.sendText = function(...args) {
  return sendText(this._client, ...args);
};
TelegramClient.prototype.sendTyping = function(...args) {
  return sendTyping(this._client, ...args);
};
TelegramClient.prototype.sendVote = function(...args) {
  return sendVote(this._client, ...args);
};
TelegramClient.prototype.setTyping = function(...args) {
  return setTyping(this._client, ...args);
};
TelegramClient.prototype.toggleTodoCompleted = function(...args) {
  return toggleTodoCompleted(this._client, ...args);
};
TelegramClient.prototype.translateMessage = function(...args) {
  return translateMessage(this._client, ...args);
};
TelegramClient.prototype.translateText = function(...args) {
  return translateText(this._client, ...args);
};
TelegramClient.prototype.unpinAllMessages = function(...args) {
  return unpinAllMessages(this._client, ...args);
};
TelegramClient.prototype.unpinMessage = function(...args) {
  return unpinMessage(this._client, ...args);
};
TelegramClient.prototype.getCollectibleInfo = function(...args) {
  return getCollectibleInfo(this._client, ...args);
};
TelegramClient.prototype.initTakeoutSession = function(...args) {
  return initTakeoutSession(this._client, ...args);
};
TelegramClient.prototype._normalizePrivacyRules = function(...args) {
  return _normalizePrivacyRules(this._client, ...args);
};
TelegramClient.prototype.changeCloudPassword = function(...args) {
  return changeCloudPassword(this._client, ...args);
};
TelegramClient.prototype.enableCloudPassword = function(...args) {
  return enableCloudPassword(this._client, ...args);
};
TelegramClient.prototype.verifyPasswordEmail = function(...args) {
  return verifyPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.resendPasswordEmail = function(...args) {
  return resendPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.cancelPasswordEmail = function(...args) {
  return cancelPasswordEmail(this._client, ...args);
};
TelegramClient.prototype.removeCloudPassword = function(...args) {
  return removeCloudPassword(this._client, ...args);
};
TelegramClient.prototype.applyBoost = function(...args) {
  return applyBoost(this._client, ...args);
};
TelegramClient.prototype.canApplyBoost = function(...args) {
  return canApplyBoost(this._client, ...args);
};
TelegramClient.prototype.createBusinessChatLink = function(...args) {
  return createBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.editBusinessChatLink = function(...args) {
  return editBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.deleteBusinessChatLink = function(...args) {
  return deleteBusinessChatLink(this._client, ...args);
};
TelegramClient.prototype.getBoostStats = function(...args) {
  return getBoostStats(this._client, ...args);
};
TelegramClient.prototype.getBoosts = function(...args) {
  return getBoosts(this._client, ...args);
};
TelegramClient.prototype.getBusinessChatLinks = function(...args) {
  return getBusinessChatLinks(this._client, ...args);
};
TelegramClient.prototype.getBusinessConnection = function(...args) {
  return getBusinessConnection(this._client, ...args);
};
TelegramClient.prototype.getMyBoostSlots = function(...args) {
  return getMyBoostSlots(this._client, ...args);
};
TelegramClient.prototype.getStarsTransactions = function(...args) {
  return getStarsTransactions(this._client, ...args);
};
TelegramClient.prototype.iterBoosters = function(...args) {
  return iterBoosters(this._client, ...args);
};
TelegramClient.prototype.iterStarsTransactions = function(...args) {
  return iterStarsTransactions(this._client, ...args);
};
TelegramClient.prototype.setBusinessIntro = function(...args) {
  return setBusinessIntro(this._client, ...args);
};
TelegramClient.prototype.setBusinessWorkHours = function(...args) {
  return setBusinessWorkHours(this._client, ...args);
};
TelegramClient.prototype.addStickerToSet = function(...args) {
  return addStickerToSet(this._client, ...args);
};
TelegramClient.prototype.createStickerSet = function(...args) {
  return createStickerSet(this._client, ...args);
};
TelegramClient.prototype.deleteStickerFromSet = function(...args) {
  return deleteStickerFromSet(this._client, ...args);
};
TelegramClient.prototype.getCustomEmojis = function(...args) {
  return getCustomEmojis(this._client, ...args);
};
TelegramClient.prototype.getCustomEmojisFromMessages = function(...args) {
  return getCustomEmojisFromMessages(this._client, ...args);
};
TelegramClient.prototype.getInstalledStickers = function(...args) {
  return getInstalledStickers(this._client, ...args);
};
TelegramClient.prototype.getMyStickerSets = function(...args) {
  return getMyStickerSets(this._client, ...args);
};
TelegramClient.prototype.getStickerSet = function(...args) {
  return getStickerSet(this._client, ...args);
};
TelegramClient.prototype.moveStickerInSet = function(...args) {
  return moveStickerInSet(this._client, ...args);
};
TelegramClient.prototype.replaceStickerInSet = function(...args) {
  return replaceStickerInSet(this._client, ...args);
};
TelegramClient.prototype.setChatStickerSet = function(...args) {
  return setChatStickerSet(this._client, ...args);
};
TelegramClient.prototype.setStickerSetThumb = function(...args) {
  return setStickerSetThumb(this._client, ...args);
};
TelegramClient.prototype.canSendStory = function(...args) {
  return canSendStory(this._client, ...args);
};
TelegramClient.prototype.deleteStories = function(...args) {
  return deleteStories(this._client, ...args);
};
TelegramClient.prototype.editStory = function(...args) {
  return editStory(this._client, ...args);
};
TelegramClient.prototype.getAllStories = function(...args) {
  return getAllStories(this._client, ...args);
};
TelegramClient.prototype.getPeerStories = function(...args) {
  return getPeerStories(this._client, ...args);
};
TelegramClient.prototype.getProfileStories = function(...args) {
  return getProfileStories(this._client, ...args);
};
TelegramClient.prototype.getStoriesById = function(...args) {
  return getStoriesById(this._client, ...args);
};
TelegramClient.prototype.getStoriesInteractions = function(...args) {
  return getStoriesInteractions(this._client, ...args);
};
TelegramClient.prototype.getStoryLink = function(...args) {
  return getStoryLink(this._client, ...args);
};
TelegramClient.prototype.getStoryViewers = function(...args) {
  return getStoryViewers(this._client, ...args);
};
TelegramClient.prototype.hideMyStoriesViews = function(...args) {
  return hideMyStoriesViews(this._client, ...args);
};
TelegramClient.prototype.incrementStoriesViews = function(...args) {
  return incrementStoriesViews(this._client, ...args);
};
TelegramClient.prototype.iterAllStories = function(...args) {
  return iterAllStories(this._client, ...args);
};
TelegramClient.prototype.iterProfileStories = function(...args) {
  return iterProfileStories(this._client, ...args);
};
TelegramClient.prototype.iterStoryViewers = function(...args) {
  return iterStoryViewers(this._client, ...args);
};
TelegramClient.prototype.readStories = function(...args) {
  return readStories(this._client, ...args);
};
TelegramClient.prototype.sendStoryReaction = function(...args) {
  return sendStoryReaction(this._client, ...args);
};
TelegramClient.prototype.sendStory = function(...args) {
  return sendStory(this._client, ...args);
};
TelegramClient.prototype.togglePeerStoriesArchived = function(...args) {
  return togglePeerStoriesArchived(this._client, ...args);
};
TelegramClient.prototype.toggleStoriesPinned = function(...args) {
  return toggleStoriesPinned(this._client, ...args);
};
TelegramClient.prototype.blockUser = function(...args) {
  return blockUser(this._client, ...args);
};
TelegramClient.prototype.deleteProfilePhotos = function(...args) {
  return deleteProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.editCloseFriendsRaw = function(...args) {
  return editCloseFriendsRaw(this._client, ...args);
};
TelegramClient.prototype.editCloseFriends = function(...args) {
  return editCloseFriends(this._client, ...args);
};
TelegramClient.prototype.getCommonChats = function(...args) {
  return getCommonChats(this._client, ...args);
};
TelegramClient.prototype.getGlobalTtl = function(...args) {
  return getGlobalTtl(this._client, ...args);
};
TelegramClient.prototype.getMe = function(...args) {
  return getMe(this._client, ...args);
};
TelegramClient.prototype.getMyUsername = function(...args) {
  return getMyUsername(this._client, ...args);
};
TelegramClient.prototype.getPeerSettings = function(...args) {
  return getPeerSettings(this._client, ...args);
};
TelegramClient.prototype.getProfilePhoto = function(...args) {
  return getProfilePhoto(this._client, ...args);
};
TelegramClient.prototype.getProfilePhotos = function(...args) {
  return getProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.getSavedMusic = function(...args) {
  return getSavedMusic(this._client, ...args);
};
TelegramClient.prototype.getUsers = function(...args) {
  return getUsers(this._client, ...args);
};
TelegramClient.prototype.isPeerAvailable = function(...args) {
  return isPeerAvailable(this._client, ...args);
};
TelegramClient.prototype.iterProfilePhotos = function(...args) {
  return iterProfilePhotos(this._client, ...args);
};
TelegramClient.prototype.resolvePeerMany = function(...args) {
  return resolvePeerMany(this._client, ...args);
};
TelegramClient.prototype.resolvePeer = function(...args) {
  return resolvePeer(this._client, ...args);
};
TelegramClient.prototype.resolveUser = function(...args) {
  return resolveUser(this._client, ...args);
};
TelegramClient.prototype.resolveChannel = function(...args) {
  return resolveChannel(this._client, ...args);
};
TelegramClient.prototype.resolvePhoneNumber = function(...args) {
  return resolvePhoneNumber(this._client, ...args);
};
TelegramClient.prototype.saveMusicToProfile = function(...args) {
  return saveMusicToProfile(this._client, ...args);
};
TelegramClient.prototype.unsaveMusicFromProfile = function(...args) {
  return unsaveMusicFromProfile(this._client, ...args);
};
TelegramClient.prototype.setEmojiStatus = function(...args) {
  return setEmojiStatus(this._client, ...args);
};
TelegramClient.prototype.setGlobalTtl = function(...args) {
  return setGlobalTtl(this._client, ...args);
};
TelegramClient.prototype.setMyBirthday = function(...args) {
  return setMyBirthday(this._client, ...args);
};
TelegramClient.prototype.setMyProfilePhoto = function(...args) {
  return setMyProfilePhoto(this._client, ...args);
};
TelegramClient.prototype.setMyUsername = function(...args) {
  return setMyUsername(this._client, ...args);
};
TelegramClient.prototype.sendOnline = function(...args) {
  return sendOnline(this._client, ...args);
};
TelegramClient.prototype.setOnline = function(...args) {
  return setOnline(this._client, ...args);
};
TelegramClient.prototype.unblockUser = function(...args) {
  return unblockUser(this._client, ...args);
};
TelegramClient.prototype.updateProfile = function(...args) {
  return updateProfile(this._client, ...args);
};
TelegramClient.prototype.prepare = function(...args) {
  return this._client.prepare(...args);
};
TelegramClient.prototype.connect = function(...args) {
  return this._client.connect(...args);
};
TelegramClient.prototype.disconnect = function(...args) {
  return this._client.disconnect(...args);
};
TelegramClient.prototype.destroy = function(...args) {
  return this._client.destroy(...args);
};
TelegramClient.prototype.notifyLoggedIn = function(...args) {
  return this._client.notifyLoggedIn(...args);
};
TelegramClient.prototype.notifyLoggedOut = function(...args) {
  return this._client.notifyLoggedOut(...args);
};
TelegramClient.prototype.notifyChannelOpened = function(...args) {
  return this._client.notifyChannelOpened(...args);
};
TelegramClient.prototype.notifyChannelClosed = function(...args) {
  return this._client.notifyChannelClosed(...args);
};
TelegramClient.prototype.startUpdatesLoop = function(...args) {
  return this._client.startUpdatesLoop(...args);
};
TelegramClient.prototype.stopUpdatesLoop = function(...args) {
  return this._client.stopUpdatesLoop(...args);
};
TelegramClient.prototype.call = function(...args) {
  return this._client.call(...args);
};
TelegramClient.prototype.importSession = function(...args) {
  return this._client.importSession(...args);
};
TelegramClient.prototype.exportSession = function(...args) {
  return this._client.exportSession(...args);
};
TelegramClient.prototype.handleClientUpdate = function(...args) {
  return this._client.handleClientUpdate(...args);
};
TelegramClient.prototype.getApiCredentials = function(...args) {
  return this._client.getApiCredentials(...args);
};
TelegramClient.prototype.getPoolSize = function(...args) {
  return this._client.getPoolSize(...args);
};
TelegramClient.prototype.getPrimaryDcId = function(...args) {
  return this._client.getPrimaryDcId(...args);
};
TelegramClient.prototype.computeSrpParams = function(...args) {
  return this._client.computeSrpParams(...args);
};
TelegramClient.prototype.computeNewPasswordHash = function(...args) {
  return this._client.computeNewPasswordHash(...args);
};
TelegramClient.prototype.changePrimaryDc = function(...args) {
  return this._client.changePrimaryDc(...args);
};
TelegramClient.prototype.getMtprotoMessageId = function(...args) {
  return this._client.getMtprotoMessageId(...args);
};
TelegramClient.prototype.recreateDc = function(...args) {
  return this._client.recreateDc(...args);
};
export {
  BaseTelegramClient,
  TelegramClient
};
