// @mtcute/core/index.js
import { default as default2 } from "long";
import { getAllPeersFrom, getBarePeerId, getBasicPeerType, getMarkedPeerId, parseMarkedPeerId, toggleChannelIdMark } from "./utils/peer-utils.js";
import * as innerMtp from "./tl/inner-mtp.js";
import * as innerTl from "./tl/inner-tl.js";
import { TelegramStorageManager } from "./highlevel/storage/storage.js";
import { SentCode } from "./highlevel/types/auth/sent-code.js";
import * as inner from "./highlevel/types/bots/command-scope/inner.js";
import { GameHighScore } from "./highlevel/types/bots/game-high-score.js";
import * as factories from "./highlevel/types/bots/inline-message/factories.js";
import * as factories$1 from "./highlevel/types/bots/inline-result/factories.js";
import { BotKeyboardBuilder } from "./highlevel/types/bots/keyboards/builder.js";
import * as factories$2 from "./highlevel/types/bots/keyboards/factories.js";
import { WebviewResult } from "./highlevel/types/bots/webview.js";
import { Conversation } from "./highlevel/types/conversation.js";
import { MtEmptyError, MtInvalidPeerTypeError, MtMessageNotFoundError, MtPeerNotFoundError } from "./highlevel/types/errors.js";
import { FileLocation } from "./highlevel/types/files/file-location.js";
import { isUploadedFile } from "./highlevel/types/files/uploaded-file.js";
import { WebDocument } from "./highlevel/types/files/web-document.js";
import { Audio } from "./highlevel/types/media/audio.js";
import { Contact } from "./highlevel/types/media/contact.js";
import { Dice } from "./highlevel/types/media/dice.js";
import { Document, RawDocument } from "./highlevel/types/media/document.js";
import { ExtendedMediaPreview } from "./highlevel/types/media/extended-media.js";
import { Game } from "./highlevel/types/media/game.js";
import * as factories$3 from "./highlevel/types/media/input-media/factories.js";
import { Invoice } from "./highlevel/types/media/invoice.js";
import { LiveLocation, Location, RawLocation } from "./highlevel/types/media/location.js";
import { PaidMedia } from "./highlevel/types/media/paid-media.js";
import { Photo } from "./highlevel/types/media/photo.js";
import { Poll, PollAnswer } from "./highlevel/types/media/poll.js";
import { MASK_POSITION_POINT_TO_TL, Sticker } from "./highlevel/types/media/sticker.js";
import { MediaStory } from "./highlevel/types/media/story.js";
import { Thumbnail } from "./highlevel/types/media/thumbnail.js";
import { TodoItem, TodoList } from "./highlevel/types/media/todo.js";
import { Venue } from "./highlevel/types/media/venue.js";
import { Video } from "./highlevel/types/media/video.js";
import { Voice } from "./highlevel/types/media/voice.js";
import { WebPage, WebPageMedia } from "./highlevel/types/media/web-page.js";
import { Dialog } from "./highlevel/types/messages/dialog.js";
import { DraftMessage } from "./highlevel/types/messages/draft-message.js";
import { FactCheck } from "./highlevel/types/messages/fact-check.js";
import { normalizeInputMessageId } from "./highlevel/types/messages/input-message-id.js";
import { _messageActionFromTl } from "./highlevel/types/messages/message-action.js";
import { MessageEffect } from "./highlevel/types/messages/message-effect.js";
import { MessageEntity } from "./highlevel/types/messages/message-entity.js";
import { MessageForwardInfo } from "./highlevel/types/messages/message-forward.js";
import { _messageMediaFromTl } from "./highlevel/types/messages/message-media.js";
import { MessageReactions } from "./highlevel/types/messages/message-reactions.js";
import { MessageRepliesInfo } from "./highlevel/types/messages/message-replies.js";
import { Message } from "./highlevel/types/messages/message.js";
import { RepliedMessageInfo } from "./highlevel/types/messages/replied-message.js";
import { RichMessage } from "./highlevel/types/messages/rich/rich-message.js";
import * as inner$1 from "./highlevel/types/messages/rich/inner.js";
import { SearchFilters } from "./highlevel/types/messages/search-filters.js";
import { SuggestedPostInfo } from "./highlevel/types/messages/suggested-post.js";
import { CollectibleInfo } from "./highlevel/types/misc/collectible-info.js";
import { inputTextToTl } from "./highlevel/types/misc/entities.js";
import * as bundle from "./highlevel/types/misc/input-privacy-rule/bundle.js";
import { StickerSet, normalizeInputStickerSet } from "./highlevel/types/misc/sticker-set.js";
import { TakeoutSession } from "./highlevel/types/misc/takeout-session.js";
import { BotInfo } from "./highlevel/types/peers/bot-info.js";
import { BotVerification } from "./highlevel/types/peers/bot-verification.js";
import { ChatColors } from "./highlevel/types/peers/chat-colors.js";
import { _actionFromTl } from "./highlevel/types/peers/chat-event/actions.js";
import { ChatEvent } from "./highlevel/types/peers/chat-event/index.js";
import { ChatInviteLinkMember } from "./highlevel/types/peers/chat-invite-link-member.js";
import { ChatInviteLink } from "./highlevel/types/peers/chat-invite-link.js";
import { ChatMember } from "./highlevel/types/peers/chat-member.js";
import { ChatPermissions } from "./highlevel/types/peers/chat-permissions.js";
import { ChatPhoto, ChatPhotoSize } from "./highlevel/types/peers/chat-photo.js";
import { ChatPreview } from "./highlevel/types/peers/chat-preview.js";
import { Chat } from "./highlevel/types/peers/chat.js";
import { ChatlistPreview } from "./highlevel/types/peers/chatlist-preview.js";
import { ForumTopic } from "./highlevel/types/peers/forum-topic.js";
import { FullChat } from "./highlevel/types/peers/full-chat.js";
import { FullUser } from "./highlevel/types/peers/full-user.js";
import { PeerSettings } from "./highlevel/types/peers/peer-settings.js";
import { parsePeer } from "./highlevel/types/peers/peer.js";
import { PeersIndex } from "./highlevel/types/peers/peers-index.js";
import { User } from "./highlevel/types/peers/user.js";
import { BoostSlot } from "./highlevel/types/premium/boost-slot.js";
import { BoostStats } from "./highlevel/types/premium/boost-stats.js";
import { Boost } from "./highlevel/types/premium/boost.js";
import { BusinessAccount } from "./highlevel/types/premium/business-account.js";
import { BusinessChatLink } from "./highlevel/types/premium/business-chat-link.js";
import { BusinessConnection } from "./highlevel/types/premium/business-connection.js";
import { BusinessIntro } from "./highlevel/types/premium/business-intro.js";
import { BusinessLocation } from "./highlevel/types/premium/business-location.js";
import { BusinessWorkHours, businessWorkHoursDaysToRaw } from "./highlevel/types/premium/business-work-hours.js";
import { SavedStarGift } from "./highlevel/types/premium/saved-star-gift.js";
import { StarGiftUnique, StarGiftUniqueAttribute, StarGiftUniqueBackdrop, StarGiftUniqueOriginalDetails } from "./highlevel/types/premium/star-gift-unique.js";
import { StarGiftValue } from "./highlevel/types/premium/star-gift-value.js";
import { StarGift } from "./highlevel/types/premium/star-gift.js";
import { StarsStatus } from "./highlevel/types/premium/stars-status.js";
import { StarsTransaction } from "./highlevel/types/premium/stars-transaction.js";
import { EmojiStatus, EmojiStatusCollectible } from "./highlevel/types/reactions/emoji-status.js";
import { PaidPeerReaction, PeerReaction } from "./highlevel/types/reactions/peer-reaction.js";
import { ReactionCount } from "./highlevel/types/reactions/reaction-count.js";
import { normalizeInputReaction, toReactionEmoji } from "./highlevel/types/reactions/types.js";
import { AllStories } from "./highlevel/types/stories/all-stories.js";
import { StoryElement } from "./highlevel/types/stories/interactive/input.js";
import { StoryInteractiveChannelPost } from "./highlevel/types/stories/interactive/channel-post.js";
import { StoryInteractiveLocation } from "./highlevel/types/stories/interactive/location.js";
import { StoryInteractiveReaction } from "./highlevel/types/stories/interactive/reaction.js";
import { StoryInteractiveStarGift } from "./highlevel/types/stories/interactive/star-gift.js";
import { StoryInteractiveUrl } from "./highlevel/types/stories/interactive/url.js";
import { StoryInteractiveVenue } from "./highlevel/types/stories/interactive/venue.js";
import { StoryInteractiveWeather } from "./highlevel/types/stories/interactive/weather.js";
import { _storyInteractiveElementFromTl } from "./highlevel/types/stories/interactive/index.js";
import { PeerStories } from "./highlevel/types/stories/peer-stories.js";
import { StoriesStealthMode } from "./highlevel/types/stories/stealth-mode.js";
import { StoryInteractions } from "./highlevel/types/stories/story-interactions.js";
import { StoryRepost, StoryViewer, StoryViewersList } from "./highlevel/types/stories/story-viewer.js";
import { Story } from "./highlevel/types/stories/story.js";
import { BotChatJoinRequestUpdate } from "./highlevel/types/updates/bot-chat-join-request.js";
import { BotGuestChatQuery } from "./highlevel/types/updates/bot-guest-chat-query.js";
import { BotReactionCountUpdate, BotReactionUpdate } from "./highlevel/types/updates/bot-reaction.js";
import { BotStoppedUpdate } from "./highlevel/types/updates/bot-stopped.js";
import { BusinessCallbackQuery, CallbackQuery, InlineCallbackQuery } from "./highlevel/types/updates/callback-query.js";
import { BusinessMessage } from "./highlevel/types/updates/business-message.js";
import { ChatJoinRequestUpdate } from "./highlevel/types/updates/chat-join-request.js";
import { ChatMemberUpdate } from "./highlevel/types/updates/chat-member-update.js";
import { ChosenInlineResult } from "./highlevel/types/updates/chosen-inline-result.js";
import { DeleteBusinessMessageUpdate } from "./highlevel/types/updates/delete-business-message-update.js";
import { DeleteMessageUpdate } from "./highlevel/types/updates/delete-message-update.js";
import { DeleteStoryUpdate } from "./highlevel/types/updates/delete-story-update.js";
import { HistoryReadUpdate } from "./highlevel/types/updates/history-read-update.js";
import { InlineQuery } from "./highlevel/types/updates/inline-query.js";
import { PollUpdate } from "./highlevel/types/updates/poll-update.js";
import { PollVoteUpdate } from "./highlevel/types/updates/poll-vote.js";
import { PreCheckoutQuery } from "./highlevel/types/updates/pre-checkout-query.js";
import { StoryUpdate } from "./highlevel/types/updates/story-update.js";
import { UserStatusUpdate } from "./highlevel/types/updates/user-status-update.js";
import { UserTypingUpdate } from "./highlevel/types/updates/user-typing-update.js";
import { UpdatesManager } from "./highlevel/updates/manager.js";
import { RawUpdateInfo } from "./highlevel/updates/types.js";
import { MtClient } from "./network/client.js";
import * as bundle$1 from "./network/middlewares/bundle.js";
import { BaseMtProxyTransport } from "./network/mtproxy/index.js";
import { SessionConnection } from "./network/session-connection.js";
import { TransportError } from "./network/transports/abstract.js";
import { IntermediatePacketCodec, PaddedIntermediatePacketCodec } from "./network/transports/intermediate.js";
import { ObfuscatedPacketCodec } from "./network/transports/obfuscated.js";
import { createProxyTransportFactory } from "./network/transports/proxy.js";
import { BaseStorageDriver } from "./storage/driver.js";
import { MemoryStorageDriver } from "./storage/memory/driver.js";
import { MemoryAuthKeysRepository } from "./storage/memory/repository/auth-keys.js";
import { MemoryKeyValueRepository } from "./storage/memory/repository/kv.js";
import { MemoryPeersRepository } from "./storage/memory/repository/peers.js";
import { MemoryRefMessagesRepository } from "./storage/memory/repository/ref-messages.js";
import { MemoryStorage } from "./storage/memory/index.js";
import { BaseSqliteStorageDriver } from "./storage/sqlite/driver.js";
import { BaseSqliteStorage } from "./storage/sqlite/index.js";
import { StorageManager } from "./storage/storage.js";
import { MtArgumentError, MtSecurityError, MtTimeoutError, MtTypeAssertionError, MtUnsupportedError, MtcuteError } from "./types/errors.js";
import { assertNever } from "./types/utils.js";
export {
  AllStories,
  Audio,
  BaseMtProxyTransport,
  BaseSqliteStorage,
  BaseSqliteStorageDriver,
  BaseStorageDriver,
  Boost,
  BoostSlot,
  BoostStats,
  BotChatJoinRequestUpdate,
  inner as BotCommands,
  BotGuestChatQuery,
  BotInfo,
  factories$1 as BotInline,
  factories as BotInlineMessage,
  factories$2 as BotKeyboard,
  BotKeyboardBuilder,
  BotReactionCountUpdate,
  BotReactionUpdate,
  BotStoppedUpdate,
  BotVerification,
  BusinessAccount,
  BusinessCallbackQuery,
  BusinessChatLink,
  BusinessConnection,
  BusinessIntro,
  BusinessLocation,
  BusinessMessage,
  BusinessWorkHours,
  CallbackQuery,
  Chat,
  ChatColors,
  ChatEvent,
  ChatInviteLink,
  ChatInviteLinkMember,
  ChatJoinRequestUpdate,
  ChatMember,
  ChatMemberUpdate,
  ChatPermissions,
  ChatPhoto,
  ChatPhotoSize,
  ChatPreview,
  ChatlistPreview,
  ChosenInlineResult,
  CollectibleInfo,
  Contact,
  Conversation,
  DeleteBusinessMessageUpdate,
  DeleteMessageUpdate,
  DeleteStoryUpdate,
  Dialog,
  Dice,
  Document,
  DraftMessage,
  EmojiStatus,
  EmojiStatusCollectible,
  ExtendedMediaPreview,
  FactCheck,
  FileLocation,
  ForumTopic,
  FullChat,
  FullUser,
  Game,
  GameHighScore,
  HistoryReadUpdate,
  InlineCallbackQuery,
  InlineQuery,
  factories$3 as InputMedia,
  IntermediatePacketCodec,
  Invoice,
  LiveLocation,
  Location,
  default2 as Long,
  MASK_POSITION_POINT_TO_TL,
  MediaStory,
  MemoryAuthKeysRepository,
  MemoryKeyValueRepository,
  MemoryPeersRepository,
  MemoryRefMessagesRepository,
  MemoryStorage,
  MemoryStorageDriver,
  Message,
  MessageEffect,
  MessageEntity,
  MessageForwardInfo,
  MessageReactions,
  MessageRepliesInfo,
  MtArgumentError,
  MtClient,
  MtEmptyError,
  MtInvalidPeerTypeError,
  MtMessageNotFoundError,
  MtPeerNotFoundError,
  MtSecurityError,
  MtTimeoutError,
  MtTypeAssertionError,
  MtUnsupportedError,
  MtcuteError,
  ObfuscatedPacketCodec,
  PaddedIntermediatePacketCodec,
  PaidMedia,
  PaidPeerReaction,
  PeerReaction,
  PeerSettings,
  PeerStories,
  PeersIndex,
  Photo,
  Poll,
  PollAnswer,
  PollUpdate,
  PollVoteUpdate,
  PreCheckoutQuery,
  bundle as PrivacyRule,
  RawDocument,
  RawLocation,
  RawUpdateInfo,
  ReactionCount,
  RepliedMessageInfo,
  inner$1 as Rich,
  RichMessage,
  SavedStarGift,
  SearchFilters,
  SentCode,
  SessionConnection,
  StarGift,
  StarGiftUnique,
  StarGiftUniqueAttribute,
  StarGiftUniqueBackdrop,
  StarGiftUniqueOriginalDetails,
  StarGiftValue,
  StarsStatus,
  StarsTransaction,
  Sticker,
  StickerSet,
  StorageManager,
  StoriesStealthMode,
  Story,
  StoryElement,
  StoryInteractions,
  StoryInteractiveChannelPost,
  StoryInteractiveLocation,
  StoryInteractiveReaction,
  StoryInteractiveStarGift,
  StoryInteractiveUrl,
  StoryInteractiveVenue,
  StoryInteractiveWeather,
  StoryRepost,
  StoryUpdate,
  StoryViewer,
  StoryViewersList,
  SuggestedPostInfo,
  TakeoutSession,
  TelegramStorageManager,
  Thumbnail,
  TodoItem,
  TodoList,
  TransportError,
  UpdatesManager,
  User,
  UserStatusUpdate,
  UserTypingUpdate,
  Venue,
  Video,
  Voice,
  WebDocument,
  WebPage,
  WebPageMedia,
  WebviewResult,
  _actionFromTl,
  _messageActionFromTl,
  _messageMediaFromTl,
  _storyInteractiveElementFromTl,
  assertNever,
  businessWorkHoursDaysToRaw,
  createProxyTransportFactory,
  getAllPeersFrom,
  getBarePeerId,
  getBasicPeerType,
  getMarkedPeerId,
  inputTextToTl,
  isUploadedFile,
  innerMtp as mtp,
  bundle$1 as networkMiddlewares,
  normalizeInputMessageId,
  normalizeInputReaction,
  normalizeInputStickerSet,
  parseMarkedPeerId,
  parsePeer,
  innerTl as tl,
  toReactionEmoji,
  toggleChannelIdMark
};
