// @mtcute/core/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const peerUtils = require("./utils/peer-utils.cjs");
const innerMtp = require("./tl/inner-mtp.cjs");
const innerTl = require("./tl/inner-tl.cjs");
const storage = require("./highlevel/storage/storage.cjs");
const sentCode = require("./highlevel/types/auth/sent-code.cjs");
const inner = require("./highlevel/types/bots/command-scope/inner.cjs");
const gameHighScore = require("./highlevel/types/bots/game-high-score.cjs");
const factories = require("./highlevel/types/bots/inline-message/factories.cjs");
const factories$1 = require("./highlevel/types/bots/inline-result/factories.cjs");
const builder = require("./highlevel/types/bots/keyboards/builder.cjs");
const factories$2 = require("./highlevel/types/bots/keyboards/factories.cjs");
const webview = require("./highlevel/types/bots/webview.cjs");
const conversation = require("./highlevel/types/conversation.cjs");
const errors = require("./highlevel/types/errors.cjs");
const fileLocation = require("./highlevel/types/files/file-location.cjs");
const uploadedFile = require("./highlevel/types/files/uploaded-file.cjs");
const webDocument = require("./highlevel/types/files/web-document.cjs");
const audio = require("./highlevel/types/media/audio.cjs");
const contact = require("./highlevel/types/media/contact.cjs");
const dice = require("./highlevel/types/media/dice.cjs");
const document = require("./highlevel/types/media/document.cjs");
const extendedMedia = require("./highlevel/types/media/extended-media.cjs");
const game = require("./highlevel/types/media/game.cjs");
const factories$3 = require("./highlevel/types/media/input-media/factories.cjs");
const invoice = require("./highlevel/types/media/invoice.cjs");
const location = require("./highlevel/types/media/location.cjs");
const paidMedia = require("./highlevel/types/media/paid-media.cjs");
const photo = require("./highlevel/types/media/photo.cjs");
const poll = require("./highlevel/types/media/poll.cjs");
const sticker = require("./highlevel/types/media/sticker.cjs");
const story = require("./highlevel/types/media/story.cjs");
const thumbnail = require("./highlevel/types/media/thumbnail.cjs");
const todo = require("./highlevel/types/media/todo.cjs");
const venue = require("./highlevel/types/media/venue.cjs");
const video = require("./highlevel/types/media/video.cjs");
const voice = require("./highlevel/types/media/voice.cjs");
const webPage = require("./highlevel/types/media/web-page.cjs");
const dialog = require("./highlevel/types/messages/dialog.cjs");
const draftMessage = require("./highlevel/types/messages/draft-message.cjs");
const factCheck = require("./highlevel/types/messages/fact-check.cjs");
const inputMessageId = require("./highlevel/types/messages/input-message-id.cjs");
const messageAction = require("./highlevel/types/messages/message-action.cjs");
const messageEffect = require("./highlevel/types/messages/message-effect.cjs");
const messageEntity = require("./highlevel/types/messages/message-entity.cjs");
const messageForward = require("./highlevel/types/messages/message-forward.cjs");
const messageMedia = require("./highlevel/types/messages/message-media.cjs");
const messageReactions = require("./highlevel/types/messages/message-reactions.cjs");
const messageReplies = require("./highlevel/types/messages/message-replies.cjs");
const message = require("./highlevel/types/messages/message.cjs");
const repliedMessage = require("./highlevel/types/messages/replied-message.cjs");
const richMessage = require("./highlevel/types/messages/rich/rich-message.cjs");
const inner$1 = require("./highlevel/types/messages/rich/inner.cjs");
const searchFilters = require("./highlevel/types/messages/search-filters.cjs");
const suggestedPost = require("./highlevel/types/messages/suggested-post.cjs");
const collectibleInfo = require("./highlevel/types/misc/collectible-info.cjs");
const entities = require("./highlevel/types/misc/entities.cjs");
const bundle = require("./highlevel/types/misc/input-privacy-rule/bundle.cjs");
const stickerSet = require("./highlevel/types/misc/sticker-set.cjs");
const takeoutSession = require("./highlevel/types/misc/takeout-session.cjs");
const botInfo = require("./highlevel/types/peers/bot-info.cjs");
const botVerification = require("./highlevel/types/peers/bot-verification.cjs");
const chatColors = require("./highlevel/types/peers/chat-colors.cjs");
const actions = require("./highlevel/types/peers/chat-event/actions.cjs");
const index = require("./highlevel/types/peers/chat-event/index.cjs");
const chatInviteLinkMember = require("./highlevel/types/peers/chat-invite-link-member.cjs");
const chatInviteLink = require("./highlevel/types/peers/chat-invite-link.cjs");
const chatMember = require("./highlevel/types/peers/chat-member.cjs");
const chatPermissions = require("./highlevel/types/peers/chat-permissions.cjs");
const chatPhoto = require("./highlevel/types/peers/chat-photo.cjs");
const chatPreview = require("./highlevel/types/peers/chat-preview.cjs");
const chat = require("./highlevel/types/peers/chat.cjs");
const chatlistPreview = require("./highlevel/types/peers/chatlist-preview.cjs");
const forumTopic = require("./highlevel/types/peers/forum-topic.cjs");
const fullChat = require("./highlevel/types/peers/full-chat.cjs");
const fullUser = require("./highlevel/types/peers/full-user.cjs");
const peerSettings = require("./highlevel/types/peers/peer-settings.cjs");
const peer = require("./highlevel/types/peers/peer.cjs");
const peersIndex = require("./highlevel/types/peers/peers-index.cjs");
const user = require("./highlevel/types/peers/user.cjs");
const boostSlot = require("./highlevel/types/premium/boost-slot.cjs");
const boostStats = require("./highlevel/types/premium/boost-stats.cjs");
const boost = require("./highlevel/types/premium/boost.cjs");
const businessAccount = require("./highlevel/types/premium/business-account.cjs");
const businessChatLink = require("./highlevel/types/premium/business-chat-link.cjs");
const businessConnection = require("./highlevel/types/premium/business-connection.cjs");
const businessIntro = require("./highlevel/types/premium/business-intro.cjs");
const businessLocation = require("./highlevel/types/premium/business-location.cjs");
const businessWorkHours = require("./highlevel/types/premium/business-work-hours.cjs");
const savedStarGift = require("./highlevel/types/premium/saved-star-gift.cjs");
const starGiftUnique = require("./highlevel/types/premium/star-gift-unique.cjs");
const starGiftValue = require("./highlevel/types/premium/star-gift-value.cjs");
const starGift = require("./highlevel/types/premium/star-gift.cjs");
const starsStatus = require("./highlevel/types/premium/stars-status.cjs");
const starsTransaction = require("./highlevel/types/premium/stars-transaction.cjs");
const emojiStatus = require("./highlevel/types/reactions/emoji-status.cjs");
const peerReaction = require("./highlevel/types/reactions/peer-reaction.cjs");
const reactionCount = require("./highlevel/types/reactions/reaction-count.cjs");
const types = require("./highlevel/types/reactions/types.cjs");
const allStories = require("./highlevel/types/stories/all-stories.cjs");
const input = require("./highlevel/types/stories/interactive/input.cjs");
const channelPost = require("./highlevel/types/stories/interactive/channel-post.cjs");
const location$1 = require("./highlevel/types/stories/interactive/location.cjs");
const reaction = require("./highlevel/types/stories/interactive/reaction.cjs");
const starGift$1 = require("./highlevel/types/stories/interactive/star-gift.cjs");
const url = require("./highlevel/types/stories/interactive/url.cjs");
const venue$1 = require("./highlevel/types/stories/interactive/venue.cjs");
const weather = require("./highlevel/types/stories/interactive/weather.cjs");
const index$1 = require("./highlevel/types/stories/interactive/index.cjs");
const peerStories = require("./highlevel/types/stories/peer-stories.cjs");
const stealthMode = require("./highlevel/types/stories/stealth-mode.cjs");
const storyInteractions = require("./highlevel/types/stories/story-interactions.cjs");
const storyViewer = require("./highlevel/types/stories/story-viewer.cjs");
const story$1 = require("./highlevel/types/stories/story.cjs");
const botChatJoinRequest = require("./highlevel/types/updates/bot-chat-join-request.cjs");
const botGuestChatQuery = require("./highlevel/types/updates/bot-guest-chat-query.cjs");
const botReaction = require("./highlevel/types/updates/bot-reaction.cjs");
const botStopped = require("./highlevel/types/updates/bot-stopped.cjs");
const callbackQuery = require("./highlevel/types/updates/callback-query.cjs");
const businessMessage = require("./highlevel/types/updates/business-message.cjs");
const chatJoinRequest = require("./highlevel/types/updates/chat-join-request.cjs");
const chatMemberUpdate = require("./highlevel/types/updates/chat-member-update.cjs");
const chosenInlineResult = require("./highlevel/types/updates/chosen-inline-result.cjs");
const deleteBusinessMessageUpdate = require("./highlevel/types/updates/delete-business-message-update.cjs");
const deleteMessageUpdate = require("./highlevel/types/updates/delete-message-update.cjs");
const deleteStoryUpdate = require("./highlevel/types/updates/delete-story-update.cjs");
const historyReadUpdate = require("./highlevel/types/updates/history-read-update.cjs");
const inlineQuery = require("./highlevel/types/updates/inline-query.cjs");
const pollUpdate = require("./highlevel/types/updates/poll-update.cjs");
const pollVote = require("./highlevel/types/updates/poll-vote.cjs");
const preCheckoutQuery = require("./highlevel/types/updates/pre-checkout-query.cjs");
const storyUpdate = require("./highlevel/types/updates/story-update.cjs");
const userStatusUpdate = require("./highlevel/types/updates/user-status-update.cjs");
const userTypingUpdate = require("./highlevel/types/updates/user-typing-update.cjs");
const manager = require("./highlevel/updates/manager.cjs");
const types$1 = require("./highlevel/updates/types.cjs");
const client = require("./network/client.cjs");
const bundle$1 = require("./network/middlewares/bundle.cjs");
const index$2 = require("./network/mtproxy/index.cjs");
const sessionConnection = require("./network/session-connection.cjs");
const abstract = require("./network/transports/abstract.cjs");
const intermediate = require("./network/transports/intermediate.cjs");
const obfuscated = require("./network/transports/obfuscated.cjs");
const proxy = require("./network/transports/proxy.cjs");
const driver = require("./storage/driver.cjs");
const driver$1 = require("./storage/memory/driver.cjs");
const authKeys = require("./storage/memory/repository/auth-keys.cjs");
const kv = require("./storage/memory/repository/kv.cjs");
const peers = require("./storage/memory/repository/peers.cjs");
const refMessages = require("./storage/memory/repository/ref-messages.cjs");
const index$3 = require("./storage/memory/index.cjs");
const driver$2 = require("./storage/sqlite/driver.cjs");
const index$4 = require("./storage/sqlite/index.cjs");
const storage$1 = require("./storage/storage.cjs");
const errors$1 = require("./types/errors.cjs");
const utils = require("./types/utils.cjs");
exports.Long = Long;
exports.getAllPeersFrom = peerUtils.getAllPeersFrom;
exports.getBarePeerId = peerUtils.getBarePeerId;
exports.getBasicPeerType = peerUtils.getBasicPeerType;
exports.getMarkedPeerId = peerUtils.getMarkedPeerId;
exports.parseMarkedPeerId = peerUtils.parseMarkedPeerId;
exports.toggleChannelIdMark = peerUtils.toggleChannelIdMark;
exports.mtp = innerMtp;
exports.tl = innerTl;
exports.TelegramStorageManager = storage.TelegramStorageManager;
exports.SentCode = sentCode.SentCode;
exports.BotCommands = inner;
exports.GameHighScore = gameHighScore.GameHighScore;
exports.BotInlineMessage = factories;
exports.BotInline = factories$1;
exports.BotKeyboardBuilder = builder.BotKeyboardBuilder;
exports.BotKeyboard = factories$2;
exports.WebviewResult = webview.WebviewResult;
exports.Conversation = conversation.Conversation;
exports.MtEmptyError = errors.MtEmptyError;
exports.MtInvalidPeerTypeError = errors.MtInvalidPeerTypeError;
exports.MtMessageNotFoundError = errors.MtMessageNotFoundError;
exports.MtPeerNotFoundError = errors.MtPeerNotFoundError;
exports.FileLocation = fileLocation.FileLocation;
exports.isUploadedFile = uploadedFile.isUploadedFile;
exports.WebDocument = webDocument.WebDocument;
exports.Audio = audio.Audio;
exports.Contact = contact.Contact;
exports.Dice = dice.Dice;
exports.Document = document.Document;
exports.RawDocument = document.RawDocument;
exports.ExtendedMediaPreview = extendedMedia.ExtendedMediaPreview;
exports.Game = game.Game;
exports.InputMedia = factories$3;
exports.Invoice = invoice.Invoice;
exports.LiveLocation = location.LiveLocation;
exports.Location = location.Location;
exports.RawLocation = location.RawLocation;
exports.PaidMedia = paidMedia.PaidMedia;
exports.Photo = photo.Photo;
exports.Poll = poll.Poll;
exports.PollAnswer = poll.PollAnswer;
exports.MASK_POSITION_POINT_TO_TL = sticker.MASK_POSITION_POINT_TO_TL;
exports.Sticker = sticker.Sticker;
exports.MediaStory = story.MediaStory;
exports.Thumbnail = thumbnail.Thumbnail;
exports.TodoItem = todo.TodoItem;
exports.TodoList = todo.TodoList;
exports.Venue = venue.Venue;
exports.Video = video.Video;
exports.Voice = voice.Voice;
exports.WebPage = webPage.WebPage;
exports.WebPageMedia = webPage.WebPageMedia;
exports.Dialog = dialog.Dialog;
exports.DraftMessage = draftMessage.DraftMessage;
exports.FactCheck = factCheck.FactCheck;
exports.normalizeInputMessageId = inputMessageId.normalizeInputMessageId;
exports._messageActionFromTl = messageAction._messageActionFromTl;
exports.MessageEffect = messageEffect.MessageEffect;
exports.MessageEntity = messageEntity.MessageEntity;
exports.MessageForwardInfo = messageForward.MessageForwardInfo;
exports._messageMediaFromTl = messageMedia._messageMediaFromTl;
exports.MessageReactions = messageReactions.MessageReactions;
exports.MessageRepliesInfo = messageReplies.MessageRepliesInfo;
exports.Message = message.Message;
exports.RepliedMessageInfo = repliedMessage.RepliedMessageInfo;
exports.RichMessage = richMessage.RichMessage;
exports.Rich = inner$1;
exports.SearchFilters = searchFilters.SearchFilters;
exports.SuggestedPostInfo = suggestedPost.SuggestedPostInfo;
exports.CollectibleInfo = collectibleInfo.CollectibleInfo;
exports.inputTextToTl = entities.inputTextToTl;
exports.PrivacyRule = bundle;
exports.StickerSet = stickerSet.StickerSet;
exports.normalizeInputStickerSet = stickerSet.normalizeInputStickerSet;
exports.TakeoutSession = takeoutSession.TakeoutSession;
exports.BotInfo = botInfo.BotInfo;
exports.BotVerification = botVerification.BotVerification;
exports.ChatColors = chatColors.ChatColors;
exports._actionFromTl = actions._actionFromTl;
exports.ChatEvent = index.ChatEvent;
exports.ChatInviteLinkMember = chatInviteLinkMember.ChatInviteLinkMember;
exports.ChatInviteLink = chatInviteLink.ChatInviteLink;
exports.ChatMember = chatMember.ChatMember;
exports.ChatPermissions = chatPermissions.ChatPermissions;
exports.ChatPhoto = chatPhoto.ChatPhoto;
exports.ChatPhotoSize = chatPhoto.ChatPhotoSize;
exports.ChatPreview = chatPreview.ChatPreview;
exports.Chat = chat.Chat;
exports.ChatlistPreview = chatlistPreview.ChatlistPreview;
exports.ForumTopic = forumTopic.ForumTopic;
exports.FullChat = fullChat.FullChat;
exports.FullUser = fullUser.FullUser;
exports.PeerSettings = peerSettings.PeerSettings;
exports.parsePeer = peer.parsePeer;
exports.PeersIndex = peersIndex.PeersIndex;
exports.User = user.User;
exports.BoostSlot = boostSlot.BoostSlot;
exports.BoostStats = boostStats.BoostStats;
exports.Boost = boost.Boost;
exports.BusinessAccount = businessAccount.BusinessAccount;
exports.BusinessChatLink = businessChatLink.BusinessChatLink;
exports.BusinessConnection = businessConnection.BusinessConnection;
exports.BusinessIntro = businessIntro.BusinessIntro;
exports.BusinessLocation = businessLocation.BusinessLocation;
exports.BusinessWorkHours = businessWorkHours.BusinessWorkHours;
exports.businessWorkHoursDaysToRaw = businessWorkHours.businessWorkHoursDaysToRaw;
exports.SavedStarGift = savedStarGift.SavedStarGift;
exports.StarGiftUnique = starGiftUnique.StarGiftUnique;
exports.StarGiftUniqueAttribute = starGiftUnique.StarGiftUniqueAttribute;
exports.StarGiftUniqueBackdrop = starGiftUnique.StarGiftUniqueBackdrop;
exports.StarGiftUniqueOriginalDetails = starGiftUnique.StarGiftUniqueOriginalDetails;
exports.StarGiftValue = starGiftValue.StarGiftValue;
exports.StarGift = starGift.StarGift;
exports.StarsStatus = starsStatus.StarsStatus;
exports.StarsTransaction = starsTransaction.StarsTransaction;
exports.EmojiStatus = emojiStatus.EmojiStatus;
exports.EmojiStatusCollectible = emojiStatus.EmojiStatusCollectible;
exports.PaidPeerReaction = peerReaction.PaidPeerReaction;
exports.PeerReaction = peerReaction.PeerReaction;
exports.ReactionCount = reactionCount.ReactionCount;
exports.normalizeInputReaction = types.normalizeInputReaction;
exports.toReactionEmoji = types.toReactionEmoji;
exports.AllStories = allStories.AllStories;
exports.StoryElement = input.StoryElement;
exports.StoryInteractiveChannelPost = channelPost.StoryInteractiveChannelPost;
exports.StoryInteractiveLocation = location$1.StoryInteractiveLocation;
exports.StoryInteractiveReaction = reaction.StoryInteractiveReaction;
exports.StoryInteractiveStarGift = starGift$1.StoryInteractiveStarGift;
exports.StoryInteractiveUrl = url.StoryInteractiveUrl;
exports.StoryInteractiveVenue = venue$1.StoryInteractiveVenue;
exports.StoryInteractiveWeather = weather.StoryInteractiveWeather;
exports._storyInteractiveElementFromTl = index$1._storyInteractiveElementFromTl;
exports.PeerStories = peerStories.PeerStories;
exports.StoriesStealthMode = stealthMode.StoriesStealthMode;
exports.StoryInteractions = storyInteractions.StoryInteractions;
exports.StoryRepost = storyViewer.StoryRepost;
exports.StoryViewer = storyViewer.StoryViewer;
exports.StoryViewersList = storyViewer.StoryViewersList;
exports.Story = story$1.Story;
exports.BotChatJoinRequestUpdate = botChatJoinRequest.BotChatJoinRequestUpdate;
exports.BotGuestChatQuery = botGuestChatQuery.BotGuestChatQuery;
exports.BotReactionCountUpdate = botReaction.BotReactionCountUpdate;
exports.BotReactionUpdate = botReaction.BotReactionUpdate;
exports.BotStoppedUpdate = botStopped.BotStoppedUpdate;
exports.BusinessCallbackQuery = callbackQuery.BusinessCallbackQuery;
exports.CallbackQuery = callbackQuery.CallbackQuery;
exports.InlineCallbackQuery = callbackQuery.InlineCallbackQuery;
exports.BusinessMessage = businessMessage.BusinessMessage;
exports.ChatJoinRequestUpdate = chatJoinRequest.ChatJoinRequestUpdate;
exports.ChatMemberUpdate = chatMemberUpdate.ChatMemberUpdate;
exports.ChosenInlineResult = chosenInlineResult.ChosenInlineResult;
exports.DeleteBusinessMessageUpdate = deleteBusinessMessageUpdate.DeleteBusinessMessageUpdate;
exports.DeleteMessageUpdate = deleteMessageUpdate.DeleteMessageUpdate;
exports.DeleteStoryUpdate = deleteStoryUpdate.DeleteStoryUpdate;
exports.HistoryReadUpdate = historyReadUpdate.HistoryReadUpdate;
exports.InlineQuery = inlineQuery.InlineQuery;
exports.PollUpdate = pollUpdate.PollUpdate;
exports.PollVoteUpdate = pollVote.PollVoteUpdate;
exports.PreCheckoutQuery = preCheckoutQuery.PreCheckoutQuery;
exports.StoryUpdate = storyUpdate.StoryUpdate;
exports.UserStatusUpdate = userStatusUpdate.UserStatusUpdate;
exports.UserTypingUpdate = userTypingUpdate.UserTypingUpdate;
exports.UpdatesManager = manager.UpdatesManager;
exports.RawUpdateInfo = types$1.RawUpdateInfo;
exports.MtClient = client.MtClient;
exports.networkMiddlewares = bundle$1;
exports.BaseMtProxyTransport = index$2.BaseMtProxyTransport;
exports.SessionConnection = sessionConnection.SessionConnection;
exports.TransportError = abstract.TransportError;
exports.IntermediatePacketCodec = intermediate.IntermediatePacketCodec;
exports.PaddedIntermediatePacketCodec = intermediate.PaddedIntermediatePacketCodec;
exports.ObfuscatedPacketCodec = obfuscated.ObfuscatedPacketCodec;
exports.createProxyTransportFactory = proxy.createProxyTransportFactory;
exports.BaseStorageDriver = driver.BaseStorageDriver;
exports.MemoryStorageDriver = driver$1.MemoryStorageDriver;
exports.MemoryAuthKeysRepository = authKeys.MemoryAuthKeysRepository;
exports.MemoryKeyValueRepository = kv.MemoryKeyValueRepository;
exports.MemoryPeersRepository = peers.MemoryPeersRepository;
exports.MemoryRefMessagesRepository = refMessages.MemoryRefMessagesRepository;
exports.MemoryStorage = index$3.MemoryStorage;
exports.BaseSqliteStorageDriver = driver$2.BaseSqliteStorageDriver;
exports.BaseSqliteStorage = index$4.BaseSqliteStorage;
exports.StorageManager = storage$1.StorageManager;
exports.MtArgumentError = errors$1.MtArgumentError;
exports.MtSecurityError = errors$1.MtSecurityError;
exports.MtTimeoutError = errors$1.MtTimeoutError;
exports.MtTypeAssertionError = errors$1.MtTypeAssertionError;
exports.MtUnsupportedError = errors$1.MtUnsupportedError;
exports.MtcuteError = errors$1.MtcuteError;
exports.assertNever = utils.assertNever;
