// @mtcute/core/methods.js
import { checkPassword } from "./highlevel/methods/auth/check-password.js";
import { getPasswordHint } from "./highlevel/methods/auth/get-password-hint.js";
import { logOut } from "./highlevel/methods/auth/log-out.js";
import { recoverPassword } from "./highlevel/methods/auth/recover-password.js";
import { resendCode } from "./highlevel/methods/auth/resend-code.js";
import { sendCode } from "./highlevel/methods/auth/send-code.js";
import { sendRecoveryCode } from "./highlevel/methods/auth/send-recovery-code.js";
import { signInBot } from "./highlevel/methods/auth/sign-in-bot.js";
import { signInQr } from "./highlevel/methods/auth/sign-in-qr.js";
import { signIn } from "./highlevel/methods/auth/sign-in.js";
import { startTest } from "./highlevel/methods/auth/start-test.js";
import { start } from "./highlevel/methods/auth/start.js";
import { isSelfPeer } from "./highlevel/methods/auth/utils.js";
import { answerBotGuestChatQuery } from "./highlevel/methods/bots/answer-bot-guest-chat-query.js";
import { answerCallbackQuery } from "./highlevel/methods/bots/answer-callback-query.js";
import { answerInlineQuery } from "./highlevel/methods/bots/answer-inline-query.js";
import { answerPreCheckoutQuery } from "./highlevel/methods/bots/answer-pre-checkout-query.js";
import { deleteMyCommands } from "./highlevel/methods/bots/delete-my-commands.js";
import { getBotInfo } from "./highlevel/methods/bots/get-bot-info.js";
import { getBotMenuButton } from "./highlevel/methods/bots/get-bot-menu-button.js";
import { getCallbackAnswer } from "./highlevel/methods/bots/get-callback-answer.js";
import { getGameHighScores, getInlineGameHighScores } from "./highlevel/methods/bots/get-game-high-scores.js";
import { getMyCommands } from "./highlevel/methods/bots/get-my-commands.js";
import { closeWebview, openWebview } from "./highlevel/methods/bots/open-webview.js";
import { prepareInlineMessage } from "./highlevel/methods/bots/prepare-inline-message.js";
import { setBotInfo } from "./highlevel/methods/bots/set-bot-info.js";
import { setBotMenuButton } from "./highlevel/methods/bots/set-bot-menu-button.js";
import { setGameScore, setInlineGameScore } from "./highlevel/methods/bots/set-game-score.js";
import { setMyCommands } from "./highlevel/methods/bots/set-my-commands.js";
import { setMyDefaultRights } from "./highlevel/methods/bots/set-my-default-rights.js";
import { toggleEmojiStatusPermission } from "./highlevel/methods/bots/toggle-emoji-status-permission.js";
import { addChatMembers } from "./highlevel/methods/chats/add-chat-members.js";
import { archiveChats } from "./highlevel/methods/chats/archive-chats.js";
import { banChatMember } from "./highlevel/methods/chats/ban-chat-member.js";
import { createChannel } from "./highlevel/methods/chats/create-channel.js";
import { createGroup } from "./highlevel/methods/chats/create-group.js";
import { createSupergroup } from "./highlevel/methods/chats/create-supergroup.js";
import { deleteChannel } from "./highlevel/methods/chats/delete-channel.js";
import { deleteChatPhoto } from "./highlevel/methods/chats/delete-chat-photo.js";
import { deleteGroup } from "./highlevel/methods/chats/delete-group.js";
import { deleteHistory } from "./highlevel/methods/chats/delete-history.js";
import { deleteUserHistory } from "./highlevel/methods/chats/delete-user-history.js";
import { editAdminRights } from "./highlevel/methods/chats/edit-admin-rights.js";
import { editChatMemberRank } from "./highlevel/methods/chats/edit-chat-member-rank.js";
import { getChatEventLog } from "./highlevel/methods/chats/get-chat-event-log.js";
import { getChatMember } from "./highlevel/methods/chats/get-chat-member.js";
import { getChatMembers } from "./highlevel/methods/chats/get-chat-members.js";
import { getChatPreview } from "./highlevel/methods/chats/get-chat-preview.js";
import { getChat } from "./highlevel/methods/chats/get-chat.js";
import { getChats } from "./highlevel/methods/chats/get-chats.js";
import { getCreatorAfterLeave } from "./highlevel/methods/chats/get-creator-after-leave.js";
import { getFullChat } from "./highlevel/methods/chats/get-full-chat.js";
import { getFullUser } from "./highlevel/methods/chats/get-full-user.js";
import { getMessageAuthor } from "./highlevel/methods/chats/get-message-author.js";
import { getPeer } from "./highlevel/methods/chats/get-peer.js";
import { getPeers } from "./highlevel/methods/chats/get-peers.js";
import { getSimilarChannels } from "./highlevel/methods/chats/get-similar-channels.js";
import { getUser } from "./highlevel/methods/chats/get-user.js";
import { iterChatEventLog } from "./highlevel/methods/chats/iter-chat-event-log.js";
import { iterChatMembers } from "./highlevel/methods/chats/iter-chat-members.js";
import { joinChat } from "./highlevel/methods/chats/join-chat.js";
import { kickChatMember } from "./highlevel/methods/chats/kick-chat-member.js";
import { leaveChat } from "./highlevel/methods/chats/leave-chat.js";
import { markChatUnread } from "./highlevel/methods/chats/mark-chat-unread.js";
import { closeChat, openChat } from "./highlevel/methods/chats/open-chat.js";
import { reorderUsernames } from "./highlevel/methods/chats/reorder-usernames.js";
import { restrictChatMember } from "./highlevel/methods/chats/restrict-chat-member.js";
import { saveDraft } from "./highlevel/methods/chats/save-draft.js";
import { setChatColor } from "./highlevel/methods/chats/set-chat-color.js";
import { setChatDefaultPermissions } from "./highlevel/methods/chats/set-chat-default-permissions.js";
import { setChatDescription } from "./highlevel/methods/chats/set-chat-description.js";
import { setChatPhoto } from "./highlevel/methods/chats/set-chat-photo.js";
import { setChatTitle } from "./highlevel/methods/chats/set-chat-title.js";
import { setChatTtl } from "./highlevel/methods/chats/set-chat-ttl.js";
import { setChatUsername } from "./highlevel/methods/chats/set-chat-username.js";
import { setSlowMode } from "./highlevel/methods/chats/set-slow-mode.js";
import { toggleContentProtection } from "./highlevel/methods/chats/toggle-content-protection.js";
import { toggleFragmentUsername } from "./highlevel/methods/chats/toggle-fragment-username.js";
import { toggleJoinRequests } from "./highlevel/methods/chats/toggle-join-requests.js";
import { toggleJoinToSend } from "./highlevel/methods/chats/toggle-join-to-send.js";
import { transferChatOwnership } from "./highlevel/methods/chats/transfer-chat-ownership.js";
import { unarchiveChats } from "./highlevel/methods/chats/unarchive-chats.js";
import { unbanChatMember } from "./highlevel/methods/chats/unban-chat-member.js";
import { addContact } from "./highlevel/methods/contacts/add-contact.js";
import { deleteContacts } from "./highlevel/methods/contacts/delete-contacts.js";
import { getContacts } from "./highlevel/methods/contacts/get-contacts.js";
import { importContacts } from "./highlevel/methods/contacts/import-contacts.js";
import { setContactNote } from "./highlevel/methods/contacts/set-contact-note.js";
import { createFolder } from "./highlevel/methods/dialogs/create-folder.js";
import { deleteFolder } from "./highlevel/methods/dialogs/delete-folder.js";
import { editFolder } from "./highlevel/methods/dialogs/edit-folder.js";
import { findDialogs } from "./highlevel/methods/dialogs/find-dialogs.js";
import { findFolder } from "./highlevel/methods/dialogs/find-folder.js";
import { getChatlistPreview } from "./highlevel/methods/dialogs/get-chatlist-preview.js";
import { getFolders } from "./highlevel/methods/dialogs/get-folders.js";
import { getPeerDialogs } from "./highlevel/methods/dialogs/get-peer-dialogs.js";
import { iterDialogs } from "./highlevel/methods/dialogs/iter-dialogs.js";
import { joinChatlist } from "./highlevel/methods/dialogs/join-chatlist.js";
import { setFoldersOrder } from "./highlevel/methods/dialogs/set-folders-order.js";
import { downloadAsBuffer } from "./highlevel/methods/files/download-buffer.js";
import { downloadChunk } from "./highlevel/methods/files/download-chunk.js";
import { _normalizeFileDownloadLocation, downloadAsIterable } from "./highlevel/methods/files/download-iterable.js";
import { downloadAsStream } from "./highlevel/methods/files/download-stream.js";
import { _normalizeInputFile } from "./highlevel/methods/files/normalize-input-file.js";
import { _normalizeInputMedia } from "./highlevel/methods/files/normalize-input-media.js";
import { uploadFile } from "./highlevel/methods/files/upload-file.js";
import { uploadMedia } from "./highlevel/methods/files/upload-media.js";
import { createForumTopic } from "./highlevel/methods/forums/create-forum-topic.js";
import { deleteForumTopicHistory } from "./highlevel/methods/forums/delete-forum-topic-history.js";
import { editForumTopic } from "./highlevel/methods/forums/edit-forum-topic.js";
import { getForumTopicsById } from "./highlevel/methods/forums/get-forum-topics-by-id.js";
import { getForumTopics } from "./highlevel/methods/forums/get-forum-topics.js";
import { iterForumTopics } from "./highlevel/methods/forums/iter-forum-topics.js";
import { reorderPinnedForumTopics } from "./highlevel/methods/forums/reorder-pinned-forum-topics.js";
import { toggleForumTopicClosed } from "./highlevel/methods/forums/toggle-forum-topic-closed.js";
import { toggleForumTopicPinned } from "./highlevel/methods/forums/toggle-forum-topic-pinned.js";
import { updateForumSettings } from "./highlevel/methods/forums/toggle-forum.js";
import { toggleGeneralTopicHidden } from "./highlevel/methods/forums/toggle-general-topic-hidden.js";
import { _normalizeInputStarGift } from "./highlevel/methods/gifts/_normalize-input-star-gift.js";
import { acceptStarGift } from "./highlevel/methods/gifts/accept-star-gift.js";
import { buyResaleGift } from "./highlevel/methods/gifts/buy-resale-gift.js";
import { getStarGiftResaleOptions } from "./highlevel/methods/gifts/get-resale-star-gifts.js";
import { getSavedStarGiftsById } from "./highlevel/methods/gifts/get-saved-star-gifts-by-id.js";
import { getSavedStarGifts } from "./highlevel/methods/gifts/get-saved-star-gifts.js";
import { getStarGiftOptions } from "./highlevel/methods/gifts/get-star-gift-options.js";
import { getStarGiftUpgradeOptions } from "./highlevel/methods/gifts/get-star-gift-upgrade-options.js";
import { getStarGiftValue } from "./highlevel/methods/gifts/get-star-gift-value.js";
import { getStarGiftWithdrawalUrl } from "./highlevel/methods/gifts/get-star-gift-withdrawal-url.js";
import { getUniqueStarGift } from "./highlevel/methods/gifts/get-unique-star-gift.js";
import { iterSavedStarGifts } from "./highlevel/methods/gifts/iter-saved-star-gifts.js";
import { togglePinnedStarGifts } from "./highlevel/methods/gifts/pin-star-gift.js";
import { prepayStarGiftUpgrade } from "./highlevel/methods/gifts/prepay-star-gift-upgrade.js";
import { resolveStarGiftOffer } from "./highlevel/methods/gifts/resolve-star-gift-offer.js";
import { sendStarGiftOffer } from "./highlevel/methods/gifts/send-star-gift-offer.js";
import { sendStarGift } from "./highlevel/methods/gifts/send-star-gift.js";
import { setResaleStarGiftPrice } from "./highlevel/methods/gifts/set-gift-resale-price.js";
import { transferStarGift } from "./highlevel/methods/gifts/transfer-star-gift.js";
import { upgradeStarGift } from "./highlevel/methods/gifts/upgrade-star-gift.js";
import { createInviteLink } from "./highlevel/methods/invite-links/create-invite-link.js";
import { editInviteLink } from "./highlevel/methods/invite-links/edit-invite-link.js";
import { exportInviteLink } from "./highlevel/methods/invite-links/export-invite-link.js";
import { getInviteLinkMembers } from "./highlevel/methods/invite-links/get-invite-link-members.js";
import { getInviteLink } from "./highlevel/methods/invite-links/get-invite-link.js";
import { getInviteLinks } from "./highlevel/methods/invite-links/get-invite-links.js";
import { getPrimaryInviteLink } from "./highlevel/methods/invite-links/get-primary-invite-link.js";
import { hideAllJoinRequests } from "./highlevel/methods/invite-links/hide-all-join-requests.js";
import { hideJoinRequest } from "./highlevel/methods/invite-links/hide-join-request.js";
import { iterInviteLinkMembers } from "./highlevel/methods/invite-links/iter-invite-link-members.js";
import { iterInviteLinks } from "./highlevel/methods/invite-links/iter-invite-links.js";
import { revokeInviteLink } from "./highlevel/methods/invite-links/revoke-invite-link.js";
import { appendTodoList } from "./highlevel/methods/messages/append-todo-list.js";
import { closePoll } from "./highlevel/methods/messages/close-poll.js";
import { createRichStreamingDraft } from "./highlevel/methods/messages/create-rich-streaming-draft.js";
import { createStreamingDraft } from "./highlevel/methods/messages/create-streaming-draft.js";
import { deleteMessages, deleteMessagesById } from "./highlevel/methods/messages/delete-messages.js";
import { deleteScheduledMessages } from "./highlevel/methods/messages/delete-scheduled-messages.js";
import { editInlineMessage } from "./highlevel/methods/messages/edit-inline-message.js";
import { editMessage } from "./highlevel/methods/messages/edit-message.js";
import { forwardMessages, forwardMessagesById } from "./highlevel/methods/messages/forward-messages.js";
import { getAllScheduledMessages } from "./highlevel/methods/messages/get-all-scheduled-messages.js";
import { getAvailableMessageEffects } from "./highlevel/methods/messages/get-available-effects.js";
import { getCallbackQueryMessage } from "./highlevel/methods/messages/get-callback-query-message.js";
import { getDiscussionMessage } from "./highlevel/methods/messages/get-discussion-message.js";
import { getFactCheck } from "./highlevel/methods/messages/get-fact-check.js";
import { getHistory } from "./highlevel/methods/messages/get-history.js";
import { getMessageByLink } from "./highlevel/methods/messages/get-message-by-link.js";
import { getMessageGroup } from "./highlevel/methods/messages/get-message-group.js";
import { getMessageReactions, getMessageReactionsById } from "./highlevel/methods/messages/get-message-reactions.js";
import { getMessagesUnsafe } from "./highlevel/methods/messages/get-messages-unsafe.js";
import { getMessages } from "./highlevel/methods/messages/get-messages.js";
import { getReactionUsers } from "./highlevel/methods/messages/get-reaction-users.js";
import { getReplyTo } from "./highlevel/methods/messages/get-reply-to.js";
import { getScheduledMessages } from "./highlevel/methods/messages/get-scheduled-messages.js";
import { getWebPagePreview } from "./highlevel/methods/messages/get-web-page-preview.js";
import { iterHistory } from "./highlevel/methods/messages/iter-history.js";
import { iterReactionUsers } from "./highlevel/methods/messages/iter-reaction-users.js";
import { iterSearchGlobal } from "./highlevel/methods/messages/iter-search-global.js";
import { iterSearchMessages } from "./highlevel/methods/messages/iter-search-messages.js";
import { pinMessage } from "./highlevel/methods/messages/pin-message.js";
import { readHistory } from "./highlevel/methods/messages/read-history.js";
import { readReactions } from "./highlevel/methods/messages/read-reactions.js";
import { searchGlobal } from "./highlevel/methods/messages/search-global.js";
import { iterSearchHashtag, searchHashtag } from "./highlevel/methods/messages/search-hashtag.js";
import { searchMessages } from "./highlevel/methods/messages/search-messages.js";
import { searchPostsGlobal } from "./highlevel/methods/messages/search-posts-global.js";
import { answerMedia, answerMediaGroup, answerText } from "./highlevel/methods/messages/send-answer.js";
import { commentMedia, commentMediaGroup, commentText } from "./highlevel/methods/messages/send-comment.js";
import { sendCopyGroup } from "./highlevel/methods/messages/send-copy-group.js";
import { sendCopy } from "./highlevel/methods/messages/send-copy.js";
import { sendMediaGroup } from "./highlevel/methods/messages/send-media-group.js";
import { sendMedia } from "./highlevel/methods/messages/send-media.js";
import { sendPaidReaction } from "./highlevel/methods/messages/send-paid-reaction.js";
import { quoteWithMedia, quoteWithMediaGroup, quoteWithText } from "./highlevel/methods/messages/send-quote.js";
import { sendReaction } from "./highlevel/methods/messages/send-reaction.js";
import { replyMedia, replyMediaGroup, replyText } from "./highlevel/methods/messages/send-reply.js";
import { sendRichMessage } from "./highlevel/methods/messages/send-rich-message.js";
import { sendScheduled } from "./highlevel/methods/messages/send-scheduled.js";
import { sendText } from "./highlevel/methods/messages/send-text.js";
import { sendTyping } from "./highlevel/methods/messages/send-typing.js";
import { sendVote } from "./highlevel/methods/messages/send-vote.js";
import { setTyping } from "./highlevel/methods/messages/set-typing.js";
import { toggleTodoCompleted } from "./highlevel/methods/messages/toggle-todo-completed.js";
import { translateMessage } from "./highlevel/methods/messages/translate-message.js";
import { translateText } from "./highlevel/methods/messages/translate-text.js";
import { unpinAllMessages } from "./highlevel/methods/messages/unpin-all-messages.js";
import { unpinMessage } from "./highlevel/methods/messages/unpin-message.js";
import { getCollectibleInfo } from "./highlevel/methods/misc/get-collectible-info.js";
import { initTakeoutSession } from "./highlevel/methods/misc/init-takeout-session.js";
import { _normalizePrivacyRules } from "./highlevel/methods/misc/normalize-privacy-rules.js";
import { changeCloudPassword } from "./highlevel/methods/password/change-cloud-password.js";
import { enableCloudPassword } from "./highlevel/methods/password/enable-cloud-password.js";
import { cancelPasswordEmail, resendPasswordEmail, verifyPasswordEmail } from "./highlevel/methods/password/password-email.js";
import { removeCloudPassword } from "./highlevel/methods/password/remove-cloud-password.js";
import { applyBoost } from "./highlevel/methods/premium/apply-boost.js";
import { canApplyBoost } from "./highlevel/methods/premium/can-apply-boost.js";
import { createBusinessChatLink } from "./highlevel/methods/premium/create-business-chat-link.js";
import { deleteBusinessChatLink, editBusinessChatLink } from "./highlevel/methods/premium/edit-business-chat-link.js";
import { getBoostStats } from "./highlevel/methods/premium/get-boost-stats.js";
import { getBoosts } from "./highlevel/methods/premium/get-boosts.js";
import { getBusinessChatLinks } from "./highlevel/methods/premium/get-business-chat-links.js";
import { getBusinessConnection } from "./highlevel/methods/premium/get-business-connection.js";
import { getMyBoostSlots } from "./highlevel/methods/premium/get-my-boost-slots.js";
import { getStarsTransactions } from "./highlevel/methods/premium/get-stars-transactions.js";
import { iterBoosters } from "./highlevel/methods/premium/iter-boosters.js";
import { iterStarsTransactions } from "./highlevel/methods/premium/iter-stars-transactions.js";
import { setBusinessIntro } from "./highlevel/methods/premium/set-business-intro.js";
import { setBusinessWorkHours } from "./highlevel/methods/premium/set-business-work-hours.js";
import { addStickerToSet } from "./highlevel/methods/stickers/add-sticker-to-set.js";
import { createStickerSet } from "./highlevel/methods/stickers/create-sticker-set.js";
import { deleteStickerFromSet } from "./highlevel/methods/stickers/delete-sticker-from-set.js";
import { getCustomEmojis, getCustomEmojisFromMessages } from "./highlevel/methods/stickers/get-custom-emojis.js";
import { getInstalledStickers } from "./highlevel/methods/stickers/get-installed-stickers.js";
import { getMyStickerSets } from "./highlevel/methods/stickers/get-my-sticker-sets.js";
import { getStickerSet } from "./highlevel/methods/stickers/get-sticker-set.js";
import { moveStickerInSet } from "./highlevel/methods/stickers/move-sticker-in-set.js";
import { replaceStickerInSet } from "./highlevel/methods/stickers/replace-sticker-in-set.js";
import { setChatStickerSet } from "./highlevel/methods/stickers/set-chat-sticker-set.js";
import { setStickerSetThumb } from "./highlevel/methods/stickers/set-sticker-set-thumb.js";
import { canSendStory } from "./highlevel/methods/stories/can-send-story.js";
import { deleteStories } from "./highlevel/methods/stories/delete-stories.js";
import { editStory } from "./highlevel/methods/stories/edit-story.js";
import { getAllStories } from "./highlevel/methods/stories/get-all-stories.js";
import { getPeerStories } from "./highlevel/methods/stories/get-peer-stories.js";
import { getProfileStories } from "./highlevel/methods/stories/get-profile-stories.js";
import { getStoriesById } from "./highlevel/methods/stories/get-stories-by-id.js";
import { getStoriesInteractions } from "./highlevel/methods/stories/get-stories-interactions.js";
import { getStoryLink } from "./highlevel/methods/stories/get-story-link.js";
import { getStoryViewers } from "./highlevel/methods/stories/get-story-viewers.js";
import { hideMyStoriesViews } from "./highlevel/methods/stories/hide-my-stories-views.js";
import { incrementStoriesViews } from "./highlevel/methods/stories/increment-stories-views.js";
import { iterAllStories } from "./highlevel/methods/stories/iter-all-stories.js";
import { iterProfileStories } from "./highlevel/methods/stories/iter-profile-stories.js";
import { iterStoryViewers } from "./highlevel/methods/stories/iter-story-viewers.js";
import { readStories } from "./highlevel/methods/stories/read-stories.js";
import { sendStoryReaction } from "./highlevel/methods/stories/send-story-reaction.js";
import { sendStory } from "./highlevel/methods/stories/send-story.js";
import { togglePeerStoriesArchived } from "./highlevel/methods/stories/toggle-peer-stories-archived.js";
import { toggleStoriesPinned } from "./highlevel/methods/stories/toggle-stories-pinned.js";
import { blockUser } from "./highlevel/methods/users/block-user.js";
import { deleteProfilePhotos } from "./highlevel/methods/users/delete-profile-photos.js";
import { editCloseFriends, editCloseFriendsRaw } from "./highlevel/methods/users/edit-close-friends.js";
import { getCommonChats } from "./highlevel/methods/users/get-common-chats.js";
import { getGlobalTtl } from "./highlevel/methods/users/get-global-ttl.js";
import { getMe } from "./highlevel/methods/users/get-me.js";
import { getMyUsername } from "./highlevel/methods/users/get-my-username.js";
import { getPeerSettings } from "./highlevel/methods/users/get-peer-settings.js";
import { getProfilePhoto } from "./highlevel/methods/users/get-profile-photo.js";
import { getProfilePhotos } from "./highlevel/methods/users/get-profile-photos.js";
import { getSavedMusic } from "./highlevel/methods/users/get-saved-music.js";
import { getUsers } from "./highlevel/methods/users/get-users.js";
import { isPeerAvailable } from "./highlevel/methods/users/is-peer-available.js";
import { iterProfilePhotos } from "./highlevel/methods/users/iter-profile-photos.js";
import { resolvePeerMany } from "./highlevel/methods/users/resolve-peer-many.js";
import { resolveChannel, resolvePeer, resolveUser } from "./highlevel/methods/users/resolve-peer.js";
import { resolvePhoneNumber } from "./highlevel/methods/users/resolve-phone-number.js";
import { saveMusicToProfile, unsaveMusicFromProfile } from "./highlevel/methods/users/save-music-to-profile.js";
import { setEmojiStatus } from "./highlevel/methods/users/set-emoji-status.js";
import { setGlobalTtl } from "./highlevel/methods/users/set-global-ttl.js";
import { setMyBirthday } from "./highlevel/methods/users/set-my-birthday.js";
import { setMyProfilePhoto } from "./highlevel/methods/users/set-my-profile-photo.js";
import { setMyUsername } from "./highlevel/methods/users/set-my-username.js";
import { sendOnline, setOnline } from "./highlevel/methods/users/set-online.js";
import { unblockUser } from "./highlevel/methods/users/unblock-user.js";
import { updateProfile } from "./highlevel/methods/users/update-profile.js";
export {
  _normalizeFileDownloadLocation,
  _normalizeInputFile,
  _normalizeInputMedia,
  _normalizeInputStarGift,
  _normalizePrivacyRules,
  acceptStarGift,
  addChatMembers,
  addContact,
  addStickerToSet,
  answerBotGuestChatQuery,
  answerCallbackQuery,
  answerInlineQuery,
  answerMedia,
  answerMediaGroup,
  answerPreCheckoutQuery,
  answerText,
  appendTodoList,
  applyBoost,
  archiveChats,
  banChatMember,
  blockUser,
  buyResaleGift,
  canApplyBoost,
  canSendStory,
  cancelPasswordEmail,
  changeCloudPassword,
  checkPassword,
  closeChat,
  closePoll,
  closeWebview,
  commentMedia,
  commentMediaGroup,
  commentText,
  createBusinessChatLink,
  createChannel,
  createFolder,
  createForumTopic,
  createGroup,
  createInviteLink,
  createRichStreamingDraft,
  createStickerSet,
  createStreamingDraft,
  createSupergroup,
  deleteBusinessChatLink,
  deleteChannel,
  deleteChatPhoto,
  deleteContacts,
  deleteFolder,
  deleteForumTopicHistory,
  deleteGroup,
  deleteHistory,
  deleteMessages,
  deleteMessagesById,
  deleteMyCommands,
  deleteProfilePhotos,
  deleteScheduledMessages,
  deleteStickerFromSet,
  deleteStories,
  deleteUserHistory,
  downloadAsBuffer,
  downloadAsIterable,
  downloadAsStream,
  downloadChunk,
  editAdminRights,
  editBusinessChatLink,
  editChatMemberRank,
  editCloseFriends,
  editCloseFriendsRaw,
  editFolder,
  editForumTopic,
  editInlineMessage,
  editInviteLink,
  editMessage,
  editStory,
  enableCloudPassword,
  exportInviteLink,
  findDialogs,
  findFolder,
  forwardMessages,
  forwardMessagesById,
  getAllScheduledMessages,
  getAllStories,
  getAvailableMessageEffects,
  getBoostStats,
  getBoosts,
  getBotInfo,
  getBotMenuButton,
  getBusinessChatLinks,
  getBusinessConnection,
  getCallbackAnswer,
  getCallbackQueryMessage,
  getChat,
  getChatEventLog,
  getChatMember,
  getChatMembers,
  getChatPreview,
  getChatlistPreview,
  getChats,
  getCollectibleInfo,
  getCommonChats,
  getContacts,
  getCreatorAfterLeave,
  getCustomEmojis,
  getCustomEmojisFromMessages,
  getDiscussionMessage,
  getFactCheck,
  getFolders,
  getForumTopics,
  getForumTopicsById,
  getFullChat,
  getFullUser,
  getGameHighScores,
  getGlobalTtl,
  getHistory,
  getInlineGameHighScores,
  getInstalledStickers,
  getInviteLink,
  getInviteLinkMembers,
  getInviteLinks,
  getMe,
  getMessageAuthor,
  getMessageByLink,
  getMessageGroup,
  getMessageReactions,
  getMessageReactionsById,
  getMessages,
  getMessagesUnsafe,
  getMyBoostSlots,
  getMyCommands,
  getMyStickerSets,
  getMyUsername,
  getPasswordHint,
  getPeer,
  getPeerDialogs,
  getPeerSettings,
  getPeerStories,
  getPeers,
  getPrimaryInviteLink,
  getProfilePhoto,
  getProfilePhotos,
  getProfileStories,
  getReactionUsers,
  getReplyTo,
  getSavedMusic,
  getSavedStarGifts,
  getSavedStarGiftsById,
  getScheduledMessages,
  getSimilarChannels,
  getStarGiftOptions,
  getStarGiftResaleOptions,
  getStarGiftUpgradeOptions,
  getStarGiftValue,
  getStarGiftWithdrawalUrl,
  getStarsTransactions,
  getStickerSet,
  getStoriesById,
  getStoriesInteractions,
  getStoryLink,
  getStoryViewers,
  getUniqueStarGift,
  getUser,
  getUsers,
  getWebPagePreview,
  hideAllJoinRequests,
  hideJoinRequest,
  hideMyStoriesViews,
  importContacts,
  incrementStoriesViews,
  initTakeoutSession,
  isPeerAvailable,
  isSelfPeer,
  iterAllStories,
  iterBoosters,
  iterChatEventLog,
  iterChatMembers,
  iterDialogs,
  iterForumTopics,
  iterHistory,
  iterInviteLinkMembers,
  iterInviteLinks,
  iterProfilePhotos,
  iterProfileStories,
  iterReactionUsers,
  iterSavedStarGifts,
  iterSearchGlobal,
  iterSearchHashtag,
  iterSearchMessages,
  iterStarsTransactions,
  iterStoryViewers,
  joinChat,
  joinChatlist,
  kickChatMember,
  leaveChat,
  logOut,
  markChatUnread,
  moveStickerInSet,
  openChat,
  openWebview,
  pinMessage,
  prepareInlineMessage,
  prepayStarGiftUpgrade,
  quoteWithMedia,
  quoteWithMediaGroup,
  quoteWithText,
  readHistory,
  readReactions,
  readStories,
  recoverPassword,
  removeCloudPassword,
  reorderPinnedForumTopics,
  reorderUsernames,
  replaceStickerInSet,
  replyMedia,
  replyMediaGroup,
  replyText,
  resendCode,
  resendPasswordEmail,
  resolveChannel,
  resolvePeer,
  resolvePeerMany,
  resolvePhoneNumber,
  resolveStarGiftOffer,
  resolveUser,
  restrictChatMember,
  revokeInviteLink,
  saveDraft,
  saveMusicToProfile,
  searchGlobal,
  searchHashtag,
  searchMessages,
  searchPostsGlobal,
  sendCode,
  sendCopy,
  sendCopyGroup,
  sendMedia,
  sendMediaGroup,
  sendOnline,
  sendPaidReaction,
  sendReaction,
  sendRecoveryCode,
  sendRichMessage,
  sendScheduled,
  sendStarGift,
  sendStarGiftOffer,
  sendStory,
  sendStoryReaction,
  sendText,
  sendTyping,
  sendVote,
  setBotInfo,
  setBotMenuButton,
  setBusinessIntro,
  setBusinessWorkHours,
  setChatColor,
  setChatDefaultPermissions,
  setChatDescription,
  setChatPhoto,
  setChatStickerSet,
  setChatTitle,
  setChatTtl,
  setChatUsername,
  setContactNote,
  setEmojiStatus,
  setFoldersOrder,
  setGameScore,
  setGlobalTtl,
  setInlineGameScore,
  setMyBirthday,
  setMyCommands,
  setMyDefaultRights,
  setMyProfilePhoto,
  setMyUsername,
  setOnline,
  setResaleStarGiftPrice,
  setSlowMode,
  setStickerSetThumb,
  setTyping,
  signIn,
  signInBot,
  signInQr,
  start,
  startTest,
  toggleContentProtection,
  toggleEmojiStatusPermission,
  toggleForumTopicClosed,
  toggleForumTopicPinned,
  toggleFragmentUsername,
  toggleGeneralTopicHidden,
  toggleJoinRequests,
  toggleJoinToSend,
  togglePeerStoriesArchived,
  togglePinnedStarGifts,
  toggleStoriesPinned,
  toggleTodoCompleted,
  transferChatOwnership,
  transferStarGift,
  translateMessage,
  translateText,
  unarchiveChats,
  unbanChatMember,
  unblockUser,
  unpinAllMessages,
  unpinMessage,
  unsaveMusicFromProfile,
  updateForumSettings,
  updateProfile,
  upgradeStarGift,
  uploadFile,
  uploadMedia,
  verifyPasswordEmail
};
