// @mtcute/core/utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const floodControl = require("./network/flood-control.cjs");
const reader = require("./tl/binary/reader.cjs");
const writer = require("./tl/binary/writer.cjs");
const bigintUtils = require("./utils/bigint-utils.cjs");
const compat = require("./utils/binary/compat.cjs");
const serialization = require("./utils/binary/serialization.cjs");
const dateEntityFormat = require("./utils/date-entity-format.cjs");
const dcs = require("./utils/dcs.cjs");
const earlyTimer = require("./utils/early-timer.cjs");
const functionUtils = require("./utils/function-utils.cjs");
const logger = require("./utils/logger.cjs");
const longUtils = require("./utils/long-utils.cjs");
const miscUtils$1 = require("./utils/misc-utils.cjs");
const peerUtils$1 = require("./utils/peer-utils.cjs");
const sortedArray = require("./utils/sorted-array.cjs");
const tlJson = require("./utils/tl-json.cjs");
const typeAssertions = require("./utils/type-assertions.cjs");
const tlRuntime = require("@mtcute/tl-runtime");
const convertFileId = require("./highlevel/utils/convert-file-id.cjs");
const entities = require("./highlevel/utils/entities.cjs");
const fileUtils = require("./highlevel/utils/file-utils.cjs");
const inlineUtils = require("./highlevel/utils/inline-utils.cjs");
const inspectable = require("./highlevel/utils/inspectable.cjs");
const miscUtils = require("./highlevel/utils/misc-utils.cjs");
const peerUtils = require("./highlevel/utils/peer-utils.cjs");
const rpsMeter = require("./highlevel/utils/rps-meter.cjs");
const streamUtils = require("./highlevel/utils/stream-utils.cjs");
const stringSession = require("./highlevel/utils/string-session.cjs");
const voiceUtils = require("./highlevel/utils/voice-utils.cjs");
const abstract = require("./utils/crypto/abstract.cjs");
const factorization = require("./utils/crypto/factorization.cjs");
const keys = require("./utils/crypto/keys.cjs");
const millerRabin = require("./utils/crypto/miller-rabin.cjs");
const mtproto = require("./utils/crypto/mtproto.cjs");
const password = require("./utils/crypto/password.cjs");
const bundle = require("./utils/links/bundle.cjs");
exports.ConnectionFloodController = floodControl.ConnectionFloodController;
exports.DEFAULT_FLOOD_LIMITS = floodControl.DEFAULT_FLOOD_LIMITS;
exports.DEFAULT_MTPROTO_ERROR_LIMITS = floodControl.DEFAULT_MTPROTO_ERROR_LIMITS;
exports.DEFAULT_SANITY_LIMITS = floodControl.DEFAULT_SANITY_LIMITS;
exports.FloodControl = floodControl.FloodControl;
exports.__tlReaderMap = reader.__tlReaderMap;
exports.__tlWriterMap = writer.__tlWriterMap;
exports.randomBigInt = bigintUtils.randomBigInt;
exports.randomBigIntBits = bigintUtils.randomBigIntBits;
exports.randomBigIntInRange = bigintUtils.randomBigIntInRange;
exports.__tlReaderMapWithCompat = compat.__tlReaderMapWithCompat;
exports.deserializeObjectWithCompat = compat.deserializeObjectWithCompat;
exports.deserializePeersIndexWithCompat = compat.deserializePeersIndexWithCompat;
exports.deserializeObject = serialization.deserializeObject;
exports.deserializePeersIndex = serialization.deserializePeersIndex;
exports.serializeObject = serialization.serializeObject;
exports.serializePeersIndex = serialization.serializePeersIndex;
exports.dateEntityFormatToString = dateEntityFormat.dateEntityFormatToString;
exports.parseDateEntityFormat = dateEntityFormat.parseDateEntityFormat;
exports.defaultProductionDc = dcs.defaultProductionDc;
exports.defaultProductionIpv6Dc = dcs.defaultProductionIpv6Dc;
exports.defaultTestDc = dcs.defaultTestDc;
exports.defaultTestIpv6Dc = dcs.defaultTestIpv6Dc;
exports.parseBasicDcOption = dcs.parseBasicDcOption;
exports.serializeBasicDcOption = dcs.serializeBasicDcOption;
exports.EarlyTimer = earlyTimer.EarlyTimer;
exports.asyncResettable = functionUtils.asyncResettable;
exports.throttle = functionUtils.throttle;
exports.LogManager = logger.LogManager;
exports.Logger = logger.Logger;
exports.LongMap = longUtils.LongMap;
exports.LongSet = longUtils.LongSet;
exports.compareLongs = longUtils.compareLongs;
exports.longFromBuffer = longUtils.longFromBuffer;
exports.longFromFastString = longUtils.longFromFastString;
exports.longToFastString = longUtils.longToFastString;
exports.randomLong = longUtils.randomLong;
exports.removeFromLongArray = longUtils.removeFromLongArray;
exports.dropUndefined = miscUtils$1.dropUndefined;
exports.getRandomInt = miscUtils$1.getRandomInt;
exports.sleepWithAbort = miscUtils$1.sleepWithAbort;
exports.getAllPeersFrom = peerUtils$1.getAllPeersFrom;
exports.getBarePeerId = peerUtils$1.getBarePeerId;
exports.getBasicPeerType = peerUtils$1.getBasicPeerType;
exports.getMarkedPeerId = peerUtils$1.getMarkedPeerId;
exports.parseMarkedPeerId = peerUtils$1.parseMarkedPeerId;
exports.toggleChannelIdMark = peerUtils$1.toggleChannelIdMark;
exports.SortedArray = sortedArray.SortedArray;
exports.jsonToTlJson = tlJson.jsonToTlJson;
exports.tlJsonToJson = tlJson.tlJsonToJson;
exports.assertTrue = typeAssertions.assertTrue;
exports.assertTypeIs = typeAssertions.assertTypeIs;
exports.assertTypeIsNot = typeAssertions.assertTypeIsNot;
exports.hasValueAtKey = typeAssertions.hasValueAtKey;
exports.isPresent = typeAssertions.isPresent;
exports.isTlRpcError = typeAssertions.isTlRpcError;
exports.mtpAssertTypeIs = typeAssertions.mtpAssertTypeIs;
exports.fileIdToEncryptedFile = convertFileId.fileIdToEncryptedFile;
exports.fileIdToInputDocument = convertFileId.fileIdToInputDocument;
exports.fileIdToInputFileLocation = convertFileId.fileIdToInputFileLocation;
exports.fileIdToInputPhoto = convertFileId.fileIdToInputPhoto;
exports.fileIdToInputWebFileLocation = convertFileId.fileIdToInputWebFileLocation;
exports.fileIdToSecureFile = convertFileId.fileIdToSecureFile;
exports.joinTextWithEntities = entities.joinTextWithEntities;
exports.richTextToTextWithEntities = entities.richTextToTextWithEntities;
exports.textWithEntitiesToRichText = entities.textWithEntitiesToRichText;
exports.determinePartSize = fileUtils.determinePartSize;
exports.extractFileName = fileUtils.extractFileName;
exports.inflateSvgPath = fileUtils.inflateSvgPath;
exports.isProbablyPlainText = fileUtils.isProbablyPlainText;
exports.strippedPhotoToJpg = fileUtils.strippedPhotoToJpg;
exports.svgPathToFile = fileUtils.svgPathToFile;
exports.encodeInlineMessageId = inlineUtils.encodeInlineMessageId;
exports.normalizeInlineId = inlineUtils.normalizeInlineId;
exports.parseInlineMessageId = inlineUtils.parseInlineMessageId;
exports.customInspectSymbol = inspectable.customInspectSymbol;
exports.makeInspectable = inspectable.makeInspectable;
exports.makeArrayPaginated = miscUtils.makeArrayPaginated;
exports.makeArrayPaginatedWithMeta = miscUtils.makeArrayPaginatedWithMeta;
exports.makeArrayWithTotal = miscUtils.makeArrayWithTotal;
exports.normalizeDate = miscUtils.normalizeDate;
exports.normalizeMessageId = miscUtils.normalizeMessageId;
exports.normalizePhoneNumber = miscUtils.normalizePhoneNumber;
exports.resolveMaybeDynamic = miscUtils.resolveMaybeDynamic;
exports.INVITE_LINK_REGEX = peerUtils.INVITE_LINK_REGEX;
exports.extractUsernames = peerUtils.extractUsernames;
exports.inputPeerToPeer = peerUtils.inputPeerToPeer;
exports.isInputPeerChannel = peerUtils.isInputPeerChannel;
exports.isInputPeerChat = peerUtils.isInputPeerChat;
exports.isInputPeerUser = peerUtils.isInputPeerUser;
exports.toInputChannel = peerUtils.toInputChannel;
exports.toInputPeer = peerUtils.toInputPeer;
exports.toInputUser = peerUtils.toInputUser;
exports.RpsMeter = rpsMeter.RpsMeter;
exports.streamToBuffer = streamUtils.streamToBuffer;
exports.readStringSession = stringSession.readStringSession;
exports.writeStringSession = stringSession.writeStringSession;
exports.decodeWaveform = voiceUtils.decodeWaveform;
exports.encodeWaveform = voiceUtils.encodeWaveform;
exports.BaseCryptoProvider = abstract.BaseCryptoProvider;
exports.factorizePQSync = factorization.factorizePQSync;
exports.addPublicKey = keys.addPublicKey;
exports.findKeyByFingerprints = keys.findKeyByFingerprints;
exports.parsePublicKey = keys.parsePublicKey;
exports.millerRabin = millerRabin.millerRabin;
exports.createAesIgeForMessage = mtproto.createAesIgeForMessage;
exports.createAesIgeForMessageOld = mtproto.createAesIgeForMessageOld;
exports.generateKeyAndIvFromNonce = mtproto.generateKeyAndIvFromNonce;
exports.computeNewPasswordHash = password.computeNewPasswordHash;
exports.computePasswordHash = password.computePasswordHash;
exports.computeSrpParams = password.computeSrpParams;
exports.links = bundle;
Object.keys(tlRuntime).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => tlRuntime[k]
  });
});
