// @mtcute/core/utils.js
import { ConnectionFloodController, DEFAULT_FLOOD_LIMITS, DEFAULT_MTPROTO_ERROR_LIMITS, DEFAULT_SANITY_LIMITS, FloodControl } from "./network/flood-control.js";
import { __tlReaderMap } from "./tl/binary/reader.js";
import { __tlWriterMap } from "./tl/binary/writer.js";
import { randomBigInt, randomBigIntBits, randomBigIntInRange } from "./utils/bigint-utils.js";
import { __tlReaderMapWithCompat, deserializeObjectWithCompat, deserializePeersIndexWithCompat } from "./utils/binary/compat.js";
import { deserializeObject, deserializePeersIndex, serializeObject, serializePeersIndex } from "./utils/binary/serialization.js";
import { dateEntityFormatToString, parseDateEntityFormat } from "./utils/date-entity-format.js";
import { defaultProductionDc, defaultProductionIpv6Dc, defaultTestDc, defaultTestIpv6Dc, parseBasicDcOption, serializeBasicDcOption } from "./utils/dcs.js";
import { EarlyTimer } from "./utils/early-timer.js";
import { asyncResettable, throttle } from "./utils/function-utils.js";
import { LogManager, Logger } from "./utils/logger.js";
import { LongMap, LongSet, compareLongs, longFromBuffer, longFromFastString, longToFastString, randomLong, removeFromLongArray } from "./utils/long-utils.js";
import { dropUndefined, getRandomInt, sleepWithAbort } from "./utils/misc-utils.js";
import { getAllPeersFrom, getBarePeerId, getBasicPeerType, getMarkedPeerId, parseMarkedPeerId, toggleChannelIdMark } from "./utils/peer-utils.js";
import { SortedArray } from "./utils/sorted-array.js";
import { jsonToTlJson, tlJsonToJson } from "./utils/tl-json.js";
import { assertTrue, assertTypeIs, assertTypeIsNot, hasValueAtKey, isPresent, isTlRpcError, mtpAssertTypeIs } from "./utils/type-assertions.js";
export * from "@mtcute/tl-runtime";
import { fileIdToEncryptedFile, fileIdToInputDocument, fileIdToInputFileLocation, fileIdToInputPhoto, fileIdToInputWebFileLocation, fileIdToSecureFile } from "./highlevel/utils/convert-file-id.js";
import { joinTextWithEntities, richTextToTextWithEntities, textWithEntitiesToRichText } from "./highlevel/utils/entities.js";
import { determinePartSize, extractFileName, inflateSvgPath, isProbablyPlainText, strippedPhotoToJpg, svgPathToFile } from "./highlevel/utils/file-utils.js";
import { encodeInlineMessageId, normalizeInlineId, parseInlineMessageId } from "./highlevel/utils/inline-utils.js";
import { customInspectSymbol, makeInspectable } from "./highlevel/utils/inspectable.js";
import { makeArrayPaginated, makeArrayPaginatedWithMeta, makeArrayWithTotal, normalizeDate, normalizeMessageId, normalizePhoneNumber, resolveMaybeDynamic } from "./highlevel/utils/misc-utils.js";
import { INVITE_LINK_REGEX, extractUsernames, inputPeerToPeer, isInputPeerChannel, isInputPeerChat, isInputPeerUser, toInputChannel, toInputPeer, toInputUser } from "./highlevel/utils/peer-utils.js";
import { RpsMeter } from "./highlevel/utils/rps-meter.js";
import { streamToBuffer } from "./highlevel/utils/stream-utils.js";
import { readStringSession, writeStringSession } from "./highlevel/utils/string-session.js";
import { decodeWaveform, encodeWaveform } from "./highlevel/utils/voice-utils.js";
import { BaseCryptoProvider } from "./utils/crypto/abstract.js";
import { factorizePQSync } from "./utils/crypto/factorization.js";
import { addPublicKey, findKeyByFingerprints, parsePublicKey } from "./utils/crypto/keys.js";
import { millerRabin } from "./utils/crypto/miller-rabin.js";
import { createAesIgeForMessage, createAesIgeForMessageOld, generateKeyAndIvFromNonce } from "./utils/crypto/mtproto.js";
import { computeNewPasswordHash, computePasswordHash, computeSrpParams } from "./utils/crypto/password.js";
import * as bundle from "./utils/links/bundle.js";
export {
  BaseCryptoProvider,
  ConnectionFloodController,
  DEFAULT_FLOOD_LIMITS,
  DEFAULT_MTPROTO_ERROR_LIMITS,
  DEFAULT_SANITY_LIMITS,
  EarlyTimer,
  FloodControl,
  INVITE_LINK_REGEX,
  LogManager,
  Logger,
  LongMap,
  LongSet,
  RpsMeter,
  SortedArray,
  __tlReaderMap,
  __tlReaderMapWithCompat,
  __tlWriterMap,
  addPublicKey,
  assertTrue,
  assertTypeIs,
  assertTypeIsNot,
  asyncResettable,
  compareLongs,
  computeNewPasswordHash,
  computePasswordHash,
  computeSrpParams,
  createAesIgeForMessage,
  createAesIgeForMessageOld,
  customInspectSymbol,
  dateEntityFormatToString,
  decodeWaveform,
  defaultProductionDc,
  defaultProductionIpv6Dc,
  defaultTestDc,
  defaultTestIpv6Dc,
  deserializeObject,
  deserializeObjectWithCompat,
  deserializePeersIndex,
  deserializePeersIndexWithCompat,
  determinePartSize,
  dropUndefined,
  encodeInlineMessageId,
  encodeWaveform,
  extractFileName,
  extractUsernames,
  factorizePQSync,
  fileIdToEncryptedFile,
  fileIdToInputDocument,
  fileIdToInputFileLocation,
  fileIdToInputPhoto,
  fileIdToInputWebFileLocation,
  fileIdToSecureFile,
  findKeyByFingerprints,
  generateKeyAndIvFromNonce,
  getAllPeersFrom,
  getBarePeerId,
  getBasicPeerType,
  getMarkedPeerId,
  getRandomInt,
  hasValueAtKey,
  inflateSvgPath,
  inputPeerToPeer,
  isInputPeerChannel,
  isInputPeerChat,
  isInputPeerUser,
  isPresent,
  isProbablyPlainText,
  isTlRpcError,
  joinTextWithEntities,
  jsonToTlJson,
  bundle as links,
  longFromBuffer,
  longFromFastString,
  longToFastString,
  makeArrayPaginated,
  makeArrayPaginatedWithMeta,
  makeArrayWithTotal,
  makeInspectable,
  millerRabin,
  mtpAssertTypeIs,
  normalizeDate,
  normalizeInlineId,
  normalizeMessageId,
  normalizePhoneNumber,
  parseBasicDcOption,
  parseDateEntityFormat,
  parseInlineMessageId,
  parseMarkedPeerId,
  parsePublicKey,
  randomBigInt,
  randomBigIntBits,
  randomBigIntInRange,
  randomLong,
  readStringSession,
  removeFromLongArray,
  resolveMaybeDynamic,
  richTextToTextWithEntities,
  serializeBasicDcOption,
  serializeObject,
  serializePeersIndex,
  sleepWithAbort,
  streamToBuffer,
  strippedPhotoToJpg,
  svgPathToFile,
  textWithEntitiesToRichText,
  throttle,
  tlJsonToJson,
  toInputChannel,
  toInputPeer,
  toInputUser,
  toggleChannelIdMark,
  writeStringSession
};
