// @fuman/utils/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const index$2 = require("./misc/index.cjs");
const index = require("./arrays/typed/index.cjs");
const index$1 = require("./arrays/u8/index.cjs");
const asyncInterval = require("./async/async-interval.cjs");
const asyncLock = require("./async/async-lock.cjs");
const asyncQueue = require("./async/async-queue.cjs");
const asyncResource = require("./async/async-resource.cjs");
const conditionVariable = require("./async/condition-variable.cjs");
const deferred = require("./async/deferred.cjs");
const emitter = require("./async/emitter.cjs");
const pool = require("./async/pool.cjs");
const sleep = require("./async/sleep.cjs");
const timers = require("./async/timers.cjs");
const bundle = require("./bigint/bundle.cjs");
const base64 = require("./encodings/base64.cjs");
const hex = require("./encodings/hex.cjs");
const utf8 = require("./encodings/utf8.cjs");
const bundle$1 = require("./iter/bundle.cjs");
const assert = require("./misc/assert.cjs");
const composer = require("./misc/composer.cjs");
const guards = require("./misc/guards.cjs");
const objects = require("./misc/objects.cjs");
const strings = require("./misc/strings.cjs");
const customMap = require("./structures/custom-map.cjs");
const customSet = require("./structures/custom-set.cjs");
const deque = require("./structures/deque.cjs");
const lruMap = require("./structures/lru-map.cjs");
const lruSet = require("./structures/lru-set.cjs");
const error = require("./types/error.cjs");
exports.noop = index$2.noop;
exports.typed = index;
exports.u8 = index$1;
exports.AsyncInterval = asyncInterval.AsyncInterval;
exports.AsyncLock = asyncLock.AsyncLock;
exports.AsyncQueue = asyncQueue.AsyncQueue;
exports.AsyncResource = asyncResource.AsyncResource;
exports.ConditionVariable = conditionVariable.ConditionVariable;
exports.Deferred = deferred.Deferred;
exports.DeferredTracked = deferred.DeferredTracked;
exports.Emitter = emitter.Emitter;
exports.AggregateError = pool.AggregateError;
exports.asyncPool = pool.asyncPool;
exports.parallelMap = pool.parallelMap;
exports.sleep = sleep.sleep;
exports.timers = timers;
exports.bigint = bundle;
exports.base64 = base64;
exports.hex = hex;
exports.utf8 = utf8;
exports.iter = bundle$1;
exports.asNonNull = assert.asNonNull;
exports.assert = assert.assert;
exports.assertHasKey = assert.assertHasKey;
exports.assertMatches = assert.assertMatches;
exports.assertNonNull = assert.assertNonNull;
exports.unsafeCastType = assert.unsafeCastType;
exports.composeMiddlewares = composer.composeMiddlewares;
exports.isBigInt = guards.isBigInt;
exports.isBoolean = guards.isBoolean;
exports.isFalsy = guards.isFalsy;
exports.isFunction = guards.isFunction;
exports.isNotNull = guards.isNotNull;
exports.isNotUndefined = guards.isNotUndefined;
exports.isNumber = guards.isNumber;
exports.isObject = guards.isObject;
exports.isString = guards.isString;
exports.isSymbol = guards.isSymbol;
exports.isTruthy = guards.isTruthy;
exports.clearUndefined = objects.clearUndefined;
exports.deepMerge = objects.deepMerge;
exports.objectEntries = objects.objectEntries;
exports.objectKeys = objects.objectKeys;
exports.assertEndsWith = strings.assertEndsWith;
exports.assertStartsWith = strings.assertStartsWith;
exports.splitOnce = strings.splitOnce;
exports.CustomMap = customMap.CustomMap;
exports.CustomSet = customSet.CustomSet;
exports.Deque = deque.Deque;
exports.LruMap = lruMap.LruMap;
exports.LruSet = lruSet.LruSet;
exports.NotImplementedError = error.NotImplementedError;
exports.notImplemented = error.notImplemented;
exports.todo = error.notImplemented;
exports.unknownToError = error.unknownToError;
exports.unreachable = error.unreachable;
