// @fuman/net/fake.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const errors = require("./errors.cjs");
class FakeConnection {
  constructor(address) {
    this.address = address;
  }
  rx = io.Bytes.alloc();
  tx = io.Bytes.alloc();
  closed = false;
  cv = new utils.ConditionVariable();
  get localAddress() {
    return this.address;
  }
  get remoteAddress() {
    return this.address;
  }
  close() {
    this.closed = true;
    this.cv.notify();
  }
  async read(into) {
    if (this.closed) throw new errors.ConnectionClosedError();
    if (!this.rx.available) await this.cv.wait();
    if (this.closed) throw new errors.ConnectionClosedError();
    return this.rx.read(into);
  }
  async write(bytes) {
    await this.tx.write(bytes);
    this.cv.notify();
  }
  writeIntoRx(bytes) {
    this.rx.writeSync(bytes.length).set(bytes);
    this.cv.notify();
  }
  getTx() {
    return this.tx.result();
  }
}
exports.FakeConnection = FakeConnection;
