// @fuman/net/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("./errors.cjs");
const fake = require("./fake.cjs");
const reconnection = require("./reconnection.cjs");
const websocket = require("./websocket.cjs");
const bundle = require("./ip/bundle.cjs");
const connect = require("./proxy/http/connect.cjs");
const types = require("./proxy/http/types.cjs");
const index = require("./proxy/http/index.cjs");
const connect$1 = require("./proxy/socks/connect.cjs");
const types$1 = require("./proxy/socks/types.cjs");
const index$1 = require("./proxy/socks/index.cjs");
exports.ConnectionClosedError = errors.ConnectionClosedError;
exports.ListenerClosedError = errors.ListenerClosedError;
exports.FakeConnection = fake.FakeConnection;
exports.PersistentConnection = reconnection.PersistentConnection;
exports.defaultReconnectionStrategy = reconnection.defaultReconnectionStrategy;
exports.WebSocketConnection = websocket.WebSocketConnection;
exports.WebSocketConnectionClosedError = websocket.WebSocketConnectionClosedError;
exports.WebSocketConnectionFramed = websocket.WebSocketConnectionFramed;
exports.connectWs = websocket.connectWs;
exports.connectWsFramed = websocket.connectWsFramed;
exports.ip = bundle;
exports.performHttpProxyHandshake = connect.performHttpProxyHandshake;
exports.HttpProxyConnectionError = types.HttpProxyConnectionError;
exports.withHttpProxy = index.withHttpProxy;
exports.performSocksHandshake = connect$1.performSocksHandshake;
exports.SocksProxyConnectionError = types$1.SocksProxyConnectionError;
exports.withSocksProxy = index$1.withSocksProxy;
