// @fuman/net/index.js
import { ConnectionClosedError, ListenerClosedError } from "./errors.js";
import { FakeConnection } from "./fake.js";
import { PersistentConnection, defaultReconnectionStrategy } from "./reconnection.js";
import { WebSocketConnection, WebSocketConnectionClosedError, WebSocketConnectionFramed, connectWs, connectWsFramed } from "./websocket.js";
import * as bundle from "./ip/bundle.js";
import { performHttpProxyHandshake } from "./proxy/http/connect.js";
import { HttpProxyConnectionError } from "./proxy/http/types.js";
import { withHttpProxy } from "./proxy/http/index.js";
import { performSocksHandshake } from "./proxy/socks/connect.js";
import { SocksProxyConnectionError } from "./proxy/socks/types.js";
import { withSocksProxy } from "./proxy/socks/index.js";
export {
  ConnectionClosedError,
  FakeConnection,
  HttpProxyConnectionError,
  ListenerClosedError,
  PersistentConnection,
  SocksProxyConnectionError,
  WebSocketConnection,
  WebSocketConnectionClosedError,
  WebSocketConnectionFramed,
  connectWs,
  connectWsFramed,
  defaultReconnectionStrategy,
  bundle as ip,
  performHttpProxyHandshake,
  performSocksHandshake,
  withHttpProxy,
  withSocksProxy
};
