// @fuman/net/reconnection.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("./errors.cjs");
const defaultReconnectionStrategy = ({ previousWait }) => {
  if (previousWait === null) return 0;
  if (previousWait === 0) return 1e3;
  if (previousWait >= 5e3) return 5e3;
  return Math.min(5e3, previousWait + 1e3);
};
function defaultOnError(err) {
  return err instanceof errors.ConnectionClosedError ? "reconnect" : "close";
}
class PersistentConnection {
  constructor(params) {
    this.params = params;
    this.#strategy = params.strategy ?? defaultReconnectionStrategy;
    this.#connect = params.connect;
    this.#onError = params.onError ?? defaultOnError;
  }
  #state = {
    previousWait: null,
    lastError: null,
    consequentFails: 0
  };
  #connect;
  #lastAddress;
  #connection;
  #connecting = false;
  #strategy;
  #onError;
  // boolean represents whether the timer is clean
  // true - resolved because timer is up
  // false - resolved because .close()/.reconnect() was called
  #sleep;
  #closed;
  get isConnected() {
    return this.#connection !== void 0;
  }
  get isConnecting() {
    return this.#connection === void 0 && this.#connecting;
  }
  get isWaiting() {
    return this.#connection === void 0 && this.#lastAddress !== void 0 && !this.#connecting;
  }
  get connection() {
    return this.#connection ?? null;
  }
  get state() {
    return this.#state;
  }
  #resetState() {
    this.#state.previousWait = null;
    this.#state.lastError = null;
    this.#state.consequentFails = 0;
    this.#connecting = false;
  }
  async #loop() {
    while (true) {
      try {
        this.#connecting = true;
        this.#connection = await this.#connect(this.#lastAddress);
        if (this.#closed) {
          this.#closed.resolve();
          break;
        }
        this.#resetState();
        await this.params.onOpen?.(this.#connection);
        this.#connection.close();
        this.#connection = void 0;
        break;
      } catch (err) {
        const oldConnection = this.#connection;
        this.#connection = void 0;
        await this.params.onClose?.();
        if (this.#closed) {
          this.#closed.resolve();
          break;
        }
        const action = await this.#onError(err, oldConnection ?? null, this.#state);
        if (action === "close") {
          break;
        }
        const wait = action === "reconnect-now" ? 0 : this.#strategy(this.#state);
        if (wait === false) {
          break;
        }
        this.params.onWait?.(wait);
        if (wait !== 0) {
          this.#sleep = new utils.Deferred();
          const timer = utils.timers.setTimeout(this.#sleep.resolve, wait, true);
          const sleepResult = await this.#sleep.promise;
          this.#sleep = void 0;
          if (!sleepResult) {
            utils.timers.clearTimeout(timer);
            if (this.#closed != null) {
              this.#closed.resolve();
              break;
            } else {
              continue;
            }
          }
        }
        this.#state.previousWait = wait;
        this.#state.consequentFails = 1;
        continue;
      }
    }
    this.#lastAddress = void 0;
    this.#resetState();
  }
  connect(address) {
    if (this.#lastAddress !== void 0 && this.#lastAddress !== address) {
      throw new Error("Connection is already open to another address");
    }
    this.#closed = void 0;
    this.#lastAddress = address;
    void this.#loop();
  }
  /**
   * @param force  Whether to close the existing connection if there is one
   */
  reconnect(force) {
    if (this.#sleep) {
      this.#sleep.resolve(false);
    } else if (this.#connection && force) {
      this.#connection.close();
    }
  }
  async close() {
    if (this.#closed) return this.#closed.promise;
    if (this.#lastAddress == null) return;
    this.#closed = new utils.Deferred();
    if (this.#sleep) {
      this.#sleep.resolve(false);
    } else if (this.#connection) {
      this.#connection.close();
    } else if (this.#connecting) ;
    return this.#closed.promise;
  }
  async changeTransport(connect) {
    this.#connect = connect;
    const addr = this.#lastAddress;
    await this.close();
    if (addr != null) {
      this.connect(addr);
    }
  }
}
exports.PersistentConnection = PersistentConnection;
exports.defaultReconnectionStrategy = defaultReconnectionStrategy;
