// @fuman/net/websocket.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const errors = require("./errors.cjs");
class WebSocketConnectionClosedError extends errors.ConnectionClosedError {
  constructor(code, reason) {
    super(`code ${code} (${reason})`);
    this.code = code;
    this.reason = reason;
  }
}
function eventToError(event) {
  if (event instanceof Error) {
    return event;
  }
  return "error" in event ? event.error : new Error("unknown WebSocket error", { cause: event });
}
class WebSocketConnectionBase {
  constructor(socket) {
    this.socket = socket;
    this._cv = new utils.ConditionVariable();
    socket.addEventListener("message", (event) => {
      this.onMessage(event);
      this._cv.notify();
    });
    socket.addEventListener("close", (event) => {
      if (this._error) return;
      this._error = new WebSocketConnectionClosedError(event.code, event.reason);
      this._cv.notify();
    });
    socket.addEventListener("error", (event) => {
      if (this._error) return;
      this._error = eventToError(event);
      this._cv.notify();
    });
  }
  _error = null;
  _cv;
  get remoteAddress() {
    return this.socket.url;
  }
  get localAddress() {
    throw new Error(".localAddress is not available for WebSockets");
  }
  close() {
    this.socket?.close();
    this._error = new errors.ConnectionClosedError();
    this._cv.notify();
  }
  closeWithCode(code, reason) {
    this.socket.close(code, reason);
    this._error = new WebSocketConnectionClosedError(code, reason ?? "");
    this._cv.notify();
  }
}
class WebSocketConnection extends WebSocketConnectionBase {
  #buffer = io.Bytes.alloc(0);
  onMessage(event) {
    if (typeof event.data === "string") {
      const buf = this.#buffer.writeSync(utils.utf8.encodedLength(event.data));
      utils.utf8.encoder.encodeInto(event.data, buf);
    } else {
      const u8 = new Uint8Array(event.data);
      this.#buffer.writeSync(u8.length).set(u8);
    }
    this.#buffer.disposeWriteSync();
  }
  async read(into) {
    if (this.#buffer.available > 0) {
      const size2 = Math.min(this.#buffer.available, into.length);
      into.set(this.#buffer.readSync(size2));
      this.#buffer.reclaim();
      return size2;
    }
    if (this._error !== null) throw this._error;
    await this._cv.wait();
    if (this._error !== null) throw this._error;
    const size = Math.min(this.#buffer.available, into.length);
    into.set(this.#buffer.readSync(size));
    this.#buffer.reclaim();
    return size;
  }
  async write(bytes) {
    if (this._error) throw this._error;
    if (!bytes.length) return;
    this.socket.send(bytes);
  }
}
class WebSocketConnectionFramed extends WebSocketConnectionBase {
  #buffer = new utils.Deque();
  onMessage(event) {
    if (typeof event.data === "string") {
      this.#buffer.pushBack(event.data);
    } else {
      this.#buffer.pushBack(new Uint8Array(event.data));
    }
  }
  async readFrame() {
    if (!this.#buffer.isEmpty()) {
      return this.#buffer.popFront();
    }
    if (this._error !== null) throw this._error;
    await this._cv.wait();
    if (this._error !== null) throw this._error;
    return this.#buffer.popFront();
  }
  async writeFrame(data) {
    if (this._error) throw this._error;
    this.socket.send(data);
  }
}
const connectWs = (endpoint) => {
  const {
    url,
    implementation: WebSocketImpl = WebSocket,
    protocols
  } = endpoint;
  return new Promise((resolve, reject) => {
    const socket = new WebSocketImpl(url, protocols);
    socket.binaryType = "arraybuffer";
    const onError = (event) => {
      socket.removeEventListener("error", onError);
      reject(eventToError(event));
    };
    socket.addEventListener("error", onError);
    socket.addEventListener("open", () => {
      socket.removeEventListener("error", onError);
      resolve(new WebSocketConnection(socket));
    });
  });
};
async function connectWsFramed(endpoint) {
  const {
    url,
    implementation: WebSocketImpl = WebSocket,
    protocols
  } = endpoint;
  return new Promise((resolve, reject) => {
    const socket = new WebSocketImpl(url, protocols);
    socket.binaryType = "arraybuffer";
    const onError = (event) => {
      socket.removeEventListener("error", onError);
      reject(eventToError(event));
    };
    socket.addEventListener("error", onError);
    socket.addEventListener("open", () => {
      socket.removeEventListener("error", onError);
      resolve(new WebSocketConnectionFramed(socket));
    });
  });
}
exports.WebSocketConnection = WebSocketConnection;
exports.WebSocketConnectionClosedError = WebSocketConnectionClosedError;
exports.WebSocketConnectionFramed = WebSocketConnectionFramed;
exports.connectWs = connectWs;
exports.connectWsFramed = connectWsFramed;
