// @mtcute/core/highlevel/types/reactions/peer-reaction.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
import { toReactionEmoji } from "./types.js";
import { makeInspectable } from "../../utils/inspectable.js";
class PeerReaction {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Emoji representing the reaction
   */
  get emoji() {
    return toReactionEmoji(this.raw.reaction);
  }
  /**
   * Whether this is a big reaction
   */
  get big() {
    return this.raw.big;
  }
  /**
   * Whether this reaction is unread by the current user
   */
  get unread() {
    return this.raw.unread;
  }
  /**
   * ID of the peer who has reacted
   */
  get peerId() {
    return getMarkedPeerId(this.raw.peerId);
  }
  /**
   * User who has reacted
   */
  get peer() {
    return parsePeer(this.raw.peerId, this._peers);
  }
}
const PeerReaction$1 = /* @__PURE__ */ memoizeGetters(PeerReaction, ["peer"]);
const PeerReaction$2 = /* @__PURE__ */ makeInspectable(PeerReaction$1);
class PaidPeerReaction {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this reaction is from the current user */
  get my() {
    return this.raw.my;
  }
  /** Whether this reaction was sent anonymously */
  get anonymous() {
    return this.raw.anonymous;
  }
  /**
   * If this reaction was not sent anonymously,
   * this field will contain the user who sent it
   */
  get peer() {
    if (!this.raw.peerId) return null;
    return parsePeer(this.raw.peerId, this._peers);
  }
  /** Number of reactions sent by this user */
  get count() {
    return this.raw.count;
  }
}
const PaidPeerReaction$1 = /* @__PURE__ */ memoizeGetters(PaidPeerReaction, ["peer"]);
const PaidPeerReaction$2 = /* @__PURE__ */ makeInspectable(PaidPeerReaction$1);
export {
  PaidPeerReaction$2 as PaidPeerReaction,
  PeerReaction$2 as PeerReaction
};
