// @mtcute/core/highlevel/types/reactions/peer-reaction.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
const types = require("./types.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class PeerReaction {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Emoji representing the reaction
   */
  get emoji() {
    return types.toReactionEmoji(this.raw.reaction);
  }
  /**
   * Whether this is a big reaction
   */
  get big() {
    return this.raw.big;
  }
  /**
   * Whether this reaction is unread by the current user
   */
  get unread() {
    return this.raw.unread;
  }
  /**
   * ID of the peer who has reacted
   */
  get peerId() {
    return peerUtils.getMarkedPeerId(this.raw.peerId);
  }
  /**
   * User who has reacted
   */
  get peer() {
    return peer.parsePeer(this.raw.peerId, this._peers);
  }
}
const PeerReaction$1 = /* @__PURE__ */ memoize.memoizeGetters(PeerReaction, ["peer"]);
const PeerReaction$2 = /* @__PURE__ */ inspectable.makeInspectable(PeerReaction$1);
class PaidPeerReaction {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this reaction is from the current user */
  get my() {
    return this.raw.my;
  }
  /** Whether this reaction was sent anonymously */
  get anonymous() {
    return this.raw.anonymous;
  }
  /**
   * If this reaction was not sent anonymously,
   * this field will contain the user who sent it
   */
  get peer() {
    if (!this.raw.peerId) return null;
    return peer.parsePeer(this.raw.peerId, this._peers);
  }
  /** Number of reactions sent by this user */
  get count() {
    return this.raw.count;
  }
}
const PaidPeerReaction$1 = /* @__PURE__ */ memoize.memoizeGetters(PaidPeerReaction, ["peer"]);
const PaidPeerReaction$2 = /* @__PURE__ */ inspectable.makeInspectable(PaidPeerReaction$1);
exports.PaidPeerReaction = PaidPeerReaction$2;
exports.PeerReaction = PeerReaction$2;
