// @mtcute/core/highlevel/types/reactions/reaction-count.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const types = require("./types.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class ReactionCount {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Emoji representing the reaction
   */
  get emoji() {
    return types.toReactionEmoji(this.raw.reaction);
  }
  /** Whether this is a paid reaction */
  get isPaid() {
    return this.raw.reaction._ === "reactionPaid";
  }
  /**
   * Number of users who reacted with this emoji
   */
  get count() {
    return this.raw.count;
  }
  /**
   * If the current user has reacted with this emoji,
   * this field will contain the order in which the
   * reaction was added.
   */
  get order() {
    return this.raw.chosenOrder ?? null;
  }
}
const ReactionCount$1 = /* @__PURE__ */ inspectable.makeInspectable(ReactionCount);
exports.ReactionCount = ReactionCount$1;
