// @mtcute/file-id/parse.js
import { base64 } from "@fuman/utils";
import { TlBinaryReader } from "@mtcute/tl-runtime";
import { telegramRleDecode } from "./utils.js";
import { PERSISTENT_ID_VERSION_OLD, PERSISTENT_ID_VERSION, UnsupportedError, CURRENT_VERSION, FileType, InvalidFileIdError, FILE_REFERENCE_FLAG, WEB_LOCATION_FLAG } from "./types-inner.js";
function parseWebFileLocation(reader) {
  return {
    _: "web",
    url: reader.string(),
    accessHash: reader.long()
  };
}
function parsePhotoSizeSource(reader) {
  const variant = reader.int();
  switch (variant) {
    case 0:
      return {
        _: "legacy",
        secret: reader.long()
      };
    case 1: {
      const fileType = reader.int();
      if (fileType < 0 || fileType >= FileType.Size) {
        throw new UnsupportedError(
          `Unsupported file type: ${fileType} (${base64.encode(reader.uint8View)})`
        );
      }
      const thumbnailType = reader.int();
      if (thumbnailType < 0 || thumbnailType > 255) {
        throw new InvalidFileIdError(
          `Wrong thumbnail type: ${thumbnailType} (${base64.encode(reader.uint8View)})`
        );
      }
      return {
        _: "thumbnail",
        fileType,
        thumbnailType: String.fromCharCode(thumbnailType)
      };
    }
    case 2:
    case 3:
      return {
        _: "dialogPhoto",
        big: variant === 3,
        id: reader.int53(),
        accessHash: reader.long()
      };
    case 4:
      return {
        _: "stickerSetThumbnail",
        id: reader.long(),
        accessHash: reader.long()
      };
    case 5: {
      const res = {
        _: "fullLegacy",
        volumeId: reader.long(),
        secret: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 6:
    case 7: {
      const res = {
        _: "dialogPhotoLegacy",
        big: variant === 7,
        id: reader.int53(),
        accessHash: reader.long(),
        volumeId: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 8: {
      const res = {
        _: "stickerSetThumbnailLegacy",
        id: reader.long(),
        accessHash: reader.long(),
        volumeId: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 9:
      return {
        _: "stickerSetThumbnailVersion",
        id: reader.long(),
        accessHash: reader.long(),
        version: reader.int()
      };
    default:
      throw new UnsupportedError(
        `Unsupported photo size source ${variant} (${base64.encode(reader.uint8View)})`
      );
  }
}
function parsePhotoFileLocation(reader, version) {
  const id = reader.long();
  const accessHash = reader.long();
  let source;
  if (version >= 32) {
    source = parsePhotoSizeSource(reader);
  } else {
    const volumeId = reader.long();
    let localId = 0;
    if (version >= 22) {
      source = parsePhotoSizeSource(reader);
      localId = reader.int();
    } else {
      source = {
        _: "fullLegacy",
        secret: reader.long(),
        localId: reader.int(),
        volumeId
      };
    }
    switch (source._) {
      case "legacy":
        source = {
          _: "fullLegacy",
          secret: reader.long(),
          localId: reader.int(),
          volumeId
        };
        break;
      case "fullLegacy":
      case "thumbnail":
        break;
      case "dialogPhoto":
        source = {
          _: "dialogPhotoLegacy",
          id: source.id,
          accessHash: source.accessHash,
          big: source.big,
          localId,
          volumeId
        };
        break;
      case "stickerSetThumbnail":
        source = {
          _: "stickerSetThumbnailLegacy",
          id: source.id,
          accessHash: source.accessHash,
          localId,
          volumeId
        };
        break;
      default:
        throw new InvalidFileIdError("Invalid PhotoSizeSource in legacy PhotoRemoteFileLocation");
    }
  }
  return {
    _: "photo",
    id,
    accessHash,
    source
  };
}
function parseCommonFileLocation(reader) {
  return {
    _: "common",
    id: reader.long(),
    accessHash: reader.long()
  };
}
function fromPersistentIdV23(binary, version) {
  if (version < 0 || version > CURRENT_VERSION) {
    throw new UnsupportedError(`Unsupported file ID v3 subversion: ${version} (${base64.encode(binary)})`);
  }
  binary = telegramRleDecode(binary);
  const reader = TlBinaryReader.manual(binary);
  let fileType = reader.int();
  const isWeb = Boolean(fileType & WEB_LOCATION_FLAG);
  const hasFileReference = Boolean(fileType & FILE_REFERENCE_FLAG);
  fileType &= -16777217;
  fileType &= -33554433;
  if (fileType < 0 || fileType >= FileType.Size) {
    throw new UnsupportedError(`Unsupported file type: ${fileType} (${base64.encode(binary)})`);
  }
  const dcId = reader.int();
  let fileReference = null;
  if (hasFileReference) {
    fileReference = reader.bytes();
    if (fileReference.length === 1 && fileReference[0] === 35) {
      fileReference = null;
    }
  }
  let location;
  if (isWeb) {
    location = parseWebFileLocation(reader);
  } else {
    switch (fileType) {
      case FileType.Photo:
      case FileType.ProfilePhoto:
      case FileType.Thumbnail:
      case FileType.EncryptedThumbnail:
      case FileType.Wallpaper: {
        location = parsePhotoFileLocation(reader, version);
        switch (location.source._) {
          case "thumbnail":
            if (location.source.fileType !== fileType || fileType !== FileType.Photo && fileType !== FileType.Thumbnail && fileType !== FileType.EncryptedThumbnail) {
              throw new InvalidFileIdError("Invalid FileType in PhotoRemoteFileLocation Thumbnail");
            }
            break;
          case "dialogPhoto":
          case "dialogPhotoLegacy":
            if (fileType !== FileType.ProfilePhoto) {
              throw new InvalidFileIdError("Invalid FileType in PhotoRemoteFileLocation DialogPhoto");
            }
            break;
          case "stickerSetThumbnail":
          case "stickerSetThumbnailLegacy":
          case "stickerSetThumbnailVersion":
            if (fileType !== FileType.Thumbnail) {
              throw new InvalidFileIdError(
                "Invalid FileType in PhotoRemoteFileLocation StickerSetThumbnail"
              );
            }
            break;
        }
        break;
      }
      case FileType.Video:
      case FileType.VoiceNote:
      case FileType.Document:
      case FileType.Sticker:
      case FileType.Audio:
      case FileType.Animation:
      case FileType.Encrypted:
      case FileType.VideoNote:
      case FileType.SecureRaw:
      case FileType.Secure:
      case FileType.Background:
      case FileType.DocumentAsFile: {
        location = parseCommonFileLocation(reader);
        break;
      }
      default:
        throw new UnsupportedError(`Invalid file type: ${fileType} (${base64.encode(binary)})`);
    }
  }
  return {
    _: "remoteFileLocation",
    dcId,
    type: fileType,
    fileReference,
    location
  };
}
function fromPersistentIdV2(binary) {
  return fromPersistentIdV23(binary.subarray(0, -1), 0);
}
function fromPersistentIdV3(binary) {
  const subversion = binary[binary.length - 2];
  return fromPersistentIdV23(binary.subarray(0, -2), subversion);
}
function parseFileId(fileId) {
  if (typeof fileId === "string") fileId = base64.decode(fileId, true);
  const version = fileId[fileId.length - 1];
  if (version === PERSISTENT_ID_VERSION_OLD) {
    return fromPersistentIdV2(fileId);
  }
  if (version === PERSISTENT_ID_VERSION) {
    return fromPersistentIdV3(fileId);
  }
  throw new UnsupportedError(`Unsupported file ID version: ${version} (${base64.encode(fileId)})`);
}
export {
  parseFileId
};
