// @mtcute/file-id/parse-unique.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/file-id] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/file-id] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const utils$1 = require("./utils.cjs");
const typesInner = require("./types-inner.cjs");
function parsePhotoUniqueId(reader, binary) {
  const remaining = binary.length - reader.pos;
  if (remaining === 13) {
    if (binary[reader.pos] === 2) {
      reader.pos += 1;
      return {
        _: "photoStickerSetVersion",
        stickerSetId: reader.long(),
        stickerSetVersion: reader.int()
      };
    }
  }
  if (remaining === 9) {
    return {
      _: "photoId",
      id: reader.long(),
      subType: binary[reader.pos]
    };
  }
  if (remaining === 12) {
    const marker = reader.int();
    if (marker === 100) {
      return { _: "photoLegacy", secret: reader.long() };
    }
    reader.pos -= 4;
    return {
      _: "photoVolumeId",
      volumeId: reader.long(),
      localId: reader.int()
    };
  }
  if (remaining === 20) {
    const marker = reader.int();
    if (marker === 150) {
      return {
        _: "photoStickerSet",
        stickerSetId: reader.long(),
        stickerSetAccessHash: reader.long()
      };
    }
    throw new typesInner.InvalidFileIdError(`Unexpected photo unique file ID marker: ${marker}`);
  }
  throw new typesInner.InvalidFileIdError(
    `Unexpected photo unique file ID size: ${remaining} (${utils.base64.encode(binary)})`
  );
}
function parseUniqueFileId(fileId) {
  if (typeof fileId === "string") fileId = utils.base64.decode(fileId, true);
  const binary = utils$1.telegramRleDecode(fileId);
  const reader = tlRuntime.TlBinaryReader.manual(binary);
  const type = reader.int();
  if (type < 0 || type > 5) {
    throw new typesInner.UnsupportedError(`Unsupported unique file ID type: ${type} (${utils.base64.encode(fileId)})`);
  }
  switch (type) {
    case typesInner.UniqueFileIdType.Web:
      return { type, url: reader.string() };
    case typesInner.UniqueFileIdType.Photo:
      return { type, location: parsePhotoUniqueId(reader, binary) };
    case typesInner.UniqueFileIdType.Document:
      return { type, id: reader.long() };
    case typesInner.UniqueFileIdType.Secure:
      return { type, id: reader.long() };
    case typesInner.UniqueFileIdType.Encrypted:
      return { type, id: reader.long() };
    case typesInner.UniqueFileIdType.Temp:
      return { type, id: reader.long() };
    default:
      throw new typesInner.UnsupportedError(`Unsupported unique file ID type: ${type}`);
  }
}
exports.parseUniqueFileId = parseUniqueFileId;
