// @mtcute/file-id/parse.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/file-id] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/file-id] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const utils$1 = require("./utils.cjs");
const typesInner = require("./types-inner.cjs");
function parseWebFileLocation(reader) {
  return {
    _: "web",
    url: reader.string(),
    accessHash: reader.long()
  };
}
function parsePhotoSizeSource(reader) {
  const variant = reader.int();
  switch (variant) {
    case 0:
      return {
        _: "legacy",
        secret: reader.long()
      };
    case 1: {
      const fileType = reader.int();
      if (fileType < 0 || fileType >= typesInner.FileType.Size) {
        throw new typesInner.UnsupportedError(
          `Unsupported file type: ${fileType} (${utils.base64.encode(reader.uint8View)})`
        );
      }
      const thumbnailType = reader.int();
      if (thumbnailType < 0 || thumbnailType > 255) {
        throw new typesInner.InvalidFileIdError(
          `Wrong thumbnail type: ${thumbnailType} (${utils.base64.encode(reader.uint8View)})`
        );
      }
      return {
        _: "thumbnail",
        fileType,
        thumbnailType: String.fromCharCode(thumbnailType)
      };
    }
    case 2:
    case 3:
      return {
        _: "dialogPhoto",
        big: variant === 3,
        id: reader.int53(),
        accessHash: reader.long()
      };
    case 4:
      return {
        _: "stickerSetThumbnail",
        id: reader.long(),
        accessHash: reader.long()
      };
    case 5: {
      const res = {
        _: "fullLegacy",
        volumeId: reader.long(),
        secret: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new typesInner.InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 6:
    case 7: {
      const res = {
        _: "dialogPhotoLegacy",
        big: variant === 7,
        id: reader.int53(),
        accessHash: reader.long(),
        volumeId: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new typesInner.InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 8: {
      const res = {
        _: "stickerSetThumbnailLegacy",
        id: reader.long(),
        accessHash: reader.long(),
        volumeId: reader.long(),
        localId: reader.int()
      };
      if (res.localId < 0) {
        throw new typesInner.InvalidFileIdError("Wrong local_id (< 0)");
      }
      return res;
    }
    case 9:
      return {
        _: "stickerSetThumbnailVersion",
        id: reader.long(),
        accessHash: reader.long(),
        version: reader.int()
      };
    default:
      throw new typesInner.UnsupportedError(
        `Unsupported photo size source ${variant} (${utils.base64.encode(reader.uint8View)})`
      );
  }
}
function parsePhotoFileLocation(reader, version) {
  const id = reader.long();
  const accessHash = reader.long();
  let source;
  if (version >= 32) {
    source = parsePhotoSizeSource(reader);
  } else {
    const volumeId = reader.long();
    let localId = 0;
    if (version >= 22) {
      source = parsePhotoSizeSource(reader);
      localId = reader.int();
    } else {
      source = {
        _: "fullLegacy",
        secret: reader.long(),
        localId: reader.int(),
        volumeId
      };
    }
    switch (source._) {
      case "legacy":
        source = {
          _: "fullLegacy",
          secret: reader.long(),
          localId: reader.int(),
          volumeId
        };
        break;
      case "fullLegacy":
      case "thumbnail":
        break;
      case "dialogPhoto":
        source = {
          _: "dialogPhotoLegacy",
          id: source.id,
          accessHash: source.accessHash,
          big: source.big,
          localId,
          volumeId
        };
        break;
      case "stickerSetThumbnail":
        source = {
          _: "stickerSetThumbnailLegacy",
          id: source.id,
          accessHash: source.accessHash,
          localId,
          volumeId
        };
        break;
      default:
        throw new typesInner.InvalidFileIdError("Invalid PhotoSizeSource in legacy PhotoRemoteFileLocation");
    }
  }
  return {
    _: "photo",
    id,
    accessHash,
    source
  };
}
function parseCommonFileLocation(reader) {
  return {
    _: "common",
    id: reader.long(),
    accessHash: reader.long()
  };
}
function fromPersistentIdV23(binary, version) {
  if (version < 0 || version > typesInner.CURRENT_VERSION) {
    throw new typesInner.UnsupportedError(`Unsupported file ID v3 subversion: ${version} (${utils.base64.encode(binary)})`);
  }
  binary = utils$1.telegramRleDecode(binary);
  const reader = tlRuntime.TlBinaryReader.manual(binary);
  let fileType = reader.int();
  const isWeb = Boolean(fileType & typesInner.WEB_LOCATION_FLAG);
  const hasFileReference = Boolean(fileType & typesInner.FILE_REFERENCE_FLAG);
  fileType &= -16777217;
  fileType &= -33554433;
  if (fileType < 0 || fileType >= typesInner.FileType.Size) {
    throw new typesInner.UnsupportedError(`Unsupported file type: ${fileType} (${utils.base64.encode(binary)})`);
  }
  const dcId = reader.int();
  let fileReference = null;
  if (hasFileReference) {
    fileReference = reader.bytes();
    if (fileReference.length === 1 && fileReference[0] === 35) {
      fileReference = null;
    }
  }
  let location;
  if (isWeb) {
    location = parseWebFileLocation(reader);
  } else {
    switch (fileType) {
      case typesInner.FileType.Photo:
      case typesInner.FileType.ProfilePhoto:
      case typesInner.FileType.Thumbnail:
      case typesInner.FileType.EncryptedThumbnail:
      case typesInner.FileType.Wallpaper: {
        location = parsePhotoFileLocation(reader, version);
        switch (location.source._) {
          case "thumbnail":
            if (location.source.fileType !== fileType || fileType !== typesInner.FileType.Photo && fileType !== typesInner.FileType.Thumbnail && fileType !== typesInner.FileType.EncryptedThumbnail) {
              throw new typesInner.InvalidFileIdError("Invalid FileType in PhotoRemoteFileLocation Thumbnail");
            }
            break;
          case "dialogPhoto":
          case "dialogPhotoLegacy":
            if (fileType !== typesInner.FileType.ProfilePhoto) {
              throw new typesInner.InvalidFileIdError("Invalid FileType in PhotoRemoteFileLocation DialogPhoto");
            }
            break;
          case "stickerSetThumbnail":
          case "stickerSetThumbnailLegacy":
          case "stickerSetThumbnailVersion":
            if (fileType !== typesInner.FileType.Thumbnail) {
              throw new typesInner.InvalidFileIdError(
                "Invalid FileType in PhotoRemoteFileLocation StickerSetThumbnail"
              );
            }
            break;
        }
        break;
      }
      case typesInner.FileType.Video:
      case typesInner.FileType.VoiceNote:
      case typesInner.FileType.Document:
      case typesInner.FileType.Sticker:
      case typesInner.FileType.Audio:
      case typesInner.FileType.Animation:
      case typesInner.FileType.Encrypted:
      case typesInner.FileType.VideoNote:
      case typesInner.FileType.SecureRaw:
      case typesInner.FileType.Secure:
      case typesInner.FileType.Background:
      case typesInner.FileType.DocumentAsFile: {
        location = parseCommonFileLocation(reader);
        break;
      }
      default:
        throw new typesInner.UnsupportedError(`Invalid file type: ${fileType} (${utils.base64.encode(binary)})`);
    }
  }
  return {
    _: "remoteFileLocation",
    dcId,
    type: fileType,
    fileReference,
    location
  };
}
function fromPersistentIdV2(binary) {
  return fromPersistentIdV23(binary.subarray(0, -1), 0);
}
function fromPersistentIdV3(binary) {
  const subversion = binary[binary.length - 2];
  return fromPersistentIdV23(binary.subarray(0, -2), subversion);
}
function parseFileId(fileId) {
  if (typeof fileId === "string") fileId = utils.base64.decode(fileId, true);
  const version = fileId[fileId.length - 1];
  if (version === typesInner.PERSISTENT_ID_VERSION_OLD) {
    return fromPersistentIdV2(fileId);
  }
  if (version === typesInner.PERSISTENT_ID_VERSION) {
    return fromPersistentIdV3(fileId);
  }
  throw new typesInner.UnsupportedError(`Unsupported file ID version: ${version} (${utils.base64.encode(fileId)})`);
}
exports.parseFileId = parseFileId;
