// @mtcute/file-id/parse-unique.js
import { base64 } from "@fuman/utils";
import { TlBinaryReader } from "@mtcute/tl-runtime";
import { telegramRleDecode } from "./utils.js";
import { UnsupportedError, UniqueFileIdType, InvalidFileIdError } from "./types-inner.js";
function parsePhotoUniqueId(reader, binary) {
  const remaining = binary.length - reader.pos;
  if (remaining === 13) {
    if (binary[reader.pos] === 2) {
      reader.pos += 1;
      return {
        _: "photoStickerSetVersion",
        stickerSetId: reader.long(),
        stickerSetVersion: reader.int()
      };
    }
  }
  if (remaining === 9) {
    return {
      _: "photoId",
      id: reader.long(),
      subType: binary[reader.pos]
    };
  }
  if (remaining === 12) {
    const marker = reader.int();
    if (marker === 100) {
      return { _: "photoLegacy", secret: reader.long() };
    }
    reader.pos -= 4;
    return {
      _: "photoVolumeId",
      volumeId: reader.long(),
      localId: reader.int()
    };
  }
  if (remaining === 20) {
    const marker = reader.int();
    if (marker === 150) {
      return {
        _: "photoStickerSet",
        stickerSetId: reader.long(),
        stickerSetAccessHash: reader.long()
      };
    }
    throw new InvalidFileIdError(`Unexpected photo unique file ID marker: ${marker}`);
  }
  throw new InvalidFileIdError(
    `Unexpected photo unique file ID size: ${remaining} (${base64.encode(binary)})`
  );
}
function parseUniqueFileId(fileId) {
  if (typeof fileId === "string") fileId = base64.decode(fileId, true);
  const binary = telegramRleDecode(fileId);
  const reader = TlBinaryReader.manual(binary);
  const type = reader.int();
  if (type < 0 || type > 5) {
    throw new UnsupportedError(`Unsupported unique file ID type: ${type} (${base64.encode(fileId)})`);
  }
  switch (type) {
    case UniqueFileIdType.Web:
      return { type, url: reader.string() };
    case UniqueFileIdType.Photo:
      return { type, location: parsePhotoUniqueId(reader, binary) };
    case UniqueFileIdType.Document:
      return { type, id: reader.long() };
    case UniqueFileIdType.Secure:
      return { type, id: reader.long() };
    case UniqueFileIdType.Encrypted:
      return { type, id: reader.long() };
    case UniqueFileIdType.Temp:
      return { type, id: reader.long() };
    default:
      throw new UnsupportedError(`Unsupported unique file ID type: ${type}`);
  }
}
export {
  parseUniqueFileId
};
