// @mtcute/file-id/types-inner.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/file-id] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/file-id] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const PERSISTENT_ID_VERSION_OLD = 2;
const PERSISTENT_ID_VERSION = 4;
const WEB_LOCATION_FLAG = 1 << 24;
const FILE_REFERENCE_FLAG = 1 << 25;
const CURRENT_VERSION = 58;
class FileIdError extends Error {
}
class UnsupportedError extends FileIdError {
}
class InvalidFileIdError extends FileIdError {
}
class ConversionError extends FileIdError {
  constructor(to) {
    super(`Cannot convert given File ID to ${to}`);
  }
}
var FileType = /* @__PURE__ */ ((FileType2) => {
  FileType2[FileType2["Thumbnail"] = 0] = "Thumbnail";
  FileType2[FileType2["ProfilePhoto"] = 1] = "ProfilePhoto";
  FileType2[FileType2["Photo"] = 2] = "Photo";
  FileType2[FileType2["VoiceNote"] = 3] = "VoiceNote";
  FileType2[FileType2["Video"] = 4] = "Video";
  FileType2[FileType2["Document"] = 5] = "Document";
  FileType2[FileType2["Encrypted"] = 6] = "Encrypted";
  FileType2[FileType2["Temp"] = 7] = "Temp";
  FileType2[FileType2["Sticker"] = 8] = "Sticker";
  FileType2[FileType2["Audio"] = 9] = "Audio";
  FileType2[FileType2["Animation"] = 10] = "Animation";
  FileType2[FileType2["EncryptedThumbnail"] = 11] = "EncryptedThumbnail";
  FileType2[FileType2["Wallpaper"] = 12] = "Wallpaper";
  FileType2[FileType2["VideoNote"] = 13] = "VideoNote";
  FileType2[FileType2["SecureRaw"] = 14] = "SecureRaw";
  FileType2[FileType2["Secure"] = 15] = "Secure";
  FileType2[FileType2["Background"] = 16] = "Background";
  FileType2[FileType2["DocumentAsFile"] = 17] = "DocumentAsFile";
  FileType2[FileType2["Size"] = 18] = "Size";
  FileType2[FileType2["None"] = 19] = "None";
  return FileType2;
})(FileType || {});
var UniqueFileIdType = /* @__PURE__ */ ((UniqueFileIdType2) => {
  UniqueFileIdType2[UniqueFileIdType2["Web"] = 0] = "Web";
  UniqueFileIdType2[UniqueFileIdType2["Photo"] = 1] = "Photo";
  UniqueFileIdType2[UniqueFileIdType2["Document"] = 2] = "Document";
  UniqueFileIdType2[UniqueFileIdType2["Secure"] = 3] = "Secure";
  UniqueFileIdType2[UniqueFileIdType2["Encrypted"] = 4] = "Encrypted";
  UniqueFileIdType2[UniqueFileIdType2["Temp"] = 5] = "Temp";
  return UniqueFileIdType2;
})(UniqueFileIdType || {});
function isFileIdLike(obj) {
  return typeof obj === "string" || obj !== null && typeof obj === "object" && obj._ === "remoteFileLocation";
}
exports.CURRENT_VERSION = CURRENT_VERSION;
exports.ConversionError = ConversionError;
exports.FILE_REFERENCE_FLAG = FILE_REFERENCE_FLAG;
exports.FileIdError = FileIdError;
exports.FileType = FileType;
exports.InvalidFileIdError = InvalidFileIdError;
exports.PERSISTENT_ID_VERSION = PERSISTENT_ID_VERSION;
exports.PERSISTENT_ID_VERSION_OLD = PERSISTENT_ID_VERSION_OLD;
exports.UniqueFileIdType = UniqueFileIdType;
exports.UnsupportedError = UnsupportedError;
exports.WEB_LOCATION_FLAG = WEB_LOCATION_FLAG;
exports.isFileIdLike = isFileIdLike;
