// @mtcute/core/utils/links/bundle.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const bots = require("./bots.cjs");
const calls = require("./calls.cjs");
const chat = require("./chat.cjs");
const misc = require("./misc.cjs");
const proxy = require("./proxy.cjs");
const stickers = require("./stickers.cjs");
const user = require("./user.cjs");
exports.botAddToChannel = bots.botAddToChannel;
exports.botAddToGroup = bots.botAddToGroup;
exports.botGame = bots.botGame;
exports.botStart = bots.botStart;
exports.botWebApp = bots.botWebApp;
exports.videoChat = calls.videoChat;
exports.chatFolder = chat.chatFolder;
exports.chatInvite = chat.chatInvite;
exports.message = chat.message;
exports.boost = misc.boost;
exports.folder = misc.folder;
exports.share = misc.share;
exports.mtproxy = proxy.mtproxy;
exports.socks5 = proxy.socks5;
exports.stickerset = stickers.stickerset;
exports.phoneNumber = user.phoneNumber;
exports.publicUsername = user.publicUsername;
exports.temporaryProfile = user.temporaryProfile;
