// @mtcute/core/utils/links/calls.js
import { deeplinkBuilder } from "./common.js";
const videoChat = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: ({ username, inviteHash, isLivestream }) => [
    "resolve",
    {
      domain: username,
      [isLivestream ? "livestream" : "videochat"]: inviteHash || true
    }
  ],
  externalBuild: ({ username, inviteHash, isLivestream }) => [
    username,
    {
      [isLivestream ? "livestream" : "videochat"]: inviteHash || true
    }
  ],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const domain = query.get("domain");
    if (!domain) return null;
    const livestream = query.get("livestream");
    const videochat = query.get("videochat");
    const voicechat = query.get("voicechat");
    if (livestream === null && videochat === null && voicechat === null) return null;
    const inviteHash = livestream || videochat || voicechat;
    return {
      username: domain,
      inviteHash: inviteHash || void 0,
      isLivestream: livestream !== null
    };
  },
  externalParse: (path, query) => {
    if (path.length <= 1 || path.includes("/") || path[0] === "+") return null;
    const livestream = query.get("livestream");
    const videochat = query.get("videochat");
    const voicechat = query.get("voicechat");
    if (livestream === null && videochat === null && voicechat === null) return null;
    const inviteHash = livestream || videochat || voicechat;
    return {
      username: path,
      inviteHash: inviteHash || void 0,
      isLivestream: livestream !== null
    };
  }
});
export {
  videoChat
};
