// @mtcute/core/utils/links/calls.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("./common.cjs");
const videoChat = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ username, inviteHash, isLivestream }) => [
    "resolve",
    {
      domain: username,
      [isLivestream ? "livestream" : "videochat"]: inviteHash || true
    }
  ],
  externalBuild: ({ username, inviteHash, isLivestream }) => [
    username,
    {
      [isLivestream ? "livestream" : "videochat"]: inviteHash || true
    }
  ],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const domain = query.get("domain");
    if (!domain) return null;
    const livestream = query.get("livestream");
    const videochat = query.get("videochat");
    const voicechat = query.get("voicechat");
    if (livestream === null && videochat === null && voicechat === null) return null;
    const inviteHash = livestream || videochat || voicechat;
    return {
      username: domain,
      inviteHash: inviteHash || void 0,
      isLivestream: livestream !== null
    };
  },
  externalParse: (path, query) => {
    if (path.length <= 1 || path.includes("/") || path[0] === "+") return null;
    const livestream = query.get("livestream");
    const videochat = query.get("videochat");
    const voicechat = query.get("voicechat");
    if (livestream === null && videochat === null && voicechat === null) return null;
    const inviteHash = livestream || videochat || voicechat;
    return {
      username: path,
      inviteHash: inviteHash || void 0,
      isLivestream: livestream !== null
    };
  }
});
exports.videoChat = videoChat;
