// @mtcute/core/utils/links/chat.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("./common.cjs");
const chatInvite = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ hash }) => ["join", { invite: hash }],
  internalParse: (path, query) => {
    if (path !== "join") return null;
    const invite = query.get("invite");
    if (!invite) return null;
    return { hash: invite };
  },
  externalBuild: ({ hash }) => [`+${hash}`, null],
  externalParse: (path) => {
    const m = path.match(/^(?:\+|joinchat\/)([\w-]+)$/);
    if (!m) return null;
    if (m[1].match(/^\d+$/)) {
      return null;
    }
    return { hash: m[1] };
  }
});
const chatFolder = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ slug }) => ["addlist", { slug }],
  internalParse: (path, query) => {
    if (path !== "addlist") return null;
    const slug = query.get("slug");
    if (!slug) return null;
    return { slug };
  },
  externalBuild: ({ slug }) => [`addlist/${slug}`, null],
  externalParse: (path) => {
    const [prefix, slug] = path.split("/");
    if (prefix !== "addlist") return null;
    return { slug };
  }
});
function parseMediaTimestamp(timestamp) {
  let m;
  if (m = timestamp.match(/^(\d+)$/)) {
    return Number(m[1]);
  }
  if (m = timestamp.match(/^(\d+):(\d{1,2})$/)) {
    return Number(m[1]) * 60 + Number(m[2]);
  }
  if (m = timestamp.match(/^(?:(\d+)h)?(?:(\d{1,2})m)?(\d{1,2})s$/)) {
    return (Number(m[1]) || 0) * 3600 + (Number(m[2]) || 0) * 60 + (Number(m[3]) || 0);
  }
  return void 0;
}
const message = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: (params) => {
    const common2 = {
      post: params.id,
      thread: params.threadId,
      comment: params.commentId,
      t: params.mediaTimestamp,
      single: params.single ? "" : void 0
    };
    if ("username" in params) {
      return ["resolve", { domain: params.username, ...common2 }];
    }
    return ["privatepost", { channel: params.channelId, ...common2 }];
  },
  internalParse: (path, query) => {
    const common2 = {
      id: Number(query.get("post")),
      threadId: query.has("thread") ? Number(query.get("thread")) : void 0,
      commentId: query.has("comment") ? Number(query.get("comment")) : void 0,
      mediaTimestamp: query.has("t") ? parseMediaTimestamp(query.get("t")) : void 0,
      single: query.has("single")
    };
    if (path === "resolve") {
      const username = query.get("domain");
      if (!username) return null;
      return { username, ...common2 };
    }
    if (path === "privatepost") {
      const channelId = Number(query.get("channel"));
      if (!channelId) return null;
      return { channelId, ...common2 };
    }
    return null;
  },
  externalBuild: (params) => {
    const common2 = {
      comment: params.commentId,
      t: params.mediaTimestamp,
      single: params.single ? "" : void 0
    };
    if ("username" in params) {
      if (params.threadId) {
        return [`${params.username}/${params.threadId}/${params.id}`, common2];
      }
      return [`${params.username}/${params.id}`, common2];
    }
    if (params.threadId) {
      return [`c/${params.channelId}/${params.threadId}/${params.id}`, common2];
    }
    return [`c/${params.channelId}/${params.id}`, common2];
  },
  externalParse: (path, query) => {
    const chunks = path.split("/");
    if (chunks.length < 2) return null;
    const id = Number(chunks[chunks.length - 1]);
    if (Number.isNaN(id)) return null;
    const common2 = {
      id,
      commentId: query.has("comment") ? Number(query.get("comment")) : void 0,
      mediaTimestamp: query.has("t") ? parseMediaTimestamp(query.get("t")) : void 0,
      single: query.has("single")
    };
    if (chunks[0] === "c") {
      const channelId = Number(chunks[1]);
      if (Number.isNaN(channelId)) return null;
      return {
        channelId,
        threadId: chunks[3] ? Number(chunks[2]) : void 0,
        ...common2
      };
    }
    const username = chunks[0];
    if (username[0] === "+") return null;
    return {
      username,
      threadId: chunks[2] ? Number(chunks[1]) : void 0,
      ...common2
    };
  }
});
exports.chatFolder = chatFolder;
exports.chatInvite = chatInvite;
exports.message = message;
