// @mtcute/core/utils/links/misc.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("./common.cjs");
const share = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ url, text }) => ["msg_url", { url, text }],
  internalParse: (path, query) => {
    if (path !== "msg_url") return null;
    const url = query.get("url");
    if (!url) return null;
    const text = query.get("text");
    return { url, text: text || void 0 };
  },
  externalBuild: ({ url, text }) => ["share", { url, text }],
  externalParse: (path, query) => {
    if (path !== "share") return null;
    const url = query.get("url");
    if (!url) return null;
    const text = query.get("text");
    return { url, text: text || void 0 };
  }
});
const boost = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: (params) => {
    if ("username" in params) {
      return ["boost", { domain: params.username }];
    }
    return ["boost", { channel: params.channelId }];
  },
  internalParse: (path, query) => {
    if (path !== "boost") return null;
    const username = query.get("domain");
    if (username) {
      return { username };
    }
    const channelId = Number(query.get("channel"));
    if (!Number.isNaN(channelId)) {
      return { channelId: Number(channelId) };
    }
    return null;
  },
  externalBuild: (params) => {
    if ("username" in params) {
      return [params.username, { boost: true }];
    }
    return [`c/${params.channelId}`, { boost: true }];
  },
  externalParse: (path, query) => {
    if (!query.has("boost")) return null;
    if (path.startsWith("c/")) {
      const channelId = Number(path.slice(2));
      if (Number.isNaN(channelId)) return null;
      return { channelId };
    }
    if (path.includes("/")) return null;
    return { username: path };
  }
});
const folder = /* @__PURE__ */ common.deeplinkBuilder({
  // tg://addlist?slug=XXX
  internalBuild: ({ slug }) => ["addlist", { slug }],
  internalParse: (path, query) => {
    if (path !== "addlist") return null;
    const slug = query.get("slug");
    if (!slug) return null;
    return { slug };
  },
  // https://t.me/addlist/XXX
  externalBuild: ({ slug }) => [`addlist/${slug}`, null],
  externalParse: (path) => {
    const [prefix, slug] = path.split("/");
    if (prefix !== "addlist") return null;
    return { slug };
  }
});
exports.boost = boost;
exports.folder = folder;
exports.share = share;
