// @mtcute/core/utils/links/proxy.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("./common.cjs");
const mtproxy = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: (params) => ["proxy", params],
  externalBuild: (params) => ["proxy", params],
  internalParse: (path, query) => {
    if (path !== "proxy") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const secret = query.get("secret");
    if (!server || Number.isNaN(port) || !secret) return null;
    return { server, port, secret };
  },
  externalParse: (path, query) => {
    if (path !== "proxy") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const secret = query.get("secret");
    if (!server || Number.isNaN(port) || !secret) return null;
    return { server, port, secret };
  }
});
const socks5 = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: (params) => ["socks", params],
  externalBuild: (params) => ["socks", params],
  internalParse: (path, query) => {
    if (path !== "socks") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const user = query.get("user");
    const pass = query.get("pass");
    if (!server || Number.isNaN(port)) return null;
    return { server, port, user: user || void 0, pass: pass || void 0 };
  },
  externalParse: (path, query) => {
    if (path !== "socks") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const user = query.get("user");
    const pass = query.get("pass");
    if (!server || Number.isNaN(port)) return null;
    return { server, port, user: user || void 0, pass: pass || void 0 };
  }
});
exports.mtproxy = mtproxy;
exports.socks5 = socks5;
