// @mtcute/core/utils/links/proxy.js
import { deeplinkBuilder } from "./common.js";
const mtproxy = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: (params) => ["proxy", params],
  externalBuild: (params) => ["proxy", params],
  internalParse: (path, query) => {
    if (path !== "proxy") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const secret = query.get("secret");
    if (!server || Number.isNaN(port) || !secret) return null;
    return { server, port, secret };
  },
  externalParse: (path, query) => {
    if (path !== "proxy") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const secret = query.get("secret");
    if (!server || Number.isNaN(port) || !secret) return null;
    return { server, port, secret };
  }
});
const socks5 = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: (params) => ["socks", params],
  externalBuild: (params) => ["socks", params],
  internalParse: (path, query) => {
    if (path !== "socks") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const user = query.get("user");
    const pass = query.get("pass");
    if (!server || Number.isNaN(port)) return null;
    return { server, port, user: user || void 0, pass: pass || void 0 };
  },
  externalParse: (path, query) => {
    if (path !== "socks") return null;
    const server = query.get("server");
    const port = Number(query.get("port"));
    const user = query.get("user");
    const pass = query.get("pass");
    if (!server || Number.isNaN(port)) return null;
    return { server, port, user: user || void 0, pass: pass || void 0 };
  }
});
export {
  mtproxy,
  socks5
};
