// @mtcute/core/utils/links/user.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("./common.cjs");
const publicUsername = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ username }) => ["resolve", { domain: username }],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const domain = query.get("domain");
    if (!domain) return null;
    if ([...query.keys()].length > 1) return null;
    return { username: domain };
  },
  externalBuild: ({ username }) => [username, null],
  externalParse: (path, query) => {
    if (path.length <= 1 || path.includes("/") || path[0] === "+") return null;
    if ([...query.keys()].length > 0) return null;
    return { username: path };
  }
});
const temporaryProfile = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ token }) => ["contact", { token }],
  internalParse: (path, query) => {
    if (path !== "contact") return null;
    const token = query.get("token");
    if (!token) return null;
    return { token };
  },
  externalBuild: ({ token }) => [`contact/${token}`, null],
  externalParse: (path) => {
    const [prefix, token] = path.split("/");
    if (prefix !== "contact") return null;
    return { token };
  }
});
const phoneNumber = /* @__PURE__ */ common.deeplinkBuilder({
  internalBuild: ({ phone }) => ["resolve", { phone }],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const phone = query.get("phone");
    if (!phone) return null;
    return { phone };
  },
  externalBuild: ({ phone }) => [`+${phone}`, null],
  externalParse: (path) => {
    const m = path.match(/^\+(\d+)$/);
    if (!m) return null;
    return { phone: m[1] };
  }
});
exports.phoneNumber = phoneNumber;
exports.publicUsername = publicUsername;
exports.temporaryProfile = temporaryProfile;
