// @mtcute/core/utils/links/user.js
import { deeplinkBuilder } from "./common.js";
const publicUsername = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: ({ username }) => ["resolve", { domain: username }],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const domain = query.get("domain");
    if (!domain) return null;
    if ([...query.keys()].length > 1) return null;
    return { username: domain };
  },
  externalBuild: ({ username }) => [username, null],
  externalParse: (path, query) => {
    if (path.length <= 1 || path.includes("/") || path[0] === "+") return null;
    if ([...query.keys()].length > 0) return null;
    return { username: path };
  }
});
const temporaryProfile = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: ({ token }) => ["contact", { token }],
  internalParse: (path, query) => {
    if (path !== "contact") return null;
    const token = query.get("token");
    if (!token) return null;
    return { token };
  },
  externalBuild: ({ token }) => [`contact/${token}`, null],
  externalParse: (path) => {
    const [prefix, token] = path.split("/");
    if (prefix !== "contact") return null;
    return { token };
  }
});
const phoneNumber = /* @__PURE__ */ deeplinkBuilder({
  internalBuild: ({ phone }) => ["resolve", { phone }],
  internalParse: (path, query) => {
    if (path !== "resolve") return null;
    const phone = query.get("phone");
    if (!phone) return null;
    return { phone };
  },
  externalBuild: ({ phone }) => [`+${phone}`, null],
  externalParse: (path) => {
    const m = path.match(/^\+(\d+)$/);
    if (!m) return null;
    return { phone: m[1] };
  }
});
export {
  phoneNumber,
  publicUsername,
  temporaryProfile
};
