// @mtcute/node/client.js
import { createInterface } from "node:readline";
import { BaseTelegramClient as BaseTelegramClient$1, TelegramClient as TelegramClient$1 } from "@mtcute/core/client.js";
import { downloadToFile } from "./methods/download-file.js";
import { downloadAsNodeStream } from "./methods/download-node-stream.js";
import { SqliteStorage } from "./sqlite/index.js";
import { NodeCryptoProvider } from "./utils/crypto.js";
import { NodePlatform } from "./utils/platform.js";
import { TcpTransport } from "./utils/tcp.js";
class BaseTelegramClient extends BaseTelegramClient$1 {
  constructor(opts) {
    super({
      crypto: new NodeCryptoProvider(),
      transport: new TcpTransport(),
      platform: new NodePlatform(),
      ...opts,
      storage: typeof opts.storage === "string" ? new SqliteStorage(opts.storage) : opts.storage ?? new SqliteStorage("client.session")
    });
  }
}
class TelegramClient extends TelegramClient$1 {
  constructor(opts) {
    if ("client" in opts) {
      super(opts);
      return;
    }
    super({
      client: new BaseTelegramClient(opts),
      disableUpdates: opts.disableUpdates,
      skipConversationUpdates: opts.skipConversationUpdates,
      updates: opts.updates
    });
  }
  _rl;
  /**
   * Tiny wrapper over Node `readline` package
   * for simpler user input for `.start()` method.
   *
   * Associated `readline` interface is closed
   * after `start()` returns, or with the client.
   *
   * @param text  Text of the question
   */
  input(text) {
    if (!this._rl) {
      this._rl = createInterface({
        input: process.stdin,
        output: process.stdout
      });
    }
    return new Promise((res) => this._rl?.question(text, res));
  }
  destroy() {
    this._rl?.close();
    return super.destroy();
  }
  start(params = {}) {
    if (!params.botToken) {
      if (!params.phone) params.phone = () => this.input("phone > ");
      if (!params.code) params.code = () => this.input("code > ");
      if (!params.password) {
        params.password = () => this.input("2fa password > ");
      }
    }
    return super.start(params).then((user) => {
      if (this._rl) {
        this._rl.close();
        delete this._rl;
      }
      return user;
    });
  }
  downloadToFile(filename, location, params) {
    return downloadToFile(this, filename, location, params);
  }
  downloadAsNodeStream(location, params) {
    return downloadAsNodeStream(this, location, params);
  }
}
export {
  BaseTelegramClient,
  TelegramClient
};
