// @mtcute/node/client.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_readline = require("node:readline");
const client_js = require("@mtcute/core/client.js");
const downloadFile = require("./methods/download-file.cjs");
const downloadNodeStream = require("./methods/download-node-stream.cjs");
const index = require("./sqlite/index.cjs");
const crypto = require("./utils/crypto.cjs");
const platform = require("./utils/platform.cjs");
const tcp = require("./utils/tcp.cjs");
class BaseTelegramClient extends client_js.BaseTelegramClient {
  constructor(opts) {
    super({
      crypto: new crypto.NodeCryptoProvider(),
      transport: new tcp.TcpTransport(),
      platform: new platform.NodePlatform(),
      ...opts,
      storage: typeof opts.storage === "string" ? new index.SqliteStorage(opts.storage) : opts.storage ?? new index.SqliteStorage("client.session")
    });
  }
}
class TelegramClient extends client_js.TelegramClient {
  constructor(opts) {
    if ("client" in opts) {
      super(opts);
      return;
    }
    super({
      client: new BaseTelegramClient(opts),
      disableUpdates: opts.disableUpdates,
      skipConversationUpdates: opts.skipConversationUpdates,
      updates: opts.updates
    });
  }
  _rl;
  /**
   * Tiny wrapper over Node `readline` package
   * for simpler user input for `.start()` method.
   *
   * Associated `readline` interface is closed
   * after `start()` returns, or with the client.
   *
   * @param text  Text of the question
   */
  input(text) {
    if (!this._rl) {
      this._rl = node_readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });
    }
    return new Promise((res) => this._rl?.question(text, res));
  }
  destroy() {
    this._rl?.close();
    return super.destroy();
  }
  start(params = {}) {
    if (!params.botToken) {
      if (!params.phone) params.phone = () => this.input("phone > ");
      if (!params.code) params.code = () => this.input("code > ");
      if (!params.password) {
        params.password = () => this.input("2fa password > ");
      }
    }
    return super.start(params).then((user) => {
      if (this._rl) {
        this._rl.close();
        delete this._rl;
      }
      return user;
    });
  }
  downloadToFile(filename, location, params) {
    return downloadFile.downloadToFile(this, filename, location, params);
  }
  downloadAsNodeStream(location, params) {
    return downloadNodeStream.downloadAsNodeStream(this, location, params);
  }
}
exports.BaseTelegramClient = BaseTelegramClient;
exports.TelegramClient = TelegramClient;
