// @mtcute/node/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const client = require("./client.cjs");
const index = require("./sqlite/index.cjs");
const platform = require("./utils/platform.cjs");
const proxies = require("./utils/proxies.cjs");
const tcp = require("./utils/tcp.cjs");
const worker = require("./worker.cjs");
const core = require("@mtcute/core");
const htmlParser = require("@mtcute/html-parser");
const markdownParser = require("@mtcute/markdown-parser");
const driver = require("./sqlite/driver.cjs");
const net = require("@fuman/net");
exports.BaseTelegramClient = client.BaseTelegramClient;
exports.TelegramClient = client.TelegramClient;
exports.SqliteStorage = index.SqliteStorage;
exports.NodePlatform = platform.NodePlatform;
exports.HttpProxyTcpTransport = proxies.HttpProxyTcpTransport;
exports.MtProxyTcpTransport = proxies.MtProxyTcpTransport;
exports.SocksProxyTcpTransport = proxies.SocksProxyTcpTransport;
exports.proxyTransportFromUrl = proxies.proxyTransportFromUrl;
exports.TcpTransport = tcp.TcpTransport;
exports.TelegramWorker = worker.TelegramWorker;
exports.TelegramWorkerPort = worker.TelegramWorkerPort;
exports.SqliteStorageDriver = driver.SqliteStorageDriver;
Object.defineProperty(exports, "HttpProxyConnectionError", {
  enumerable: true,
  get: () => net.HttpProxyConnectionError
});
Object.defineProperty(exports, "SocksProxyConnectionError", {
  enumerable: true,
  get: () => net.SocksProxyConnectionError
});
Object.keys(core).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => core[k]
  });
});
Object.keys(htmlParser).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => htmlParser[k]
  });
});
Object.keys(markdownParser).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => markdownParser[k]
  });
});
