// @mtcute/node/worker.js
import { parentPort, Worker, MessagePort } from "node:worker_threads";
import { TelegramWorker as TelegramWorker$1, TelegramWorkerPort as TelegramWorkerPort$1 } from "@mtcute/core/worker.js";
import { NodePlatform } from "./utils/platform.js";
class TelegramWorker extends TelegramWorker$1 {
  registerWorker(handler) {
    if (!parentPort) {
      throw new Error("TelegramWorker must be created from a worker thread");
    }
    const port = parentPort;
    const respond = port.postMessage.bind(port);
    const messageHandler = (message) => handler(message, respond);
    port.on("message", messageHandler);
    return [respond, () => port.off("message", messageHandler)];
  }
}
class TelegramWorkerPort extends TelegramWorkerPort$1 {
  constructor(options) {
    super({
      worker: options.worker,
      workerId: options.workerId,
      platform: new NodePlatform()
    });
  }
  connectToWorker(worker, handler) {
    if (!(worker instanceof Worker) && !(worker instanceof MessagePort)) {
      throw new TypeError("Only worker_threads and MessagePorts are supported");
    }
    const send = worker.postMessage.bind(worker);
    worker.on("message", handler);
    return [
      send,
      () => {
        worker.off("message", handler);
      }
    ];
  }
}
export {
  TelegramWorker,
  TelegramWorkerPort
};
