// @mtcute/node/worker.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_worker_threads = require("node:worker_threads");
const worker_js = require("@mtcute/core/worker.js");
const platform = require("./utils/platform.cjs");
class TelegramWorker extends worker_js.TelegramWorker {
  registerWorker(handler) {
    if (!node_worker_threads.parentPort) {
      throw new Error("TelegramWorker must be created from a worker thread");
    }
    const port = node_worker_threads.parentPort;
    const respond = port.postMessage.bind(port);
    const messageHandler = (message) => handler(message, respond);
    port.on("message", messageHandler);
    return [respond, () => port.off("message", messageHandler)];
  }
}
class TelegramWorkerPort extends worker_js.TelegramWorkerPort {
  constructor(options) {
    super({
      worker: options.worker,
      workerId: options.workerId,
      platform: new platform.NodePlatform()
    });
  }
  connectToWorker(worker, handler) {
    if (!(worker instanceof node_worker_threads.Worker) && !(worker instanceof node_worker_threads.MessagePort)) {
      throw new TypeError("Only worker_threads and MessagePorts are supported");
    }
    const send = worker.postMessage.bind(worker);
    worker.on("message", handler);
    return [
      send,
      () => {
        worker.off("message", handler);
      }
    ];
  }
}
exports.TelegramWorker = TelegramWorker;
exports.TelegramWorkerPort = TelegramWorkerPort;
