// @mtcute/html-parser/index.js
import { parseDateEntityFormat, dateEntityFormatToString } from "@mtcute/core/utils.js";
import { Parser } from "htmlparser2";
import Long from "long";
const MENTION_REGEX = /^tg:\/\/user\?id=(\d+)(?:&hash=(-?[0-9a-fA-F]+)(?:&|$)|&|$)/;
function escape(str, quote = false) {
  str = str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  if (quote) str = str.replace(/"/g, "&quot;");
  return str;
}
function dedent(strings) {
  const joined = strings.join("\0");
  const lines = joined.split("\n");
  let minIndent = Infinity;
  for (let i = 1; i < lines.length; i++) {
    const match = lines[i].match(/^(\s*)\S/);
    if (match) minIndent = Math.min(minIndent, match[1].length);
  }
  if (minIndent === Infinity) minIndent = 0;
  const result = [];
  let remainder = joined;
  for (let i = 0; i < strings.length; i++) {
    const part = remainder.slice(0, strings[i].length);
    remainder = remainder.slice(strings[i].length + 1);
    const partLines = part.split("\n");
    for (let j = 0; j < partLines.length; j++) {
      if (i === 0 && j === 0) continue;
      partLines[j] = partLines[j].slice(minIndent);
    }
    result.push(partLines.join("\n"));
  }
  if (result[0].startsWith("\n")) result[0] = result[0].slice(1);
  result[result.length - 1] = result[result.length - 1].replace(/\n[ \t]*$/, "");
  return result;
}
function parse(keepWhitespace, strings, ...sub) {
  const stacks = {};
  const entities = [];
  let plainText = "";
  let pendingText = "";
  let isInsideAttrib = false;
  function processPendingText(tagEnd = false, forceKeepWhitespace = false) {
    if (!pendingText.length) return;
    if (!stacks.pre?.length && !keepWhitespace && !forceKeepWhitespace) {
      pendingText = pendingText.replace(/[^\S\u00A0]+/g, " ");
      if (tagEnd) pendingText = pendingText.trimEnd();
      if (!plainText.length || plainText.match(/\s$/)) {
        pendingText = pendingText.trimStart();
      }
    }
    for (const ents of Object.values(stacks)) {
      for (const ent of ents) {
        ent.length += pendingText.length;
      }
    }
    plainText += pendingText;
    pendingText = "";
  }
  const parser = new Parser({
    onopentag(name, attribs) {
      name = name.toLowerCase();
      processPendingText();
      if (stacks.pre?.length && name !== "pre") {
        if (name === "code" && attribs.class) {
          const match = attribs.class.match(/language-(\S+)/);
          if (match) {
            const preEntity = stacks.pre[stacks.pre.length - 1];
            preEntity.language = match[1];
          }
        }
        return;
      }
      let entity;
      switch (name) {
        case "br":
          pendingText += "\n";
          processPendingText(true, true);
          return;
        case "b":
        case "strong":
          entity = {
            _: "messageEntityBold",
            offset: plainText.length,
            length: 0
          };
          break;
        case "i":
        case "em":
          entity = {
            _: "messageEntityItalic",
            offset: plainText.length,
            length: 0
          };
          break;
        case "u":
        case "ins":
          entity = {
            _: "messageEntityUnderline",
            offset: plainText.length,
            length: 0
          };
          break;
        case "s":
        case "del":
        case "strike":
          entity = {
            _: "messageEntityStrike",
            offset: plainText.length,
            length: 0
          };
          break;
        case "blockquote":
          entity = {
            _: "messageEntityBlockquote",
            offset: plainText.length,
            length: 0,
            collapsed: "collapsible" in attribs || "expandable" in attribs
          };
          break;
        case "code":
          entity = {
            _: "messageEntityCode",
            offset: plainText.length,
            length: 0
          };
          break;
        case "pre":
          entity = {
            _: "messageEntityPre",
            offset: plainText.length,
            length: 0,
            language: attribs.language ?? ""
          };
          break;
        case "spoiler":
        case "tg-spoiler":
          entity = {
            _: "messageEntitySpoiler",
            offset: plainText.length,
            length: 0
          };
          break;
        case "span":
          if (attribs.class !== "tg-spoiler") return;
          entity = {
            _: "messageEntitySpoiler",
            offset: plainText.length,
            length: 0
          };
          break;
        case "emoji":
        case "tg-emoji": {
          const id = attribs.id || attribs["emoji-id"];
          if (!id || !id.match(/^-?\d+$/)) return;
          entity = {
            _: "messageEntityCustomEmoji",
            offset: plainText.length,
            length: 0,
            documentId: Long.fromString(id)
          };
          break;
        }
        case "tg-time": {
          const unix = attribs.unix;
          if (!unix || !unix.match(/^\d+$/)) return;
          entity = {
            _: "messageEntityFormattedDate",
            offset: plainText.length,
            length: 0,
            date: Number.parseInt(unix),
            ...attribs.format ? parseDateEntityFormat(attribs.format) : {}
          };
          break;
        }
        case "time": {
          const datetime = attribs.datetime;
          if (!datetime) return;
          const date = Math.floor(new Date(datetime).getTime() / 1e3);
          if (Number.isNaN(date)) return;
          entity = {
            _: "messageEntityFormattedDate",
            offset: plainText.length,
            length: 0,
            date,
            ...attribs.format ? parseDateEntityFormat(attribs.format) : {}
          };
          break;
        }
        case "a": {
          let url = attribs.href;
          if (!url) return;
          const mention = MENTION_REGEX.exec(url);
          if (mention) {
            const id = Number.parseInt(mention[1]);
            const accessHash = mention[2];
            if (accessHash) {
              entity = {
                _: "inputMessageEntityMentionName",
                offset: plainText.length,
                length: 0,
                userId: {
                  _: "inputUser",
                  userId: id,
                  accessHash: Long.fromString(accessHash, false, 16)
                }
              };
            } else {
              entity = {
                _: "messageEntityMentionName",
                offset: plainText.length,
                length: 0,
                userId: id
              };
            }
          } else {
            if (url.match(/^\/\//)) url = `http:${url}`;
            entity = {
              _: "messageEntityTextUrl",
              offset: plainText.length,
              length: 0,
              url
            };
          }
          break;
        }
        default:
          return;
      }
      if (!(name in stacks)) {
        stacks[name] = [];
      }
      stacks[name].push(entity);
    },
    onclosetag(name) {
      processPendingText(true);
      name = name.toLowerCase();
      if (name !== "pre" && stacks.pre?.length) return;
      const entity = stacks[name]?.pop();
      if (!entity) return;
      if (name !== "pre" || !stacks.pre?.length) {
        entities.push(entity);
      }
    },
    ontext(data) {
      pendingText += data;
    }
  });
  const onattribname = parser.onattribname;
  const onattribend = parser.onattribend;
  parser.onattribname = function(startIdx, endIdx) {
    onattribname.call(this, startIdx, endIdx);
    isInsideAttrib = true;
  };
  parser.onattribend = function(quote, endIdx) {
    onattribend.call(this, quote, endIdx);
    isInsideAttrib = false;
  };
  if (typeof strings === "string") strings = [strings];
  if (keepWhitespace) strings = dedent(strings);
  sub.forEach((it, idx) => {
    parser.write(strings[idx]);
    if (typeof it === "boolean" || !it) return;
    if (isInsideAttrib) {
      let text;
      if (typeof it === "string") {
        text = it;
      } else if (typeof it === "number") {
        text = it.toString();
      } else if (Long.isLong(it)) {
        text = it.toString(10);
      } else {
        text = it.text;
      }
      parser.write(escape(text, true));
      return;
    }
    if (typeof it === "string") {
      processPendingText();
      pendingText += it;
      processPendingText(false, true);
    } else if (Long.isLong(it) || typeof it === "number") {
      pendingText += it.toString(10);
    } else {
      const text = it.text;
      const innerEntities = "raw" in it ? [it.raw] : it.entities;
      processPendingText();
      const baseOffset = plainText.length;
      pendingText += text;
      if (innerEntities) {
        for (const ent of innerEntities) {
          entities.push({ ...ent, offset: ent.offset + baseOffset });
        }
      }
      processPendingText(false, true);
    }
  });
  parser.write(strings[strings.length - 1]);
  processPendingText(true);
  return {
    text: plainText.replace(/\u00A0/g, " "),
    entities
  };
}
function _unparse(keepWhitespace, text, entities, params, entitiesOffset = 0, offset = 0, length = text.length) {
  if (!text) return text;
  if (!entities.length || entities.length === entitiesOffset) {
    if (keepWhitespace) return escape(text);
    return escape(text).replace(/\n/g, "<br>").replace(/ {2,}/g, (match) => {
      return "&nbsp;".repeat(match.length);
    });
  }
  const end = offset + length;
  const html2 = [];
  let lastOffset = 0;
  for (let i = entitiesOffset; i < entities.length; i++) {
    const entity = entities[i];
    if (entity.offset >= end) break;
    let entOffset = entity.offset;
    let length2 = entity.length;
    if (entOffset < 0) {
      length2 += entOffset;
      entOffset = 0;
    }
    let relativeOffset = entOffset - offset;
    if (relativeOffset > lastOffset) {
      html2.push(escape(text.substring(lastOffset, relativeOffset)));
    } else if (relativeOffset < lastOffset) {
      length2 -= lastOffset - relativeOffset;
      relativeOffset = lastOffset;
    }
    if (length2 <= 0 || relativeOffset >= end || relativeOffset < 0) {
      continue;
    }
    let skip = false;
    const substr = text.substr(relativeOffset, length2);
    if (!substr) continue;
    const type = entity._;
    let entityText;
    if (type === "messageEntityPre") {
      entityText = substr;
    } else {
      entityText = _unparse(keepWhitespace, substr, entities, params, i + 1, offset + relativeOffset, length2);
    }
    switch (type) {
      case "messageEntityBold":
      case "messageEntityItalic":
      case "messageEntityUnderline":
      case "messageEntityStrike":
      case "messageEntityCode":
      case "messageEntitySpoiler":
        {
          const tag = {
            messageEntityBold: "b",
            messageEntityItalic: "i",
            messageEntityUnderline: "u",
            messageEntityStrike: "s",
            messageEntityCode: "code",
            messageEntitySpoiler: "spoiler"
          }[type];
          html2.push(`<${tag}>${entityText}</${tag}>`);
        }
        break;
      case "messageEntityBlockquote":
        html2.push(`<blockquote${entity.collapsed ? " collapsible" : ""}>${entityText}</blockquote>`);
        break;
      case "messageEntityPre":
        html2.push(
          `<pre${entity.language ? ` language="${entity.language}"` : ""}>${params.syntaxHighlighter && entity.language ? params.syntaxHighlighter(entityText, entity.language) : entityText}</pre>`
        );
        break;
      case "messageEntityEmail":
        html2.push(`<a href="mailto:${entityText}">${entityText}</a>`);
        break;
      case "messageEntityUrl":
        html2.push(`<a href="${entityText}">${entityText}</a>`);
        break;
      case "messageEntityTextUrl":
        html2.push(`<a href="${escape(entity.url, true)}">${entityText}</a>`);
        break;
      case "messageEntityMentionName":
        html2.push(`<a href="tg://user?id=${entity.userId}">${entityText}</a>`);
        break;
      case "messageEntityFormattedDate": {
        const fmt = dateEntityFormatToString(entity);
        html2.push(`<tg-time unix="${entity.date}"${fmt ? ` format="${fmt}"` : ""}>${entityText}</tg-time>`);
        break;
      }
      default:
        skip = true;
        break;
    }
    lastOffset = relativeOffset + (skip ? 0 : length2);
  }
  html2.push(escape(text.substr(lastOffset)));
  return html2.join("");
}
function unparse(keepWhitespace, input, options) {
  if (typeof input === "string") {
    return _unparse(keepWhitespace, input, [], options ?? {});
  }
  return _unparse(keepWhitespace, input.text, input.entities ?? [], options ?? {});
}
const html = Object.assign(
  parse.bind(null, false),
  { escape, unparse: unparse.bind(null, false) }
);
const thtml = Object.assign(
  parse.bind(null, true),
  { escape, unparse: unparse.bind(null, true) }
);
export {
  html,
  thtml
};
