// @mtcute/core/highlevel/types/media/document.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const memoize = require("../../utils/memoize.cjs");
const thumbnail = require("./thumbnail.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const fileLocation = require("../files/file-location.cjs");
class RawDocument extends fileLocation.FileLocation {
  constructor(raw) {
    super(
      {
        _: "inputDocumentFileLocation",
        id: raw.id,
        fileReference: raw.fileReference,
        accessHash: raw.accessHash,
        thumbSize: ""
      },
      raw.size,
      raw.dcId
    );
    this.raw = raw;
    this.raw = raw;
  }
  /**
   * Original file name, extracted from the document
   * attributes.
   */
  get fileName() {
    const attr = this.raw.attributes.find((it) => it._ === "documentAttributeFilename");
    return attr?.fileName ?? null;
  }
  /**
   * File MIME type, as defined by the sender.
   */
  get mimeType() {
    return this.raw.mimeType;
  }
  /**
   * Date the document was sent
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Available thumbnails, if any.
   *
   * If there are no thumbnails, the array will be empty.
   */
  get thumbnails() {
    const arr = [];
    this.raw.thumbs?.forEach((sz) => arr.push(new thumbnail.Thumbnail(this.raw, sz)));
    this.raw.videoThumbs?.forEach((sz) => arr.push(new thumbnail.Thumbnail(this.raw, sz)));
    return arr;
  }
  /**
   * Get a thumbnail by its type.
   *
   * Thumbnail types are described in the
   * [Telegram docs](https://core.telegram.org/api/files#image-thumbnail-types),
   * and are also available as static members of {@link Thumbnail} for convenience.
   *
   * @param type  Thumbnail type
   */
  getThumbnail(type) {
    return this.thumbnails.find((it) => {
      if (it.raw._ === "videoSizeEmojiMarkup" || it.raw._ === "videoSizeStickerMarkup") {
        return false;
      }
      return it.raw.type === type;
    }) ?? null;
  }
  /**
   * Input document TL object generated from this object,
   * to be used with methods that use it
   */
  get inputDocument() {
    return {
      _: "inputDocument",
      id: this.raw.id,
      accessHash: this.raw.accessHash,
      fileReference: this.raw.fileReference
    };
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaDocument",
      id: this.inputDocument
    };
  }
  _fileIdType() {
    return fileId.tdFileId.FileType.Document;
  }
  /**
   * Get TDLib and Bot API compatible File ID
   * representing this document.
   */
  get fileId() {
    return fileId.toFileId({
      type: this._fileIdType(),
      dcId: this.raw.dcId,
      fileReference: this.raw.fileReference,
      location: {
        _: "common",
        id: this.raw.id,
        accessHash: this.raw.accessHash
      }
    });
  }
  /**
   * Get a unique File ID representing this document.
   */
  get uniqueFileId() {
    return fileId.toUniqueFileId(fileId.tdFileId.FileType.Document, {
      _: "common",
      id: this.raw.id
    });
  }
}
class Document extends RawDocument {
  type = "document";
}
const Document$1 = /* @__PURE__ */ memoize.memoizeGetters(Document, ["fileName", "thumbnails", "fileId", "uniqueFileId"]);
const Document$2 = /* @__PURE__ */ inspectable.makeInspectable(Document$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument"]);
exports.Document = Document$2;
exports.RawDocument = RawDocument;
