// @mtcute/core/highlevel/types/media/document.js
import { tdFileId, toFileId, toUniqueFileId } from "@mtcute/file-id";
import { memoizeGetters } from "../../utils/memoize.js";
import { Thumbnail as Thumbnail$2 } from "./thumbnail.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { FileLocation as FileLocation$1 } from "../files/file-location.js";
class RawDocument extends FileLocation$1 {
  constructor(raw) {
    super(
      {
        _: "inputDocumentFileLocation",
        id: raw.id,
        fileReference: raw.fileReference,
        accessHash: raw.accessHash,
        thumbSize: ""
      },
      raw.size,
      raw.dcId
    );
    this.raw = raw;
    this.raw = raw;
  }
  /**
   * Original file name, extracted from the document
   * attributes.
   */
  get fileName() {
    const attr = this.raw.attributes.find((it) => it._ === "documentAttributeFilename");
    return attr?.fileName ?? null;
  }
  /**
   * File MIME type, as defined by the sender.
   */
  get mimeType() {
    return this.raw.mimeType;
  }
  /**
   * Date the document was sent
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Available thumbnails, if any.
   *
   * If there are no thumbnails, the array will be empty.
   */
  get thumbnails() {
    const arr = [];
    this.raw.thumbs?.forEach((sz) => arr.push(new Thumbnail$2(this.raw, sz)));
    this.raw.videoThumbs?.forEach((sz) => arr.push(new Thumbnail$2(this.raw, sz)));
    return arr;
  }
  /**
   * Get a thumbnail by its type.
   *
   * Thumbnail types are described in the
   * [Telegram docs](https://core.telegram.org/api/files#image-thumbnail-types),
   * and are also available as static members of {@link Thumbnail} for convenience.
   *
   * @param type  Thumbnail type
   */
  getThumbnail(type) {
    return this.thumbnails.find((it) => {
      if (it.raw._ === "videoSizeEmojiMarkup" || it.raw._ === "videoSizeStickerMarkup") {
        return false;
      }
      return it.raw.type === type;
    }) ?? null;
  }
  /**
   * Input document TL object generated from this object,
   * to be used with methods that use it
   */
  get inputDocument() {
    return {
      _: "inputDocument",
      id: this.raw.id,
      accessHash: this.raw.accessHash,
      fileReference: this.raw.fileReference
    };
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaDocument",
      id: this.inputDocument
    };
  }
  _fileIdType() {
    return tdFileId.FileType.Document;
  }
  /**
   * Get TDLib and Bot API compatible File ID
   * representing this document.
   */
  get fileId() {
    return toFileId({
      type: this._fileIdType(),
      dcId: this.raw.dcId,
      fileReference: this.raw.fileReference,
      location: {
        _: "common",
        id: this.raw.id,
        accessHash: this.raw.accessHash
      }
    });
  }
  /**
   * Get a unique File ID representing this document.
   */
  get uniqueFileId() {
    return toUniqueFileId(tdFileId.FileType.Document, {
      _: "common",
      id: this.raw.id
    });
  }
}
class Document extends RawDocument {
  type = "document";
}
const Document$1 = /* @__PURE__ */ memoizeGetters(Document, ["fileName", "thumbnails", "fileId", "uniqueFileId"]);
const Document$2 = /* @__PURE__ */ makeInspectable(Document$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument"]);
export {
  Document$2 as Document,
  RawDocument
};
