// @mtcute/core/highlevel/types/media/extended-media.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Thumbnail as Thumbnail$2 } from "./thumbnail.js";
class ExtendedMediaPreview {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Width of the preview, in pixels (if available, else 0)
   */
  get width() {
    return this.raw.w ?? 0;
  }
  /**
   * Height of the preview, in pixels (if available, else 0)
   */
  get height() {
    return this.raw.h ?? 0;
  }
  get thumbnail() {
    if (!this.raw.thumb) {
      return null;
    }
    return new Thumbnail$2(this.raw, this.raw.thumb);
  }
  /**
   * If this is a video, the duration of the video,
   * in seconds (if available, else 0)
   */
  get videoDuration() {
    return this.raw.videoDuration ?? 0;
  }
}
const ExtendedMediaPreview$1 = /* @__PURE__ */ memoizeGetters(ExtendedMediaPreview, ["thumbnail"]);
const ExtendedMediaPreview$2 = /* @__PURE__ */ makeInspectable(ExtendedMediaPreview$1);
export {
  ExtendedMediaPreview$2 as ExtendedMediaPreview
};
