// @mtcute/core/highlevel/types/media/extended-media.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const thumbnail = require("./thumbnail.cjs");
class ExtendedMediaPreview {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Width of the preview, in pixels (if available, else 0)
   */
  get width() {
    return this.raw.w ?? 0;
  }
  /**
   * Height of the preview, in pixels (if available, else 0)
   */
  get height() {
    return this.raw.h ?? 0;
  }
  get thumbnail() {
    if (!this.raw.thumb) {
      return null;
    }
    return new thumbnail.Thumbnail(this.raw, this.raw.thumb);
  }
  /**
   * If this is a video, the duration of the video,
   * in seconds (if available, else 0)
   */
  get videoDuration() {
    return this.raw.videoDuration ?? 0;
  }
}
const ExtendedMediaPreview$1 = /* @__PURE__ */ memoize.memoizeGetters(ExtendedMediaPreview, ["thumbnail"]);
const ExtendedMediaPreview$2 = /* @__PURE__ */ inspectable.makeInspectable(ExtendedMediaPreview$1);
exports.ExtendedMediaPreview = ExtendedMediaPreview$2;
