// @mtcute/core/highlevel/types/media/game.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const photo = require("./photo.cjs");
const video = require("./video.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Game {
  constructor(game) {
    this.game = game;
  }
  type = "game";
  /**
   * Unique identifier of the game.
   */
  get id() {
    return this.game.id;
  }
  /**
   * Title of the game
   */
  get title() {
    return this.game.title;
  }
  /**
   * Description of the game
   */
  get description() {
    return this.game.description;
  }
  /**
   * Unique short name of the game
   */
  get shortName() {
    return this.game.shortName;
  }
  /**
   * Photo that will be displayed in the game message in chats
   */
  get photo() {
    if (this.game.photo._ === "photoEmpty") return null;
    return new photo.Photo(this.game.photo);
  }
  /**
   * Animation that will be displayed in the game message in chats
   */
  get animation() {
    if (this.game.document?._ !== "document") return null;
    const attr = this.game.document.attributes.find((it) => it._ === "documentAttributeVideo");
    if (!attr) {
      return null;
    }
    return new video.Video(this.game.document, attr);
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaGame",
      id: {
        _: "inputGameID",
        id: this.game.id,
        accessHash: this.game.accessHash
      }
    };
  }
}
const Game$1 = /* @__PURE__ */ memoize.memoizeGetters(Game, ["photo", "animation"]);
const Game$2 = /* @__PURE__ */ inspectable.makeInspectable(Game$1, void 0, ["inputMedia"]);
exports.Game = Game$2;
