// @mtcute/core/highlevel/types/media/game.js
import { memoizeGetters } from "../../utils/memoize.js";
import { Photo as Photo$2 } from "./photo.js";
import { Video as Video$2 } from "./video.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Game {
  constructor(game) {
    this.game = game;
  }
  type = "game";
  /**
   * Unique identifier of the game.
   */
  get id() {
    return this.game.id;
  }
  /**
   * Title of the game
   */
  get title() {
    return this.game.title;
  }
  /**
   * Description of the game
   */
  get description() {
    return this.game.description;
  }
  /**
   * Unique short name of the game
   */
  get shortName() {
    return this.game.shortName;
  }
  /**
   * Photo that will be displayed in the game message in chats
   */
  get photo() {
    if (this.game.photo._ === "photoEmpty") return null;
    return new Photo$2(this.game.photo);
  }
  /**
   * Animation that will be displayed in the game message in chats
   */
  get animation() {
    if (this.game.document?._ !== "document") return null;
    const attr = this.game.document.attributes.find((it) => it._ === "documentAttributeVideo");
    if (!attr) {
      return null;
    }
    return new Video$2(this.game.document, attr);
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaGame",
      id: {
        _: "inputGameID",
        id: this.game.id,
        accessHash: this.game.accessHash
      }
    };
  }
}
const Game$1 = /* @__PURE__ */ memoizeGetters(Game, ["photo", "animation"]);
const Game$2 = /* @__PURE__ */ makeInspectable(Game$1, void 0, ["inputMedia"]);
export {
  Game$2 as Game
};
