// @mtcute/core/highlevel/types/media/invoice.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const memoize = require("../../utils/memoize.cjs");
const webDocument = require("../files/web-document.cjs");
const extendedMedia = require("./extended-media.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Invoice {
  constructor(raw, _extendedMedia) {
    this.raw = raw;
    this._extendedMedia = _extendedMedia;
  }
  type = "invoice";
  /**
   * Whether the shipping address was requested
   */
  isShippingAddressRequested() {
    return this.raw.shippingAddressRequested;
  }
  /**
   * Whether this is an example (test) invoice
   */
  isTest() {
    return this.raw.test;
  }
  /**
   * Product name, 1-32 characters
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Product description, 1-255 characters
   */
  get description() {
    return this.raw.description;
  }
  /**
   * URL of the product photo for the invoice
   */
  get photo() {
    if (!this.raw.photo) return null;
    return new webDocument.WebDocument(this.raw.photo);
  }
  /**
   * Message ID of receipt
   */
  get receiptMessageId() {
    return this.raw.receiptMsgId ?? null;
  }
  /**
   * Three-letter ISO 4217 currency code
   */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Total price in the smallest units of the currency
   * (integer, not float/double). For example, for a price
   * of `US$ 1.45` `amount = 145`.
   *
   * See the exp parameter in [currencies.json](https://core.telegram.org/bots/payments/currencies.json),
   * it shows the number of digits past the decimal point
   * for each currency (2 for the majority of currencies).
   */
  get amount() {
    return this.raw.totalAmount;
  }
  /**
   * Unique bot deep-linking parameter that can be used to generate this invoice
   */
  get startParam() {
    return this.raw.startParam;
  }
  /**
   * If this invoice has extended media
   */
  get extendedMediaState() {
    if (!this.raw.extendedMedia) return "none";
    if (this.raw.extendedMedia._ === "messageExtendedMediaPreview") {
      return "preview";
    }
    return "full";
  }
  /**
   * Get the invoice's extended media preview.
   * Only available if {@link extendedMediaState} is `preview`.
   * Otherwise, throws an error.
   */
  get extendedMediaPreview() {
    if (this.raw.extendedMedia?._ !== "messageExtendedMediaPreview") {
      throw new errors.MtArgumentError("No extended media preview available");
    }
    return new extendedMedia.ExtendedMediaPreview(this.raw.extendedMedia);
  }
  /**
   * Get the invoice's extended media.
   * Only available if {@link extendedMediaState} is `full`.
   * Otherwise, throws an error.
   */
  get extendedMedia() {
    if (!this._extendedMedia) {
      throw new errors.MtArgumentError("No extended media available");
    }
    return this._extendedMedia;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}.
   *
   * Invoice can't provide an input media, since some
   * of the data is not available to the user,
   * which is required to send it. This getter
   * is only provided to allow using `msg.media.inputMedia`
   */
  get inputMedia() {
    throw new errors.MtArgumentError("Invoice cannot provide an InputMedia");
  }
}
const Invoice$1 = /* @__PURE__ */ memoize.memoizeGetters(Invoice, ["extendedMediaPreview", "photo"]);
const Invoice$2 = /* @__PURE__ */ inspectable.makeInspectable(Invoice$1, void 0, ["inputMedia"]);
exports.Invoice = Invoice$2;
