// @mtcute/core/highlevel/types/media/invoice.js
import { MtArgumentError } from "../../../types/errors.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { WebDocument as WebDocument$1 } from "../files/web-document.js";
import { ExtendedMediaPreview as ExtendedMediaPreview$2 } from "./extended-media.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Invoice {
  constructor(raw, _extendedMedia) {
    this.raw = raw;
    this._extendedMedia = _extendedMedia;
  }
  type = "invoice";
  /**
   * Whether the shipping address was requested
   */
  isShippingAddressRequested() {
    return this.raw.shippingAddressRequested;
  }
  /**
   * Whether this is an example (test) invoice
   */
  isTest() {
    return this.raw.test;
  }
  /**
   * Product name, 1-32 characters
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Product description, 1-255 characters
   */
  get description() {
    return this.raw.description;
  }
  /**
   * URL of the product photo for the invoice
   */
  get photo() {
    if (!this.raw.photo) return null;
    return new WebDocument$1(this.raw.photo);
  }
  /**
   * Message ID of receipt
   */
  get receiptMessageId() {
    return this.raw.receiptMsgId ?? null;
  }
  /**
   * Three-letter ISO 4217 currency code
   */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Total price in the smallest units of the currency
   * (integer, not float/double). For example, for a price
   * of `US$ 1.45` `amount = 145`.
   *
   * See the exp parameter in [currencies.json](https://core.telegram.org/bots/payments/currencies.json),
   * it shows the number of digits past the decimal point
   * for each currency (2 for the majority of currencies).
   */
  get amount() {
    return this.raw.totalAmount;
  }
  /**
   * Unique bot deep-linking parameter that can be used to generate this invoice
   */
  get startParam() {
    return this.raw.startParam;
  }
  /**
   * If this invoice has extended media
   */
  get extendedMediaState() {
    if (!this.raw.extendedMedia) return "none";
    if (this.raw.extendedMedia._ === "messageExtendedMediaPreview") {
      return "preview";
    }
    return "full";
  }
  /**
   * Get the invoice's extended media preview.
   * Only available if {@link extendedMediaState} is `preview`.
   * Otherwise, throws an error.
   */
  get extendedMediaPreview() {
    if (this.raw.extendedMedia?._ !== "messageExtendedMediaPreview") {
      throw new MtArgumentError("No extended media preview available");
    }
    return new ExtendedMediaPreview$2(this.raw.extendedMedia);
  }
  /**
   * Get the invoice's extended media.
   * Only available if {@link extendedMediaState} is `full`.
   * Otherwise, throws an error.
   */
  get extendedMedia() {
    if (!this._extendedMedia) {
      throw new MtArgumentError("No extended media available");
    }
    return this._extendedMedia;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}.
   *
   * Invoice can't provide an input media, since some
   * of the data is not available to the user,
   * which is required to send it. This getter
   * is only provided to allow using `msg.media.inputMedia`
   */
  get inputMedia() {
    throw new MtArgumentError("Invoice cannot provide an InputMedia");
  }
}
const Invoice$1 = /* @__PURE__ */ memoizeGetters(Invoice, ["extendedMediaPreview", "photo"]);
const Invoice$2 = /* @__PURE__ */ makeInspectable(Invoice$1, void 0, ["inputMedia"]);
export {
  Invoice$2 as Invoice
};
