// @mtcute/core/highlevel/types/media/location.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const fileLocation = require("../files/file-location.cjs");
class RawLocation {
  constructor(geo) {
    this.geo = geo;
  }
  /**
   * Geo point latitude
   */
  get latitude() {
    return this.geo.lat;
  }
  /**
   * Geo point longitude
   */
  get longitude() {
    return this.geo.long;
  }
  /**
   * Accuracy radius in meters.
   */
  get radius() {
    return this.geo.accuracyRadius ?? 0;
  }
  /**
   * Create {@link FileLocation} containing
   * server-generated image with the map preview
   */
  preview(params = {}) {
    return new fileLocation.FileLocation({
      _: "inputWebFileGeoPointLocation",
      geoPoint: {
        _: "inputGeoPoint",
        lat: this.geo.lat,
        long: this.geo.long,
        accuracyRadius: this.geo.accuracyRadius
      },
      accessHash: this.geo.accessHash,
      w: params.width ?? 128,
      h: params.height ?? 128,
      zoom: params.zoom ?? 15,
      scale: params.scale ?? 1
    });
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaGeoPoint",
      geoPoint: {
        _: "inputGeoPoint",
        lat: this.geo.lat,
        long: this.geo.long,
        accuracyRadius: this.geo.accuracyRadius
      }
    };
  }
}
class Location extends RawLocation {
  type = "location";
}
class LiveLocation extends RawLocation {
  constructor(live) {
    super(live.geo);
    this.live = live;
  }
  type = "live_location";
  /**
   * A direction in which the location moves, in degrees; 1-360
   */
  get heading() {
    return this.live.heading ?? null;
  }
  /**
   * Validity period of provided geolocation
   */
  get period() {
    return this.live.period;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   *
   * Note that using this will result in an
   * independent live geolocation, which
   * will not be auto-updated with the current
   */
  get inputMedia() {
    return {
      _: "inputMediaGeoLive",
      geoPoint: {
        _: "inputGeoPoint",
        lat: this.geo.lat,
        long: this.geo.long,
        accuracyRadius: this.geo.accuracyRadius
      },
      heading: this.live.heading,
      period: this.live.period,
      proximityNotificationRadius: this.live.proximityNotificationRadius
    };
  }
  // todo: api to subscribe for real-time updates
}
const Location$1 = /* @__PURE__ */ inspectable.makeInspectable(Location, void 0, ["inputMedia"]);
const LiveLocation$1 = /* @__PURE__ */ inspectable.makeInspectable(LiveLocation, void 0, ["inputMedia"]);
exports.LiveLocation = LiveLocation$1;
exports.Location = Location$1;
exports.RawLocation = RawLocation;
