// @mtcute/core/highlevel/types/media/paid-media.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { ExtendedMediaPreview as ExtendedMediaPreview$2 } from "./extended-media.js";
class PaidMedia {
  constructor(raw, _extendedMedia) {
    this.raw = raw;
    this._extendedMedia = _extendedMedia;
  }
  type = "paid";
  /** Whether this media was paid for */
  get isPaid() {
    return this._extendedMedia !== void 0;
  }
  /** Price of the media (in Telegram Stars) */
  get price() {
    return this.raw.starsAmount;
  }
  /**
   * Get the available media previews.
   *
   * If the media is already paid for, this will return an empty array.
   */
  get previews() {
    const res = [];
    this.raw.extendedMedia.forEach((m) => {
      if (m._ !== "messageExtendedMediaPreview") return;
      res.push(new ExtendedMediaPreview$2(m));
    });
    return res;
  }
  /**
   * Get the available full media.
   *
   * If the media is not paid for, this will return an empty array.
   */
  get medias() {
    return this._extendedMedia;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}.
   *
   * Will throw if the media is not paid for.
   */
  get inputMedia() {
    if (!this.isPaid) {
      throw new Error("Cannot get input media for non-paid media");
    }
    return {
      _: "inputMediaPaidMedia",
      starsAmount: this.raw.starsAmount,
      extendedMedia: this._extendedMedia.map((m) => m.inputMedia)
    };
  }
}
const PaidMedia$1 = /* @__PURE__ */ makeInspectable(PaidMedia, void 0, ["inputMedia"]);
const PaidMedia$2 = /* @__PURE__ */ memoizeGetters(PaidMedia$1, ["previews", "inputMedia"]);
export {
  PaidMedia$2 as PaidMedia
};
