// @mtcute/core/highlevel/types/media/paid-media.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const extendedMedia = require("./extended-media.cjs");
class PaidMedia {
  constructor(raw, _extendedMedia) {
    this.raw = raw;
    this._extendedMedia = _extendedMedia;
  }
  type = "paid";
  /** Whether this media was paid for */
  get isPaid() {
    return this._extendedMedia !== void 0;
  }
  /** Price of the media (in Telegram Stars) */
  get price() {
    return this.raw.starsAmount;
  }
  /**
   * Get the available media previews.
   *
   * If the media is already paid for, this will return an empty array.
   */
  get previews() {
    const res = [];
    this.raw.extendedMedia.forEach((m) => {
      if (m._ !== "messageExtendedMediaPreview") return;
      res.push(new extendedMedia.ExtendedMediaPreview(m));
    });
    return res;
  }
  /**
   * Get the available full media.
   *
   * If the media is not paid for, this will return an empty array.
   */
  get medias() {
    return this._extendedMedia;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}.
   *
   * Will throw if the media is not paid for.
   */
  get inputMedia() {
    if (!this.isPaid) {
      throw new Error("Cannot get input media for non-paid media");
    }
    return {
      _: "inputMediaPaidMedia",
      starsAmount: this.raw.starsAmount,
      extendedMedia: this._extendedMedia.map((m) => m.inputMedia)
    };
  }
}
const PaidMedia$1 = /* @__PURE__ */ inspectable.makeInspectable(PaidMedia, void 0, ["inputMedia"]);
const PaidMedia$2 = /* @__PURE__ */ memoize.memoizeGetters(PaidMedia$1, ["previews", "inputMedia"]);
exports.PaidMedia = PaidMedia$2;
