// @mtcute/core/highlevel/types/media/photo.js
import { MtArgumentError } from "../../../types/errors.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parseDocument } from "./document-utils.js";
import { Thumbnail as Thumbnail$2 } from "./thumbnail.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { FileLocation as FileLocation$1 } from "../files/file-location.js";
class Photo extends FileLocation$1 {
  constructor(raw, media) {
    const location = {
      _: "inputPhotoFileLocation",
      id: raw.id,
      fileReference: raw.fileReference,
      accessHash: raw.accessHash,
      thumbSize: ""
    };
    let size;
    let width;
    let height;
    let bestSize;
    const progressive = raw.sizes.find((it) => it._ === "photoSizeProgressive");
    if (progressive) {
      location.thumbSize = progressive.type;
      size = Math.max(...progressive.sizes);
      width = progressive.w;
      height = progressive.h;
      bestSize = progressive;
    } else {
      let max = null;
      for (const sz of raw.sizes) {
        if (sz._ === "photoSize" && (!max || sz.size > max.size)) {
          max = sz;
        }
      }
      if (max) {
        location.thumbSize = max.type;
        size = max.size;
        width = max.w;
        height = max.h;
        bestSize = max;
      } else {
        throw new MtArgumentError("Photo does not have any sizes");
      }
    }
    super(location, size, raw.dcId);
    this.raw = raw;
    this.media = media;
    this._bestSize = bestSize;
    this.width = width;
    this.height = height;
    this.type = "photo";
  }
  type = "photo";
  /** Biggest available photo width */
  width;
  /** Biggest available photo height */
  height;
  _bestSize;
  /**
   * Photo ID
   */
  get id() {
    return this.raw.id;
  }
  /** Date this photo was sent */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this photo is an animated profile picture
   */
  get isAnimatedAvatar() {
    return Boolean(this.raw.videoSizes?.some((s) => s._ === "videoSize" && s.type === "u"));
  }
  /**
   * Whether this photo is an animated profile picture, built from an emoji/sticker markup
   */
  get isMarkupAvatar() {
    return Boolean(
      this.raw.videoSizes?.some((s) => s._ === "videoSizeEmojiMarkup" || s._ === "videoSizeStickerMarkup")
    );
  }
  /** Whether this photo is hidden with a spoiler */
  get hasSpoiler() {
    return this.media?.spoiler ?? false;
  }
  /** For self-destructing photos, TTL in seconds */
  get ttlSeconds() {
    return this.media?.ttlSeconds ?? null;
  }
  /**
   * Available thumbnails.
   *
   * **Note**: This list will also contain the largest thumbnail that is
   * represented by the current object.
   */
  get thumbnails() {
    const res = this.raw.sizes.map((sz) => new Thumbnail$2(this.raw, sz));
    this.raw.videoSizes?.forEach((sz) => res.push(new Thumbnail$2(this.raw, sz)));
    return res;
  }
  /**
   * Get a photo thumbnail by its type.
   *
   * Thumbnail types are described in the
   * [Telegram docs](https://core.telegram.org/api/files#image-thumbnail-types),
   * and are also available as static members of {@link Thumbnail} for convenience.
   *
   * @param type  Thumbnail type
   */
  getThumbnail(type) {
    return this.thumbnails.find((it) => it.type === type) ?? null;
  }
  /**
   * Get TDLib and Bot API compatible File ID
   * representing this photo's best thumbnail.
   */
  get fileId() {
    if (!this._bestSize) {
      throw new MtArgumentError("Cannot get File ID for this photo");
    }
    return this.getThumbnail(this._bestSize.type).fileId;
  }
  /**
   * Get TDLib and Bot API compatible Unique File ID
   * representing this photo's best thumbnail.
   */
  get uniqueFileId() {
    if (!this._bestSize) {
      throw new MtArgumentError("Cannot get File ID for this photo");
    }
    return this.getThumbnail(this._bestSize.type).uniqueFileId;
  }
  get livePhotoVideo() {
    if (!this.media?.video) return null;
    if (this.media.video._ !== "document") return null;
    const parsed = parseDocument(this.media.video);
    if (parsed.type !== "video") return null;
    return parsed;
  }
  /**
   * Input photo generated from this object.
   */
  get inputPhoto() {
    return {
      _: "inputPhoto",
      id: this.raw.id,
      accessHash: this.raw.accessHash,
      fileReference: this.raw.fileReference
    };
  }
  /**
   * Input media generated from this object,
   * to be used in {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   */
  get inputMedia() {
    return {
      _: "inputMediaPhoto",
      id: this.inputPhoto
    };
  }
}
const Photo$1 = /* @__PURE__ */ memoizeGetters(Photo, ["thumbnails", "fileId", "uniqueFileId"]);
const Photo$2 = /* @__PURE__ */ makeInspectable(Photo$1, ["fileSize", "dcId", "width", "height"], ["inputMedia", "inputPhoto"]);
export {
  Photo$2 as Photo
};
