// @mtcute/core/highlevel/types/media/poll.js
import Long from "long";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { MessageEntity as MessageEntity$2 } from "../messages/message-entity.js";
import { _messageMediaFromTl } from "../messages/message-media.js";
import { makeInspectable } from "../../utils/inspectable.js";
class PollAnswer {
  constructor(raw, result) {
    this.raw = raw;
    this.result = result;
  }
  /**
   * Answer text
   */
  get text() {
    return this.raw.text.text;
  }
  /**
   * Format entities for {@link text}, currently may only contain custom emojis
   */
  get textEntities() {
    return this.raw.text.entities.map((ent) => new MessageEntity$2(ent, this.raw.text.text));
  }
  /**
   * Answer data, to be passed to
   * {@link TelegramClient.sendVote}
   */
  get data() {
    return this.raw.option;
  }
  /**
   * Number of people who has chosen this result.
   * If not available (i.e. not voted yet)
   *
   * @default
   */
  get voters() {
    return this.result?.voters ?? 0;
  }
  /**
   * Whether this answer was chosen by the current user
   */
  get chosen() {
    return Boolean(this.result?.chosen);
  }
  /**
   * Whether this answer is correct (for quizzes).
   * Not available before choosing an answer
   *
   * @default
   */
  get correct() {
    return Boolean(this.result?.correct);
  }
  /** Media attached to this answer, if any */
  get media() {
    return this.raw.media ? _messageMediaFromTl(null, this.raw.media) : null;
  }
}
const PollAnswer$1 = /* @__PURE__ */ memoizeGetters(PollAnswer, ["textEntities", "media"]);
const PollAnswer$2 = /* @__PURE__ */ makeInspectable(PollAnswer$1);
class Poll {
  constructor(raw, _peers, results, attachedMedia) {
    this.raw = raw;
    this._peers = _peers;
    this.results = results;
    this.attachedMedia = attachedMedia;
  }
  type = "poll";
  /**
   * Unique identifier of the poll
   */
  get id() {
    return this.raw.id;
  }
  /**
   * Poll question
   */
  get question() {
    return this.raw.question.text;
  }
  /**
   * Format entities for {@link question} (currently may only contain custom emojis)
   */
  get questionEntities() {
    return this.raw.question.entities.map((ent) => new MessageEntity$2(ent, this.raw.question.text));
  }
  /**
   * List of answers in this poll
   */
  get answers() {
    const results = this.results?.results;
    return this.raw.answers.map((ans, idx) => {
      assertTypeIs("Poll.answers", ans, "pollAnswer");
      if (results) {
        return new PollAnswer(ans, results[idx]);
      }
      return new PollAnswer(ans);
    });
  }
  /**
   * Total number of voters in this poll, if available
   */
  get voters() {
    return this.results?.totalVoters ?? 0;
  }
  /**
   * Whether this poll is closed, i.e. does not
   * accept votes anymore
   */
  get isClosed() {
    return this.raw.closed;
  }
  /**
   * Whether this poll is public, i.e. you
   * list of voters is publicly available
   */
  get isPublic() {
    return this.raw.publicVoters;
  }
  /**
   * Whether this is a quiz
   */
  get isQuiz() {
    return this.raw.quiz;
  }
  /**
   * Whether this poll accepts multiple answers
   */
  get isMultiple() {
    return this.raw.multipleChoice;
  }
  /** Whether the current user is the creator of this poll */
  get isCreator() {
    return this.raw.creator;
  }
  /** Whether new options can be suggested to this poll */
  get canAddAnswers() {
    return !this.raw.closed && !this.raw.openAnswers;
  }
  /** Whether retracting the vote is disabled in this poll */
  get isRevotingDisabled() {
    return this.raw.revotingDisabled;
  }
  /** Whether answers to this poll should be shuffled before showing to the user */
  get shuffleAnswers() {
    return this.raw.shuffleAnswers;
  }
  /** Whether the results of this poll are hidden until the end of the poll */
  get hideResultsUntilClose() {
    return this.raw.hideResultsUntilClose;
  }
  /** Whether the poll has unread votes */
  get hasUnreaVotes() {
    return this.results?.hasUnreadVotes ?? false;
  }
  /** Whether voting in this poll is restricted to channel subscribers */
  get isSubscribersOnly() {
    return this.raw.subscribersOnly;
  }
  /** ISO2 country codes the poll is restricted to */
  get countries() {
    return this.raw.countriesIso2 ?? null;
  }
  /** Whether the current user can view extended statistics for this poll */
  get canViewStats() {
    return this.results?.canViewStats ?? false;
  }
  /**
   * Solution for the quiz, only available
   * in case you have already answered
   */
  get solution() {
    return this.results?.solution ?? null;
  }
  /**
   * Format entities for {@link solution}, only available
   * in case you have already answered
   */
  get solutionEntities() {
    if (!this.results) return null;
    const res = [];
    if (this.results.solutionEntities?.length) {
      for (const ent of this.results.solutionEntities) {
        res.push(new MessageEntity$2(ent, this.results.solution));
      }
    }
    return res;
  }
  /** Media attached to the solution, if any */
  get solutionMedia() {
    return this.results?.solutionMedia ? _messageMediaFromTl(null, this.results.solutionMedia) : null;
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   *
   * A few notes:
   *  - Using this will result in an
   *    independent poll, which will not
   *    be auto-updated with the current.
   *  - If this is a quiz, a normal poll
   *    will be returned since the client does not
   *    know the correct answer.
   *  - This always returns a non-closed poll,
   *    even if the current poll was closed
   */
  get inputMedia() {
    return {
      _: "inputMediaPoll",
      attachedMedia: this.attachedMedia?.inputMedia,
      solutionMedia: this.solutionMedia?.inputMedia,
      poll: {
        _: "poll",
        closed: false,
        id: Long.ZERO,
        publicVoters: this.raw.publicVoters,
        multipleChoice: this.raw.multipleChoice,
        question: this.raw.question,
        answers: this.raw.answers,
        closePeriod: this.raw.closePeriod,
        closeDate: this.raw.closeDate,
        openAnswers: this.raw.openAnswers,
        revotingDisabled: this.raw.revotingDisabled,
        shuffleAnswers: this.raw.shuffleAnswers,
        hideResultsUntilClose: this.raw.hideResultsUntilClose,
        subscribersOnly: this.raw.subscribersOnly,
        countriesIso2: this.raw.countriesIso2,
        hash: Long.ZERO
      }
    };
  }
}
const Poll$1 = /* @__PURE__ */ memoizeGetters(Poll, ["answers", "solutionEntities", "questionEntities", "solutionMedia"]);
const Poll$2 = /* @__PURE__ */ makeInspectable(Poll$1, ["attachedMedia"], ["inputMedia"]);
export {
  Poll$2 as Poll,
  PollAnswer$2 as PollAnswer
};
