// @mtcute/core/highlevel/types/media/sticker.js
import { tdFileId } from "@mtcute/file-id";
import { MtArgumentError } from "../../../types/errors.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { RawDocument } from "./document.js";
import { makeInspectable } from "../../utils/inspectable.js";
const MASK_POSITION_POINT_TO_TL = {
  forehead: 0,
  eyes: 1,
  mouth: 2,
  chin: 3
};
const MASK_POS = ["forehead", "eyes", "mouth", "chin"];
class Sticker extends RawDocument {
  constructor(doc, attr, attr2) {
    super(doc);
    this.attr = attr;
    this.attr2 = attr2;
  }
  type = "sticker";
  _fileIdType() {
    return tdFileId.FileType.Sticker;
  }
  /**
   * Sticker width in pixels
   */
  get width() {
    return this.attr2?.w ?? 512;
  }
  /**
   * Sticker height in pixels
   */
  get height() {
    return this.attr2?.h ?? 512;
  }
  /**
   * Whether this sticker is a premium sticker
   * (has premium fullscreen animation)
   */
  get isPremiumSticker() {
    return Boolean(this.raw.videoThumbs?.some((s) => s._ === "videoSize" && s.type === "f"));
  }
  /**
   * Whether this sticker is a valid sticker.
   *
   * If it is not, then this is probably a WEBP photo
   * that Telegram treats as a sticker.
   */
  get isValidSticker() {
    return this.attr2 !== void 0 && (this.attr2.w === 512 || this.attr2.h === 512);
  }
  /**
   * Primary emoji associated with this sticker,
   * that is displayed in dialogs list.
   *
   * If there is none, empty string is returned.
   *
   * **Note:** This only contains at most one emoji.
   * Some stickers have multiple associated emojis,
   * but only one is returned here. This is Telegram's
   * limitation! Use {@link getAllEmojis} instead.
   *
   * For custom emojis, this alt should be used as a fallback
   * text that will be "behind" the custom emoji entity.
   */
  get emoji() {
    return this.attr.alt;
  }
  /**
   * Whether this custom emoji can be used by non-premium users.
   * `false` if this is not a custom emoji.
   *
   * > Not sure if there are any such stickers currently.
   */
  get customEmojiFree() {
    return this.attr._ === "documentAttributeCustomEmoji" ? this.attr.free ?? false : false;
  }
  /**
   * If this is a custom emoji, its unique ID
   * that can be used in {@link TelegramClient#getCustomEmojis}
   */
  get customEmojiId() {
    if (this.attr._ !== "documentAttributeCustomEmoji") {
      throw new MtArgumentError("This is not a custom emoji");
    }
    return this.raw.id;
  }
  /**
   * Type of the sticker
   */
  get stickerType() {
    if (this.attr._ === "documentAttributeSticker") {
      return this.attr.mask ? "mask" : "sticker";
    } else if (this.attr._ === "documentAttributeCustomEmoji") {
      return "emoji";
    }
    return "sticker";
  }
  /**
   * Type of the file representing the sticker
   */
  get sourceType() {
    if (this.attr2?._ === "documentAttributeVideo") {
      return "video";
    }
    return this.mimeType === "application/x-tgsticker" ? "animated" : "static";
  }
  /**
   * Whether this sticker has an associated public sticker set.
   */
  get hasStickerSet() {
    return this.attr.stickerset._ === "inputStickerSetID";
  }
  /**
   * Input sticker set that it associated with this sticker, if available.
   */
  get inputStickerSet() {
    return this.attr.stickerset._ === "inputStickerSetEmpty" ? null : this.attr.stickerset;
  }
  /**
   * Position where this mask should be placed
   */
  get maskPosition() {
    if (this.attr._ !== "documentAttributeSticker" || !this.attr.maskCoords) {
      return null;
    }
    const raw = this.attr.maskCoords;
    return {
      point: MASK_POS[raw.n],
      x: raw.x,
      y: raw.y,
      scale: raw.zoom
    };
  }
}
const Sticker$1 = /* @__PURE__ */ memoizeGetters(Sticker, ["fileName", "thumbnails", "fileId", "uniqueFileId", "maskPosition"]);
const Sticker$2 = /* @__PURE__ */ makeInspectable(Sticker$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument"]);
export {
  MASK_POSITION_POINT_TO_TL,
  Sticker$2 as Sticker
};
