// @mtcute/core/highlevel/types/media/story.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const peer = require("../peers/peer.cjs");
const story = require("../stories/story.cjs");
class MediaStory {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  type = "story";
  /**
   * Whether this story was automatically forwarded to you
   * because you were mentioned in it
   */
  get isMention() {
    return this.raw.viaMention;
  }
  /**
   * Peer who has posted this story
   */
  get peer() {
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the story
   */
  get storyId() {
    return this.raw.id;
  }
  /**
   * Contents of the story. May not be available, in which case the story
   * should be fetched using {@link getStoriesById}
   */
  get story() {
    if (this.raw.story?._ !== "storyItem") return null;
    return new story.Story(this.raw.story, this._peers);
  }
  get inputMedia() {
    return {
      _: "inputMediaStory",
      peer: this.peer.inputPeer,
      id: this.raw.id
    };
  }
}
const MediaStory$1 = /* @__PURE__ */ inspectable.makeInspectable(MediaStory, void 0, ["inputMedia"]);
exports.MediaStory = MediaStory$1;
