// @mtcute/core/highlevel/types/media/story.js
import { makeInspectable } from "../../utils/inspectable.js";
import { parsePeer } from "../peers/peer.js";
import { Story as Story$2 } from "../stories/story.js";
class MediaStory {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  type = "story";
  /**
   * Whether this story was automatically forwarded to you
   * because you were mentioned in it
   */
  get isMention() {
    return this.raw.viaMention;
  }
  /**
   * Peer who has posted this story
   */
  get peer() {
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the story
   */
  get storyId() {
    return this.raw.id;
  }
  /**
   * Contents of the story. May not be available, in which case the story
   * should be fetched using {@link getStoriesById}
   */
  get story() {
    if (this.raw.story?._ !== "storyItem") return null;
    return new Story$2(this.raw.story, this._peers);
  }
  get inputMedia() {
    return {
      _: "inputMediaStory",
      peer: this.peer.inputPeer,
      id: this.raw.id
    };
  }
}
const MediaStory$1 = /* @__PURE__ */ makeInspectable(MediaStory, void 0, ["inputMedia"]);
export {
  MediaStory$1 as MediaStory
};
