// @mtcute/core/highlevel/types/media/thumbnail.js
import { toFileId, tdFileId, toUniqueFileId } from "@mtcute/file-id";
import Long from "long";
import { MtTypeAssertionError, MtArgumentError } from "../../../types/errors.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { strippedPhotoToJpg, inflateSvgPath, svgPathToFile } from "../../utils/file-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { FileLocation as FileLocation$1 } from "../files/file-location.js";
class Thumbnail extends FileLocation$1 {
  // see: https://core.telegram.org/api/files#image-thumbnail-types
  static THUMB_100x100_BOX = "s";
  static THUMB_320x320_BOX = "m";
  static THUMB_800x800_BOX = "x";
  static THUMB_1280x1280_BOX = "y";
  static THUMB_2560x2560_BOX = "w";
  static THUMB_160x160_CROP = "a";
  static THUMB_320x320_CROP = "b";
  static THUMB_640x640_CROP = "c";
  static THUMB_1280x1280_CROP = "d";
  static THUMB_STRIP = "i";
  static THUMB_OUTLINE = "j";
  /** Animated profile pictures preview */
  static THUMB_VIDEO_PROFILE = "u";
  /** Trimmed and downscaled video previews */
  static THUMB_VIDEO_PREVIEW = "v";
  /** Fullscreen animation for Premium stickers */
  static THUMB_VIDEO_FULLSCREEN = "f";
  /** (non-standard) Emoji-based markup for profile photo */
  static THUMB_EMOJI_MARKUP = "pfp_em";
  /** (non-standard) Sticker-based markup for profile photo */
  static THUMB_STICKER_MARKUP = "pfp_st";
  raw;
  /**
   * Thumbnail width
   * (`NaN` for {@link THUMB_OUTLINE} and {@link THUMB_STRIP})
   */
  width;
  /**
   * Thumbnail height
   * (`NaN` for {@link THUMB_OUTLINE} and {@link THUMB_STRIP})
   */
  height;
  _path;
  _media;
  constructor(media, sz) {
    switch (sz._) {
      case "photoSizeEmpty":
      case "photoCachedSize":
        throw new MtTypeAssertionError("sz", "not (photoSizeEmpty | photoCachedSize)", sz._);
    }
    let location;
    let size;
    let width;
    let height;
    switch (sz._) {
      case "photoStrippedSize":
        location = strippedPhotoToJpg(sz.bytes);
        width = height = Number.NaN;
        size = location.length;
        break;
      case "photoPathSize": {
        const imageSize = media._ === "document" ? media.attributes.find((it) => it._ === "documentAttributeImageSize") : void 0;
        location = () => svgPathToFile(this._path, imageSize);
        if (imageSize) {
          width = imageSize.w;
          height = imageSize.h;
        } else {
          width = height = 512;
        }
        size = Infinity;
        break;
      }
      case "videoSizeEmojiMarkup":
      case "videoSizeStickerMarkup":
        location = () => {
          throw new MtArgumentError("Cannot download thumbnail with emoji/sticker markup, try other size");
        };
        width = height = Number.NaN;
        size = Infinity;
        break;
      default:
        if (media._ === "stickerSet") {
          location = {
            _: "inputStickerSetThumb",
            stickerset: {
              _: "inputStickerSetID",
              id: media.id,
              accessHash: media.accessHash
            },
            thumbVersion: media.thumbVersion
          };
        } else if (media._ === "messageExtendedMediaPreview") {
          throw new MtTypeAssertionError("messageExtendedMediaPreview#thumb", "photoStrippedSize", sz._);
        } else {
          location = {
            _: media._ === "photo" ? "inputPhotoFileLocation" : "inputDocumentFileLocation",
            id: media.id,
            fileReference: media.fileReference,
            accessHash: media.accessHash,
            thumbSize: sz.type
          };
        }
        width = sz.w;
        height = sz.h;
        size = sz._ === "photoSizeProgressive" ? Math.max(...sz.sizes) : sz.size;
        break;
    }
    let dcId;
    if (media._ === "stickerSet") {
      dcId = media.thumbDcId;
    } else if (media._ === "messageExtendedMediaPreview") {
      dcId = 0;
    } else {
      dcId = media.dcId;
    }
    super(location, size, dcId);
    this.raw = sz;
    this.width = width;
    this.height = height;
    this._media = media;
    if (sz._ === "photoPathSize") {
      this._path = inflateSvgPath(sz.bytes);
    }
  }
  get isVideo() {
    return this.raw._ === "videoSize";
  }
  /**
   * Thumbnail type
   */
  get type() {
    if (this.raw._ === "videoSizeEmojiMarkup") {
      return Thumbnail.THUMB_EMOJI_MARKUP;
    }
    if (this.raw._ === "videoSizeStickerMarkup") {
      return Thumbnail.THUMB_STICKER_MARKUP;
    }
    return this.raw.type;
  }
  /**
   * If {@link raw} is `tl.RawPhotoPathSize` (i.e. `raw.type === Thumbnail.THUMB_OUTLINE`),
   * this property will return raw SVG path of the preview.
   *
   * When downloading path thumbnails, a valid SVG file is returned.
   *
   * See also: https://core.telegram.org/api/files#vector-thumbnails
   *
   * @throws MtTypeAssertionError  In case {@link raw} is not `tl.RawPhotoPathSize`
   */
  get path() {
    assertTypeIs("Thumbnail#path", this.raw, "photoPathSize");
    return this._path;
  }
  /**
   * Get TDLib and Bot API compatible File ID
   * representing this thumbnail.
   *
   * > **Note:** You can't use this file id to send a thumbnail,
   * > only to download it.
   */
  get fileId() {
    if (this.raw._ !== "photoSize" && this.raw._ !== "photoSizeProgressive" && this.raw._ !== "videoSize" || this._media._ === "messageExtendedMediaPreview") {
      throw new MtArgumentError(`Cannot generate a file ID for "${this.type}"`);
    }
    if (this._media._ === "stickerSet") {
      return toFileId({
        type: tdFileId.FileType.Thumbnail,
        dcId: this.dcId,
        fileReference: null,
        location: {
          _: "photo",
          id: Long.ZERO,
          accessHash: Long.ZERO,
          source: {
            _: "stickerSetThumbnailVersion",
            id: this._media.id,
            accessHash: this._media.accessHash,
            version: this._media.thumbVersion
          }
        }
      });
    }
    return toFileId({
      type: this._media._ === "photo" ? tdFileId.FileType.Photo : tdFileId.FileType.Thumbnail,
      dcId: this.dcId,
      fileReference: this._media.fileReference,
      location: {
        _: "photo",
        id: this._media.id,
        accessHash: this._media.accessHash,
        source: {
          _: "thumbnail",
          fileType: this._media._ === "photo" ? tdFileId.FileType.Photo : tdFileId.FileType.Thumbnail,
          thumbnailType: this.raw.type === "u" ? "\0" : this.raw.type
        }
      }
    });
  }
  /**
   * Get a unique File ID representing this thumbnail.
   */
  get uniqueFileId() {
    if (this.raw._ !== "photoSize" && this.raw._ !== "photoSizeProgressive" && this.raw._ !== "videoSize" || this._media._ === "messageExtendedMediaPreview") {
      throw new MtArgumentError(`Cannot generate a unique file ID for "${this.type}"`);
    }
    if (this._media._ === "stickerSet") {
      return toUniqueFileId(tdFileId.FileType.Thumbnail, {
        _: "photo",
        id: Long.ZERO,
        source: {
          _: "stickerSetThumbnailVersion",
          id: this._media.id,
          accessHash: this._media.accessHash,
          version: this._media.thumbVersion
        }
      });
    }
    return toUniqueFileId(this._media._ === "photo" ? tdFileId.FileType.Photo : tdFileId.FileType.Thumbnail, {
      _: "photo",
      id: this._media.id,
      source: {
        _: "thumbnail",
        fileType: this._media._ === "photo" ? tdFileId.FileType.Photo : tdFileId.FileType.Thumbnail,
        thumbnailType: this.raw.type === "u" ? "\0" : this.raw.type
      }
    });
  }
}
const Thumbnail$1 = /* @__PURE__ */ memoizeGetters(Thumbnail, ["fileId", "uniqueFileId"]);
const Thumbnail$2 = /* @__PURE__ */ makeInspectable(Thumbnail$1, ["fileSize", "dcId", "width", "height"], ["path"]);
export {
  Thumbnail$2 as Thumbnail
};
