// @mtcute/core/highlevel/types/media/todo.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
class TodoItem {
  constructor(raw, _peers, completion) {
    this.raw = raw;
    this._peers = _peers;
    this.completion = completion;
  }
  get id() {
    return this.raw.id;
  }
  get text() {
    return this.raw.title;
  }
  get isCompleted() {
    return this.completion != null;
  }
  get completedBy() {
    if (!this.completion) return null;
    const peer = parsePeer(this.completion.completedBy, this._peers);
    if (!peer) return null;
    return peer;
  }
  get completedDate() {
    if (!this.completion) return null;
    return new Date(this.completion.date * 1e3);
  }
}
const TodoItem$1 = /* @__PURE__ */ makeInspectable(TodoItem);
const TodoItem$2 = /* @__PURE__ */ memoizeGetters(TodoItem$1, ["completedBy", "completedDate"]);
class TodoList {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  type = "todo";
  get othersCanAppend() {
    return this.raw.todo.othersCanAppend;
  }
  get othersCanComplete() {
    return this.raw.todo.othersCanComplete;
  }
  get title() {
    return this.raw.todo.title;
  }
  get items() {
    const completionsMap = /* @__PURE__ */ new Map();
    this.raw.completions?.forEach((c) => {
      if (c._ === "todoCompletion") {
        completionsMap.set(c.id, c);
      }
    });
    return this.raw.todo.list.map((i) => new TodoItem(i, this._peers, completionsMap.get(i.id)));
  }
  get inputMedia() {
    return {
      _: "inputMediaTodo",
      todo: this.raw.todo
    };
  }
}
const TodoList$1 = /* @__PURE__ */ makeInspectable(TodoList);
const TodoList$2 = /* @__PURE__ */ memoizeGetters(TodoList$1, ["items"]);
export {
  TodoItem$2 as TodoItem,
  TodoList$2 as TodoList
};
