// @mtcute/core/highlevel/types/media/todo.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
class TodoItem {
  constructor(raw, _peers, completion) {
    this.raw = raw;
    this._peers = _peers;
    this.completion = completion;
  }
  get id() {
    return this.raw.id;
  }
  get text() {
    return this.raw.title;
  }
  get isCompleted() {
    return this.completion != null;
  }
  get completedBy() {
    if (!this.completion) return null;
    const peer$1 = peer.parsePeer(this.completion.completedBy, this._peers);
    if (!peer$1) return null;
    return peer$1;
  }
  get completedDate() {
    if (!this.completion) return null;
    return new Date(this.completion.date * 1e3);
  }
}
const TodoItem$1 = /* @__PURE__ */ inspectable.makeInspectable(TodoItem);
const TodoItem$2 = /* @__PURE__ */ memoize.memoizeGetters(TodoItem$1, ["completedBy", "completedDate"]);
class TodoList {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  type = "todo";
  get othersCanAppend() {
    return this.raw.todo.othersCanAppend;
  }
  get othersCanComplete() {
    return this.raw.todo.othersCanComplete;
  }
  get title() {
    return this.raw.todo.title;
  }
  get items() {
    const completionsMap = /* @__PURE__ */ new Map();
    this.raw.completions?.forEach((c) => {
      if (c._ === "todoCompletion") {
        completionsMap.set(c.id, c);
      }
    });
    return this.raw.todo.list.map((i) => new TodoItem(i, this._peers, completionsMap.get(i.id)));
  }
  get inputMedia() {
    return {
      _: "inputMediaTodo",
      todo: this.raw.todo
    };
  }
}
const TodoList$1 = /* @__PURE__ */ inspectable.makeInspectable(TodoList);
const TodoList$2 = /* @__PURE__ */ memoize.memoizeGetters(TodoList$1, ["items"]);
exports.TodoItem = TodoItem$2;
exports.TodoList = TodoList$2;
