// @mtcute/core/highlevel/types/media/venue.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const memoize = require("../../utils/memoize.cjs");
const location = require("./location.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Venue {
  constructor(raw) {
    this.raw = raw;
  }
  type = "venue";
  /**
   * Geolocation of the venue
   */
  get location() {
    typeAssertions.assertTypeIs("Venue#location", this.raw.geo, "geoPoint");
    return new location.Location(this.raw.geo);
  }
  /**
   * Venue name
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Venue address
   */
  get address() {
    return this.raw.address;
  }
  /**
   * When available, source from where this venue was acquired
   */
  get source() {
    if (!this.raw.provider) return null;
    return {
      provider: this.raw.provider,
      id: this.raw.venueId,
      type: this.raw.venueType
    };
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   *
   * A few notes:
   *  - Using this will result in an
   *    independent poll, which will not
   *    be auto-updated with the current.
   *  - If this is a quiz, a normal poll
   *    will be returned since the client does not
   *    know the correct answer.
   *  - This always returns a non-closed poll,
   *    even if the current poll was closed
   */
  get inputMedia() {
    return {
      _: "inputMediaVenue",
      geoPoint: this.raw.geo._ === "geoPoint" ? {
        _: "inputGeoPoint",
        lat: this.raw.geo.lat,
        long: this.raw.geo.long,
        accuracyRadius: this.raw.geo.accuracyRadius
      } : { _: "inputGeoPointEmpty" },
      title: this.raw.title,
      address: this.raw.address,
      provider: this.raw.provider,
      venueId: this.raw.venueId,
      venueType: this.raw.venueType
    };
  }
}
const Venue$1 = /* @__PURE__ */ memoize.memoizeGetters(Venue, ["location", "source"]);
const Venue$2 = /* @__PURE__ */ inspectable.makeInspectable(Venue$1, void 0, ["inputMedia"]);
exports.Venue = Venue$2;
