// @mtcute/core/highlevel/types/media/venue.js
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Location as Location$1 } from "./location.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Venue {
  constructor(raw) {
    this.raw = raw;
  }
  type = "venue";
  /**
   * Geolocation of the venue
   */
  get location() {
    assertTypeIs("Venue#location", this.raw.geo, "geoPoint");
    return new Location$1(this.raw.geo);
  }
  /**
   * Venue name
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Venue address
   */
  get address() {
    return this.raw.address;
  }
  /**
   * When available, source from where this venue was acquired
   */
  get source() {
    if (!this.raw.provider) return null;
    return {
      provider: this.raw.provider,
      id: this.raw.venueId,
      type: this.raw.venueType
    };
  }
  /**
   * Input media TL object generated from this object,
   * to be used inside {@link InputMediaLike} and
   * {@link TelegramClient.sendMedia}
   *
   * A few notes:
   *  - Using this will result in an
   *    independent poll, which will not
   *    be auto-updated with the current.
   *  - If this is a quiz, a normal poll
   *    will be returned since the client does not
   *    know the correct answer.
   *  - This always returns a non-closed poll,
   *    even if the current poll was closed
   */
  get inputMedia() {
    return {
      _: "inputMediaVenue",
      geoPoint: this.raw.geo._ === "geoPoint" ? {
        _: "inputGeoPoint",
        lat: this.raw.geo.lat,
        long: this.raw.geo.long,
        accuracyRadius: this.raw.geo.accuracyRadius
      } : { _: "inputGeoPointEmpty" },
      title: this.raw.title,
      address: this.raw.address,
      provider: this.raw.provider,
      venueId: this.raw.venueId,
      venueType: this.raw.venueType
    };
  }
}
const Venue$1 = /* @__PURE__ */ memoizeGetters(Venue, ["location", "source"]);
const Venue$2 = /* @__PURE__ */ makeInspectable(Venue$1, void 0, ["inputMedia"]);
export {
  Venue$2 as Venue
};
