// @mtcute/core/highlevel/types/media/video.js
import { tdFileId } from "@mtcute/file-id";
import { memoizeGetters } from "../../utils/memoize.js";
import { RawDocument } from "./document.js";
import { Photo as Photo$2 } from "./photo.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Video extends RawDocument {
  constructor(doc, attr, media) {
    super(doc);
    this.attr = attr;
    this.media = media;
  }
  type = "video";
  _fileIdType() {
    if (this.isRound) return tdFileId.FileType.VideoNote;
    if (this.isAnimation) return tdFileId.FileType.Animation;
    return tdFileId.FileType.Video;
  }
  /**
   * Video width in pixels
   */
  get width() {
    return this.attr.w;
  }
  /**
   * Video height in pixels
   */
  get height() {
    return this.attr.h;
  }
  /**
   * Video duration in seconds.
   *
   * `0` for legacy GIFs
   */
  get duration() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.duration : 0;
  }
  /**
   * Whether this video is an animated GIF
   * (represented either by actual GIF or a silent MP4 video)
   */
  get isAnimation() {
    return this.attr._ === "documentAttributeImageSize" || this.raw.attributes.some((it) => it._ === "documentAttributeAnimated");
  }
  /**
   * Whether this video is a round video message (aka video note)
   */
  get isRound() {
    return this.attr._ === "documentAttributeVideo" && Boolean(this.attr.roundMessage);
  }
  /**
   * Whether this video is a legacy GIF (i.e. its MIME is `image/gif`)
   */
  get isLegacyGif() {
    return this.attr._ === "documentAttributeImageSize";
  }
  /** Whether this video is hidden with a spoiler */
  get hasSpoiler() {
    return this.media?.spoiler ?? false;
  }
  /** For self-destructing videos, TTL in seconds */
  get ttlSeconds() {
    return this.media?.ttlSeconds ?? null;
  }
  get videoStartTs() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.videoStartTs ?? null : null;
  }
  get codec() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.videoCodec ?? null : null;
  }
  get videoCover() {
    if (!this.media || this.media.videoCover?._ !== "photo") return null;
    return new Photo$2(this.media.videoCover);
  }
  get videoTimestamp() {
    return this.media?.videoTimestamp ?? null;
  }
}
const Video$1 = /* @__PURE__ */ memoizeGetters(Video, ["fileName", "thumbnails", "fileId", "uniqueFileId", "isAnimation"]);
const Video$2 = /* @__PURE__ */ makeInspectable(Video$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument"]);
export {
  Video$2 as Video
};
