// @mtcute/core/highlevel/types/media/video.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const memoize = require("../../utils/memoize.cjs");
const document = require("./document.cjs");
const photo = require("./photo.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Video extends document.RawDocument {
  constructor(doc, attr, media) {
    super(doc);
    this.attr = attr;
    this.media = media;
  }
  type = "video";
  _fileIdType() {
    if (this.isRound) return fileId.tdFileId.FileType.VideoNote;
    if (this.isAnimation) return fileId.tdFileId.FileType.Animation;
    return fileId.tdFileId.FileType.Video;
  }
  /**
   * Video width in pixels
   */
  get width() {
    return this.attr.w;
  }
  /**
   * Video height in pixels
   */
  get height() {
    return this.attr.h;
  }
  /**
   * Video duration in seconds.
   *
   * `0` for legacy GIFs
   */
  get duration() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.duration : 0;
  }
  /**
   * Whether this video is an animated GIF
   * (represented either by actual GIF or a silent MP4 video)
   */
  get isAnimation() {
    return this.attr._ === "documentAttributeImageSize" || this.raw.attributes.some((it) => it._ === "documentAttributeAnimated");
  }
  /**
   * Whether this video is a round video message (aka video note)
   */
  get isRound() {
    return this.attr._ === "documentAttributeVideo" && Boolean(this.attr.roundMessage);
  }
  /**
   * Whether this video is a legacy GIF (i.e. its MIME is `image/gif`)
   */
  get isLegacyGif() {
    return this.attr._ === "documentAttributeImageSize";
  }
  /** Whether this video is hidden with a spoiler */
  get hasSpoiler() {
    return this.media?.spoiler ?? false;
  }
  /** For self-destructing videos, TTL in seconds */
  get ttlSeconds() {
    return this.media?.ttlSeconds ?? null;
  }
  get videoStartTs() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.videoStartTs ?? null : null;
  }
  get codec() {
    return this.attr._ === "documentAttributeVideo" ? this.attr.videoCodec ?? null : null;
  }
  get videoCover() {
    if (!this.media || this.media.videoCover?._ !== "photo") return null;
    return new photo.Photo(this.media.videoCover);
  }
  get videoTimestamp() {
    return this.media?.videoTimestamp ?? null;
  }
}
const Video$1 = /* @__PURE__ */ memoize.memoizeGetters(Video, ["fileName", "thumbnails", "fileId", "uniqueFileId", "isAnimation"]);
const Video$2 = /* @__PURE__ */ inspectable.makeInspectable(Video$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument"]);
exports.Video = Video$2;
