// @mtcute/core/highlevel/types/media/voice.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const memoize = require("../../utils/memoize.cjs");
const voiceUtils = require("../../utils/voice-utils.cjs");
const document = require("./document.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Voice extends document.RawDocument {
  constructor(doc, attr, media) {
    super(doc);
    this.attr = attr;
    this.media = media;
  }
  type = "voice";
  _fileIdType() {
    return fileId.tdFileId.FileType.VoiceNote;
  }
  /**
   * Duration of the voice note in seconds.
   */
  get duration() {
    return this.attr.duration;
  }
  /**
   * Voice note's waveform
   *
   * Represented with integers in range [0, 31],
   * usually has length of 100
   */
  get waveform() {
    return voiceUtils.decodeWaveform(this.attr.waveform);
  }
  /** For self-destructing voice notes, TTL in seconds */
  get ttlSeconds() {
    return this.media?.ttlSeconds ?? null;
  }
}
const Voice$1 = /* @__PURE__ */ memoize.memoizeGetters(Voice, ["fileName", "thumbnails", "fileId", "uniqueFileId", "waveform"]);
const Voice$2 = /* @__PURE__ */ inspectable.makeInspectable(Voice$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument", "waveform"]);
exports.Voice = Voice$2;
