// @mtcute/core/highlevel/types/media/voice.js
import { tdFileId } from "@mtcute/file-id";
import { memoizeGetters } from "../../utils/memoize.js";
import { decodeWaveform } from "../../utils/voice-utils.js";
import { RawDocument } from "./document.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Voice extends RawDocument {
  constructor(doc, attr, media) {
    super(doc);
    this.attr = attr;
    this.media = media;
  }
  type = "voice";
  _fileIdType() {
    return tdFileId.FileType.VoiceNote;
  }
  /**
   * Duration of the voice note in seconds.
   */
  get duration() {
    return this.attr.duration;
  }
  /**
   * Voice note's waveform
   *
   * Represented with integers in range [0, 31],
   * usually has length of 100
   */
  get waveform() {
    return decodeWaveform(this.attr.waveform);
  }
  /** For self-destructing voice notes, TTL in seconds */
  get ttlSeconds() {
    return this.media?.ttlSeconds ?? null;
  }
}
const Voice$1 = /* @__PURE__ */ memoizeGetters(Voice, ["fileName", "thumbnails", "fileId", "uniqueFileId", "waveform"]);
const Voice$2 = /* @__PURE__ */ makeInspectable(Voice$1, ["fileSize", "dcId"], ["inputMedia", "inputDocument", "waveform"]);
export {
  Voice$2 as Voice
};
