// @mtcute/core/highlevel/types/messages/dialog.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { assertTypeIsNot, hasValueAtKey } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
import { PeersIndex } from "../peers/peers-index.js";
import { DraftMessage as DraftMessage$2 } from "./draft-message.js";
import { Message as Message$2 } from "./message.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Dialog {
  constructor(raw, _peers, _messages) {
    this.raw = raw;
    this._peers = _peers;
    this._messages = _messages;
  }
  /**
   * Parse a list of dialogs from a TL object
   *
   * @param client  Client instance
   * @param dialogs  TL object
   * @param limit  Maximum number of dialogs to parse
   */
  static parseTlDialogs(dialogs, limit) {
    assertTypeIsNot("parseDialogs", dialogs, "messages.dialogsNotModified");
    const peers = PeersIndex.from(dialogs);
    const messages = /* @__PURE__ */ new Map();
    dialogs.messages.forEach((msg) => {
      if (!msg.peerId) return;
      messages.set(getMarkedPeerId(msg.peerId), msg);
    });
    const arr = dialogs.dialogs.filter(hasValueAtKey("_", "dialog")).map((it) => new Dialog(it, peers, messages));
    if (limit) {
      return arr.slice(0, limit);
    }
    return arr;
  }
  /**
   * Find pinned dialogs from a list of dialogs
   *
   * @param dialogs  Dialogs list
   * @param folder  If passed, status of pin will be checked against this folder, and not globally
   */
  static findPinned(dialogs, folder) {
    if (folder) {
      const index = {};
      folder.pinnedPeers.forEach((peer) => {
        index[getMarkedPeerId(peer)] = true;
      });
      return dialogs.filter((i) => index[i.peer.id]);
    }
    return dialogs.filter((i) => i.isPinned);
  }
  /**
   * Create a filter predicate for the given Folder.
   * Returned predicate can be used in `Array.filter()`
   *
   * @param folder  Folder to filter for
   * @param excludePinned  Whether to exclude pinned folders
   */
  static filterFolder(folder, excludePinned = true) {
    if (folder._ === "dialogFilterDefault") {
      return () => true;
    }
    const pinned = {};
    const include = {};
    const exclude = {};
    if (excludePinned) {
      folder.pinnedPeers.forEach((peer) => {
        pinned[getMarkedPeerId(peer)] = true;
      });
    }
    folder.includePeers.forEach((peer) => {
      include[getMarkedPeerId(peer)] = true;
    });
    if (folder._ === "dialogFilterChatlist") {
      return (dialog) => {
        const peerId = dialog.peer.id;
        if (excludePinned && pinned[peerId]) return false;
        return include[peerId] || pinned[peerId];
      };
    }
    folder.excludePeers.forEach((peer) => {
      exclude[getMarkedPeerId(peer)] = true;
    });
    return (dialog) => {
      const peer = dialog.peer;
      const peerId = dialog.peer.id;
      if (include[peerId]) return true;
      if (exclude[peerId] || excludePinned && pinned[peerId]) {
        return false;
      }
      if (folder.excludeRead && !dialog.isUnread) return false;
      if (folder.excludeMuted && dialog.isMuted) return false;
      if (folder.excludeArchived && dialog.isArchived) return false;
      if (folder.contacts && peer.type === "user" && peer.isContact) {
        return true;
      }
      if (folder.nonContacts && peer.type === "user" && !peer.isContact) {
        return true;
      }
      if (folder.groups && peer.type === "chat" && peer.isGroup) {
        return true;
      }
      if (folder.broadcasts && peer.type === "chat" && peer.chatType === "channel") return true;
      if (folder.bots && peer.type === "user" && peer.isBot) return true;
      return false;
    };
  }
  /**
   * Whether this dialog is pinned
   */
  get isPinned() {
    return this.raw.pinned;
  }
  /**
   * Whether this chat was manually marked as unread
   */
  get isManuallyUnread() {
    return this.raw.unreadMark;
  }
  /**
   * Whether this chat should be considered unread
   * (i.e. has more than 1 unread message, or has
   * a "manually unread" mark)
   */
  get isUnread() {
    return this.raw.unreadMark || this.raw.unreadCount > 1;
  }
  /**
   * Whether this dialog is muted.
   *
   * If `null`, the default account-level setting should be used.
   */
  get isMuted() {
    return this.raw.notifySettings.silent ?? null;
  }
  /**
   * Whether this dialog is archived
   */
  get isArchived() {
    return this.raw.folderId === 1;
  }
  /**
   * Peer that this dialog represents
   */
  get peer() {
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * The latest message sent in this chat (if any)
   */
  get lastMessage() {
    const cid = this.peer.id;
    if (this._messages.has(cid)) {
      const msg = this._messages.get(cid);
      if (msg._ === "messageEmpty") return null;
      return new Message$2(msg, this._peers);
    }
    return null;
  }
  /**
   * ID of the last read outgoing message in this dialog
   */
  get lastReadOutgoing() {
    return this.raw.readOutboxMaxId;
  }
  /**
   * ID of the last read ingoing message in this dialog
   */
  get lastReadIngoing() {
    return this.raw.readInboxMaxId;
  }
  /**
   * ID of the last read message in this dialog
   */
  get lastRead() {
    return Math.max(this.raw.readOutboxMaxId, this.raw.readInboxMaxId);
  }
  /**
   * Number of unread messages
   */
  get unreadCount() {
    return this.raw.unreadCount;
  }
  /**
   * Number of unread mentions
   */
  get unreadMentionsCount() {
    return this.raw.unreadMentionsCount;
  }
  /**
   * Number of unread reactions
   */
  get unreadReactionsCount() {
    return this.raw.unreadReactionsCount;
  }
  /** Number of unread poll votes */
  get unreadPollVotesCount() {
    return this.raw.unreadPollVotesCount;
  }
  /**
   * Draft message in this dialog
   */
  get draftMessage() {
    if (this.raw.draft?._ === "draftMessage") {
      return new DraftMessage$2(this.raw.draft);
    }
    return null;
  }
  /**
   * TTL period of all messages in this dialog
   */
  get ttlPeriod() {
    return this.raw.ttlPeriod ?? null;
  }
}
const Dialog$1 = /* @__PURE__ */ memoizeGetters(Dialog, ["peer", "lastMessage", "draftMessage"]);
const Dialog$2 = /* @__PURE__ */ makeInspectable(Dialog$1);
export {
  Dialog$2 as Dialog
};
