// @mtcute/core/highlevel/types/messages/dialog.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
const peersIndex = require("../peers/peers-index.cjs");
const draftMessage = require("./draft-message.cjs");
const message = require("./message.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Dialog {
  constructor(raw, _peers, _messages) {
    this.raw = raw;
    this._peers = _peers;
    this._messages = _messages;
  }
  /**
   * Parse a list of dialogs from a TL object
   *
   * @param client  Client instance
   * @param dialogs  TL object
   * @param limit  Maximum number of dialogs to parse
   */
  static parseTlDialogs(dialogs, limit) {
    typeAssertions.assertTypeIsNot("parseDialogs", dialogs, "messages.dialogsNotModified");
    const peers = peersIndex.PeersIndex.from(dialogs);
    const messages = /* @__PURE__ */ new Map();
    dialogs.messages.forEach((msg) => {
      if (!msg.peerId) return;
      messages.set(peerUtils.getMarkedPeerId(msg.peerId), msg);
    });
    const arr = dialogs.dialogs.filter(typeAssertions.hasValueAtKey("_", "dialog")).map((it) => new Dialog(it, peers, messages));
    if (limit) {
      return arr.slice(0, limit);
    }
    return arr;
  }
  /**
   * Find pinned dialogs from a list of dialogs
   *
   * @param dialogs  Dialogs list
   * @param folder  If passed, status of pin will be checked against this folder, and not globally
   */
  static findPinned(dialogs, folder) {
    if (folder) {
      const index = {};
      folder.pinnedPeers.forEach((peer2) => {
        index[peerUtils.getMarkedPeerId(peer2)] = true;
      });
      return dialogs.filter((i) => index[i.peer.id]);
    }
    return dialogs.filter((i) => i.isPinned);
  }
  /**
   * Create a filter predicate for the given Folder.
   * Returned predicate can be used in `Array.filter()`
   *
   * @param folder  Folder to filter for
   * @param excludePinned  Whether to exclude pinned folders
   */
  static filterFolder(folder, excludePinned = true) {
    if (folder._ === "dialogFilterDefault") {
      return () => true;
    }
    const pinned = {};
    const include = {};
    const exclude = {};
    if (excludePinned) {
      folder.pinnedPeers.forEach((peer2) => {
        pinned[peerUtils.getMarkedPeerId(peer2)] = true;
      });
    }
    folder.includePeers.forEach((peer2) => {
      include[peerUtils.getMarkedPeerId(peer2)] = true;
    });
    if (folder._ === "dialogFilterChatlist") {
      return (dialog) => {
        const peerId = dialog.peer.id;
        if (excludePinned && pinned[peerId]) return false;
        return include[peerId] || pinned[peerId];
      };
    }
    folder.excludePeers.forEach((peer2) => {
      exclude[peerUtils.getMarkedPeerId(peer2)] = true;
    });
    return (dialog) => {
      const peer2 = dialog.peer;
      const peerId = dialog.peer.id;
      if (include[peerId]) return true;
      if (exclude[peerId] || excludePinned && pinned[peerId]) {
        return false;
      }
      if (folder.excludeRead && !dialog.isUnread) return false;
      if (folder.excludeMuted && dialog.isMuted) return false;
      if (folder.excludeArchived && dialog.isArchived) return false;
      if (folder.contacts && peer2.type === "user" && peer2.isContact) {
        return true;
      }
      if (folder.nonContacts && peer2.type === "user" && !peer2.isContact) {
        return true;
      }
      if (folder.groups && peer2.type === "chat" && peer2.isGroup) {
        return true;
      }
      if (folder.broadcasts && peer2.type === "chat" && peer2.chatType === "channel") return true;
      if (folder.bots && peer2.type === "user" && peer2.isBot) return true;
      return false;
    };
  }
  /**
   * Whether this dialog is pinned
   */
  get isPinned() {
    return this.raw.pinned;
  }
  /**
   * Whether this chat was manually marked as unread
   */
  get isManuallyUnread() {
    return this.raw.unreadMark;
  }
  /**
   * Whether this chat should be considered unread
   * (i.e. has more than 1 unread message, or has
   * a "manually unread" mark)
   */
  get isUnread() {
    return this.raw.unreadMark || this.raw.unreadCount > 1;
  }
  /**
   * Whether this dialog is muted.
   *
   * If `null`, the default account-level setting should be used.
   */
  get isMuted() {
    return this.raw.notifySettings.silent ?? null;
  }
  /**
   * Whether this dialog is archived
   */
  get isArchived() {
    return this.raw.folderId === 1;
  }
  /**
   * Peer that this dialog represents
   */
  get peer() {
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * The latest message sent in this chat (if any)
   */
  get lastMessage() {
    const cid = this.peer.id;
    if (this._messages.has(cid)) {
      const msg = this._messages.get(cid);
      if (msg._ === "messageEmpty") return null;
      return new message.Message(msg, this._peers);
    }
    return null;
  }
  /**
   * ID of the last read outgoing message in this dialog
   */
  get lastReadOutgoing() {
    return this.raw.readOutboxMaxId;
  }
  /**
   * ID of the last read ingoing message in this dialog
   */
  get lastReadIngoing() {
    return this.raw.readInboxMaxId;
  }
  /**
   * ID of the last read message in this dialog
   */
  get lastRead() {
    return Math.max(this.raw.readOutboxMaxId, this.raw.readInboxMaxId);
  }
  /**
   * Number of unread messages
   */
  get unreadCount() {
    return this.raw.unreadCount;
  }
  /**
   * Number of unread mentions
   */
  get unreadMentionsCount() {
    return this.raw.unreadMentionsCount;
  }
  /**
   * Number of unread reactions
   */
  get unreadReactionsCount() {
    return this.raw.unreadReactionsCount;
  }
  /** Number of unread poll votes */
  get unreadPollVotesCount() {
    return this.raw.unreadPollVotesCount;
  }
  /**
   * Draft message in this dialog
   */
  get draftMessage() {
    if (this.raw.draft?._ === "draftMessage") {
      return new draftMessage.DraftMessage(this.raw.draft);
    }
    return null;
  }
  /**
   * TTL period of all messages in this dialog
   */
  get ttlPeriod() {
    return this.raw.ttlPeriod ?? null;
  }
}
const Dialog$1 = /* @__PURE__ */ memoize.memoizeGetters(Dialog, ["peer", "lastMessage", "draftMessage"]);
const Dialog$2 = /* @__PURE__ */ inspectable.makeInspectable(Dialog$1);
exports.Dialog = Dialog$2;
