// @mtcute/core/highlevel/types/messages/draft-message.js
import { memoizeGetters } from "../../utils/memoize.js";
import { MessageEntity as MessageEntity$2 } from "./message-entity.js";
import { makeInspectable } from "../../utils/inspectable.js";
class DraftMessage {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Text of the draft message
   */
  get text() {
    return this.raw.message;
  }
  /**
   * Information about replies/quotes in this message
   */
  get replyToMessage() {
    if (this.raw.replyTo?._ !== "inputReplyToMessage") return null;
    return this.raw.replyTo;
  }
  /**
   * Date of the last time this draft was updated
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether no webpage preview will be generated
   */
  get disableWebPreview() {
    return this.raw.noWebpage;
  }
  /**
   * Message text entities (may be empty)
   */
  get entities() {
    const entities = [];
    if (this.raw.entities?.length) {
      for (const ent of this.raw.entities) {
        entities.push(new MessageEntity$2(ent, this.raw.message));
      }
    }
    return entities;
  }
}
const DraftMessage$1 = /* @__PURE__ */ memoizeGetters(DraftMessage, ["entities"]);
const DraftMessage$2 = /* @__PURE__ */ makeInspectable(DraftMessage$1);
export {
  DraftMessage$2 as DraftMessage
};
