// @mtcute/core/highlevel/types/messages/draft-message.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const messageEntity = require("./message-entity.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class DraftMessage {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Text of the draft message
   */
  get text() {
    return this.raw.message;
  }
  /**
   * Information about replies/quotes in this message
   */
  get replyToMessage() {
    if (this.raw.replyTo?._ !== "inputReplyToMessage") return null;
    return this.raw.replyTo;
  }
  /**
   * Date of the last time this draft was updated
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether no webpage preview will be generated
   */
  get disableWebPreview() {
    return this.raw.noWebpage;
  }
  /**
   * Message text entities (may be empty)
   */
  get entities() {
    const entities = [];
    if (this.raw.entities?.length) {
      for (const ent of this.raw.entities) {
        entities.push(new messageEntity.MessageEntity(ent, this.raw.message));
      }
    }
    return entities;
  }
}
const DraftMessage$1 = /* @__PURE__ */ memoize.memoizeGetters(DraftMessage, ["entities"]);
const DraftMessage$2 = /* @__PURE__ */ inspectable.makeInspectable(DraftMessage$1);
exports.DraftMessage = DraftMessage$2;
