// @mtcute/core/highlevel/types/messages/fact-check.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { MessageEntity as MessageEntity$2 } from "./message-entity.js";
class FactCheck {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Text of the fact-check
   */
  get text() {
    return this.raw.text?.text ?? "";
  }
  /**
   * Entities contained in the fact-check text
   */
  get entities() {
    const entities = [];
    if (this.raw.text?.entities) {
      for (const ent of this.raw.text.entities) {
        entities.push(new MessageEntity$2(ent, this.raw.text.text));
      }
    }
    return entities;
  }
  /**
   * Country for which the fact-check is relevant
   */
  get country() {
    return this.raw.country ?? null;
  }
  /**
   * Whether this information might be outdated
   * and should be re-fetched manually
   */
  get shouldRecheck() {
    return this.raw.needCheck;
  }
  /**
   * Hash of the fact-check
   */
  get hash() {
    return this.raw.hash;
  }
}
const FactCheck$1 = /* @__PURE__ */ memoizeGetters(FactCheck, ["entities"]);
const FactCheck$2 = /* @__PURE__ */ makeInspectable(FactCheck$1);
export {
  FactCheck$2 as FactCheck
};
