// @mtcute/core/highlevel/types/messages/fact-check.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const messageEntity = require("./message-entity.cjs");
class FactCheck {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Text of the fact-check
   */
  get text() {
    return this.raw.text?.text ?? "";
  }
  /**
   * Entities contained in the fact-check text
   */
  get entities() {
    const entities = [];
    if (this.raw.text?.entities) {
      for (const ent of this.raw.text.entities) {
        entities.push(new messageEntity.MessageEntity(ent, this.raw.text.text));
      }
    }
    return entities;
  }
  /**
   * Country for which the fact-check is relevant
   */
  get country() {
    return this.raw.country ?? null;
  }
  /**
   * Whether this information might be outdated
   * and should be re-fetched manually
   */
  get shouldRecheck() {
    return this.raw.needCheck;
  }
  /**
   * Hash of the fact-check
   */
  get hash() {
    return this.raw.hash;
  }
}
const FactCheck$1 = /* @__PURE__ */ memoize.memoizeGetters(FactCheck, ["entities"]);
const FactCheck$2 = /* @__PURE__ */ inspectable.makeInspectable(FactCheck$1);
exports.FactCheck = FactCheck$2;
