// @mtcute/core/highlevel/types/messages/message-effect.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const documentUtils = require("../media/document-utils.cjs");
class MessageEffect {
  constructor(raw, documentsMap) {
    this.raw = raw;
    this.documentsMap = documentsMap;
  }
  /** Whether Telegram Premium is required to use this effect */
  get isPremiumRequired() {
    return this.raw.premiumRequired;
  }
  /** ID of this effect */
  get id() {
    return this.raw.id;
  }
  /** Emoji representint this reaction */
  get emoji() {
    return this.raw.emoticon;
  }
  /** Sticker representing a static icon for this effect (if any) */
  get staticIcon() {
    if (!this.raw.staticIconId) return null;
    const document = this.documentsMap.get(this.raw.staticIconId);
    if (!document) return null;
    const parsed = documentUtils.parseSticker(document);
    if (!parsed) return null;
    return parsed;
  }
  /** Animated icon representing the effect  */
  get icon() {
    const document = this.documentsMap.get(this.raw.effectStickerId);
    if (!document) {
      throw new errors.MtTypeAssertionError("MessageEffect.effect", "document", "null");
    }
    const parsed = documentUtils.parseSticker(document);
    if (!parsed) {
      throw new errors.MtTypeAssertionError("MessageEffect.effect", "sticker", "null");
    }
    return parsed;
  }
  /** The animation itself */
  get animation() {
    if (!this.raw.effectAnimationId) return null;
    const document = this.documentsMap.get(this.raw.effectAnimationId);
    if (!document) return null;
    const parsed = documentUtils.parseSticker(document);
    if (!parsed) return null;
    return parsed;
  }
}
const MessageEffect$1 = /* @__PURE__ */ memoize.memoizeGetters(MessageEffect, ["staticIcon", "icon", "animation"]);
const MessageEffect$2 = /* @__PURE__ */ inspectable.makeInspectable(MessageEffect$1);
exports.MessageEffect = MessageEffect$2;
