// @mtcute/core/highlevel/types/messages/message-replies.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
class MessageRepliesInfo {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether this message is a channel post that has a comments thread
   * in the linked discussion group
   */
  get hasComments() {
    return this.raw.comments;
  }
  /**
   * Total number of replies
   */
  get count() {
    return this.raw.replies;
  }
  /**
   * Whether this reply thread has unread messages
   */
  get hasUnread() {
    return this.raw.readMaxId !== void 0 && this.raw.readMaxId !== this.raw.maxId;
  }
  /**
   * ID of the last message in the thread (if any)
   */
  get lastMessageId() {
    return this.raw.maxId ?? null;
  }
  /**
   * ID of the last read message in the thread (if any)
   */
  get lastReadMessageId() {
    return this.raw.readMaxId ?? null;
  }
  /**
   * ID of the discussion group for the post
   *
   * `null` if the post is not a channel post
   */
  get discussion() {
    if (!this.raw.channelId) return null;
    return peerUtils.getMarkedPeerId(this.raw.channelId, "channel");
  }
  /**
   * Last few commenters to the post (usually 3)
   */
  get repliers() {
    return this.raw.recentRepliers?.map((it) => peer.parsePeer(it, this._peers)) ?? [];
  }
}
const MessageRepliesInfo$1 = /* @__PURE__ */ memoize.memoizeGetters(MessageRepliesInfo, ["discussion", "repliers"]);
const MessageRepliesInfo$2 = /* @__PURE__ */ inspectable.makeInspectable(MessageRepliesInfo$1);
exports.MessageRepliesInfo = MessageRepliesInfo$2;
