// @mtcute/core/highlevel/types/messages/message-forward.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
class MessageForwardInfo {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Date the original message was sent
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Sender of the original message (either user or a channel)
   * or their name (for users with private forwards)
   */
  get sender() {
    if (this.raw.fromName) {
      return {
        type: "anonymous",
        displayName: this.raw.fromName
      };
    }
    if (this.raw.fromId) {
      return peer.parsePeer(this.raw.fromId, this._peers);
    }
    throw new errors.MtTypeAssertionError("MessageForwardInfo", "to have fromId or fromName", "neither");
  }
  /**
   * For "saved" messages (i.e. messages forwarded to yourself,
   * "Saved Messages"), the peer where the message was originally sent.
   *
   * `null` for other messages, you might want to use {@link sender} instead
   */
  fromChat() {
    if (!this.raw.savedFromPeer) return null;
    return peer.parsePeer(this.raw.savedFromPeer, this._peers);
  }
  /**
   * For messages forwarded from channels,
   * identifier of the original message in the channel
   *
   * (only availale if {@link fromChat} is not `null`)
   */
  get fromMessageId() {
    return this.raw.savedFromMsgId ?? null;
  }
  /**
   * For messages forwarded from channels,
   * signature of the post author (if present)
   */
  get signature() {
    return this.raw.postAuthor ?? null;
  }
}
const MessageForwardInfo$1 = /* @__PURE__ */ memoize.memoizeGetters(MessageForwardInfo, ["sender", "fromChat"]);
const MessageForwardInfo$2 = /* @__PURE__ */ inspectable.makeInspectable(MessageForwardInfo$1);
exports.MessageForwardInfo = MessageForwardInfo$2;
