// @mtcute/core/highlevel/types/messages/message-reactions.js
import { memoizeGetters } from "../../utils/memoize.js";
import { PeerReaction as PeerReaction$2, PaidPeerReaction as PaidPeerReaction$2 } from "../reactions/peer-reaction.js";
import { ReactionCount as ReactionCount$1 } from "../reactions/reaction-count.js";
import { makeInspectable } from "../../utils/inspectable.js";
class MessageReactions {
  constructor(messageId, chatId, raw, _peers) {
    this.messageId = messageId;
    this.chatId = chatId;
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether you can use {@link getUsers}
   * (or {@link TelegramClient.getReactionUsers})
   * to get the users who reacted to this message
   */
  get usersVisible() {
    return this.raw.canSeeList;
  }
  /**
   * Reactions on the message, along with their counts
   */
  get reactions() {
    return this.raw.results.map((it) => new ReactionCount$1(it));
  }
  /**
   * Recently reacted users.
   * To get a full list of users, use {@link getUsers}
   */
  get recentReactions() {
    return this.raw.recentReactions?.map((reaction) => new PeerReaction$2(reaction, this._peers)) ?? [];
  }
  /** Leaderboard of paid reactions to the message */
  get paidReactions() {
    return this.raw.topReactors?.map((reaction) => new PaidPeerReaction$2(reaction, this._peers)) ?? [];
  }
}
const MessageReactions$1 = /* @__PURE__ */ memoizeGetters(MessageReactions, ["reactions", "recentReactions"]);
const MessageReactions$2 = /* @__PURE__ */ makeInspectable(MessageReactions$1);
export {
  MessageReactions$2 as MessageReactions
};
