// @mtcute/core/highlevel/types/messages/message-entity.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class MessageEntity {
  constructor(raw, _text) {
    this.raw = raw;
    this._text = _text;
  }
  /**
   * Offset in UTF-16 code units to the start of the entity.
   *
   * Since JS strings are UTF-16, you can use this as-is
   */
  get offset() {
    return this.raw.offset;
  }
  /**
   * Length of the entity in UTF-16 code units.
   *
   * Since JS strings are UTF-16, you can use this as-is
   */
  get length() {
    return this.raw.length;
  }
  /**
   * Kind of the entity (see {@link MessageEntityParams})
   */
  get kind() {
    return this.params.kind;
  }
  /**
   * Params of the entity
   */
  get params() {
    switch (this.raw._) {
      case "messageEntityMention":
        return { kind: "mention" };
      case "messageEntityHashtag":
        return { kind: "hashtag" };
      case "messageEntityCashtag":
        return { kind: "cashtag" };
      case "messageEntityBotCommand":
        return { kind: "bot_command" };
      case "messageEntityUrl":
        return { kind: "url" };
      case "messageEntityEmail":
        return { kind: "email" };
      case "messageEntityPhone":
        return { kind: "phone_number" };
      case "messageEntityBold":
        return { kind: "bold" };
      case "messageEntityItalic":
        return { kind: "italic" };
      case "messageEntityUnderline":
        return { kind: "underline" };
      case "messageEntityStrike":
        return { kind: "strikethrough" };
      case "messageEntitySpoiler":
        return { kind: "spoiler" };
      case "messageEntityCode":
        return { kind: "code" };
      case "messageEntityPre":
        return { kind: "pre", language: this.raw.language };
      case "messageEntityTextUrl":
        return { kind: "text_link", url: this.raw.url };
      case "messageEntityMentionName":
        return { kind: "text_mention", userId: this.raw.userId };
      case "messageEntityBlockquote":
        return { kind: "blockquote", collapsible: this.raw.collapsed };
      case "messageEntityCustomEmoji":
        return { kind: "emoji", emojiId: this.raw.documentId };
      case "messageEntityBankCard":
        return { kind: "bank_card" };
    }
    return { kind: "unknown" };
  }
  /**
   * Text contained in this entity.
   *
   * > **Note**: This does not take into account that entities may overlap,
   * > and is only useful for simple cases.
   */
  get text() {
    if (!this._text) return "";
    return this._text.slice(this.raw.offset, this.raw.offset + this.raw.length);
  }
  /**
   * Checks if this entity is of the given type, and adjusts the type accordingly.
   * @param kind
   * @returns
   */
  is(kind) {
    return this.params.kind === kind;
  }
}
const MessageEntity$1 = /* @__PURE__ */ memoize.memoizeGetters(MessageEntity, ["params"]);
const MessageEntity$2 = /* @__PURE__ */ inspectable.makeInspectable(MessageEntity$1);
exports.MessageEntity = MessageEntity$2;
