// @mtcute/core/highlevel/types/messages/message-effect.js
import { MtTypeAssertionError } from "../../../types/errors.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parseSticker } from "../media/document-utils.js";
class MessageEffect {
  constructor(raw, documentsMap) {
    this.raw = raw;
    this.documentsMap = documentsMap;
  }
  /** Whether Telegram Premium is required to use this effect */
  get isPremiumRequired() {
    return this.raw.premiumRequired;
  }
  /** ID of this effect */
  get id() {
    return this.raw.id;
  }
  /** Emoji representint this reaction */
  get emoji() {
    return this.raw.emoticon;
  }
  /** Sticker representing a static icon for this effect (if any) */
  get staticIcon() {
    if (!this.raw.staticIconId) return null;
    const document = this.documentsMap.get(this.raw.staticIconId);
    if (!document) return null;
    const parsed = parseSticker(document);
    if (!parsed) return null;
    return parsed;
  }
  /** Animated icon representing the effect  */
  get icon() {
    const document = this.documentsMap.get(this.raw.effectStickerId);
    if (!document) {
      throw new MtTypeAssertionError("MessageEffect.effect", "document", "null");
    }
    const parsed = parseSticker(document);
    if (!parsed) {
      throw new MtTypeAssertionError("MessageEffect.effect", "sticker", "null");
    }
    return parsed;
  }
  /** The animation itself */
  get animation() {
    if (!this.raw.effectAnimationId) return null;
    const document = this.documentsMap.get(this.raw.effectAnimationId);
    if (!document) return null;
    const parsed = parseSticker(document);
    if (!parsed) return null;
    return parsed;
  }
}
const MessageEffect$1 = /* @__PURE__ */ memoizeGetters(MessageEffect, ["staticIcon", "icon", "animation"]);
const MessageEffect$2 = /* @__PURE__ */ makeInspectable(MessageEffect$1);
export {
  MessageEffect$2 as MessageEffect
};
