// @mtcute/core/highlevel/types/messages/message-replies.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
class MessageRepliesInfo {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether this message is a channel post that has a comments thread
   * in the linked discussion group
   */
  get hasComments() {
    return this.raw.comments;
  }
  /**
   * Total number of replies
   */
  get count() {
    return this.raw.replies;
  }
  /**
   * Whether this reply thread has unread messages
   */
  get hasUnread() {
    return this.raw.readMaxId !== void 0 && this.raw.readMaxId !== this.raw.maxId;
  }
  /**
   * ID of the last message in the thread (if any)
   */
  get lastMessageId() {
    return this.raw.maxId ?? null;
  }
  /**
   * ID of the last read message in the thread (if any)
   */
  get lastReadMessageId() {
    return this.raw.readMaxId ?? null;
  }
  /**
   * ID of the discussion group for the post
   *
   * `null` if the post is not a channel post
   */
  get discussion() {
    if (!this.raw.channelId) return null;
    return getMarkedPeerId(this.raw.channelId, "channel");
  }
  /**
   * Last few commenters to the post (usually 3)
   */
  get repliers() {
    return this.raw.recentRepliers?.map((it) => parsePeer(it, this._peers)) ?? [];
  }
}
const MessageRepliesInfo$1 = /* @__PURE__ */ memoizeGetters(MessageRepliesInfo, ["discussion", "repliers"]);
const MessageRepliesInfo$2 = /* @__PURE__ */ makeInspectable(MessageRepliesInfo$1);
export {
  MessageRepliesInfo$2 as MessageRepliesInfo
};
