// @mtcute/core/highlevel/types/messages/message-forward.js
import { MtTypeAssertionError } from "../../../types/errors.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
class MessageForwardInfo {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Date the original message was sent
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Sender of the original message (either user or a channel)
   * or their name (for users with private forwards)
   */
  get sender() {
    if (this.raw.fromName) {
      return {
        type: "anonymous",
        displayName: this.raw.fromName
      };
    }
    if (this.raw.fromId) {
      return parsePeer(this.raw.fromId, this._peers);
    }
    throw new MtTypeAssertionError("MessageForwardInfo", "to have fromId or fromName", "neither");
  }
  /**
   * For "saved" messages (i.e. messages forwarded to yourself,
   * "Saved Messages"), the peer where the message was originally sent.
   *
   * `null` for other messages, you might want to use {@link sender} instead
   */
  fromChat() {
    if (!this.raw.savedFromPeer) return null;
    return parsePeer(this.raw.savedFromPeer, this._peers);
  }
  /**
   * For messages forwarded from channels,
   * identifier of the original message in the channel
   *
   * (only availale if {@link fromChat} is not `null`)
   */
  get fromMessageId() {
    return this.raw.savedFromMsgId ?? null;
  }
  /**
   * For messages forwarded from channels,
   * signature of the post author (if present)
   */
  get signature() {
    return this.raw.postAuthor ?? null;
  }
}
const MessageForwardInfo$1 = /* @__PURE__ */ memoizeGetters(MessageForwardInfo, ["sender", "fromChat"]);
const MessageForwardInfo$2 = /* @__PURE__ */ makeInspectable(MessageForwardInfo$1);
export {
  MessageForwardInfo$2 as MessageForwardInfo
};
