// @mtcute/core/highlevel/types/messages/message-action.js
import Long from "long";
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { Photo as Photo$2 } from "../media/photo.js";
import { parsePeer } from "../peers/peer.js";
import { User as User$2 } from "../peers/user.js";
import { StarGiftUnique as StarGiftUnique$2 } from "../premium/star-gift-unique.js";
import { StarGift as StarGift$2 } from "../premium/star-gift.js";
function _messageActionFromTl(act) {
  switch (act._) {
    case "messageActionChatCreate":
      return {
        type: "chat_created",
        title: act.title,
        users: act.users
      };
    case "messageActionChannelCreate":
      return {
        type: "channel_created",
        title: act.title
      };
    case "messageActionChatMigrateTo":
      return {
        type: "chat_migrate_to",
        id: act.channelId
      };
    case "messageActionChannelMigrateFrom":
      return {
        type: "channel_migrate_from",
        id: act.chatId,
        title: act.title
      };
    case "messageActionPinMessage":
      return {
        type: "message_pinned"
      };
    case "messageActionHistoryClear":
      return {
        type: "history_cleared"
      };
    case "messageActionGameScore":
      return {
        type: "game_score",
        gameId: act.gameId,
        score: act.score
      };
    case "messageActionContactSignUp":
      return {
        type: "contact_joined"
      };
    case "messageActionChatEditTitle":
      return {
        type: "title_changed",
        title: act.title
      };
    case "messageActionChatEditPhoto":
      if (act.photo._ === "photoEmpty") {
        return { type: "photo_deleted" };
      }
      return {
        type: "photo_changed",
        photo: new Photo$2(act.photo)
      };
    case "messageActionChatDeletePhoto":
      return {
        type: "photo_deleted"
      };
    case "messageActionChatAddUser":
      return {
        type: "users_added",
        users: act.users
      };
    case "messageActionChatDeleteUser":
      if (this.raw.fromId?._ === "peerUser" && act.userId === this.raw.fromId.userId) {
        return {
          type: "user_left"
        };
      }
      return {
        type: "user_removed",
        user: act.userId
      };
    case "messageActionChatJoinedByLink":
      return {
        type: "user_joined_link",
        inviter: act.inviterId
      };
    case "messageActionPaymentSentMe":
      return {
        type: "payment_received",
        currency: act.currency,
        amount: act.totalAmount,
        payload: act.payload,
        info: act.info,
        shippingOptionId: act.shippingOptionId,
        charge: act.charge,
        subscriptionUntilDate: act.subscriptionUntilDate ? new Date(act.subscriptionUntilDate * 1e3) : void 0
      };
    case "messageActionPaymentSent":
      return {
        type: "payment_sent",
        currency: act.currency,
        amount: act.totalAmount,
        subscriptionUntilDate: act.subscriptionUntilDate ? new Date(act.subscriptionUntilDate * 1e3) : void 0
      };
    case "messageActionPhoneCall":
      return {
        type: "call",
        id: act.callId,
        isVideo: Boolean(act.video),
        reason: act.reason,
        duration: act.duration ?? 0
      };
    case "messageActionScreenshotTaken":
      return {
        type: "screenshot_taken"
      };
    case "messageActionBotAllowed":
      return {
        type: "bot_allowed",
        domain: act.domain
      };
    case "messageActionGeoProximityReached":
      if (act.fromId._ !== "peerUser" || act.toId._ !== "peerUser") {
        return null;
      }
      return {
        type: "geo_proximity",
        targetId: act.toId.userId,
        userId: act.fromId.userId,
        distance: act.distance
      };
    case "messageActionGroupCall":
      if (act.duration) {
        return {
          type: "group_call_ended",
          call: act.call,
          duration: act.duration
        };
      }
      return {
        type: "group_call_started",
        call: act.call
      };
    case "messageActionGroupCallScheduled":
      return {
        type: "group_call_scheduled",
        call: act.call,
        date: new Date(act.scheduleDate * 1e3)
      };
    case "messageActionInviteToGroupCall":
      return {
        type: "group_call_invite",
        call: act.call,
        userIds: act.users
      };
    case "messageActionSetMessagesTTL":
      return {
        type: "ttl_changed",
        period: act.period
      };
    case "messageActionTopicCreate":
      return {
        type: "topic_created",
        title: act.title,
        iconColor: act.iconColor,
        iconCustomEmoji: act.iconEmojiId,
        titleMissing: act.titleMissing
      };
    case "messageActionTopicEdit":
      return {
        type: "topic_edited",
        title: act.title,
        iconCustomEmoji: act.iconEmojiId,
        closed: act.closed,
        hidden: act.hidden
      };
    case "messageActionCustomAction":
      return {
        type: "custom",
        action: act.message
      };
    case "messageActionSetChatTheme":
      if (act.theme._ === "chatThemeUniqueGift") {
        if (act.theme.gift._ !== "starGiftUnique") return null;
        return {
          type: "theme_changed_unique",
          gift: new StarGiftUnique$2(act.theme.gift, this._peers),
          themeSettings: act.theme.themeSettings
        };
      }
      return {
        type: "theme_changed",
        emoji: act.theme.emoticon
      };
    case "messageActionChatJoinedByRequest":
      return {
        type: "user_joined_approved"
      };
    case "messageActionWebViewDataSent":
      return {
        type: "webview_sent",
        text: act.text
      };
    case "messageActionWebViewDataSentMe":
      return {
        type: "webview_received",
        text: act.text,
        data: act.data
      };
    case "messageActionGiftPremium":
      return {
        type: "premium_gifted",
        currency: act.currency,
        amount: act.amount.toNumber(),
        days: act.days,
        crypto: act.cryptoAmount ? {
          currency: act.cryptoCurrency,
          amount: act.cryptoAmount.toNumber()
        } : void 0,
        message: act.message
      };
    case "messageActionSuggestProfilePhoto":
      if (act.photo._ !== "photo") return null;
      return {
        type: "photo_suggested",
        photo: new Photo$2(act.photo)
      };
    case "messageActionRequestedPeerSentMe":
      return {
        type: "peer_sent",
        buttonId: act.buttonId,
        peers: act.peers
      };
    case "messageActionRequestedPeer":
      return {
        type: "peer_chosen",
        buttonId: act.buttonId,
        peers: act.peers.map((it) => parsePeer(it, this._peers))
      };
    case "messageActionSetChatWallPaper":
      return {
        type: "wallpaper_changed",
        same: act.same,
        forBoth: act.forBoth,
        wallpaper: act.wallpaper
      };
    case "messageActionGiftCode":
      return {
        type: "gift_code",
        raw: act,
        message: act.message
      };
    case "messageActionGiveawayLaunch":
      return {
        type: "giveaway_started"
      };
    case "messageActionGiveawayResults":
      return {
        type: "giveaway_ended",
        winners: act.winnersCount,
        undistributed: act.unclaimedCount
      };
    case "messageActionBoostApply":
      return {
        type: "boost_apply",
        count: act.boosts
      };
    case "messageActionPaymentRefunded":
      return {
        type: "payment_refunded",
        currency: act.currency,
        peerId: getMarkedPeerId(act.peer),
        amount: act.totalAmount,
        payload: act.payload,
        charge: act.charge
      };
    case "messageActionGiftStars":
      return {
        type: "stars_gifted",
        isCrypto: Boolean(act.cryptoCurrency),
        currency: act.cryptoCurrency ?? act.currency,
        amount: act.cryptoAmount ?? act.amount,
        stars: act.stars,
        transactionId: act.transactionId
      };
    case "messageActionPrizeStars":
      return {
        type: "stars_prize",
        claimed: !act.unclaimed,
        stars: act.stars,
        transactionId: act.transactionId,
        boostPeer: parsePeer(act.boostPeer, this._peers),
        giveawayMessageId: act.giveawayMsgId
      };
    case "messageActionStarGift":
      if (act.gift._ !== "starGift") return null;
      if (act.prepaidUpgrade) {
        return {
          type: "star_gift_upgrade_paid",
          gift: new StarGift$2(act.gift, this._peers),
          giftMsgId: act.giftMsgId ?? void 0,
          upgradeMsgId: act.upgradeMsgId ?? null
        };
      }
      return {
        type: "star_gift_sent",
        nameHidden: act.nameHidden,
        saved: act.saved,
        converted: act.converted,
        convertStars: act.convertStars ?? Long.ZERO,
        refunded: act.refunded,
        gift: new StarGift$2(act.gift, this._peers),
        message: act.message ?? null,
        canUpgrade: act.canUpgrade,
        upgraded: act.upgraded,
        upgradeStars: act.upgradeStars ?? Long.ZERO,
        upgradeMsgId: act.upgradeMsgId ?? null,
        fromAuction: act.auctionAcquired,
        giftNum: act.giftNum,
        toId: act.toId ? parsePeer(act.toId, this._peers) : null,
        savedId: act.savedId,
        prepaidUpgradeHash: act.prepaidUpgradeHash ?? void 0
      };
    case "messageActionStarGiftUnique": {
      if (act.gift._ !== "starGiftUnique") return null;
      const common = {
        saved: act.saved,
        refunded: act.refunded,
        gift: new StarGiftUnique$2(act.gift, this._peers),
        canTransferAt: act.canTransferAt ? new Date(act.canTransferAt * 1e3) : void 0,
        canResellAt: act.canResellAt ? new Date(act.canResellAt * 1e3) : void 0,
        canExportAt: act.canExportAt ? new Date(act.canExportAt * 1e3) : void 0,
        canCraftAt: act.canCraftAt ? new Date(act.canCraftAt * 1e3) : void 0,
        savedId: act.savedId,
        dropDetailsStars: act.dropOriginalDetailsStars
      };
      if (act.assigned) {
        return {
          type: "star_gift_assigned",
          ...common
        };
      }
      if (act.craft) {
        return {
          type: "star_gift_crafted",
          ...common
        };
      }
      if (act.fromOffer) {
        return {
          type: "star_gift_bought_offer",
          ...common
        };
      }
      if (act.resaleAmount) {
        return {
          type: "star_gift_bought_resale",
          ...common,
          amount: act.resaleAmount
        };
      }
      if (act.upgrade) {
        return {
          type: "star_gift_upgraded",
          ...common
        };
      }
      return {
        type: "star_gift_transferred",
        ...common
      };
    }
    case "messageActionPaidMessagesPrice":
      return {
        type: "messages_paid",
        stars: act.stars
      };
    case "messageActionPaidMessagesRefunded":
      return {
        type: "messages_refunded",
        stars: act.stars,
        count: act.count
      };
    case "messageActionTodoCompletions":
      return {
        type: "todo_completions",
        completed: act.completed,
        uncompleted: act.incompleted
      };
    case "messageActionTodoAppendTasks":
      return {
        type: "todo_append_tasks",
        list: act.list
      };
    case "messageActionGiftTon":
      return {
        type: "ton_gift",
        currency: act.currency,
        amount: act.amount,
        cryptoCurrency: act.cryptoCurrency,
        cryptoAmount: act.cryptoAmount,
        transactionId: act.transactionId
      };
    case "messageActionSuggestBirthday":
      return {
        type: "suggest_birthday",
        birthday: act.birthday
      };
    case "messageActionStarGiftPurchaseOffer":
      if (act.gift._ !== "starGiftUnique") return null;
      return {
        type: "star_gift_purchase_offer",
        status: act.accepted ? "accepted" : act.declined ? "declined" : "pending",
        gift: new StarGiftUnique$2(act.gift, this._peers),
        price: act.price,
        expiresAt: new Date(act.expiresAt * 1e3)
      };
    case "messageActionStarGiftPurchaseOfferDeclined":
      if (act.gift._ !== "starGiftUnique") return null;
      return {
        type: "star_gift_purchase_offer_declined",
        expired: act.expired,
        gift: new StarGiftUnique$2(act.gift, this._peers),
        price: act.price
      };
    case "messageActionNewCreatorPending":
      return {
        type: "new_creator_pending",
        newCreator: new User$2(this._peers.user(act.newCreatorId))
      };
    case "messageActionChangeCreator":
      return {
        type: "change_creator",
        newCreator: new User$2(this._peers.user(act.newCreatorId))
      };
    case "messageActionNoForwardsToggle":
      return {
        type: "no_forwards_toggle",
        prevValue: act.prevValue,
        newValue: act.newValue
      };
    case "messageActionNoForwardsRequest":
      return {
        type: "no_forwards_request",
        expired: act.expired,
        prevValue: act.prevValue,
        newValue: act.newValue
      };
    case "messageActionPollAppendAnswer":
      assertTypeIsNot("messageActionPollAppendAnswer", act.answer, "inputPollAnswer");
      return {
        type: "poll_append_answer",
        answer: act.answer
      };
    case "messageActionPollDeleteAnswer":
      assertTypeIsNot("messageActionPollAppendAnswer", act.answer, "inputPollAnswer");
      return {
        type: "poll_delete_answer",
        answer: act.answer
      };
    case "messageActionManagedBotCreated":
      return {
        type: "managed_bot_created",
        botId: act.botId
      };
    default:
      return null;
  }
}
export {
  _messageActionFromTl
};
