// @mtcute/core/highlevel/types/messages/message-reactions.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const peerReaction = require("../reactions/peer-reaction.cjs");
const reactionCount = require("../reactions/reaction-count.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class MessageReactions {
  constructor(messageId, chatId, raw, _peers) {
    this.messageId = messageId;
    this.chatId = chatId;
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether you can use {@link getUsers}
   * (or {@link TelegramClient.getReactionUsers})
   * to get the users who reacted to this message
   */
  get usersVisible() {
    return this.raw.canSeeList;
  }
  /**
   * Reactions on the message, along with their counts
   */
  get reactions() {
    return this.raw.results.map((it) => new reactionCount.ReactionCount(it));
  }
  /**
   * Recently reacted users.
   * To get a full list of users, use {@link getUsers}
   */
  get recentReactions() {
    return this.raw.recentReactions?.map((reaction) => new peerReaction.PeerReaction(reaction, this._peers)) ?? [];
  }
  /** Leaderboard of paid reactions to the message */
  get paidReactions() {
    return this.raw.topReactors?.map((reaction) => new peerReaction.PaidPeerReaction(reaction, this._peers)) ?? [];
  }
}
const MessageReactions$1 = /* @__PURE__ */ memoize.memoizeGetters(MessageReactions, ["reactions", "recentReactions"]);
const MessageReactions$2 = /* @__PURE__ */ inspectable.makeInspectable(MessageReactions$1);
exports.MessageReactions = MessageReactions$2;
