// @mtcute/core/highlevel/types/messages/replied-message.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
import { MessageEntity as MessageEntity$2 } from "./message-entity.js";
import { _messageMediaFromTl } from "./message-media.js";
class RepliedMessageInfo {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this message is a reply to another scheduled message */
  get isScheduled() {
    return this.raw.replyToScheduled;
  }
  /** Whether this message was sent to a forum topic */
  get isForumTopic() {
    return this.raw.forumTopic;
  }
  /** Whether this message is quoting another message */
  get isQuote() {
    return this.raw.quote;
  }
  /** Origin of the replied-to message */
  get origin() {
    if (!this.raw.replyToMsgId) return "private";
    if (!this.raw.replyFrom) return "same_chat";
    return "other_chat";
  }
  /**
   * Helper method to check {@link origin} that will also
   * narrow the type of this object respectively
   */
  originIs(origin) {
    if (this.origin === origin) {
      switch (origin) {
        case "same_chat":
          return true;
        case "other_chat":
          return this.raw.replyToPeerId !== void 0;
        case "private":
          return this.raw.replyFrom?.fromId !== void 0;
      }
    }
    return false;
  }
  /**
   * For non-`private` origin, ID of the replied-to message in the original chat.
   */
  get id() {
    return this.raw.replyToMsgId ?? null;
  }
  /** ID of the replies thread where this message belongs to */
  get threadId() {
    if (this.raw.forumTopic) {
      return this.raw.replyToTopId ?? this.raw.replyToMsgId ?? null;
    }
    return this.raw.replyToTopId ?? null;
  }
  /** If this is a reply to a specific todo item, ID of that item */
  get todoItemId() {
    return this.raw.todoItemId ?? null;
  }
  /** If this is a reply to a specific poll option, its ID */
  get pollOption() {
    return this.raw.pollOption ?? null;
  }
  /**
   * If replied-to message is available, chat where the message was sent.
   *
   * If `null`, the message was sent in the same chat.
   */
  get chat() {
    if (!this.raw.replyToPeerId || !this.raw.replyFrom) {
      return null;
    }
    return parsePeer(this.raw.replyToPeerId, this._peers);
  }
  /**
   * Sender of the replied-to message (either user or a channel)
   * or their name (for users with private forwards).
   *
   * For replies to channel messages, this will be the channel itself.
   *
   * `null` if the sender is not available (for `same_chat` origin)
   */
  get sender() {
    const { replyFrom, replyToPeerId } = this.raw;
    if (!replyFrom && !replyToPeerId) return null;
    if (replyFrom?.fromName) {
      return {
        type: "anonymous",
        displayName: replyFrom.fromName
      };
    }
    const peer = replyFrom?.fromId ?? replyToPeerId;
    if (peer) {
      return parsePeer(peer, this._peers);
    }
    return null;
  }
  /**
   * For non-`same_chat` origin, date the original message was sent.
   */
  get date() {
    if (!this.raw.replyFrom?.date) return null;
    return new Date(this.raw.replyFrom.date * 1e3);
  }
  /**
   * If this message is a quote, text of the quote.
   *
   * For non-`same_chat` origin, this will be the full text of the
   * replied-to message in case `.isQuote` is `false`
   */
  get quoteText() {
    return this.raw.quoteText ?? "";
  }
  /** Message entities contained in the quote */
  get quoteEntities() {
    return this.raw.quoteEntities?.map((e) => new MessageEntity$2(e)) ?? [];
  }
  /**
   * Offset of the start of the {@link quoteText} in the replied-to message.
   *
   * Note that this offset should only be used as a hint, as the actual
   * quote offset may be different due to message being edited after the quote
   *
   * `null` if not available, in which case it should be assumed that the quote
   * starts at `.indexOf(quoteText)` of the replied-to message text.
   */
  get quoteOffset() {
    if (!this.raw.quoteOffset) return null;
    return this.raw.quoteOffset;
  }
  /**
   * Media contained in the replied-to message
   *
   * Only available for non-`same_chat` origin
   */
  get media() {
    if (!this.raw.replyMedia) return null;
    return _messageMediaFromTl(this._peers, this.raw.replyMedia);
  }
}
const RepliedMessageInfo$1 = /* @__PURE__ */ memoizeGetters(RepliedMessageInfo, ["chat", "sender", "quoteEntities", "media"]);
const RepliedMessageInfo$2 = /* @__PURE__ */ makeInspectable(RepliedMessageInfo$1);
export {
  RepliedMessageInfo$2 as RepliedMessageInfo
};
