// @mtcute/core/storage/memory/repository/auth-keys.js
class MemoryAuthKeysRepository {
  constructor(_driver) {
    this._driver = _driver;
    this.state = this._driver.getState("authKeys", () => ({
      authKeys: /* @__PURE__ */ new Map(),
      authKeysTemp: /* @__PURE__ */ new Map(),
      authKeysTempExpiry: /* @__PURE__ */ new Map()
    }));
  }
  state;
  set(dc, key) {
    if (key) {
      this.state.authKeys.set(dc, key);
    } else {
      this.state.authKeys.delete(dc);
    }
  }
  get(dc) {
    return this.state.authKeys.get(dc) ?? null;
  }
  setTemp(dc, idx, key, expires) {
    const k = `${dc}:${idx}`;
    if (key) {
      this.state.authKeysTemp.set(k, key);
      this.state.authKeysTempExpiry.set(k, expires);
    } else {
      this.state.authKeysTemp.delete(k);
      this.state.authKeysTempExpiry.delete(k);
    }
  }
  getTemp(dc, idx, now) {
    const k = `${dc}:${idx}`;
    if (now > (this.state.authKeysTempExpiry.get(k) ?? 0)) {
      return null;
    }
    return this.state.authKeysTemp.get(k) ?? null;
  }
  deleteByDc(dc) {
    this.state.authKeys.delete(dc);
    for (const key of this.state.authKeysTemp.keys()) {
      if (key.startsWith(`${dc}:`)) {
        this.state.authKeysTemp.delete(key);
        this.state.authKeysTempExpiry.delete(key);
      }
    }
  }
  deleteAll() {
    this.state.authKeys.clear();
    this.state.authKeysTemp.clear();
    this.state.authKeysTempExpiry.clear();
  }
}
export {
  MemoryAuthKeysRepository
};
