// @mtcute/core/storage/memory/repository/peers.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class MemoryPeersRepository {
  constructor(_driver) {
    this._driver = _driver;
    this.state = this._driver.getState("peers", () => ({
      entities: /* @__PURE__ */ new Map(),
      usernameIndex: /* @__PURE__ */ new Map(),
      phoneIndex: /* @__PURE__ */ new Map()
    }));
  }
  state;
  store(peer) {
    const old = this.state.entities.get(peer.id);
    if (old) {
      old.usernames.forEach((username) => {
        this.state.usernameIndex.delete(username);
      });
      if (old.phone) {
        this.state.phoneIndex.delete(old.phone);
      }
    }
    if (peer.usernames) {
      for (const username of peer.usernames) {
        this.state.usernameIndex.set(username, peer.id);
      }
    }
    if (peer.phone) this.state.phoneIndex.set(peer.phone, peer.id);
    this.state.entities.set(peer.id, peer);
  }
  getById(id) {
    const ent = this.state.entities.get(id);
    if (!ent) return null;
    return ent;
  }
  getByUsername(username) {
    const id = this.state.usernameIndex.get(username.toLowerCase());
    if (!id) return null;
    const ent = this.state.entities.get(id);
    if (!ent || ent.isMin) return null;
    return ent;
  }
  getByPhone(phone) {
    const id = this.state.phoneIndex.get(phone);
    if (!id) return null;
    const ent = this.state.entities.get(id);
    if (!ent || ent.isMin) return null;
    return ent;
  }
  deleteAll() {
    this.state.entities.clear();
    this.state.phoneIndex.clear();
    this.state.usernameIndex.clear();
  }
}
exports.MemoryPeersRepository = MemoryPeersRepository;
