// @mtcute/core/storage/memory/repository/ref-messages.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class MemoryRefMessagesRepository {
  constructor(_driver) {
    this._driver = _driver;
    this.state = this._driver.getState("refMessages", () => ({
      refs: /* @__PURE__ */ new Map()
    }));
  }
  state;
  store(peerId, chatId, msgId) {
    if (!this.state.refs.has(peerId)) {
      this.state.refs.set(peerId, /* @__PURE__ */ new Set());
    }
    this.state.refs.get(peerId).add(`${chatId}:${msgId}`);
  }
  getByPeer(peerId) {
    const refs = this.state.refs.get(peerId);
    if (!refs?.size) return null;
    const [ref] = refs;
    const [chatId, msgId] = ref.split(":");
    return [Number(chatId), Number(msgId)];
  }
  delete(chatId, msgIds) {
    for (const refs of this.state.refs.values()) {
      for (const msg of msgIds) {
        refs.delete(`${chatId}:${msg}`);
      }
    }
  }
  deleteByPeer(peerId) {
    this.state.refs.delete(peerId);
  }
  deleteAll() {
    this.state.refs.clear();
  }
}
exports.MemoryRefMessagesRepository = MemoryRefMessagesRepository;
