// @mtcute/core/highlevel/types/updates/bot-reaction.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
const reactionCount = require("../reactions/reaction-count.cjs");
const types = require("../reactions/types.cjs");
class BotReactionUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat where the reaction has been changed
   */
  get chat() {
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the message where the reaction has been changed
   */
  get messageId() {
    return this.raw.msgId;
  }
  /**
   * Date when the reaction has been changed
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * ID of the user who has set/removed the reaction
   */
  get actor() {
    return peer.parsePeer(this.raw.actor, this._peers);
  }
  /**
   * List of reactions before the change
   */
  get before() {
    return this.raw.oldReactions.map((it) => types.toReactionEmoji(it));
  }
  /**
   * List of reactions after the change
   */
  get after() {
    return this.raw.newReactions.map((it) => types.toReactionEmoji(it));
  }
}
const BotReactionUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(BotReactionUpdate, ["chat", "actor", "before", "after"]);
const BotReactionUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(BotReactionUpdate$1);
class BotReactionCountUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat where the reaction has been changed
   */
  get chat() {
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the message where the reaction has been changed
   */
  get messageId() {
    return this.raw.msgId;
  }
  /**
   * Date when the reaction has been changed
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * The new list of reactions to the message
   */
  get reactions() {
    return this.raw.reactions.map((it) => new reactionCount.ReactionCount(it));
  }
}
const BotReactionCountUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(BotReactionCountUpdate, ["chat"]);
const BotReactionCountUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(BotReactionCountUpdate$1);
exports.BotReactionCountUpdate = BotReactionCountUpdate$2;
exports.BotReactionUpdate = BotReactionUpdate$2;
