// @mtcute/core/highlevel/types/updates/bot-chat-join-request.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const chatInviteLink = require("../peers/chat-invite-link.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const chat = require("../peers/chat.cjs");
const user = require("../peers/user.cjs");
class BotChatJoinRequestUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat ID where the user is requesting to join.
   */
  get chatId() {
    return peerUtils.getMarkedPeerId(this.raw.peer);
  }
  /**
   * Object containing the chat information.
   */
  get chat() {
    return new chat.Chat(this._peers.chat(peerUtils.getBarePeerId(this.raw.peer)));
  }
  /**
   * ID of the user who requested to join the chat.
   */
  get userId() {
    return this.raw.userId;
  }
  /**
   * Object containing the user information.
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * Bio of the user who requested to join the chat.
   */
  get userBio() {
    return this.raw.about;
  }
  /**
   * Date when the request was sent.
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Invite link used to request joining.
   */
  get invite() {
    return new chatInviteLink.ChatInviteLink(this.raw.invite);
  }
}
const BotChatJoinRequestUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(BotChatJoinRequestUpdate, ["chat", "user", "invite"]);
const BotChatJoinRequestUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(BotChatJoinRequestUpdate$1);
exports.BotChatJoinRequestUpdate = BotChatJoinRequestUpdate$2;
