// @mtcute/core/highlevel/types/updates/bot-chat-join-request.js
import { getMarkedPeerId, getBarePeerId } from "../../../utils/peer-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { ChatInviteLink as ChatInviteLink$2 } from "../peers/chat-invite-link.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { Chat as Chat$2 } from "../peers/chat.js";
import { User as User$2 } from "../peers/user.js";
class BotChatJoinRequestUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat ID where the user is requesting to join.
   */
  get chatId() {
    return getMarkedPeerId(this.raw.peer);
  }
  /**
   * Object containing the chat information.
   */
  get chat() {
    return new Chat$2(this._peers.chat(getBarePeerId(this.raw.peer)));
  }
  /**
   * ID of the user who requested to join the chat.
   */
  get userId() {
    return this.raw.userId;
  }
  /**
   * Object containing the user information.
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * Bio of the user who requested to join the chat.
   */
  get userBio() {
    return this.raw.about;
  }
  /**
   * Date when the request was sent.
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Invite link used to request joining.
   */
  get invite() {
    return new ChatInviteLink$2(this.raw.invite);
  }
}
const BotChatJoinRequestUpdate$1 = /* @__PURE__ */ memoizeGetters(BotChatJoinRequestUpdate, ["chat", "user", "invite"]);
const BotChatJoinRequestUpdate$2 = /* @__PURE__ */ makeInspectable(BotChatJoinRequestUpdate$1);
export {
  BotChatJoinRequestUpdate$2 as BotChatJoinRequestUpdate
};
