// @mtcute/core/highlevel/types/updates/bot-reaction.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "../peers/peer.js";
import { ReactionCount as ReactionCount$1 } from "../reactions/reaction-count.js";
import { toReactionEmoji } from "../reactions/types.js";
class BotReactionUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat where the reaction has been changed
   */
  get chat() {
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the message where the reaction has been changed
   */
  get messageId() {
    return this.raw.msgId;
  }
  /**
   * Date when the reaction has been changed
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * ID of the user who has set/removed the reaction
   */
  get actor() {
    return parsePeer(this.raw.actor, this._peers);
  }
  /**
   * List of reactions before the change
   */
  get before() {
    return this.raw.oldReactions.map((it) => toReactionEmoji(it));
  }
  /**
   * List of reactions after the change
   */
  get after() {
    return this.raw.newReactions.map((it) => toReactionEmoji(it));
  }
}
const BotReactionUpdate$1 = /* @__PURE__ */ memoizeGetters(BotReactionUpdate, ["chat", "actor", "before", "after"]);
const BotReactionUpdate$2 = /* @__PURE__ */ makeInspectable(BotReactionUpdate$1);
class BotReactionCountUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Chat where the reaction has been changed
   */
  get chat() {
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * ID of the message where the reaction has been changed
   */
  get messageId() {
    return this.raw.msgId;
  }
  /**
   * Date when the reaction has been changed
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * The new list of reactions to the message
   */
  get reactions() {
    return this.raw.reactions.map((it) => new ReactionCount$1(it));
  }
}
const BotReactionCountUpdate$1 = /* @__PURE__ */ memoizeGetters(BotReactionCountUpdate, ["chat"]);
const BotReactionCountUpdate$2 = /* @__PURE__ */ makeInspectable(BotReactionCountUpdate$1);
export {
  BotReactionCountUpdate$2 as BotReactionCountUpdate,
  BotReactionUpdate$2 as BotReactionUpdate
};
