// @mtcute/core/highlevel/types/updates/callback-query.js
import { utf8 } from "@fuman/utils";
import { MtArgumentError } from "../../../types/errors.js";
import { encodeInlineMessageId } from "../../utils/inline-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Message as Message$2 } from "../messages/message.js";
import { parsePeer } from "../peers/peer.js";
import { User as User$2 } from "../peers/user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class BaseCallbackQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * ID of this callback query
   */
  get id() {
    return this.raw.queryId;
  }
  /**
   * User who has pressed the button
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * Unique ID, that represents the chat to which the inline
   * message was sent. Does *not* contain actual chat ID.
   */
  get uniqueChatId() {
    return this.raw.chatInstance;
  }
  /**
   * Data that was contained in the callback button, if any
   *
   * Note that this field is defined by the client, and a bad
   * client can send arbitrary data in this field.
   */
  get data() {
    return this.raw.data ?? null;
  }
  /**
   * Data that was contained in the callback button, if any,
   * parsed as a UTF8 string
   *
   * Note that this field is defined by the client, and a bad
   * client can send arbitrary data in this field.
   */
  get dataStr() {
    if (!this.raw.data) return null;
    return utf8.decoder.decode(this.raw.data);
  }
  /**
   * In case this message was from {@link InputInlineResultGame},
   * or the button was {@link BotKeyboard.game},
   * short name of the game that should be returned.
   */
  get game() {
    if (this.raw._ === "updateBusinessBotCallbackQuery") return null;
    return this.raw.gameShortName ?? null;
  }
}
class CallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /**
   * Chat where the originating message was sent
   */
  get chat() {
    if (this.raw._ !== "updateBotCallbackQuery") {
      throw new MtArgumentError("Cannot get message id for inline callback");
    }
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * Identifier of the message containing the button which was clicked.
   */
  get messageId() {
    return this.raw.msgId;
  }
}
const CallbackQuery$1 = /* @__PURE__ */ memoizeGetters(CallbackQuery, ["user", "dataStr", "chat"]);
const CallbackQuery$2 = /* @__PURE__ */ makeInspectable(CallbackQuery$1);
class InlineCallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /**
   * Identifier of the previously sent inline message,
   * that contained the button which was clicked.
   * This ID can be used in `TelegramClient.editInlineMessage`
   */
  get inlineMessageId() {
    return this.raw.msgId;
  }
  /**
   * Identifier of the previously sent inline message,
   * that contained the button which was clicked,
   * as a TDLib and Bot API compatible string.
   * Can be used instead of {@link inlineMessageId} in
   * case you want to store it in some storage.
   */
  get inlineMessageIdStr() {
    return encodeInlineMessageId(this.raw.msgId);
  }
}
const InlineCallbackQuery$1 = /* @__PURE__ */ memoizeGetters(InlineCallbackQuery, ["user", "dataStr", "inlineMessageIdStr"]);
const InlineCallbackQuery$2 = /* @__PURE__ */ makeInspectable(InlineCallbackQuery$1);
class BusinessCallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /** ID of the business connection */
  get connectionId() {
    return this.raw.connectionId;
  }
  /** Message containing the button */
  get message() {
    return new Message$2(this.raw.message, this._peers);
  }
  /** Message that {@link message} is a reply to (if any) */
  get replyToMessage() {
    if (!this.raw.replyToMessage) return null;
    return new Message$2(this.raw.replyToMessage, this._peers);
  }
}
const BusinessCallbackQuery$1 = /* @__PURE__ */ memoizeGetters(BusinessCallbackQuery, ["user", "dataStr", "message", "replyToMessage"]);
const BusinessCallbackQuery$2 = /* @__PURE__ */ makeInspectable(BusinessCallbackQuery$1);
export {
  BusinessCallbackQuery$2 as BusinessCallbackQuery,
  CallbackQuery$2 as CallbackQuery,
  InlineCallbackQuery$2 as InlineCallbackQuery
};
