// @mtcute/core/highlevel/types/updates/callback-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("../../../types/errors.cjs");
const inlineUtils = require("../../utils/inline-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const message = require("../messages/message.cjs");
const peer = require("../peers/peer.cjs");
const user = require("../peers/user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class BaseCallbackQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * ID of this callback query
   */
  get id() {
    return this.raw.queryId;
  }
  /**
   * User who has pressed the button
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * Unique ID, that represents the chat to which the inline
   * message was sent. Does *not* contain actual chat ID.
   */
  get uniqueChatId() {
    return this.raw.chatInstance;
  }
  /**
   * Data that was contained in the callback button, if any
   *
   * Note that this field is defined by the client, and a bad
   * client can send arbitrary data in this field.
   */
  get data() {
    return this.raw.data ?? null;
  }
  /**
   * Data that was contained in the callback button, if any,
   * parsed as a UTF8 string
   *
   * Note that this field is defined by the client, and a bad
   * client can send arbitrary data in this field.
   */
  get dataStr() {
    if (!this.raw.data) return null;
    return utils.utf8.decoder.decode(this.raw.data);
  }
  /**
   * In case this message was from {@link InputInlineResultGame},
   * or the button was {@link BotKeyboard.game},
   * short name of the game that should be returned.
   */
  get game() {
    if (this.raw._ === "updateBusinessBotCallbackQuery") return null;
    return this.raw.gameShortName ?? null;
  }
}
class CallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /**
   * Chat where the originating message was sent
   */
  get chat() {
    if (this.raw._ !== "updateBotCallbackQuery") {
      throw new errors.MtArgumentError("Cannot get message id for inline callback");
    }
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * Identifier of the message containing the button which was clicked.
   */
  get messageId() {
    return this.raw.msgId;
  }
}
const CallbackQuery$1 = /* @__PURE__ */ memoize.memoizeGetters(CallbackQuery, ["user", "dataStr", "chat"]);
const CallbackQuery$2 = /* @__PURE__ */ inspectable.makeInspectable(CallbackQuery$1);
class InlineCallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /**
   * Identifier of the previously sent inline message,
   * that contained the button which was clicked.
   * This ID can be used in `TelegramClient.editInlineMessage`
   */
  get inlineMessageId() {
    return this.raw.msgId;
  }
  /**
   * Identifier of the previously sent inline message,
   * that contained the button which was clicked,
   * as a TDLib and Bot API compatible string.
   * Can be used instead of {@link inlineMessageId} in
   * case you want to store it in some storage.
   */
  get inlineMessageIdStr() {
    return inlineUtils.encodeInlineMessageId(this.raw.msgId);
  }
}
const InlineCallbackQuery$1 = /* @__PURE__ */ memoize.memoizeGetters(InlineCallbackQuery, ["user", "dataStr", "inlineMessageIdStr"]);
const InlineCallbackQuery$2 = /* @__PURE__ */ inspectable.makeInspectable(InlineCallbackQuery$1);
class BusinessCallbackQuery extends BaseCallbackQuery {
  constructor(raw, _peers) {
    super(raw, _peers);
    this.raw = raw;
  }
  /** ID of the business connection */
  get connectionId() {
    return this.raw.connectionId;
  }
  /** Message containing the button */
  get message() {
    return new message.Message(this.raw.message, this._peers);
  }
  /** Message that {@link message} is a reply to (if any) */
  get replyToMessage() {
    if (!this.raw.replyToMessage) return null;
    return new message.Message(this.raw.replyToMessage, this._peers);
  }
}
const BusinessCallbackQuery$1 = /* @__PURE__ */ memoize.memoizeGetters(BusinessCallbackQuery, ["user", "dataStr", "message", "replyToMessage"]);
const BusinessCallbackQuery$2 = /* @__PURE__ */ inspectable.makeInspectable(BusinessCallbackQuery$1);
exports.BusinessCallbackQuery = BusinessCallbackQuery$2;
exports.CallbackQuery = CallbackQuery$2;
exports.InlineCallbackQuery = InlineCallbackQuery$2;
