// @mtcute/core/highlevel/types/updates/chat-join-request.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const user = require("../peers/user.cjs");
class ChatJoinRequestUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  // in this update, peers index only contains
  // recent requesters, not the chat
  /**
   * Marked ID of the chat/channel
   */
  get chatId() {
    return peerUtils.getMarkedPeerId(this.raw.peer);
  }
  /**
   * IDs of the users who recently requested to join the chat
   */
  get recentRequestersIds() {
    return this.raw.recentRequesters;
  }
  /**
   * Users who recently requested to join the chat
   */
  get recentRequesters() {
    return this.raw.recentRequesters.map((id) => new user.User(this._peers.user(id)));
  }
  /**
   * Total number of pending requests
   */
  get totalPending() {
    return this.raw.requestsPending;
  }
}
const ChatJoinRequestUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatJoinRequestUpdate, ["recentRequesters"]);
const ChatJoinRequestUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatJoinRequestUpdate$1);
exports.ChatJoinRequestUpdate = ChatJoinRequestUpdate$2;
