// @mtcute/core/highlevel/types/updates/chat-member-update.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { ChatInviteLink as ChatInviteLink$2 } from "../peers/chat-invite-link.js";
import { ChatMember as ChatMember$2 } from "../peers/chat-member.js";
import { Chat as Chat$2 } from "../peers/chat.js";
import { User as User$2 } from "../peers/user.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { isAnyChatParticipant } from "../../../tl/inner-tl.js";
function extractPeerId(raw) {
  if (!raw) return 0;
  if (isAnyChatParticipant(raw)) {
    return raw.userId;
  }
  switch (raw._) {
    case "channelParticipant":
    case "channelParticipantSelf":
    case "channelParticipantCreator":
    case "channelParticipantAdmin":
      return raw.userId;
    default:
      return getMarkedPeerId(raw.peer);
  }
}
class ChatMemberUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Date of the event
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this is an update about current user
   */
  get isSelf() {
    return this.user.isSelf;
  }
  /**
   * Type of the update
   *
   * @link ChatMemberUpdate.Type
   */
  get type() {
    const old = this.raw.prevParticipant;
    const cur = this.raw.newParticipant;
    const oldId = extractPeerId(old);
    const curId = extractPeerId(cur);
    const actorId = this.raw.actorId;
    if (!old && cur) {
      if (cur._ === "channelParticipantBanned") return "kicked";
      return actorId === curId ? "joined" : "added";
    }
    if (old && !cur) {
      if (actorId === oldId) return "left";
      if (old._ === "channelParticipantBanned") {
        return "unkicked";
      }
      return "kicked";
    }
    if (!old || !cur) return "other";
    switch (old._) {
      case "chatParticipant":
      case "channelParticipant":
        switch (cur._) {
          case "chatParticipantAdmin":
          case "channelParticipantAdmin":
            return "promoted";
          case "channelParticipantBanned":
            if (cur.left) return "kicked";
            return "restricted";
        }
        break;
      case "chatParticipantCreator":
      case "channelParticipantCreator":
        return "old_owner";
    }
    switch (cur._) {
      case "chatParticipantCreator":
      case "channelParticipantCreator":
        return "new_owner";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipant") {
      return "unrestricted";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipantAdmin") {
      return "unrestricted_promoted";
    }
    if (old._ === "channelParticipantAdmin" && cur._ === "channelParticipant") {
      return "demoted";
    }
    if (old._ === "channelParticipantAdmin" && cur._ === "channelParticipantBanned") {
      return cur.left ? "demoted_kicked" : "demoted_restricted";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipantBanned" && old.left !== cur.left) {
      if (actorId === curId) {
        return cur.left ? "left" : "joined";
      }
      return cur.left ? "kicked" : "added";
    }
    return "other";
  }
  /**
   * Chat in which this event has occurred
   */
  get chat() {
    const id = this.raw._ === "updateChannelParticipant" ? this.raw.channelId : this.raw.chatId;
    return new Chat$2(this._peers.chat(id));
  }
  /**
   * Performer of the action which resulted in this update.
   *
   * Can be chat/channel administrator or the {@link user} themself.
   */
  get actor() {
    return new User$2(this._peers.user(this.raw.actorId));
  }
  /**
   * User representing the chat member whose status was changed.
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /** Whether this is a self-made action (i.e. actor == user) */
  get isSelfMade() {
    return this.raw.actorId === this.raw.userId;
  }
  /**
   * Previous (old) information about chat member.
   */
  get oldMember() {
    if (!this.raw.prevParticipant) return null;
    return new ChatMember$2(this.raw.prevParticipant, this._peers);
  }
  /**
   * Current (new) information about chat member.
   */
  get newMember() {
    if (!this.raw.newParticipant) return null;
    return new ChatMember$2(this.raw.newParticipant, this._peers);
  }
  /**
   * In case this is a "join" event, invite link that was used to join (if any)
   */
  get inviteLink() {
    if (!this.raw.invite) return null;
    return new ChatInviteLink$2(this.raw.invite);
  }
}
const ChatMemberUpdate$1 = /* @__PURE__ */ memoizeGetters(ChatMemberUpdate, ["type", "chat", "actor", "user", "oldMember", "newMember", "inviteLink"]);
const ChatMemberUpdate$2 = /* @__PURE__ */ makeInspectable(ChatMemberUpdate$1);
export {
  ChatMemberUpdate$2 as ChatMemberUpdate
};
