// @mtcute/core/highlevel/types/updates/chat-join-request.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { User as User$2 } from "../peers/user.js";
class ChatJoinRequestUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  // in this update, peers index only contains
  // recent requesters, not the chat
  /**
   * Marked ID of the chat/channel
   */
  get chatId() {
    return getMarkedPeerId(this.raw.peer);
  }
  /**
   * IDs of the users who recently requested to join the chat
   */
  get recentRequestersIds() {
    return this.raw.recentRequesters;
  }
  /**
   * Users who recently requested to join the chat
   */
  get recentRequesters() {
    return this.raw.recentRequesters.map((id) => new User$2(this._peers.user(id)));
  }
  /**
   * Total number of pending requests
   */
  get totalPending() {
    return this.raw.requestsPending;
  }
}
const ChatJoinRequestUpdate$1 = /* @__PURE__ */ memoizeGetters(ChatJoinRequestUpdate, ["recentRequesters"]);
const ChatJoinRequestUpdate$2 = /* @__PURE__ */ makeInspectable(ChatJoinRequestUpdate$1);
export {
  ChatJoinRequestUpdate$2 as ChatJoinRequestUpdate
};
