// @mtcute/core/highlevel/types/updates/chat-member-update.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const chatInviteLink = require("../peers/chat-invite-link.cjs");
const chatMember = require("../peers/chat-member.cjs");
const chat = require("../peers/chat.cjs");
const user = require("../peers/user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
function extractPeerId(raw) {
  if (!raw) return 0;
  if (innerTl.isAnyChatParticipant(raw)) {
    return raw.userId;
  }
  switch (raw._) {
    case "channelParticipant":
    case "channelParticipantSelf":
    case "channelParticipantCreator":
    case "channelParticipantAdmin":
      return raw.userId;
    default:
      return peerUtils.getMarkedPeerId(raw.peer);
  }
}
class ChatMemberUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Date of the event
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this is an update about current user
   */
  get isSelf() {
    return this.user.isSelf;
  }
  /**
   * Type of the update
   *
   * @link ChatMemberUpdate.Type
   */
  get type() {
    const old = this.raw.prevParticipant;
    const cur = this.raw.newParticipant;
    const oldId = extractPeerId(old);
    const curId = extractPeerId(cur);
    const actorId = this.raw.actorId;
    if (!old && cur) {
      if (cur._ === "channelParticipantBanned") return "kicked";
      return actorId === curId ? "joined" : "added";
    }
    if (old && !cur) {
      if (actorId === oldId) return "left";
      if (old._ === "channelParticipantBanned") {
        return "unkicked";
      }
      return "kicked";
    }
    if (!old || !cur) return "other";
    switch (old._) {
      case "chatParticipant":
      case "channelParticipant":
        switch (cur._) {
          case "chatParticipantAdmin":
          case "channelParticipantAdmin":
            return "promoted";
          case "channelParticipantBanned":
            if (cur.left) return "kicked";
            return "restricted";
        }
        break;
      case "chatParticipantCreator":
      case "channelParticipantCreator":
        return "old_owner";
    }
    switch (cur._) {
      case "chatParticipantCreator":
      case "channelParticipantCreator":
        return "new_owner";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipant") {
      return "unrestricted";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipantAdmin") {
      return "unrestricted_promoted";
    }
    if (old._ === "channelParticipantAdmin" && cur._ === "channelParticipant") {
      return "demoted";
    }
    if (old._ === "channelParticipantAdmin" && cur._ === "channelParticipantBanned") {
      return cur.left ? "demoted_kicked" : "demoted_restricted";
    }
    if (old._ === "channelParticipantBanned" && cur._ === "channelParticipantBanned" && old.left !== cur.left) {
      if (actorId === curId) {
        return cur.left ? "left" : "joined";
      }
      return cur.left ? "kicked" : "added";
    }
    return "other";
  }
  /**
   * Chat in which this event has occurred
   */
  get chat() {
    const id = this.raw._ === "updateChannelParticipant" ? this.raw.channelId : this.raw.chatId;
    return new chat.Chat(this._peers.chat(id));
  }
  /**
   * Performer of the action which resulted in this update.
   *
   * Can be chat/channel administrator or the {@link user} themself.
   */
  get actor() {
    return new user.User(this._peers.user(this.raw.actorId));
  }
  /**
   * User representing the chat member whose status was changed.
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /** Whether this is a self-made action (i.e. actor == user) */
  get isSelfMade() {
    return this.raw.actorId === this.raw.userId;
  }
  /**
   * Previous (old) information about chat member.
   */
  get oldMember() {
    if (!this.raw.prevParticipant) return null;
    return new chatMember.ChatMember(this.raw.prevParticipant, this._peers);
  }
  /**
   * Current (new) information about chat member.
   */
  get newMember() {
    if (!this.raw.newParticipant) return null;
    return new chatMember.ChatMember(this.raw.newParticipant, this._peers);
  }
  /**
   * In case this is a "join" event, invite link that was used to join (if any)
   */
  get inviteLink() {
    if (!this.raw.invite) return null;
    return new chatInviteLink.ChatInviteLink(this.raw.invite);
  }
}
const ChatMemberUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatMemberUpdate, ["type", "chat", "actor", "user", "oldMember", "newMember", "inviteLink"]);
const ChatMemberUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatMemberUpdate$1);
exports.ChatMemberUpdate = ChatMemberUpdate$2;
