// @mtcute/core/highlevel/types/updates/chosen-inline-result.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inlineUtils = require("../../utils/inline-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const location = require("../media/location.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const user = require("../peers/user.cjs");
class ChosenInlineResult {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique identifier of the chosen result,
   * as set in `InputInlineResult.id`
   */
  get id() {
    return this.raw.id;
  }
  /**
   * User who has chosen the query
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * The query that was previously sent by the user,
   * which was used to obtain this result
   */
  get query() {
    return this.raw.query;
  }
  /**
   * Sender location, only applicable to bots that requested user location
   */
  get location() {
    if (this.raw.geo?._ !== "geoPoint") return null;
    return new location.Location(this.raw.geo);
  }
  /**
   * Identifier of the sent inline message,
   * which can be used in `TelegramClient.editInlineMessage`
   *
   * > **Note**: this is only available in case the `InputInlineMessage`
   * > contained a reply keyboard markup.
   */
  get messageId() {
    return this.raw.msgId ?? null;
  }
  /**
   * Identifier of the sent inline message
   * as a TDLib and Bot API compatible string.
   * Can be used instead of {@link messageId} in
   * case you want to store it in some storage.
   *
   * > **Note**: this is only available in case the `InputInlineMessage`
   * > contained a reply keyboard markup.
   */
  get messageIdStr() {
    if (!this.raw.msgId) return null;
    return inlineUtils.encodeInlineMessageId(this.raw.msgId);
  }
}
const ChosenInlineResult$1 = /* @__PURE__ */ memoize.memoizeGetters(ChosenInlineResult, ["user", "location"]);
const ChosenInlineResult$2 = /* @__PURE__ */ inspectable.makeInspectable(ChosenInlineResult$1);
exports.ChosenInlineResult = ChosenInlineResult$2;
