// @mtcute/core/highlevel/types/updates/chosen-inline-result.js
import { encodeInlineMessageId } from "../../utils/inline-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Location as Location$1 } from "../media/location.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { User as User$2 } from "../peers/user.js";
class ChosenInlineResult {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique identifier of the chosen result,
   * as set in `InputInlineResult.id`
   */
  get id() {
    return this.raw.id;
  }
  /**
   * User who has chosen the query
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * The query that was previously sent by the user,
   * which was used to obtain this result
   */
  get query() {
    return this.raw.query;
  }
  /**
   * Sender location, only applicable to bots that requested user location
   */
  get location() {
    if (this.raw.geo?._ !== "geoPoint") return null;
    return new Location$1(this.raw.geo);
  }
  /**
   * Identifier of the sent inline message,
   * which can be used in `TelegramClient.editInlineMessage`
   *
   * > **Note**: this is only available in case the `InputInlineMessage`
   * > contained a reply keyboard markup.
   */
  get messageId() {
    return this.raw.msgId ?? null;
  }
  /**
   * Identifier of the sent inline message
   * as a TDLib and Bot API compatible string.
   * Can be used instead of {@link messageId} in
   * case you want to store it in some storage.
   *
   * > **Note**: this is only available in case the `InputInlineMessage`
   * > contained a reply keyboard markup.
   */
  get messageIdStr() {
    if (!this.raw.msgId) return null;
    return encodeInlineMessageId(this.raw.msgId);
  }
}
const ChosenInlineResult$1 = /* @__PURE__ */ memoizeGetters(ChosenInlineResult, ["user", "location"]);
const ChosenInlineResult$2 = /* @__PURE__ */ makeInspectable(ChosenInlineResult$1);
export {
  ChosenInlineResult$2 as ChosenInlineResult
};
