// @mtcute/core/highlevel/types/updates/history-read-update.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class HistoryReadUpdate {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Whether this update is about an "outbox" read
   * (i.e. a message you have sent earlier was read)
   */
  get isOutbox() {
    switch (this.raw._) {
      case "updateReadChannelDiscussionOutbox":
      case "updateReadHistoryOutbox":
      case "updateReadChannelOutbox":
        return true;
      default:
        return false;
    }
  }
  /**
   * Whether this update is about messages in a thread
   * (e.g. a comments thread or a topic in a forum)
   */
  get isDiscussion() {
    switch (this.raw._) {
      case "updateReadChannelDiscussionOutbox":
      case "updateReadChannelDiscussionInbox":
        return true;
      default:
        return false;
    }
  }
  /**
   * Whether this update is about messages in a "monoforum"
   * (i.e. a chat used for "direct channel messages")
   */
  get isMonoForum() {
    switch (this.raw._) {
      case "updateReadMonoForumOutbox":
      case "updateReadMonoForumInbox":
        return true;
      default:
        return false;
    }
  }
  /**
   * Marked peer ID of the chat where the messages were read.
   */
  get chatId() {
    switch (this.raw._) {
      case "updateReadHistoryOutbox":
      case "updateReadHistoryInbox":
        return peerUtils.getMarkedPeerId(this.raw.peer);
      case "updateReadChannelOutbox":
      case "updateReadChannelInbox":
      case "updateReadChannelDiscussionOutbox":
      case "updateReadChannelDiscussionInbox":
      case "updateReadMonoForumOutbox":
      case "updateReadMonoForumInbox":
        return peerUtils.toggleChannelIdMark(this.raw.channelId);
    }
  }
  /**
   * For inbox updates (i.e. `isOutbox = false`),
   * number of messages that are still unread in the chat.
   *
   * For other updates, `0`
   */
  get unreadCount() {
    switch (this.raw._) {
      case "updateReadHistoryInbox":
      case "updateReadChannelInbox":
        return this.raw.stillUnreadCount;
      // `updateReadChannelDiscussionInbox` does not contain that data for some reason
      case "updateReadChannelDiscussionInbox":
      case "updateReadHistoryOutbox":
      case "updateReadChannelOutbox":
      case "updateReadChannelDiscussionOutbox":
      case "updateReadMonoForumOutbox":
      case "updateReadMonoForumInbox":
        return 0;
    }
  }
  /**
   * ID of the last read message.
   *
   * Note that if `isDiscussion == true`, this contains the ID of the
   * last read message inside that thread, and not in the group itself.
   */
  get maxReadId() {
    switch (this.raw._) {
      case "updateReadHistoryOutbox":
      case "updateReadHistoryInbox":
      case "updateReadChannelOutbox":
      case "updateReadChannelInbox":
        return this.raw.maxId;
      case "updateReadChannelDiscussionOutbox":
      case "updateReadChannelDiscussionInbox":
      case "updateReadMonoForumInbox":
      case "updateReadMonoForumOutbox":
        return this.raw.readMaxId;
    }
  }
  /**
   * ID of the thread/topic (i.e. ID of the top message).
   *
   * For non-thread updates, 0.
   */
  get threadId() {
    switch (this.raw._) {
      case "updateReadHistoryOutbox":
      case "updateReadChannelOutbox":
      case "updateReadChannelInbox":
      case "updateReadMonoForumInbox":
      case "updateReadMonoForumOutbox":
        return 0;
      case "updateReadHistoryInbox":
        return this.raw.topMsgId ?? 0;
      case "updateReadChannelDiscussionOutbox":
      case "updateReadChannelDiscussionInbox":
        return this.raw.topMsgId;
    }
  }
}
const HistoryReadUpdate$1 = /* @__PURE__ */ inspectable.makeInspectable(HistoryReadUpdate);
exports.HistoryReadUpdate = HistoryReadUpdate$1;
