// @mtcute/core/highlevel/types/updates/inline-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const location = require("../media/location.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const user = require("../peers/user.cjs");
const PEER_TYPE_MAP = {
  inlineQueryPeerTypeBroadcast: "channel",
  inlineQueryPeerTypeChat: "group",
  inlineQueryPeerTypeMegagroup: "supergroup",
  inlineQueryPeerTypePM: "user",
  inlineQueryPeerTypeSameBotPM: "bot",
  inlineQueryPeerTypeBotPM: "bot"
};
class InlineQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique query ID
   */
  get id() {
    return this.raw.queryId;
  }
  /**
   * User who sent this query
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * Text of the query (0-512 characters)
   */
  get query() {
    return this.raw.query;
  }
  /**
   * Attached geolocation.
   *
   * Only used in case the bot requested user location
   */
  get location() {
    if (this.raw.geo?._ !== "geoPoint") return null;
    return new location.Location(this.raw.geo);
  }
  /**
   * Inline query scroll offset, controlled by the bot
   */
  get offset() {
    return this.raw.offset;
  }
  /**
   * Peer type from which this query was sent.
   *
   * Can be:
   *  - `bot`: Query was sent in this bot's PM
   *  - `user`: Query was sent in somebody's PM
   *  - `group`: Query was sent in a legacy group
   *  - `supergroup`: Query was sent in a supergroup
   *  - `channel`: Query was sent in a channel
   *  - `null`, in case this information is not available
   */
  get peerType() {
    return this.raw.peerType ? PEER_TYPE_MAP[this.raw.peerType._] : null;
  }
}
const InlineQuery$1 = /* @__PURE__ */ memoize.memoizeGetters(InlineQuery, ["user", "location"]);
const InlineQuery$2 = /* @__PURE__ */ inspectable.makeInspectable(InlineQuery$1);
exports.InlineQuery = InlineQuery$2;
