// @mtcute/core/highlevel/types/updates/inline-query.js
import { memoizeGetters } from "../../utils/memoize.js";
import { Location as Location$1 } from "../media/location.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { User as User$2 } from "../peers/user.js";
const PEER_TYPE_MAP = {
  inlineQueryPeerTypeBroadcast: "channel",
  inlineQueryPeerTypeChat: "group",
  inlineQueryPeerTypeMegagroup: "supergroup",
  inlineQueryPeerTypePM: "user",
  inlineQueryPeerTypeSameBotPM: "bot",
  inlineQueryPeerTypeBotPM: "bot"
};
class InlineQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique query ID
   */
  get id() {
    return this.raw.queryId;
  }
  /**
   * User who sent this query
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * Text of the query (0-512 characters)
   */
  get query() {
    return this.raw.query;
  }
  /**
   * Attached geolocation.
   *
   * Only used in case the bot requested user location
   */
  get location() {
    if (this.raw.geo?._ !== "geoPoint") return null;
    return new Location$1(this.raw.geo);
  }
  /**
   * Inline query scroll offset, controlled by the bot
   */
  get offset() {
    return this.raw.offset;
  }
  /**
   * Peer type from which this query was sent.
   *
   * Can be:
   *  - `bot`: Query was sent in this bot's PM
   *  - `user`: Query was sent in somebody's PM
   *  - `group`: Query was sent in a legacy group
   *  - `supergroup`: Query was sent in a supergroup
   *  - `channel`: Query was sent in a channel
   *  - `null`, in case this information is not available
   */
  get peerType() {
    return this.raw.peerType ? PEER_TYPE_MAP[this.raw.peerType._] : null;
  }
}
const InlineQuery$1 = /* @__PURE__ */ memoizeGetters(InlineQuery, ["user", "location"]);
const InlineQuery$2 = /* @__PURE__ */ makeInspectable(InlineQuery$1);
export {
  InlineQuery$2 as InlineQuery
};
