// @mtcute/core/highlevel/types/updates/poll-update.js
import Long from "long";
import { memoizeGetters } from "../../utils/memoize.js";
import { Poll as Poll$2 } from "../media/poll.js";
import { makeInspectable } from "../../utils/inspectable.js";
class PollUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique poll ID
   */
  get pollId() {
    return this.raw.pollId;
  }
  /**
   * Whether this is a shortened version of update, not containing the poll itself.
   */
  get isShort() {
    return this.raw.poll === void 0;
  }
  /**
   * The poll.
   *
   * When {@link isShort} is set, mtcute creates a stub poll
   * with empty question, answers and flags
   * (like `quiz`, `public`, etc.)
   *
   * If you need access to them, you should
   * map the {@link pollId} with full poll on your side
   * (e.g. in a database) and fetch from there.
   *
   * Bot API and TDLib do basically the same internally,
   * and thus are able to always provide them,
   * but mtcute currently does not have a way to do that.
   */
  get poll() {
    let poll = this.raw.poll;
    if (!poll) {
      poll = {
        _: "poll",
        id: this.raw.pollId,
        question: { _: "textWithEntities", text: "", entities: [] },
        answers: this.raw.results.results?.map((res) => ({
          _: "pollAnswer",
          text: {
            _: "textWithEntities",
            text: "",
            entities: []
          },
          option: res.option
        })) ?? [],
        hash: Long.ZERO
      };
    }
    return new Poll$2(poll, this._peers, this.raw.results);
  }
}
const PollUpdate$1 = /* @__PURE__ */ memoizeGetters(PollUpdate, ["poll"]);
const PollUpdate$2 = /* @__PURE__ */ makeInspectable(PollUpdate$1);
export {
  PollUpdate$2 as PollUpdate
};
