// @mtcute/core/highlevel/types/updates/poll-update.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const memoize = require("../../utils/memoize.cjs");
const poll = require("../media/poll.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class PollUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique poll ID
   */
  get pollId() {
    return this.raw.pollId;
  }
  /**
   * Whether this is a shortened version of update, not containing the poll itself.
   */
  get isShort() {
    return this.raw.poll === void 0;
  }
  /**
   * The poll.
   *
   * When {@link isShort} is set, mtcute creates a stub poll
   * with empty question, answers and flags
   * (like `quiz`, `public`, etc.)
   *
   * If you need access to them, you should
   * map the {@link pollId} with full poll on your side
   * (e.g. in a database) and fetch from there.
   *
   * Bot API and TDLib do basically the same internally,
   * and thus are able to always provide them,
   * but mtcute currently does not have a way to do that.
   */
  get poll() {
    let poll$1 = this.raw.poll;
    if (!poll$1) {
      poll$1 = {
        _: "poll",
        id: this.raw.pollId,
        question: { _: "textWithEntities", text: "", entities: [] },
        answers: this.raw.results.results?.map((res) => ({
          _: "pollAnswer",
          text: {
            _: "textWithEntities",
            text: "",
            entities: []
          },
          option: res.option
        })) ?? [],
        hash: Long.ZERO
      };
    }
    return new poll.Poll(poll$1, this._peers, this.raw.results);
  }
}
const PollUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(PollUpdate, ["poll"]);
const PollUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(PollUpdate$1);
exports.PollUpdate = PollUpdate$2;
