// @mtcute/core/highlevel/types/updates/poll-vote.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const memoize = require("../../utils/memoize.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const peer = require("../peers/peer.cjs");
class PollVoteUpdate {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Unique poll ID
   */
  get pollId() {
    return this.raw.pollId;
  }
  /**
   * Peer who has voted
   */
  get peer() {
    return peer.parsePeer(this.raw.peer, this._peers);
  }
  /**
   * Answers that the user has chosen.
   *
   * Note that due to incredible Telegram APIs, you
   * have to have the poll cached to be able to properly
   * tell which answers were chosen, since in the API
   * there are just arbitrary `Buffer`s, which are
   * defined by the client.
   *
   * However, most of the major implementations
   * (tested with TDLib and Bot API, official apps
   * for Android, Desktop, iOS/macOS) and mtcute
   * (by default) create `option` as a one-byte `Buffer`,
   * incrementing from `48` (ASCII `0`) up to `57` (ASCII `9`),
   * and ASCII representation would define index in the array.
   * Meaning, if `chosen[0][0] === 48` or `chosen[0].toString() === '0'`,
   * then the first answer (indexed with `0`) was chosen. To get the index,
   * you simply subtract `48` from the first byte.
   *
   * This might break at any time, but seems to be consistent for now.
   * To get chosen answer indexes derived as before, use {@link chosenIndexesAuto}.
   */
  get chosen() {
    return this.raw.options;
  }
  /**
   * Indexes of the chosen answers, derived based on observations
   * described in {@link chosen}.
   * This might break at any time, but seems to be consistent for now.
   *
   * If something does not add up, {@link MtUnsupportedError} is thrown
   */
  get chosenIndexesAuto() {
    return this.raw.options.map((buf) => {
      if (buf.length > 1) {
        throw new errors.MtUnsupportedError("option had >1 byte");
      }
      if (buf[0] < 48 || buf[0] > 57) {
        throw new errors.MtUnsupportedError("option had first byte out of 0-9 range");
      }
      return buf[0] - 48;
    });
  }
}
const PollVoteUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(PollVoteUpdate, ["peer"]);
const PollVoteUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(PollVoteUpdate$1);
exports.PollVoteUpdate = PollVoteUpdate$2;
