// @mtcute/core/highlevel/types/updates/pre-checkout-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const user = require("../peers/user.cjs");
class PreCheckoutQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * ID of the query
   */
  get queryId() {
    return this.raw.queryId;
  }
  /**
   * ID of the user who sent the query
   */
  get userId() {
    return this.raw.userId;
  }
  /**
   * User who sent the query
   */
  get user() {
    return new user.User(this._peers.user(this.userId));
  }
  /**
   * Bot-defined payload of the original invoice
   * (see {@link InputMediaInvoice.payload})
   */
  get payload() {
    return this.raw.payload;
  }
  /**
   * User-provided payment info (like name, phone, shipping address, etc.)
   */
  get paymentInfo() {
    if (!this.raw.info) return null;
    return this.raw.info;
  }
  /**
   * Currency of the payment
   */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Total price of the payment
   */
  get totalAmount() {
    return this.raw.totalAmount;
  }
}
const PreCheckoutQuery$1 = /* @__PURE__ */ memoize.memoizeGetters(PreCheckoutQuery, ["user"]);
const PreCheckoutQuery$2 = /* @__PURE__ */ inspectable.makeInspectable(PreCheckoutQuery$1);
exports.PreCheckoutQuery = PreCheckoutQuery$2;
