// @mtcute/core/highlevel/types/updates/pre-checkout-query.js
import { memoizeGetters } from "../../utils/memoize.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { User as User$2 } from "../peers/user.js";
class PreCheckoutQuery {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * ID of the query
   */
  get queryId() {
    return this.raw.queryId;
  }
  /**
   * ID of the user who sent the query
   */
  get userId() {
    return this.raw.userId;
  }
  /**
   * User who sent the query
   */
  get user() {
    return new User$2(this._peers.user(this.userId));
  }
  /**
   * Bot-defined payload of the original invoice
   * (see {@link InputMediaInvoice.payload})
   */
  get payload() {
    return this.raw.payload;
  }
  /**
   * User-provided payment info (like name, phone, shipping address, etc.)
   */
  get paymentInfo() {
    if (!this.raw.info) return null;
    return this.raw.info;
  }
  /**
   * Currency of the payment
   */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Total price of the payment
   */
  get totalAmount() {
    return this.raw.totalAmount;
  }
}
const PreCheckoutQuery$1 = /* @__PURE__ */ memoizeGetters(PreCheckoutQuery, ["user"]);
const PreCheckoutQuery$2 = /* @__PURE__ */ makeInspectable(PreCheckoutQuery$1);
export {
  PreCheckoutQuery$2 as PreCheckoutQuery
};
