// @mtcute/core/highlevel/types/updates/user-status-update.js
import { memoizeGetters } from "../../utils/memoize.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { User as User$2 } from "../peers/user.js";
class UserStatusUpdate {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * ID of the user whose status has updated
   */
  get userId() {
    return this.raw.userId;
  }
  get _parsedStatus() {
    return User$2.parseStatus(this.raw.status);
  }
  /**
   * User's new Last Seen & Online status
   */
  get status() {
    return this._parsedStatus.status;
  }
  /**
   * Last time this user was seen online.
   * Only available if {@link status} is `offline`
   */
  get lastOnline() {
    return this._parsedStatus.lastOnline;
  }
  /**
   * Time when this user will automatically go offline.
   * Only available if {@link status} is `online`
   */
  get nextOffline() {
    return this._parsedStatus.nextOffline;
  }
}
const UserStatusUpdate$1 = /* @__PURE__ */ memoizeGetters(UserStatusUpdate, ["_parsedStatus"]);
const UserStatusUpdate$2 = /* @__PURE__ */ makeInspectable(UserStatusUpdate$1, void 0, ["_parsedStatus"]);
export {
  UserStatusUpdate$2 as UserStatusUpdate
};
