// @mtcute/core/highlevel/types/updates/user-typing-update.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("../../../types/utils.cjs");
const peerUtils = require("../../../utils/peer-utils.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class UserTypingUpdate {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * ID of the user whose typing status changed
   */
  get userId() {
    return this.raw._ === "updateUserTyping" ? this.raw.userId : peerUtils.getBarePeerId(this.raw.fromId);
  }
  /**
   * Marked ID of the chat where the user is typing,
   *
   * If the user is typing in PMs, this will
   * equal to {@link userId}
   */
  get chatId() {
    switch (this.raw._) {
      case "updateUserTyping":
        return this.raw.userId;
      case "updateChatUserTyping":
        return -this.raw.chatId;
      case "updateChannelUserTyping":
        return peerUtils.toggleChannelIdMark(this.raw.channelId);
    }
  }
  /** ID of the thread/topic where the user is typing, if applicable */
  get threadId() {
    switch (this.raw._) {
      case "updateUserTyping":
        return this.raw.topMsgId ?? null;
      case "updateChatUserTyping":
        return null;
      case "updateChannelUserTyping":
        return this.raw.topMsgId ?? null;
    }
  }
  /**
   * Type of the chat where this event has occurred
   */
  get chatType() {
    switch (this.raw._) {
      case "updateUserTyping":
        return "user";
      case "updateChatUserTyping":
        return "chat";
      case "updateChannelUserTyping":
        return "channel";
    }
  }
  /**
   * Current typing status
   */
  get status() {
    switch (this.raw.action._) {
      case "sendMessageTypingAction":
      case "sendMessageTextDraftAction":
      case "inputSendMessageRichMessageDraftAction":
      case "sendMessageRichMessageDraftAction":
        return "typing";
      case "sendMessageCancelAction":
        return "cancel";
      case "sendMessageRecordVideoAction":
        return "record_video";
      case "sendMessageUploadVideoAction":
        return "upload_video";
      case "sendMessageRecordAudioAction":
        return "record_voice";
      case "sendMessageUploadAudioAction":
        return "upload_voice";
      case "sendMessageUploadPhotoAction":
        return "upload_photo";
      case "sendMessageUploadDocumentAction":
        return "upload_document";
      case "sendMessageGeoLocationAction":
        return "geo";
      case "sendMessageGamePlayAction":
        return "game";
      case "sendMessageChooseContactAction":
        return "contact";
      case "sendMessageRecordRoundAction":
        return "record_round";
      case "sendMessageUploadRoundAction":
        return "upload_round";
      case "speakingInGroupCallAction":
        return "speak_call";
      case "sendMessageHistoryImportAction":
        return "history_import";
      case "sendMessageChooseStickerAction":
        return "sticker";
      case "sendMessageEmojiInteraction":
        return "interaction";
      case "sendMessageEmojiInteractionSeen":
        return "interaction_seen";
      default:
        utils.assertNever(this.raw.action);
    }
  }
}
const UserTypingUpdate$1 = /* @__PURE__ */ inspectable.makeInspectable(UserTypingUpdate);
exports.UserTypingUpdate = UserTypingUpdate$1;
