// @mtcute/core/highlevel/types/updates/user-status-update.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const user = require("../peers/user.cjs");
class UserStatusUpdate {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * ID of the user whose status has updated
   */
  get userId() {
    return this.raw.userId;
  }
  get _parsedStatus() {
    return user.User.parseStatus(this.raw.status);
  }
  /**
   * User's new Last Seen & Online status
   */
  get status() {
    return this._parsedStatus.status;
  }
  /**
   * Last time this user was seen online.
   * Only available if {@link status} is `offline`
   */
  get lastOnline() {
    return this._parsedStatus.lastOnline;
  }
  /**
   * Time when this user will automatically go offline.
   * Only available if {@link status} is `online`
   */
  get nextOffline() {
    return this._parsedStatus.nextOffline;
  }
}
const UserStatusUpdate$1 = /* @__PURE__ */ memoize.memoizeGetters(UserStatusUpdate, ["_parsedStatus"]);
const UserStatusUpdate$2 = /* @__PURE__ */ inspectable.makeInspectable(UserStatusUpdate$1, void 0, ["_parsedStatus"]);
exports.UserStatusUpdate = UserStatusUpdate$2;
