// @mtcute/core/highlevel/utils/convert-file-id.js
import { parseFileId, tdFileId } from "@mtcute/file-id";
import Long from "long";
import { assertNever } from "../../types/utils.js";
import { parseMarkedPeerId } from "../../utils/peer-utils.js";
const FileType = tdFileId.FileType;
const EMPTY_BUFFER = new Uint8Array(0);
function dialogPhotoToInputPeer(dialog) {
  const markedPeerId = dialog.id;
  const [peerType, peerId] = parseMarkedPeerId(markedPeerId);
  if (peerType === "user") {
    return {
      _: "inputPeerUser",
      userId: peerId,
      accessHash: dialog.accessHash
    };
  } else if (peerType === "chat") {
    return {
      _: "inputPeerChat",
      chatId: peerId
    };
  }
  return {
    _: "inputPeerChannel",
    channelId: peerId,
    accessHash: dialog.accessHash
  };
}
function fileIdToInputWebFileLocation(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  if (fileId.location._ !== "web") {
    throw new tdFileId.ConversionError("inputWebFileLocation");
  }
  return {
    _: "inputWebFileLocation",
    url: fileId.location.url,
    accessHash: fileId.location.accessHash
  };
}
function fileIdToInputFileLocation(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  const loc = fileId.location;
  switch (loc._) {
    case "web":
      throw new tdFileId.ConversionError("InputFileLocation");
    case "photo": {
      switch (loc.source._) {
        case "legacy":
          if (!fileId.fileReference) {
            throw new tdFileId.InvalidFileIdError("Expected legacy photo to have file reference");
          }
          return {
            _: "inputPhotoLegacyFileLocation",
            fileReference: fileId.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            volumeId: Long.ZERO,
            localId: 0,
            secret: loc.source.secret
          };
        case "thumbnail":
          if (!fileId.fileReference) {
            throw new tdFileId.InvalidFileIdError("Expected thumbnail photo to have file reference");
          }
          if (loc.source.fileType !== FileType.Photo && loc.source.fileType !== FileType.Thumbnail) {
            throw new tdFileId.InvalidFileIdError("Expected a thumbnail to have a correct file type");
          }
          return {
            _: loc.source.fileType === FileType.Photo ? "inputPhotoFileLocation" : "inputDocumentFileLocation",
            fileReference: fileId.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            thumbSize: loc.source.thumbnailType
          };
        case "dialogPhoto":
          return {
            _: "inputPeerPhotoFileLocation",
            big: loc.source.big,
            peer: dialogPhotoToInputPeer(loc.source),
            photoId: loc.id
          };
        case "stickerSetThumbnail":
          return {
            _: "inputStickerSetThumb",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            thumbVersion: 0
          };
        case "fullLegacy":
          if (!fileId.fileReference) {
            throw new tdFileId.InvalidFileIdError("Expected legacy photo to have file reference");
          }
          return {
            _: "inputPhotoLegacyFileLocation",
            fileReference: fileId.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            volumeId: loc.source.volumeId,
            localId: loc.source.localId,
            secret: loc.source.secret
          };
        case "dialogPhotoLegacy":
          return {
            _: "inputPeerPhotoFileLocationLegacy",
            big: loc.source.big,
            peer: dialogPhotoToInputPeer(loc.source),
            volumeId: loc.source.volumeId,
            localId: loc.source.localId
          };
        case "stickerSetThumbnailLegacy":
          return {
            _: "inputStickerSetThumbLegacy",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            volumeId: loc.source.volumeId,
            localId: loc.source.localId
          };
        case "stickerSetThumbnailVersion":
          return {
            _: "inputStickerSetThumb",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            thumbVersion: loc.source.version
          };
        default:
          assertNever(loc.source);
      }
      throw new tdFileId.ConversionError("inputFileLocation");
    }
    case "common": {
      if (!fileId.fileReference) {
        throw new tdFileId.InvalidFileIdError("Expected common to have file reference");
      }
      if (fileId.type === FileType.Encrypted) {
        return {
          _: "inputEncryptedFileLocation",
          id: loc.id,
          accessHash: loc.accessHash
        };
      } else if (fileId.type === FileType.Secure || fileId.type === FileType.SecureRaw) {
        return {
          _: "inputSecureFileLocation",
          id: loc.id,
          accessHash: loc.accessHash
        };
      }
      return {
        _: "inputDocumentFileLocation",
        fileReference: fileId.fileReference,
        id: loc.id,
        accessHash: loc.accessHash,
        thumbSize: ""
      };
    }
    default:
      assertNever();
  }
}
function fileIdToInputDocument(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  if (fileId.location._ !== "common" || fileId.type === FileType.Secure || fileId.type === FileType.SecureRaw || fileId.type === FileType.Encrypted) {
    throw new tdFileId.ConversionError("inputDocument");
  }
  let fileRef = fileId.fileReference;
  if (!fileRef) {
    if (fileId.type === FileType.Sticker) {
      fileRef = EMPTY_BUFFER;
    } else {
      throw new tdFileId.InvalidFileIdError("Expected document to have file reference");
    }
  }
  return {
    _: "inputDocument",
    fileReference: fileRef,
    id: fileId.location.id,
    accessHash: fileId.location.accessHash
  };
}
function fileIdToInputPhoto(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  if (fileId.location._ !== "photo") {
    throw new tdFileId.ConversionError("inputPhoto");
  }
  if (!fileId.fileReference) {
    throw new tdFileId.InvalidFileIdError("Expected photo to have file reference");
  }
  return {
    _: "inputPhoto",
    fileReference: fileId.fileReference,
    id: fileId.location.id,
    accessHash: fileId.location.accessHash
  };
}
function fileIdToEncryptedFile(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  if (fileId.location._ !== "common" || fileId.type !== FileType.Encrypted) {
    throw new tdFileId.ConversionError("inputEncryptedFile");
  }
  return {
    _: "inputEncryptedFile",
    id: fileId.location.id,
    accessHash: fileId.location.accessHash
  };
}
function fileIdToSecureFile(fileId) {
  if (typeof fileId === "string") fileId = parseFileId(fileId);
  if (fileId.location._ !== "common" || fileId.type !== FileType.Secure && fileId.type !== FileType.SecureRaw) {
    throw new tdFileId.ConversionError("inputSecureFile");
  }
  return {
    _: "inputSecureFile",
    id: fileId.location.id,
    accessHash: fileId.location.accessHash
  };
}
export {
  fileIdToEncryptedFile,
  fileIdToInputDocument,
  fileIdToInputFileLocation,
  fileIdToInputPhoto,
  fileIdToInputWebFileLocation,
  fileIdToSecureFile
};
