// @mtcute/core/highlevel/utils/convert-file-id.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const Long = require("long");
const utils = require("../../types/utils.cjs");
const peerUtils = require("../../utils/peer-utils.cjs");
const FileType = fileId.tdFileId.FileType;
const EMPTY_BUFFER = new Uint8Array(0);
function dialogPhotoToInputPeer(dialog) {
  const markedPeerId = dialog.id;
  const [peerType, peerId] = peerUtils.parseMarkedPeerId(markedPeerId);
  if (peerType === "user") {
    return {
      _: "inputPeerUser",
      userId: peerId,
      accessHash: dialog.accessHash
    };
  } else if (peerType === "chat") {
    return {
      _: "inputPeerChat",
      chatId: peerId
    };
  }
  return {
    _: "inputPeerChannel",
    channelId: peerId,
    accessHash: dialog.accessHash
  };
}
function fileIdToInputWebFileLocation(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  if (fileId$1.location._ !== "web") {
    throw new fileId.tdFileId.ConversionError("inputWebFileLocation");
  }
  return {
    _: "inputWebFileLocation",
    url: fileId$1.location.url,
    accessHash: fileId$1.location.accessHash
  };
}
function fileIdToInputFileLocation(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  const loc = fileId$1.location;
  switch (loc._) {
    case "web":
      throw new fileId.tdFileId.ConversionError("InputFileLocation");
    case "photo": {
      switch (loc.source._) {
        case "legacy":
          if (!fileId$1.fileReference) {
            throw new fileId.tdFileId.InvalidFileIdError("Expected legacy photo to have file reference");
          }
          return {
            _: "inputPhotoLegacyFileLocation",
            fileReference: fileId$1.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            volumeId: Long.ZERO,
            localId: 0,
            secret: loc.source.secret
          };
        case "thumbnail":
          if (!fileId$1.fileReference) {
            throw new fileId.tdFileId.InvalidFileIdError("Expected thumbnail photo to have file reference");
          }
          if (loc.source.fileType !== FileType.Photo && loc.source.fileType !== FileType.Thumbnail) {
            throw new fileId.tdFileId.InvalidFileIdError("Expected a thumbnail to have a correct file type");
          }
          return {
            _: loc.source.fileType === FileType.Photo ? "inputPhotoFileLocation" : "inputDocumentFileLocation",
            fileReference: fileId$1.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            thumbSize: loc.source.thumbnailType
          };
        case "dialogPhoto":
          return {
            _: "inputPeerPhotoFileLocation",
            big: loc.source.big,
            peer: dialogPhotoToInputPeer(loc.source),
            photoId: loc.id
          };
        case "stickerSetThumbnail":
          return {
            _: "inputStickerSetThumb",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            thumbVersion: 0
          };
        case "fullLegacy":
          if (!fileId$1.fileReference) {
            throw new fileId.tdFileId.InvalidFileIdError("Expected legacy photo to have file reference");
          }
          return {
            _: "inputPhotoLegacyFileLocation",
            fileReference: fileId$1.fileReference,
            id: loc.id,
            accessHash: loc.accessHash,
            volumeId: loc.source.volumeId,
            localId: loc.source.localId,
            secret: loc.source.secret
          };
        case "dialogPhotoLegacy":
          return {
            _: "inputPeerPhotoFileLocationLegacy",
            big: loc.source.big,
            peer: dialogPhotoToInputPeer(loc.source),
            volumeId: loc.source.volumeId,
            localId: loc.source.localId
          };
        case "stickerSetThumbnailLegacy":
          return {
            _: "inputStickerSetThumbLegacy",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            volumeId: loc.source.volumeId,
            localId: loc.source.localId
          };
        case "stickerSetThumbnailVersion":
          return {
            _: "inputStickerSetThumb",
            stickerset: {
              _: "inputStickerSetID",
              id: loc.source.id,
              accessHash: loc.source.accessHash
            },
            thumbVersion: loc.source.version
          };
        default:
          utils.assertNever(loc.source);
      }
      throw new fileId.tdFileId.ConversionError("inputFileLocation");
    }
    case "common": {
      if (!fileId$1.fileReference) {
        throw new fileId.tdFileId.InvalidFileIdError("Expected common to have file reference");
      }
      if (fileId$1.type === FileType.Encrypted) {
        return {
          _: "inputEncryptedFileLocation",
          id: loc.id,
          accessHash: loc.accessHash
        };
      } else if (fileId$1.type === FileType.Secure || fileId$1.type === FileType.SecureRaw) {
        return {
          _: "inputSecureFileLocation",
          id: loc.id,
          accessHash: loc.accessHash
        };
      }
      return {
        _: "inputDocumentFileLocation",
        fileReference: fileId$1.fileReference,
        id: loc.id,
        accessHash: loc.accessHash,
        thumbSize: ""
      };
    }
    default:
      utils.assertNever();
  }
}
function fileIdToInputDocument(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  if (fileId$1.location._ !== "common" || fileId$1.type === FileType.Secure || fileId$1.type === FileType.SecureRaw || fileId$1.type === FileType.Encrypted) {
    throw new fileId.tdFileId.ConversionError("inputDocument");
  }
  let fileRef = fileId$1.fileReference;
  if (!fileRef) {
    if (fileId$1.type === FileType.Sticker) {
      fileRef = EMPTY_BUFFER;
    } else {
      throw new fileId.tdFileId.InvalidFileIdError("Expected document to have file reference");
    }
  }
  return {
    _: "inputDocument",
    fileReference: fileRef,
    id: fileId$1.location.id,
    accessHash: fileId$1.location.accessHash
  };
}
function fileIdToInputPhoto(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  if (fileId$1.location._ !== "photo") {
    throw new fileId.tdFileId.ConversionError("inputPhoto");
  }
  if (!fileId$1.fileReference) {
    throw new fileId.tdFileId.InvalidFileIdError("Expected photo to have file reference");
  }
  return {
    _: "inputPhoto",
    fileReference: fileId$1.fileReference,
    id: fileId$1.location.id,
    accessHash: fileId$1.location.accessHash
  };
}
function fileIdToEncryptedFile(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  if (fileId$1.location._ !== "common" || fileId$1.type !== FileType.Encrypted) {
    throw new fileId.tdFileId.ConversionError("inputEncryptedFile");
  }
  return {
    _: "inputEncryptedFile",
    id: fileId$1.location.id,
    accessHash: fileId$1.location.accessHash
  };
}
function fileIdToSecureFile(fileId$1) {
  if (typeof fileId$1 === "string") fileId$1 = fileId.parseFileId(fileId$1);
  if (fileId$1.location._ !== "common" || fileId$1.type !== FileType.Secure && fileId$1.type !== FileType.SecureRaw) {
    throw new fileId.tdFileId.ConversionError("inputSecureFile");
  }
  return {
    _: "inputSecureFile",
    id: fileId$1.location.id,
    accessHash: fileId$1.location.accessHash
  };
}
exports.fileIdToEncryptedFile = fileIdToEncryptedFile;
exports.fileIdToInputDocument = fileIdToInputDocument;
exports.fileIdToInputFileLocation = fileIdToInputFileLocation;
exports.fileIdToInputPhoto = fileIdToInputPhoto;
exports.fileIdToInputWebFileLocation = fileIdToInputWebFileLocation;
exports.fileIdToSecureFile = fileIdToSecureFile;
