// @mtcute/core/highlevel/utils/inline-utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const utils$1 = require("../../types/utils.cjs");
function parseInlineMessageId(id) {
  const buf = utils.base64.decode(id, true);
  const reader = tlRuntime.TlBinaryReader.manual(buf);
  if (buf.length === 20) {
    return {
      _: "inputBotInlineMessageID",
      dcId: reader.int(),
      id: reader.long(),
      accessHash: reader.long()
    };
  }
  return {
    _: "inputBotInlineMessageID64",
    dcId: reader.int(),
    ownerId: reader.long(),
    id: reader.int(),
    accessHash: reader.long()
  };
}
function encodeInlineMessageId(id) {
  let writer;
  switch (id._) {
    case "inputBotInlineMessageID":
      writer = tlRuntime.TlBinaryWriter.manual(20);
      writer.int(id.dcId);
      writer.long(id.id);
      writer.long(id.accessHash);
      break;
    case "inputBotInlineMessageID64":
      writer = tlRuntime.TlBinaryWriter.manual(24);
      writer.int(id.dcId);
      writer.long(id.ownerId);
      writer.int(id.id);
      writer.long(id.accessHash);
      break;
    default:
      utils$1.assertNever();
  }
  return utils.base64.encode(writer.result(), true);
}
function normalizeInlineId(id) {
  if (typeof id === "string") {
    return parseInlineMessageId(id);
  }
  return id;
}
exports.encodeInlineMessageId = encodeInlineMessageId;
exports.normalizeInlineId = normalizeInlineId;
exports.parseInlineMessageId = parseInlineMessageId;
