// @mtcute/core/highlevel/utils/misc-utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../types/errors.cjs");
function normalizePhoneNumber(phone) {
  phone = phone.trim().replace(/[+()\s-]/g, "");
  if (!phone.match(/^\d+$/)) throw new errors.MtArgumentError("Invalid phone number");
  return phone;
}
async function resolveMaybeDynamic(val) {
  return typeof val === "function" ? await val() : await val;
}
function makeArrayWithTotal(arr, total) {
  const a = arr;
  a.total = total;
  return a;
}
function makeArrayPaginated(arr, total, next) {
  const a = arr;
  a.total = total;
  a.next = next;
  return a;
}
function makeArrayPaginatedWithMeta(arr, total, meta, next) {
  const a = arr;
  a.total = total;
  a.next = next;
  Object.assign(a, meta);
  return a;
}
function normalizeDate(date) {
  return date ? ~~((typeof date === "number" ? date : date.getTime()) / 1e3) : void 0;
}
function normalizeMessageId(msg) {
  if (!msg) return void 0;
  return typeof msg === "number" ? msg : msg.id;
}
exports.makeArrayPaginated = makeArrayPaginated;
exports.makeArrayPaginatedWithMeta = makeArrayPaginatedWithMeta;
exports.makeArrayWithTotal = makeArrayWithTotal;
exports.normalizeDate = normalizeDate;
exports.normalizeMessageId = normalizeMessageId;
exports.normalizePhoneNumber = normalizePhoneNumber;
exports.resolveMaybeDynamic = resolveMaybeDynamic;
