// @mtcute/core/highlevel/utils/misc-utils.js
import { MtArgumentError } from "../../types/errors.js";
function normalizePhoneNumber(phone) {
  phone = phone.trim().replace(/[+()\s-]/g, "");
  if (!phone.match(/^\d+$/)) throw new MtArgumentError("Invalid phone number");
  return phone;
}
async function resolveMaybeDynamic(val) {
  return typeof val === "function" ? await val() : await val;
}
function makeArrayWithTotal(arr, total) {
  const a = arr;
  a.total = total;
  return a;
}
function makeArrayPaginated(arr, total, next) {
  const a = arr;
  a.total = total;
  a.next = next;
  return a;
}
function makeArrayPaginatedWithMeta(arr, total, meta, next) {
  const a = arr;
  a.total = total;
  a.next = next;
  Object.assign(a, meta);
  return a;
}
function normalizeDate(date) {
  return date ? ~~((typeof date === "number" ? date : date.getTime()) / 1e3) : void 0;
}
function normalizeMessageId(msg) {
  if (!msg) return void 0;
  return typeof msg === "number" ? msg : msg.id;
}
export {
  makeArrayPaginated,
  makeArrayPaginatedWithMeta,
  makeArrayWithTotal,
  normalizeDate,
  normalizeMessageId,
  normalizePhoneNumber,
  resolveMaybeDynamic
};
