// @mtcute/core/highlevel/utils/peer-utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils$1 = require("@fuman/utils");
const utils = require("../../types/utils.cjs");
const errors = require("../types/errors.cjs");
const innerTl = require("../../tl/inner-tl.cjs");
const INVITE_LINK_REGEX = /^(?:https?:\/\/)?(?:www\.)?t(?:elegram)?\.(?:org|me|dog)\/(?:joinchat\/|\+)([\w-]+)$/i;
function toInputPeer(res) {
  if (innerTl.isAnyInputPeer(res)) return res;
  switch (res._) {
    case "user": {
      return {
        _: "inputPeerUser",
        userId: res.id,
        accessHash: utils$1.asNonNull(res.accessHash)
      };
    }
    case "chat": {
      return {
        _: "inputPeerChat",
        chatId: res.id
      };
    }
    case "channel": {
      return {
        _: "inputPeerChannel",
        channelId: res.id,
        accessHash: utils$1.asNonNull(res.accessHash)
      };
    }
    case "inputChannelEmpty":
    case "inputUserEmpty":
      return { _: "inputPeerEmpty" };
    case "inputUser":
      return {
        _: "inputPeerUser",
        userId: res.userId,
        accessHash: res.accessHash
      };
    case "inputUserSelf":
      return { _: "inputPeerSelf" };
    case "inputChannel":
      return {
        _: "inputPeerChannel",
        channelId: res.channelId,
        accessHash: res.accessHash
      };
    case "inputChannelFromMessage":
      return {
        _: "inputPeerChannelFromMessage",
        channelId: res.channelId,
        msgId: res.msgId,
        peer: res.peer
      };
    case "inputUserFromMessage":
      return {
        _: "inputPeerUserFromMessage",
        userId: res.userId,
        msgId: res.msgId,
        peer: res.peer
      };
    default:
      utils.assertNever();
  }
}
function toInputUser(res, input) {
  if (innerTl.isAnyInputUser(res)) return res;
  switch (res._) {
    case "inputPeerSelf":
      return { _: "inputUserSelf" };
    case "inputPeerUser":
      return {
        _: "inputUser",
        userId: res.userId,
        accessHash: res.accessHash
      };
    case "inputPeerUserFromMessage":
      return {
        _: "inputUserFromMessage",
        userId: res.userId,
        msgId: res.msgId,
        peer: res.peer
      };
  }
  throw new errors.MtInvalidPeerTypeError(input ?? res, "user");
}
function toInputChannel(res, input) {
  if (innerTl.isAnyInputChannel(res)) return res;
  switch (res._) {
    case "inputPeerChannel":
      return {
        _: "inputChannel",
        channelId: res.channelId,
        accessHash: res.accessHash
      };
    case "inputPeerChannelFromMessage":
      return {
        _: "inputChannelFromMessage",
        channelId: res.channelId,
        msgId: res.msgId,
        peer: res.peer
      };
  }
  throw new errors.MtInvalidPeerTypeError(input ?? res, "channel");
}
function isInputPeerUser(obj) {
  switch (obj._) {
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
    case "inputPeerSelf":
      return true;
  }
  return false;
}
function isInputPeerChannel(obj) {
  switch (obj._) {
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
      return true;
  }
  return false;
}
function isInputPeerChat(obj) {
  return obj._ === "inputPeerChat";
}
function inputPeerToPeer(inp) {
  switch (inp._) {
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
      return { _: "peerUser", userId: inp.userId };
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
      return { _: "peerChannel", channelId: inp.channelId };
    case "inputPeerChat":
      return { _: "peerChat", chatId: inp.chatId };
    default:
      throw new errors.MtInvalidPeerTypeError(inp, `Cannot convert ${inp._} to peer`);
  }
}
function extractUsernames(obj) {
  if (obj.usernames?.length) return obj.usernames.map((x) => x.username.toLowerCase());
  if (obj.username) return [obj.username.toLowerCase()];
  return [];
}
exports.INVITE_LINK_REGEX = INVITE_LINK_REGEX;
exports.extractUsernames = extractUsernames;
exports.inputPeerToPeer = inputPeerToPeer;
exports.isInputPeerChannel = isInputPeerChannel;
exports.isInputPeerChat = isInputPeerChat;
exports.isInputPeerUser = isInputPeerUser;
exports.toInputChannel = toInputChannel;
exports.toInputPeer = toInputPeer;
exports.toInputUser = toInputUser;
