// @mtcute/core/highlevel/utils/peer-utils.js
import { asNonNull } from "@fuman/utils";
import { assertNever } from "../../types/utils.js";
import { MtInvalidPeerTypeError } from "../types/errors.js";
import { isAnyInputPeer, isAnyInputUser, isAnyInputChannel } from "../../tl/inner-tl.js";
const INVITE_LINK_REGEX = /^(?:https?:\/\/)?(?:www\.)?t(?:elegram)?\.(?:org|me|dog)\/(?:joinchat\/|\+)([\w-]+)$/i;
function toInputPeer(res) {
  if (isAnyInputPeer(res)) return res;
  switch (res._) {
    case "user": {
      return {
        _: "inputPeerUser",
        userId: res.id,
        accessHash: asNonNull(res.accessHash)
      };
    }
    case "chat": {
      return {
        _: "inputPeerChat",
        chatId: res.id
      };
    }
    case "channel": {
      return {
        _: "inputPeerChannel",
        channelId: res.id,
        accessHash: asNonNull(res.accessHash)
      };
    }
    case "inputChannelEmpty":
    case "inputUserEmpty":
      return { _: "inputPeerEmpty" };
    case "inputUser":
      return {
        _: "inputPeerUser",
        userId: res.userId,
        accessHash: res.accessHash
      };
    case "inputUserSelf":
      return { _: "inputPeerSelf" };
    case "inputChannel":
      return {
        _: "inputPeerChannel",
        channelId: res.channelId,
        accessHash: res.accessHash
      };
    case "inputChannelFromMessage":
      return {
        _: "inputPeerChannelFromMessage",
        channelId: res.channelId,
        msgId: res.msgId,
        peer: res.peer
      };
    case "inputUserFromMessage":
      return {
        _: "inputPeerUserFromMessage",
        userId: res.userId,
        msgId: res.msgId,
        peer: res.peer
      };
    default:
      assertNever();
  }
}
function toInputUser(res, input) {
  if (isAnyInputUser(res)) return res;
  switch (res._) {
    case "inputPeerSelf":
      return { _: "inputUserSelf" };
    case "inputPeerUser":
      return {
        _: "inputUser",
        userId: res.userId,
        accessHash: res.accessHash
      };
    case "inputPeerUserFromMessage":
      return {
        _: "inputUserFromMessage",
        userId: res.userId,
        msgId: res.msgId,
        peer: res.peer
      };
  }
  throw new MtInvalidPeerTypeError(input ?? res, "user");
}
function toInputChannel(res, input) {
  if (isAnyInputChannel(res)) return res;
  switch (res._) {
    case "inputPeerChannel":
      return {
        _: "inputChannel",
        channelId: res.channelId,
        accessHash: res.accessHash
      };
    case "inputPeerChannelFromMessage":
      return {
        _: "inputChannelFromMessage",
        channelId: res.channelId,
        msgId: res.msgId,
        peer: res.peer
      };
  }
  throw new MtInvalidPeerTypeError(input ?? res, "channel");
}
function isInputPeerUser(obj) {
  switch (obj._) {
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
    case "inputPeerSelf":
      return true;
  }
  return false;
}
function isInputPeerChannel(obj) {
  switch (obj._) {
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
      return true;
  }
  return false;
}
function isInputPeerChat(obj) {
  return obj._ === "inputPeerChat";
}
function inputPeerToPeer(inp) {
  switch (inp._) {
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
      return { _: "peerUser", userId: inp.userId };
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
      return { _: "peerChannel", channelId: inp.channelId };
    case "inputPeerChat":
      return { _: "peerChat", chatId: inp.chatId };
    default:
      throw new MtInvalidPeerTypeError(inp, `Cannot convert ${inp._} to peer`);
  }
}
function extractUsernames(obj) {
  if (obj.usernames?.length) return obj.usernames.map((x) => x.username.toLowerCase());
  if (obj.username) return [obj.username.toLowerCase()];
  return [];
}
export {
  INVITE_LINK_REGEX,
  extractUsernames,
  inputPeerToPeer,
  isInputPeerChannel,
  isInputPeerChat,
  isInputPeerUser,
  toInputChannel,
  toInputPeer,
  toInputUser
};
