// @mtcute/core/highlevel/utils/query-batcher.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
function batchedQuery(params) {
  const { inputKey, outputKey, fetch, maxBatchSize = Infinity, maxConcurrent = 1, retrySingleOnError } = params;
  const symbol = Symbol("batchedQueryState");
  function getState(client_) {
    const client = client_;
    if (!client[symbol]) {
      client[symbol] = {
        waiters: /* @__PURE__ */ new Map(),
        fetchingKeys: /* @__PURE__ */ new Set(),
        retryQueue: new utils.Deque(),
        numRunning: 0
      };
    }
    return client[symbol];
  }
  function addWaiter(client, waiters, item) {
    const key = inputKey(item, client);
    let arr = waiters.get(key);
    if (!arr) {
      arr = [];
      waiters.set(key, arr);
    }
    return new Promise((resolve, reject) => {
      arr.push([item, resolve, reject]);
    });
  }
  function popWaiters(waiters, key) {
    const arr = waiters.get(key);
    if (!arr) return [];
    waiters.delete(key);
    return arr;
  }
  function startLoops(client, state) {
    for (let i = state.numRunning; i <= maxConcurrent; i++) {
      processPending(client, state);
    }
  }
  function processPending(client, state) {
    const { waiters, fetchingKeys, retryQueue } = state;
    if (state.numRunning >= maxConcurrent) return;
    const request = [];
    const requestKeys = [];
    let isRetryRequest = false;
    if (retryQueue.length > 0) {
      isRetryRequest = true;
      const it = retryQueue.popFront();
      request.push(it);
      const key = inputKey(it, client);
      requestKeys.push(key);
      fetchingKeys.add(key);
    } else {
      for (const it of waiters.keys()) {
        if (fetchingKeys.has(it)) continue;
        request.push(waiters.get(it)[0][0]);
        requestKeys.push(it);
        fetchingKeys.add(it);
        if (request.length === maxBatchSize) break;
      }
      if (request.length === 0) return;
    }
    state.numRunning += 1;
    fetch(client, request).then((res) => {
      const receivedKeys = /* @__PURE__ */ new Set();
      for (const it of res) {
        const key = outputKey(it, client);
        receivedKeys.add(key);
        for (const waiter of popWaiters(waiters, key)) {
          waiter[1](it);
        }
        fetchingKeys.delete(key);
      }
      for (const key of requestKeys) {
        if (!receivedKeys.has(key)) {
          for (const waiter of popWaiters(waiters, key)) {
            waiter[1](null);
          }
          fetchingKeys.delete(key);
        }
      }
    }).catch((err) => {
      if (retrySingleOnError && !isRetryRequest) {
        const retry = retrySingleOnError(request, err);
        if (retry === true) {
          for (const it of request) {
            retryQueue.pushBack(it);
          }
          return;
        } else if (Array.isArray(retry)) {
          for (const req of retry) {
            const requestKeysIdx = request.indexOf(req);
            if (requestKeysIdx < 0) continue;
            retryQueue.pushBack(req);
            request.splice(requestKeysIdx, 1);
            requestKeys.splice(requestKeysIdx, 1);
          }
        }
      }
      for (const key of requestKeys) {
        for (const waiter of popWaiters(waiters, key)) {
          waiter[2](err);
        }
        fetchingKeys.delete(key);
      }
    }).then(() => {
      state.numRunning -= 1;
      if (waiters.size > 0) processPending(client, state);
    });
  }
  return function(client, item) {
    const state = getState(client);
    const promise = addWaiter(client, state.waiters, item);
    startLoops(client, state);
    return promise;
  };
}
exports.batchedQuery = batchedQuery;
