// @mtcute/core/highlevel/utils/string-session.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const dcs = require("../../utils/dcs.cjs");
const errors = require("../../types/errors.cjs");
function writeStringSession(data) {
  const writer = tlRuntime.TlBinaryWriter.manual(512);
  const version = data.version;
  if (version !== 3) {
    throw new errors.MtArgumentError(`Unsupported string session version: ${version}`);
  }
  let flags = 0;
  if (data.self) {
    flags |= 1;
  }
  writer.uint8View[0] = version;
  writer.pos += 1;
  if (data.primaryDcs.media !== data.primaryDcs.main) {
    flags |= 4;
  }
  writer.int(flags);
  writer.bytes(dcs.serializeBasicDcOption(data.primaryDcs.main));
  if (data.primaryDcs.media !== data.primaryDcs.main) {
    writer.bytes(dcs.serializeBasicDcOption(data.primaryDcs.media));
  }
  if (data.self) {
    writer.int53(data.self.userId);
    writer.boolean(data.self.isBot);
  }
  writer.bytes(data.authKey);
  return utils.base64.encode(writer.result(), true);
}
function readStringSession(data) {
  if (typeof data !== "string") {
    return {
      version: 3,
      primaryDcs: data.primaryDcs ?? (data.testMode ? dcs.defaultTestDc : dcs.defaultProductionDc),
      self: data.self ?? null,
      authKey: data.authKey
    };
  }
  const buf = utils.base64.decode(data, true);
  const version = buf[0];
  if (version !== 3) {
    throw new Error(`Invalid session string (version = ${version})`);
  }
  const reader = tlRuntime.TlBinaryReader.manual(buf, 1);
  const flags = reader.int();
  const hasSelf = flags & 1;
  const testModeOld = Boolean(flags & 2);
  const hasMedia = version >= 2 && Boolean(flags & 4);
  const primaryDc = dcs.parseBasicDcOption(reader.bytes());
  if (primaryDc === null) {
    throw new errors.MtArgumentError("Invalid session string (failed to parse primaryDc)");
  }
  const primaryMediaDc = hasMedia ? dcs.parseBasicDcOption(reader.bytes()) : primaryDc;
  if (primaryMediaDc === null) {
    throw new errors.MtArgumentError("Invalid session string (failed to parse primaryMediaDc)");
  }
  if (testModeOld) {
    primaryDc.testMode = true;
    primaryMediaDc.testMode = true;
  } else if (primaryDc.testMode !== primaryMediaDc.testMode) {
    throw new errors.MtArgumentError("Primary DC and primary media DC must have the same test mode flag");
  }
  let self = null;
  if (hasSelf) {
    const selfId = reader.int53();
    const selfBot = reader.boolean();
    self = {
      userId: selfId,
      isBot: selfBot,
      // todo: we should make sure we fetch this from the server at first start
      isPremium: false,
      usernames: []
    };
  }
  const key = reader.bytes();
  return {
    version,
    primaryDcs: {
      main: primaryDc,
      media: primaryMediaDc
    },
    self,
    authKey: key
  };
}
exports.readStringSession = readStringSession;
exports.writeStringSession = writeStringSession;
