// @mtcute/core/highlevel/types/auth/sent-code.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const sentCodeMap = {
  "auth.sentCodeTypeApp": "app",
  "auth.sentCodeTypeCall": "call",
  "auth.sentCodeTypeFlashCall": "flash_call",
  "auth.sentCodeTypeSms": "sms",
  "auth.sentCodeTypeMissedCall": "missed_call",
  "auth.sentCodeTypeEmailCode": "email",
  "auth.sentCodeTypeSetUpEmailRequired": "email_required",
  "auth.sentCodeTypeFragmentSms": "fragment",
  "auth.sentCodeTypeFirebaseSms": "firebase",
  "auth.sentCodeTypeSmsWord": "sms_word",
  "auth.sentCodeTypeSmsPhrase": "sms_phrase"
};
const nextCodeMap = {
  "auth.codeTypeCall": "call",
  "auth.codeTypeFlashCall": "flash_call",
  "auth.codeTypeSms": "sms",
  "auth.codeTypeMissedCall": "missed_call",
  "auth.codeTypeFragmentSms": "fragment"
};
class SentCode {
  constructor(raw) {
    this.raw = raw;
  }
  /**
   * Type of currently sent confirmation code
   */
  get type() {
    return sentCodeMap[this.raw.type._];
  }
  /**
   * Type of the confirmation code that will be sent
   * if you call {@link TelegramClient.resendCode}.
   */
  get nextType() {
    return this.raw.nextType ? nextCodeMap[this.raw.nextType._] : "none";
  }
  /**
   * Confirmation code identifier used for the next authorization steps
   * (like {@link TelegramClient.signIn} and {@link TelegramClient.signUp})
   */
  get phoneCodeHash() {
    return this.raw.phoneCodeHash;
  }
  /**
   * Delay in seconds to wait before calling {@link TelegramClient.resendCode}
   */
  get timeout() {
    return this.raw.timeout ?? 0;
  }
  /**
   * If the code is sent via SMS with a word/phrase, this field *may* contain the beginning of the message
   */
  get beginning() {
    switch (this.raw.type._) {
      case "auth.sentCodeTypeSmsPhrase":
      case "auth.sentCodeTypeSmsWord":
        return this.raw.type.beginning;
      default:
        return void 0;
    }
  }
  /**
   * Length of the code (0 for flash calls)
   */
  get length() {
    return "length" in this.raw.type ? this.raw.type.length : 0;
  }
}
const SentCode$1 = /* @__PURE__ */ inspectable.makeInspectable(SentCode);
exports.SentCode = SentCode$1;
