// @mtcute/tl-runtime/reader.js
import { utf8 } from "@fuman/utils";
import Long from "long";
const TWO_PWR_32_DBL = (1 << 16) * (1 << 16);
class TlUnknownObjectError extends Error {
  constructor(objectId) {
    super(`Unknown object id: 0x${objectId.toString(16)}`);
    this.objectId = objectId;
  }
}
class TlBinaryReader {
  /**
   * @param objectsMap  Readers map
   * @param data  Buffer to read from
   * @param start  Position to start reading from
   */
  constructor(objectsMap, data, start = 0) {
    this.objectsMap = objectsMap;
    if (ArrayBuffer.isView(data)) {
      this.pos = start;
      this.dataView = new DataView(data.buffer, data.byteOffset, data.byteLength);
      this.uint8View = new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else {
      this.pos = start;
      this.dataView = new DataView(data);
      this.uint8View = new Uint8Array(data);
    }
  }
  dataView;
  uint8View;
  pos = 0;
  _objectMapper;
  /**
   * Create a new reader without objects map for manual usage
   *
   * @param data  Buffer to read from
   * @param start  Position to start reading from
   */
  static manual(data, start = 0) {
    return new TlBinaryReader(void 0, data, start);
  }
  /**
   * Deserialize a single object
   *
   * @param objectsMap  Readers map
   * @param data  Buffer to read from
   * @param start  Position to start reading from
   */
  static deserializeObject(objectsMap, data, start = 0) {
    return new TlBinaryReader(objectsMap, data, start).object();
  }
  int() {
    const res = this.dataView.getInt32(this.pos, true);
    this.pos += 4;
    return res;
  }
  uint() {
    const res = this.dataView.getUint32(this.pos, true);
    this.pos += 4;
    return res;
  }
  /**
   * Get the next {@link uint} without advancing the reader cursor
   */
  peekUint() {
    return this.dataView.getUint32(this.pos, true);
  }
  int53() {
    const res = (this.dataView.getInt32(this.pos, true) >>> 0) + TWO_PWR_32_DBL * this.dataView.getInt32(this.pos + 4, true);
    this.pos += 8;
    return res;
  }
  long(unsigned = false) {
    const lo = this.dataView.getInt32(this.pos, true);
    const hi = this.dataView.getInt32(this.pos + 4, true);
    this.pos += 8;
    return new Long(lo, hi, unsigned);
  }
  float() {
    const res = this.dataView.getFloat32(this.pos, true);
    this.pos += 4;
    return res;
  }
  double() {
    const res = this.dataView.getFloat64(this.pos, true);
    this.pos += 8;
    return res;
  }
  boolean() {
    const val = this.uint();
    if (val === 3162085175) return false;
    if (val === 2574415285) return true;
    throw new Error(`Expected either boolTrue or boolFalse, got 0x${val.toString(16)}`);
  }
  /**
   * Read raw bytes of the given length
   * @param bytes  Length of the buffer to read
   */
  raw(bytes = -1) {
    if (bytes === -1) bytes = this.uint8View.length - this.pos;
    return this.uint8View.subarray(this.pos, this.pos += bytes);
  }
  int128() {
    return this.uint8View.subarray(this.pos, this.pos += 16);
  }
  int256() {
    return this.uint8View.subarray(this.pos, this.pos += 32);
  }
  bytes() {
    const firstByte = this.uint8View[this.pos++];
    let length;
    let padding;
    if (firstByte === 254) {
      length = this.uint8View[this.pos++] | this.uint8View[this.pos++] << 8 | this.uint8View[this.pos++] << 16;
      padding = length % 4;
    } else {
      length = firstByte;
      padding = (length + 1) % 4;
    }
    const data = this.raw(length);
    if (padding > 0) this.pos += 4 - padding;
    return data;
  }
  string() {
    return utf8.decoder.decode(this.bytes());
  }
  object(id = this.uint()) {
    if (id === 481674261) {
      return this.vector(this.object, true);
    }
    if (id === 3162085175) return false;
    if (id === 2574415285) return true;
    if (id === 1072550713) return true;
    if (id === 1450380236) return null;
    const reader = this.objectsMap[id];
    if (!reader) {
      this.seek(-4);
      const pos = this.pos;
      const error = new TlUnknownObjectError(id);
      this.pos = pos;
      throw error;
    }
    return reader(this);
  }
  vector(reader = this.object, bare = false) {
    if (!bare) {
      const uint = this.uint();
      if (uint !== 481674261) {
        throw new Error(
          `Invalid object code, expected 0x1cb5c415 (vector), got 0x${uint.toString(16)} at ${this.pos - 4}`
        );
      }
    }
    const length = this.uint();
    const ret = [];
    for (let i = 0; i < length; i++) ret.push(reader.call(this));
    return ret;
  }
  /**
   * Advance the reader cursor by the given amount of bytes
   *
   * @param delta  Amount of bytes to advance (can be negative)
   */
  seek(delta) {
    this.seekTo(this.pos + delta);
  }
  /**
   * Seek to the given position
   *
   * @param pos  Position to seek to
   */
  seekTo(pos) {
    if (pos >= this.uint8View.length || pos < 0) {
      throw new RangeError("New position is out of range");
    }
    this.pos = pos;
  }
}
export {
  TlBinaryReader,
  TlUnknownObjectError
};
