// @mtcute/tl-runtime/writer.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/tl-runtime] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/tl-runtime] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const TWO_PWR_32_DBL = (1 << 16) * (1 << 16);
class TlSerializationCounter {
  /**
   * @param objectMap  Writers map
   */
  constructor(objectMap) {
    this.objectMap = objectMap;
  }
  count = 0;
  /**
   * Count bytes required to serialize the given object.
   *
   * @param objectMap  Writers map
   * @param obj  Object to count bytes for
   */
  static countNeededBytes(objectMap, obj) {
    const cnt = new TlSerializationCounter(objectMap);
    cnt.object(obj);
    return cnt.count;
  }
  /**
   * Count overhead in bytes for the given number of bytes when
   * encoded as `bytes` TL type.
   *
   * @param size  Number of bytes
   */
  static countBytesOverhead(size) {
    let res = 0;
    let padding;
    if (size <= 253) {
      res += 1;
      padding = (size + 1) % 4;
    } else {
      res += 4;
      padding = size % 4;
    }
    if (padding > 0) res += 4 - padding;
    return res;
  }
  boolean() {
    this.count += 4;
  }
  double() {
    this.count += 8;
  }
  float() {
    this.count += 4;
  }
  int128() {
    this.count += 16;
  }
  int256() {
    this.count += 32;
  }
  int() {
    this.count += 4;
  }
  uint() {
    this.count += 4;
  }
  int53() {
    this.count += 8;
  }
  long() {
    this.count += 8;
  }
  null() {
    this.count += 4;
  }
  raw(val) {
    this.count += val.byteLength;
  }
  bytes(val) {
    this.count += TlSerializationCounter.countBytesOverhead(val.length) + val.length;
  }
  string(val) {
    const length = utils.utf8.encodedLength(val);
    this.count += TlSerializationCounter.countBytesOverhead(length) + length;
  }
  object(obj) {
    if (!this.objectMap[obj._]) throw new Error(`Unknown object ${obj._}`);
    this.objectMap[obj._](this, obj);
  }
  vector(fn, items) {
    this.count += 8;
    items.forEach((it) => fn.call(this, it));
  }
}
class TlBinaryWriter {
  /**
   * @param objectMap  Writers map
   * @param buffer  Buffer to write to
   * @param start  Position to start writing at
   */
  constructor(objectMap, data, start = 0) {
    this.objectMap = objectMap;
    if (ArrayBuffer.isView(data)) {
      this.pos = start;
      this.dataView = new DataView(data.buffer, data.byteOffset, data.byteLength);
      this.uint8View = new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else {
      this.pos = start;
      this.dataView = new DataView(data);
      this.uint8View = new Uint8Array(data);
    }
  }
  dataView;
  uint8View;
  /**
   * Current position in the buffer.
   */
  pos;
  /**
   * Create a new writer with the given size.
   *
   * @param objectMap  Writers map
   * @param size  Size of the writer's buffer
   */
  static alloc(objectMap, size) {
    return new TlBinaryWriter(objectMap, new ArrayBuffer(size));
  }
  /**
   * Create a new writer without objects map for manual usage
   *
   * @param buffer  Buffer to write to, or its size
   * @param start  Position to start writing at
   */
  static manual(buffer, start = 0) {
    if (typeof buffer === "number") buffer = new ArrayBuffer(buffer);
    return new TlBinaryWriter(void 0, buffer, start);
  }
  /**
   * Serialize a single object
   *
   * @param objectMap  Writers map
   * @param obj  Object to serialize
   * @param knownSize  In case the size is known, pass it here
   */
  static serializeObject(objectMap, obj, knownSize = -1) {
    if (knownSize === -1) {
      knownSize = objectMap._staticSize[obj._] || TlSerializationCounter.countNeededBytes(objectMap, obj);
    }
    const writer = TlBinaryWriter.alloc(objectMap, knownSize);
    writer.object(obj);
    return writer.uint8View;
  }
  int(val) {
    this.dataView.setInt32(this.pos, val, true);
    this.pos += 4;
  }
  uint(val) {
    this.dataView.setUint32(this.pos, val, true);
    this.pos += 4;
  }
  int53(val) {
    this.dataView.setInt32(this.pos, val % TWO_PWR_32_DBL | 0, true);
    if (val < 0) {
      this.dataView.setInt32(this.pos + 4, val / TWO_PWR_32_DBL - 1 | 0, true);
    } else {
      this.dataView.setInt32(this.pos + 4, val / TWO_PWR_32_DBL | 0, true);
    }
    this.pos += 8;
  }
  null() {
    this.uint(1450380236);
  }
  long(val) {
    this.dataView.setInt32(this.pos, val.low, true);
    this.dataView.setInt32(this.pos + 4, val.high, true);
    this.pos += 8;
  }
  float(val) {
    this.dataView.setFloat32(this.pos, val, true);
    this.pos += 4;
  }
  double(val) {
    this.dataView.setFloat64(this.pos, val, true);
    this.pos += 8;
  }
  boolean(val) {
    this.dataView.setInt32(this.pos, val ? 2574415285 : 3162085175, true);
    this.pos += 4;
  }
  /**
   * Write raw bytes to the buffer
   * @param val  Buffer to write
   */
  raw(val) {
    this.uint8View.set(val, this.pos);
    this.pos += val.byteLength;
  }
  int128(val) {
    if (val.byteLength !== 16) throw new Error("Invalid int128 length");
    this.raw(val);
  }
  int256(val) {
    if (val.byteLength !== 32) throw new Error("Invalid int256 length");
    this.raw(val);
  }
  bytes(val) {
    const length = val.byteLength;
    let padding;
    if (length <= 253) {
      this.uint8View[this.pos++] = length;
      padding = (length + 1) % 4;
    } else {
      this.uint8View[this.pos++] = 254;
      this.uint8View[this.pos++] = length & 255;
      this.uint8View[this.pos++] = length >> 8 & 255;
      this.uint8View[this.pos++] = length >> 16 & 255;
      padding = length % 4;
    }
    this.uint8View.set(val, this.pos);
    this.pos += length;
    if (padding > 0) {
      padding = 4 - padding;
      while (padding--) this.uint8View[this.pos++] = 0;
    }
  }
  string(val) {
    this.bytes(utils.utf8.encoder.encode(val));
  }
  // hot path, avoid additional runtime checks
  object(obj) {
    const fn = this.objectMap[obj._];
    if (!fn) throw new Error(`Unknown object ${obj._}`);
    fn(this, obj);
  }
  vector(fn, val, bare) {
    if (!bare) this.uint(481674261);
    this.uint(val.length);
    val.forEach((it) => fn.call(this, it, bare));
  }
  /**
   * Get the resulting buffer
   */
  result() {
    return this.uint8View.subarray(0, this.pos);
  }
}
exports.TlBinaryWriter = TlBinaryWriter;
exports.TlSerializationCounter = TlSerializationCounter;
