// @mtcute/core/highlevel/types/misc/collectible-info.js
import { makeInspectable } from "../../utils/inspectable.js";
class CollectibleInfo {
  constructor(raw) {
    this.raw = raw;
  }
  /** Date when the item was purchased */
  get purchaseDate() {
    return new Date(this.raw.purchaseDate * 1e3);
  }
  /** Crypto currency used to purchase the item */
  get cryptoCurrency() {
    return this.raw.cryptoCurrency;
  }
  /**
   * Amount of crypto currency used to purchase the item,
   * in the smallest units
   */
  get cryptoAmount() {
    return this.raw.cryptoAmount;
  }
  /** Fiat currency to which the crypto currency was converted */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Converted amount in fiat currency,
   * in the smallest units (e.g. cents)
   */
  get amount() {
    return this.raw.amount;
  }
  /** URL to the collectible on Fragment */
  get url() {
    return this.raw.url;
  }
}
const CollectibleInfo$1 = /* @__PURE__ */ makeInspectable(CollectibleInfo);
export {
  CollectibleInfo$1 as CollectibleInfo
};
