// @mtcute/core/highlevel/types/misc/collectible-info.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
class CollectibleInfo {
  constructor(raw) {
    this.raw = raw;
  }
  /** Date when the item was purchased */
  get purchaseDate() {
    return new Date(this.raw.purchaseDate * 1e3);
  }
  /** Crypto currency used to purchase the item */
  get cryptoCurrency() {
    return this.raw.cryptoCurrency;
  }
  /**
   * Amount of crypto currency used to purchase the item,
   * in the smallest units
   */
  get cryptoAmount() {
    return this.raw.cryptoAmount;
  }
  /** Fiat currency to which the crypto currency was converted */
  get currency() {
    return this.raw.currency;
  }
  /**
   * Converted amount in fiat currency,
   * in the smallest units (e.g. cents)
   */
  get amount() {
    return this.raw.amount;
  }
  /** URL to the collectible on Fragment */
  get url() {
    return this.raw.url;
  }
}
const CollectibleInfo$1 = /* @__PURE__ */ inspectable.makeInspectable(CollectibleInfo);
exports.CollectibleInfo = CollectibleInfo$1;
