// @mtcute/core/highlevel/types/misc/sticker-set.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const longUtils = require("../../../utils/long-utils.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const memoize = require("../../utils/memoize.cjs");
const errors$1 = require("../errors.cjs");
const documentUtils = require("../media/document-utils.cjs");
const thumbnail = require("../media/thumbnail.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
function normalizeInputStickerSet(input) {
  if (typeof input === "string") {
    return {
      _: "inputStickerSetShortName",
      shortName: input
    };
  }
  if ("_" in input) return input;
  if (input instanceof StickerSet) return input.inputStickerSet;
  if ("dice" in input) {
    return {
      _: "inputStickerSetDice",
      emoticon: input.dice
    };
  }
  switch (input.system) {
    case "animated":
      return { _: "inputStickerSetAnimatedEmoji" };
    case "animated_animations":
      return { _: "inputStickerSetAnimatedEmojiAnimations" };
    case "premium_gifts":
      return { _: "inputStickerSetPremiumGifts" };
    case "generic_animations":
      return { _: "inputStickerSetEmojiGenericAnimations" };
    case "default_statuses":
      return { _: "inputStickerSetEmojiDefaultStatuses" };
    case "default_topic_icons":
      return { _: "inputStickerSetEmojiDefaultTopicIcons" };
    case "default_channel_statuses":
      return { _: "inputStickerSetEmojiChannelDefaultStatuses" };
  }
}
function parseStickerOrThrow(doc) {
  typeAssertions.assertTypeIs("parseStickerOrThrow", doc, "document");
  const sticker = documentUtils.parseSticker(doc);
  if (!sticker) {
    throw new errors.MtTypeAssertionError("full.documents", "sticker", "not a sticker");
  }
  return sticker;
}
class StickerSet {
  brief;
  full;
  cover;
  /**
   * Whether this object contains information about stickers inside the set
   */
  isFull;
  constructor(raw) {
    if (raw._ === "messages.stickerSet") {
      this.full = raw;
      this.brief = raw.set;
    } else if (raw._ === "stickerSet") {
      this.brief = raw;
    } else if (innerTl.isAnyStickerSetCovered(raw)) {
      this.cover = raw;
      this.brief = raw.set;
    } else {
      throw new errors.MtTypeAssertionError("StickerSet", "messages.stickerSet | stickerSet", raw._);
    }
    this.isFull = raw._ === "messages.stickerSet";
  }
  /**
   * Whether this sticker set was archived
   * (due to too many saved stickers in the current account)
   */
  get isArchived() {
    return this.brief.archived;
  }
  /**
   * Whether this stickerset is official
   */
  get isOfficial() {
    return this.brief.official;
  }
  /**
   * Whether this sticker set was created by the current user
   */
  get isCreator() {
    return this.brief.creator;
  }
  /**
   * Type of the stickers in this set
   */
  get type() {
    if (this.brief.masks) {
      return "mask";
    }
    if (this.brief.emojis) {
      return "emoji";
    }
    return "sticker";
  }
  /**
   * Date when this sticker set was installed
   */
  get installedDate() {
    return this.brief.installedDate ? new Date(this.brief.installedDate * 1e3) : null;
  }
  /**
   * Number of stickers in this sticker set
   */
  get count() {
    return this.brief.count;
  }
  /**
   * Input sticker set to be used in other API methods
   */
  get inputStickerSet() {
    return {
      _: "inputStickerSetID",
      id: this.brief.id,
      accessHash: this.brief.accessHash
    };
  }
  /**
   * Title of the sticker set
   */
  get title() {
    return this.brief.title;
  }
  /**
   * Short name of sticker set to use in `tg://addstickers?set=short_name`
   * or `https://t.me/addstickers/short_name`
   */
  get shortName() {
    return this.brief.shortName;
  }
  /**
   * List of stickers inside this sticker set
   *
   * @throws MtEmptyError
   *     In case this object does not contain info about stickers (i.e. {@link isFull} = false)
   */
  get stickers() {
    if (!this.full) throw new errors$1.MtEmptyError();
    const stickers = [];
    const index = new longUtils.LongMap();
    this.full.documents.forEach((doc) => {
      const sticker = parseStickerOrThrow(doc);
      const info = {
        alt: sticker.emoji,
        emoji: "",
        // populated later
        sticker
      };
      stickers.push(info);
      index.set(doc.id, info);
    });
    this.full.packs.forEach((pack) => {
      pack.documents.forEach((id) => {
        const item = index.get(id);
        if (item) {
          item.emoji += pack.emoticon;
        }
      });
    });
    return stickers;
  }
  /** Cover stickers of the sticker set. Not the same as {@link thumbnails} */
  get covers() {
    if (!this.cover) return [];
    switch (this.cover._) {
      case "stickerSetCovered":
        return [parseStickerOrThrow(this.cover.cover)];
      case "stickerSetMultiCovered":
        return this.cover.covers.map((it) => parseStickerOrThrow(it));
      case "stickerSetFullCovered":
        return this.cover.documents.map((it) => parseStickerOrThrow(it));
      case "stickerSetNoCovered":
        return [];
    }
  }
  /**
   * Available sticker set thumbnails.
   *
   * Returns empty array if not available
   * (i.e. first sticker should be used as thumbnail)
   */
  get thumbnails() {
    return this.brief.thumbs?.map((sz) => new thumbnail.Thumbnail(this.brief, sz)) ?? [];
  }
  /**
   * Get a sticker set thumbnail by its type.
   *
   * Thumbnail types are described in the
   * [Telegram docs](https://core.telegram.org/api/files#image-thumbnail-types),
   * and are also available as static members of {@link Thumbnail} for convenience.
   *
   * @param type  Thumbnail type
   */
  getThumbnail(type) {
    return this.thumbnails.find((it) => it.type === type) ?? null;
  }
  /**
   * Find stickers given their emoji.
   *
   * @param emoji  Emoji to search for
   * @throws MtEmptyError
   *     In case this object does not contain info about stickers (i.e. {@link isFull} = false)
   */
  getStickersByEmoji(emoji) {
    return this.stickers.filter((it) => it.alt === emoji || it.emoji.includes(emoji));
  }
  _getInputDocument(idx) {
    if (!this.full) throw new errors$1.MtEmptyError();
    if (idx < 0) idx = this.full.documents.length + idx;
    const doc = this.full.documents[idx];
    if (!doc || doc._ !== "document") {
      throw new RangeError(`Sticker set does not have sticker ${idx}`);
    }
    return {
      _: "inputDocument",
      id: doc.id,
      accessHash: doc.accessHash,
      fileReference: doc.fileReference
    };
  }
}
const StickerSet$1 = /* @__PURE__ */ memoize.memoizeGetters(StickerSet, ["thumbnails", "stickers"]);
const StickerSet$2 = /* @__PURE__ */ inspectable.makeInspectable(StickerSet$1, ["isFull"]);
exports.StickerSet = StickerSet$2;
exports.normalizeInputStickerSet = normalizeInputStickerSet;
