// @mtcute/core/highlevel/types/misc/takeout-session.js
import { assertTrue } from "../../../utils/type-assertions.js";
import { makeInspectable } from "../../utils/inspectable.js";
class TakeoutSession {
  constructor(client, session) {
    this.client = client;
    this.id = session.id;
  }
  /**
   * Takeout session id
   */
  id;
  /**
   * Make an API call using this takeout session
   *
   * This method just wraps the query into `invokeWithTakeout`
   * and passes the control down to {@link TelegramClient.call}.
   *
   * @param message  RPC method to call
   * @param params  Additional call parameters
   */
  async call(message, params) {
    let query = {
      _: "invokeWithTakeout",
      takeoutId: this.id,
      query: message
    };
    if (params?.withMessageRange) {
      query = {
        _: "invokeWithMessagesRange",
        range: params.withMessageRange,
        query
      };
    }
    return this.client.call(query, params);
  }
  /**
   * Create a proxy over {@link TelegramClient}
   * that will use this takeout session to call methods.
   *
   * You can optionally provide a function to check if some
   * RPC method should be called via a takeout session or directly,
   * otherwise all methods are called through the takeout session.
   */
  createProxy(params) {
    const boundCall = params ? (obj, theirParams) => {
      const mergedParams = theirParams ? { ...params, ...theirParams } : params;
      if (!params.predicate || params.predicate?.(obj)) {
        return this.call(obj, mergedParams);
      }
      return this.client.call(obj, params);
    } : this.call.bind(this);
    return new Proxy(this.client, {
      get(target, prop, receiver) {
        if (prop === "call") return boundCall;
        return Reflect.get(target, prop, receiver);
      }
    });
  }
  /**
   * Finish account takeout session
   *
   * @param success  Whether the data was successfully exported
   */
  async finish(success = true) {
    const r = await this.call({
      _: "account.finishTakeoutSession",
      success
    });
    assertTrue("account.finishTakeoutSession", r);
  }
}
const TakeoutSession$1 = /* @__PURE__ */ makeInspectable(TakeoutSession);
export {
  TakeoutSession$1 as TakeoutSession
};
