// @mtcute/core/utils/crypto/keys.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const rsaKeys = require("../../tl/binary/rsa-keys.cjs");
const asn1Parser = require("../binary/asn1-parser.cjs");
function parsePublicKey(crypto, key, old = false) {
  const asn1 = asn1Parser.parseAsn1(asn1Parser.parsePemContents(key));
  const modulus = asn1.children?.[0].value;
  const exponent = asn1.children?.[1].value;
  if (!modulus || !exponent) throw new Error("Invalid public key");
  const writer = tlRuntime.TlBinaryWriter.manual(512);
  writer.bytes(modulus);
  writer.bytes(exponent);
  const data = writer.result();
  const sha = crypto.sha1(data);
  const fp = utils.hex.encode(sha.slice(-8).reverse());
  return {
    modulus: utils.hex.encode(modulus),
    exponent: utils.hex.encode(exponent),
    fingerprint: fp,
    old
  };
}
function addPublicKey(crypto, key, old = false) {
  const parsed = parsePublicKey(crypto, key, old);
  rsaKeys.__publicKeyIndex[parsed.fingerprint] = parsed;
}
function findKeyByFingerprints(fingerprints, allowOld = false) {
  for (let fp of fingerprints) {
    if (typeof fp !== "string") {
      fp = fp.toUnsigned().toString(16);
    }
    if (fp in rsaKeys.__publicKeyIndex) {
      if (rsaKeys.__publicKeyIndex[fp].old && !allowOld) continue;
      return rsaKeys.__publicKeyIndex[fp];
    }
  }
  if (!allowOld) return findKeyByFingerprints(fingerprints, true);
  return null;
}
exports.addPublicKey = addPublicKey;
exports.findKeyByFingerprints = findKeyByFingerprints;
exports.parsePublicKey = parsePublicKey;
