// @mtcute/core/utils/crypto/keys.js
import { hex } from "@fuman/utils";
import { TlBinaryWriter } from "@mtcute/tl-runtime";
import { __publicKeyIndex } from "../../tl/binary/rsa-keys.js";
import { parseAsn1, parsePemContents } from "../binary/asn1-parser.js";
function parsePublicKey(crypto, key, old = false) {
  const asn1 = parseAsn1(parsePemContents(key));
  const modulus = asn1.children?.[0].value;
  const exponent = asn1.children?.[1].value;
  if (!modulus || !exponent) throw new Error("Invalid public key");
  const writer = TlBinaryWriter.manual(512);
  writer.bytes(modulus);
  writer.bytes(exponent);
  const data = writer.result();
  const sha = crypto.sha1(data);
  const fp = hex.encode(sha.slice(-8).reverse());
  return {
    modulus: hex.encode(modulus),
    exponent: hex.encode(exponent),
    fingerprint: fp,
    old
  };
}
function addPublicKey(crypto, key, old = false) {
  const parsed = parsePublicKey(crypto, key, old);
  __publicKeyIndex[parsed.fingerprint] = parsed;
}
function findKeyByFingerprints(fingerprints, allowOld = false) {
  for (let fp of fingerprints) {
    if (typeof fp !== "string") {
      fp = fp.toUnsigned().toString(16);
    }
    if (fp in __publicKeyIndex) {
      if (__publicKeyIndex[fp].old && !allowOld) continue;
      return __publicKeyIndex[fp];
    }
  }
  if (!allowOld) return findKeyByFingerprints(fingerprints, true);
  return null;
}
export {
  addPublicKey,
  findKeyByFingerprints,
  parsePublicKey
};
