// @mtcute/core/utils/crypto/password.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("../../types/errors.cjs");
const typeAssertions = require("../type-assertions.cjs");
async function computePasswordHash(crypto, password, salt1, salt2) {
  const SH = (data, salt) => crypto.sha256(utils.u8.concat3(salt, data, salt));
  const PH1 = (pwd, salt12, salt22) => SH(SH(pwd, salt12), salt22);
  return SH(await crypto.pbkdf2(PH1(password, salt1, salt2), salt1, 1e5), salt2);
}
async function computeNewPasswordHash(crypto, algo, password) {
  typeAssertions.assertTypeIs("account.getPassword", algo, "passwordKdfAlgoSHA256SHA256PBKDF2HMACSHA512iter100000SHA256ModPow");
  const salt1 = utils.u8.alloc(algo.salt1.length + 32);
  salt1.set(algo.salt1);
  crypto.randomFill(salt1.subarray(algo.salt1.length));
  algo.salt1 = salt1;
  const _x = await computePasswordHash(crypto, utils.utf8.encoder.encode(password), algo.salt1, algo.salt2);
  const g = BigInt(algo.g);
  const p = utils.bigint.fromBytes(algo.p);
  const x = utils.bigint.fromBytes(_x);
  return utils.bigint.toBytes(utils.bigint.modPowBinary(g, x, p), 256);
}
async function computeSrpParams(crypto, request, password) {
  if (!request.currentAlgo || request.currentAlgo._ !== "passwordKdfAlgoSHA256SHA256PBKDF2HMACSHA512iter100000SHA256ModPow") {
    throw new errors.MtUnsupportedError(`Unknown algo ${request.currentAlgo?._}`);
  }
  const algo = request.currentAlgo;
  if (!request.srpB) {
    throw new errors.MtSecurityError("SRP_B is not present in the request");
  }
  if (!request.srpId) {
    throw new errors.MtSecurityError("SRP_ID is not present in the request");
  }
  const g = BigInt(algo.g);
  const _g = utils.bigint.toBytes(g, 256);
  const p = utils.bigint.fromBytes(algo.p);
  const gB = utils.bigint.fromBytes(request.srpB);
  const a = utils.bigint.fromBytes(crypto.randomBytes(256));
  const gA = utils.bigint.modPowBinary(g, a, p);
  const _gA = utils.bigint.toBytes(gA, 256);
  const H = (data) => crypto.sha256(data);
  const _k = crypto.sha256(utils.u8.concat2(algo.p, _g));
  const _u = crypto.sha256(utils.u8.concat2(_gA, request.srpB));
  const _x = await computePasswordHash(crypto, utils.utf8.encoder.encode(password), algo.salt1, algo.salt2);
  const k = utils.bigint.fromBytes(_k);
  const u = utils.bigint.fromBytes(_u);
  const x = utils.bigint.fromBytes(_x);
  const v = utils.bigint.modPowBinary(g, x, p);
  const kV = k * v % p;
  let t = gB - kV;
  if (t < 0n) t += p;
  const sA = utils.bigint.modPowBinary(t, a + u * x, p);
  const _kA = H(utils.bigint.toBytes(sA, 256));
  const _M1 = H(utils.u8.concat([
    utils.u8.xor(H(algo.p), H(_g)),
    H(algo.salt1),
    H(algo.salt2),
    _gA,
    request.srpB,
    _kA
  ]));
  return {
    _: "inputCheckPasswordSRP",
    srpId: request.srpId,
    A: _gA,
    M1: _M1
  };
}
exports.computeNewPasswordHash = computeNewPasswordHash;
exports.computePasswordHash = computePasswordHash;
exports.computeSrpParams = computeSrpParams;
