// @fuman/node/net/connection.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const net = require("@fuman/net");
const utils = require("@fuman/utils");
class NodeSocketConnection {
  constructor(socket, options = {}) {
    this.socket = socket;
    if (socket.pending) {
      throw new Error("socket is not connected");
    }
    if (socket.destroyed) {
      throw new Error("socket is destroyed");
    }
    this.#buffer = io.Bytes.alloc(options.bufferSize);
    socket.resume();
    socket.on("data", (data) => {
      this.#buffer.writeSync(data.length).set(data);
      this.#cv.notify();
    });
    socket.on("close", () => {
      this.#error = new net.ConnectionClosedError();
      this.#cv.notify();
    });
    socket.on("error", (error) => {
      this.#error = utils.unknownToError(error);
      this.#cv.notify();
    });
  }
  #error = null;
  #buffer;
  #cv = new utils.ConditionVariable();
  close() {
    this.socket?.destroy();
    this.#error = new net.ConnectionClosedError();
    this.#cv.notify();
  }
  async read(into) {
    if (this.#buffer.available > 0) {
      const size2 = Math.min(this.#buffer.available, into.length);
      into.set(this.#buffer.readSync(size2));
      this.#buffer.reclaim();
      return size2;
    }
    if (this.#error !== null) throw this.#error;
    await this.#cv.wait();
    if (this.#error !== null) throw this.#error;
    const size = Math.min(this.#buffer.available, into.length);
    into.set(this.#buffer.readSync(size));
    this.#buffer.reclaim();
    return size;
  }
  async write(bytes) {
    if (this.#error) throw this.#error;
    return new Promise((resolve, reject) => {
      this.socket.write(bytes, (error) => {
        if (error) reject(error);
        else resolve();
      });
    });
  }
}
class TcpConnection extends NodeSocketConnection {
  get localAddress() {
    return {
      address: utils.asNonNull(this.socket.localAddress),
      port: utils.asNonNull(this.socket.localPort)
    };
  }
  get remoteAddress() {
    return {
      address: utils.asNonNull(this.socket.remoteAddress),
      port: utils.asNonNull(this.socket.remotePort)
    };
  }
  setKeepAlive(keepAlive) {
    this.socket.setKeepAlive(keepAlive);
  }
  setNoDelay(noDelay) {
    this.socket.setNoDelay(noDelay);
  }
}
class TlsConnection extends NodeSocketConnection {
  get localAddress() {
    return {
      address: utils.asNonNull(this.socket.localAddress),
      port: utils.asNonNull(this.socket.localPort)
    };
  }
  get remoteAddress() {
    return {
      address: utils.asNonNull(this.socket.remoteAddress),
      port: utils.asNonNull(this.socket.remotePort)
    };
  }
  setKeepAlive(keepAlive) {
    this.socket.setKeepAlive(keepAlive);
  }
  setNoDelay(noDelay) {
    this.socket.setNoDelay(noDelay);
  }
  getAlpnProtocol() {
    const proto = this.socket.alpnProtocol;
    if (proto === false) return null;
    return proto;
  }
}
exports.TcpConnection = TcpConnection;
exports.TlsConnection = TlsConnection;
