// @fuman/node/net/connection.js
import { Bytes } from "@fuman/io";
import { ConnectionClosedError } from "@fuman/net";
import { asNonNull, unknownToError, ConditionVariable } from "@fuman/utils";
class NodeSocketConnection {
  constructor(socket, options = {}) {
    this.socket = socket;
    if (socket.pending) {
      throw new Error("socket is not connected");
    }
    if (socket.destroyed) {
      throw new Error("socket is destroyed");
    }
    this.#buffer = Bytes.alloc(options.bufferSize);
    socket.resume();
    socket.on("data", (data) => {
      this.#buffer.writeSync(data.length).set(data);
      this.#cv.notify();
    });
    socket.on("close", () => {
      this.#error = new ConnectionClosedError();
      this.#cv.notify();
    });
    socket.on("error", (error) => {
      this.#error = unknownToError(error);
      this.#cv.notify();
    });
  }
  #error = null;
  #buffer;
  #cv = new ConditionVariable();
  close() {
    this.socket?.destroy();
    this.#error = new ConnectionClosedError();
    this.#cv.notify();
  }
  async read(into) {
    if (this.#buffer.available > 0) {
      const size2 = Math.min(this.#buffer.available, into.length);
      into.set(this.#buffer.readSync(size2));
      this.#buffer.reclaim();
      return size2;
    }
    if (this.#error !== null) throw this.#error;
    await this.#cv.wait();
    if (this.#error !== null) throw this.#error;
    const size = Math.min(this.#buffer.available, into.length);
    into.set(this.#buffer.readSync(size));
    this.#buffer.reclaim();
    return size;
  }
  async write(bytes) {
    if (this.#error) throw this.#error;
    return new Promise((resolve, reject) => {
      this.socket.write(bytes, (error) => {
        if (error) reject(error);
        else resolve();
      });
    });
  }
}
class TcpConnection extends NodeSocketConnection {
  get localAddress() {
    return {
      address: asNonNull(this.socket.localAddress),
      port: asNonNull(this.socket.localPort)
    };
  }
  get remoteAddress() {
    return {
      address: asNonNull(this.socket.remoteAddress),
      port: asNonNull(this.socket.remotePort)
    };
  }
  setKeepAlive(keepAlive) {
    this.socket.setKeepAlive(keepAlive);
  }
  setNoDelay(noDelay) {
    this.socket.setNoDelay(noDelay);
  }
}
class TlsConnection extends NodeSocketConnection {
  get localAddress() {
    return {
      address: asNonNull(this.socket.localAddress),
      port: asNonNull(this.socket.localPort)
    };
  }
  get remoteAddress() {
    return {
      address: asNonNull(this.socket.remoteAddress),
      port: asNonNull(this.socket.remotePort)
    };
  }
  setKeepAlive(keepAlive) {
    this.socket.setKeepAlive(keepAlive);
  }
  setNoDelay(noDelay) {
    this.socket.setNoDelay(noDelay);
  }
  getAlpnProtocol() {
    const proto = this.socket.alpnProtocol;
    if (proto === false) return null;
    return proto;
  }
}
export {
  TcpConnection,
  TlsConnection
};
