// @fuman/node/net/functions.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_net = require("node:net");
const node_tls = require("node:tls");
const connection = require("./connection.cjs");
const listener = require("./listener.cjs");
function awaitConnect(socket) {
  return new Promise((resolve, reject) => {
    socket.on("error", reject);
    socket.once("connect", () => {
      socket.off("error", reject);
      resolve();
    });
  });
}
const connectTcp = async ({ address, port }) => {
  const socket = new node_net.Socket();
  socket.connect(port, address);
  await awaitConnect(socket);
  return new connection.TcpConnection(socket);
};
const connectTls = async (options) => {
  const { address, port, sni, caCerts, alpnProtocols, extraOptions } = options;
  const socket = node_tls.connect({
    host: address,
    port,
    ca: caCerts,
    ALPNProtocols: alpnProtocols,
    servername: sni,
    ...extraOptions
  });
  await awaitConnect(socket);
  return new connection.TlsConnection(socket);
};
const listenTcp = async ({ address, port }) => {
  return new Promise((resolve, reject) => {
    const server = new node_net.Server();
    server.on("error", reject);
    server.listen(port, address, () => {
      server.off("error", reject);
      resolve(new listener.TcpListener(server));
    });
  });
};
const upgradeTls = async (conn, options) => {
  return new Promise((resolve, reject) => {
    const onError = (error) => {
      reject(error);
      conn.close();
    };
    const tlsSocket = node_tls.connect({
      socket: conn.socket,
      ca: options.caCerts,
      ALPNProtocols: options.alpnProtocols,
      servername: options.sni,
      ...options.extraOptions
    }, () => {
      tlsSocket.off("error", onError);
      resolve(new connection.TlsConnection(tlsSocket));
    });
    tlsSocket.on("error", onError);
  });
};
function hostToSecureContextOptions(host) {
  const { key, cert, caCerts, extraOptions } = host;
  return {
    key,
    cert,
    ca: caCerts,
    ...extraOptions
  };
}
const listenTls = async (options) => {
  const hosts = options.hosts ?? [options];
  let listenOptions;
  if (hosts.length === 1) {
    listenOptions = {
      ALPNProtocols: options.alpnProtocols,
      ...hostToSecureContextOptions(hosts[0])
    };
  } else {
    const secureContexts = /* @__PURE__ */ new Map();
    for (const host of hosts) {
      if (host.sni == null) throw new Error("SNI is required for multi-host setups");
      secureContexts.set(host.sni, node_tls.createSecureContext(hostToSecureContextOptions(host)));
    }
    listenOptions = {
      ALPNProtocols: options.alpnProtocols,
      SNICallback: (hostname, cb) => {
        const ctx = secureContexts.get(hostname);
        if (ctx) {
          cb(null, ctx);
          return;
        }
        cb(new Error("No matching host found"));
      }
    };
  }
  const socket = node_tls.createServer(listenOptions);
  socket.listen(options.port, options.address);
  return new listener.TlsListener(socket);
};
exports.connectTcp = connectTcp;
exports.connectTls = connectTls;
exports.listenTcp = listenTcp;
exports.listenTls = listenTls;
exports.upgradeTls = upgradeTls;
