// @fuman/node/net/functions.js
import { Socket, Server } from "node:net";
import { connect, createSecureContext, createServer } from "node:tls";
import { TcpConnection, TlsConnection } from "./connection.js";
import { TcpListener, TlsListener } from "./listener.js";
function awaitConnect(socket) {
  return new Promise((resolve, reject) => {
    socket.on("error", reject);
    socket.once("connect", () => {
      socket.off("error", reject);
      resolve();
    });
  });
}
const connectTcp = async ({ address, port }) => {
  const socket = new Socket();
  socket.connect(port, address);
  await awaitConnect(socket);
  return new TcpConnection(socket);
};
const connectTls = async (options) => {
  const { address, port, sni, caCerts, alpnProtocols, extraOptions } = options;
  const socket = connect({
    host: address,
    port,
    ca: caCerts,
    ALPNProtocols: alpnProtocols,
    servername: sni,
    ...extraOptions
  });
  await awaitConnect(socket);
  return new TlsConnection(socket);
};
const listenTcp = async ({ address, port }) => {
  return new Promise((resolve, reject) => {
    const server = new Server();
    server.on("error", reject);
    server.listen(port, address, () => {
      server.off("error", reject);
      resolve(new TcpListener(server));
    });
  });
};
const upgradeTls = async (conn, options) => {
  return new Promise((resolve, reject) => {
    const onError = (error) => {
      reject(error);
      conn.close();
    };
    const tlsSocket = connect({
      socket: conn.socket,
      ca: options.caCerts,
      ALPNProtocols: options.alpnProtocols,
      servername: options.sni,
      ...options.extraOptions
    }, () => {
      tlsSocket.off("error", onError);
      resolve(new TlsConnection(tlsSocket));
    });
    tlsSocket.on("error", onError);
  });
};
function hostToSecureContextOptions(host) {
  const { key, cert, caCerts, extraOptions } = host;
  return {
    key,
    cert,
    ca: caCerts,
    ...extraOptions
  };
}
const listenTls = async (options) => {
  const hosts = options.hosts ?? [options];
  let listenOptions;
  if (hosts.length === 1) {
    listenOptions = {
      ALPNProtocols: options.alpnProtocols,
      ...hostToSecureContextOptions(hosts[0])
    };
  } else {
    const secureContexts = /* @__PURE__ */ new Map();
    for (const host of hosts) {
      if (host.sni == null) throw new Error("SNI is required for multi-host setups");
      secureContexts.set(host.sni, createSecureContext(hostToSecureContextOptions(host)));
    }
    listenOptions = {
      ALPNProtocols: options.alpnProtocols,
      SNICallback: (hostname, cb) => {
        const ctx = secureContexts.get(hostname);
        if (ctx) {
          cb(null, ctx);
          return;
        }
        cb(new Error("No matching host found"));
      }
    };
  }
  const socket = createServer(listenOptions);
  socket.listen(options.port, options.address);
  return new TlsListener(socket);
};
export {
  connectTcp,
  connectTls,
  listenTcp,
  listenTls,
  upgradeTls
};
