// @fuman/node/net/listener.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const net = require("@fuman/net");
const utils = require("@fuman/utils");
const connection = require("./connection.cjs");
class NodeListener {
  constructor(server) {
    this.server = server;
    server.on("close", () => {
      this._waiter?.reject(new net.ListenerClosedError());
    });
    server.on("error", (error) => {
      this._waiter?.reject(error);
    });
  }
  #closed = false;
  _waiter;
  get address() {
    return this.mapAddress(this.server.address());
  }
  async accept() {
    if (this.#closed) {
      throw new net.ListenerClosedError();
    }
    this._waiter = new utils.Deferred();
    const connection2 = await this._waiter.promise;
    this._waiter = void 0;
    return connection2;
  }
  close() {
    this._waiter?.reject(new net.ListenerClosedError());
    this._waiter = void 0;
    this.server.close();
  }
}
class TcpListener extends NodeListener {
  constructor(server) {
    super(server);
    this.server = server;
    server.on("connection", (socket) => {
      if (!this._waiter) {
        socket.destroy();
        return;
      }
      this._waiter.resolve(new connection.TcpConnection(socket));
      this._waiter = void 0;
    });
  }
  mapAddress(addr) {
    if (addr === null || typeof addr === "string") {
      throw new Error("listener is not bound");
    }
    return {
      address: addr.address,
      port: addr.port
    };
  }
  mapConnection(socket) {
    return new connection.TcpConnection(socket);
  }
}
class TlsListener extends NodeListener {
  constructor(server) {
    super(server);
    this.server = server;
    server.on("secureConnection", (socket) => {
      if (!this._waiter) {
        socket.destroy();
        return;
      }
      this._waiter.resolve(new connection.TlsConnection(socket));
      this._waiter = void 0;
    });
  }
  mapAddress(addr) {
    if (addr === null || typeof addr === "string") {
      throw new Error("listener is not bound");
    }
    return {
      address: addr.address,
      port: addr.port
    };
  }
}
exports.TcpListener = TcpListener;
exports.TlsListener = TlsListener;
