// @fuman/node/net/listener.js
import { ListenerClosedError } from "@fuman/net";
import { Deferred } from "@fuman/utils";
import { TcpConnection, TlsConnection } from "./connection.js";
class NodeListener {
  constructor(server) {
    this.server = server;
    server.on("close", () => {
      this._waiter?.reject(new ListenerClosedError());
    });
    server.on("error", (error) => {
      this._waiter?.reject(error);
    });
  }
  #closed = false;
  _waiter;
  get address() {
    return this.mapAddress(this.server.address());
  }
  async accept() {
    if (this.#closed) {
      throw new ListenerClosedError();
    }
    this._waiter = new Deferred();
    const connection = await this._waiter.promise;
    this._waiter = void 0;
    return connection;
  }
  close() {
    this._waiter?.reject(new ListenerClosedError());
    this._waiter = void 0;
    this.server.close();
  }
}
class TcpListener extends NodeListener {
  constructor(server) {
    super(server);
    this.server = server;
    server.on("connection", (socket) => {
      if (!this._waiter) {
        socket.destroy();
        return;
      }
      this._waiter.resolve(new TcpConnection(socket));
      this._waiter = void 0;
    });
  }
  mapAddress(addr) {
    if (addr === null || typeof addr === "string") {
      throw new Error("listener is not bound");
    }
    return {
      address: addr.address,
      port: addr.port
    };
  }
  mapConnection(socket) {
    return new TcpConnection(socket);
  }
}
class TlsListener extends NodeListener {
  constructor(server) {
    super(server);
    this.server = server;
    server.on("secureConnection", (socket) => {
      if (!this._waiter) {
        socket.destroy();
        return;
      }
      this._waiter.resolve(new TlsConnection(socket));
      this._waiter = void 0;
    });
  }
  mapAddress(addr) {
    if (addr === null || typeof addr === "string") {
      throw new Error("listener is not bound");
    }
    return {
      address: addr.address,
      port: addr.port
    };
  }
}
export {
  TcpListener,
  TlsListener
};
