// @mtcute/core/highlevel/methods/files/download-stream.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const downloadIterable = require("./download-iterable.cjs");
const fileLocation = require("../../types/files/file-location.cjs");
function downloadAsStream(client, location, params) {
  if (location instanceof fileLocation.FileLocation && ArrayBuffer.isView(location.location)) {
    const buf = location.location;
    return new ReadableStream({
      start(controller) {
        controller.enqueue(buf);
        controller.close();
      }
    });
  }
  const cancel = new AbortController();
  if (params?.abortSignal) {
    params?.abortSignal.addEventListener("abort", () => {
      cancel.abort();
    });
  }
  const backpressureCv = new utils.ConditionVariable();
  return new ReadableStream({
    start(controller) {
      (async () => {
        const params_ = params ?? {};
        params_.throttle = () => {
          if (controller.desiredSize != null && controller.desiredSize <= 0) {
            return backpressureCv.wait();
          }
        };
        for await (const chunk of downloadIterable.downloadAsIterable(client, location, params)) {
          controller.enqueue(chunk);
        }
        controller.close();
      })().catch((e) => controller.error(e));
    },
    pull() {
      backpressureCv.notify();
    },
    cancel() {
      cancel.abort();
    }
  }, {
    highWaterMark: params?.highWaterMark,
    size: (chunk) => chunk?.length ?? 0
  });
}
exports.downloadAsStream = downloadAsStream;
