// @mtcute/core/highlevel/methods/files/download-iterable.js
import { ConditionVariable } from "@fuman/utils";
import { parseFileId } from "@mtcute/file-id";
import { MtArgumentError, MtUnsupportedError } from "../../../types/errors.js";
import { toggleChannelIdMark } from "../../../utils/peer-utils.js";
import { fileIdToInputWebFileLocation, fileIdToInputFileLocation } from "../../utils/convert-file-id.js";
import { determinePartSize } from "../../utils/file-utils.js";
import { FileLocation as FileLocation$1 } from "../../types/files/file-location.js";
import { MtPeerNotFoundError } from "../../types/errors.js";
import { isAnyInputWebFileLocation, RpcError } from "../../../tl/inner-tl.js";
const SMALL_FILE_MAX_SIZE = 131072;
const REQUESTS_PER_CONNECTION = 3;
async function _normalizeFileDownloadLocation(client, input) {
  let location;
  let dcId;
  let fileSize;
  if (input instanceof FileLocation$1) {
    let locationInner = input.location;
    if (typeof locationInner === "function") {
      locationInner = locationInner();
    }
    if (ArrayBuffer.isView(locationInner)) {
      return {
        location: locationInner,
        dcId: input.dcId,
        fileSize: input.fileSize
      };
    }
    location = locationInner;
    dcId = input.dcId;
    fileSize = input.fileSize;
  } else if (typeof input === "string") {
    const parsed = parseFileId(input);
    if (parsed.location._ === "web") {
      location = fileIdToInputWebFileLocation(parsed);
    } else {
      location = fileIdToInputFileLocation(parsed);
    }
  } else {
    location = input;
  }
  if (location._ === "inputPeerPhotoFileLocation") {
    if (location.peer._ === "mtcute.dummyInputPeerMinUser" || location.peer._ === "mtcute.dummyInputPeerMinChannel") {
      const peerId = location.peer._ === "mtcute.dummyInputPeerMinUser" ? location.peer.userId : toggleChannelIdMark(location.peer.channelId);
      const accessHash = await client.storage.peers.getMinAccessHash(peerId);
      if (!accessHash) {
        throw new MtPeerNotFoundError(`Peer ${peerId} not found`);
      }
      location = {
        _: "inputPeerPhotoFileLocation",
        peer: location.peer._ === "mtcute.dummyInputPeerMinUser" ? {
          _: "inputPeerUser",
          userId: location.peer.userId,
          accessHash
        } : {
          _: "inputPeerChannel",
          channelId: location.peer.channelId,
          accessHash
        },
        photoId: location.photoId,
        big: location.big
      };
    }
  }
  return {
    location,
    dcId,
    fileSize
  };
}
async function* downloadAsIterable(client, input, params) {
  const offset = params?.offset ?? 0;
  if (offset % 4096 !== 0) {
    throw new MtArgumentError(`Invalid offset: ${offset}. Must be divisible by 4096`);
  }
  const normalized = await _normalizeFileDownloadLocation(client, input);
  const fileSize = params?.fileSize ?? normalized.fileSize;
  const location = normalized.location;
  if (ArrayBuffer.isView(location)) {
    yield location;
    return;
  }
  const isWeb = isAnyInputWebFileLocation(location);
  let dcId = params?.dcId ?? normalized?.dcId;
  const primaryDcId = await client.getPrimaryDcId();
  if (!dcId) dcId = primaryDcId;
  const abortSignal = params?.abortSignal;
  let ended = false;
  const partSizeKb = params?.partSize ?? (fileSize ? determinePartSize(fileSize) : 64);
  if (partSizeKb % 4 !== 0) {
    throw new MtArgumentError(`Invalid part size: ${partSizeKb}. Must be divisible by 4.`);
  }
  const chunkSize = partSizeKb * 1024;
  let limitBytes = params?.limit ?? fileSize ?? Infinity;
  if (limitBytes === 0) return;
  if (limitBytes !== Infinity && limitBytes !== fileSize) {
    if (limitBytes % 4096 !== 0) {
      throw new MtArgumentError(`Invalid limit: ${limitBytes}. Must be divisible by 4096.`);
    }
    if (1048576 % limitBytes !== 0) {
      throw new MtArgumentError(`Invalid limit: ${limitBytes}. Must be a multiple of 1MB.`);
    }
  }
  let numChunks = limitBytes === Infinity ? Infinity : ~~((limitBytes + chunkSize - offset - 1) / chunkSize);
  let nextChunkIdx = 0;
  let nextWorkerChunkIdx = 0;
  const nextChunkCv = new ConditionVariable();
  const buffer = {};
  const isSmall = fileSize && fileSize <= SMALL_FILE_MAX_SIZE || location._ === "inputPeerPhotoFileLocation";
  let connectionKind;
  if (isSmall) {
    connectionKind = dcId === primaryDcId ? "main" : "downloadSmall";
  } else {
    connectionKind = "download";
  }
  const poolSize = await client.getPoolSize(connectionKind, dcId);
  client.log.debug(
    "Downloading file of size %d from dc %d using %s connection pool (pool size: %d)",
    limitBytes,
    dcId,
    connectionKind,
    poolSize
  );
  const downloadChunk = async (chunk = nextWorkerChunkIdx++) => {
    let result;
    if (ended) {
      return;
    }
    const throttlePromise = params?.throttle?.(chunkSize);
    if (throttlePromise) {
      await throttlePromise;
    }
    try {
      result = await client.call(
        {
          _: isWeb ? "upload.getWebFile" : "upload.getFile",
          // eslint-disable-next-line
          location,
          offset: offset + chunkSize * chunk,
          limit: chunkSize
        },
        {
          dcId,
          kind: connectionKind,
          maxRetryCount: Infinity,
          // retry until explicitly aborted (or finished)
          floodSleepThreshold: Infinity,
          abortSignal
        }
      );
    } catch (e) {
      if (e instanceof DOMException && e.name === "AbortError") return;
      if (!RpcError.is(e)) throw e;
      if (e.is("FILE_MIGRATE_%d")) {
        dcId = e.newDc;
        return downloadChunk(chunk);
      } else if (e.is("FILEREF_UPGRADE_NEEDED")) {
        throw new MtUnsupportedError("File ref expired!");
      } else if (e.is("MTPROTO_CLUSTER_INVALID") && dcId != null) {
        client.log.debug("received cluster invalid error for dc %d, recreating dc", dcId);
        await client.recreateDc(dcId);
        return downloadChunk(chunk);
      } else {
        throw e;
      }
    }
    if (result._ === "upload.fileCdnRedirect") {
      throw new MtUnsupportedError("Received CDN redirect, which is not supported (yet)");
    }
    if (ended) {
      return;
    }
    if (result._ === "upload.webFile" && result.size && limitBytes === Infinity) {
      limitBytes = result.size;
      numChunks = ~~((limitBytes + chunkSize - offset - 1) / chunkSize);
    }
    buffer[chunk] = result.bytes;
    if (chunk === nextChunkIdx) {
      nextChunkCv.notify();
    }
    if (nextWorkerChunkIdx < numChunks && result.bytes.length === chunkSize) {
      return downloadChunk();
    }
  };
  let error;
  void Promise.all(Array.from({
    length: Math.min(poolSize * (isSmall ? 1 : REQUESTS_PER_CONNECTION), numChunks)
  }, downloadChunk)).catch((e) => {
    client.log.debug("download workers errored: %e", e);
    error = e;
    ended = true;
    nextChunkCv.notify();
  }).then(() => {
    client.log.debug("download workers finished");
  });
  abortSignal?.addEventListener("abort", () => {
    client.log.debug("download aborted");
    error = abortSignal.reason;
    ended = true;
    nextChunkCv.notify();
  });
  let position = offset;
  while (position < limitBytes) {
    await nextChunkCv.wait();
    if (error) throw error;
    while (nextChunkIdx in buffer) {
      const buf = buffer[nextChunkIdx];
      delete buffer[nextChunkIdx];
      position += buf.length;
      params?.progressCallback?.(position, limitBytes);
      yield buf;
      nextChunkIdx++;
      if (buf.length < chunkSize) {
        return;
      }
    }
  }
}
export {
  _normalizeFileDownloadLocation,
  downloadAsIterable
};
