// @mtcute/core/highlevel/methods/files/download-stream.js
import { ConditionVariable } from "@fuman/utils";
import { downloadAsIterable } from "./download-iterable.js";
import { FileLocation as FileLocation$1 } from "../../types/files/file-location.js";
function downloadAsStream(client, location, params) {
  if (location instanceof FileLocation$1 && ArrayBuffer.isView(location.location)) {
    const buf = location.location;
    return new ReadableStream({
      start(controller) {
        controller.enqueue(buf);
        controller.close();
      }
    });
  }
  const cancel = new AbortController();
  if (params?.abortSignal) {
    params?.abortSignal.addEventListener("abort", () => {
      cancel.abort();
    });
  }
  const backpressureCv = new ConditionVariable();
  return new ReadableStream({
    start(controller) {
      (async () => {
        const params_ = params ?? {};
        params_.throttle = () => {
          if (controller.desiredSize != null && controller.desiredSize <= 0) {
            return backpressureCv.wait();
          }
        };
        for await (const chunk of downloadAsIterable(client, location, params)) {
          controller.enqueue(chunk);
        }
        controller.close();
      })().catch((e) => controller.error(e));
    },
    pull() {
      backpressureCv.notify();
    },
    cancel() {
      cancel.abort();
    }
  }, {
    highWaterMark: params?.highWaterMark,
    size: (chunk) => chunk?.length ?? 0
  });
}
export {
  downloadAsStream
};
