// @mtcute/core/highlevel/methods/files/normalize-input-media.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const Long = require("long");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const uploadedFile = require("../../types/files/uploaded-file.cjs");
const entities = require("../../types/misc/entities.cjs");
const convertFileId = require("../../utils/convert-file-id.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const voiceUtils = require("../../utils/voice-utils.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const normalizeInputFile = require("./normalize-input-file.cjs");
const uploadFile = require("./upload-file.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
async function _normalizeInputMedia(client, media, params = {}, uploadMedia = false) {
  if (innerTl.isAnyInputMedia(media)) {
    switch (media._) {
      case "inputMediaStory":
        return {
          ...media,
          peer: await resolvePeer.resolvePeer(client, media.peer)
        };
    }
    return media;
  }
  if (innerTl.isAnyInputPhoto(media)) {
    typeAssertions.assertTypeIs("normalizeInputMedia (@ photo)", media, "inputPhoto");
    return { _: "inputMediaPhoto", id: media };
  }
  if (innerTl.isAnyInputDocument(media)) {
    typeAssertions.assertTypeIs("normalizeInputMedia (@ document)", media, "inputDocument");
    return { _: "inputMediaDocument", id: media };
  }
  if (media.type === "venue") {
    return {
      _: "inputMediaVenue",
      geoPoint: {
        _: "inputGeoPoint",
        lat: media.latitude,
        long: media.longitude
      },
      title: media.title,
      address: media.address,
      provider: media.source?.provider ?? "",
      venueId: media.source?.id ?? "",
      venueType: media.source?.type ?? ""
    };
  }
  if (media.type === "geo") {
    return {
      _: "inputMediaGeoPoint",
      geoPoint: {
        _: "inputGeoPoint",
        lat: media.latitude,
        long: media.longitude
      }
    };
  }
  if (media.type === "geo_live") {
    return {
      _: "inputMediaGeoLive",
      geoPoint: {
        _: "inputGeoPoint",
        lat: media.latitude,
        long: media.longitude
      },
      stopped: media.stopped,
      heading: media.heading,
      period: media.period,
      proximityNotificationRadius: media.proximityNotificationRadius
    };
  }
  if (media.type === "dice") {
    return {
      _: "inputMediaDice",
      emoticon: media.emoji
    };
  }
  if (media.type === "contact") {
    return {
      _: "inputMediaContact",
      phoneNumber: media.phone,
      firstName: media.firstName,
      lastName: media.lastName ?? "",
      vcard: media.vcard ?? ""
    };
  }
  if (media.type === "game") {
    return {
      _: "inputMediaGame",
      id: typeof media.game === "string" ? {
        _: "inputGameShortName",
        botId: { _: "inputUserSelf" },
        shortName: media.game
      } : media.game
    };
  }
  if (media.type === "invoice") {
    return {
      _: "inputMediaInvoice",
      title: media.title,
      description: media.description,
      photo: typeof media.photo === "string" ? {
        _: "inputWebDocument",
        url: media.photo,
        mimeType: "image/jpeg",
        size: 0,
        attributes: []
      } : media.photo,
      invoice: media.invoice,
      payload: media.payload,
      provider: media.token,
      providerData: {
        _: "dataJSON",
        data: JSON.stringify(media.providerData)
      },
      startParam: media.startParam,
      extendedMedia: media.extendedMedia ? await _normalizeInputMedia(client, media.extendedMedia, params) : void 0
    };
  }
  if (media.type === "poll" || media.type === "quiz") {
    const answers = await Promise.all(media.answers.map((ans) => {
      if (typeof ans === "string") return { _: "inputPollAnswer", text: entities.inputTextToTl(ans) };
      if (innerTl.isAnyPollAnswer(ans)) return ans;
      if ("media" in ans && ans.media) {
        return _normalizeInputMedia(client, ans.media, params, true).then((m) => ({
          _: "inputPollAnswer",
          text: entities.inputTextToTl(ans.text),
          media: m
        }));
      }
      return {
        _: "inputPollAnswer",
        text: entities.inputTextToTl(ans.text)
      };
    }));
    let correct;
    let solution;
    let solutionEntities;
    const [attachedMedia, solutionMedia] = await Promise.all([
      media.attachedMedia ? _normalizeInputMedia(client, media.attachedMedia, params, true) : void 0,
      media.type === "quiz" && media.solutionMedia ? _normalizeInputMedia(client, media.solutionMedia, params, true) : void 0
    ]);
    if (media.type === "quiz") {
      let input2 = media.correct;
      if (!Array.isArray(input2)) input2 = [input2];
      correct = input2;
      if (media.solution) {
        [solution, solutionEntities] = await normalizeText._normalizeInputText(client, media.solution);
      }
    }
    return {
      _: "inputMediaPoll",
      attachedMedia,
      solutionMedia,
      poll: {
        _: "poll",
        closed: media.closed,
        id: Long.ZERO,
        publicVoters: media.public,
        multipleChoice: media.multiple,
        quiz: media.type === "quiz",
        question: entities.inputTextToTl(media.question),
        answers,
        closePeriod: media.closePeriod,
        closeDate: miscUtils.normalizeDate(media.closeDate),
        openAnswers: media.openAnswers,
        revotingDisabled: media.disableRevoting,
        shuffleAnswers: media.shuffleAnswers,
        hideResultsUntilClose: media.hideResultsUntilClose,
        hash: Long.ZERO
      },
      correctAnswers: correct,
      solution,
      solutionEntities
    };
  }
  if (media.type === "story") {
    return {
      _: "inputMediaStory",
      peer: await resolvePeer.resolvePeer(client, media.peer),
      id: media.id
    };
  }
  if (media.type === "webpage") {
    return {
      _: "inputMediaWebPage",
      forceLargeMedia: media.size === "large",
      forceSmallMedia: media.size === "small",
      optional: !media.required,
      url: media.url
    };
  }
  if (media.type === "paid") {
    let medias;
    if (Array.isArray(media.media)) {
      medias = await Promise.all(media.media.map((m) => _normalizeInputMedia(client, m, params)));
    } else {
      medias = [await _normalizeInputMedia(client, media.media, params)];
    }
    return {
      _: "inputMediaPaidMedia",
      starsAmount: Long.isLong(media.starsAmount) ? media.starsAmount : Long.fromNumber(media.starsAmount),
      extendedMedia: medias,
      payload: media.payload
    };
  }
  if (media.type === "todo") {
    return {
      _: "inputMediaTodo",
      todo: {
        _: "todoList",
        othersCanAppend: media.othersCanAppend,
        othersCanComplete: media.othersCanComplete,
        title: entities.inputTextToTl(media.title),
        list: media.items.map((it, idx) => ({
          _: "todoItem",
          id: idx,
          title: entities.inputTextToTl(it)
        }))
      }
    };
  }
  let inputFile;
  let thumb;
  let mime = "application/octet-stream";
  const upload = async (file) => {
    let sendMime;
    if (media.type === "sticker") {
      sendMime = media.isAnimated ? "application/x-tgsticker" : "image/webp";
    } else {
      sendMime = media.fileMime;
    }
    const uploaded = await uploadFile.uploadFile(client, {
      file,
      progressCallback: params.progressCallback,
      fileName: media.fileName,
      fileMime: sendMime,
      fileSize: media.fileSize,
      requireFileSize: media.type === "photo",
      requireExtension: media.type === "photo",
      abortSignal: params.abortSignal
    });
    inputFile = uploaded.inputFile;
    mime = uploaded.mime;
  };
  let videoCover;
  if (media.type === "video" && media.cover) {
    const inputMedia = await _normalizeInputMedia(client, media.cover, params);
    typeAssertions.assertTypeIs("uploadMediaIfNeeded", inputMedia, "inputMediaPhoto");
    videoCover = inputMedia.id;
  }
  let livePhotoVideo;
  if (media.type === "photo" && media.livePhotoVideo) {
    const inputMedia = await _normalizeInputMedia(client, media.livePhotoVideo, params, true);
    typeAssertions.assertTypeIs("uploadMediaIfNeeded", inputMedia, "inputMediaDocument");
    livePhotoVideo = inputMedia.id;
  }
  const uploadPeer = params.uploadPeer ?? { _: "inputPeerSelf" };
  const uploadMediaIfNeeded = async (inputMedia, photo) => {
    if (!uploadMedia) return inputMedia;
    const res = await client.call({
      _: "messages.uploadMedia",
      peer: uploadPeer,
      media: inputMedia,
      businessConnectionId: params.businessConnectionId
    }, { abortSignal: params.abortSignal });
    if (photo) {
      typeAssertions.assertTypeIs("normalizeInputMedia (@ messages.uploadMedia)", res, "messageMediaPhoto");
      typeAssertions.assertTypeIs("normalizeInputMedia (@ messages.uploadMedia)", res.photo, "photo");
      return {
        _: "inputMediaPhoto",
        id: {
          _: "inputPhoto",
          id: res.photo.id,
          accessHash: res.photo.accessHash,
          fileReference: res.photo.fileReference
        },
        ttlSeconds: media.ttlSeconds,
        spoiler: media.type === "video" && media.spoiler,
        livePhoto: Boolean(livePhotoVideo),
        video: livePhotoVideo
      };
    }
    typeAssertions.assertTypeIs("normalizeInputMedia (@ messages.uploadMedia)", res, "messageMediaDocument");
    typeAssertions.assertTypeIs("normalizeInputMedia (@ messages.uploadMedia)", res.document, "document");
    return {
      _: "inputMediaDocument",
      id: {
        _: "inputDocument",
        id: res.document.id,
        accessHash: res.document.accessHash,
        fileReference: res.document.fileReference
      },
      ttlSeconds: media.ttlSeconds,
      spoiler: media.type === "video" && media.spoiler,
      videoCover,
      videoTimestamp: media.type === "video" ? media.timestamp : void 0
    };
  };
  const input = media.file;
  if (fileId.tdFileId.isFileIdLike(input)) {
    const spoiler = (media.type === "video" || media.type === "photo") && media.spoiler;
    if (typeof input === "string" && input.match(/^https?:\/\//)) {
      return uploadMediaIfNeeded(
        {
          _: media.type === "photo" ? "inputMediaPhotoExternal" : "inputMediaDocumentExternal",
          url: input,
          ttlSeconds: media.ttlSeconds,
          spoiler
        },
        media.type === "photo"
      );
    } else if (typeof input === "string" && input.match(/^file:/)) {
      await upload(input.substring(5));
    } else {
      const parsed = typeof input === "string" ? fileId.parseFileId(input) : input;
      if (parsed.location._ === "photo") {
        return {
          _: "inputMediaPhoto",
          id: convertFileId.fileIdToInputPhoto(parsed),
          ttlSeconds: media.ttlSeconds,
          spoiler
        };
      } else if (parsed.location._ === "web") {
        return uploadMediaIfNeeded(
          {
            _: parsed.type === fileId.tdFileId.FileType.Photo ? "inputMediaPhotoExternal" : "inputMediaDocumentExternal",
            url: parsed.location.url,
            ttlSeconds: media.ttlSeconds,
            spoiler
          },
          parsed.type === fileId.tdFileId.FileType.Photo
        );
      }
      return {
        _: "inputMediaDocument",
        id: convertFileId.fileIdToInputDocument(parsed),
        ttlSeconds: media.ttlSeconds,
        spoiler
      };
    }
  } else if (typeof input === "object" && innerTl.isAnyInputMedia(input)) {
    return input;
  } else if (uploadedFile.isUploadedFile(input)) {
    inputFile = input.inputFile;
    mime = input.mime;
  } else if (typeof input === "object" && innerTl.isAnyInputFile(input)) {
    inputFile = input;
  } else {
    await upload(input);
  }
  if (!inputFile) throw new Error("should not happen");
  if (media.type === "photo") {
    return uploadMediaIfNeeded(
      {
        _: "inputMediaUploadedPhoto",
        file: inputFile,
        livePhoto: Boolean(livePhotoVideo),
        video: livePhotoVideo,
        ttlSeconds: media.ttlSeconds,
        spoiler: media.spoiler
      },
      true
    );
  }
  if ("thumb" in media && media.thumb) {
    thumb = await normalizeInputFile._normalizeInputFile(client, media.thumb, { abortSignal: params.abortSignal });
  }
  const attributes = [];
  if (media.type !== "voice") {
    attributes.push({
      _: "documentAttributeFilename",
      fileName: media.fileName || (inputFile._ === "inputFileStoryDocument" ? "story" : inputFile.name)
    });
  }
  if (media.type === "video") {
    const attr = {
      _: "documentAttributeVideo",
      duration: media.duration || 0,
      w: media.width || 0,
      h: media.height || 0,
      supportsStreaming: media.supportsStreaming,
      roundMessage: media.isRound
    };
    if (media.isRound) {
      if (attr.w === 0) attr.w = 1;
      if (attr.h === 0) attr.h = 1;
      if (attr.duration === 0) attr.duration = 1;
    }
    attributes.push(attr);
    if (media.isAnimated) {
      attributes.push({ _: "documentAttributeAnimated" });
    }
  }
  if (media.type === "audio" || media.type === "voice") {
    attributes.push({
      _: "documentAttributeAudio",
      voice: media.type === "voice",
      duration: media.duration || 0,
      title: media.type === "audio" ? media.title : void 0,
      performer: media.type === "audio" ? media.performer : void 0,
      waveform: media.type === "voice" && media.waveform ? voiceUtils.encodeWaveform(media.waveform) : void 0
    });
  }
  if (media.type === "sticker") {
    attributes.push({
      _: "documentAttributeSticker",
      stickerset: {
        _: "inputStickerSetEmpty"
      },
      alt: media.alt ?? ""
    });
  }
  return uploadMediaIfNeeded(
    {
      _: "inputMediaUploadedDocument",
      nosoundVideo: media.type === "video" && !media.isAnimated,
      forceFile: media.type === "document",
      file: inputFile,
      thumb,
      mimeType: mime,
      attributes,
      ttlSeconds: media.ttlSeconds,
      spoiler: media.type === "video" && media.spoiler,
      videoTimestamp: media.type === "video" ? media.timestamp : void 0,
      videoCover
    },
    false
  );
}
exports._normalizeInputMedia = _normalizeInputMedia;
