// @mtcute/core/highlevel/methods/files/upload-file.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const errors = require("../../../types/errors.cjs");
const longUtils = require("../../../utils/long-utils.cjs");
const fileType = require("../../utils/file-type.cjs");
const fileUtils = require("../../utils/file-utils.cjs");
const OVERRIDE_MIME = {
  // tg doesn't interpret `audio/opus` files as voice messages for some reason
  "audio/opus": "audio/ogg"
};
const SMALL_FILE_MAX_SIZE = 131072;
const BIG_FILE_MIN_SIZE = 10485760;
const DEFAULT_FILE_NAME = "unnamed";
const REQUESTS_PER_CONNECTION = 3;
const MAX_PART_COUNT = 4e3;
const MAX_PART_COUNT_PREMIUM = 8e3;
const HAS_FILE = typeof File !== "undefined";
const HAS_RESPONSE = typeof Response !== "undefined";
const HAS_URL = typeof URL !== "undefined";
const HAS_BLOB = typeof Blob !== "undefined";
async function uploadFile(client, params) {
  let file = params.file;
  let fileSize = -1;
  let fileName = params.fileName;
  let fileMime = params.fileMime;
  const abortSignal = params.abortSignal;
  if (client.platform.normalizeFile) {
    const res = await client.platform.normalizeFile(file);
    if (res?.file) {
      file = res.file;
      if (res.fileSize) fileSize = res.fileSize;
      if (res.fileName) fileName = res.fileName;
    }
  }
  if (ArrayBuffer.isView(file)) {
    fileSize = file.length;
    file = io.Bytes.from(file);
  }
  if (HAS_FILE && file instanceof File) {
    fileName = file.name;
    fileSize = file.size;
    fileMime = file.type;
    file = file.stream();
  }
  if (HAS_URL && file instanceof URL) {
    file = await fetch(file);
  }
  if (HAS_BLOB && file instanceof Blob) {
    fileSize = file.size;
    file = file.stream();
  }
  if (HAS_RESPONSE && file instanceof Response) {
    const length = Number.parseInt(file.headers.get("content-length") || "0");
    if (!Number.isNaN(length) && length) fileSize = length;
    fileMime = file.headers.get("content-type")?.split(";")[0];
    const disposition = file.headers.get("content-disposition");
    if (disposition) {
      const idx2 = disposition.indexOf("filename=");
      if (idx2 > -1) {
        const raw = disposition.slice(idx2 + 9).split(";")[0];
        fileName = JSON.parse(raw);
      }
    }
    if (fileName === DEFAULT_FILE_NAME) {
      const url = new URL(file.url);
      const name = url.pathname.split("/").pop();
      if (name && name.includes(".")) {
        fileName = name;
      }
    }
    if (!file.body) {
      throw new errors.MtArgumentError("Fetch response contains `null` body");
    }
    file = file.body;
  }
  if (file instanceof ReadableStream) {
    file = io.webReadableToFuman(file);
  } else if (!(typeof file === "object" && "read" in file)) {
    throw new errors.MtArgumentError("Could not convert input `file` to stream!");
  }
  let readable = file;
  if (fileSize === -1 && params.fileSize) fileSize = params.fileSize;
  if (fileSize === -1 && params.requireFileSize) {
    const buffer = await io.read.async.untilEnd(readable);
    fileSize = buffer.length;
    readable = io.Bytes.from(buffer);
  }
  let partSizeKb = params.partSize;
  if (!partSizeKb) {
    if (fileSize === -1) {
      partSizeKb = params.estimatedSize ? fileUtils.determinePartSize(params.estimatedSize, true) : 64;
    } else {
      partSizeKb = fileUtils.determinePartSize(fileSize, true);
    }
  }
  if (partSizeKb > 512) {
    throw new errors.MtArgumentError(`Invalid part size: ${partSizeKb}KB`);
  }
  const partSize = partSizeKb * 1024;
  let partCount = fileSize === -1 ? -1 : ~~((fileSize + partSize - 1) / partSize);
  const maxPartCount = client.storage.self.getCached()?.isPremium ? MAX_PART_COUNT_PREMIUM : MAX_PART_COUNT;
  if (partCount > maxPartCount) {
    throw new errors.MtArgumentError(`File is too large (max ${maxPartCount} parts, got ${partCount})`);
  }
  const isBig = fileSize === -1 || fileSize > BIG_FILE_MIN_SIZE;
  const isSmall = fileSize !== -1 && fileSize < SMALL_FILE_MAX_SIZE;
  const connectionKind = isSmall ? "main" : "upload";
  const connectionPoolSize = Math.min(await client.getPoolSize(connectionKind), partCount);
  const requestsPerConnection = params.requestsPerConnection ?? REQUESTS_PER_CONNECTION;
  client.log.debug(
    "uploading %d bytes file in %d chunks, each %d bytes in %s connection pool of size %d",
    fileSize,
    partCount,
    partSize,
    connectionKind,
    connectionPoolSize
  );
  const fileId = longUtils.randomLong();
  const lock = new utils.AsyncLock();
  let pos = 0;
  let idx = 0;
  let readableEnded = false;
  const uploadNextPart = async () => {
    if (abortSignal?.aborted) return;
    const thisIdx = idx++;
    await lock.acquire();
    if (readableEnded) {
      lock.release();
      return;
    }
    let part;
    try {
      part = await io.read.async.exactly(readable, partSize, "truncate");
    } finally {
      lock.release();
    }
    if (abortSignal?.aborted) return;
    const ended = part.length < partSize;
    if (ended) {
      readableEnded = true;
      if (fileSize === -1) {
        fileSize = pos + part.length;
        partCount = ~~((fileSize + partSize - 1) / partSize);
      } else if (thisIdx !== partCount - 1) {
        throw new errors.MtArgumentError(`Unexpected EOS (there were only ${idx} parts, but expected ${partCount})`);
      }
      client.log.debug("readable ended, file size = %d, part count = %d", fileSize, partCount);
    }
    if (thisIdx === 0 && fileMime === void 0) {
      const mime = fileType.guessFileMime(part);
      if (mime) {
        fileMime = mime;
      } else {
        const isPlainText = fileUtils.isProbablyPlainText(part.slice(0, 8));
        fileMime = isPlainText ? "text/plain" : "application/octet-stream";
      }
    }
    const request = isBig ? {
      _: "upload.saveBigFilePart",
      fileId,
      filePart: thisIdx,
      fileTotalParts: partCount,
      bytes: part
    } : {
      _: "upload.saveFilePart",
      fileId,
      filePart: thisIdx,
      bytes: part
    };
    const result = await client.call(request, { kind: connectionKind, abortSignal });
    if (!result) throw new Error(`Failed to upload part ${idx}`);
    if (abortSignal?.aborted) return;
    pos += part.length;
    params.progressCallback?.(pos, fileSize);
    if (idx === partCount) return;
    return uploadNextPart();
  };
  let poolSize = partCount === -1 ? 1 : connectionPoolSize * requestsPerConnection;
  if (partCount !== -1 && poolSize > partCount) poolSize = partCount;
  await Promise.all(Array.from({ length: poolSize }, uploadNextPart));
  abortSignal?.throwIfAborted();
  if (fileName === void 0) {
    const ext = fileType.MIME_TO_EXTENSION[fileMime];
    fileName = ext ? `${DEFAULT_FILE_NAME}.${ext}` : DEFAULT_FILE_NAME;
  } else if (params.requireExtension) {
    const extFromMime = fileType.MIME_TO_EXTENSION[fileMime];
    const idx2 = fileName.lastIndexOf(".");
    const extFromName = idx2 === -1 ? void 0 : fileName.slice(idx2 + 1);
    if (!extFromName) {
      if (!extFromMime) {
        throw new errors.MtArgumentError(`File name does not have an extension, and we cannot guess it from the mime type (${fileMime})`);
      }
      fileName = `${fileName}.${extFromMime}`;
    }
  }
  let inputFile;
  if (isBig) {
    inputFile = {
      _: "inputFileBig",
      id: fileId,
      parts: partCount,
      name: fileName
    };
  } else {
    inputFile = {
      _: "inputFile",
      id: fileId,
      parts: partCount,
      name: fileName,
      md5Checksum: ""
      // tdlib doesn't do this, why should we?
    };
  }
  if (fileMime in OVERRIDE_MIME) fileMime = OVERRIDE_MIME[fileMime];
  return {
    inputFile,
    size: fileSize,
    mime: fileMime
  };
}
exports.uploadFile = uploadFile;
