// @mtcute/core/highlevel/methods/dialogs/iter-dialogs.js
import Long from "long";
import { MtArgumentError } from "../../../types/errors.js";
import { normalizeDate } from "../../utils/misc-utils.js";
import { _normalizeInputFolder } from "./get-folders.js";
import { Dialog as Dialog$2 } from "../../types/messages/dialog.js";
async function* iterDialogs(client, params) {
  if (!params) params = {};
  const { limit = Infinity, chunkSize = 100, folder, filter, pinned = "include" } = params;
  let { offsetId = 0, offsetPeer = { _: "inputPeerEmpty" }, archived = "exclude" } = params;
  let offsetDate = normalizeDate(params.offsetDate) ?? 0;
  let localFilters_;
  if (folder) {
    localFilters_ = await _normalizeInputFolder(client, folder);
  }
  if (filter) {
    if (localFilters_ && localFilters_._ !== "dialogFilterDefault") {
      localFilters_ = {
        ...localFilters_,
        ...filter
      };
    } else {
      localFilters_ = {
        _: "dialogFilter",
        id: 0,
        title: { _: "textWithEntities", text: "", entities: [] },
        pinnedPeers: [],
        includePeers: [],
        excludePeers: [],
        ...params.filter
      };
    }
  }
  if (localFilters_?._ === "dialogFilterDefault") {
    localFilters_ = void 0;
  }
  const localFilters = localFilters_;
  if (localFilters?._ === "dialogFilterChatlist") {
    if (offsetId !== 0 || offsetDate !== 0 || offsetPeer._ !== "inputPeerEmpty") {
      throw new MtArgumentError("Cannot use offset parameters with chatlist filters");
    }
    let remaining = Math.min(limit, localFilters.includePeers.length + localFilters.pinnedPeers.length);
    if (pinned === "include" || pinned === "only") {
      const peers = [];
      for (const peer of localFilters.pinnedPeers) {
        if (remaining <= 0) break;
        remaining--;
        peers.push({
          _: "inputDialogPeer",
          peer
        });
      }
      const res = await client.call({
        _: "messages.getPeerDialogs",
        peers
      });
      res.dialogs.forEach((dialog) => dialog.pinned = true);
      yield* Dialog$2.parseTlDialogs(res);
    }
    if (pinned === "only" || remaining <= 0) {
      return;
    }
    let offset = 0;
    while (remaining > 0) {
      const peers = [];
      for (let i = 0; i < chunkSize; i++) {
        if (remaining <= 0) break;
        remaining--;
        peers.push({
          _: "inputDialogPeer",
          peer: localFilters.includePeers[offset + i]
        });
      }
      offset += chunkSize;
      const res = await client.call({
        _: "messages.getPeerDialogs",
        peers
      });
      yield* Dialog$2.parseTlDialogs(res);
    }
    return;
  }
  if (localFilters) {
    archived = localFilters.excludeArchived ? "exclude" : "keep";
  }
  const fetchPinnedDialogsFromFolder = async () => {
    if (!localFilters || !localFilters.pinnedPeers.length) {
      return null;
    }
    const res = await client.call({
      _: "messages.getPeerDialogs",
      peers: localFilters.pinnedPeers.map((peer) => ({
        _: "inputDialogPeer",
        peer
      }))
    });
    res.dialogs.forEach((dialog) => dialog.pinned = true);
    return res;
  };
  if (pinned === "only") {
    let res;
    if (localFilters) {
      res = await fetchPinnedDialogsFromFolder();
    } else {
      res = await client.call({
        _: "messages.getPinnedDialogs",
        folderId: archived === "exclude" ? 0 : 1
      });
    }
    if (res) yield* Dialog$2.parseTlDialogs(res, limit);
    return;
  }
  let current = 0;
  if (localFilters?.pinnedPeers.length && pinned === "include" && offsetId === 0 && offsetDate === 0 && offsetPeer._ === "inputPeerEmpty") {
    const res = await fetchPinnedDialogsFromFolder();
    if (res) {
      const dialogs = Dialog$2.parseTlDialogs(res, limit);
      for (const d of dialogs) {
        yield d;
        if (++current >= limit) return;
      }
    }
  }
  const filterFolder = localFilters ? Dialog$2.filterFolder(localFilters, pinned !== "keep") : void 0;
  let folderId;
  if (archived === "keep") {
    folderId = void 0;
  } else if (archived === "only") {
    folderId = 1;
  } else {
    folderId = 0;
  }
  for (; ; ) {
    const dialogs = Dialog$2.parseTlDialogs(
      await client.call({
        _: "messages.getDialogs",
        excludePinned: params.pinned === "exclude",
        folderId,
        offsetDate,
        offsetId,
        offsetPeer,
        limit: filterFolder ? chunkSize : Math.min(limit - current, chunkSize),
        hash: Long.ZERO
      })
    );
    if (!dialogs.length) return;
    const last = dialogs[dialogs.length - 1];
    offsetPeer = last.peer.inputPeer;
    offsetId = last.raw.topMessage;
    if (last.lastMessage) {
      offsetDate = last.lastMessage.raw.date;
    }
    for (const d of dialogs) {
      if (filterFolder && !filterFolder(d)) continue;
      if (d.isPinned) {
        if (archived === "exclude" && d.isArchived) continue;
        if (archived === "only" && !d.isArchived) continue;
      }
      yield d;
      if (++current >= limit) return;
    }
  }
}
export {
  iterDialogs
};
