// @mtcute/core/highlevel/methods/bots/answer-inline-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const factories = require("../../types/bots/inline-result/factories.cjs");
async function answerInlineQuery(client, queryId, results, params) {
  const { cacheTime = 300, gallery, private: priv, nextOffset, switchPm, switchWebview } = params ?? {};
  const [defaultGallery, tlResults] = await factories._convertToTl(client, results);
  await client.call({
    _: "messages.setInlineBotResults",
    queryId: Long.isLong(queryId) ? queryId : queryId.id,
    results: tlResults,
    cacheTime,
    gallery: gallery ?? defaultGallery,
    private: priv,
    nextOffset,
    switchPm: switchPm ? {
      _: "inlineBotSwitchPM",
      text: switchPm.text,
      startParam: switchPm.parameter
    } : void 0,
    switchWebview: switchWebview ? {
      _: "inlineBotWebView",
      text: switchWebview.text,
      url: switchWebview.url
    } : void 0
  });
}
exports.answerInlineQuery = answerInlineQuery;
