// @mtcute/core/highlevel/methods/bots/answer-bot-guest-chat-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const factories = require("../../types/bots/inline-result/factories.cjs");
async function answerBotGuestChatQuery(client, queryId, result) {
  if (!("id" in result)) {
    result = {
      type: "article",
      id: "guest",
      title: "guest",
      message: result
    };
  }
  const [, tlResults] = await factories._convertToTl(client, [result]);
  return client.call({
    _: "messages.setBotGuestChatResult",
    queryId: Long.isLong(queryId) ? queryId : queryId.id,
    result: tlResults[0]
  });
}
exports.answerBotGuestChatQuery = answerBotGuestChatQuery;
