// @mtcute/core/highlevel/methods/bots/open-webview.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils$1 = require("@fuman/utils");
const Long = require("long");
const utils = require("../../../types/utils.cjs");
const longUtils = require("../../../utils/long-utils.cjs");
const webview = require("../../types/bots/webview.cjs");
const peerUtils = require("../../utils/peer-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const _getWebviewTimerId = (queryId) => `webview:${longUtils.longToFastString(queryId)}`;
async function openWebview(client, params) {
  const {
    webview: webview$1,
    bot,
    chat,
    theme,
    platform
  } = params;
  const botPeer = await resolvePeer.resolveUser(client, bot);
  let themeObj;
  if (theme) {
    if ("_" in theme) {
      themeObj = theme;
    } else {
      themeObj = {
        _: "dataJSON",
        data: JSON.stringify(theme)
      };
    }
  }
  switch (webview$1.type) {
    case "main": {
      const chatPeer = chat != null ? await resolvePeer.resolvePeer(client, chat) : { _: "inputPeerEmpty" };
      const res = await client.call({
        _: "messages.requestMainWebView",
        compact: webview$1.compact,
        peer: chatPeer,
        bot: botPeer,
        startParam: webview$1.startParam,
        themeParams: themeObj,
        platform
      });
      return new webview.WebviewResult(res);
    }
    case "from_reply_keyboard":
    case "from_switch_inline":
    case "from_side_menu": {
      const res = await client.call({
        _: "messages.requestSimpleWebView",
        bot: botPeer,
        url: webview$1.type === "from_side_menu" ? void 0 : webview$1.url,
        themeParams: themeObj,
        platform
      });
      return new webview.WebviewResult(res);
    }
    case "from_inline_keyboard":
    case "from_bot_menu":
    case "from_attach_menu": {
      const chatPeer = chat != null ? await resolvePeer.resolvePeer(client, chat) : peerUtils.toInputPeer(botPeer);
      const tlReplyTo = typeof webview$1.replyTo === "number" ? {
        _: "inputReplyToMessage",
        replyToMsgId: webview$1.replyTo
      } : webview$1.replyTo;
      const tlSendAs = webview$1.sendAs !== void 0 ? await resolvePeer.resolvePeer(client, webview$1.sendAs) : void 0;
      const res = await client.call({
        _: "messages.requestWebView",
        bot: botPeer,
        peer: chatPeer,
        url: webview$1.type === "from_attach_menu" ? void 0 : webview$1.url,
        compact: webview$1.compact,
        themeParams: themeObj,
        platform,
        replyTo: tlReplyTo,
        sendAs: tlSendAs,
        silent: webview$1.silent
      });
      const withTimer = res.queryId != null && !webview$1.fireAndForget;
      const result = new webview.WebviewResult(res, withTimer);
      if (withTimer) {
        const queryId = utils$1.asNonNull(result.queryId);
        const timerId = _getWebviewTimerId(queryId);
        await client.timers.upsert({
          kind: "rpc",
          key: timerId,
          interval: 6e4,
          request: {
            _: "messages.prolongWebView",
            silent: webview$1.silent,
            peer: chatPeer,
            bot: botPeer,
            queryId,
            replyTo: tlReplyTo,
            sendAs: tlSendAs
          }
        });
      }
      return result;
    }
    case "from_link": {
      const chatPeer = chat != null ? await resolvePeer.resolvePeer(client, chat) : peerUtils.toInputPeer(botPeer);
      const res = await client.call({
        _: "messages.requestAppWebView",
        writeAllowed: webview$1.allowWrite,
        compact: webview$1.compact,
        peer: chatPeer,
        app: {
          _: "inputBotAppShortName",
          botId: botPeer,
          shortName: webview$1.shortName
        },
        startParam: webview$1.startParam,
        themeParams: themeObj,
        platform
      });
      const withTimer = res.queryId != null && !webview$1.fireAndForget;
      const result = new webview.WebviewResult(res, withTimer);
      if (withTimer) {
        const queryId = utils$1.asNonNull(result.queryId);
        const timerId = _getWebviewTimerId(queryId);
        await client.timers.upsert({
          kind: "rpc",
          key: timerId,
          interval: 6e4,
          request: {
            _: "messages.prolongWebView",
            peer: chatPeer,
            bot: botPeer,
            queryId
          }
        });
      }
      return result;
    }
    default:
      utils.assertNever();
  }
}
async function closeWebview(client, webview2) {
  const timerId = _getWebviewTimerId(Long.isLong(webview2) ? webview2 : utils$1.asNonNull(webview2.queryId));
  await client.timers.cancel(timerId);
}
exports.closeWebview = closeWebview;
exports.openWebview = openWebview;
