// @mtcute/core/highlevel/methods/bots/open-webview.js
import { asNonNull } from "@fuman/utils";
import Long from "long";
import { assertNever } from "../../../types/utils.js";
import { longToFastString } from "../../../utils/long-utils.js";
import { WebviewResult as WebviewResult$1 } from "../../types/bots/webview.js";
import { toInputPeer } from "../../utils/peer-utils.js";
import { resolveUser, resolvePeer } from "../users/resolve-peer.js";
const _getWebviewTimerId = (queryId) => `webview:${longToFastString(queryId)}`;
async function openWebview(client, params) {
  const {
    webview,
    bot,
    chat,
    theme,
    platform
  } = params;
  const botPeer = await resolveUser(client, bot);
  let themeObj;
  if (theme) {
    if ("_" in theme) {
      themeObj = theme;
    } else {
      themeObj = {
        _: "dataJSON",
        data: JSON.stringify(theme)
      };
    }
  }
  switch (webview.type) {
    case "main": {
      const chatPeer = chat != null ? await resolvePeer(client, chat) : { _: "inputPeerEmpty" };
      const res = await client.call({
        _: "messages.requestMainWebView",
        compact: webview.compact,
        peer: chatPeer,
        bot: botPeer,
        startParam: webview.startParam,
        themeParams: themeObj,
        platform
      });
      return new WebviewResult$1(res);
    }
    case "from_reply_keyboard":
    case "from_switch_inline":
    case "from_side_menu": {
      const res = await client.call({
        _: "messages.requestSimpleWebView",
        bot: botPeer,
        url: webview.type === "from_side_menu" ? void 0 : webview.url,
        themeParams: themeObj,
        platform
      });
      return new WebviewResult$1(res);
    }
    case "from_inline_keyboard":
    case "from_bot_menu":
    case "from_attach_menu": {
      const chatPeer = chat != null ? await resolvePeer(client, chat) : toInputPeer(botPeer);
      const tlReplyTo = typeof webview.replyTo === "number" ? {
        _: "inputReplyToMessage",
        replyToMsgId: webview.replyTo
      } : webview.replyTo;
      const tlSendAs = webview.sendAs !== void 0 ? await resolvePeer(client, webview.sendAs) : void 0;
      const res = await client.call({
        _: "messages.requestWebView",
        bot: botPeer,
        peer: chatPeer,
        url: webview.type === "from_attach_menu" ? void 0 : webview.url,
        compact: webview.compact,
        themeParams: themeObj,
        platform,
        replyTo: tlReplyTo,
        sendAs: tlSendAs,
        silent: webview.silent
      });
      const withTimer = res.queryId != null && !webview.fireAndForget;
      const result = new WebviewResult$1(res, withTimer);
      if (withTimer) {
        const queryId = asNonNull(result.queryId);
        const timerId = _getWebviewTimerId(queryId);
        await client.timers.upsert({
          kind: "rpc",
          key: timerId,
          interval: 6e4,
          request: {
            _: "messages.prolongWebView",
            silent: webview.silent,
            peer: chatPeer,
            bot: botPeer,
            queryId,
            replyTo: tlReplyTo,
            sendAs: tlSendAs
          }
        });
      }
      return result;
    }
    case "from_link": {
      const chatPeer = chat != null ? await resolvePeer(client, chat) : toInputPeer(botPeer);
      const res = await client.call({
        _: "messages.requestAppWebView",
        writeAllowed: webview.allowWrite,
        compact: webview.compact,
        peer: chatPeer,
        app: {
          _: "inputBotAppShortName",
          botId: botPeer,
          shortName: webview.shortName
        },
        startParam: webview.startParam,
        themeParams: themeObj,
        platform
      });
      const withTimer = res.queryId != null && !webview.fireAndForget;
      const result = new WebviewResult$1(res, withTimer);
      if (withTimer) {
        const queryId = asNonNull(result.queryId);
        const timerId = _getWebviewTimerId(queryId);
        await client.timers.upsert({
          kind: "rpc",
          key: timerId,
          interval: 6e4,
          request: {
            _: "messages.prolongWebView",
            peer: chatPeer,
            bot: botPeer,
            queryId
          }
        });
      }
      return result;
    }
    default:
      assertNever();
  }
}
async function closeWebview(client, webview) {
  const timerId = _getWebviewTimerId(Long.isLong(webview) ? webview : asNonNull(webview.queryId));
  await client.timers.cancel(timerId);
}
export {
  closeWebview,
  openWebview
};
