// @mtcute/core/highlevel/methods/bots/set-game-score.js
import { assertTrue } from "../../../utils/type-assertions.js";
import { normalizeInlineId } from "../../utils/inline-utils.js";
import { _findMessageInUpdate } from "../messages/find-in-update.js";
import { resolveUser, resolvePeer } from "../users/resolve-peer.js";
import { normalizeInputMessageId } from "../../types/messages/input-message-id.js";
async function setGameScore(client, params) {
  const { userId, score, noEdit, force, shouldDispatch } = params;
  const { chatId, message } = normalizeInputMessageId(params);
  const user = await resolveUser(client, userId);
  const chat = await resolvePeer(client, chatId);
  const res = await client.call({
    _: "messages.setGameScore",
    peer: chat,
    id: message,
    userId: user,
    score,
    editMessage: !noEdit,
    force
  });
  return _findMessageInUpdate(client, res, true, !shouldDispatch);
}
async function setInlineGameScore(client, params) {
  const { messageId, userId, score, noEdit, force } = params;
  const user = await resolveUser(client, userId);
  const id = normalizeInlineId(messageId);
  const r = await client.call(
    {
      _: "messages.setInlineGameScore",
      id,
      userId: user,
      score,
      editMessage: !noEdit,
      force
    },
    { dcId: id.dcId }
  );
  assertTrue("messages.setInlineGameScore", r);
}
export {
  setGameScore,
  setInlineGameScore
};
