// @mtcute/core/highlevel/methods/bots/set-game-score.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const inlineUtils = require("../../utils/inline-utils.cjs");
const findInUpdate = require("../messages/find-in-update.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const inputMessageId = require("../../types/messages/input-message-id.cjs");
async function setGameScore(client, params) {
  const { userId, score, noEdit, force, shouldDispatch } = params;
  const { chatId, message } = inputMessageId.normalizeInputMessageId(params);
  const user = await resolvePeer.resolveUser(client, userId);
  const chat = await resolvePeer.resolvePeer(client, chatId);
  const res = await client.call({
    _: "messages.setGameScore",
    peer: chat,
    id: message,
    userId: user,
    score,
    editMessage: !noEdit,
    force
  });
  return findInUpdate._findMessageInUpdate(client, res, true, !shouldDispatch);
}
async function setInlineGameScore(client, params) {
  const { messageId, userId, score, noEdit, force } = params;
  const user = await resolvePeer.resolveUser(client, userId);
  const id = inlineUtils.normalizeInlineId(messageId);
  const r = await client.call(
    {
      _: "messages.setInlineGameScore",
      id,
      userId: user,
      score,
      editMessage: !noEdit,
      force
    },
    { dcId: id.dcId }
  );
  typeAssertions.assertTrue("messages.setInlineGameScore", r);
}
exports.setGameScore = setGameScore;
exports.setInlineGameScore = setInlineGameScore;
