// @mtcute/core/highlevel/types/bots/inline-message/factories.js
import { assertNever } from "../../../../types/utils.js";
import { _normalizeInputRichMessage } from "../../../methods/messages/normalize-rich-message.js";
import { _normalizeInputText } from "../../../methods/misc/normalize-text.js";
import { _convertToTl as _convertToTl$1 } from "../keyboards/factories.js";
function text(text2, params = {}) {
  const ret = params;
  ret.type = "text";
  ret.text = text2;
  return ret;
}
function rich(rich2, params = {}) {
  const ret = params;
  ret.type = "rich";
  ret.content = rich2;
  return ret;
}
function media(params = {}) {
  const ret = params;
  ret.type = "media";
  return ret;
}
function geo(params) {
  const ret = params;
  ret.type = "geo";
  return ret;
}
function geoLive(params) {
  const ret = params;
  ret.type = "geo_live";
  return ret;
}
function venue(params) {
  const ret = params;
  ret.type = "venue";
  return ret;
}
function game(params) {
  const ret = params;
  ret.type = "game";
  return ret;
}
function contact(params) {
  const ret = params;
  ret.type = "contact";
  return ret;
}
function webpage(params) {
  const ret = params;
  ret.type = "webpage";
  return ret;
}
async function _convertToTl(client, obj) {
  switch (obj.type) {
    case "text": {
      const [message, entities] = await _normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageText",
        message,
        entities,
        replyMarkup: _convertToTl$1(obj.replyMarkup),
        invertMedia: obj.invertMedia
      };
    }
    case "rich":
      return {
        _: "inputBotInlineMessageRichMessage",
        replyMarkup: _convertToTl$1(obj.replyMarkup),
        richMessage: "_" in obj.content ? obj.content : await _normalizeInputRichMessage(client, { _: "inputPeerEmpty" }, obj.content, {})
      };
    case "media": {
      const [message, entities] = await _normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageMediaAuto",
        message,
        entities,
        replyMarkup: _convertToTl$1(obj.replyMarkup),
        invertMedia: obj.invertMedia
      };
    }
    case "geo":
    case "geo_live":
      return {
        _: "inputBotInlineMessageMediaGeo",
        geoPoint: {
          _: "inputGeoPoint",
          lat: obj.latitude,
          long: obj.longitude
        },
        // fields will be `undefined` if this is a `geo`
        heading: obj.heading,
        period: obj.period,
        proximityNotificationRadius: obj.proximityNotificationRadius,
        replyMarkup: _convertToTl$1(obj.replyMarkup)
      };
    case "venue":
      return {
        _: "inputBotInlineMessageMediaVenue",
        geoPoint: {
          _: "inputGeoPoint",
          lat: obj.latitude,
          long: obj.longitude
        },
        title: obj.title,
        address: obj.address,
        provider: obj.source?.provider ?? "",
        venueId: obj.source?.id ?? "",
        venueType: obj.source?.type ?? "",
        replyMarkup: _convertToTl$1(obj.replyMarkup)
      };
    case "game":
      return {
        _: "inputBotInlineMessageGame",
        replyMarkup: _convertToTl$1(obj.replyMarkup)
      };
    case "contact":
      return {
        _: "inputBotInlineMessageMediaContact",
        phoneNumber: obj.phone,
        firstName: obj.firstName,
        lastName: obj.lastName ?? "",
        vcard: obj.vcard ?? "",
        replyMarkup: _convertToTl$1(obj.replyMarkup)
      };
    case "webpage": {
      const [message, entities] = await _normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageMediaWebPage",
        message,
        entities,
        replyMarkup: _convertToTl$1(obj.replyMarkup),
        invertMedia: obj.invertMedia,
        forceLargeMedia: obj.size === "large",
        forceSmallMedia: obj.size === "small",
        optional: !obj.required,
        url: obj.url
      };
    }
    default:
      assertNever();
  }
}
export {
  _convertToTl,
  contact,
  game,
  geo,
  geoLive,
  media,
  rich,
  text,
  venue,
  webpage
};
