// @mtcute/core/highlevel/types/bots/inline-message/factories.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("../../../../types/utils.cjs");
const normalizeRichMessage = require("../../../methods/messages/normalize-rich-message.cjs");
const normalizeText = require("../../../methods/misc/normalize-text.cjs");
const factories = require("../keyboards/factories.cjs");
function text(text2, params = {}) {
  const ret = params;
  ret.type = "text";
  ret.text = text2;
  return ret;
}
function rich(rich2, params = {}) {
  const ret = params;
  ret.type = "rich";
  ret.content = rich2;
  return ret;
}
function media(params = {}) {
  const ret = params;
  ret.type = "media";
  return ret;
}
function geo(params) {
  const ret = params;
  ret.type = "geo";
  return ret;
}
function geoLive(params) {
  const ret = params;
  ret.type = "geo_live";
  return ret;
}
function venue(params) {
  const ret = params;
  ret.type = "venue";
  return ret;
}
function game(params) {
  const ret = params;
  ret.type = "game";
  return ret;
}
function contact(params) {
  const ret = params;
  ret.type = "contact";
  return ret;
}
function webpage(params) {
  const ret = params;
  ret.type = "webpage";
  return ret;
}
async function _convertToTl(client, obj) {
  switch (obj.type) {
    case "text": {
      const [message, entities] = await normalizeText._normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageText",
        message,
        entities,
        replyMarkup: factories._convertToTl(obj.replyMarkup),
        invertMedia: obj.invertMedia
      };
    }
    case "rich":
      return {
        _: "inputBotInlineMessageRichMessage",
        replyMarkup: factories._convertToTl(obj.replyMarkup),
        richMessage: "_" in obj.content ? obj.content : await normalizeRichMessage._normalizeInputRichMessage(client, { _: "inputPeerEmpty" }, obj.content, {})
      };
    case "media": {
      const [message, entities] = await normalizeText._normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageMediaAuto",
        message,
        entities,
        replyMarkup: factories._convertToTl(obj.replyMarkup),
        invertMedia: obj.invertMedia
      };
    }
    case "geo":
    case "geo_live":
      return {
        _: "inputBotInlineMessageMediaGeo",
        geoPoint: {
          _: "inputGeoPoint",
          lat: obj.latitude,
          long: obj.longitude
        },
        // fields will be `undefined` if this is a `geo`
        heading: obj.heading,
        period: obj.period,
        proximityNotificationRadius: obj.proximityNotificationRadius,
        replyMarkup: factories._convertToTl(obj.replyMarkup)
      };
    case "venue":
      return {
        _: "inputBotInlineMessageMediaVenue",
        geoPoint: {
          _: "inputGeoPoint",
          lat: obj.latitude,
          long: obj.longitude
        },
        title: obj.title,
        address: obj.address,
        provider: obj.source?.provider ?? "",
        venueId: obj.source?.id ?? "",
        venueType: obj.source?.type ?? "",
        replyMarkup: factories._convertToTl(obj.replyMarkup)
      };
    case "game":
      return {
        _: "inputBotInlineMessageGame",
        replyMarkup: factories._convertToTl(obj.replyMarkup)
      };
    case "contact":
      return {
        _: "inputBotInlineMessageMediaContact",
        phoneNumber: obj.phone,
        firstName: obj.firstName,
        lastName: obj.lastName ?? "",
        vcard: obj.vcard ?? "",
        replyMarkup: factories._convertToTl(obj.replyMarkup)
      };
    case "webpage": {
      const [message, entities] = await normalizeText._normalizeInputText(client, obj.text);
      return {
        _: "inputBotInlineMessageMediaWebPage",
        message,
        entities,
        replyMarkup: factories._convertToTl(obj.replyMarkup),
        invertMedia: obj.invertMedia,
        forceLargeMedia: obj.size === "large",
        forceSmallMedia: obj.size === "small",
        optional: !obj.required,
        url: obj.url
      };
    }
    default:
      utils.assertNever();
  }
}
exports._convertToTl = _convertToTl;
exports.contact = contact;
exports.game = game;
exports.geo = geo;
exports.geoLive = geoLive;
exports.media = media;
exports.rich = rich;
exports.text = text;
exports.venue = venue;
exports.webpage = webpage;
