// @fuman/utils/structures/custom-map.js
import { maybeWrapIterator } from "./_iter.js";
class CustomMap {
  #map;
  #mapperTo;
  #mapperFrom;
  clear;
  constructor(externalToInternal, internalToExternal) {
    this.#mapperTo = externalToInternal;
    this.#mapperFrom = internalToExternal;
    const map = this.#map = /* @__PURE__ */ new Map();
    this.clear = map.clear.bind(this.#map);
  }
  getInternalMap() {
    return this.#map;
  }
  delete(key) {
    return this.#map.delete(this.#mapperTo(key));
  }
  forEach(callbackfn, thisArg) {
    return this.#map.forEach((value, key) => {
      callbackfn.call(thisArg, value, this.#mapperFrom(key), this);
    });
  }
  get(key) {
    return this.#map.get(this.#mapperTo(key));
  }
  has(key) {
    return this.#map.has(this.#mapperTo(key));
  }
  set(key, value) {
    this.#map.set(this.#mapperTo(key), value);
    return this;
  }
  get size() {
    return this.#map.size;
  }
  entries() {
    const inner = this.#map.entries();
    const iterator = {
      [Symbol.iterator]: () => iterator,
      next: () => {
        const { done, value } = inner.next();
        if (done) return { done, value };
        return {
          done,
          value: [this.#mapperFrom(value[0]), value[1]]
        };
      }
    };
    return maybeWrapIterator(iterator);
  }
  keys() {
    const inner = this.#map.keys();
    const iterator = {
      [Symbol.iterator]: () => iterator,
      next: () => {
        const { done, value } = inner.next();
        if (done) return { done, value };
        return {
          done,
          value: this.#mapperFrom(value)
        };
      }
    };
    return maybeWrapIterator(iterator);
  }
  values() {
    const inner = this.#map.values();
    const iterator = {
      [Symbol.iterator]: () => iterator,
      next: () => {
        const { done, value } = inner.next();
        if (done) return { done, value };
        return {
          done,
          value
        };
      }
    };
    return maybeWrapIterator(iterator);
  }
  [Symbol.iterator]() {
    return this.entries();
  }
  get [Symbol.toStringTag]() {
    return this.#map[Symbol.toStringTag];
  }
}
export {
  CustomMap
};
