// @fuman/utils/structures/custom-set.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const _iter = require("./_iter.cjs");
class CustomSet {
  #set;
  #mapperTo;
  #mapperFrom;
  clear;
  constructor(externalToInternal, internalToExternal) {
    this.#mapperTo = externalToInternal;
    this.#mapperFrom = internalToExternal;
    const set = this.#set = /* @__PURE__ */ new Set();
    this.clear = set.clear.bind(set);
  }
  get size() {
    return this.#set.size;
  }
  add(value) {
    this.#set.add(this.#mapperTo(value));
    return this;
  }
  delete(value) {
    return this.#set.delete(this.#mapperTo(value));
  }
  forEach(callbackfn, thisArg) {
    this.#set.forEach((value) => {
      const mapped = this.#mapperFrom(value);
      callbackfn.call(thisArg, mapped, mapped, this);
    });
  }
  has(value) {
    return this.#set.has(this.#mapperTo(value));
  }
  entries() {
    const inner = this.#set.entries();
    const iterator = {
      [Symbol.iterator]: () => iterator,
      next: () => {
        const { done, value } = inner.next();
        if (done) return { done, value };
        const mapped = this.#mapperFrom(value[0]);
        return {
          done,
          value: [mapped, mapped]
        };
      }
    };
    return _iter.maybeWrapIterator(iterator);
  }
  keys() {
    const inner = this.#set.keys();
    const iterator = {
      [Symbol.iterator]: () => iterator,
      next: () => {
        const { done, value } = inner.next();
        if (done) return { done, value: void 0 };
        return {
          done,
          value: this.#mapperFrom(value)
        };
      }
    };
    return _iter.maybeWrapIterator(iterator);
  }
  values() {
    return this.keys();
  }
  union() {
    throw new Error("Method not supported.");
  }
  intersection() {
    throw new Error("Method not supported.");
  }
  difference() {
    throw new Error("Method not supported.");
  }
  symmetricDifference() {
    throw new Error("Method not supported.");
  }
  isSubsetOf() {
    throw new Error("Method not supported.");
  }
  isSupersetOf() {
    throw new Error("Method not supported.");
  }
  isDisjointFrom() {
    throw new Error("Method not supported.");
  }
  [Symbol.iterator]() {
    return this.keys();
  }
  get [Symbol.toStringTag]() {
    return this.#set[Symbol.toStringTag];
  }
  getInternalSet() {
    return this.#set;
  }
}
exports.CustomSet = CustomSet;
