// @fuman/utils/structures/deque.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const LOG2 = /* @__PURE__ */ Math.log(2);
function _nextPowerOf2(num) {
  const log2 = Math.log(num) / LOG2;
  const nextPow2 = 1 << log2 + 1;
  return Math.max(nextPow2, 4);
}
class Deque {
  _list;
  _head;
  _tail;
  _capacityMask;
  _capacity;
  constructor(array, options = {}) {
    this._head = 0;
    this._tail = 0;
    this._capacity = options.capacity;
    if (array) {
      this.#fromArray(array);
    } else {
      this._capacityMask = 3;
      this._list = Array.from({ length: 4 });
    }
  }
  /**
   * Return the number of items on the list, or 0 if empty.
   */
  get length() {
    if (this._head === this._tail) return 0;
    if (this._head < this._tail) return this._tail - this._head;
    else return this._capacityMask + 1 - (this._head - this._tail);
  }
  /**
   * Returns the item at the specified index from the list.
   * 0 is the first element, 1 is the second, and so on...
   * Elements at negative values are that many from the end: -1 is one before the end
   * (the last element), -2 is two before the end (one before last), etc.
   * Returns undefined if the index is out of bounds.
   *
   * @param index
   */
  at(index) {
    let i = index;
    if (i !== (i | 0)) {
      return void 0;
    }
    const len = this.length;
    if (i >= len || i < -len) return void 0;
    if (i < 0) i += len;
    i = this._head + i & this._capacityMask;
    return this._list[i];
  }
  /**
   * Returns the first item in the list without removing it.
   */
  peekFront() {
    if (this._head === this._tail) return void 0;
    return this._list[this._head];
  }
  /**
   * Returns the last item in the list without removing it.
   */
  peekBack() {
    if (this._head === this._tail) return void 0;
    return this._list[this._tail - 1];
  }
  /**
   * Add an item at the beginning of the list.
   * @param item
   */
  pushFront(item) {
    const len = this._list.length;
    this._head = this._head - 1 + len & this._capacityMask;
    this._list[this._head] = item;
    if (this._tail === this._head) this.#growArray();
    if (this._capacity !== void 0 && this.length > this._capacity) this.popBack();
    if (this._head < this._tail) return this._tail - this._head;
    else return this._capacityMask + 1 - (this._head - this._tail);
  }
  /**
   * Add an item to the end of the list.
   * @param item
   */
  pushBack(item) {
    const tail = this._tail;
    this._list[tail] = item;
    this._tail = tail + 1 & this._capacityMask;
    if (this._tail === this._head) {
      this.#growArray();
    }
    if (this._capacity !== void 0 && this.length > this._capacity) {
      this.popFront();
    }
    if (this._head < this._tail) return this._tail - this._head;
    else return this._capacityMask + 1 - (this._head - this._tail);
  }
  /**
   * Remove and return the last item on the list.
   * Returns undefined if the list is empty.
   */
  popBack() {
    const tail = this._tail;
    if (tail === this._head) return void 0;
    const len = this._list.length;
    this._tail = tail - 1 + len & this._capacityMask;
    const item = this._list[this._tail];
    this._list[this._tail] = void 0;
    if (this._head < 2 && tail > 1e4 && tail <= len >>> 2) this.#shrinkArray();
    return item;
  }
  /**
   * Remove and return the first item on the list,
   * Returns undefined if the list is empty.
   */
  popFront() {
    const head = this._head;
    if (head === this._tail) return void 0;
    const item = this._list[head];
    this._list[head] = void 0;
    this._head = head + 1 & this._capacityMask;
    if (head < 2 && this._tail > 1e4 && this._tail <= this._list.length >>> 2) this.#shrinkArray();
    return item;
  }
  #remove(i, size) {
    const len = this._list.length;
    let k;
    if (i < size / 2) {
      for (k = i; k > 0; k--) {
        this._list[i] = this._list[i = i - 1 + len & this._capacityMask];
      }
      this._list[i] = void 0;
      this._head = this._head + 1 + len & this._capacityMask;
    } else {
      for (k = size - 1 - i; k >= 0; k--) {
        this._list[i] = this._list[i = i + 1 + len & this._capacityMask];
      }
      this._list[i] = void 0;
      this._tail = this._tail - 1 + len & this._capacityMask;
    }
  }
  /**
   * Remove and return the item at the specified index from the list.
   * Returns undefined if the list is empty.
   * @param index
   */
  removeOne(index) {
    let i = index;
    if (this._head === this._tail) return void 0;
    const size = this.length;
    if (i >= size || i < -size) return void 0;
    if (i < 0) i += size;
    i = this._head + i & this._capacityMask;
    const item = this._list[i];
    this.#remove(i, size);
    return item;
  }
  removeBy(predicate) {
    const mask = this._capacityMask;
    let i = this._head;
    let val;
    while ((val = this._list[i]) !== void 0) {
      if (predicate(val)) {
        this.#remove(i, this.length);
        return;
      }
      i = i + 1 & mask;
    }
  }
  /**
   * Soft clear - does not reset capacity.
   */
  clear() {
    this._list = Array.from({ length: this._list.length });
    this._head = 0;
    this._tail = 0;
  }
  indexOf(item) {
    let i = this._head;
    while (i !== this._tail) {
      if (this._list[i] === item) {
        return i - this._head & this._capacityMask;
      }
      i = i + 1 & this._capacityMask;
    }
    return -1;
  }
  findIndex(predicate) {
    let i = this._head;
    while (i !== this._tail) {
      if (predicate(this._list[i])) {
        return i - this._head & this._capacityMask;
      }
      i = i + 1 & this._capacityMask;
    }
    return -1;
  }
  find(predicate) {
    let i = this._head;
    while (i !== this._tail) {
      if (predicate(this._list[i])) {
        return this._list[i];
      }
      i = i + 1 & this._capacityMask;
    }
    return void 0;
  }
  includes(item) {
    return this.indexOf(item) !== -1;
  }
  /**
   * Returns true or false whether the list is empty.
   */
  isEmpty() {
    return this._head === this._tail;
  }
  /**
   * Returns an array of all queue items.
   */
  toArray() {
    return this.#copyArray(false, this.length);
  }
  /**
   * Fills the queue with items from an array
   * For use in the constructor
   * @param array
   * @private
   */
  #fromArray(array) {
    const length = array.length;
    const capacity = _nextPowerOf2(length);
    const list = this._list = new Array(capacity);
    this._capacityMask = capacity - 1;
    this._tail = length;
    for (let i = 0; i < length; i++) list[i] = array[i];
  }
  #growArray() {
    if (this._head !== 0) {
      const newList = this.#copyArray(true, this._list.length << 1);
      this._tail = this._list.length;
      this._head = 0;
      this._list = newList;
    } else {
      this._tail = this._list.length;
      this._list.length <<= 1;
    }
    this._capacityMask = this._capacityMask << 1 | 1;
  }
  #shrinkArray() {
    this._list.length >>>= 1;
    this._capacityMask >>>= 1;
  }
  #copyArray(fullCopy, size) {
    const src = this._list;
    const capacity = src.length;
    const length = this.length;
    size = size | length;
    if (size === length && this._head < this._tail) {
      return this._list.slice(this._head, this._tail);
    }
    const dest = new Array(size);
    let k = 0;
    let i;
    if (fullCopy || this._head > this._tail) {
      for (i = this._head; i < capacity; i++) dest[k++] = src[i];
      for (i = 0; i < this._tail; i++) dest[k++] = src[i];
    } else {
      for (i = this._head; i < this._tail; i++) dest[k++] = src[i];
    }
    return dest;
  }
  [Symbol.iterator]() {
    let i = this._head;
    return {
      next: () => {
        if (i === this._tail) return { done: true, value: void 0 };
        const item = this._list[i];
        i = i + 1 & this._capacityMask;
        return { done: false, value: item };
      }
    };
  }
}
exports.Deque = Deque;
