// @fuman/utils/structures/lru-map.js
class LruMap {
  #capacity;
  #first;
  #last;
  #map;
  #size = 0;
  constructor(capacity, MapImpl = Map) {
    this.#capacity = capacity;
    this.#map = new MapImpl();
  }
  #markUsed(item) {
    if (item === this.#first) {
      return;
    }
    if (item.p) {
      if (item === this.#last) {
        this.#last = item.p;
      }
      item.p.n = item.n;
    }
    if (item.n) {
      item.n.p = item.p;
    }
    item.p = void 0;
    item.n = this.#first;
    if (this.#first) {
      this.#first.p = item;
    }
    this.#first = item;
  }
  get(key) {
    const item = this.#map.get(key);
    if (!item) return void 0;
    this.#markUsed(item);
    return item.v;
  }
  has(key) {
    return this.#map.has(key);
  }
  #remove(item) {
    if (item.p) {
      this.#last = item.p;
      this.#last.n = void 0;
    } else {
      this.#last = void 0;
      this.#first = void 0;
    }
    item.p = item.n = void 0;
    this.#map.delete(item.k);
    this.#size -= 1;
  }
  set(key, value) {
    let item = this.#map.get(key);
    if (item) {
      item.v = value;
      this.#markUsed(item);
      return;
    }
    item = {
      k: key,
      v: value,
      // for jit to optimize stuff
      n: void 0,
      p: void 0
    };
    this.#map.set(key, item);
    if (this.#first) {
      this.#first.p = item;
      item.n = this.#first;
    } else {
      this.#last = item;
    }
    this.#first = item;
    this.#size += 1;
    if (this.#size > this.#capacity) {
      const oldest = this.#last;
      if (oldest) {
        this.#remove(oldest);
      }
    }
  }
  delete(key) {
    const item = this.#map.get(key);
    if (item) this.#remove(item);
  }
  clear() {
    this.#map.clear();
    this.#first = void 0;
    this.#last = void 0;
    this.#size = 0;
  }
}
export {
  LruMap
};
