// @mtcute/node/utils/crypto.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_crypto = require("node:crypto");
const promises = require("node:fs/promises");
const node_module = require("node:module");
const node_zlib = require("node:zlib");
const utils_js = require("@mtcute/core/utils.js");
const wasm = require("@mtcute/wasm");
var _documentCurrentScript = typeof document !== "undefined" ? document.currentScript : null;
class NodeCryptoProvider extends utils_js.BaseCryptoProvider {
  createAesCtr(key, iv) {
    const cipher = node_crypto.createCipheriv(`aes-${key.length * 8}-ctr`, key, iv);
    const update = (data) => cipher.update(data);
    return {
      process: update
    };
  }
  pbkdf2(password, salt, iterations, keylen = 64, algo = "sha512") {
    return new Promise(
      (resolve, reject) => node_crypto.pbkdf2(password, salt, iterations, keylen, algo, (err, buf) => err !== null ? reject(err) : resolve(buf))
    );
  }
  sha1(data) {
    return node_crypto.createHash("sha1").update(data).digest();
  }
  sha256(data) {
    return node_crypto.createHash("sha256").update(data).digest();
  }
  hmacSha256(data, key) {
    return node_crypto.createHmac("sha256", key).update(data).digest();
  }
  gzip(data, maxSize) {
    try {
      return node_zlib.deflateSync(data, {
        maxOutputLength: maxSize
      });
    } catch (e) {
      if (e.code === "ERR_BUFFER_TOO_LARGE") {
        return null;
      }
      throw e;
    }
  }
  gunzip(data) {
    return node_zlib.gunzipSync(data);
  }
  randomFill(buf) {
    node_crypto.randomFillSync(buf);
  }
  async initialize() {
    const require$1 = node_module.createRequire(typeof document === "undefined" ? require("url").pathToFileURL(__filename).href : _documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === "SCRIPT" && _documentCurrentScript.src || new URL("utils/crypto.cjs", document.baseURI).href);
    const file = wasm.SIMD_AVAILABLE ? "mtcute-simd.wasm" : "mtcute.wasm";
    const wasmFile = require$1.resolve(`@mtcute/wasm/${file}`);
    const wasm$1 = await promises.readFile(wasmFile);
    wasm.initSync(wasm$1);
  }
  createAesIge(key, iv) {
    return {
      encrypt(data) {
        return wasm.ige256Encrypt(data, key, iv);
      },
      decrypt(data) {
        return wasm.ige256Decrypt(data, key, iv);
      }
    };
  }
}
exports.NodeCryptoProvider = NodeCryptoProvider;
