// @mtcute/node/utils/crypto.js
import { createCipheriv, pbkdf2, createHash, createHmac, randomFillSync } from "node:crypto";
import { readFile } from "node:fs/promises";
import { createRequire } from "node:module";
import { deflateSync, gunzipSync } from "node:zlib";
import { BaseCryptoProvider } from "@mtcute/core/utils.js";
import { initSync, ige256Decrypt, ige256Encrypt, SIMD_AVAILABLE } from "@mtcute/wasm";
class NodeCryptoProvider extends BaseCryptoProvider {
  createAesCtr(key, iv) {
    const cipher = createCipheriv(`aes-${key.length * 8}-ctr`, key, iv);
    const update = (data) => cipher.update(data);
    return {
      process: update
    };
  }
  pbkdf2(password, salt, iterations, keylen = 64, algo = "sha512") {
    return new Promise(
      (resolve, reject) => pbkdf2(password, salt, iterations, keylen, algo, (err, buf) => err !== null ? reject(err) : resolve(buf))
    );
  }
  sha1(data) {
    return createHash("sha1").update(data).digest();
  }
  sha256(data) {
    return createHash("sha256").update(data).digest();
  }
  hmacSha256(data, key) {
    return createHmac("sha256", key).update(data).digest();
  }
  gzip(data, maxSize) {
    try {
      return deflateSync(data, {
        maxOutputLength: maxSize
      });
    } catch (e) {
      if (e.code === "ERR_BUFFER_TOO_LARGE") {
        return null;
      }
      throw e;
    }
  }
  gunzip(data) {
    return gunzipSync(data);
  }
  randomFill(buf) {
    randomFillSync(buf);
  }
  async initialize() {
    const require2 = createRequire(import.meta.url);
    const file = SIMD_AVAILABLE ? "mtcute-simd.wasm" : "mtcute.wasm";
    const wasmFile = require2.resolve(`@mtcute/wasm/${file}`);
    const wasm = await readFile(wasmFile);
    initSync(wasm);
  }
  createAesIge(key, iv) {
    return {
      encrypt(data) {
        return ige256Encrypt(data, key, iv);
      },
      decrypt(data) {
        return ige256Decrypt(data, key, iv);
      }
    };
  }
}
export {
  NodeCryptoProvider
};
