// @mtcute/node/utils/platform.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const os = require("node:os");
const normalizeFile = require("./normalize-file.cjs");
const exitHook = require("./exit-hook.cjs");
const logging = require("./logging.cjs");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const os__namespace = /* @__PURE__ */ _interopNamespaceDefault(os);
class NodePlatform {
  getDeviceModel() {
    return `Node.js/${process.version} (${os__namespace.type()} ${os__namespace.arch()})`;
  }
  getDefaultLogLevel() {
    const envLogLevel = Number.parseInt(process.env.MTCUTE_LOG_LEVEL ?? "");
    if (!Number.isNaN(envLogLevel)) {
      return envLogLevel;
    }
    return null;
  }
}
NodePlatform.prototype.log = logging.defaultLoggingHandler;
NodePlatform.prototype.beforeExit = exitHook.beforeExit;
NodePlatform.prototype.normalizeFile = normalizeFile.normalizeFile;
exports.NodePlatform = NodePlatform;
