// @mtcute/node/utils/proxies.js
import { performHttpProxyHandshake, performSocksHandshake } from "@fuman/net";
import { HttpProxyConnectionError, SocksProxyConnectionError } from "@fuman/net";
import { connectTls, connectTcp } from "@fuman/node";
import { createProxyTransportFactory, IntermediatePacketCodec, BaseMtProxyTransport } from "@mtcute/core";
class HttpProxyTcpTransport {
  constructor(proxy) {
    this.proxy = proxy;
  }
  async connect(dc) {
    let conn;
    if (this.proxy.tls) {
      conn = await connectTls({
        address: this.proxy.host,
        port: this.proxy.port,
        extraOptions: this.proxy.tlsOptions
      });
    } else {
      conn = await connectTcp({
        address: this.proxy.host,
        port: this.proxy.port
      });
    }
    await performHttpProxyHandshake(conn, conn, this.proxy, {
      address: dc.ipAddress,
      port: dc.port
    });
    return conn;
  }
  packetCodec() {
    return new IntermediatePacketCodec();
  }
}
class SocksProxyTcpTransport {
  constructor(proxy) {
    this.proxy = proxy;
  }
  async connect(dc) {
    const conn = await connectTcp({
      address: this.proxy.host,
      port: this.proxy.port
    });
    await performSocksHandshake(conn, conn, this.proxy, {
      address: dc.ipAddress,
      port: dc.port
    });
    return conn;
  }
  packetCodec() {
    return new IntermediatePacketCodec();
  }
}
class MtProxyTcpTransport extends BaseMtProxyTransport {
  async _connectTcp(endpoint) {
    return connectTcp(endpoint);
  }
}
const proxyTransportFromUrl = createProxyTransportFactory({
  SocksProxyTcpTransport,
  HttpProxyTcpTransport,
  MtProxyTcpTransport
});
export {
  HttpProxyConnectionError,
  HttpProxyTcpTransport,
  MtProxyTcpTransport,
  SocksProxyConnectionError,
  SocksProxyTcpTransport,
  proxyTransportFromUrl
};
