// @mtcute/node/utils/proxies.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/node] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/node] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const net = require("@fuman/net");
const node = require("@fuman/node");
const core = require("@mtcute/core");
class HttpProxyTcpTransport {
  constructor(proxy) {
    this.proxy = proxy;
  }
  async connect(dc) {
    let conn;
    if (this.proxy.tls) {
      conn = await node.connectTls({
        address: this.proxy.host,
        port: this.proxy.port,
        extraOptions: this.proxy.tlsOptions
      });
    } else {
      conn = await node.connectTcp({
        address: this.proxy.host,
        port: this.proxy.port
      });
    }
    await net.performHttpProxyHandshake(conn, conn, this.proxy, {
      address: dc.ipAddress,
      port: dc.port
    });
    return conn;
  }
  packetCodec() {
    return new core.IntermediatePacketCodec();
  }
}
class SocksProxyTcpTransport {
  constructor(proxy) {
    this.proxy = proxy;
  }
  async connect(dc) {
    const conn = await node.connectTcp({
      address: this.proxy.host,
      port: this.proxy.port
    });
    await net.performSocksHandshake(conn, conn, this.proxy, {
      address: dc.ipAddress,
      port: dc.port
    });
    return conn;
  }
  packetCodec() {
    return new core.IntermediatePacketCodec();
  }
}
class MtProxyTcpTransport extends core.BaseMtProxyTransport {
  async _connectTcp(endpoint) {
    return node.connectTcp(endpoint);
  }
}
const proxyTransportFromUrl = core.createProxyTransportFactory({
  SocksProxyTcpTransport,
  HttpProxyTcpTransport,
  MtProxyTcpTransport
});
Object.defineProperty(exports, "HttpProxyConnectionError", {
  enumerable: true,
  get: () => net.HttpProxyConnectionError
});
Object.defineProperty(exports, "SocksProxyConnectionError", {
  enumerable: true,
  get: () => net.SocksProxyConnectionError
});
exports.HttpProxyTcpTransport = HttpProxyTcpTransport;
exports.MtProxyTcpTransport = MtProxyTcpTransport;
exports.SocksProxyTcpTransport = SocksProxyTcpTransport;
exports.proxyTransportFromUrl = proxyTransportFromUrl;
