// @fuman/io/read/adapters.js
import { Bytes } from "../bytes.js";
import { isByobCapableStream } from "../streams.js";
function fumanSyncReadableToAsync(readable) {
  return {
    async read(into) {
      const buf = readable.readSync(into.length);
      into.set(buf);
      return buf.length;
    }
  };
}
function webReadableToFuman(readable) {
  if (isByobCapableStream(readable)) {
    const reader2 = readable.getReader({ mode: "byob" });
    return {
      async read(into) {
        const temp = new Uint8Array(into.length);
        const { value, done } = await reader2.read(temp);
        if (done) return 0;
        into.set(value);
        return value.byteLength;
      },
      close() {
        reader2.cancel().catch(() => {
        });
      }
    };
  }
  const reader = readable.getReader();
  const buf = Bytes.alloc();
  return {
    async read(into) {
      if (buf.available === 0) {
        const { value, done } = await reader.read();
        if (done) return 0;
        buf.writeSync(value.byteLength).set(value);
        buf.disposeWriteSync();
      }
      const read = Math.min(into.length, buf.available);
      into.set(buf.readSync(read), 0);
      buf.reclaim();
      return read;
    },
    close() {
      reader.cancel().catch(() => {
      });
    }
  };
}
function fumanReadableToWeb(readable) {
  return new ReadableStream({
    type: "bytes",
    async pull(controller) {
      if (controller.byobRequest) {
        if (!controller.byobRequest.view) {
          controller.byobRequest.respond(0);
          return;
        }
        const view = new Uint8Array(
          controller.byobRequest.view.buffer,
          controller.byobRequest.view.byteOffset,
          controller.byobRequest.view.byteLength
        );
        const nread2 = await readable.read(view);
        if (nread2 === 0) {
          controller.close();
          controller.byobRequest.respond(0);
          return;
        }
        controller.byobRequest.respond(nread2);
        return;
      }
      const buf = new Uint8Array(1024 * 32);
      let nread = 0;
      while (nread === 0) {
        nread = await readable.read(buf);
        if (nread === 0) break;
        controller.enqueue(buf.subarray(0, nread));
      }
      if (nread === 0) {
        controller.close();
      }
    }
  });
}
export {
  fumanReadableToWeb,
  fumanSyncReadableToAsync,
  webReadableToFuman
};
