// @fuman/io/read/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const index = require("./async/index.cjs");
const numbers = require("./numbers.cjs");
const strings = require("./strings.cjs");
exports.async = index;
exports.float32be = numbers.float32be;
exports.float32le = numbers.float32le;
exports.float64be = numbers.float64be;
exports.float64le = numbers.float64le;
exports.int16be = numbers.int16be;
exports.int16le = numbers.int16le;
exports.int24be = numbers.int24be;
exports.int24le = numbers.int24le;
exports.int32be = numbers.int32be;
exports.int32le = numbers.int32le;
exports.int64be = numbers.int64be;
exports.int64le = numbers.int64le;
exports.int8 = numbers.int8;
exports.intbe = numbers.intbe;
exports.intle = numbers.intle;
exports.uint16be = numbers.uint16be;
exports.uint16le = numbers.uint16le;
exports.uint24be = numbers.uint24be;
exports.uint24le = numbers.uint24le;
exports.uint32be = numbers.uint32be;
exports.uint32le = numbers.uint32le;
exports.uint64be = numbers.uint64be;
exports.uint64le = numbers.uint64le;
exports.uint8 = numbers.uint8;
exports.uintbe = numbers.uintbe;
exports.uintle = numbers.uintle;
exports.cUtf8String = strings.cUtf8String;
exports.exactly = strings.exactly;
exports.lengthPrefixed = strings.lengthPrefixed;
exports.rawString = strings.rawString;
exports.untilCondition = strings.untilCondition;
exports.untilEnd = strings.untilEnd;
exports.untilZero = strings.untilZero;
exports.utf16beString = strings.utf16beString;
exports.utf16leString = strings.utf16leString;
exports.utf8String = strings.utf8String;
