// @fuman/io/read/numbers.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const _utils = require("../_utils.cjs");
const errors = require("../errors.cjs");
const strings = require("./strings.cjs");
function uint8(readable) {
  return strings.exactly(readable, 1)[0];
}
function _maybeRead(readable, size) {
  if (ArrayBuffer.isView(readable)) {
    if (readable.byteLength < size) {
      throw new errors.PartialReadError(readable, size);
    }
    return readable;
  }
  return strings.exactly(readable, size);
}
function int8(readable) {
  const val = _maybeRead(readable, 1)[0];
  if (!(val & 128)) return val;
  return -(255 - val + 1);
}
function uint16le(readable) {
  const buffer = _maybeRead(readable, 2);
  return buffer[0] | buffer[1] << 8;
}
function uint16be(readable) {
  const buffer = _maybeRead(readable, 2);
  return buffer[0] << 8 | buffer[1];
}
function uint24le(readable) {
  const buffer = _maybeRead(readable, 3);
  return buffer[0] | buffer[1] << 8 | buffer[2] << 16;
}
function uint24be(readable) {
  const buffer = _maybeRead(readable, 3);
  return buffer[0] << 16 | buffer[1] << 8 | buffer[2];
}
function uint32le(readable) {
  const buffer = _maybeRead(readable, 4);
  return (buffer[0] | buffer[1] << 8 | buffer[2] << 16) + buffer[3] * 16777216;
}
function uint32be(readable) {
  const buffer = _maybeRead(readable, 4);
  return buffer[0] * 16777216 + (buffer[1] << 16 | buffer[2] << 8 | buffer[3]);
}
function uint64le(readable) {
  const buffer = _maybeRead(readable, 8);
  const lo = (buffer[0] | buffer[1] << 8 | buffer[2] << 16) + buffer[3] * 16777216;
  const hi = (buffer[4] | buffer[5] << 8 | buffer[6] << 16) + buffer[7] * 16777216;
  return BigInt(lo) | BigInt(hi) << 32n;
}
function uint64be(readable) {
  const buffer = _maybeRead(readable, 8);
  const lo = buffer[0] * 16777216 + (buffer[1] << 16 | buffer[2] << 8 | buffer[3]);
  const hi = buffer[4] * 16777216 + (buffer[5] << 16 | buffer[6] << 8 | buffer[7]);
  return BigInt(lo) << 32n | BigInt(hi);
}
function int16le(readable) {
  const buffer = _maybeRead(readable, 2);
  const val = buffer[0] | buffer[1] << 8;
  return val & 32768 ? val | 4294901760 : val;
}
function int16be(readable) {
  const buffer = _maybeRead(readable, 2);
  const val = buffer[0] << 8 | buffer[1];
  return val & 32768 ? val | 4294901760 : val;
}
function int24le(readable) {
  const buffer = _maybeRead(readable, 3);
  const val = buffer[0] | buffer[1] << 8 | buffer[2] << 16;
  return val & 8388608 ? val | 4278190080 : val;
}
function int24be(readable) {
  const buffer = _maybeRead(readable, 3);
  const val = buffer[0] << 16 | buffer[1] << 8 | buffer[2];
  return val & 8388608 ? val | 4278190080 : val;
}
function int32le(readable) {
  const buffer = _maybeRead(readable, 4);
  return buffer[0] | buffer[1] << 8 | buffer[2] << 16 | buffer[3] << 24;
}
function int32be(readable) {
  const buffer = _maybeRead(readable, 4);
  return buffer[0] << 24 | buffer[1] << 16 | buffer[2] << 8 | buffer[3];
}
function int64le(readable) {
  const buffer = _maybeRead(readable, 8);
  const val = buffer[4] + buffer[5] * 2 ** 8 + buffer[6] * 2 ** 16 + (buffer[7] << 24);
  return (BigInt(val) << 32n) + BigInt(buffer[0] + buffer[1] * 2 ** 8 + buffer[2] * 2 ** 16 + buffer[3] * 2 ** 24);
}
function int64be(readable) {
  const buffer = _maybeRead(readable, 8);
  const val = (buffer[0] << 24) + buffer[1] * 2 ** 16 + buffer[2] * 2 ** 8 + buffer[3];
  return (BigInt(val) << 32n) + BigInt(buffer[4] * 2 ** 24 + buffer[5] * 2 ** 16 + buffer[6] * 2 ** 8 + buffer[7]);
}
function uintle(readable, size) {
  const buffer = _maybeRead(readable, size);
  let val = BigInt(buffer[0]);
  let mul = 1;
  let i = 0;
  while (++i < size && (mul *= 256)) {
    val |= BigInt(buffer[i] * mul);
  }
  return val;
}
function uintbe(readable, size) {
  const buffer = _maybeRead(readable, size);
  let val = BigInt(buffer[--size]);
  let mul = 1;
  while (size > 0 && (mul *= 256)) {
    val |= BigInt(buffer[--size] * mul);
  }
  return val;
}
function intbe(readable, size) {
  const buffer = _maybeRead(readable, size);
  let i = size - 1;
  let mul = 1;
  let val = BigInt(buffer[i]);
  while (i > 0 && (mul *= 256)) {
    val |= BigInt(buffer[--i] * mul);
  }
  mul *= 128;
  if (val >= mul) val -= 2n ** BigInt(8 * size);
  return val;
}
function intle(readable, size) {
  const buffer = _maybeRead(readable, size);
  let val = BigInt(buffer[0]);
  let mul = 1;
  let i = 0;
  while (++i < size && (mul *= 256)) {
    val |= BigInt(buffer[i] * mul);
  }
  mul *= 128;
  if (val >= mul) val -= 2n ** BigInt(8 * size);
  return val;
}
function float32le(readable) {
  const buffer = _maybeRead(readable, 4);
  return _utils.getDv(buffer).getFloat32(buffer.byteOffset, true);
}
function float32be(readable) {
  const buffer = _maybeRead(readable, 4);
  return _utils.getDv(buffer).getFloat32(buffer.byteOffset, false);
}
function float64le(readable) {
  const buffer = _maybeRead(readable, 8);
  return _utils.getDv(buffer).getFloat64(buffer.byteOffset, true);
}
function float64be(readable) {
  const buffer = _maybeRead(readable, 8);
  return _utils.getDv(buffer).getFloat64(buffer.byteOffset, false);
}
exports.float32be = float32be;
exports.float32le = float32le;
exports.float64be = float64be;
exports.float64le = float64le;
exports.int16be = int16be;
exports.int16le = int16le;
exports.int24be = int24be;
exports.int24le = int24le;
exports.int32be = int32be;
exports.int32le = int32le;
exports.int64be = int64be;
exports.int64le = int64le;
exports.int8 = int8;
exports.intbe = intbe;
exports.intle = intle;
exports.uint16be = uint16be;
exports.uint16le = uint16le;
exports.uint24be = uint24be;
exports.uint24le = uint24le;
exports.uint32be = uint32be;
exports.uint32le = uint32le;
exports.uint64be = uint64be;
exports.uint64le = uint64le;
exports.uint8 = uint8;
exports.uintbe = uintbe;
exports.uintle = uintle;
