// @fuman/io/read/strings.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const bytes = require("../bytes.cjs");
const errors = require("../errors.cjs");
function exactly(readable, length) {
  const buf = readable.readSync(length);
  if (buf.length < length) {
    throw new errors.PartialReadError(buf, length);
  }
  return buf;
}
function rawString(readable, length) {
  const buf = readable instanceof Uint8Array ? readable : exactly(readable, length);
  let result = "";
  for (let i = 0; i < length; i++) {
    result += String.fromCharCode(buf[i]);
  }
  return result;
}
function utf8String(readable, length) {
  return utils.utf8.decoder.decode(exactly(readable, length));
}
function utf16beString(readable, byteLength) {
  if (byteLength % 2 !== 0) {
    throw new Error("utf16beString: byteLength must be even");
  }
  const buf = readable instanceof Uint8Array ? readable : exactly(readable, byteLength);
  let result = "";
  for (let i = 0; i < byteLength; i += 2) {
    result += String.fromCharCode(buf[i + 1] | buf[i] << 8);
  }
  return result;
}
function utf16leString(readable, byteLength) {
  if (byteLength % 2 !== 0) {
    throw new Error("utf16leString: byteLength must be even");
  }
  const buf = readable instanceof Uint8Array ? readable : exactly(readable, byteLength);
  let result = "";
  for (let i = 0; i < byteLength; i += 2) {
    result += String.fromCharCode(buf[i] | buf[i + 1] << 8);
  }
  return result;
}
function untilCondition(readable, condition) {
  const buf = bytes.Bytes.alloc();
  while (true) {
    const byte = readable.readSync(1);
    if (byte.length === 0) {
      throw new errors.PartialReadError(buf.result(), buf.available + 1);
    }
    buf.writeSync(1)[0] = byte[0];
    if (condition(byte[0])) {
      break;
    }
  }
  return buf.result();
}
function untilEnd(readable, chunkSize = 1024) {
  const buf = bytes.Bytes.alloc(chunkSize);
  while (true) {
    const chunk = readable.readSync(chunkSize);
    if (chunk.length === 0) {
      break;
    }
    buf.writeSync(chunk.length).set(chunk);
  }
  return buf.result();
}
function untilZero(readable) {
  return untilCondition(readable, (byte) => byte === 0);
}
function cUtf8String(readable) {
  const buf = untilZero(readable);
  return utils.utf8.decoder.decode(buf.subarray(0, buf.length - 1));
}
function lengthPrefixed(readLength, readable) {
  const length = readLength(readable);
  return exactly(readable, length);
}
exports.cUtf8String = cUtf8String;
exports.exactly = exactly;
exports.lengthPrefixed = lengthPrefixed;
exports.rawString = rawString;
exports.untilCondition = untilCondition;
exports.untilEnd = untilEnd;
exports.untilZero = untilZero;
exports.utf16beString = utf16beString;
exports.utf16leString = utf16leString;
exports.utf8String = utf8String;
