// @mtcute/core/network/mtproxy/_fake-tls.js
import { Bytes, read } from "@fuman/io";
import { u8, typed, bigint } from "@fuman/utils";
const MAX_TLS_PACKET_LENGTH = 2878;
const MAX_TLS_HELLO_SIZE = 4096;
const KEY_MOD = 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEDn;
const QUAD_RES_MOD = 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEDn;
const QUAD_RES_POW = 0x3FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF6n;
function _getY2(x, mod) {
  let y = x;
  y = (y + 486662n) % mod;
  y = y * x % mod;
  y = (y + 1n) % mod;
  y = y * x % mod;
  return y;
}
function _getDoubleX(x, mod) {
  let denominator = _getY2(x, mod);
  denominator = denominator * 4n % mod;
  let numerator = x * x % mod;
  numerator = (numerator - 1n) % mod;
  numerator = numerator * numerator % mod;
  denominator = bigint.modInv(denominator, mod);
  numerator = numerator * denominator % mod;
  return numerator;
}
function _isQuadraticResidue(a) {
  const r = bigint.modPowBinary(a, QUAD_RES_POW, QUAD_RES_MOD);
  return r === 1n;
}
function executeTlsOperations(h) {
  h.string([22, 3, 1]);
  h.beginScope();
  h.string([1, 0]);
  h.beginScope();
  h.string([3, 3]);
  h.zero(32);
  h.string([32]);
  h.random(32);
  h.string([0, 32]);
  h.grease(0);
  h.string([
    19,
    1,
    19,
    2,
    19,
    3,
    192,
    43,
    192,
    47,
    192,
    44,
    192,
    48,
    204,
    169,
    204,
    168,
    192,
    19,
    192,
    20,
    0,
    156,
    0,
    157,
    0,
    47,
    0,
    53,
    1,
    0
  ]);
  h.beginScope();
  h.grease(2);
  h.string([0, 0]);
  h.permutation([
    (w) => {
      w.string([0, 0]);
      w.beginScope();
      w.beginScope();
      w.string([0]);
      w.beginScope();
      w.domain();
      w.endScope();
      w.endScope();
      w.endScope();
    },
    (w) => w.string([0, 5, 0, 5, 1, 0, 0, 0, 0]),
    // status_request
    (w) => {
      w.string([0, 10, 0, 12, 0, 10]);
      w.grease(4);
      w.string([17, 236, 0, 29, 0, 23, 0, 24]);
    },
    (w) => w.string([0, 11, 0, 2, 1, 0]),
    // ec_point_formats
    (w) => w.string([
      // signature_algorithms
      0,
      13,
      0,
      18,
      0,
      16,
      4,
      3,
      8,
      4,
      4,
      1,
      5,
      3,
      8,
      5,
      5,
      1,
      8,
      6,
      6,
      1
    ]),
    (w) => w.string([
      // application_layer_protocol_negotiation (ALPN)
      0,
      16,
      0,
      14,
      0,
      12,
      2,
      104,
      50,
      8,
      104,
      116,
      116,
      112,
      47,
      49,
      46,
      49
    ]),
    (w) => w.string([0, 18, 0, 0]),
    // signed_certificate_timestamp
    (w) => w.string([0, 23, 0, 0]),
    // extended_master_secret
    (w) => w.string([0, 27, 0, 3, 2, 0, 2]),
    // compress_certificate
    (w) => w.string([0, 35, 0, 0]),
    // session_ticket
    (w) => {
      w.string([0, 43, 0, 7, 6]);
      w.grease(6);
      w.string([3, 4, 3, 3]);
    },
    (w) => w.string([0, 45, 0, 2, 1, 1]),
    // psk_key_exchange_modes
    (w) => {
      w.string([0, 51, 4, 239, 4, 237]);
      w.grease(4);
      w.string([0, 1, 0, 17, 236, 4, 192]);
      w.mlKem768Key();
      w.key();
      w.string([0, 29, 0, 32]);
      w.key();
    },
    (w) => w.string([68, 205, 0, 5, 0, 3, 2, 104, 50]),
    // application_settings (ALPS)
    (w) => {
      w.string([254, 13]);
      w.beginScope();
      w.string([0, 0, 1, 0, 1]);
      w.random(1);
      w.string([0, 32]);
      w.random(32);
      w.beginScope();
      w.echPayload();
      w.endScope();
      w.endScope();
    },
    (w) => w.string([255, 1, 0, 1, 0])
    // renegotiation_info
  ]);
  h.grease(3);
  h.string([0, 1, 0]);
  h.padding();
  h.endScope();
  h.endScope();
  h.endScope();
}
function initGrease(crypto, size) {
  const buf = crypto.randomBytes(size);
  for (let i = 0; i < size; i++) {
    buf[i] = (buf[i] & 240) + 10;
  }
  for (let i = 1; i < size; i += 2) {
    if (buf[i] === buf[i - 1]) {
      buf[i] ^= 16;
    }
  }
  return buf;
}
class TlsHelloWriter {
  constructor(crypto, domain, grease) {
    this.crypto = crypto;
    this._domain = domain;
    this.buf = u8.alloc(MAX_TLS_HELLO_SIZE);
    this.dv = typed.toDataView(this.buf);
    this._grease = grease ?? initGrease(this.crypto, 7);
  }
  buf;
  dv;
  pos = 0;
  _domain;
  _grease;
  _scopes = [];
  string(buf) {
    this.buf.set(buf, this.pos);
    this.pos += buf.length;
  }
  random(size) {
    this.string(this.crypto.randomBytes(size));
  }
  zero(size) {
    this.pos += size;
  }
  domain() {
    this.string(this._domain);
  }
  grease(seed) {
    this.buf[this.pos] = this.buf[this.pos + 1] = this._grease[seed];
    this.pos += 2;
  }
  key() {
    for (; ; ) {
      const key = this.crypto.randomBytes(32);
      key[31] &= 127;
      let x = bigint.fromBytes(key);
      const y = _getY2(x, KEY_MOD);
      if (_isQuadraticResidue(y)) {
        for (let i = 0; i < 3; i++) {
          x = _getDoubleX(x, KEY_MOD);
        }
        const key2 = bigint.toBytes(x, 32, true);
        this.string(key2);
        return;
      }
    }
  }
  // fake X25519MLKEM768 hybrid keyshare's ml-kem-768 part: 384 coefficients
  // (mod 3329) packed 3 bytes per 2 coefficients, followed by 32 random bytes
  mlKem768Key() {
    const rnd = this.crypto.randomBytes(384 * 8);
    const dv = typed.toDataView(rnd);
    for (let i = 0; i < 384; i++) {
      const a = dv.getUint32(i * 8, true) % 3329;
      const b = dv.getUint32(i * 8 + 4, true) % 3329;
      this.buf[this.pos++] = a & 255;
      this.buf[this.pos++] = (a >> 8) + ((b & 15) << 4);
      this.buf[this.pos++] = b >> 4;
    }
    this.random(32);
  }
  // GREASE ECH payload: random bytes of length 144/176/208/240
  echPayload() {
    this.random(this._randomInt(4) * 32 + 144);
  }
  beginScope() {
    this._scopes.push(this.pos);
    this.pos += 2;
  }
  endScope() {
    const begin = this._scopes.pop();
    if (begin === void 0) {
      throw new Error("endScope called without beginScope");
    }
    this.dv.setUint16(begin, this.pos - begin - 2);
  }
  // render each part into its own buffer, then emit them in random order —
  // mimics chrome's per-connection tls extension shuffling
  permutation(parts) {
    const rendered = parts.map((fn) => {
      const sub = new TlsHelloWriter(this.crypto, this._domain, this._grease);
      fn(sub);
      return sub.buf.subarray(0, sub.pos);
    });
    for (let i = rendered.length - 1; i > 0; i--) {
      const j = this._randomInt(i + 1);
      const tmp = rendered[i];
      rendered[i] = rendered[j];
      rendered[j] = tmp;
    }
    for (const part of rendered) {
      this.string(part);
    }
  }
  // tls padding extension, only emitted for hellos smaller than 513 bytes
  padding() {
    if (this.pos >= 513) return;
    const size = 513 - this.pos;
    this.string([0, 21]);
    this.beginScope();
    this.zero(size);
    this.endScope();
  }
  _randomInt(max) {
    return typed.toDataView(this.crypto.randomBytes(4)).getUint32(0, true) % max;
  }
  async finish(secret) {
    const data = this.buf.subarray(0, this.pos);
    const unixTime = ~~(Date.now() / 1e3);
    const hash = await this.crypto.hmacSha256(data, secret);
    const dv = typed.toDataView(hash);
    const old = dv.getInt32(28, true);
    dv.setInt32(28, old ^ unixTime, true);
    data.set(hash, 11);
    return data;
  }
}
async function generateFakeTlsHeader(domain, secret, crypto) {
  const writer = new TlsHelloWriter(crypto, domain);
  executeTlsOperations(writer);
  return writer.finish(secret);
}
const EMPTY_BYTES = Bytes.from(u8.empty);
class FakeTlsPacketCodec {
  // protected _stream: Buffer = Buffer.alloc(0)
  constructor(_inner) {
    this._inner = _inner;
  }
  _isFirstTls = true;
  setup(crypto, log) {
    this._inner.setup?.(crypto, log);
  }
  _tag;
  async tag() {
    this._tag = await this._inner.tag();
    return new Uint8Array(0);
  }
  async encode(packet, into) {
    const tmp = Bytes.alloc(packet.length);
    await this._inner.encode(packet, tmp);
    while (tmp.available > 0) {
      const header = new Uint8Array([23, 3, 3, 0, 0]);
      let packet2;
      if (this._isFirstTls) {
        this._isFirstTls = false;
        packet2 = u8.concat2(this._tag, tmp.readSync(MAX_TLS_PACKET_LENGTH - this._tag.length));
      } else {
        packet2 = tmp.readSync(MAX_TLS_PACKET_LENGTH);
      }
      typed.toDataView(header).setUint16(3, packet2.length);
      into.writeSync(header.length).set(header);
      into.writeSync(packet2.length).set(packet2);
      into.disposeWriteSync();
    }
  }
  async decode(reader, eof) {
    if (eof) return null;
    while (reader.available >= 5) {
      const header = reader.readSync(3);
      if (header[0] !== 23 || header[1] !== 3 || header[2] !== 3) {
        throw new Error("Invalid TLS header");
      }
      const length = read.uint16be(reader);
      if (reader.available < length) {
        reader.rewind(5);
        break;
      }
      const inner = await this._inner.decode(Bytes.from(reader.readSync(length)), eof);
      if (inner) return inner;
    }
    return this._inner.decode(EMPTY_BYTES, eof);
  }
  reset() {
    this._isFirstTls = true;
  }
}
export {
  FakeTlsPacketCodec,
  generateFakeTlsHeader
};
