// @mtcute/core/network/mtproxy/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const errors = require("../../types/errors.cjs");
const _fakeTls = require("./_fake-tls.cjs");
const obfuscated = require("../transports/obfuscated.cjs");
const intermediate = require("../transports/intermediate.cjs");
const MAX_DOMAIN_LENGTH = 182;
const TLS_START_SERVER = [
  new Uint8Array([22, 3, 3]),
  new Uint8Array([20, 3, 3, 0, 1, 1, 23, 3, 3])
];
const TLS_START_CLIENT = new Uint8Array([20, 3, 3, 0, 1, 1]);
class BaseMtProxyTransport {
  _proxy;
  _rawSecret;
  _randomPadding = false;
  _fakeTlsDomain = null;
  _crypto;
  _log;
  setup(crypto, log) {
    this._crypto = crypto;
    this._log = log.create("mtproxy");
  }
  /**
   * @param proxy  Information about the proxy
   */
  constructor(proxy) {
    this._proxy = proxy;
    let secret;
    if (ArrayBuffer.isView(proxy.secret)) {
      secret = proxy.secret;
    } else if (proxy.secret.match(/^[0-9a-f]+$/i)) {
      secret = utils.hex.decode(proxy.secret);
    } else {
      secret = utils.base64.decode(proxy.secret, true);
    }
    if (secret.length > 17 + MAX_DOMAIN_LENGTH) {
      throw new errors.MtSecurityError("Invalid secret: too long");
    }
    if (secret.length < 16) {
      throw new errors.MtSecurityError("Invalid secret: too short");
    }
    if (secret.length === 16) {
      this._rawSecret = secret;
    } else if (secret.length === 17 && secret[0] === 221) {
      this._rawSecret = secret.slice(1);
      this._randomPadding = true;
    } else if (secret.length >= 18 && secret[0] === 238) {
      this._rawSecret = secret.slice(1, 17);
      this._fakeTlsDomain = secret.slice(17);
    } else {
      throw new errors.MtUnsupportedError("Unsupported secret");
    }
  }
  getMtproxyInfo() {
    return {
      _: "inputClientProxy",
      address: this._proxy.host,
      port: this._proxy.port
    };
  }
  packetCodec(dc) {
    const proxy = {
      dcId: dc.id,
      media: Boolean(dc.mediaOnly),
      test: Boolean(dc.testMode),
      secret: this._rawSecret
    };
    if (!this._fakeTlsDomain) {
      let inner;
      if (this._randomPadding) {
        inner = new intermediate.PaddedIntermediatePacketCodec();
      } else {
        inner = new intermediate.IntermediatePacketCodec();
      }
      return new obfuscated.ObfuscatedPacketCodec(inner, proxy);
    }
    return new _fakeTls.FakeTlsPacketCodec(
      new obfuscated.ObfuscatedPacketCodec(new intermediate.PaddedIntermediatePacketCodec(), proxy)
    );
  }
  async connect() {
    const conn = await this._connectTcp({
      address: this._proxy.host,
      port: this._proxy.port
    });
    if (this._fakeTlsDomain) {
      await this._handleConnectFakeTls(conn);
    }
    return conn;
  }
  async _handleConnectFakeTls(conn) {
    this._log.debug("performing initial faketls handshake");
    const hello = await _fakeTls.generateFakeTlsHeader(this._fakeTlsDomain, this._rawSecret, this._crypto);
    const helloRand = hello.slice(11, 11 + 32);
    await conn.write(hello);
    const resp = io.Bytes.alloc(0);
    for (const first of TLS_START_SERVER) {
      const buf = await io.read.async.exactly(conn, first.length + 2);
      io.write.bytes(resp, buf);
      if (!utils.typed.equal(buf.slice(0, first.length), first)) {
        throw new errors.MtSecurityError("Server hello is invalid");
      }
      const skipSize = utils.typed.toDataView(buf).getUint16(first.length);
      io.write.bytes(resp, await io.read.async.exactly(conn, skipSize));
    }
    const respBuf = resp.result();
    const respRand = respBuf.slice(11, 11 + 32);
    const hash = await this._crypto.hmacSha256(
      utils.u8.concat([
        helloRand,
        respBuf.slice(0, 11),
        utils.u8.alloc(32),
        respBuf.slice(11 + 32)
      ]),
      this._rawSecret
    );
    if (!utils.typed.equal(hash, respRand)) {
      throw new errors.MtSecurityError("Response hash is invalid");
    }
    await conn.write(TLS_START_CLIENT);
    this._log.debug("faketls handshake done");
  }
}
exports.BaseMtProxyTransport = BaseMtProxyTransport;
