// @mtcute/core/network/mtproxy/index.js
import { Bytes, read, write } from "@fuman/io";
import { hex, base64, typed, u8 } from "@fuman/utils";
import { MtSecurityError, MtUnsupportedError } from "../../types/errors.js";
import { FakeTlsPacketCodec, generateFakeTlsHeader } from "./_fake-tls.js";
import { ObfuscatedPacketCodec } from "../transports/obfuscated.js";
import { PaddedIntermediatePacketCodec, IntermediatePacketCodec } from "../transports/intermediate.js";
const MAX_DOMAIN_LENGTH = 182;
const TLS_START_SERVER = [
  new Uint8Array([22, 3, 3]),
  new Uint8Array([20, 3, 3, 0, 1, 1, 23, 3, 3])
];
const TLS_START_CLIENT = new Uint8Array([20, 3, 3, 0, 1, 1]);
class BaseMtProxyTransport {
  _proxy;
  _rawSecret;
  _randomPadding = false;
  _fakeTlsDomain = null;
  _crypto;
  _log;
  setup(crypto, log) {
    this._crypto = crypto;
    this._log = log.create("mtproxy");
  }
  /**
   * @param proxy  Information about the proxy
   */
  constructor(proxy) {
    this._proxy = proxy;
    let secret;
    if (ArrayBuffer.isView(proxy.secret)) {
      secret = proxy.secret;
    } else if (proxy.secret.match(/^[0-9a-f]+$/i)) {
      secret = hex.decode(proxy.secret);
    } else {
      secret = base64.decode(proxy.secret, true);
    }
    if (secret.length > 17 + MAX_DOMAIN_LENGTH) {
      throw new MtSecurityError("Invalid secret: too long");
    }
    if (secret.length < 16) {
      throw new MtSecurityError("Invalid secret: too short");
    }
    if (secret.length === 16) {
      this._rawSecret = secret;
    } else if (secret.length === 17 && secret[0] === 221) {
      this._rawSecret = secret.slice(1);
      this._randomPadding = true;
    } else if (secret.length >= 18 && secret[0] === 238) {
      this._rawSecret = secret.slice(1, 17);
      this._fakeTlsDomain = secret.slice(17);
    } else {
      throw new MtUnsupportedError("Unsupported secret");
    }
  }
  getMtproxyInfo() {
    return {
      _: "inputClientProxy",
      address: this._proxy.host,
      port: this._proxy.port
    };
  }
  packetCodec(dc) {
    const proxy = {
      dcId: dc.id,
      media: Boolean(dc.mediaOnly),
      test: Boolean(dc.testMode),
      secret: this._rawSecret
    };
    if (!this._fakeTlsDomain) {
      let inner;
      if (this._randomPadding) {
        inner = new PaddedIntermediatePacketCodec();
      } else {
        inner = new IntermediatePacketCodec();
      }
      return new ObfuscatedPacketCodec(inner, proxy);
    }
    return new FakeTlsPacketCodec(
      new ObfuscatedPacketCodec(new PaddedIntermediatePacketCodec(), proxy)
    );
  }
  async connect() {
    const conn = await this._connectTcp({
      address: this._proxy.host,
      port: this._proxy.port
    });
    if (this._fakeTlsDomain) {
      await this._handleConnectFakeTls(conn);
    }
    return conn;
  }
  async _handleConnectFakeTls(conn) {
    this._log.debug("performing initial faketls handshake");
    const hello = await generateFakeTlsHeader(this._fakeTlsDomain, this._rawSecret, this._crypto);
    const helloRand = hello.slice(11, 11 + 32);
    await conn.write(hello);
    const resp = Bytes.alloc(0);
    for (const first of TLS_START_SERVER) {
      const buf = await read.async.exactly(conn, first.length + 2);
      write.bytes(resp, buf);
      if (!typed.equal(buf.slice(0, first.length), first)) {
        throw new MtSecurityError("Server hello is invalid");
      }
      const skipSize = typed.toDataView(buf).getUint16(first.length);
      write.bytes(resp, await read.async.exactly(conn, skipSize));
    }
    const respBuf = resp.result();
    const respRand = respBuf.slice(11, 11 + 32);
    const hash = await this._crypto.hmacSha256(
      u8.concat([
        helloRand,
        respBuf.slice(0, 11),
        u8.alloc(32),
        respBuf.slice(11 + 32)
      ]),
      this._rawSecret
    );
    if (!typed.equal(hash, respRand)) {
      throw new MtSecurityError("Response hash is invalid");
    }
    await conn.write(TLS_START_CLIENT);
    this._log.debug("faketls handshake done");
  }
}
export {
  BaseMtProxyTransport
};
