// @mtcute/core/highlevel/types/bots/webview.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
class WebviewResult {
  constructor(raw, shouldBeClosed = false) {
    this.raw = raw;
    this.shouldBeClosed = shouldBeClosed;
  }
  /** Whether the webview should be opened as a landscape fullscreen app */
  get isFullscreen() {
    return this.raw.fullscreen;
  }
  /** Whether the webview should be opened in "compact" mode */
  get isCompact() {
    return !this.raw.fullsize;
  }
  /**
   * Webview session ID (only returned by inline button mini apps, menu button mini apps, attachment menu mini apps).
   */
  get queryId() {
    return this.raw.queryId ?? null;
  }
  /** Actual URL to open */
  get url() {
    return this.raw.url;
  }
}
const WebviewResult$1 = /* @__PURE__ */ inspectable.makeInspectable(WebviewResult, ["shouldBeClosed"]);
exports.WebviewResult = WebviewResult$1;
