// @mtcute/core/highlevel/types/media/input-media/factories.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
function animation(file, params = {}) {
  const ret = params;
  ret.type = "video";
  ret.file = file;
  ret.isAnimated = true;
  return ret;
}
function audio(file, params = {}) {
  const ret = params;
  ret.type = "audio";
  ret.file = file;
  return ret;
}
function document(file, params = {}) {
  const ret = params;
  ret.type = "document";
  ret.file = file;
  return ret;
}
function photo(file, params = {}) {
  const ret = params;
  ret.type = "photo";
  ret.file = file;
  return ret;
}
function video(file, params = {}) {
  const ret = params;
  ret.type = "video";
  ret.file = file;
  return ret;
}
function voice(file, params = {}) {
  const ret = params;
  ret.type = "voice";
  ret.file = file;
  return ret;
}
function sticker(file, params = {}) {
  const ret = params;
  ret.type = "sticker";
  ret.file = file;
  return ret;
}
function venue(params) {
  const ret = params;
  ret.type = "venue";
  return ret;
}
function geo(latitude, longitude, params = {}) {
  const ret = params;
  ret.type = "geo";
  ret.latitude = latitude;
  ret.longitude = longitude;
  return ret;
}
function geoLive(latitude, longitude, params = {}) {
  const ret = params;
  ret.type = "geo_live";
  ret.latitude = latitude;
  ret.longitude = longitude;
  return ret;
}
function dice(emoji, params) {
  const ret = params;
  ret.type = "dice";
  ret.emoji = emoji;
  return ret;
}
function contact(params) {
  const ret = params;
  ret.type = "contact";
  return ret;
}
function game(game2) {
  return {
    type: "game",
    game: game2
  };
}
function invoice(params) {
  const ret = params;
  ret.type = "invoice";
  return ret;
}
function poll(params) {
  const ret = params;
  ret.type = "poll";
  return ret;
}
function quiz(params) {
  const ret = params;
  ret.type = "quiz";
  return ret;
}
function story(params) {
  const ret = params;
  ret.type = "story";
  return ret;
}
function webpage(url, params = {}) {
  const ret = params;
  ret.type = "webpage";
  ret.url = url;
  return ret;
}
function paid(params) {
  const ret = params;
  ret.type = "paid";
  return ret;
}
function todo(params) {
  const ret = params;
  ret.type = "todo";
  return ret;
}
function auto(file, params = {}) {
  const ret = params;
  ret.type = "auto";
  ret.file = file;
  return ret;
}
exports.animation = animation;
exports.audio = audio;
exports.auto = auto;
exports.contact = contact;
exports.dice = dice;
exports.document = document;
exports.game = game;
exports.geo = geo;
exports.geoLive = geoLive;
exports.invoice = invoice;
exports.paid = paid;
exports.photo = photo;
exports.poll = poll;
exports.quiz = quiz;
exports.sticker = sticker;
exports.story = story;
exports.todo = todo;
exports.venue = venue;
exports.video = video;
exports.voice = voice;
exports.webpage = webpage;
