// @mtcute/core/highlevel/types/media/input-media/factories.js
function animation(file, params = {}) {
  const ret = params;
  ret.type = "video";
  ret.file = file;
  ret.isAnimated = true;
  return ret;
}
function audio(file, params = {}) {
  const ret = params;
  ret.type = "audio";
  ret.file = file;
  return ret;
}
function document(file, params = {}) {
  const ret = params;
  ret.type = "document";
  ret.file = file;
  return ret;
}
function photo(file, params = {}) {
  const ret = params;
  ret.type = "photo";
  ret.file = file;
  return ret;
}
function video(file, params = {}) {
  const ret = params;
  ret.type = "video";
  ret.file = file;
  return ret;
}
function voice(file, params = {}) {
  const ret = params;
  ret.type = "voice";
  ret.file = file;
  return ret;
}
function sticker(file, params = {}) {
  const ret = params;
  ret.type = "sticker";
  ret.file = file;
  return ret;
}
function venue(params) {
  const ret = params;
  ret.type = "venue";
  return ret;
}
function geo(latitude, longitude, params = {}) {
  const ret = params;
  ret.type = "geo";
  ret.latitude = latitude;
  ret.longitude = longitude;
  return ret;
}
function geoLive(latitude, longitude, params = {}) {
  const ret = params;
  ret.type = "geo_live";
  ret.latitude = latitude;
  ret.longitude = longitude;
  return ret;
}
function dice(emoji, params) {
  const ret = params;
  ret.type = "dice";
  ret.emoji = emoji;
  return ret;
}
function contact(params) {
  const ret = params;
  ret.type = "contact";
  return ret;
}
function game(game2) {
  return {
    type: "game",
    game: game2
  };
}
function invoice(params) {
  const ret = params;
  ret.type = "invoice";
  return ret;
}
function poll(params) {
  const ret = params;
  ret.type = "poll";
  return ret;
}
function quiz(params) {
  const ret = params;
  ret.type = "quiz";
  return ret;
}
function story(params) {
  const ret = params;
  ret.type = "story";
  return ret;
}
function webpage(url, params = {}) {
  const ret = params;
  ret.type = "webpage";
  ret.url = url;
  return ret;
}
function paid(params) {
  const ret = params;
  ret.type = "paid";
  return ret;
}
function todo(params) {
  const ret = params;
  ret.type = "todo";
  return ret;
}
function auto(file, params = {}) {
  const ret = params;
  ret.type = "auto";
  ret.file = file;
  return ret;
}
export {
  animation,
  audio,
  auto,
  contact,
  dice,
  document,
  game,
  geo,
  geoLive,
  invoice,
  paid,
  photo,
  poll,
  quiz,
  sticker,
  story,
  todo,
  venue,
  video,
  voice,
  webpage
};
