// @fuman/io/write/strings.js
import { utf8 } from "@fuman/utils";
function bytes(writable, bytes2) {
  writable.writeSync(bytes2.length).set(bytes2);
  writable.disposeWriteSync();
}
function bytesReversed(writable, bytes2) {
  const buf = writable.writeSync(bytes2.length);
  buf.set(bytes2);
  buf.reverse();
  writable.disposeWriteSync();
}
function rawString(writable, str) {
  const buf = writable.writeSync(str.length);
  for (let i = 0; i < str.length; i++) {
    buf[i] = str.charCodeAt(i);
  }
  writable.disposeWriteSync();
}
function utf8String(writable, str) {
  const len = utf8.encodedLength(str);
  const buf = writable.writeSync(len);
  utf8.encoder.encodeInto(str, buf);
  writable.disposeWriteSync();
}
function utf16beString(writable, str) {
  const buf = writable.writeSync(str.length * 2);
  for (let i = 0, j = 0; i < str.length; i++, j += 2) {
    const char = str.charCodeAt(i);
    buf[j + 1] = char & 255;
    buf[j] = char >> 8 & 255;
  }
  writable.disposeWriteSync();
}
function utf16leString(writable, str) {
  const buf = writable.writeSync(str.length * 2);
  for (let i = 0, j = 0; i < str.length; i++, j += 2) {
    const char = str.charCodeAt(i);
    buf[j] = char & 255;
    buf[j + 1] = char >> 8 & 255;
  }
  writable.disposeWriteSync();
}
function cUtf8String(writable, str) {
  const len = utf8.encodedLength(str);
  const buf = writable.writeSync(len + 1);
  utf8.encoder.encodeInto(str, buf);
  buf[len] = 0;
  writable.disposeWriteSync();
}
export {
  bytes,
  bytesReversed,
  cUtf8String,
  rawString,
  utf16beString,
  utf16leString,
  utf8String
};
