// @fuman/io/write/strings.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
function bytes(writable, bytes2) {
  writable.writeSync(bytes2.length).set(bytes2);
  writable.disposeWriteSync();
}
function bytesReversed(writable, bytes2) {
  const buf = writable.writeSync(bytes2.length);
  buf.set(bytes2);
  buf.reverse();
  writable.disposeWriteSync();
}
function rawString(writable, str) {
  const buf = writable.writeSync(str.length);
  for (let i = 0; i < str.length; i++) {
    buf[i] = str.charCodeAt(i);
  }
  writable.disposeWriteSync();
}
function utf8String(writable, str) {
  const len = utils.utf8.encodedLength(str);
  const buf = writable.writeSync(len);
  utils.utf8.encoder.encodeInto(str, buf);
  writable.disposeWriteSync();
}
function utf16beString(writable, str) {
  const buf = writable.writeSync(str.length * 2);
  for (let i = 0, j = 0; i < str.length; i++, j += 2) {
    const char = str.charCodeAt(i);
    buf[j + 1] = char & 255;
    buf[j] = char >> 8 & 255;
  }
  writable.disposeWriteSync();
}
function utf16leString(writable, str) {
  const buf = writable.writeSync(str.length * 2);
  for (let i = 0, j = 0; i < str.length; i++, j += 2) {
    const char = str.charCodeAt(i);
    buf[j] = char & 255;
    buf[j + 1] = char >> 8 & 255;
  }
  writable.disposeWriteSync();
}
function cUtf8String(writable, str) {
  const len = utils.utf8.encodedLength(str);
  const buf = writable.writeSync(len + 1);
  utils.utf8.encoder.encodeInto(str, buf);
  buf[len] = 0;
  writable.disposeWriteSync();
}
exports.bytes = bytes;
exports.bytesReversed = bytesReversed;
exports.cUtf8String = cUtf8String;
exports.rawString = rawString;
exports.utf16beString = utf16beString;
exports.utf16leString = utf16leString;
exports.utf8String = utf8String;
