// @mtcute/core/tl/inner-mtp.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
function _isAny() {
  const a = arguments;
  return function(o) {
    if (typeof o !== "object" || o === null) return false;
    for (let i = 0; i < a.length; i++) if (o._ === a[i]) return true;
    return false;
  };
}
const LAYER = 0;
const isAnyResPQ = /* @__PURE__ */ _isAny("mt_resPQ", "mt_req_pq_multi");
const isAnyP_Q_inner_data = /* @__PURE__ */ _isAny("mt_p_q_inner_data_dc", "mt_p_q_inner_data_temp_dc");
const isAnyServer_DH_Params = /* @__PURE__ */ _isAny("mt_server_DH_params_ok", "mt_req_DH_params");
const isAnyServer_DH_inner_data = /* @__PURE__ */ _isAny("mt_server_DH_inner_data");
const isAnyClient_DH_Inner_Data = /* @__PURE__ */ _isAny("mt_client_DH_inner_data");
const isAnySet_client_DH_params_answer = /* @__PURE__ */ _isAny("mt_dh_gen_ok", "mt_dh_gen_retry", "mt_dh_gen_fail", "mt_set_client_DH_params");
const isAnyBindAuthKeyInner = /* @__PURE__ */ _isAny("mt_bind_auth_key_inner");
const isAnyRpcError = /* @__PURE__ */ _isAny("mt_rpc_error");
const isAnyRpcDropAnswer = /* @__PURE__ */ _isAny("mt_rpc_answer_unknown", "mt_rpc_answer_dropped_running", "mt_rpc_answer_dropped", "mt_rpc_drop_answer");
const isAnyFutureSalt = /* @__PURE__ */ _isAny("mt_future_salt");
const isAnyFutureSalts = /* @__PURE__ */ _isAny("mt_future_salts", "mt_get_future_salts");
const isAnyPong = /* @__PURE__ */ _isAny("mt_pong", "mt_ping", "mt_ping_delay_disconnect");
const isAnyDestroySessionRes = /* @__PURE__ */ _isAny("mt_destroy_session_ok", "mt_destroy_session_none", "mt_destroy_session");
const isAnyNewSession = /* @__PURE__ */ _isAny("mt_new_session_created");
const isAnyMsgsAck = /* @__PURE__ */ _isAny("mt_msgs_ack");
const isAnyBadMsgNotification = /* @__PURE__ */ _isAny("mt_bad_msg_notification", "mt_bad_server_salt");
const isAnyMsgResendReq = /* @__PURE__ */ _isAny("mt_msg_resend_req");
const isAnyMsgsStateReq = /* @__PURE__ */ _isAny("mt_msgs_state_req");
const isAnyMsgsStateInfo = /* @__PURE__ */ _isAny("mt_msgs_state_info");
const isAnyMsgsAllInfo = /* @__PURE__ */ _isAny("mt_msgs_all_info");
const isAnyMsgDetailedInfo = /* @__PURE__ */ _isAny("mt_msg_detailed_info", "mt_msg_new_detailed_info");
const isAnyDestroyAuthKeyRes = /* @__PURE__ */ _isAny("mt_destroy_auth_key_ok", "mt_destroy_auth_key_none", "mt_destroy_auth_key_fail", "mt_destroy_auth_key");
const isAnyHttpWait = /* @__PURE__ */ _isAny("mt_http_wait");
exports.LAYER = LAYER;
exports.isAnyBadMsgNotification = isAnyBadMsgNotification;
exports.isAnyBindAuthKeyInner = isAnyBindAuthKeyInner;
exports.isAnyClient_DH_Inner_Data = isAnyClient_DH_Inner_Data;
exports.isAnyDestroyAuthKeyRes = isAnyDestroyAuthKeyRes;
exports.isAnyDestroySessionRes = isAnyDestroySessionRes;
exports.isAnyFutureSalt = isAnyFutureSalt;
exports.isAnyFutureSalts = isAnyFutureSalts;
exports.isAnyHttpWait = isAnyHttpWait;
exports.isAnyMsgDetailedInfo = isAnyMsgDetailedInfo;
exports.isAnyMsgResendReq = isAnyMsgResendReq;
exports.isAnyMsgsAck = isAnyMsgsAck;
exports.isAnyMsgsAllInfo = isAnyMsgsAllInfo;
exports.isAnyMsgsStateInfo = isAnyMsgsStateInfo;
exports.isAnyMsgsStateReq = isAnyMsgsStateReq;
exports.isAnyNewSession = isAnyNewSession;
exports.isAnyP_Q_inner_data = isAnyP_Q_inner_data;
exports.isAnyPong = isAnyPong;
exports.isAnyResPQ = isAnyResPQ;
exports.isAnyRpcDropAnswer = isAnyRpcDropAnswer;
exports.isAnyRpcError = isAnyRpcError;
exports.isAnyServer_DH_Params = isAnyServer_DH_Params;
exports.isAnyServer_DH_inner_data = isAnyServer_DH_inner_data;
exports.isAnySet_client_DH_params_answer = isAnySet_client_DH_params_answer;
