// @mtcute/core/tl/inner-tl.js
function _isAny() {
  const a = arguments;
  return function(o) {
    if (typeof o !== "object" || o === null) return false;
    for (let i = 0; i < a.length; i++) if (o._ === a[i]) return true;
    return false;
  };
}
const LAYER = 227;
class RpcError extends Error {
  constructor(code, text) {
    super("Telegram API error " + code + ": " + text);
    this.code = code;
    this.text = text;
  }
  static is(err, text) {
    return err.constructor === RpcError && (!text || err.text === text);
  }
  is(text) {
    return this.text === text;
  }
}
RpcError.fromTl = function(obj) {
  const err = new RpcError(obj.errorCode, obj.errorMessage);
  let match;
  if ((match = err.text.match(/^FLOOD_WAIT_(\d+)$/)) != null) {
    err.text = "FLOOD_WAIT_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^FILE_MIGRATE_(\d+)$/)) != null) {
    err.text = "FILE_MIGRATE_%d";
    err.newDc = parseInt(match[1]);
  } else if ((match = err.text.match(/^NETWORK_MIGRATE_(\d+)$/)) != null) {
    err.text = "NETWORK_MIGRATE_%d";
    err.newDc = parseInt(match[1]);
  } else if ((match = err.text.match(/^PHONE_MIGRATE_(\d+)$/)) != null) {
    err.text = "PHONE_MIGRATE_%d";
    err.newDc = parseInt(match[1]);
  } else if ((match = err.text.match(/^STATS_MIGRATE_(\d+)$/)) != null) {
    err.text = "STATS_MIGRATE_%d";
    err.dc = parseInt(match[1]);
  } else if ((match = err.text.match(/^EMAIL_UNCONFIRMED_(\d+)$/)) != null) {
    err.text = "EMAIL_UNCONFIRMED_%d";
    err.codeLength = parseInt(match[1]);
  } else if ((match = err.text.match(/^PASSWORD_TOO_FRESH_(\d+)$/)) != null) {
    err.text = "PASSWORD_TOO_FRESH_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^SESSION_TOO_FRESH_(\d+)$/)) != null) {
    err.text = "SESSION_TOO_FRESH_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^PREVIOUS_CHAT_IMPORT_ACTIVE_WAIT_(\d+)MIN$/)) != null) {
    err.text = "PREVIOUS_CHAT_IMPORT_ACTIVE_WAIT_%dMIN";
    err.minutes = parseInt(match[1]);
  } else if ((match = err.text.match(/^2FA_CONFIRM_WAIT_(\d+)$/)) != null) {
    err.text = "2FA_CONFIRM_WAIT_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^SLOWMODE_WAIT_(\d+)$/)) != null) {
    err.text = "SLOWMODE_WAIT_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^TAKEOUT_INIT_DELAY_(\d+)$/)) != null) {
    err.text = "TAKEOUT_INIT_DELAY_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^FILE_PART_(\d+)_MISSING$/)) != null) {
    err.text = "FILE_PART_%d_MISSING";
    err.which = parseInt(match[1]);
  } else if ((match = err.text.match(/^FLOOD_TEST_PHONE_WAIT_(\d+)$/)) != null) {
    err.text = "FLOOD_TEST_PHONE_WAIT_%d";
    err.seconds = parseInt(match[1]);
  } else if ((match = err.text.match(/^INTERDC_(\d+)_CALL_ERROR$/)) != null) {
    err.text = "INTERDC_%d_CALL_ERROR";
    err.dc = parseInt(match[1]);
  } else if ((match = err.text.match(/^INTERDC_(\d+)_CALL_RICH_ERROR$/)) != null) {
    err.text = "INTERDC_%d_CALL_RICH_ERROR";
    err.dc = parseInt(match[1]);
  } else if ((match = err.text.match(/^USER_MIGRATE_(\d+)$/)) != null) {
    err.text = "USER_MIGRATE_%d";
    err.newDc = parseInt(match[1]);
  }
  return err;
};
RpcError.BAD_REQUEST = 400;
RpcError.UNAUTHORIZED = 401;
RpcError.FORBIDDEN = 403;
RpcError.NOT_FOUND = 404;
RpcError.FLOOD = 420;
RpcError.SEE_OTHER = 303;
RpcError.NOT_ACCEPTABLE = 406;
RpcError.INTERNAL = 500;
const isAnyError = /* @__PURE__ */ _isAny("error");
const isAnyInputFileLocation = /* @__PURE__ */ _isAny("inputPeerPhotoFileLocationLegacy", "inputStickerSetThumbLegacy", "inputFileLocation", "inputEncryptedFileLocation", "inputDocumentFileLocation", "inputSecureFileLocation", "inputTakeoutFileLocation", "inputPhotoFileLocation", "inputPhotoLegacyFileLocation", "inputPeerPhotoFileLocation", "inputStickerSetThumb", "inputGroupCallStream");
const isAnyInputPeer = /* @__PURE__ */ _isAny("inputPeerEmpty", "inputPeerSelf", "inputPeerChat", "inputPeerUser", "inputPeerChannel", "inputPeerUserFromMessage", "inputPeerChannelFromMessage", "mtcute.dummyInputPeerMinUser", "mtcute.dummyInputPeerMinChannel");
const isAnyInputUser = /* @__PURE__ */ _isAny("inputUserEmpty", "inputUserSelf", "inputUser", "inputUserFromMessage");
const isAnyInputContact = /* @__PURE__ */ _isAny("inputPhoneContact");
const isAnyInputFile = /* @__PURE__ */ _isAny("inputFile", "inputFileBig", "inputFileStoryDocument");
const isAnyInputMedia = /* @__PURE__ */ _isAny("inputMediaEmpty", "inputMediaUploadedPhoto", "inputMediaPhoto", "inputMediaGeoPoint", "inputMediaContact", "inputMediaUploadedDocument", "inputMediaDocument", "inputMediaVenue", "inputMediaPhotoExternal", "inputMediaDocumentExternal", "inputMediaGame", "inputMediaInvoice", "inputMediaGeoLive", "inputMediaPoll", "inputMediaDice", "inputMediaStory", "inputMediaWebPage", "inputMediaPaidMedia", "inputMediaTodo", "inputMediaStakeDice");
const isAnyInputChatPhoto = /* @__PURE__ */ _isAny("inputChatPhotoEmpty", "inputChatUploadedPhoto", "inputChatPhoto");
const isAnyInputGeoPoint = /* @__PURE__ */ _isAny("inputGeoPointEmpty", "inputGeoPoint");
const isAnyInputPhoto = /* @__PURE__ */ _isAny("inputPhotoEmpty", "inputPhoto");
const isAnyPeer = /* @__PURE__ */ _isAny("peerUser", "peerChat", "peerChannel");
const isAnyUser = /* @__PURE__ */ _isAny("userEmpty", "user");
const isAnyUserProfilePhoto = /* @__PURE__ */ _isAny("userProfilePhotoEmpty", "userProfilePhoto");
const isAnyUserStatus = /* @__PURE__ */ _isAny("userStatusEmpty", "userStatusOnline", "userStatusOffline", "userStatusRecently", "userStatusLastWeek", "userStatusLastMonth");
const isAnyChat = /* @__PURE__ */ _isAny("chatEmpty", "chat", "chatForbidden", "channel", "channelForbidden");
const isAnyChatFull = /* @__PURE__ */ _isAny("chatFull", "channelFull");
const isAnyChatParticipant = /* @__PURE__ */ _isAny("chatParticipant", "chatParticipantCreator", "chatParticipantAdmin");
const isAnyChatParticipants = /* @__PURE__ */ _isAny("chatParticipantsForbidden", "chatParticipants");
const isAnyChatPhoto = /* @__PURE__ */ _isAny("chatPhotoEmpty", "chatPhoto");
const isAnyMessage = /* @__PURE__ */ _isAny("messageEmpty", "message", "messageService");
const isAnyMessageMedia = /* @__PURE__ */ _isAny("messageMediaEmpty", "messageMediaPhoto", "messageMediaGeo", "messageMediaContact", "messageMediaUnsupported", "messageMediaDocument", "messageMediaWebPage", "messageMediaVenue", "messageMediaGame", "messageMediaInvoice", "messageMediaGeoLive", "messageMediaPoll", "messageMediaDice", "messageMediaStory", "messageMediaGiveaway", "messageMediaGiveawayResults", "messageMediaPaidMedia", "messageMediaToDo", "messageMediaVideoStream");
const isAnyMessageAction = /* @__PURE__ */ _isAny("messageActionEmpty", "messageActionChatCreate", "messageActionChatEditTitle", "messageActionChatEditPhoto", "messageActionChatDeletePhoto", "messageActionChatAddUser", "messageActionChatDeleteUser", "messageActionChatJoinedByLink", "messageActionChannelCreate", "messageActionChatMigrateTo", "messageActionChannelMigrateFrom", "messageActionPinMessage", "messageActionHistoryClear", "messageActionGameScore", "messageActionPaymentSentMe", "messageActionPaymentSent", "messageActionPhoneCall", "messageActionScreenshotTaken", "messageActionCustomAction", "messageActionBotAllowed", "messageActionSecureValuesSentMe", "messageActionSecureValuesSent", "messageActionContactSignUp", "messageActionGeoProximityReached", "messageActionGroupCall", "messageActionInviteToGroupCall", "messageActionSetMessagesTTL", "messageActionGroupCallScheduled", "messageActionSetChatTheme", "messageActionChatJoinedByRequest", "messageActionWebViewDataSentMe", "messageActionWebViewDataSent", "messageActionGiftPremium", "messageActionTopicCreate", "messageActionTopicEdit", "messageActionSuggestProfilePhoto", "messageActionRequestedPeer", "messageActionSetChatWallPaper", "messageActionGiftCode", "messageActionGiveawayLaunch", "messageActionGiveawayResults", "messageActionBoostApply", "messageActionRequestedPeerSentMe", "messageActionPaymentRefunded", "messageActionGiftStars", "messageActionPrizeStars", "messageActionStarGift", "messageActionStarGiftUnique", "messageActionPaidMessagesRefunded", "messageActionPaidMessagesPrice", "messageActionConferenceCall", "messageActionTodoCompletions", "messageActionTodoAppendTasks", "messageActionSuggestedPostApproval", "messageActionSuggestedPostSuccess", "messageActionSuggestedPostRefund", "messageActionGiftTon", "messageActionSuggestBirthday", "messageActionStarGiftPurchaseOffer", "messageActionStarGiftPurchaseOfferDeclined", "messageActionNewCreatorPending", "messageActionChangeCreator", "messageActionNoForwardsToggle", "messageActionNoForwardsRequest", "messageActionPollAppendAnswer", "messageActionPollDeleteAnswer", "messageActionManagedBotCreated");
const isAnyDialog = /* @__PURE__ */ _isAny("dialog", "dialogFolder");
const isAnyPhoto = /* @__PURE__ */ _isAny("photoEmpty", "photo");
const isAnyPhotoSize = /* @__PURE__ */ _isAny("photoSizeEmpty", "photoSize", "photoCachedSize", "photoStrippedSize", "photoSizeProgressive", "photoPathSize");
const isAnyGeoPoint = /* @__PURE__ */ _isAny("geoPointEmpty", "geoPoint");
const isAnyInputNotifyPeer = /* @__PURE__ */ _isAny("inputNotifyPeer", "inputNotifyUsers", "inputNotifyChats", "inputNotifyBroadcasts", "inputNotifyForumTopic");
const isAnyInputPeerNotifySettings = /* @__PURE__ */ _isAny("inputPeerNotifySettings");
const isAnyPeerNotifySettings = /* @__PURE__ */ _isAny("peerNotifySettings");
const isAnyPeerSettings = /* @__PURE__ */ _isAny("peerSettings");
const isAnyWallPaper = /* @__PURE__ */ _isAny("wallPaper", "wallPaperNoFile");
const isAnyReportReason = /* @__PURE__ */ _isAny("inputReportReasonSpam", "inputReportReasonViolence", "inputReportReasonPornography", "inputReportReasonChildAbuse", "inputReportReasonOther", "inputReportReasonCopyright", "inputReportReasonGeoIrrelevant", "inputReportReasonFake", "inputReportReasonIllegalDrugs", "inputReportReasonPersonalDetails");
const isAnyUserFull = /* @__PURE__ */ _isAny("userFull");
const isAnyContact = /* @__PURE__ */ _isAny("contact");
const isAnyImportedContact = /* @__PURE__ */ _isAny("importedContact");
const isAnyContactStatus = /* @__PURE__ */ _isAny("contactStatus");
const isAnyMessagesFilter = /* @__PURE__ */ _isAny("inputMessagesFilterEmpty", "inputMessagesFilterPhotos", "inputMessagesFilterVideo", "inputMessagesFilterPhotoVideo", "inputMessagesFilterDocument", "inputMessagesFilterUrl", "inputMessagesFilterGif", "inputMessagesFilterVoice", "inputMessagesFilterMusic", "inputMessagesFilterChatPhotos", "inputMessagesFilterPhoneCalls", "inputMessagesFilterRoundVoice", "inputMessagesFilterRoundVideo", "inputMessagesFilterMyMentions", "inputMessagesFilterGeo", "inputMessagesFilterContacts", "inputMessagesFilterPinned", "inputMessagesFilterPoll");
const isAnyUpdate = /* @__PURE__ */ _isAny("updateNewMessage", "updateMessageID", "updateDeleteMessages", "updateUserTyping", "updateChatUserTyping", "updateChatParticipants", "updateUserStatus", "updateUserName", "updateNewAuthorization", "updateNewEncryptedMessage", "updateEncryptedChatTyping", "updateEncryption", "updateEncryptedMessagesRead", "updateChatParticipantAdd", "updateChatParticipantDelete", "updateDcOptions", "updateNotifySettings", "updateServiceNotification", "updatePrivacy", "updateUserPhone", "updateReadHistoryInbox", "updateReadHistoryOutbox", "updateWebPage", "updateReadMessagesContents", "updateChannelTooLong", "updateChannel", "updateNewChannelMessage", "updateReadChannelInbox", "updateDeleteChannelMessages", "updateChannelMessageViews", "updateChatParticipantAdmin", "updateNewStickerSet", "updateStickerSetsOrder", "updateStickerSets", "updateSavedGifs", "updateBotInlineQuery", "updateBotInlineSend", "updateEditChannelMessage", "updateBotCallbackQuery", "updateEditMessage", "updateInlineBotCallbackQuery", "updateReadChannelOutbox", "updateDraftMessage", "updateReadFeaturedStickers", "updateRecentStickers", "updateConfig", "updatePtsChanged", "updateChannelWebPage", "updateDialogPinned", "updatePinnedDialogs", "updateBotWebhookJSON", "updateBotWebhookJSONQuery", "updateBotShippingQuery", "updateBotPrecheckoutQuery", "updatePhoneCall", "updateLangPackTooLong", "updateLangPack", "updateFavedStickers", "updateChannelReadMessagesContents", "updateContactsReset", "updateChannelAvailableMessages", "updateDialogUnreadMark", "updateMessagePoll", "updateChatDefaultBannedRights", "updateFolderPeers", "updatePeerSettings", "updatePeerLocated", "updateNewScheduledMessage", "updateDeleteScheduledMessages", "updateTheme", "updateGeoLiveViewed", "updateLoginToken", "updateMessagePollVote", "updateDialogFilter", "updateDialogFilterOrder", "updateDialogFilters", "updatePhoneCallSignalingData", "updateChannelMessageForwards", "updateReadChannelDiscussionInbox", "updateReadChannelDiscussionOutbox", "updatePeerBlocked", "updateChannelUserTyping", "updatePinnedMessages", "updatePinnedChannelMessages", "updateChat", "updateGroupCallParticipants", "updateGroupCall", "updatePeerHistoryTTL", "updateChatParticipant", "updateChannelParticipant", "updateBotStopped", "updateGroupCallConnection", "updateBotCommands", "updatePendingJoinRequests", "updateBotChatInviteRequester", "updateMessageReactions", "updateAttachMenuBots", "updateWebViewResultSent", "updateBotMenuButton", "updateSavedRingtones", "updateTranscribedAudio", "updateReadFeaturedEmojiStickers", "updateUserEmojiStatus", "updateRecentEmojiStatuses", "updateRecentReactions", "updateMoveStickerSetToTop", "updateMessageExtendedMedia", "updateUser", "updateAutoSaveSettings", "updateStory", "updateReadStories", "updateStoryID", "updateStoriesStealthMode", "updateSentStoryReaction", "updateBotChatBoost", "updateChannelViewForumAsMessages", "updatePeerWallpaper", "updateBotMessageReaction", "updateBotMessageReactions", "updateSavedDialogPinned", "updatePinnedSavedDialogs", "updateSavedReactionTags", "updateSmsJob", "updateQuickReplies", "updateNewQuickReply", "updateDeleteQuickReply", "updateQuickReplyMessage", "updateDeleteQuickReplyMessages", "updateBotBusinessConnect", "updateBotNewBusinessMessage", "updateBotEditBusinessMessage", "updateBotDeleteBusinessMessage", "updateNewStoryReaction", "updateStarsBalance", "updateBusinessBotCallbackQuery", "updateStarsRevenueStatus", "updateBotPurchasedPaidMedia", "updatePaidReactionPrivacy", "updateSentPhoneCode", "updateGroupCallChainBlocks", "updateReadMonoForumInbox", "updateReadMonoForumOutbox", "updateMonoForumNoPaidException", "updateGroupCallMessage", "updateGroupCallEncryptedMessage", "updatePinnedForumTopic", "updatePinnedForumTopics", "updateDeleteGroupCallMessages", "updateStarGiftAuctionState", "updateStarGiftAuctionUserState", "updateEmojiGameInfo", "updateStarGiftCraftFail", "updateChatParticipantRank", "updateManagedBot", "updateBotGuestChatQuery", "updateAiComposeTones", "updateJoinChatWebViewDecision", "updateNewBotConnection", "updateWebBrowserSettings", "updateWebBrowserException", "updateChannelPinnedTopic", "updateChannelPinnedTopics", "mtcute.dummyUpdate");
const isAnyUpdates = /* @__PURE__ */ _isAny("updatesTooLong", "updateShortMessage", "updateShortChatMessage", "updateShort", "updatesCombined", "updates", "updateShortSentMessage");
const isAnyDcOption = /* @__PURE__ */ _isAny("dcOption");
const isAnyConfig = /* @__PURE__ */ _isAny("config");
const isAnyNearestDc = /* @__PURE__ */ _isAny("nearestDc");
const isAnyEncryptedChat = /* @__PURE__ */ _isAny("encryptedChatEmpty", "encryptedChatWaiting", "encryptedChatRequested", "encryptedChat", "encryptedChatDiscarded");
const isAnyInputEncryptedChat = /* @__PURE__ */ _isAny("inputEncryptedChat");
const isAnyEncryptedFile = /* @__PURE__ */ _isAny("encryptedFileEmpty", "encryptedFile");
const isAnyInputEncryptedFile = /* @__PURE__ */ _isAny("inputEncryptedFileEmpty", "inputEncryptedFileUploaded", "inputEncryptedFile", "inputEncryptedFileBigUploaded");
const isAnyEncryptedMessage = /* @__PURE__ */ _isAny("encryptedMessage", "encryptedMessageService");
const isAnyInputDocument = /* @__PURE__ */ _isAny("inputDocumentEmpty", "inputDocument");
const isAnyDocument = /* @__PURE__ */ _isAny("documentEmpty", "document");
const isAnyNotifyPeer = /* @__PURE__ */ _isAny("notifyPeer", "notifyUsers", "notifyChats", "notifyBroadcasts", "notifyForumTopic");
const isAnySendMessageAction = /* @__PURE__ */ _isAny("sendMessageTypingAction", "sendMessageCancelAction", "sendMessageRecordVideoAction", "sendMessageUploadVideoAction", "sendMessageRecordAudioAction", "sendMessageUploadAudioAction", "sendMessageUploadPhotoAction", "sendMessageUploadDocumentAction", "sendMessageGeoLocationAction", "sendMessageChooseContactAction", "sendMessageGamePlayAction", "sendMessageRecordRoundAction", "sendMessageUploadRoundAction", "speakingInGroupCallAction", "sendMessageHistoryImportAction", "sendMessageChooseStickerAction", "sendMessageEmojiInteraction", "sendMessageEmojiInteractionSeen", "sendMessageTextDraftAction", "inputSendMessageRichMessageDraftAction", "sendMessageRichMessageDraftAction");
const isAnyInputPrivacyKey = /* @__PURE__ */ _isAny("inputPrivacyKeyStatusTimestamp", "inputPrivacyKeyChatInvite", "inputPrivacyKeyPhoneCall", "inputPrivacyKeyPhoneP2P", "inputPrivacyKeyForwards", "inputPrivacyKeyProfilePhoto", "inputPrivacyKeyPhoneNumber", "inputPrivacyKeyAddedByPhone", "inputPrivacyKeyVoiceMessages", "inputPrivacyKeyAbout", "inputPrivacyKeyBirthday", "inputPrivacyKeyStarGiftsAutoSave", "inputPrivacyKeyNoPaidMessages", "inputPrivacyKeySavedMusic");
const isAnyPrivacyKey = /* @__PURE__ */ _isAny("privacyKeyStatusTimestamp", "privacyKeyChatInvite", "privacyKeyPhoneCall", "privacyKeyPhoneP2P", "privacyKeyForwards", "privacyKeyProfilePhoto", "privacyKeyPhoneNumber", "privacyKeyAddedByPhone", "privacyKeyVoiceMessages", "privacyKeyAbout", "privacyKeyBirthday", "privacyKeyStarGiftsAutoSave", "privacyKeyNoPaidMessages", "privacyKeySavedMusic");
const isAnyInputPrivacyRule = /* @__PURE__ */ _isAny("inputPrivacyValueAllowContacts", "inputPrivacyValueAllowAll", "inputPrivacyValueAllowUsers", "inputPrivacyValueDisallowContacts", "inputPrivacyValueDisallowAll", "inputPrivacyValueDisallowUsers", "inputPrivacyValueAllowChatParticipants", "inputPrivacyValueDisallowChatParticipants", "inputPrivacyValueAllowCloseFriends", "inputPrivacyValueAllowPremium", "inputPrivacyValueAllowBots", "inputPrivacyValueDisallowBots");
const isAnyPrivacyRule = /* @__PURE__ */ _isAny("privacyValueAllowContacts", "privacyValueAllowAll", "privacyValueAllowUsers", "privacyValueDisallowContacts", "privacyValueDisallowAll", "privacyValueDisallowUsers", "privacyValueAllowChatParticipants", "privacyValueDisallowChatParticipants", "privacyValueAllowCloseFriends", "privacyValueAllowPremium", "privacyValueAllowBots", "privacyValueDisallowBots");
const isAnyAccountDaysTTL = /* @__PURE__ */ _isAny("accountDaysTTL");
const isAnyDocumentAttribute = /* @__PURE__ */ _isAny("documentAttributeImageSize", "documentAttributeAnimated", "documentAttributeSticker", "documentAttributeVideo", "documentAttributeAudio", "documentAttributeFilename", "documentAttributeHasStickers", "documentAttributeCustomEmoji");
const isAnyStickerPack = /* @__PURE__ */ _isAny("stickerPack");
const isAnyWebPage = /* @__PURE__ */ _isAny("webPageEmpty", "webPagePending", "webPage", "webPageNotModified");
const isAnyAuthorization = /* @__PURE__ */ _isAny("authorization");
const isAnyReceivedNotifyMessage = /* @__PURE__ */ _isAny("receivedNotifyMessage");
const isAnyExportedChatInvite = /* @__PURE__ */ _isAny("chatInviteExported", "chatInvitePublicJoinRequests");
const isAnyChatInvite = /* @__PURE__ */ _isAny("chatInviteAlready", "chatInvite", "chatInvitePeek");
const isAnyInputStickerSet = /* @__PURE__ */ _isAny("inputStickerSetEmpty", "inputStickerSetID", "inputStickerSetShortName", "inputStickerSetAnimatedEmoji", "inputStickerSetDice", "inputStickerSetAnimatedEmojiAnimations", "inputStickerSetPremiumGifts", "inputStickerSetEmojiGenericAnimations", "inputStickerSetEmojiDefaultStatuses", "inputStickerSetEmojiDefaultTopicIcons", "inputStickerSetEmojiChannelDefaultStatuses", "inputStickerSetTonGifts");
const isAnyStickerSet = /* @__PURE__ */ _isAny("stickerSet");
const isAnyBotCommand = /* @__PURE__ */ _isAny("botCommand");
const isAnyBotInfo = /* @__PURE__ */ _isAny("botInfo");
const isAnyKeyboardButton = /* @__PURE__ */ _isAny("keyboardButton", "keyboardButtonUrl", "keyboardButtonCallback", "keyboardButtonRequestPhone", "keyboardButtonRequestGeoLocation", "keyboardButtonSwitchInline", "keyboardButtonGame", "keyboardButtonBuy", "keyboardButtonUrlAuth", "inputKeyboardButtonUrlAuth", "keyboardButtonRequestPoll", "inputKeyboardButtonUserProfile", "keyboardButtonUserProfile", "keyboardButtonWebView", "keyboardButtonSimpleWebView", "keyboardButtonRequestPeer", "inputKeyboardButtonRequestPeer", "keyboardButtonCopy");
const isAnyKeyboardButtonRow = /* @__PURE__ */ _isAny("keyboardButtonRow");
const isAnyReplyMarkup = /* @__PURE__ */ _isAny("replyKeyboardHide", "replyKeyboardForceReply", "replyKeyboardMarkup", "replyInlineMarkup");
const isAnyMessageEntity = /* @__PURE__ */ _isAny("messageEntityUnknown", "messageEntityMention", "messageEntityHashtag", "messageEntityBotCommand", "messageEntityUrl", "messageEntityEmail", "messageEntityBold", "messageEntityItalic", "messageEntityCode", "messageEntityPre", "messageEntityTextUrl", "messageEntityMentionName", "inputMessageEntityMentionName", "messageEntityPhone", "messageEntityCashtag", "messageEntityUnderline", "messageEntityStrike", "messageEntityBankCard", "messageEntitySpoiler", "messageEntityCustomEmoji", "messageEntityBlockquote", "messageEntityFormattedDate", "messageEntityDiffInsert", "messageEntityDiffReplace", "messageEntityDiffDelete");
const isAnyInputChannel = /* @__PURE__ */ _isAny("inputChannelEmpty", "inputChannel", "inputChannelFromMessage");
const isAnyMessageRange = /* @__PURE__ */ _isAny("messageRange");
const isAnyChannelMessagesFilter = /* @__PURE__ */ _isAny("channelMessagesFilterEmpty", "channelMessagesFilter");
const isAnyChannelParticipant = /* @__PURE__ */ _isAny("channelParticipant", "channelParticipantSelf", "channelParticipantCreator", "channelParticipantAdmin", "channelParticipantBanned", "channelParticipantLeft");
const isAnyChannelParticipantsFilter = /* @__PURE__ */ _isAny("channelParticipantsRecent", "channelParticipantsAdmins", "channelParticipantsKicked", "channelParticipantsBots", "channelParticipantsBanned", "channelParticipantsSearch", "channelParticipantsContacts", "channelParticipantsMentions");
const isAnyInputBotInlineMessage = /* @__PURE__ */ _isAny("inputBotInlineMessageMediaAuto", "inputBotInlineMessageText", "inputBotInlineMessageMediaGeo", "inputBotInlineMessageMediaVenue", "inputBotInlineMessageMediaContact", "inputBotInlineMessageGame", "inputBotInlineMessageMediaInvoice", "inputBotInlineMessageMediaWebPage", "inputBotInlineMessageRichMessage");
const isAnyInputBotInlineResult = /* @__PURE__ */ _isAny("inputBotInlineResult", "inputBotInlineResultPhoto", "inputBotInlineResultDocument", "inputBotInlineResultGame");
const isAnyBotInlineMessage = /* @__PURE__ */ _isAny("botInlineMessageMediaAuto", "botInlineMessageText", "botInlineMessageMediaGeo", "botInlineMessageMediaVenue", "botInlineMessageMediaContact", "botInlineMessageMediaInvoice", "botInlineMessageMediaWebPage", "botInlineMessageRichMessage");
const isAnyBotInlineResult = /* @__PURE__ */ _isAny("botInlineResult", "botInlineMediaResult");
const isAnyExportedMessageLink = /* @__PURE__ */ _isAny("exportedMessageLink");
const isAnyMessageFwdHeader = /* @__PURE__ */ _isAny("messageFwdHeader");
const isAnyInputBotInlineMessageID = /* @__PURE__ */ _isAny("inputBotInlineMessageID", "inputBotInlineMessageID64");
const isAnyInlineBotSwitchPM = /* @__PURE__ */ _isAny("inlineBotSwitchPM");
const isAnyTopPeer = /* @__PURE__ */ _isAny("topPeer");
const isAnyTopPeerCategory = /* @__PURE__ */ _isAny("topPeerCategoryBotsPM", "topPeerCategoryBotsInline", "topPeerCategoryCorrespondents", "topPeerCategoryGroups", "topPeerCategoryChannels", "topPeerCategoryPhoneCalls", "topPeerCategoryForwardUsers", "topPeerCategoryForwardChats", "topPeerCategoryBotsApp", "topPeerCategoryBotsGuestChat");
const isAnyTopPeerCategoryPeers = /* @__PURE__ */ _isAny("topPeerCategoryPeers");
const isAnyDraftMessage = /* @__PURE__ */ _isAny("draftMessageEmpty", "draftMessage");
const isAnyStickerSetCovered = /* @__PURE__ */ _isAny("stickerSetCovered", "stickerSetMultiCovered", "stickerSetFullCovered", "stickerSetNoCovered");
const isAnyMaskCoords = /* @__PURE__ */ _isAny("maskCoords");
const isAnyInputStickeredMedia = /* @__PURE__ */ _isAny("inputStickeredMediaPhoto", "inputStickeredMediaDocument");
const isAnyGame = /* @__PURE__ */ _isAny("game");
const isAnyInputGame = /* @__PURE__ */ _isAny("inputGameID", "inputGameShortName");
const isAnyHighScore = /* @__PURE__ */ _isAny("highScore");
const isAnyRichText = /* @__PURE__ */ _isAny("textEmpty", "textPlain", "textBold", "textItalic", "textUnderline", "textStrike", "textFixed", "textUrl", "textEmail", "textConcat", "textSubscript", "textSuperscript", "textMarked", "textPhone", "textImage", "textAnchor", "textMath", "textCustomEmoji", "textSpoiler", "textMention", "textHashtag", "textBotCommand", "textCashtag", "textAutoUrl", "textAutoEmail", "textAutoPhone", "textBankCard", "textMentionName", "textDate");
const isAnyPageBlock = /* @__PURE__ */ _isAny("pageBlockUnsupported", "pageBlockTitle", "pageBlockSubtitle", "pageBlockAuthorDate", "pageBlockHeader", "pageBlockSubheader", "pageBlockParagraph", "pageBlockPreformatted", "pageBlockFooter", "pageBlockDivider", "pageBlockAnchor", "pageBlockList", "pageBlockBlockquote", "pageBlockPullquote", "pageBlockPhoto", "pageBlockVideo", "pageBlockCover", "pageBlockEmbed", "pageBlockEmbedPost", "pageBlockCollage", "pageBlockSlideshow", "pageBlockChannel", "pageBlockAudio", "pageBlockKicker", "pageBlockTable", "pageBlockOrderedList", "pageBlockDetails", "pageBlockRelatedArticles", "pageBlockMap", "pageBlockHeading1", "pageBlockHeading2", "pageBlockHeading3", "pageBlockHeading4", "pageBlockHeading5", "pageBlockHeading6", "pageBlockMath", "pageBlockThinking", "inputPageBlockMap", "pageBlockBlockquoteBlocks");
const isAnyPhoneCallDiscardReason = /* @__PURE__ */ _isAny("phoneCallDiscardReasonMissed", "phoneCallDiscardReasonDisconnect", "phoneCallDiscardReasonHangup", "phoneCallDiscardReasonBusy", "phoneCallDiscardReasonMigrateConferenceCall");
const isAnyDataJSON = /* @__PURE__ */ _isAny("dataJSON");
const isAnyLabeledPrice = /* @__PURE__ */ _isAny("labeledPrice");
const isAnyInvoice = /* @__PURE__ */ _isAny("invoice");
const isAnyPaymentCharge = /* @__PURE__ */ _isAny("paymentCharge");
const isAnyPostAddress = /* @__PURE__ */ _isAny("postAddress");
const isAnyPaymentRequestedInfo = /* @__PURE__ */ _isAny("paymentRequestedInfo");
const isAnyPaymentSavedCredentials = /* @__PURE__ */ _isAny("paymentSavedCredentialsCard");
const isAnyWebDocument = /* @__PURE__ */ _isAny("webDocument", "webDocumentNoProxy");
const isAnyInputWebDocument = /* @__PURE__ */ _isAny("inputWebDocument");
const isAnyInputWebFileLocation = /* @__PURE__ */ _isAny("inputWebFileLocation", "inputWebFileGeoPointLocation", "inputWebFileAudioAlbumThumbLocation");
const isAnyInputPaymentCredentials = /* @__PURE__ */ _isAny("inputPaymentCredentialsSaved", "inputPaymentCredentials", "inputPaymentCredentialsApplePay", "inputPaymentCredentialsGooglePay");
const isAnyShippingOption = /* @__PURE__ */ _isAny("shippingOption");
const isAnyInputStickerSetItem = /* @__PURE__ */ _isAny("inputStickerSetItem");
const isAnyInputPhoneCall = /* @__PURE__ */ _isAny("inputPhoneCall");
const isAnyPhoneCall = /* @__PURE__ */ _isAny("phoneCallEmpty", "phoneCallWaiting", "phoneCallRequested", "phoneCallAccepted", "phoneCall", "phoneCallDiscarded");
const isAnyPhoneConnection = /* @__PURE__ */ _isAny("phoneConnection", "phoneConnectionWebrtc");
const isAnyPhoneCallProtocol = /* @__PURE__ */ _isAny("phoneCallProtocol");
const isAnyCdnPublicKey = /* @__PURE__ */ _isAny("cdnPublicKey");
const isAnyCdnConfig = /* @__PURE__ */ _isAny("cdnConfig");
const isAnyLangPackString = /* @__PURE__ */ _isAny("langPackString", "langPackStringPluralized", "langPackStringDeleted");
const isAnyLangPackDifference = /* @__PURE__ */ _isAny("langPackDifference");
const isAnyLangPackLanguage = /* @__PURE__ */ _isAny("langPackLanguage");
const isAnyChannelAdminLogEventAction = /* @__PURE__ */ _isAny("channelAdminLogEventActionChangeTitle", "channelAdminLogEventActionChangeAbout", "channelAdminLogEventActionChangeUsername", "channelAdminLogEventActionChangePhoto", "channelAdminLogEventActionToggleInvites", "channelAdminLogEventActionToggleSignatures", "channelAdminLogEventActionUpdatePinned", "channelAdminLogEventActionEditMessage", "channelAdminLogEventActionDeleteMessage", "channelAdminLogEventActionParticipantJoin", "channelAdminLogEventActionParticipantLeave", "channelAdminLogEventActionParticipantInvite", "channelAdminLogEventActionParticipantToggleBan", "channelAdminLogEventActionParticipantToggleAdmin", "channelAdminLogEventActionChangeStickerSet", "channelAdminLogEventActionTogglePreHistoryHidden", "channelAdminLogEventActionDefaultBannedRights", "channelAdminLogEventActionStopPoll", "channelAdminLogEventActionChangeLinkedChat", "channelAdminLogEventActionChangeLocation", "channelAdminLogEventActionToggleSlowMode", "channelAdminLogEventActionStartGroupCall", "channelAdminLogEventActionDiscardGroupCall", "channelAdminLogEventActionParticipantMute", "channelAdminLogEventActionParticipantUnmute", "channelAdminLogEventActionToggleGroupCallSetting", "channelAdminLogEventActionParticipantJoinByInvite", "channelAdminLogEventActionExportedInviteDelete", "channelAdminLogEventActionExportedInviteRevoke", "channelAdminLogEventActionExportedInviteEdit", "channelAdminLogEventActionParticipantVolume", "channelAdminLogEventActionChangeHistoryTTL", "channelAdminLogEventActionParticipantJoinByRequest", "channelAdminLogEventActionToggleNoForwards", "channelAdminLogEventActionSendMessage", "channelAdminLogEventActionChangeAvailableReactions", "channelAdminLogEventActionChangeUsernames", "channelAdminLogEventActionToggleForum", "channelAdminLogEventActionCreateTopic", "channelAdminLogEventActionEditTopic", "channelAdminLogEventActionDeleteTopic", "channelAdminLogEventActionPinTopic", "channelAdminLogEventActionToggleAntiSpam", "channelAdminLogEventActionChangePeerColor", "channelAdminLogEventActionChangeProfilePeerColor", "channelAdminLogEventActionChangeWallpaper", "channelAdminLogEventActionChangeEmojiStatus", "channelAdminLogEventActionChangeEmojiStickerSet", "channelAdminLogEventActionToggleSignatureProfiles", "channelAdminLogEventActionParticipantSubExtend", "channelAdminLogEventActionToggleAutotranslation", "channelAdminLogEventActionParticipantEditRank");
const isAnyChannelAdminLogEvent = /* @__PURE__ */ _isAny("channelAdminLogEvent");
const isAnyChannelAdminLogEventsFilter = /* @__PURE__ */ _isAny("channelAdminLogEventsFilter");
const isAnyPopularContact = /* @__PURE__ */ _isAny("popularContact");
const isAnyRecentMeUrl = /* @__PURE__ */ _isAny("recentMeUrlUnknown", "recentMeUrlUser", "recentMeUrlChat", "recentMeUrlChatInvite", "recentMeUrlStickerSet");
const isAnyInputSingleMedia = /* @__PURE__ */ _isAny("inputSingleMedia");
const isAnyWebAuthorization = /* @__PURE__ */ _isAny("webAuthorization");
const isAnyInputMessage = /* @__PURE__ */ _isAny("inputMessageID", "inputMessageReplyTo", "inputMessagePinned", "inputMessageCallbackQuery");
const isAnyInputDialogPeer = /* @__PURE__ */ _isAny("inputDialogPeer", "inputDialogPeerFolder");
const isAnyDialogPeer = /* @__PURE__ */ _isAny("dialogPeer", "dialogPeerFolder");
const isAnyFileHash = /* @__PURE__ */ _isAny("fileHash");
const isAnyInputClientProxy = /* @__PURE__ */ _isAny("inputClientProxy");
const isAnyInputSecureFile = /* @__PURE__ */ _isAny("inputSecureFileUploaded", "inputSecureFile");
const isAnySecureFile = /* @__PURE__ */ _isAny("secureFileEmpty", "secureFile");
const isAnySecureData = /* @__PURE__ */ _isAny("secureData");
const isAnySecurePlainData = /* @__PURE__ */ _isAny("securePlainPhone", "securePlainEmail");
const isAnySecureValueType = /* @__PURE__ */ _isAny("secureValueTypePersonalDetails", "secureValueTypePassport", "secureValueTypeDriverLicense", "secureValueTypeIdentityCard", "secureValueTypeInternalPassport", "secureValueTypeAddress", "secureValueTypeUtilityBill", "secureValueTypeBankStatement", "secureValueTypeRentalAgreement", "secureValueTypePassportRegistration", "secureValueTypeTemporaryRegistration", "secureValueTypePhone", "secureValueTypeEmail");
const isAnySecureValue = /* @__PURE__ */ _isAny("secureValue");
const isAnyInputSecureValue = /* @__PURE__ */ _isAny("inputSecureValue");
const isAnySecureValueHash = /* @__PURE__ */ _isAny("secureValueHash");
const isAnySecureValueError = /* @__PURE__ */ _isAny("secureValueErrorData", "secureValueErrorFrontSide", "secureValueErrorReverseSide", "secureValueErrorSelfie", "secureValueErrorFile", "secureValueErrorFiles", "secureValueError", "secureValueErrorTranslationFile", "secureValueErrorTranslationFiles");
const isAnySecureCredentialsEncrypted = /* @__PURE__ */ _isAny("secureCredentialsEncrypted");
const isAnySavedContact = /* @__PURE__ */ _isAny("savedPhoneContact");
const isAnyPasswordKdfAlgo = /* @__PURE__ */ _isAny("passwordKdfAlgoUnknown", "passwordKdfAlgoSHA256SHA256PBKDF2HMACSHA512iter100000SHA256ModPow");
const isAnySecurePasswordKdfAlgo = /* @__PURE__ */ _isAny("securePasswordKdfAlgoUnknown", "securePasswordKdfAlgoPBKDF2HMACSHA512iter100000", "securePasswordKdfAlgoSHA512");
const isAnySecureSecretSettings = /* @__PURE__ */ _isAny("secureSecretSettings");
const isAnyInputCheckPasswordSRP = /* @__PURE__ */ _isAny("inputCheckPasswordEmpty", "inputCheckPasswordSRP");
const isAnySecureRequiredType = /* @__PURE__ */ _isAny("secureRequiredType", "secureRequiredTypeOneOf");
const isAnyInputAppEvent = /* @__PURE__ */ _isAny("inputAppEvent");
const isAnyJSONObjectValue = /* @__PURE__ */ _isAny("jsonObjectValue");
const isAnyJSONValue = /* @__PURE__ */ _isAny("jsonNull", "jsonBool", "jsonNumber", "jsonString", "jsonArray", "jsonObject");
const isAnyPageTableCell = /* @__PURE__ */ _isAny("pageTableCell");
const isAnyPageTableRow = /* @__PURE__ */ _isAny("pageTableRow");
const isAnyPageCaption = /* @__PURE__ */ _isAny("pageCaption");
const isAnyPageListItem = /* @__PURE__ */ _isAny("pageListItemText", "pageListItemBlocks");
const isAnyPageListOrderedItem = /* @__PURE__ */ _isAny("pageListOrderedItemText", "pageListOrderedItemBlocks");
const isAnyPageRelatedArticle = /* @__PURE__ */ _isAny("pageRelatedArticle");
const isAnyPage = /* @__PURE__ */ _isAny("page");
const isAnyPollAnswer = /* @__PURE__ */ _isAny("pollAnswer", "inputPollAnswer");
const isAnyPoll = /* @__PURE__ */ _isAny("poll");
const isAnyPollAnswerVoters = /* @__PURE__ */ _isAny("pollAnswerVoters");
const isAnyPollResults = /* @__PURE__ */ _isAny("pollResults");
const isAnyChatOnlines = /* @__PURE__ */ _isAny("chatOnlines");
const isAnyStatsURL = /* @__PURE__ */ _isAny("statsURL");
const isAnyChatAdminRights = /* @__PURE__ */ _isAny("chatAdminRights");
const isAnyChatBannedRights = /* @__PURE__ */ _isAny("chatBannedRights");
const isAnyInputWallPaper = /* @__PURE__ */ _isAny("inputWallPaper", "inputWallPaperSlug", "inputWallPaperNoFile");
const isAnyCodeSettings = /* @__PURE__ */ _isAny("codeSettings");
const isAnyWallPaperSettings = /* @__PURE__ */ _isAny("wallPaperSettings");
const isAnyAutoDownloadSettings = /* @__PURE__ */ _isAny("autoDownloadSettings");
const isAnyEmojiKeyword = /* @__PURE__ */ _isAny("emojiKeyword", "emojiKeywordDeleted");
const isAnyEmojiKeywordsDifference = /* @__PURE__ */ _isAny("emojiKeywordsDifference");
const isAnyEmojiURL = /* @__PURE__ */ _isAny("emojiURL");
const isAnyEmojiLanguage = /* @__PURE__ */ _isAny("emojiLanguage");
const isAnyFolder = /* @__PURE__ */ _isAny("folder");
const isAnyInputFolderPeer = /* @__PURE__ */ _isAny("inputFolderPeer");
const isAnyFolderPeer = /* @__PURE__ */ _isAny("folderPeer");
const isAnyUrlAuthResult = /* @__PURE__ */ _isAny("urlAuthResultRequest", "urlAuthResultAccepted", "urlAuthResultDefault");
const isAnyChannelLocation = /* @__PURE__ */ _isAny("channelLocationEmpty", "channelLocation");
const isAnyPeerLocated = /* @__PURE__ */ _isAny("peerLocated", "peerSelfLocated");
const isAnyRestrictionReason = /* @__PURE__ */ _isAny("restrictionReason");
const isAnyInputTheme = /* @__PURE__ */ _isAny("inputTheme", "inputThemeSlug");
const isAnyTheme = /* @__PURE__ */ _isAny("theme");
const isAnyBaseTheme = /* @__PURE__ */ _isAny("baseThemeClassic", "baseThemeDay", "baseThemeNight", "baseThemeTinted", "baseThemeArctic");
const isAnyInputThemeSettings = /* @__PURE__ */ _isAny("inputThemeSettings");
const isAnyThemeSettings = /* @__PURE__ */ _isAny("themeSettings");
const isAnyWebPageAttribute = /* @__PURE__ */ _isAny("webPageAttributeTheme", "webPageAttributeStory", "webPageAttributeStickerSet", "webPageAttributeUniqueStarGift", "webPageAttributeStarGiftCollection", "webPageAttributeStarGiftAuction", "webPageAttributeAiComposeTone");
const isAnyBankCardOpenUrl = /* @__PURE__ */ _isAny("bankCardOpenUrl");
const isAnyDialogFilter = /* @__PURE__ */ _isAny("dialogFilter", "dialogFilterDefault", "dialogFilterChatlist");
const isAnyDialogFilterSuggested = /* @__PURE__ */ _isAny("dialogFilterSuggested");
const isAnyStatsDateRangeDays = /* @__PURE__ */ _isAny("statsDateRangeDays");
const isAnyStatsAbsValueAndPrev = /* @__PURE__ */ _isAny("statsAbsValueAndPrev");
const isAnyStatsPercentValue = /* @__PURE__ */ _isAny("statsPercentValue");
const isAnyStatsGraph = /* @__PURE__ */ _isAny("statsGraphAsync", "statsGraphError", "statsGraph");
const isAnyVideoSize = /* @__PURE__ */ _isAny("videoSize", "videoSizeEmojiMarkup", "videoSizeStickerMarkup");
const isAnyStatsGroupTopPoster = /* @__PURE__ */ _isAny("statsGroupTopPoster");
const isAnyStatsGroupTopAdmin = /* @__PURE__ */ _isAny("statsGroupTopAdmin");
const isAnyStatsGroupTopInviter = /* @__PURE__ */ _isAny("statsGroupTopInviter");
const isAnyGlobalPrivacySettings = /* @__PURE__ */ _isAny("globalPrivacySettings");
const isAnyMessageViews = /* @__PURE__ */ _isAny("messageViews");
const isAnyMessageReplyHeader = /* @__PURE__ */ _isAny("messageReplyHeader", "messageReplyStoryHeader");
const isAnyMessageReplies = /* @__PURE__ */ _isAny("messageReplies");
const isAnyPeerBlocked = /* @__PURE__ */ _isAny("peerBlocked");
const isAnyGroupCall = /* @__PURE__ */ _isAny("groupCallDiscarded", "groupCall");
const isAnyInputGroupCall = /* @__PURE__ */ _isAny("inputGroupCall", "inputGroupCallSlug", "inputGroupCallInviteMessage");
const isAnyGroupCallParticipant = /* @__PURE__ */ _isAny("groupCallParticipant");
const isAnyInlineQueryPeerType = /* @__PURE__ */ _isAny("inlineQueryPeerTypeSameBotPM", "inlineQueryPeerTypePM", "inlineQueryPeerTypeChat", "inlineQueryPeerTypeMegagroup", "inlineQueryPeerTypeBroadcast", "inlineQueryPeerTypeBotPM");
const isAnyChatInviteImporter = /* @__PURE__ */ _isAny("chatInviteImporter");
const isAnyChatAdminWithInvites = /* @__PURE__ */ _isAny("chatAdminWithInvites");
const isAnyGroupCallParticipantVideoSourceGroup = /* @__PURE__ */ _isAny("groupCallParticipantVideoSourceGroup");
const isAnyGroupCallParticipantVideo = /* @__PURE__ */ _isAny("groupCallParticipantVideo");
const isAnyBotCommandScope = /* @__PURE__ */ _isAny("botCommandScopeDefault", "botCommandScopeUsers", "botCommandScopeChats", "botCommandScopeChatAdmins", "botCommandScopePeer", "botCommandScopePeerAdmins", "botCommandScopePeerUser");
const isAnyChatTheme = /* @__PURE__ */ _isAny("chatTheme", "chatThemeUniqueGift");
const isAnySponsoredMessage = /* @__PURE__ */ _isAny("sponsoredMessage");
const isAnySearchResultsCalendarPeriod = /* @__PURE__ */ _isAny("searchResultsCalendarPeriod");
const isAnySearchResultsPosition = /* @__PURE__ */ _isAny("searchResultPosition");
const isAnyReactionCount = /* @__PURE__ */ _isAny("reactionCount");
const isAnyMessageReactions = /* @__PURE__ */ _isAny("messageReactions");
const isAnyAvailableReaction = /* @__PURE__ */ _isAny("availableReaction");
const isAnyMessagePeerReaction = /* @__PURE__ */ _isAny("messagePeerReaction");
const isAnyGroupCallStreamChannel = /* @__PURE__ */ _isAny("groupCallStreamChannel");
const isAnyAttachMenuBotIconColor = /* @__PURE__ */ _isAny("attachMenuBotIconColor");
const isAnyAttachMenuBotIcon = /* @__PURE__ */ _isAny("attachMenuBotIcon");
const isAnyAttachMenuBot = /* @__PURE__ */ _isAny("attachMenuBot");
const isAnyAttachMenuBots = /* @__PURE__ */ _isAny("attachMenuBotsNotModified", "attachMenuBots");
const isAnyAttachMenuBotsBot = /* @__PURE__ */ _isAny("attachMenuBotsBot");
const isAnyWebViewResult = /* @__PURE__ */ _isAny("webViewResultUrl");
const isAnyWebViewMessageSent = /* @__PURE__ */ _isAny("webViewMessageSent");
const isAnyBotMenuButton = /* @__PURE__ */ _isAny("botMenuButtonDefault", "botMenuButtonCommands", "botMenuButton");
const isAnyNotificationSound = /* @__PURE__ */ _isAny("notificationSoundDefault", "notificationSoundNone", "notificationSoundLocal", "notificationSoundRingtone");
const isAnyAttachMenuPeerType = /* @__PURE__ */ _isAny("attachMenuPeerTypeSameBotPM", "attachMenuPeerTypeBotPM", "attachMenuPeerTypePM", "attachMenuPeerTypeChat", "attachMenuPeerTypeBroadcast");
const isAnyInputInvoice = /* @__PURE__ */ _isAny("inputInvoiceMessage", "inputInvoiceSlug", "inputInvoicePremiumGiftCode", "inputInvoiceStars", "inputInvoiceChatInviteSubscription", "inputInvoiceStarGift", "inputInvoiceStarGiftUpgrade", "inputInvoiceStarGiftTransfer", "inputInvoicePremiumGiftStars", "inputInvoiceBusinessBotTransferStars", "inputInvoiceStarGiftResale", "inputInvoiceStarGiftPrepaidUpgrade", "inputInvoicePremiumAuthCode", "inputInvoiceStarGiftDropOriginalDetails", "inputInvoiceStarGiftAuctionBid");
const isAnyInputStorePaymentPurpose = /* @__PURE__ */ _isAny("inputStorePaymentPremiumSubscription", "inputStorePaymentGiftPremium", "inputStorePaymentPremiumGiftCode", "inputStorePaymentPremiumGiveaway", "inputStorePaymentStarsTopup", "inputStorePaymentStarsGift", "inputStorePaymentStarsGiveaway", "inputStorePaymentAuthCode");
const isAnyPaymentFormMethod = /* @__PURE__ */ _isAny("paymentFormMethod");
const isAnyEmojiStatus = /* @__PURE__ */ _isAny("emojiStatusEmpty", "emojiStatus", "emojiStatusCollectible", "inputEmojiStatusCollectible");
const isAnyReaction = /* @__PURE__ */ _isAny("reactionEmpty", "reactionEmoji", "reactionCustomEmoji", "reactionPaid");
const isAnyChatReactions = /* @__PURE__ */ _isAny("chatReactionsNone", "chatReactionsAll", "chatReactionsSome");
const isAnyEmailVerifyPurpose = /* @__PURE__ */ _isAny("emailVerifyPurposeLoginSetup", "emailVerifyPurposeLoginChange", "emailVerifyPurposePassport");
const isAnyEmailVerification = /* @__PURE__ */ _isAny("emailVerificationCode", "emailVerificationGoogle", "emailVerificationApple");
const isAnyPremiumSubscriptionOption = /* @__PURE__ */ _isAny("premiumSubscriptionOption");
const isAnySendAsPeer = /* @__PURE__ */ _isAny("sendAsPeer");
const isAnyMessageExtendedMedia = /* @__PURE__ */ _isAny("messageExtendedMediaPreview", "messageExtendedMedia");
const isAnyStickerKeyword = /* @__PURE__ */ _isAny("stickerKeyword");
const isAnyUsername = /* @__PURE__ */ _isAny("username");
const isAnyForumTopic = /* @__PURE__ */ _isAny("forumTopicDeleted", "forumTopic");
const isAnyDefaultHistoryTTL = /* @__PURE__ */ _isAny("defaultHistoryTTL");
const isAnyExportedContactToken = /* @__PURE__ */ _isAny("exportedContactToken");
const isAnyRequestPeerType = /* @__PURE__ */ _isAny("requestPeerTypeUser", "requestPeerTypeChat", "requestPeerTypeBroadcast", "requestPeerTypeCreateBot");
const isAnyEmojiList = /* @__PURE__ */ _isAny("emojiListNotModified", "emojiList");
const isAnyEmojiGroup = /* @__PURE__ */ _isAny("emojiGroup", "emojiGroupGreeting", "emojiGroupPremium");
const isAnyTextWithEntities = /* @__PURE__ */ _isAny("textWithEntities");
const isAnyAutoSaveSettings = /* @__PURE__ */ _isAny("autoSaveSettings");
const isAnyAutoSaveException = /* @__PURE__ */ _isAny("autoSaveException");
const isAnyInputBotApp = /* @__PURE__ */ _isAny("inputBotAppID", "inputBotAppShortName");
const isAnyBotApp = /* @__PURE__ */ _isAny("botAppNotModified", "botApp");
const isAnyInlineBotWebView = /* @__PURE__ */ _isAny("inlineBotWebView");
const isAnyReadParticipantDate = /* @__PURE__ */ _isAny("readParticipantDate");
const isAnyInputChatlist = /* @__PURE__ */ _isAny("inputChatlistDialogFilter");
const isAnyExportedChatlistInvite = /* @__PURE__ */ _isAny("exportedChatlistInvite");
const isAnyMessagePeerVote = /* @__PURE__ */ _isAny("messagePeerVote", "messagePeerVoteInputOption", "messagePeerVoteMultiple");
const isAnyStoryViews = /* @__PURE__ */ _isAny("storyViews");
const isAnyStoryItem = /* @__PURE__ */ _isAny("storyItemDeleted", "storyItemSkipped", "storyItem");
const isAnyStoryView = /* @__PURE__ */ _isAny("storyView", "storyViewPublicForward", "storyViewPublicRepost");
const isAnyInputReplyTo = /* @__PURE__ */ _isAny("inputReplyToMessage", "inputReplyToStory", "inputReplyToMonoForum");
const isAnyExportedStoryLink = /* @__PURE__ */ _isAny("exportedStoryLink");
const isAnyStoriesStealthMode = /* @__PURE__ */ _isAny("storiesStealthMode");
const isAnyMediaAreaCoordinates = /* @__PURE__ */ _isAny("mediaAreaCoordinates");
const isAnyMediaArea = /* @__PURE__ */ _isAny("mediaAreaVenue", "inputMediaAreaVenue", "mediaAreaGeoPoint", "mediaAreaSuggestedReaction", "mediaAreaChannelPost", "inputMediaAreaChannelPost", "mediaAreaUrl", "mediaAreaWeather", "mediaAreaStarGift");
const isAnyPeerStories = /* @__PURE__ */ _isAny("peerStories");
const isAnyPremiumGiftCodeOption = /* @__PURE__ */ _isAny("premiumGiftCodeOption");
const isAnyPrepaidGiveaway = /* @__PURE__ */ _isAny("prepaidGiveaway", "prepaidStarsGiveaway");
const isAnyBoost = /* @__PURE__ */ _isAny("boost");
const isAnyMyBoost = /* @__PURE__ */ _isAny("myBoost");
const isAnyStoryFwdHeader = /* @__PURE__ */ _isAny("storyFwdHeader");
const isAnyPostInteractionCounters = /* @__PURE__ */ _isAny("postInteractionCountersMessage", "postInteractionCountersStory");
const isAnyPublicForward = /* @__PURE__ */ _isAny("publicForwardMessage", "publicForwardStory");
const isAnyPeerColor = /* @__PURE__ */ _isAny("peerColor", "peerColorCollectible", "inputPeerColorCollectible");
const isAnyStoryReaction = /* @__PURE__ */ _isAny("storyReaction", "storyReactionPublicForward", "storyReactionPublicRepost");
const isAnySavedDialog = /* @__PURE__ */ _isAny("savedDialog", "monoForumDialog");
const isAnySavedReactionTag = /* @__PURE__ */ _isAny("savedReactionTag");
const isAnyOutboxReadDate = /* @__PURE__ */ _isAny("outboxReadDate");
const isAnySmsJob = /* @__PURE__ */ _isAny("smsJob");
const isAnyBusinessWeeklyOpen = /* @__PURE__ */ _isAny("businessWeeklyOpen");
const isAnyBusinessWorkHours = /* @__PURE__ */ _isAny("businessWorkHours");
const isAnyBusinessLocation = /* @__PURE__ */ _isAny("businessLocation");
const isAnyInputBusinessRecipients = /* @__PURE__ */ _isAny("inputBusinessRecipients");
const isAnyBusinessRecipients = /* @__PURE__ */ _isAny("businessRecipients");
const isAnyBusinessAwayMessageSchedule = /* @__PURE__ */ _isAny("businessAwayMessageScheduleAlways", "businessAwayMessageScheduleOutsideWorkHours", "businessAwayMessageScheduleCustom");
const isAnyInputBusinessGreetingMessage = /* @__PURE__ */ _isAny("inputBusinessGreetingMessage");
const isAnyBusinessGreetingMessage = /* @__PURE__ */ _isAny("businessGreetingMessage");
const isAnyInputBusinessAwayMessage = /* @__PURE__ */ _isAny("inputBusinessAwayMessage");
const isAnyBusinessAwayMessage = /* @__PURE__ */ _isAny("businessAwayMessage");
const isAnyTimezone = /* @__PURE__ */ _isAny("timezone");
const isAnyQuickReply = /* @__PURE__ */ _isAny("quickReply");
const isAnyInputQuickReplyShortcut = /* @__PURE__ */ _isAny("inputQuickReplyShortcut", "inputQuickReplyShortcutId");
const isAnyConnectedBot = /* @__PURE__ */ _isAny("connectedBot");
const isAnyBirthday = /* @__PURE__ */ _isAny("birthday");
const isAnyBotBusinessConnection = /* @__PURE__ */ _isAny("botBusinessConnection");
const isAnyInputBusinessIntro = /* @__PURE__ */ _isAny("inputBusinessIntro");
const isAnyBusinessIntro = /* @__PURE__ */ _isAny("businessIntro");
const isAnyInputCollectible = /* @__PURE__ */ _isAny("inputCollectibleUsername", "inputCollectiblePhone");
const isAnyInputBusinessBotRecipients = /* @__PURE__ */ _isAny("inputBusinessBotRecipients");
const isAnyBusinessBotRecipients = /* @__PURE__ */ _isAny("businessBotRecipients");
const isAnyContactBirthday = /* @__PURE__ */ _isAny("contactBirthday");
const isAnyMissingInvitee = /* @__PURE__ */ _isAny("missingInvitee");
const isAnyInputBusinessChatLink = /* @__PURE__ */ _isAny("inputBusinessChatLink");
const isAnyBusinessChatLink = /* @__PURE__ */ _isAny("businessChatLink");
const isAnyRequestedPeer = /* @__PURE__ */ _isAny("requestedPeerUser", "requestedPeerChat", "requestedPeerChannel");
const isAnySponsoredMessageReportOption = /* @__PURE__ */ _isAny("sponsoredMessageReportOption");
const isAnyReactionNotificationsFrom = /* @__PURE__ */ _isAny("reactionNotificationsFromContacts", "reactionNotificationsFromAll");
const isAnyReactionsNotifySettings = /* @__PURE__ */ _isAny("reactionsNotifySettings");
const isAnyAvailableEffect = /* @__PURE__ */ _isAny("availableEffect");
const isAnyFactCheck = /* @__PURE__ */ _isAny("factCheck");
const isAnyStarsTransactionPeer = /* @__PURE__ */ _isAny("starsTransactionPeerUnsupported", "starsTransactionPeerAppStore", "starsTransactionPeerPlayMarket", "starsTransactionPeerPremiumBot", "starsTransactionPeerFragment", "starsTransactionPeer", "starsTransactionPeerAds", "starsTransactionPeerAPI");
const isAnyStarsTopupOption = /* @__PURE__ */ _isAny("starsTopupOption");
const isAnyStarsTransaction = /* @__PURE__ */ _isAny("starsTransaction");
const isAnyFoundStory = /* @__PURE__ */ _isAny("foundStory");
const isAnyGeoPointAddress = /* @__PURE__ */ _isAny("geoPointAddress");
const isAnyStarsRevenueStatus = /* @__PURE__ */ _isAny("starsRevenueStatus");
const isAnyInputStarsTransaction = /* @__PURE__ */ _isAny("inputStarsTransaction");
const isAnyStarsGiftOption = /* @__PURE__ */ _isAny("starsGiftOption");
const isAnyBotPreviewMedia = /* @__PURE__ */ _isAny("botPreviewMedia");
const isAnyStarsSubscriptionPricing = /* @__PURE__ */ _isAny("starsSubscriptionPricing");
const isAnyStarsSubscription = /* @__PURE__ */ _isAny("starsSubscription");
const isAnyMessageReactor = /* @__PURE__ */ _isAny("messageReactor");
const isAnyStarsGiveawayOption = /* @__PURE__ */ _isAny("starsGiveawayOption");
const isAnyStarsGiveawayWinnersOption = /* @__PURE__ */ _isAny("starsGiveawayWinnersOption");
const isAnyStarGift = /* @__PURE__ */ _isAny("starGift", "starGiftUnique");
const isAnyMessageReportOption = /* @__PURE__ */ _isAny("messageReportOption");
const isAnyReportResult = /* @__PURE__ */ _isAny("reportResultChooseOption", "reportResultAddComment", "reportResultReported");
const isAnyBotAppSettings = /* @__PURE__ */ _isAny("botAppSettings");
const isAnyStarRefProgram = /* @__PURE__ */ _isAny("starRefProgram");
const isAnyConnectedBotStarRef = /* @__PURE__ */ _isAny("connectedBotStarRef");
const isAnyStarsAmount = /* @__PURE__ */ _isAny("starsAmount", "starsTonAmount");
const isAnyBotVerifierSettings = /* @__PURE__ */ _isAny("botVerifierSettings");
const isAnyBotVerification = /* @__PURE__ */ _isAny("botVerification");
const isAnyStarGiftAttribute = /* @__PURE__ */ _isAny("starGiftAttributeModel", "starGiftAttributePattern", "starGiftAttributeBackdrop", "starGiftAttributeOriginalDetails");
const isAnySavedStarGift = /* @__PURE__ */ _isAny("savedStarGift");
const isAnyInputSavedStarGift = /* @__PURE__ */ _isAny("inputSavedStarGiftUser", "inputSavedStarGiftChat", "inputSavedStarGiftSlug");
const isAnyPaidReactionPrivacy = /* @__PURE__ */ _isAny("paidReactionPrivacyDefault", "paidReactionPrivacyAnonymous", "paidReactionPrivacyPeer");
const isAnyRequirementToContact = /* @__PURE__ */ _isAny("requirementToContactEmpty", "requirementToContactPremium", "requirementToContactPaidMessages");
const isAnyBusinessBotRights = /* @__PURE__ */ _isAny("businessBotRights");
const isAnyDisallowedGiftsSettings = /* @__PURE__ */ _isAny("disallowedGiftsSettings");
const isAnySponsoredPeer = /* @__PURE__ */ _isAny("sponsoredPeer");
const isAnyStarGiftAttributeId = /* @__PURE__ */ _isAny("starGiftAttributeIdModel", "starGiftAttributeIdPattern", "starGiftAttributeIdBackdrop");
const isAnyStarGiftAttributeCounter = /* @__PURE__ */ _isAny("starGiftAttributeCounter");
const isAnyPendingSuggestion = /* @__PURE__ */ _isAny("pendingSuggestion");
const isAnyTodoItem = /* @__PURE__ */ _isAny("todoItem");
const isAnyTodoList = /* @__PURE__ */ _isAny("todoList");
const isAnyTodoCompletion = /* @__PURE__ */ _isAny("todoCompletion");
const isAnySuggestedPost = /* @__PURE__ */ _isAny("suggestedPost");
const isAnyStarsRating = /* @__PURE__ */ _isAny("starsRating");
const isAnyStarGiftCollection = /* @__PURE__ */ _isAny("starGiftCollection");
const isAnyStoryAlbum = /* @__PURE__ */ _isAny("storyAlbum");
const isAnySearchPostsFlood = /* @__PURE__ */ _isAny("searchPostsFlood");
const isAnyProfileTab = /* @__PURE__ */ _isAny("profileTabPosts", "profileTabGifts", "profileTabMedia", "profileTabFiles", "profileTabMusic", "profileTabVoice", "profileTabLinks", "profileTabGifs");
const isAnyInputChatTheme = /* @__PURE__ */ _isAny("inputChatThemeEmpty", "inputChatTheme", "inputChatThemeUniqueGift");
const isAnyStarGiftUpgradePrice = /* @__PURE__ */ _isAny("starGiftUpgradePrice");
const isAnyGroupCallMessage = /* @__PURE__ */ _isAny("groupCallMessage");
const isAnyGroupCallDonor = /* @__PURE__ */ _isAny("groupCallDonor");
const isAnyRecentStory = /* @__PURE__ */ _isAny("recentStory");
const isAnyAuctionBidLevel = /* @__PURE__ */ _isAny("auctionBidLevel");
const isAnyStarGiftAuctionState = /* @__PURE__ */ _isAny("starGiftAuctionStateNotModified", "starGiftAuctionState", "starGiftAuctionStateFinished");
const isAnyStarGiftAuctionUserState = /* @__PURE__ */ _isAny("starGiftAuctionUserState");
const isAnyStarGiftAuctionAcquiredGift = /* @__PURE__ */ _isAny("starGiftAuctionAcquiredGift");
const isAnyStarGiftActiveAuctionState = /* @__PURE__ */ _isAny("starGiftActiveAuctionState");
const isAnyInputStarGiftAuction = /* @__PURE__ */ _isAny("inputStarGiftAuction", "inputStarGiftAuctionSlug");
const isAnyPasskey = /* @__PURE__ */ _isAny("passkey");
const isAnyInputPasskeyResponse = /* @__PURE__ */ _isAny("inputPasskeyResponseRegister", "inputPasskeyResponseLogin");
const isAnyInputPasskeyCredential = /* @__PURE__ */ _isAny("inputPasskeyCredentialPublicKey", "inputPasskeyCredentialFirebasePNV");
const isAnyStarGiftBackground = /* @__PURE__ */ _isAny("starGiftBackground");
const isAnyStarGiftAuctionRound = /* @__PURE__ */ _isAny("starGiftAuctionRound", "starGiftAuctionRoundExtendable");
const isAnyStarGiftAttributeRarity = /* @__PURE__ */ _isAny("starGiftAttributeRarity", "starGiftAttributeRarityUncommon", "starGiftAttributeRarityRare", "starGiftAttributeRarityEpic", "starGiftAttributeRarityLegendary");
const isAnyKeyboardButtonStyle = /* @__PURE__ */ _isAny("keyboardButtonStyle");
const isAnyInputMessageReadMetric = /* @__PURE__ */ _isAny("inputMessageReadMetric");
const isAnyInputAiComposeTone = /* @__PURE__ */ _isAny("inputAiComposeToneDefault", "inputAiComposeToneID", "inputAiComposeToneSlug");
const isAnyAiComposeTone = /* @__PURE__ */ _isAny("aiComposeTone", "aiComposeToneDefault");
const isAnyAiComposeToneExample = /* @__PURE__ */ _isAny("aiComposeToneExample");
const isAnyJoinChatBotResult = /* @__PURE__ */ _isAny("joinChatBotResultApproved", "joinChatBotResultDeclined", "joinChatBotResultQueued", "joinChatBotResultWebView");
const isAnyWebDomainException = /* @__PURE__ */ _isAny("webDomainException");
const isAnyInputRichFile = /* @__PURE__ */ _isAny("inputRichFilePhoto", "inputRichFileDocument");
const isAnyInputRichMessage = /* @__PURE__ */ _isAny("inputRichMessage", "inputRichMessageHTML", "inputRichMessageMarkdown");
const isAnyRichMessage = /* @__PURE__ */ _isAny("richMessage");
const isAnyStorage_FileType = /* @__PURE__ */ _isAny("storage.fileUnknown", "storage.filePartial", "storage.fileJpeg", "storage.fileGif", "storage.filePng", "storage.filePdf", "storage.fileMp3", "storage.fileMov", "storage.fileMp4", "storage.fileWebp");
const isAnyAuth_SentCode = /* @__PURE__ */ _isAny("auth.sentCode", "auth.sentCodeSuccess", "auth.sentCodePaymentRequired");
const isAnyAuth_Authorization = /* @__PURE__ */ _isAny("auth.authorization", "auth.authorizationSignUpRequired");
const isAnyAuth_ExportedAuthorization = /* @__PURE__ */ _isAny("auth.exportedAuthorization");
const isAnyAuth_PasswordRecovery = /* @__PURE__ */ _isAny("auth.passwordRecovery");
const isAnyAuth_CodeType = /* @__PURE__ */ _isAny("auth.codeTypeSms", "auth.codeTypeCall", "auth.codeTypeFlashCall", "auth.codeTypeMissedCall", "auth.codeTypeFragmentSms");
const isAnyAuth_SentCodeType = /* @__PURE__ */ _isAny("auth.sentCodeTypeApp", "auth.sentCodeTypeSms", "auth.sentCodeTypeCall", "auth.sentCodeTypeFlashCall", "auth.sentCodeTypeMissedCall", "auth.sentCodeTypeEmailCode", "auth.sentCodeTypeSetUpEmailRequired", "auth.sentCodeTypeFragmentSms", "auth.sentCodeTypeFirebaseSms", "auth.sentCodeTypeSmsWord", "auth.sentCodeTypeSmsPhrase");
const isAnyAuth_LoginToken = /* @__PURE__ */ _isAny("auth.loginToken", "auth.loginTokenMigrateTo", "auth.loginTokenSuccess");
const isAnyAuth_LoggedOut = /* @__PURE__ */ _isAny("auth.loggedOut");
const isAnyAuth_PasskeyLoginOptions = /* @__PURE__ */ _isAny("auth.passkeyLoginOptions");
const isAnyContacts_Contacts = /* @__PURE__ */ _isAny("contacts.contactsNotModified", "contacts.contacts");
const isAnyContacts_ImportedContacts = /* @__PURE__ */ _isAny("contacts.importedContacts");
const isAnyContacts_Blocked = /* @__PURE__ */ _isAny("contacts.blocked", "contacts.blockedSlice");
const isAnyContacts_Found = /* @__PURE__ */ _isAny("contacts.found");
const isAnyContacts_ResolvedPeer = /* @__PURE__ */ _isAny("contacts.resolvedPeer");
const isAnyContacts_TopPeers = /* @__PURE__ */ _isAny("contacts.topPeersNotModified", "contacts.topPeers", "contacts.topPeersDisabled");
const isAnyContacts_ContactBirthdays = /* @__PURE__ */ _isAny("contacts.contactBirthdays");
const isAnyContacts_SponsoredPeers = /* @__PURE__ */ _isAny("contacts.sponsoredPeersEmpty", "contacts.sponsoredPeers");
const isAnyMessages_Dialogs = /* @__PURE__ */ _isAny("messages.dialogs", "messages.dialogsSlice", "messages.dialogsNotModified");
const isAnyMessages_Messages = /* @__PURE__ */ _isAny("messages.messages", "messages.messagesSlice", "messages.channelMessages", "messages.messagesNotModified");
const isAnyMessages_Chats = /* @__PURE__ */ _isAny("messages.chats", "messages.chatsSlice");
const isAnyMessages_ChatFull = /* @__PURE__ */ _isAny("messages.chatFull");
const isAnyMessages_AffectedHistory = /* @__PURE__ */ _isAny("messages.affectedHistory");
const isAnyMessages_DhConfig = /* @__PURE__ */ _isAny("messages.dhConfigNotModified", "messages.dhConfig");
const isAnyMessages_SentEncryptedMessage = /* @__PURE__ */ _isAny("messages.sentEncryptedMessage", "messages.sentEncryptedFile");
const isAnyMessages_Stickers = /* @__PURE__ */ _isAny("messages.stickersNotModified", "messages.stickers");
const isAnyMessages_AllStickers = /* @__PURE__ */ _isAny("messages.allStickersNotModified", "messages.allStickers");
const isAnyMessages_AffectedMessages = /* @__PURE__ */ _isAny("messages.affectedMessages");
const isAnyMessages_StickerSet = /* @__PURE__ */ _isAny("messages.stickerSet", "messages.stickerSetNotModified");
const isAnyMessages_SavedGifs = /* @__PURE__ */ _isAny("messages.savedGifsNotModified", "messages.savedGifs");
const isAnyMessages_BotResults = /* @__PURE__ */ _isAny("messages.botResults");
const isAnyMessages_BotCallbackAnswer = /* @__PURE__ */ _isAny("messages.botCallbackAnswer");
const isAnyMessages_MessageEditData = /* @__PURE__ */ _isAny("messages.messageEditData");
const isAnyMessages_PeerDialogs = /* @__PURE__ */ _isAny("messages.peerDialogs");
const isAnyMessages_FeaturedStickers = /* @__PURE__ */ _isAny("messages.featuredStickersNotModified", "messages.featuredStickers");
const isAnyMessages_RecentStickers = /* @__PURE__ */ _isAny("messages.recentStickersNotModified", "messages.recentStickers");
const isAnyMessages_ArchivedStickers = /* @__PURE__ */ _isAny("messages.archivedStickers");
const isAnyMessages_StickerSetInstallResult = /* @__PURE__ */ _isAny("messages.stickerSetInstallResultSuccess", "messages.stickerSetInstallResultArchive");
const isAnyMessages_HighScores = /* @__PURE__ */ _isAny("messages.highScores");
const isAnyMessages_FavedStickers = /* @__PURE__ */ _isAny("messages.favedStickersNotModified", "messages.favedStickers");
const isAnyMessages_FoundStickerSets = /* @__PURE__ */ _isAny("messages.foundStickerSetsNotModified", "messages.foundStickerSets");
const isAnyMessages_SearchCounter = /* @__PURE__ */ _isAny("messages.searchCounter");
const isAnyMessages_InactiveChats = /* @__PURE__ */ _isAny("messages.inactiveChats");
const isAnyMessages_VotesList = /* @__PURE__ */ _isAny("messages.votesList");
const isAnyMessages_MessageViews = /* @__PURE__ */ _isAny("messages.messageViews");
const isAnyMessages_DiscussionMessage = /* @__PURE__ */ _isAny("messages.discussionMessage");
const isAnyMessages_HistoryImport = /* @__PURE__ */ _isAny("messages.historyImport");
const isAnyMessages_HistoryImportParsed = /* @__PURE__ */ _isAny("messages.historyImportParsed");
const isAnyMessages_AffectedFoundMessages = /* @__PURE__ */ _isAny("messages.affectedFoundMessages");
const isAnyMessages_ExportedChatInvites = /* @__PURE__ */ _isAny("messages.exportedChatInvites");
const isAnyMessages_ExportedChatInvite = /* @__PURE__ */ _isAny("messages.exportedChatInvite", "messages.exportedChatInviteReplaced");
const isAnyMessages_ChatInviteImporters = /* @__PURE__ */ _isAny("messages.chatInviteImporters");
const isAnyMessages_ChatAdminsWithInvites = /* @__PURE__ */ _isAny("messages.chatAdminsWithInvites");
const isAnyMessages_CheckedHistoryImportPeer = /* @__PURE__ */ _isAny("messages.checkedHistoryImportPeer");
const isAnyMessages_SponsoredMessages = /* @__PURE__ */ _isAny("messages.sponsoredMessages", "messages.sponsoredMessagesEmpty");
const isAnyMessages_SearchResultsCalendar = /* @__PURE__ */ _isAny("messages.searchResultsCalendar");
const isAnyMessages_SearchResultsPositions = /* @__PURE__ */ _isAny("messages.searchResultsPositions");
const isAnyMessages_PeerSettings = /* @__PURE__ */ _isAny("messages.peerSettings");
const isAnyMessages_MessageReactionsList = /* @__PURE__ */ _isAny("messages.messageReactionsList");
const isAnyMessages_AvailableReactions = /* @__PURE__ */ _isAny("messages.availableReactionsNotModified", "messages.availableReactions");
const isAnyMessages_TranscribedAudio = /* @__PURE__ */ _isAny("messages.transcribedAudio");
const isAnyMessages_Reactions = /* @__PURE__ */ _isAny("messages.reactionsNotModified", "messages.reactions");
const isAnyMessages_ForumTopics = /* @__PURE__ */ _isAny("messages.forumTopics");
const isAnyMessages_EmojiGroups = /* @__PURE__ */ _isAny("messages.emojiGroupsNotModified", "messages.emojiGroups");
const isAnyMessages_TranslatedText = /* @__PURE__ */ _isAny("messages.translateResult");
const isAnyMessages_BotApp = /* @__PURE__ */ _isAny("messages.botApp");
const isAnyMessages_WebPage = /* @__PURE__ */ _isAny("messages.webPage");
const isAnyMessages_SavedDialogs = /* @__PURE__ */ _isAny("messages.savedDialogs", "messages.savedDialogsSlice", "messages.savedDialogsNotModified");
const isAnyMessages_SavedReactionTags = /* @__PURE__ */ _isAny("messages.savedReactionTagsNotModified", "messages.savedReactionTags");
const isAnyMessages_QuickReplies = /* @__PURE__ */ _isAny("messages.quickReplies", "messages.quickRepliesNotModified");
const isAnyMessages_DialogFilters = /* @__PURE__ */ _isAny("messages.dialogFilters");
const isAnyMessages_MyStickers = /* @__PURE__ */ _isAny("messages.myStickers");
const isAnyMessages_InvitedUsers = /* @__PURE__ */ _isAny("messages.invitedUsers");
const isAnyMessages_AvailableEffects = /* @__PURE__ */ _isAny("messages.availableEffectsNotModified", "messages.availableEffects");
const isAnyMessages_BotPreparedInlineMessage = /* @__PURE__ */ _isAny("messages.botPreparedInlineMessage");
const isAnyMessages_PreparedInlineMessage = /* @__PURE__ */ _isAny("messages.preparedInlineMessage");
const isAnyMessages_FoundStickers = /* @__PURE__ */ _isAny("messages.foundStickersNotModified", "messages.foundStickers");
const isAnyMessages_WebPagePreview = /* @__PURE__ */ _isAny("messages.webPagePreview");
const isAnyMessages_EmojiGameOutcome = /* @__PURE__ */ _isAny("messages.emojiGameOutcome");
const isAnyMessages_EmojiGameInfo = /* @__PURE__ */ _isAny("messages.emojiGameUnavailable", "messages.emojiGameDiceInfo");
const isAnyMessages_ComposedMessageWithAI = /* @__PURE__ */ _isAny("messages.composedMessageWithAI");
const isAnyMessages_ChatInviteJoinResult = /* @__PURE__ */ _isAny("messages.chatInviteJoinResultOk", "messages.chatInviteJoinResultWebView");
const isAnyUpdates_State = /* @__PURE__ */ _isAny("updates.state");
const isAnyUpdates_Difference = /* @__PURE__ */ _isAny("updates.differenceEmpty", "updates.difference", "updates.differenceSlice", "updates.differenceTooLong");
const isAnyUpdates_ChannelDifference = /* @__PURE__ */ _isAny("updates.channelDifferenceEmpty", "updates.channelDifferenceTooLong", "updates.channelDifference");
const isAnyPhotos_Photos = /* @__PURE__ */ _isAny("photos.photos", "photos.photosSlice");
const isAnyPhotos_Photo = /* @__PURE__ */ _isAny("photos.photo");
const isAnyUpload_File = /* @__PURE__ */ _isAny("upload.file", "upload.fileCdnRedirect");
const isAnyUpload_WebFile = /* @__PURE__ */ _isAny("upload.webFile");
const isAnyUpload_CdnFile = /* @__PURE__ */ _isAny("upload.cdnFileReuploadNeeded", "upload.cdnFile");
const isAnyHelp_AppUpdate = /* @__PURE__ */ _isAny("help.appUpdate", "help.noAppUpdate");
const isAnyHelp_InviteText = /* @__PURE__ */ _isAny("help.inviteText");
const isAnyHelp_Support = /* @__PURE__ */ _isAny("help.support");
const isAnyHelp_TermsOfService = /* @__PURE__ */ _isAny("help.termsOfService");
const isAnyHelp_RecentMeUrls = /* @__PURE__ */ _isAny("help.recentMeUrls");
const isAnyHelp_TermsOfServiceUpdate = /* @__PURE__ */ _isAny("help.termsOfServiceUpdateEmpty", "help.termsOfServiceUpdate");
const isAnyHelp_DeepLinkInfo = /* @__PURE__ */ _isAny("help.deepLinkInfoEmpty", "help.deepLinkInfo");
const isAnyHelp_PassportConfig = /* @__PURE__ */ _isAny("help.passportConfigNotModified", "help.passportConfig");
const isAnyHelp_SupportName = /* @__PURE__ */ _isAny("help.supportName");
const isAnyHelp_UserInfo = /* @__PURE__ */ _isAny("help.userInfoEmpty", "help.userInfo");
const isAnyHelp_PromoData = /* @__PURE__ */ _isAny("help.promoDataEmpty", "help.promoData");
const isAnyHelp_CountryCode = /* @__PURE__ */ _isAny("help.countryCode");
const isAnyHelp_Country = /* @__PURE__ */ _isAny("help.country");
const isAnyHelp_CountriesList = /* @__PURE__ */ _isAny("help.countriesListNotModified", "help.countriesList");
const isAnyHelp_PremiumPromo = /* @__PURE__ */ _isAny("help.premiumPromo");
const isAnyHelp_AppConfig = /* @__PURE__ */ _isAny("help.appConfigNotModified", "help.appConfig");
const isAnyHelp_PeerColorSet = /* @__PURE__ */ _isAny("help.peerColorSet", "help.peerColorProfileSet");
const isAnyHelp_PeerColorOption = /* @__PURE__ */ _isAny("help.peerColorOption");
const isAnyHelp_PeerColors = /* @__PURE__ */ _isAny("help.peerColorsNotModified", "help.peerColors");
const isAnyHelp_TimezonesList = /* @__PURE__ */ _isAny("help.timezonesListNotModified", "help.timezonesList");
const isAnyAccount_PrivacyRules = /* @__PURE__ */ _isAny("account.privacyRules");
const isAnyAccount_Authorizations = /* @__PURE__ */ _isAny("account.authorizations");
const isAnyAccount_Password = /* @__PURE__ */ _isAny("account.password");
const isAnyAccount_PasswordSettings = /* @__PURE__ */ _isAny("account.passwordSettings");
const isAnyAccount_PasswordInputSettings = /* @__PURE__ */ _isAny("account.passwordInputSettings");
const isAnyAccount_TmpPassword = /* @__PURE__ */ _isAny("account.tmpPassword");
const isAnyAccount_WebAuthorizations = /* @__PURE__ */ _isAny("account.webAuthorizations");
const isAnyAccount_AuthorizationForm = /* @__PURE__ */ _isAny("account.authorizationForm");
const isAnyAccount_SentEmailCode = /* @__PURE__ */ _isAny("account.sentEmailCode");
const isAnyAccount_Takeout = /* @__PURE__ */ _isAny("account.takeout");
const isAnyAccount_WallPapers = /* @__PURE__ */ _isAny("account.wallPapersNotModified", "account.wallPapers");
const isAnyAccount_AutoDownloadSettings = /* @__PURE__ */ _isAny("account.autoDownloadSettings");
const isAnyAccount_Themes = /* @__PURE__ */ _isAny("account.themesNotModified", "account.themes");
const isAnyAccount_ContentSettings = /* @__PURE__ */ _isAny("account.contentSettings");
const isAnyAccount_ResetPasswordResult = /* @__PURE__ */ _isAny("account.resetPasswordFailedWait", "account.resetPasswordRequestedWait", "account.resetPasswordOk");
const isAnyAccount_ChatThemes = /* @__PURE__ */ _isAny("account.chatThemesNotModified", "account.chatThemes");
const isAnyAccount_SavedRingtones = /* @__PURE__ */ _isAny("account.savedRingtonesNotModified", "account.savedRingtones");
const isAnyAccount_SavedRingtone = /* @__PURE__ */ _isAny("account.savedRingtone", "account.savedRingtoneConverted");
const isAnyAccount_EmojiStatuses = /* @__PURE__ */ _isAny("account.emojiStatusesNotModified", "account.emojiStatuses");
const isAnyAccount_EmailVerified = /* @__PURE__ */ _isAny("account.emailVerified", "account.emailVerifiedLogin");
const isAnyAccount_AutoSaveSettings = /* @__PURE__ */ _isAny("account.autoSaveSettings");
const isAnyAccount_ConnectedBots = /* @__PURE__ */ _isAny("account.connectedBots");
const isAnyAccount_BusinessChatLinks = /* @__PURE__ */ _isAny("account.businessChatLinks");
const isAnyAccount_ResolvedBusinessChatLinks = /* @__PURE__ */ _isAny("account.resolvedBusinessChatLinks");
const isAnyAccount_PaidMessagesRevenue = /* @__PURE__ */ _isAny("account.paidMessagesRevenue");
const isAnyAccount_SavedMusicIds = /* @__PURE__ */ _isAny("account.savedMusicIdsNotModified", "account.savedMusicIds");
const isAnyAccount_Passkeys = /* @__PURE__ */ _isAny("account.passkeys");
const isAnyAccount_PasskeyRegistrationOptions = /* @__PURE__ */ _isAny("account.passkeyRegistrationOptions");
const isAnyAccount_WebBrowserSettings = /* @__PURE__ */ _isAny("account.webBrowserSettingsNotModified", "account.webBrowserSettings");
const isAnyChannels_ChannelParticipants = /* @__PURE__ */ _isAny("channels.channelParticipants", "channels.channelParticipantsNotModified");
const isAnyChannels_ChannelParticipant = /* @__PURE__ */ _isAny("channels.channelParticipant");
const isAnyChannels_AdminLogResults = /* @__PURE__ */ _isAny("channels.adminLogResults");
const isAnyChannels_SendAsPeers = /* @__PURE__ */ _isAny("channels.sendAsPeers");
const isAnyChannels_SponsoredMessageReportResult = /* @__PURE__ */ _isAny("channels.sponsoredMessageReportResultChooseOption", "channels.sponsoredMessageReportResultAdsHidden", "channels.sponsoredMessageReportResultReported");
const isAnyPayments_PaymentForm = /* @__PURE__ */ _isAny("payments.paymentForm", "payments.paymentFormStars", "payments.paymentFormStarGift");
const isAnyPayments_ValidatedRequestedInfo = /* @__PURE__ */ _isAny("payments.validatedRequestedInfo");
const isAnyPayments_PaymentResult = /* @__PURE__ */ _isAny("payments.paymentResult", "payments.paymentVerificationNeeded");
const isAnyPayments_PaymentReceipt = /* @__PURE__ */ _isAny("payments.paymentReceipt", "payments.paymentReceiptStars");
const isAnyPayments_SavedInfo = /* @__PURE__ */ _isAny("payments.savedInfo");
const isAnyPayments_BankCardData = /* @__PURE__ */ _isAny("payments.bankCardData");
const isAnyPayments_ExportedInvoice = /* @__PURE__ */ _isAny("payments.exportedInvoice");
const isAnyPayments_CheckedGiftCode = /* @__PURE__ */ _isAny("payments.checkedGiftCode");
const isAnyPayments_GiveawayInfo = /* @__PURE__ */ _isAny("payments.giveawayInfo", "payments.giveawayInfoResults");
const isAnyPayments_StarsStatus = /* @__PURE__ */ _isAny("payments.starsStatus");
const isAnyPayments_StarsRevenueStats = /* @__PURE__ */ _isAny("payments.starsRevenueStats");
const isAnyPayments_StarsRevenueWithdrawalUrl = /* @__PURE__ */ _isAny("payments.starsRevenueWithdrawalUrl");
const isAnyPayments_StarsRevenueAdsAccountUrl = /* @__PURE__ */ _isAny("payments.starsRevenueAdsAccountUrl");
const isAnyPayments_StarGifts = /* @__PURE__ */ _isAny("payments.starGiftsNotModified", "payments.starGifts");
const isAnyPayments_ConnectedStarRefBots = /* @__PURE__ */ _isAny("payments.connectedStarRefBots");
const isAnyPayments_SuggestedStarRefBots = /* @__PURE__ */ _isAny("payments.suggestedStarRefBots");
const isAnyPayments_StarGiftUpgradePreview = /* @__PURE__ */ _isAny("payments.starGiftUpgradePreview");
const isAnyPayments_UniqueStarGift = /* @__PURE__ */ _isAny("payments.uniqueStarGift");
const isAnyPayments_SavedStarGifts = /* @__PURE__ */ _isAny("payments.savedStarGifts");
const isAnyPayments_StarGiftWithdrawalUrl = /* @__PURE__ */ _isAny("payments.starGiftWithdrawalUrl");
const isAnyPayments_ResaleStarGifts = /* @__PURE__ */ _isAny("payments.resaleStarGifts");
const isAnyPayments_StarGiftCollections = /* @__PURE__ */ _isAny("payments.starGiftCollectionsNotModified", "payments.starGiftCollections");
const isAnyPayments_UniqueStarGiftValueInfo = /* @__PURE__ */ _isAny("payments.uniqueStarGiftValueInfo");
const isAnyPayments_CheckCanSendGiftResult = /* @__PURE__ */ _isAny("payments.checkCanSendGiftResultOk", "payments.checkCanSendGiftResultFail");
const isAnyPayments_StarGiftAuctionState = /* @__PURE__ */ _isAny("payments.starGiftAuctionState");
const isAnyPayments_StarGiftAuctionAcquiredGifts = /* @__PURE__ */ _isAny("payments.starGiftAuctionAcquiredGifts");
const isAnyPayments_StarGiftActiveAuctions = /* @__PURE__ */ _isAny("payments.starGiftActiveAuctionsNotModified", "payments.starGiftActiveAuctions");
const isAnyPayments_StarGiftUpgradeAttributes = /* @__PURE__ */ _isAny("payments.starGiftUpgradeAttributes");
const isAnyPhone_PhoneCall = /* @__PURE__ */ _isAny("phone.phoneCall");
const isAnyPhone_GroupCall = /* @__PURE__ */ _isAny("phone.groupCall");
const isAnyPhone_GroupParticipants = /* @__PURE__ */ _isAny("phone.groupParticipants");
const isAnyPhone_JoinAsPeers = /* @__PURE__ */ _isAny("phone.joinAsPeers");
const isAnyPhone_ExportedGroupCallInvite = /* @__PURE__ */ _isAny("phone.exportedGroupCallInvite");
const isAnyPhone_GroupCallStreamChannels = /* @__PURE__ */ _isAny("phone.groupCallStreamChannels");
const isAnyPhone_GroupCallStreamRtmpUrl = /* @__PURE__ */ _isAny("phone.groupCallStreamRtmpUrl");
const isAnyPhone_GroupCallStars = /* @__PURE__ */ _isAny("phone.groupCallStars");
const isAnyStats_BroadcastStats = /* @__PURE__ */ _isAny("stats.broadcastStats");
const isAnyStats_MegagroupStats = /* @__PURE__ */ _isAny("stats.megagroupStats");
const isAnyStats_MessageStats = /* @__PURE__ */ _isAny("stats.messageStats");
const isAnyStats_StoryStats = /* @__PURE__ */ _isAny("stats.storyStats");
const isAnyStats_PublicForwards = /* @__PURE__ */ _isAny("stats.publicForwards");
const isAnyStats_PollStats = /* @__PURE__ */ _isAny("stats.pollStats");
const isAnyStickers_SuggestedShortName = /* @__PURE__ */ _isAny("stickers.suggestedShortName");
const isAnyUsers_UserFull = /* @__PURE__ */ _isAny("users.userFull");
const isAnyUsers_Users = /* @__PURE__ */ _isAny("users.users", "users.usersSlice");
const isAnyUsers_SavedMusic = /* @__PURE__ */ _isAny("users.savedMusicNotModified", "users.savedMusic");
const isAnyChatlists_ExportedChatlistInvite = /* @__PURE__ */ _isAny("chatlists.exportedChatlistInvite");
const isAnyChatlists_ExportedInvites = /* @__PURE__ */ _isAny("chatlists.exportedInvites");
const isAnyChatlists_ChatlistInvite = /* @__PURE__ */ _isAny("chatlists.chatlistInviteAlready", "chatlists.chatlistInvite");
const isAnyChatlists_ChatlistUpdates = /* @__PURE__ */ _isAny("chatlists.chatlistUpdates");
const isAnyBots_BotInfo = /* @__PURE__ */ _isAny("bots.botInfo");
const isAnyBots_PopularAppBots = /* @__PURE__ */ _isAny("bots.popularAppBots");
const isAnyBots_PreviewInfo = /* @__PURE__ */ _isAny("bots.previewInfo");
const isAnyBots_ExportedBotToken = /* @__PURE__ */ _isAny("bots.exportedBotToken");
const isAnyBots_RequestedButton = /* @__PURE__ */ _isAny("bots.requestedButton");
const isAnyBots_AccessSettings = /* @__PURE__ */ _isAny("bots.accessSettings");
const isAnyStories_AllStories = /* @__PURE__ */ _isAny("stories.allStoriesNotModified", "stories.allStories");
const isAnyStories_Stories = /* @__PURE__ */ _isAny("stories.stories");
const isAnyStories_StoryViewsList = /* @__PURE__ */ _isAny("stories.storyViewsList");
const isAnyStories_StoryViews = /* @__PURE__ */ _isAny("stories.storyViews");
const isAnyStories_PeerStories = /* @__PURE__ */ _isAny("stories.peerStories");
const isAnyStories_StoryReactionsList = /* @__PURE__ */ _isAny("stories.storyReactionsList");
const isAnyStories_FoundStories = /* @__PURE__ */ _isAny("stories.foundStories");
const isAnyStories_CanSendStoryCount = /* @__PURE__ */ _isAny("stories.canSendStoryCount");
const isAnyStories_Albums = /* @__PURE__ */ _isAny("stories.albumsNotModified", "stories.albums");
const isAnyPremium_BoostsList = /* @__PURE__ */ _isAny("premium.boostsList");
const isAnyPremium_MyBoosts = /* @__PURE__ */ _isAny("premium.myBoosts");
const isAnyPremium_BoostsStatus = /* @__PURE__ */ _isAny("premium.boostsStatus");
const isAnySmsjobs_EligibilityToJoin = /* @__PURE__ */ _isAny("smsjobs.eligibleToJoin");
const isAnySmsjobs_Status = /* @__PURE__ */ _isAny("smsjobs.status");
const isAnyFragment_CollectibleInfo = /* @__PURE__ */ _isAny("fragment.collectibleInfo");
const isAnyAicompose_Tones = /* @__PURE__ */ _isAny("aicompose.tonesNotModified", "aicompose.tones");
const isAnyMtcute_Update = /* @__PURE__ */ _isAny("updateNewMessage", "updateMessageID", "updateDeleteMessages", "updateUserTyping", "updateChatUserTyping", "updateChatParticipants", "updateUserStatus", "updateUserName", "updateNewAuthorization", "updateNewEncryptedMessage", "updateEncryptedChatTyping", "updateEncryption", "updateEncryptedMessagesRead", "updateChatParticipantAdd", "updateChatParticipantDelete", "updateDcOptions", "updateNotifySettings", "updateServiceNotification", "updatePrivacy", "updateUserPhone", "updateReadHistoryInbox", "updateReadHistoryOutbox", "updateWebPage", "updateReadMessagesContents", "updateChannelTooLong", "updateChannel", "updateNewChannelMessage", "updateReadChannelInbox", "updateDeleteChannelMessages", "updateChannelMessageViews", "updateChatParticipantAdmin", "updateNewStickerSet", "updateStickerSetsOrder", "updateStickerSets", "updateSavedGifs", "updateBotInlineQuery", "updateBotInlineSend", "updateEditChannelMessage", "updateBotCallbackQuery", "updateEditMessage", "updateInlineBotCallbackQuery", "updateReadChannelOutbox", "updateDraftMessage", "updateReadFeaturedStickers", "updateRecentStickers", "updateConfig", "updatePtsChanged", "updateChannelWebPage", "updateDialogPinned", "updatePinnedDialogs", "updateBotWebhookJSON", "updateBotWebhookJSONQuery", "updateBotShippingQuery", "updateBotPrecheckoutQuery", "updatePhoneCall", "updateLangPackTooLong", "updateLangPack", "updateFavedStickers", "updateChannelReadMessagesContents", "updateContactsReset", "updateChannelAvailableMessages", "updateDialogUnreadMark", "updateMessagePoll", "updateChatDefaultBannedRights", "updateFolderPeers", "updatePeerSettings", "updatePeerLocated", "updateNewScheduledMessage", "updateDeleteScheduledMessages", "updateTheme", "updateGeoLiveViewed", "updateLoginToken", "updateMessagePollVote", "updateDialogFilter", "updateDialogFilterOrder", "updateDialogFilters", "updatePhoneCallSignalingData", "updateChannelMessageForwards", "updateReadChannelDiscussionInbox", "updateReadChannelDiscussionOutbox", "updatePeerBlocked", "updateChannelUserTyping", "updatePinnedMessages", "updatePinnedChannelMessages", "updateChat", "updateGroupCallParticipants", "updateGroupCall", "updatePeerHistoryTTL", "updateChatParticipant", "updateChannelParticipant", "updateBotStopped", "updateGroupCallConnection", "updateBotCommands", "updatePendingJoinRequests", "updateBotChatInviteRequester", "updateMessageReactions", "updateAttachMenuBots", "updateWebViewResultSent", "updateBotMenuButton", "updateSavedRingtones", "updateTranscribedAudio", "updateReadFeaturedEmojiStickers", "updateUserEmojiStatus", "updateRecentEmojiStatuses", "updateRecentReactions", "updateMoveStickerSetToTop", "updateMessageExtendedMedia", "updateUser", "updateAutoSaveSettings", "updateStory", "updateReadStories", "updateStoryID", "updateStoriesStealthMode", "updateSentStoryReaction", "updateBotChatBoost", "updateChannelViewForumAsMessages", "updatePeerWallpaper", "updateBotMessageReaction", "updateBotMessageReactions", "updateSavedDialogPinned", "updatePinnedSavedDialogs", "updateSavedReactionTags", "updateSmsJob", "updateQuickReplies", "updateNewQuickReply", "updateDeleteQuickReply", "updateQuickReplyMessage", "updateDeleteQuickReplyMessages", "updateBotBusinessConnect", "updateBotNewBusinessMessage", "updateBotEditBusinessMessage", "updateBotDeleteBusinessMessage", "updateNewStoryReaction", "updateStarsBalance", "updateBusinessBotCallbackQuery", "updateStarsRevenueStatus", "updateBotPurchasedPaidMedia", "updatePaidReactionPrivacy", "updateSentPhoneCode", "updateGroupCallChainBlocks", "updateReadMonoForumInbox", "updateReadMonoForumOutbox", "updateMonoForumNoPaidException", "updateGroupCallMessage", "updateGroupCallEncryptedMessage", "updatePinnedForumTopic", "updatePinnedForumTopics", "updateDeleteGroupCallMessages", "updateStarGiftAuctionState", "updateStarGiftAuctionUserState", "updateEmojiGameInfo", "updateStarGiftCraftFail", "updateChatParticipantRank", "updateManagedBot", "updateBotGuestChatQuery", "updateAiComposeTones", "updateJoinChatWebViewDecision", "updateNewBotConnection", "updateWebBrowserSettings", "updateWebBrowserException", "updateChannelPinnedTopic", "updateChannelPinnedTopics", "mtcute.dummyUpdate");
const isAnyMtcute_InputPeer = /* @__PURE__ */ _isAny("inputPeerEmpty", "inputPeerSelf", "inputPeerChat", "inputPeerUser", "inputPeerChannel", "inputPeerUserFromMessage", "inputPeerChannelFromMessage", "mtcute.dummyInputPeerMinUser", "mtcute.dummyInputPeerMinChannel");
export {
  LAYER,
  RpcError,
  isAnyAccountDaysTTL,
  isAnyAccount_AuthorizationForm,
  isAnyAccount_Authorizations,
  isAnyAccount_AutoDownloadSettings,
  isAnyAccount_AutoSaveSettings,
  isAnyAccount_BusinessChatLinks,
  isAnyAccount_ChatThemes,
  isAnyAccount_ConnectedBots,
  isAnyAccount_ContentSettings,
  isAnyAccount_EmailVerified,
  isAnyAccount_EmojiStatuses,
  isAnyAccount_PaidMessagesRevenue,
  isAnyAccount_PasskeyRegistrationOptions,
  isAnyAccount_Passkeys,
  isAnyAccount_Password,
  isAnyAccount_PasswordInputSettings,
  isAnyAccount_PasswordSettings,
  isAnyAccount_PrivacyRules,
  isAnyAccount_ResetPasswordResult,
  isAnyAccount_ResolvedBusinessChatLinks,
  isAnyAccount_SavedMusicIds,
  isAnyAccount_SavedRingtone,
  isAnyAccount_SavedRingtones,
  isAnyAccount_SentEmailCode,
  isAnyAccount_Takeout,
  isAnyAccount_Themes,
  isAnyAccount_TmpPassword,
  isAnyAccount_WallPapers,
  isAnyAccount_WebAuthorizations,
  isAnyAccount_WebBrowserSettings,
  isAnyAiComposeTone,
  isAnyAiComposeToneExample,
  isAnyAicompose_Tones,
  isAnyAttachMenuBot,
  isAnyAttachMenuBotIcon,
  isAnyAttachMenuBotIconColor,
  isAnyAttachMenuBots,
  isAnyAttachMenuBotsBot,
  isAnyAttachMenuPeerType,
  isAnyAuctionBidLevel,
  isAnyAuth_Authorization,
  isAnyAuth_CodeType,
  isAnyAuth_ExportedAuthorization,
  isAnyAuth_LoggedOut,
  isAnyAuth_LoginToken,
  isAnyAuth_PasskeyLoginOptions,
  isAnyAuth_PasswordRecovery,
  isAnyAuth_SentCode,
  isAnyAuth_SentCodeType,
  isAnyAuthorization,
  isAnyAutoDownloadSettings,
  isAnyAutoSaveException,
  isAnyAutoSaveSettings,
  isAnyAvailableEffect,
  isAnyAvailableReaction,
  isAnyBankCardOpenUrl,
  isAnyBaseTheme,
  isAnyBirthday,
  isAnyBoost,
  isAnyBotApp,
  isAnyBotAppSettings,
  isAnyBotBusinessConnection,
  isAnyBotCommand,
  isAnyBotCommandScope,
  isAnyBotInfo,
  isAnyBotInlineMessage,
  isAnyBotInlineResult,
  isAnyBotMenuButton,
  isAnyBotPreviewMedia,
  isAnyBotVerification,
  isAnyBotVerifierSettings,
  isAnyBots_AccessSettings,
  isAnyBots_BotInfo,
  isAnyBots_ExportedBotToken,
  isAnyBots_PopularAppBots,
  isAnyBots_PreviewInfo,
  isAnyBots_RequestedButton,
  isAnyBusinessAwayMessage,
  isAnyBusinessAwayMessageSchedule,
  isAnyBusinessBotRecipients,
  isAnyBusinessBotRights,
  isAnyBusinessChatLink,
  isAnyBusinessGreetingMessage,
  isAnyBusinessIntro,
  isAnyBusinessLocation,
  isAnyBusinessRecipients,
  isAnyBusinessWeeklyOpen,
  isAnyBusinessWorkHours,
  isAnyCdnConfig,
  isAnyCdnPublicKey,
  isAnyChannelAdminLogEvent,
  isAnyChannelAdminLogEventAction,
  isAnyChannelAdminLogEventsFilter,
  isAnyChannelLocation,
  isAnyChannelMessagesFilter,
  isAnyChannelParticipant,
  isAnyChannelParticipantsFilter,
  isAnyChannels_AdminLogResults,
  isAnyChannels_ChannelParticipant,
  isAnyChannels_ChannelParticipants,
  isAnyChannels_SendAsPeers,
  isAnyChannels_SponsoredMessageReportResult,
  isAnyChat,
  isAnyChatAdminRights,
  isAnyChatAdminWithInvites,
  isAnyChatBannedRights,
  isAnyChatFull,
  isAnyChatInvite,
  isAnyChatInviteImporter,
  isAnyChatOnlines,
  isAnyChatParticipant,
  isAnyChatParticipants,
  isAnyChatPhoto,
  isAnyChatReactions,
  isAnyChatTheme,
  isAnyChatlists_ChatlistInvite,
  isAnyChatlists_ChatlistUpdates,
  isAnyChatlists_ExportedChatlistInvite,
  isAnyChatlists_ExportedInvites,
  isAnyCodeSettings,
  isAnyConfig,
  isAnyConnectedBot,
  isAnyConnectedBotStarRef,
  isAnyContact,
  isAnyContactBirthday,
  isAnyContactStatus,
  isAnyContacts_Blocked,
  isAnyContacts_ContactBirthdays,
  isAnyContacts_Contacts,
  isAnyContacts_Found,
  isAnyContacts_ImportedContacts,
  isAnyContacts_ResolvedPeer,
  isAnyContacts_SponsoredPeers,
  isAnyContacts_TopPeers,
  isAnyDataJSON,
  isAnyDcOption,
  isAnyDefaultHistoryTTL,
  isAnyDialog,
  isAnyDialogFilter,
  isAnyDialogFilterSuggested,
  isAnyDialogPeer,
  isAnyDisallowedGiftsSettings,
  isAnyDocument,
  isAnyDocumentAttribute,
  isAnyDraftMessage,
  isAnyEmailVerification,
  isAnyEmailVerifyPurpose,
  isAnyEmojiGroup,
  isAnyEmojiKeyword,
  isAnyEmojiKeywordsDifference,
  isAnyEmojiLanguage,
  isAnyEmojiList,
  isAnyEmojiStatus,
  isAnyEmojiURL,
  isAnyEncryptedChat,
  isAnyEncryptedFile,
  isAnyEncryptedMessage,
  isAnyError,
  isAnyExportedChatInvite,
  isAnyExportedChatlistInvite,
  isAnyExportedContactToken,
  isAnyExportedMessageLink,
  isAnyExportedStoryLink,
  isAnyFactCheck,
  isAnyFileHash,
  isAnyFolder,
  isAnyFolderPeer,
  isAnyForumTopic,
  isAnyFoundStory,
  isAnyFragment_CollectibleInfo,
  isAnyGame,
  isAnyGeoPoint,
  isAnyGeoPointAddress,
  isAnyGlobalPrivacySettings,
  isAnyGroupCall,
  isAnyGroupCallDonor,
  isAnyGroupCallMessage,
  isAnyGroupCallParticipant,
  isAnyGroupCallParticipantVideo,
  isAnyGroupCallParticipantVideoSourceGroup,
  isAnyGroupCallStreamChannel,
  isAnyHelp_AppConfig,
  isAnyHelp_AppUpdate,
  isAnyHelp_CountriesList,
  isAnyHelp_Country,
  isAnyHelp_CountryCode,
  isAnyHelp_DeepLinkInfo,
  isAnyHelp_InviteText,
  isAnyHelp_PassportConfig,
  isAnyHelp_PeerColorOption,
  isAnyHelp_PeerColorSet,
  isAnyHelp_PeerColors,
  isAnyHelp_PremiumPromo,
  isAnyHelp_PromoData,
  isAnyHelp_RecentMeUrls,
  isAnyHelp_Support,
  isAnyHelp_SupportName,
  isAnyHelp_TermsOfService,
  isAnyHelp_TermsOfServiceUpdate,
  isAnyHelp_TimezonesList,
  isAnyHelp_UserInfo,
  isAnyHighScore,
  isAnyImportedContact,
  isAnyInlineBotSwitchPM,
  isAnyInlineBotWebView,
  isAnyInlineQueryPeerType,
  isAnyInputAiComposeTone,
  isAnyInputAppEvent,
  isAnyInputBotApp,
  isAnyInputBotInlineMessage,
  isAnyInputBotInlineMessageID,
  isAnyInputBotInlineResult,
  isAnyInputBusinessAwayMessage,
  isAnyInputBusinessBotRecipients,
  isAnyInputBusinessChatLink,
  isAnyInputBusinessGreetingMessage,
  isAnyInputBusinessIntro,
  isAnyInputBusinessRecipients,
  isAnyInputChannel,
  isAnyInputChatPhoto,
  isAnyInputChatTheme,
  isAnyInputChatlist,
  isAnyInputCheckPasswordSRP,
  isAnyInputClientProxy,
  isAnyInputCollectible,
  isAnyInputContact,
  isAnyInputDialogPeer,
  isAnyInputDocument,
  isAnyInputEncryptedChat,
  isAnyInputEncryptedFile,
  isAnyInputFile,
  isAnyInputFileLocation,
  isAnyInputFolderPeer,
  isAnyInputGame,
  isAnyInputGeoPoint,
  isAnyInputGroupCall,
  isAnyInputInvoice,
  isAnyInputMedia,
  isAnyInputMessage,
  isAnyInputMessageReadMetric,
  isAnyInputNotifyPeer,
  isAnyInputPasskeyCredential,
  isAnyInputPasskeyResponse,
  isAnyInputPaymentCredentials,
  isAnyInputPeer,
  isAnyInputPeerNotifySettings,
  isAnyInputPhoneCall,
  isAnyInputPhoto,
  isAnyInputPrivacyKey,
  isAnyInputPrivacyRule,
  isAnyInputQuickReplyShortcut,
  isAnyInputReplyTo,
  isAnyInputRichFile,
  isAnyInputRichMessage,
  isAnyInputSavedStarGift,
  isAnyInputSecureFile,
  isAnyInputSecureValue,
  isAnyInputSingleMedia,
  isAnyInputStarGiftAuction,
  isAnyInputStarsTransaction,
  isAnyInputStickerSet,
  isAnyInputStickerSetItem,
  isAnyInputStickeredMedia,
  isAnyInputStorePaymentPurpose,
  isAnyInputTheme,
  isAnyInputThemeSettings,
  isAnyInputUser,
  isAnyInputWallPaper,
  isAnyInputWebDocument,
  isAnyInputWebFileLocation,
  isAnyInvoice,
  isAnyJSONObjectValue,
  isAnyJSONValue,
  isAnyJoinChatBotResult,
  isAnyKeyboardButton,
  isAnyKeyboardButtonRow,
  isAnyKeyboardButtonStyle,
  isAnyLabeledPrice,
  isAnyLangPackDifference,
  isAnyLangPackLanguage,
  isAnyLangPackString,
  isAnyMaskCoords,
  isAnyMediaArea,
  isAnyMediaAreaCoordinates,
  isAnyMessage,
  isAnyMessageAction,
  isAnyMessageEntity,
  isAnyMessageExtendedMedia,
  isAnyMessageFwdHeader,
  isAnyMessageMedia,
  isAnyMessagePeerReaction,
  isAnyMessagePeerVote,
  isAnyMessageRange,
  isAnyMessageReactions,
  isAnyMessageReactor,
  isAnyMessageReplies,
  isAnyMessageReplyHeader,
  isAnyMessageReportOption,
  isAnyMessageViews,
  isAnyMessagesFilter,
  isAnyMessages_AffectedFoundMessages,
  isAnyMessages_AffectedHistory,
  isAnyMessages_AffectedMessages,
  isAnyMessages_AllStickers,
  isAnyMessages_ArchivedStickers,
  isAnyMessages_AvailableEffects,
  isAnyMessages_AvailableReactions,
  isAnyMessages_BotApp,
  isAnyMessages_BotCallbackAnswer,
  isAnyMessages_BotPreparedInlineMessage,
  isAnyMessages_BotResults,
  isAnyMessages_ChatAdminsWithInvites,
  isAnyMessages_ChatFull,
  isAnyMessages_ChatInviteImporters,
  isAnyMessages_ChatInviteJoinResult,
  isAnyMessages_Chats,
  isAnyMessages_CheckedHistoryImportPeer,
  isAnyMessages_ComposedMessageWithAI,
  isAnyMessages_DhConfig,
  isAnyMessages_DialogFilters,
  isAnyMessages_Dialogs,
  isAnyMessages_DiscussionMessage,
  isAnyMessages_EmojiGameInfo,
  isAnyMessages_EmojiGameOutcome,
  isAnyMessages_EmojiGroups,
  isAnyMessages_ExportedChatInvite,
  isAnyMessages_ExportedChatInvites,
  isAnyMessages_FavedStickers,
  isAnyMessages_FeaturedStickers,
  isAnyMessages_ForumTopics,
  isAnyMessages_FoundStickerSets,
  isAnyMessages_FoundStickers,
  isAnyMessages_HighScores,
  isAnyMessages_HistoryImport,
  isAnyMessages_HistoryImportParsed,
  isAnyMessages_InactiveChats,
  isAnyMessages_InvitedUsers,
  isAnyMessages_MessageEditData,
  isAnyMessages_MessageReactionsList,
  isAnyMessages_MessageViews,
  isAnyMessages_Messages,
  isAnyMessages_MyStickers,
  isAnyMessages_PeerDialogs,
  isAnyMessages_PeerSettings,
  isAnyMessages_PreparedInlineMessage,
  isAnyMessages_QuickReplies,
  isAnyMessages_Reactions,
  isAnyMessages_RecentStickers,
  isAnyMessages_SavedDialogs,
  isAnyMessages_SavedGifs,
  isAnyMessages_SavedReactionTags,
  isAnyMessages_SearchCounter,
  isAnyMessages_SearchResultsCalendar,
  isAnyMessages_SearchResultsPositions,
  isAnyMessages_SentEncryptedMessage,
  isAnyMessages_SponsoredMessages,
  isAnyMessages_StickerSet,
  isAnyMessages_StickerSetInstallResult,
  isAnyMessages_Stickers,
  isAnyMessages_TranscribedAudio,
  isAnyMessages_TranslatedText,
  isAnyMessages_VotesList,
  isAnyMessages_WebPage,
  isAnyMessages_WebPagePreview,
  isAnyMissingInvitee,
  isAnyMtcute_InputPeer,
  isAnyMtcute_Update,
  isAnyMyBoost,
  isAnyNearestDc,
  isAnyNotificationSound,
  isAnyNotifyPeer,
  isAnyOutboxReadDate,
  isAnyPage,
  isAnyPageBlock,
  isAnyPageCaption,
  isAnyPageListItem,
  isAnyPageListOrderedItem,
  isAnyPageRelatedArticle,
  isAnyPageTableCell,
  isAnyPageTableRow,
  isAnyPaidReactionPrivacy,
  isAnyPasskey,
  isAnyPasswordKdfAlgo,
  isAnyPaymentCharge,
  isAnyPaymentFormMethod,
  isAnyPaymentRequestedInfo,
  isAnyPaymentSavedCredentials,
  isAnyPayments_BankCardData,
  isAnyPayments_CheckCanSendGiftResult,
  isAnyPayments_CheckedGiftCode,
  isAnyPayments_ConnectedStarRefBots,
  isAnyPayments_ExportedInvoice,
  isAnyPayments_GiveawayInfo,
  isAnyPayments_PaymentForm,
  isAnyPayments_PaymentReceipt,
  isAnyPayments_PaymentResult,
  isAnyPayments_ResaleStarGifts,
  isAnyPayments_SavedInfo,
  isAnyPayments_SavedStarGifts,
  isAnyPayments_StarGiftActiveAuctions,
  isAnyPayments_StarGiftAuctionAcquiredGifts,
  isAnyPayments_StarGiftAuctionState,
  isAnyPayments_StarGiftCollections,
  isAnyPayments_StarGiftUpgradeAttributes,
  isAnyPayments_StarGiftUpgradePreview,
  isAnyPayments_StarGiftWithdrawalUrl,
  isAnyPayments_StarGifts,
  isAnyPayments_StarsRevenueAdsAccountUrl,
  isAnyPayments_StarsRevenueStats,
  isAnyPayments_StarsRevenueWithdrawalUrl,
  isAnyPayments_StarsStatus,
  isAnyPayments_SuggestedStarRefBots,
  isAnyPayments_UniqueStarGift,
  isAnyPayments_UniqueStarGiftValueInfo,
  isAnyPayments_ValidatedRequestedInfo,
  isAnyPeer,
  isAnyPeerBlocked,
  isAnyPeerColor,
  isAnyPeerLocated,
  isAnyPeerNotifySettings,
  isAnyPeerSettings,
  isAnyPeerStories,
  isAnyPendingSuggestion,
  isAnyPhoneCall,
  isAnyPhoneCallDiscardReason,
  isAnyPhoneCallProtocol,
  isAnyPhoneConnection,
  isAnyPhone_ExportedGroupCallInvite,
  isAnyPhone_GroupCall,
  isAnyPhone_GroupCallStars,
  isAnyPhone_GroupCallStreamChannels,
  isAnyPhone_GroupCallStreamRtmpUrl,
  isAnyPhone_GroupParticipants,
  isAnyPhone_JoinAsPeers,
  isAnyPhone_PhoneCall,
  isAnyPhoto,
  isAnyPhotoSize,
  isAnyPhotos_Photo,
  isAnyPhotos_Photos,
  isAnyPoll,
  isAnyPollAnswer,
  isAnyPollAnswerVoters,
  isAnyPollResults,
  isAnyPopularContact,
  isAnyPostAddress,
  isAnyPostInteractionCounters,
  isAnyPremiumGiftCodeOption,
  isAnyPremiumSubscriptionOption,
  isAnyPremium_BoostsList,
  isAnyPremium_BoostsStatus,
  isAnyPremium_MyBoosts,
  isAnyPrepaidGiveaway,
  isAnyPrivacyKey,
  isAnyPrivacyRule,
  isAnyProfileTab,
  isAnyPublicForward,
  isAnyQuickReply,
  isAnyReaction,
  isAnyReactionCount,
  isAnyReactionNotificationsFrom,
  isAnyReactionsNotifySettings,
  isAnyReadParticipantDate,
  isAnyReceivedNotifyMessage,
  isAnyRecentMeUrl,
  isAnyRecentStory,
  isAnyReplyMarkup,
  isAnyReportReason,
  isAnyReportResult,
  isAnyRequestPeerType,
  isAnyRequestedPeer,
  isAnyRequirementToContact,
  isAnyRestrictionReason,
  isAnyRichMessage,
  isAnyRichText,
  isAnySavedContact,
  isAnySavedDialog,
  isAnySavedReactionTag,
  isAnySavedStarGift,
  isAnySearchPostsFlood,
  isAnySearchResultsCalendarPeriod,
  isAnySearchResultsPosition,
  isAnySecureCredentialsEncrypted,
  isAnySecureData,
  isAnySecureFile,
  isAnySecurePasswordKdfAlgo,
  isAnySecurePlainData,
  isAnySecureRequiredType,
  isAnySecureSecretSettings,
  isAnySecureValue,
  isAnySecureValueError,
  isAnySecureValueHash,
  isAnySecureValueType,
  isAnySendAsPeer,
  isAnySendMessageAction,
  isAnyShippingOption,
  isAnySmsJob,
  isAnySmsjobs_EligibilityToJoin,
  isAnySmsjobs_Status,
  isAnySponsoredMessage,
  isAnySponsoredMessageReportOption,
  isAnySponsoredPeer,
  isAnyStarGift,
  isAnyStarGiftActiveAuctionState,
  isAnyStarGiftAttribute,
  isAnyStarGiftAttributeCounter,
  isAnyStarGiftAttributeId,
  isAnyStarGiftAttributeRarity,
  isAnyStarGiftAuctionAcquiredGift,
  isAnyStarGiftAuctionRound,
  isAnyStarGiftAuctionState,
  isAnyStarGiftAuctionUserState,
  isAnyStarGiftBackground,
  isAnyStarGiftCollection,
  isAnyStarGiftUpgradePrice,
  isAnyStarRefProgram,
  isAnyStarsAmount,
  isAnyStarsGiftOption,
  isAnyStarsGiveawayOption,
  isAnyStarsGiveawayWinnersOption,
  isAnyStarsRating,
  isAnyStarsRevenueStatus,
  isAnyStarsSubscription,
  isAnyStarsSubscriptionPricing,
  isAnyStarsTopupOption,
  isAnyStarsTransaction,
  isAnyStarsTransactionPeer,
  isAnyStatsAbsValueAndPrev,
  isAnyStatsDateRangeDays,
  isAnyStatsGraph,
  isAnyStatsGroupTopAdmin,
  isAnyStatsGroupTopInviter,
  isAnyStatsGroupTopPoster,
  isAnyStatsPercentValue,
  isAnyStatsURL,
  isAnyStats_BroadcastStats,
  isAnyStats_MegagroupStats,
  isAnyStats_MessageStats,
  isAnyStats_PollStats,
  isAnyStats_PublicForwards,
  isAnyStats_StoryStats,
  isAnyStickerKeyword,
  isAnyStickerPack,
  isAnyStickerSet,
  isAnyStickerSetCovered,
  isAnyStickers_SuggestedShortName,
  isAnyStorage_FileType,
  isAnyStoriesStealthMode,
  isAnyStories_Albums,
  isAnyStories_AllStories,
  isAnyStories_CanSendStoryCount,
  isAnyStories_FoundStories,
  isAnyStories_PeerStories,
  isAnyStories_Stories,
  isAnyStories_StoryReactionsList,
  isAnyStories_StoryViews,
  isAnyStories_StoryViewsList,
  isAnyStoryAlbum,
  isAnyStoryFwdHeader,
  isAnyStoryItem,
  isAnyStoryReaction,
  isAnyStoryView,
  isAnyStoryViews,
  isAnySuggestedPost,
  isAnyTextWithEntities,
  isAnyTheme,
  isAnyThemeSettings,
  isAnyTimezone,
  isAnyTodoCompletion,
  isAnyTodoItem,
  isAnyTodoList,
  isAnyTopPeer,
  isAnyTopPeerCategory,
  isAnyTopPeerCategoryPeers,
  isAnyUpdate,
  isAnyUpdates,
  isAnyUpdates_ChannelDifference,
  isAnyUpdates_Difference,
  isAnyUpdates_State,
  isAnyUpload_CdnFile,
  isAnyUpload_File,
  isAnyUpload_WebFile,
  isAnyUrlAuthResult,
  isAnyUser,
  isAnyUserFull,
  isAnyUserProfilePhoto,
  isAnyUserStatus,
  isAnyUsername,
  isAnyUsers_SavedMusic,
  isAnyUsers_UserFull,
  isAnyUsers_Users,
  isAnyVideoSize,
  isAnyWallPaper,
  isAnyWallPaperSettings,
  isAnyWebAuthorization,
  isAnyWebDocument,
  isAnyWebDomainException,
  isAnyWebPage,
  isAnyWebPageAttribute,
  isAnyWebViewMessageSent,
  isAnyWebViewResult
};
