// @mtcute/core/tl/inner-mtp.js
function _isAny() {
  const a = arguments;
  return function(o) {
    if (typeof o !== "object" || o === null) return false;
    for (let i = 0; i < a.length; i++) if (o._ === a[i]) return true;
    return false;
  };
}
const LAYER = 0;
const isAnyResPQ = /* @__PURE__ */ _isAny("mt_resPQ", "mt_req_pq_multi");
const isAnyP_Q_inner_data = /* @__PURE__ */ _isAny("mt_p_q_inner_data_dc", "mt_p_q_inner_data_temp_dc");
const isAnyServer_DH_Params = /* @__PURE__ */ _isAny("mt_server_DH_params_ok", "mt_req_DH_params");
const isAnyServer_DH_inner_data = /* @__PURE__ */ _isAny("mt_server_DH_inner_data");
const isAnyClient_DH_Inner_Data = /* @__PURE__ */ _isAny("mt_client_DH_inner_data");
const isAnySet_client_DH_params_answer = /* @__PURE__ */ _isAny("mt_dh_gen_ok", "mt_dh_gen_retry", "mt_dh_gen_fail", "mt_set_client_DH_params");
const isAnyBindAuthKeyInner = /* @__PURE__ */ _isAny("mt_bind_auth_key_inner");
const isAnyRpcError = /* @__PURE__ */ _isAny("mt_rpc_error");
const isAnyRpcDropAnswer = /* @__PURE__ */ _isAny("mt_rpc_answer_unknown", "mt_rpc_answer_dropped_running", "mt_rpc_answer_dropped", "mt_rpc_drop_answer");
const isAnyFutureSalt = /* @__PURE__ */ _isAny("mt_future_salt");
const isAnyFutureSalts = /* @__PURE__ */ _isAny("mt_future_salts", "mt_get_future_salts");
const isAnyPong = /* @__PURE__ */ _isAny("mt_pong", "mt_ping", "mt_ping_delay_disconnect");
const isAnyDestroySessionRes = /* @__PURE__ */ _isAny("mt_destroy_session_ok", "mt_destroy_session_none", "mt_destroy_session");
const isAnyNewSession = /* @__PURE__ */ _isAny("mt_new_session_created");
const isAnyMsgsAck = /* @__PURE__ */ _isAny("mt_msgs_ack");
const isAnyBadMsgNotification = /* @__PURE__ */ _isAny("mt_bad_msg_notification", "mt_bad_server_salt");
const isAnyMsgResendReq = /* @__PURE__ */ _isAny("mt_msg_resend_req");
const isAnyMsgsStateReq = /* @__PURE__ */ _isAny("mt_msgs_state_req");
const isAnyMsgsStateInfo = /* @__PURE__ */ _isAny("mt_msgs_state_info");
const isAnyMsgsAllInfo = /* @__PURE__ */ _isAny("mt_msgs_all_info");
const isAnyMsgDetailedInfo = /* @__PURE__ */ _isAny("mt_msg_detailed_info", "mt_msg_new_detailed_info");
const isAnyDestroyAuthKeyRes = /* @__PURE__ */ _isAny("mt_destroy_auth_key_ok", "mt_destroy_auth_key_none", "mt_destroy_auth_key_fail", "mt_destroy_auth_key");
const isAnyHttpWait = /* @__PURE__ */ _isAny("mt_http_wait");
export {
  LAYER,
  isAnyBadMsgNotification,
  isAnyBindAuthKeyInner,
  isAnyClient_DH_Inner_Data,
  isAnyDestroyAuthKeyRes,
  isAnyDestroySessionRes,
  isAnyFutureSalt,
  isAnyFutureSalts,
  isAnyHttpWait,
  isAnyMsgDetailedInfo,
  isAnyMsgResendReq,
  isAnyMsgsAck,
  isAnyMsgsAllInfo,
  isAnyMsgsStateInfo,
  isAnyMsgsStateReq,
  isAnyNewSession,
  isAnyP_Q_inner_data,
  isAnyPong,
  isAnyResPQ,
  isAnyRpcDropAnswer,
  isAnyRpcError,
  isAnyServer_DH_Params,
  isAnyServer_DH_inner_data,
  isAnySet_client_DH_params_answer
};
