// @mtcute/core/highlevel/methods/stories/get-story-viewers.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const resolvePeer = require("../users/resolve-peer.cjs");
const storyViewer = require("../../types/stories/story-viewer.cjs");
async function getStoryViewers(client, peerId, storyId, params) {
  if (!params) params = {};
  const { onlyContacts, sortBy = "reaction", query, offset = "", limit = 100 } = params;
  const res = await client.call({
    _: "stories.getStoryViewsList",
    peer: await resolvePeer.resolvePeer(client, peerId),
    justContacts: onlyContacts,
    reactionsFirst: sortBy === "reaction",
    q: query,
    id: storyId,
    offset,
    limit
  });
  return new storyViewer.StoryViewersList(res);
}
exports.getStoryViewers = getStoryViewers;
