// @mtcute/core/highlevel/methods/stories/send-story.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const normalizeInputMedia = require("../files/normalize-input-media.cjs");
const normalizePrivacyRules = require("../misc/normalize-privacy-rules.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
async function sendStory(client, params) {
  const { peer = "me", pinned, forbidForwards, interactiveElements, period, addToAlbums } = params;
  let { media } = params;
  if (typeof media === "string") {
    media = {
      type: "auto",
      file: media
    };
  }
  const inputMedia = await normalizeInputMedia._normalizeInputMedia(client, media);
  const privacyRules = params.privacyRules ? await normalizePrivacyRules._normalizePrivacyRules(client, params.privacyRules) : [{ _: "inputPrivacyValueAllowAll" }];
  const [caption, entities] = await normalizeText._normalizeInputText(
    client,
    // some types dont have `caption` field, and ts warns us,
    // but since it's JS, they'll just be `undefined` and properly handled by the method
    params.caption || media.caption
  );
  const res = await client.call({
    _: "stories.sendStory",
    pinned,
    noforwards: forbidForwards,
    peer: await resolvePeer.resolvePeer(client, peer),
    media: inputMedia,
    mediaAreas: interactiveElements,
    caption,
    entities,
    privacyRules,
    randomId: longUtils.randomLong(),
    period,
    albums: addToAlbums
  });
  return findInUpdate._findStoryInUpdate(client, res);
}
exports.sendStory = sendStory;
