// @mtcute/core/highlevel/methods/stories/send-story.js
import { randomLong } from "../../../utils/long-utils.js";
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
import { _normalizePrivacyRules } from "../misc/normalize-privacy-rules.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findStoryInUpdate } from "./find-in-update.js";
async function sendStory(client, params) {
  const { peer = "me", pinned, forbidForwards, interactiveElements, period, addToAlbums } = params;
  let { media } = params;
  if (typeof media === "string") {
    media = {
      type: "auto",
      file: media
    };
  }
  const inputMedia = await _normalizeInputMedia(client, media);
  const privacyRules = params.privacyRules ? await _normalizePrivacyRules(client, params.privacyRules) : [{ _: "inputPrivacyValueAllowAll" }];
  const [caption, entities] = await _normalizeInputText(
    client,
    // some types dont have `caption` field, and ts warns us,
    // but since it's JS, they'll just be `undefined` and properly handled by the method
    params.caption || media.caption
  );
  const res = await client.call({
    _: "stories.sendStory",
    pinned,
    noforwards: forbidForwards,
    peer: await resolvePeer(client, peer),
    media: inputMedia,
    mediaAreas: interactiveElements,
    caption,
    entities,
    privacyRules,
    randomId: randomLong(),
    period,
    albums: addToAlbums
  });
  return _findStoryInUpdate(client, res);
}
export {
  sendStory
};
