// @mtcute/core/highlevel/methods/messages/append-todo-list.js
import { assert } from "@fuman/utils";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findMessageInUpdate } from "./find-in-update.js";
import { getMessages } from "./get-messages.js";
import { normalizeInputMessageId } from "../../types/messages/input-message-id.js";
import { inputTextToTl } from "../../types/misc/entities.js";
async function appendTodoList(client, params) {
  const { items, shouldDispatch } = params;
  const { chatId, message } = normalizeInputMessageId(params);
  let msgObject;
  if (typeof params.message === "number") {
    const [msg] = await getMessages(client, chatId, [message]);
    assert(msg != null, "message not found");
    msgObject = msg;
  } else {
    msgObject = params.message;
  }
  assert(msgObject?.media?.type === "todo", "message is not a todo list");
  const itemId = Math.max(...msgObject.media.items.map((i) => i.id)) + 1;
  const res = await client.call({
    _: "messages.appendTodoList",
    peer: await resolvePeer(client, chatId),
    msgId: message,
    list: (Array.isArray(items) ? items : [items]).map((item, index) => ({
      _: "todoItem",
      id: itemId + index,
      title: inputTextToTl(item)
    }))
  });
  return _findMessageInUpdate(client, res, false, !shouldDispatch, true);
}
export {
  appendTodoList
};
