// @mtcute/core/highlevel/methods/messages/append-todo-list.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
const getMessages = require("./get-messages.cjs");
const inputMessageId = require("../../types/messages/input-message-id.cjs");
const entities = require("../../types/misc/entities.cjs");
async function appendTodoList(client, params) {
  const { items, shouldDispatch } = params;
  const { chatId, message } = inputMessageId.normalizeInputMessageId(params);
  let msgObject;
  if (typeof params.message === "number") {
    const [msg] = await getMessages.getMessages(client, chatId, [message]);
    utils.assert(msg != null, "message not found");
    msgObject = msg;
  } else {
    msgObject = params.message;
  }
  utils.assert(msgObject?.media?.type === "todo", "message is not a todo list");
  const itemId = Math.max(...msgObject.media.items.map((i) => i.id)) + 1;
  const res = await client.call({
    _: "messages.appendTodoList",
    peer: await resolvePeer.resolvePeer(client, chatId),
    msgId: message,
    list: (Array.isArray(items) ? items : [items]).map((item, index) => ({
      _: "todoItem",
      id: itemId + index,
      title: entities.inputTextToTl(item)
    }))
  });
  return findInUpdate._findMessageInUpdate(client, res, false, !shouldDispatch, true);
}
exports.appendTodoList = appendTodoList;
