// @mtcute/core/highlevel/methods/messages/close-poll.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const errors = require("../../../types/errors.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const utils = require("../../updates/utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const poll = require("../../types/media/poll.cjs");
const inputMessageId = require("../../types/messages/input-message-id.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
async function closePoll(client, params) {
  const { chatId, message } = inputMessageId.normalizeInputMessageId(params);
  const res = await client.call({
    _: "messages.editMessage",
    peer: await resolvePeer.resolvePeer(client, chatId),
    id: message,
    media: {
      _: "inputMediaPoll",
      poll: {
        _: "poll",
        id: Long.ZERO,
        closed: true,
        question: { _: "textWithEntities", text: "", entities: [] },
        answers: [],
        hash: Long.ZERO
      }
    }
  });
  utils.assertIsUpdatesGroup("messages.editMessage", res);
  client.handleClientUpdate(res, !params.shouldDispatch);
  const upd = res.updates[0];
  typeAssertions.assertTypeIs("messages.editMessage (@ .updates[0])", upd, "updateMessagePoll");
  if (!upd.poll) {
    throw new errors.MtTypeAssertionError("messages.editMessage (@ .updates[0].poll)", "poll", "undefined");
  }
  const peers = peersIndex.PeersIndex.from(res);
  return new poll.Poll(upd.poll, peers, upd.results);
}
exports.closePoll = closePoll;
