// @mtcute/core/highlevel/methods/messages/close-poll.js
import Long from "long";
import { MtTypeAssertionError } from "../../../types/errors.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { assertIsUpdatesGroup } from "../../updates/utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { Poll as Poll$2 } from "../../types/media/poll.js";
import { normalizeInputMessageId } from "../../types/messages/input-message-id.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
async function closePoll(client, params) {
  const { chatId, message } = normalizeInputMessageId(params);
  const res = await client.call({
    _: "messages.editMessage",
    peer: await resolvePeer(client, chatId),
    id: message,
    media: {
      _: "inputMediaPoll",
      poll: {
        _: "poll",
        id: Long.ZERO,
        closed: true,
        question: { _: "textWithEntities", text: "", entities: [] },
        answers: [],
        hash: Long.ZERO
      }
    }
  });
  assertIsUpdatesGroup("messages.editMessage", res);
  client.handleClientUpdate(res, !params.shouldDispatch);
  const upd = res.updates[0];
  assertTypeIs("messages.editMessage (@ .updates[0])", upd, "updateMessagePoll");
  if (!upd.poll) {
    throw new MtTypeAssertionError("messages.editMessage (@ .updates[0].poll)", "poll", "undefined");
  }
  const peers = PeersIndex.from(res);
  return new Poll$2(upd.poll, peers, upd.results);
}
export {
  closePoll
};
