// @mtcute/core/highlevel/methods/messages/create-streaming-draft.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const entities = require("../../utils/entities.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
async function createStreamingDraft(client, chatId, params) {
  const {
    mode = "replace",
    threadId,
    businessConnectionId
  } = params ?? {};
  const randomId = longUtils.randomLong();
  const peer = await resolvePeer.resolvePeer(client, chatId);
  let previousText;
  const sendDraft = async (text) => {
    if (typeof text === "string") {
      text = { text };
    }
    if (mode === "append" && previousText) {
      text = entities.joinTextWithEntities([previousText, text]);
    }
    previousText = text;
    const [_text, entities$1 = []] = await normalizeText._normalizeInputText(client, text);
    const r = await client.call({
      _: "messages.setTyping",
      peer,
      action: {
        _: "sendMessageTextDraftAction",
        text: {
          _: "textWithEntities",
          text: _text,
          entities: entities$1
        },
        randomId
      },
      topMsgId: threadId
    }, {
      businessConnectionId
    });
    typeAssertions.assertTrue("messages.setTyping", r);
  };
  return {
    randomId,
    send: sendDraft,
    get finalText() {
      return previousText ?? { text: "", entities: [] };
    }
  };
}
exports.createStreamingDraft = createStreamingDraft;
