// @mtcute/core/highlevel/methods/messages/create-streaming-draft.js
import { randomLong } from "../../../utils/long-utils.js";
import { assertTrue } from "../../../utils/type-assertions.js";
import { joinTextWithEntities } from "../../utils/entities.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
async function createStreamingDraft(client, chatId, params) {
  const {
    mode = "replace",
    threadId,
    businessConnectionId
  } = params ?? {};
  const randomId = randomLong();
  const peer = await resolvePeer(client, chatId);
  let previousText;
  const sendDraft = async (text) => {
    if (typeof text === "string") {
      text = { text };
    }
    if (mode === "append" && previousText) {
      text = joinTextWithEntities([previousText, text]);
    }
    previousText = text;
    const [_text, entities = []] = await _normalizeInputText(client, text);
    const r = await client.call({
      _: "messages.setTyping",
      peer,
      action: {
        _: "sendMessageTextDraftAction",
        text: {
          _: "textWithEntities",
          text: _text,
          entities
        },
        randomId
      },
      topMsgId: threadId
    }, {
      businessConnectionId
    });
    assertTrue("messages.setTyping", r);
  };
  return {
    randomId,
    send: sendDraft,
    get finalText() {
      return previousText ?? { text: "", entities: [] };
    }
  };
}
export {
  createStreamingDraft
};
