// @mtcute/core/highlevel/methods/messages/edit-message.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const normalizeInputMedia = require("../files/normalize-input-media.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
const inputMessageId = require("../../types/messages/input-message-id.cjs");
const factories = require("../../types/bots/keyboards/factories.cjs");
async function editMessage(client, params) {
  const { chatId, message } = inputMessageId.normalizeInputMessageId(params);
  let content;
  let entities;
  let media;
  if (params.media) {
    media = await normalizeInputMedia._normalizeInputMedia(client, params.media, params);
    if ("caption" in params.media && params.media.caption !== void 0) {
      [content, entities] = await normalizeText._normalizeInputText(client, params.media.caption);
    }
  }
  if (params.text !== void 0) {
    [content, entities] = await normalizeText._normalizeInputText(client, params.text);
  }
  const res = await client.call({
    _: "messages.editMessage",
    id: message,
    peer: await resolvePeer.resolvePeer(client, chatId),
    noWebpage: params.disableWebPreview,
    replyMarkup: factories._convertToTl(params.replyMarkup),
    message: content,
    entities,
    media,
    invertMedia: params.invertMedia
  }, {
    businessConnectionId: params.businessConnectionId
  });
  return findInUpdate._findMessageInUpdate(client, res, true, !params.shouldDispatch);
}
exports.editMessage = editMessage;
