// @mtcute/core/highlevel/methods/messages/edit-message.js
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findMessageInUpdate } from "./find-in-update.js";
import { normalizeInputMessageId } from "../../types/messages/input-message-id.js";
import { _convertToTl } from "../../types/bots/keyboards/factories.js";
async function editMessage(client, params) {
  const { chatId, message } = normalizeInputMessageId(params);
  let content;
  let entities;
  let media;
  if (params.media) {
    media = await _normalizeInputMedia(client, params.media, params);
    if ("caption" in params.media && params.media.caption !== void 0) {
      [content, entities] = await _normalizeInputText(client, params.media.caption);
    }
  }
  if (params.text !== void 0) {
    [content, entities] = await _normalizeInputText(client, params.text);
  }
  const res = await client.call({
    _: "messages.editMessage",
    id: message,
    peer: await resolvePeer(client, chatId),
    noWebpage: params.disableWebPreview,
    replyMarkup: _convertToTl(params.replyMarkup),
    message: content,
    entities,
    media,
    invertMedia: params.invertMedia
  }, {
    businessConnectionId: params.businessConnectionId
  });
  return _findMessageInUpdate(client, res, true, !params.shouldDispatch);
}
export {
  editMessage
};
