// @mtcute/core/highlevel/methods/messages/forward-messages.js
import { MtArgumentError } from "../../../types/errors.js";
import { randomLong } from "../../../utils/long-utils.js";
import { assertIsUpdatesGroup } from "../../updates/utils.js";
import { normalizeDate } from "../../utils/misc-utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _normalizeQuickReplyShortcut } from "./send-common.js";
import { _getTypingTimerId } from "./set-typing.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
async function forwardMessagesById(client, params) {
  const {
    messages,
    toChatId,
    fromChatId,
    silent,
    schedule,
    forbidForwards,
    sendAs,
    noAuthor,
    noCaption,
    allowPaidFloodskip,
    videoTimestamp,
    toThreadId,
    toMonoforumPeer
  } = params;
  if (messages.length > 100) {
    throw new MtArgumentError("You can forward no more than 100 messages at once");
  }
  const toPeer = await resolvePeer(client, toChatId);
  await client.timers.cancel(_getTypingTimerId(toPeer));
  const res = await client.call({
    _: "messages.forwardMessages",
    toPeer,
    fromPeer: await resolvePeer(client, fromChatId),
    id: messages,
    silent,
    scheduleDate: normalizeDate(schedule),
    randomId: Array.from({ length: messages.length }, () => randomLong()),
    dropAuthor: noAuthor,
    dropMediaCaptions: noCaption,
    noforwards: forbidForwards,
    sendAs: sendAs ? await resolvePeer(client, sendAs) : void 0,
    quickReplyShortcut: _normalizeQuickReplyShortcut(params.quickReply),
    allowPaidFloodskip,
    videoTimestamp,
    allowPaidStars: params.allowPaidMessages,
    topMsgId: toThreadId,
    replyTo: toMonoforumPeer ? {
      _: "inputReplyToMonoForum",
      monoforumPeerId: await resolvePeer(client, toMonoforumPeer)
    } : void 0
  });
  assertIsUpdatesGroup("messages.forwardMessages", res);
  client.handleClientUpdate(res, !params.shouldDispatch);
  const peers = PeersIndex.from(res);
  const forwarded = [];
  res.updates.forEach((upd) => {
    switch (upd._) {
      case "updateNewMessage":
      case "updateNewChannelMessage":
      case "updateNewScheduledMessage":
        forwarded.push(new Message$2(upd.message, peers, upd._ === "updateNewScheduledMessage"));
        break;
    }
  });
  return forwarded;
}
async function forwardMessages(client, params) {
  const { messages, ...rest } = params;
  return forwardMessagesById(client, {
    ...rest,
    fromChatId: messages[0].chat.inputPeer,
    messages: messages.map((it) => it.id)
  });
}
export {
  forwardMessages,
  forwardMessagesById
};
