// @mtcute/core/highlevel/methods/messages/forward-messages.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const longUtils = require("../../../utils/long-utils.cjs");
const utils = require("../../updates/utils.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const sendCommon = require("./send-common.cjs");
const setTyping = require("./set-typing.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
async function forwardMessagesById(client, params) {
  const {
    messages,
    toChatId,
    fromChatId,
    silent,
    schedule,
    forbidForwards,
    sendAs,
    noAuthor,
    noCaption,
    allowPaidFloodskip,
    videoTimestamp,
    toThreadId,
    toMonoforumPeer
  } = params;
  if (messages.length > 100) {
    throw new errors.MtArgumentError("You can forward no more than 100 messages at once");
  }
  const toPeer = await resolvePeer.resolvePeer(client, toChatId);
  await client.timers.cancel(setTyping._getTypingTimerId(toPeer));
  const res = await client.call({
    _: "messages.forwardMessages",
    toPeer,
    fromPeer: await resolvePeer.resolvePeer(client, fromChatId),
    id: messages,
    silent,
    scheduleDate: miscUtils.normalizeDate(schedule),
    randomId: Array.from({ length: messages.length }, () => longUtils.randomLong()),
    dropAuthor: noAuthor,
    dropMediaCaptions: noCaption,
    noforwards: forbidForwards,
    sendAs: sendAs ? await resolvePeer.resolvePeer(client, sendAs) : void 0,
    quickReplyShortcut: sendCommon._normalizeQuickReplyShortcut(params.quickReply),
    allowPaidFloodskip,
    videoTimestamp,
    allowPaidStars: params.allowPaidMessages,
    topMsgId: toThreadId,
    replyTo: toMonoforumPeer ? {
      _: "inputReplyToMonoForum",
      monoforumPeerId: await resolvePeer.resolvePeer(client, toMonoforumPeer)
    } : void 0
  });
  utils.assertIsUpdatesGroup("messages.forwardMessages", res);
  client.handleClientUpdate(res, !params.shouldDispatch);
  const peers = peersIndex.PeersIndex.from(res);
  const forwarded = [];
  res.updates.forEach((upd) => {
    switch (upd._) {
      case "updateNewMessage":
      case "updateNewChannelMessage":
      case "updateNewScheduledMessage":
        forwarded.push(new message.Message(upd.message, peers, upd._ === "updateNewScheduledMessage"));
        break;
    }
  });
  return forwarded;
}
async function forwardMessages(client, params) {
  const { messages, ...rest } = params;
  return forwardMessagesById(client, {
    ...rest,
    fromChatId: messages[0].chat.inputPeer,
    messages: messages.map((it) => it.id)
  });
}
exports.forwardMessages = forwardMessages;
exports.forwardMessagesById = forwardMessagesById;
