// @mtcute/core/highlevel/methods/messages/get-history.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const defaultOffset = {
  id: 0,
  date: 0
};
const defaultOffsetReverse = {
  id: 1,
  date: 0
};
async function getHistory(client, chatId, params) {
  if (!params) params = {};
  const {
    reverse = false,
    limit = 100,
    offset: { id: offsetId, date: offsetDate } = reverse ? defaultOffsetReverse : defaultOffset,
    addOffset = 0,
    minId = 0,
    maxId = 0
  } = params;
  const addOffsetAdjusted = addOffset + (reverse ? -limit : 0);
  const peer = await resolvePeer.resolvePeer(client, chatId);
  const res = await client.call({
    _: "messages.getHistory",
    peer,
    offsetId,
    offsetDate,
    addOffset: addOffsetAdjusted,
    limit,
    maxId,
    minId,
    hash: Long.ZERO
  });
  typeAssertions.assertTypeIsNot("getHistory", res, "messages.messagesNotModified");
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new message.Message(msg, peers));
  if (reverse) msgs.reverse();
  const last = msgs[msgs.length - 1];
  const next = last ? {
    id: last.id + (reverse ? 1 : 0),
    date: last.raw.date
  } : void 0;
  return miscUtils.makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
exports.getHistory = getHistory;
