// @mtcute/core/highlevel/methods/messages/get-history.js
import Long from "long";
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
import { makeArrayPaginated } from "../../utils/misc-utils.js";
const defaultOffset = {
  id: 0,
  date: 0
};
const defaultOffsetReverse = {
  id: 1,
  date: 0
};
async function getHistory(client, chatId, params) {
  if (!params) params = {};
  const {
    reverse = false,
    limit = 100,
    offset: { id: offsetId, date: offsetDate } = reverse ? defaultOffsetReverse : defaultOffset,
    addOffset = 0,
    minId = 0,
    maxId = 0
  } = params;
  const addOffsetAdjusted = addOffset + (reverse ? -limit : 0);
  const peer = await resolvePeer(client, chatId);
  const res = await client.call({
    _: "messages.getHistory",
    peer,
    offsetId,
    offsetDate,
    addOffset: addOffsetAdjusted,
    limit,
    maxId,
    minId,
    hash: Long.ZERO
  });
  assertTypeIsNot("getHistory", res, "messages.messagesNotModified");
  const peers = PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new Message$2(msg, peers));
  if (reverse) msgs.reverse();
  const last = msgs[msgs.length - 1];
  const next = last ? {
    id: last.id + (reverse ? 1 : 0),
    date: last.raw.date
  } : void 0;
  return makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
export {
  getHistory
};
