// @mtcute/core/highlevel/methods/messages/get-discussion-message.js
import { resolvePeer } from "../users/resolve-peer.js";
import { normalizeInputMessageId } from "../../types/messages/input-message-id.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
async function _getDiscussionMessage(client, peer, message) {
  const inputPeer = await resolvePeer(client, peer);
  const res = await client.call({
    _: "messages.getDiscussionMessage",
    peer: inputPeer,
    msgId: message
  });
  if (!res.messages.length || res.messages[0]._ === "messageEmpty") {
    return [inputPeer, message];
  }
  const msg = res.messages[0];
  const chat = res.chats.find((it) => it.id === msg.peerId.channelId);
  return [
    {
      _: "inputPeerChannel",
      channelId: chat.id,
      accessHash: chat.accessHash
    },
    msg.id
  ];
}
async function getDiscussionMessage(client, params) {
  const { chatId, message } = normalizeInputMessageId(params);
  const inputPeer = await resolvePeer(client, chatId);
  const res = await client.call({
    _: "messages.getDiscussionMessage",
    peer: inputPeer,
    msgId: message
  });
  if (!res.messages.length || res.messages[0]._ === "messageEmpty") {
    return null;
  }
  const msg = res.messages[0];
  const peers = PeersIndex.from(res);
  return new Message$2(msg, peers);
}
export {
  _getDiscussionMessage,
  getDiscussionMessage
};
