// @mtcute/core/highlevel/methods/messages/normalize-rich-message.js
import { parallelMap } from "@fuman/utils";
import { MtArgumentError, MtTypeAssertionError } from "../../../types/errors.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
function _richMediaCacheKey(media) {
  if (typeof media === "string") return `auto:${media}`;
  if ("_" in media) return void 0;
  if (typeof media.file === "string") return `${media.type}:${media.file}`;
  return media;
}
async function _uploadRichMedia(client, peer, id, media, params) {
  const cacheKey = params.uploadCache != null ? _richMediaCacheKey(media) : void 0;
  if (cacheKey !== void 0) {
    const cached = params.uploadCache.get(cacheKey);
    if (cached != null) return cached;
  }
  const input = typeof media === "string" ? { type: "auto", file: media } : media;
  const normalized = await _normalizeInputMedia(client, input, {
    uploadPeer: peer,
    abortSignal: params.abortSignal,
    progressCallback: params.progressCallback?.bind(null, id)
  }, true);
  if (normalized._ !== "inputMediaPhoto" && normalized._ !== "inputMediaDocument") {
    throw new MtTypeAssertionError("sendRichMessage (upload)", "inputMediaDocument | inputMediaPhoto", normalized._);
  }
  if (cacheKey !== void 0) params.uploadCache.set(cacheKey, normalized);
  return normalized;
}
async function _normalizeInputRichMessage(client, peer, input, params) {
  if ("_" in input) return input;
  if (input.type !== "blocks") {
    let richFiles;
    if (input.attachments) {
      const medias = [...Object.entries(input.attachments)];
      const normalized = await parallelMap(
        medias,
        async ([id, media]) => [id, await _uploadRichMedia(client, peer, id, media, params)],
        { limit: 4, signal: params.abortSignal }
      );
      richFiles = normalized.map(
        ([id, media]) => media._ === "inputMediaDocument" ? { _: "inputRichFileDocument", id, document: media.id } : { _: "inputRichFilePhoto", id, photo: media.id }
      );
    }
    return input.type === "markdown" ? {
      _: "inputRichMessageMarkdown",
      rtl: input.rtl,
      noautolink: input.skipEntityDetection,
      files: richFiles,
      markdown: input.content
    } : {
      _: "inputRichMessageHTML",
      rtl: input.rtl,
      noautolink: input.skipEntityDetection,
      files: richFiles,
      html: input.content
    };
  }
  const photos = [];
  const documents = [];
  const fileBlocks = [];
  function processBlock(node) {
    if ("file" in node) {
      const block = processBlock(node.block);
      fileBlocks.push({ block, media: node.file });
      return block;
    }
    switch (node._) {
      case "pageBlockCover":
        return { ...node, cover: processBlock(node.cover) };
      case "pageBlockCollage":
      case "pageBlockSlideshow":
        return { ...node, items: node.items.map(processBlock) };
      case "pageBlockDetails":
      case "pageBlockEmbedPost":
      case "pageBlockBlockquoteBlocks":
        return { ...node, blocks: node.blocks.map(processBlock) };
      case "pageBlockList":
        return {
          ...node,
          items: node.items.map(
            (item) => item._ === "pageListItemBlocks" ? {
              ...item,
              blocks: item.blocks.map(processBlock)
            } : item
          )
        };
      case "pageBlockOrderedList":
        return {
          ...node,
          items: node.items.map(
            (item) => item._ === "pageListOrderedItemBlocks" ? { ...item, blocks: item.blocks.map(processBlock) } : item
          )
        };
      default:
        return node;
    }
  }
  const blocks = input.blocks.map(processBlock);
  const uploaded = await parallelMap(
    fileBlocks,
    async ({ block, media }, idx) => {
      const uploaded2 = await _uploadRichMedia(client, peer, `block-${idx}`, media, params);
      return { block, media: uploaded2 };
    },
    { limit: 4, signal: params.abortSignal }
  );
  for (const { block, media } of uploaded) {
    if (media._ === "inputMediaPhoto") {
      const photo = media.id;
      assertTypeIs("sendRichMessage (upload)", photo, "inputPhoto");
      photos.push(photo);
      switch (block._) {
        case "pageBlockPhoto":
          block.photoId = photo.id;
          break;
        case "pageBlockEmbed":
          block.posterPhotoId = photo.id;
          break;
        case "pageBlockEmbedPost":
          block.authorPhotoId = photo.id;
          break;
        default:
          throw new MtArgumentError(`sendRichMessage: expected photo ${photo.id} to be used in a photo block, but got ${block._}`);
      }
    } else {
      const document = media.id;
      assertTypeIs("sendRichMessage (upload)", document, "inputDocument");
      documents.push(document);
      switch (block._) {
        case "pageBlockVideo":
          block.videoId = document.id;
          break;
        case "pageBlockAudio":
          block.audioId = document.id;
          break;
        default:
          throw new MtArgumentError(`sendRichMessage: expected document ${document.id} to be used in a document block, but got ${block._}`);
      }
    }
  }
  return {
    _: "inputRichMessage",
    rtl: input.rtl,
    noautolink: input.skipEntityDetection,
    blocks,
    photos: photos.length ? photos : void 0,
    documents: documents.length ? documents : void 0
  };
}
export {
  _normalizeInputRichMessage
};
