// @mtcute/core/highlevel/methods/messages/search-global.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const searchFilters = require("../../types/messages/search-filters.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchGlobal(client, params) {
  if (!params) params = {};
  const {
    query = "",
    filter = searchFilters.SearchFilters.Empty,
    limit = 100,
    offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset,
    onlyChannels
  } = params;
  const minDate = miscUtils.normalizeDate(params.minDate) ?? 0;
  const maxDate = miscUtils.normalizeDate(params.maxDate) ?? 0;
  const res = await client.call({
    _: "messages.searchGlobal",
    q: query,
    filter,
    minDate,
    maxDate,
    offsetId,
    offsetRate,
    offsetPeer,
    limit,
    broadcastsOnly: onlyChannels
  });
  typeAssertions.assertTypeIsNot("searchGlobal", res, "messages.messagesNotModified");
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new message.Message(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return miscUtils.makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
exports.searchGlobal = searchGlobal;
