// @mtcute/core/highlevel/methods/messages/search-global.js
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { SearchFilters } from "../../types/messages/search-filters.js";
import { normalizeDate, makeArrayPaginated } from "../../utils/misc-utils.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchGlobal(client, params) {
  if (!params) params = {};
  const {
    query = "",
    filter = SearchFilters.Empty,
    limit = 100,
    offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset,
    onlyChannels
  } = params;
  const minDate = normalizeDate(params.minDate) ?? 0;
  const maxDate = normalizeDate(params.maxDate) ?? 0;
  const res = await client.call({
    _: "messages.searchGlobal",
    q: query,
    filter,
    minDate,
    maxDate,
    offsetId,
    offsetRate,
    offsetPeer,
    limit,
    broadcastsOnly: onlyChannels
  });
  assertTypeIsNot("searchGlobal", res, "messages.messagesNotModified");
  const peers = PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new Message$2(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
export {
  searchGlobal
};
