// @mtcute/core/highlevel/methods/messages/search-hashtag.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchHashtag(client, hashtag, params) {
  const { offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset, limit = 100 } = params ?? {};
  const res = await client.call({
    _: "channels.searchPosts",
    hashtag,
    offsetId,
    offsetRate,
    offsetPeer,
    limit
  });
  typeAssertions.assertTypeIsNot("searchHashtag", res, "messages.messagesNotModified");
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new message.Message(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return miscUtils.makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
async function* iterSearchHashtag(client, hashtag, params) {
  if (!params) params = {};
  const { limit = Infinity, chunkSize = 100 } = params;
  let { offset } = params;
  let current = 0;
  for (; ; ) {
    const res = await searchHashtag(client, hashtag, {
      offset,
      limit: Math.min(chunkSize, limit - current)
    });
    if (!res.length) return;
    for (const msg of res) {
      yield msg;
      if (++current >= limit) return;
    }
    if (!res.next) return;
    offset = res.next;
  }
}
exports.iterSearchHashtag = iterSearchHashtag;
exports.searchHashtag = searchHashtag;
