// @mtcute/core/highlevel/methods/messages/search-messages.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const searchFilters = require("../../types/messages/search-filters.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
async function searchMessages(client, params) {
  if (!params) params = {};
  const {
    query = "",
    chatId = { _: "inputPeerEmpty" },
    offset = 0,
    addOffset = 0,
    minId = 0,
    maxId = 0,
    threadId,
    limit = 100,
    filter = searchFilters.SearchFilters.Empty
  } = params;
  const minDate = miscUtils.normalizeDate(params.minDate) ?? 0;
  const maxDate = miscUtils.normalizeDate(params.maxDate) ?? 0;
  const peer = await resolvePeer.resolvePeer(client, chatId);
  const fromUser = params.fromUser ? await resolvePeer.resolvePeer(client, params.fromUser) : void 0;
  const res = await client.call({
    _: "messages.search",
    peer,
    q: query,
    filter,
    minDate,
    maxDate,
    offsetId: offset,
    addOffset,
    limit,
    minId,
    maxId,
    fromId: fromUser,
    topMsgId: threadId,
    hash: Long.ZERO
  });
  typeAssertions.assertTypeIsNot("searchMessages", res, "messages.messagesNotModified");
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new message.Message(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? last.id : void 0;
  return miscUtils.makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
exports.searchMessages = searchMessages;
