// @mtcute/core/highlevel/methods/messages/search-posts-global.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const message = require("../../types/messages/message.cjs");
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchPostsGlobal(client, query, params) {
  const {
    offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset,
    limit = 100,
    payStars
  } = params ?? {};
  const res = await client.call({
    _: "channels.searchPosts",
    query,
    allowPaidStars: payStars,
    offsetId,
    offsetRate,
    offsetPeer,
    limit
  });
  typeAssertions.assertTypeIsNot("searchHashtag", res, "messages.messagesNotModified");
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new message.Message(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return miscUtils.makeArrayPaginatedWithMeta(
    msgs,
    res.count ?? msgs.length,
    { limits: res.searchFlood },
    next
  );
}
exports.searchPostsGlobal = searchPostsGlobal;
