// @mtcute/core/highlevel/methods/messages/search-messages.js
import Long from "long";
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { normalizeDate, makeArrayPaginated } from "../../utils/misc-utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { SearchFilters } from "../../types/messages/search-filters.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
async function searchMessages(client, params) {
  if (!params) params = {};
  const {
    query = "",
    chatId = { _: "inputPeerEmpty" },
    offset = 0,
    addOffset = 0,
    minId = 0,
    maxId = 0,
    threadId,
    limit = 100,
    filter = SearchFilters.Empty
  } = params;
  const minDate = normalizeDate(params.minDate) ?? 0;
  const maxDate = normalizeDate(params.maxDate) ?? 0;
  const peer = await resolvePeer(client, chatId);
  const fromUser = params.fromUser ? await resolvePeer(client, params.fromUser) : void 0;
  const res = await client.call({
    _: "messages.search",
    peer,
    q: query,
    filter,
    minDate,
    maxDate,
    offsetId: offset,
    addOffset,
    limit,
    minId,
    maxId,
    fromId: fromUser,
    topMsgId: threadId,
    hash: Long.ZERO
  });
  assertTypeIsNot("searchMessages", res, "messages.messagesNotModified");
  const peers = PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new Message$2(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? last.id : void 0;
  return makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
export {
  searchMessages
};
