// @mtcute/core/highlevel/methods/messages/search-posts-global.js
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { makeArrayPaginatedWithMeta } from "../../utils/misc-utils.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchPostsGlobal(client, query, params) {
  const {
    offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset,
    limit = 100,
    payStars
  } = params ?? {};
  const res = await client.call({
    _: "channels.searchPosts",
    query,
    allowPaidStars: payStars,
    offsetId,
    offsetRate,
    offsetPeer,
    limit
  });
  assertTypeIsNot("searchHashtag", res, "messages.messagesNotModified");
  const peers = PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new Message$2(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return makeArrayPaginatedWithMeta(
    msgs,
    res.count ?? msgs.length,
    { limits: res.searchFlood },
    next
  );
}
export {
  searchPostsGlobal
};
