// @mtcute/core/highlevel/methods/messages/search-hashtag.js
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { makeArrayPaginated } from "../../utils/misc-utils.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { Message as Message$2 } from "../../types/messages/message.js";
const defaultOffset = {
  rate: 0,
  peer: { _: "inputPeerEmpty" },
  id: 0
};
async function searchHashtag(client, hashtag, params) {
  const { offset: { rate: offsetRate, peer: offsetPeer, id: offsetId } = defaultOffset, limit = 100 } = params ?? {};
  const res = await client.call({
    _: "channels.searchPosts",
    hashtag,
    offsetId,
    offsetRate,
    offsetPeer,
    limit
  });
  assertTypeIsNot("searchHashtag", res, "messages.messagesNotModified");
  const peers = PeersIndex.from(res);
  const msgs = res.messages.filter((msg) => msg._ !== "messageEmpty").map((msg) => new Message$2(msg, peers));
  const last = msgs[msgs.length - 1];
  const next = last ? {
    rate: res.nextRate ?? last.raw.date,
    peer: last.chat.inputPeer,
    id: last.id
  } : void 0;
  return makeArrayPaginated(msgs, res.count ?? msgs.length, next);
}
async function* iterSearchHashtag(client, hashtag, params) {
  if (!params) params = {};
  const { limit = Infinity, chunkSize = 100 } = params;
  let { offset } = params;
  let current = 0;
  for (; ; ) {
    const res = await searchHashtag(client, hashtag, {
      offset,
      limit: Math.min(chunkSize, limit - current)
    });
    if (!res.length) return;
    for (const msg of res) {
      yield msg;
      if (++current >= limit) return;
    }
    if (!res.next) return;
    offset = res.next;
  }
}
export {
  iterSearchHashtag,
  searchHashtag
};
