// @mtcute/core/highlevel/methods/messages/send-rich-message.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
const normalizeRichMessage = require("./normalize-rich-message.cjs");
const sendCommon = require("./send-common.cjs");
const factories = require("../../types/bots/keyboards/factories.cjs");
async function sendRichMessage(client, chatId, params) {
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await sendCommon._processCommonSendParameters(
    client,
    chatId,
    params
  );
  const richMessage = await normalizeRichMessage._normalizeInputRichMessage(client, peer, params.content, params);
  const randomId = params.randomId ?? longUtils.randomLong();
  const res = await client.call(
    {
      _: "messages.sendMessage",
      peer,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup: factories._convertToTl(params.replyMarkup),
      message: "",
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer.resolvePeer(client, params.sendAs) : void 0,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages,
      richMessage
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  return findInUpdate._findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
}
exports.sendRichMessage = sendRichMessage;
