// @mtcute/core/highlevel/methods/messages/send-text.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const longUtils = require("../../../utils/long-utils.cjs");
const peerUtils$1 = require("../../../utils/peer-utils.cjs");
const message = require("../../types/messages/message.cjs");
const utils = require("../../updates/utils.cjs");
const peerUtils = require("../../utils/peer-utils.cjs");
const batchedQueries = require("../chats/batched-queries.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
const sendCommon = require("./send-common.cjs");
const factories = require("../../types/bots/keyboards/factories.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
async function sendText(client, chatId, text, params) {
  if (!params) params = {};
  const [message$1, entities] = await normalizeText._normalizeInputText(client, text);
  const replyMarkup = factories._convertToTl(params.replyMarkup);
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await sendCommon._processCommonSendParameters(
    client,
    chatId,
    params
  );
  const randomId = params.randomId ?? longUtils.randomLong();
  const res = await client.call(
    {
      _: "messages.sendMessage",
      peer,
      noWebpage: params.disableWebPreview,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup,
      message: message$1,
      entities,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer.resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invertMedia,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  if (res._ === "updateShortSentMessage") {
    const msg2 = {
      _: "message",
      id: res.id,
      peerId: peerUtils.inputPeerToPeer(peer),
      fromId: { _: "peerUser", userId: client.storage.self.getCached().userId },
      message: message$1,
      date: res.date,
      out: res.out,
      replyMarkup,
      entities: res.entities
    };
    if (!params.shouldDispatch) {
      client.handleClientUpdate(utils.createDummyUpdate(res.pts, res.ptsCount));
    }
    const peers = new peersIndex.PeersIndex();
    const fetchPeer = async (peer2) => {
      const id = peerUtils$1.getMarkedPeerId(peer2);
      let cached = await client.storage.peers.getCompleteById(id);
      if (!cached) {
        cached = await batchedQueries._getRawPeerBatched(client, await resolvePeer.resolvePeer(client, peer2));
      }
      if (!cached) {
        throw new errors.MtTypeAssertionError("sendText (@ getFullPeerById)", "user | chat", "null");
      }
      switch (cached._) {
        case "user":
          peers.users.set(cached.id, cached);
          break;
        case "chat":
        case "chatForbidden":
        case "channel":
        case "channelForbidden":
          peers.chats.set(cached.id, cached);
          break;
        default:
          throw new errors.MtTypeAssertionError(
            "sendText (@ users.getUsers)",
            "user | chat | channel",
            // not very accurate, but good enough
            cached._
          );
      }
    };
    await Promise.all([fetchPeer(peer), fetchPeer(msg2.fromId)]);
    const ret = new message.Message(msg2, peers);
    return ret;
  }
  const msg = findInUpdate._findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
  return msg;
}
exports.sendText = sendText;
