// @mtcute/core/highlevel/methods/messages/send-common.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const peerUtils = require("../../../utils/peer-utils.cjs");
const errors$1 = require("../../types/errors.cjs");
const chainId = require("../misc/chain-id.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const getDiscussionMessage = require("./get-discussion-message.cjs");
const getMessages = require("./get-messages.cjs");
const setTyping = require("./set-typing.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
function _normalizeQuickReplyShortcut(shortcut) {
  if (!shortcut) return void 0;
  if (typeof shortcut === "number") {
    return {
      _: "inputQuickReplyShortcutId",
      shortcutId: shortcut
    };
  }
  return {
    _: "inputQuickReplyShortcut",
    shortcut
  };
}
async function _processCommonSendParameters(client, chatId, params) {
  let peer = await resolvePeer.resolvePeer(client, chatId);
  await client.timers.cancel(setTyping._getTypingTimerId(peer, params.businessConnectionId));
  let replyTo = miscUtils.normalizeMessageId(params.replyTo);
  const replyToPeer = typeof params.replyTo === "number" ? void 0 : params.replyTo?.chat.inputPeer;
  if (params.commentTo) {
    [peer, replyTo] = await getDiscussionMessage._getDiscussionMessage(client, peer, miscUtils.normalizeMessageId(params.commentTo));
  }
  if (params.mustReply) {
    if (!replyTo) {
      throw new errors.MtArgumentError("mustReply used, but replyTo was not passed");
    }
    const msg = await getMessages.getMessages(client, peer, replyTo);
    if (!msg) {
      throw new errors$1.MtMessageNotFoundError(peerUtils.getMarkedPeerId(peer), replyTo, "to reply to");
    }
  }
  if (params.replyToStory && replyTo) {
    throw new errors.MtArgumentError("replyTo/commentTo and replyToStory cannot be used together");
  }
  let tlReplyTo;
  if (replyTo) {
    tlReplyTo = {
      _: "inputReplyToMessage",
      replyToMsgId: replyTo,
      replyToPeerId: replyToPeer,
      quoteText: params.quote?.text,
      quoteEntities: params.quote?.entities,
      quoteOffset: params.quoteOffset,
      monoforumPeerId: params.toMonoforumPeer ? await resolvePeer.resolvePeer(client, params.toMonoforumPeer) : void 0,
      todoItemId: params.replyToTodoItem,
      pollOption: params.replyToPollOption
    };
  } else if (params.toMonoforumPeer) {
    tlReplyTo = {
      _: "inputReplyToMonoForum",
      monoforumPeerId: await resolvePeer.resolvePeer(client, params.toMonoforumPeer)
    };
  } else if (params.replyToStory) {
    tlReplyTo = {
      _: "inputReplyToStory",
      storyId: params.replyToStory,
      peer
    };
  }
  let scheduleDate;
  if (params.schedule === "online") {
    scheduleDate = 2147483646;
  } else if (params.schedule) {
    scheduleDate = miscUtils.normalizeDate(params.schedule);
  }
  return {
    peer,
    replyTo: tlReplyTo,
    scheduleDate,
    quickReplyShortcut: _normalizeQuickReplyShortcut(params.quickReply),
    chainId: chainId._getPeerChainId(client, peer, "send")
  };
}
exports._normalizeQuickReplyShortcut = _normalizeQuickReplyShortcut;
exports._processCommonSendParameters = _processCommonSendParameters;
