// @mtcute/core/highlevel/methods/messages/send-media.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const normalizeInputMedia = require("../files/normalize-input-media.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const findInUpdate = require("./find-in-update.cjs");
const sendCommon = require("./send-common.cjs");
const factories = require("../../types/bots/keyboards/factories.cjs");
async function sendMedia(client, chatId, media, params) {
  if (!params) params = {};
  if (typeof media === "string") {
    media = {
      type: "auto",
      file: media
    };
  }
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await sendCommon._processCommonSendParameters(
    client,
    chatId,
    params
  );
  const inputMedia = await normalizeInputMedia._normalizeInputMedia(client, media, {
    progressCallback: params.progressCallback,
    uploadPeer: peer,
    abortSignal: params.abortSignal
  });
  const [message, entities] = await normalizeText._normalizeInputText(
    client,
    // some types dont have `caption` field, and ts warns us,
    // but since it's JS, they'll just be `undefined` and properly handled by the method
    params.caption || media.caption
  );
  const replyMarkup = factories._convertToTl(params.replyMarkup);
  const randomId = params.randomId ?? longUtils.randomLong();
  const res = await client.call(
    {
      _: "messages.sendMedia",
      peer,
      media: inputMedia,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup,
      message,
      entities,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer.resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invert,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  const msg = findInUpdate._findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
  return msg;
}
exports.sendMedia = sendMedia;
