// @mtcute/core/highlevel/methods/messages/send-typing.js
import { assertNever } from "../../../types/utils.js";
import { assertTrue } from "../../../utils/type-assertions.js";
import { resolvePeer } from "../users/resolve-peer.js";
function _mapTypingStatus(status, progress = 0) {
  switch (status) {
    case "typing":
      return { _: "sendMessageTypingAction" };
    case "cancel":
      return { _: "sendMessageCancelAction" };
    case "record_video":
      return { _: "sendMessageRecordVideoAction" };
    case "upload_video":
      return { _: "sendMessageUploadVideoAction", progress };
    case "record_voice":
      return { _: "sendMessageRecordAudioAction" };
    case "upload_voice":
      return { _: "sendMessageUploadAudioAction", progress };
    case "upload_photo":
      return { _: "sendMessageUploadPhotoAction", progress };
    case "upload_document":
      return { _: "sendMessageUploadDocumentAction", progress };
    case "geo":
      return { _: "sendMessageGeoLocationAction" };
    case "contact":
      return { _: "sendMessageChooseContactAction" };
    case "game":
      return { _: "sendMessageGamePlayAction" };
    case "record_round":
      return { _: "sendMessageRecordRoundAction" };
    case "upload_round":
      return { _: "sendMessageUploadRoundAction", progress };
    case "speak_call":
      return { _: "speakingInGroupCallAction" };
    case "history_import":
      return { _: "sendMessageHistoryImportAction", progress };
    case "sticker":
      return { _: "sendMessageChooseStickerAction" };
    default:
      assertNever();
  }
}
async function sendTyping(client, chatId, status = "typing", params) {
  if (typeof status === "string") {
    status = _mapTypingStatus(status, params?.progress ?? 0);
  }
  const r = await client.call({
    _: "messages.setTyping",
    peer: await resolvePeer(client, chatId),
    action: status,
    topMsgId: params?.threadId
  }, {
    businessConnectionId: params?.businessConnectionId
  });
  assertTrue("messages.setTyping", r);
}
export {
  _mapTypingStatus,
  sendTyping
};
