// @mtcute/core/highlevel/methods/messages/send-rich-message.js
import { randomLong } from "../../../utils/long-utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findMessageInUpdate } from "./find-in-update.js";
import { _normalizeInputRichMessage } from "./normalize-rich-message.js";
import { _processCommonSendParameters } from "./send-common.js";
import { _convertToTl } from "../../types/bots/keyboards/factories.js";
async function sendRichMessage(client, chatId, params) {
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await _processCommonSendParameters(
    client,
    chatId,
    params
  );
  const richMessage = await _normalizeInputRichMessage(client, peer, params.content, params);
  const randomId = params.randomId ?? randomLong();
  const res = await client.call(
    {
      _: "messages.sendMessage",
      peer,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup: _convertToTl(params.replyMarkup),
      message: "",
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer(client, params.sendAs) : void 0,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages,
      richMessage
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  return _findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
}
export {
  sendRichMessage
};
