// @mtcute/core/highlevel/methods/messages/send-text.js
import { MtTypeAssertionError } from "../../../types/errors.js";
import { randomLong } from "../../../utils/long-utils.js";
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { Message as Message$2 } from "../../types/messages/message.js";
import { createDummyUpdate } from "../../updates/utils.js";
import { inputPeerToPeer } from "../../utils/peer-utils.js";
import { _getRawPeerBatched } from "../chats/batched-queries.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findMessageInUpdate } from "./find-in-update.js";
import { _processCommonSendParameters } from "./send-common.js";
import { _convertToTl } from "../../types/bots/keyboards/factories.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
async function sendText(client, chatId, text, params) {
  if (!params) params = {};
  const [message, entities] = await _normalizeInputText(client, text);
  const replyMarkup = _convertToTl(params.replyMarkup);
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await _processCommonSendParameters(
    client,
    chatId,
    params
  );
  const randomId = params.randomId ?? randomLong();
  const res = await client.call(
    {
      _: "messages.sendMessage",
      peer,
      noWebpage: params.disableWebPreview,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup,
      message,
      entities,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invertMedia,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  if (res._ === "updateShortSentMessage") {
    const msg2 = {
      _: "message",
      id: res.id,
      peerId: inputPeerToPeer(peer),
      fromId: { _: "peerUser", userId: client.storage.self.getCached().userId },
      message,
      date: res.date,
      out: res.out,
      replyMarkup,
      entities: res.entities
    };
    if (!params.shouldDispatch) {
      client.handleClientUpdate(createDummyUpdate(res.pts, res.ptsCount));
    }
    const peers = new PeersIndex();
    const fetchPeer = async (peer2) => {
      const id = getMarkedPeerId(peer2);
      let cached = await client.storage.peers.getCompleteById(id);
      if (!cached) {
        cached = await _getRawPeerBatched(client, await resolvePeer(client, peer2));
      }
      if (!cached) {
        throw new MtTypeAssertionError("sendText (@ getFullPeerById)", "user | chat", "null");
      }
      switch (cached._) {
        case "user":
          peers.users.set(cached.id, cached);
          break;
        case "chat":
        case "chatForbidden":
        case "channel":
        case "channelForbidden":
          peers.chats.set(cached.id, cached);
          break;
        default:
          throw new MtTypeAssertionError(
            "sendText (@ users.getUsers)",
            "user | chat | channel",
            // not very accurate, but good enough
            cached._
          );
      }
    };
    await Promise.all([fetchPeer(peer), fetchPeer(msg2.fromId)]);
    const ret = new Message$2(msg2, peers);
    return ret;
  }
  const msg = _findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
  return msg;
}
export {
  sendText
};
