// @mtcute/core/highlevel/methods/messages/send-media-group.js
import { randomLong } from "../../../utils/long-utils.js";
import { Message as Message$2 } from "../../types/messages/message.js";
import { assertIsUpdatesGroup } from "../../updates/utils.js";
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _processCommonSendParameters } from "./send-common.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
async function sendMediaGroup(client, chatId, medias, params) {
  if (!params) params = {};
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await _processCommonSendParameters(
    client,
    chatId,
    params
  );
  const multiMedia = [];
  for (let i = 0; i < medias.length; i++) {
    let media = medias[i];
    if (typeof media === "string") {
      media = {
        type: "auto",
        file: media
      };
    }
    const inputMedia = await _normalizeInputMedia(
      client,
      media,
      {
        progressCallback: params.progressCallback?.bind(null, i),
        // i have no fucking clue why we should upload it manually,
        // but otherwise Telegram throws MEDIA_INVALID
        // fuck my life
        uploadPeer: peer,
        businessConnectionId: params.businessConnectionId,
        abortSignal: params.abortSignal
      },
      true
    );
    const [message, entities] = await _normalizeInputText(
      client,
      // some types dont have `caption` field, and ts warns us,
      // but since it's JS, they'll just be `undefined` and properly handled by the method
      media.caption
    );
    multiMedia.push({
      _: "inputSingleMedia",
      randomId: randomLong(),
      media: inputMedia,
      message,
      entities
    });
  }
  const res = await client.call(
    {
      _: "messages.sendMultiMedia",
      peer,
      multiMedia,
      silent: params.silent,
      replyTo,
      scheduleDate,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invertMedia,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  assertIsUpdatesGroup("sendMediaGroup", res);
  client.handleClientUpdate(res);
  const peers = PeersIndex.from(res);
  const msgs = res.updates.filter(
    (u) => u._ === "updateNewMessage" || u._ === "updateNewChannelMessage" || u._ === "updateNewScheduledMessage"
  ).map((u) => new Message$2(u.message, peers, u._ === "updateNewScheduledMessage"));
  return msgs;
}
export {
  sendMediaGroup
};
