// @mtcute/core/highlevel/methods/messages/send-media-group.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const message = require("../../types/messages/message.cjs");
const utils = require("../../updates/utils.cjs");
const normalizeInputMedia = require("../files/normalize-input-media.cjs");
const normalizeText = require("../misc/normalize-text.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const sendCommon = require("./send-common.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
async function sendMediaGroup(client, chatId, medias, params) {
  if (!params) params = {};
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await sendCommon._processCommonSendParameters(
    client,
    chatId,
    params
  );
  const multiMedia = [];
  for (let i = 0; i < medias.length; i++) {
    let media = medias[i];
    if (typeof media === "string") {
      media = {
        type: "auto",
        file: media
      };
    }
    const inputMedia = await normalizeInputMedia._normalizeInputMedia(
      client,
      media,
      {
        progressCallback: params.progressCallback?.bind(null, i),
        // i have no fucking clue why we should upload it manually,
        // but otherwise Telegram throws MEDIA_INVALID
        // fuck my life
        uploadPeer: peer,
        businessConnectionId: params.businessConnectionId,
        abortSignal: params.abortSignal
      },
      true
    );
    const [message2, entities] = await normalizeText._normalizeInputText(
      client,
      // some types dont have `caption` field, and ts warns us,
      // but since it's JS, they'll just be `undefined` and properly handled by the method
      media.caption
    );
    multiMedia.push({
      _: "inputSingleMedia",
      randomId: longUtils.randomLong(),
      media: inputMedia,
      message: message2,
      entities
    });
  }
  const res = await client.call(
    {
      _: "messages.sendMultiMedia",
      peer,
      multiMedia,
      silent: params.silent,
      replyTo,
      scheduleDate,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer.resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invertMedia,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  utils.assertIsUpdatesGroup("sendMediaGroup", res);
  client.handleClientUpdate(res);
  const peers = peersIndex.PeersIndex.from(res);
  const msgs = res.updates.filter(
    (u) => u._ === "updateNewMessage" || u._ === "updateNewChannelMessage" || u._ === "updateNewScheduledMessage"
  ).map((u) => new message.Message(u.message, peers, u._ === "updateNewScheduledMessage"));
  return msgs;
}
exports.sendMediaGroup = sendMediaGroup;
