// @mtcute/core/highlevel/methods/messages/send-common.js
import { MtArgumentError } from "../../../types/errors.js";
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { MtMessageNotFoundError } from "../../types/errors.js";
import { _getPeerChainId } from "../misc/chain-id.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _getDiscussionMessage } from "./get-discussion-message.js";
import { getMessages } from "./get-messages.js";
import { _getTypingTimerId } from "./set-typing.js";
import { normalizeMessageId, normalizeDate } from "../../utils/misc-utils.js";
function _normalizeQuickReplyShortcut(shortcut) {
  if (!shortcut) return void 0;
  if (typeof shortcut === "number") {
    return {
      _: "inputQuickReplyShortcutId",
      shortcutId: shortcut
    };
  }
  return {
    _: "inputQuickReplyShortcut",
    shortcut
  };
}
async function _processCommonSendParameters(client, chatId, params) {
  let peer = await resolvePeer(client, chatId);
  await client.timers.cancel(_getTypingTimerId(peer, params.businessConnectionId));
  let replyTo = normalizeMessageId(params.replyTo);
  const replyToPeer = typeof params.replyTo === "number" ? void 0 : params.replyTo?.chat.inputPeer;
  if (params.commentTo) {
    [peer, replyTo] = await _getDiscussionMessage(client, peer, normalizeMessageId(params.commentTo));
  }
  if (params.mustReply) {
    if (!replyTo) {
      throw new MtArgumentError("mustReply used, but replyTo was not passed");
    }
    const msg = await getMessages(client, peer, replyTo);
    if (!msg) {
      throw new MtMessageNotFoundError(getMarkedPeerId(peer), replyTo, "to reply to");
    }
  }
  if (params.replyToStory && replyTo) {
    throw new MtArgumentError("replyTo/commentTo and replyToStory cannot be used together");
  }
  let tlReplyTo;
  if (replyTo) {
    tlReplyTo = {
      _: "inputReplyToMessage",
      replyToMsgId: replyTo,
      replyToPeerId: replyToPeer,
      quoteText: params.quote?.text,
      quoteEntities: params.quote?.entities,
      quoteOffset: params.quoteOffset,
      monoforumPeerId: params.toMonoforumPeer ? await resolvePeer(client, params.toMonoforumPeer) : void 0,
      todoItemId: params.replyToTodoItem,
      pollOption: params.replyToPollOption
    };
  } else if (params.toMonoforumPeer) {
    tlReplyTo = {
      _: "inputReplyToMonoForum",
      monoforumPeerId: await resolvePeer(client, params.toMonoforumPeer)
    };
  } else if (params.replyToStory) {
    tlReplyTo = {
      _: "inputReplyToStory",
      storyId: params.replyToStory,
      peer
    };
  }
  let scheduleDate;
  if (params.schedule === "online") {
    scheduleDate = 2147483646;
  } else if (params.schedule) {
    scheduleDate = normalizeDate(params.schedule);
  }
  return {
    peer,
    replyTo: tlReplyTo,
    scheduleDate,
    quickReplyShortcut: _normalizeQuickReplyShortcut(params.quickReply),
    chainId: _getPeerChainId(client, peer, "send")
  };
}
export {
  _normalizeQuickReplyShortcut,
  _processCommonSendParameters
};
