// @mtcute/core/highlevel/methods/messages/send-media.js
import { randomLong } from "../../../utils/long-utils.js";
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
import { _normalizeInputText } from "../misc/normalize-text.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _findMessageInUpdate } from "./find-in-update.js";
import { _processCommonSendParameters } from "./send-common.js";
import { _convertToTl } from "../../types/bots/keyboards/factories.js";
async function sendMedia(client, chatId, media, params) {
  if (!params) params = {};
  if (typeof media === "string") {
    media = {
      type: "auto",
      file: media
    };
  }
  const { peer, replyTo, scheduleDate, chainId, quickReplyShortcut } = await _processCommonSendParameters(
    client,
    chatId,
    params
  );
  const inputMedia = await _normalizeInputMedia(client, media, {
    progressCallback: params.progressCallback,
    uploadPeer: peer,
    abortSignal: params.abortSignal
  });
  const [message, entities] = await _normalizeInputText(
    client,
    // some types dont have `caption` field, and ts warns us,
    // but since it's JS, they'll just be `undefined` and properly handled by the method
    params.caption || media.caption
  );
  const replyMarkup = _convertToTl(params.replyMarkup);
  const randomId = params.randomId ?? randomLong();
  const res = await client.call(
    {
      _: "messages.sendMedia",
      peer,
      media: inputMedia,
      silent: params.silent,
      replyTo,
      randomId,
      scheduleDate,
      replyMarkup,
      message,
      entities,
      clearDraft: params.clearDraft,
      noforwards: params.forbidForwards,
      sendAs: params.sendAs ? await resolvePeer(client, params.sendAs) : void 0,
      invertMedia: params.invert,
      quickReplyShortcut,
      effect: params.effect,
      allowPaidFloodskip: params.allowPaidFloodskip,
      allowPaidStars: params.allowPaidMessages
    },
    {
      chainId,
      abortSignal: params.abortSignal,
      businessConnectionId: params.businessConnectionId
    }
  );
  const msg = _findMessageInUpdate(client, res, false, !params.shouldDispatch, false, randomId);
  return msg;
}
export {
  sendMedia
};
