// @mtcute/core/highlevel/methods/messages/set-typing.js
import { getMarkedPeerId } from "../../../utils/peer-utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { _mapTypingStatus } from "./send-typing.js";
function _getTypingTimerId(peer, businessId) {
  let base = `typing:${peer._ === "inputPeerSelf" ? "self" : getMarkedPeerId(peer)}`;
  if (businessId) base += `:b${businessId}`;
  return base;
}
const TIMER_INTERVAL = 5e3;
async function setTyping(client, params) {
  const {
    peerId,
    businessConnectionId,
    threadId
  } = params;
  let status = params.status ?? "typing";
  if (typeof status === "string") status = _mapTypingStatus(status);
  const peer = await resolvePeer(client, peerId);
  const timerId = _getTypingTimerId(peer, businessConnectionId);
  await client.timers.cancel(timerId);
  if (status._ === "sendMessageCancelAction") {
    await client.call({
      _: "messages.setTyping",
      peer,
      action: status,
      topMsgId: threadId
    }, {
      businessConnectionId
    });
    return;
  }
  await client.timers.upsert({
    kind: "rpc",
    key: timerId,
    interval: TIMER_INTERVAL,
    startNow: true,
    options: { businessConnectionId },
    request: {
      _: "messages.setTyping",
      peer,
      action: status,
      topMsgId: threadId
    }
  });
}
export {
  _getTypingTimerId,
  setTyping
};
