// @mtcute/core/highlevel/methods/messages/set-typing.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../utils/peer-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const sendTyping = require("./send-typing.cjs");
function _getTypingTimerId(peer, businessId) {
  let base = `typing:${peer._ === "inputPeerSelf" ? "self" : peerUtils.getMarkedPeerId(peer)}`;
  if (businessId) base += `:b${businessId}`;
  return base;
}
const TIMER_INTERVAL = 5e3;
async function setTyping(client, params) {
  const {
    peerId,
    businessConnectionId,
    threadId
  } = params;
  let status = params.status ?? "typing";
  if (typeof status === "string") status = sendTyping._mapTypingStatus(status);
  const peer = await resolvePeer.resolvePeer(client, peerId);
  const timerId = _getTypingTimerId(peer, businessConnectionId);
  await client.timers.cancel(timerId);
  if (status._ === "sendMessageCancelAction") {
    await client.call({
      _: "messages.setTyping",
      peer,
      action: status,
      topMsgId: threadId
    }, {
      businessConnectionId
    });
    return;
  }
  await client.timers.upsert({
    kind: "rpc",
    key: timerId,
    interval: TIMER_INTERVAL,
    startNow: true,
    options: { businessConnectionId },
    request: {
      _: "messages.setTyping",
      peer,
      action: status,
      topMsgId: threadId
    }
  });
}
exports._getTypingTimerId = _getTypingTimerId;
exports.setTyping = setTyping;
