// @mtcute/core/utils/binary/compat.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const Long = require("long");
const peersIndex = require("../../highlevel/types/peers/peers-index.cjs");
const reader = require("../../tl/binary/reader.cjs");
const reader$1 = require("../../tl/compat/reader.cjs");
const innerTl = require("../../tl/inner-tl.cjs");
function replaceType(obj, type) {
  return Object.assign(obj, { _: type });
}
function dropFields(obj, fields) {
  for (let i = 0; i < fields.length; i++) {
    delete obj[fields[i]];
  }
  return obj;
}
function mapCompatStarGiftAttribute(obj) {
  switch (obj._) {
    case "starGiftAttributeBackdrop_layer221":
      return {
        ...dropFields(obj, ["rarityPermille"]),
        _: "starGiftAttributeBackdrop",
        rarity: { _: "starGiftAttributeRarity", permille: obj.rarityPermille }
      };
    case "starGiftAttributePattern_layer221":
      return {
        ...dropFields(obj, ["rarityPermille"]),
        _: "starGiftAttributePattern",
        rarity: { _: "starGiftAttributeRarity", permille: obj.rarityPermille }
      };
    case "starGiftAttributeModel_layer221":
      return {
        ...dropFields(obj, ["rarityPermille"]),
        _: "starGiftAttributeModel",
        rarity: { _: "starGiftAttributeRarity", permille: obj.rarityPermille }
      };
    default:
      return obj;
  }
}
function mapCompatStarGift(obj) {
  switch (obj._) {
    case "starGiftUnique_layer197":
      return {
        ...obj,
        _: "starGiftUnique",
        giftId: Long.ZERO,
        ownerId: obj.ownerId ? { _: "peerUser", userId: obj.ownerId } : void 0,
        attributes: obj.attributes.map(mapCompatStarGiftAttribute)
      };
    case "starGiftUnique_layer198":
    case "starGiftUnique_layer202":
    case "starGiftUnique_layer206":
      return {
        ...obj,
        _: "starGiftUnique",
        giftId: Long.ZERO,
        attributes: obj.attributes.map(mapCompatStarGiftAttribute)
      };
    case "starGiftUnique_layer210":
      return {
        ...obj,
        _: "starGiftUnique",
        giftId: Long.ZERO,
        attributes: obj.attributes.map(mapCompatStarGiftAttribute),
        resellAmount: obj.resellStars ? [{
          _: "starsAmount",
          amount: obj.resellStars,
          nanos: 0
        }] : void 0
      };
    case "starGiftUnique_layer211":
      return {
        ...obj,
        _: "starGiftUnique",
        giftId: Long.ZERO,
        attributes: obj.attributes.map(mapCompatStarGiftAttribute)
      };
    case "starGiftUnique_layer214":
    case "starGiftUnique_layer218":
    case "starGiftUnique_layer221":
      return {
        ...obj,
        _: "starGiftUnique",
        attributes: obj.attributes.map(mapCompatStarGiftAttribute)
      };
    case "starGift_layer202":
    case "starGift_layer206":
    case "starGift_layer209":
    case "starGift_layer211":
    case "starGift_layer216":
    case "starGift_layer218":
      return replaceType(obj, "starGift");
    default:
      return obj;
  }
}
function mapCompatEmojiStatus(obj) {
  switch (obj._) {
    case "emojiStatus_layer197":
      return {
        ...obj,
        _: "emojiStatus",
        documentId: obj.documentId
      };
    case "emojiStatusUntil_layer214":
      return {
        _: "emojiStatus",
        documentId: obj.documentId,
        until: obj.until
      };
    default:
      return obj;
  }
}
function mapCompatPoll(obj) {
  switch (obj._) {
    case "poll_layer223":
    case "poll_layer224":
      return {
        ...obj,
        _: "poll",
        hash: Long.ZERO,
        // answers are mapped at runtime by wrapped compat readers
        answers: obj.answers
      };
    default:
      return obj;
  }
}
function mapCompatPageListOrderedItem(obj) {
  switch (obj._) {
    case "pageListOrderedItemBlocks_layer225":
      return {
        ...obj,
        _: "pageListOrderedItemBlocks",
        blocks: obj.blocks.map(mapCompatPageBlock)
      };
    case "pageListOrderedItemText_layer225":
      return replaceType(obj, "pageListOrderedItemText");
    default:
      return obj;
  }
}
function mapCompatPageListItem(obj) {
  switch (obj._) {
    case "pageListItemBlocks_layer225":
      return {
        ...obj,
        _: "pageListItemBlocks",
        blocks: obj.blocks.map(mapCompatPageBlock)
      };
    case "pageListItemText_layer225":
      return replaceType(obj, "pageListItemText");
    default:
      return obj;
  }
}
function mapCompatPageBlock(obj) {
  switch (obj._) {
    case "pageBlockOrderedList_layer225":
      return {
        ...obj,
        _: "pageBlockOrderedList",
        items: obj.items.map(mapCompatPageListOrderedItem)
      };
    case "pageBlockList":
      return {
        ...obj,
        items: obj.items.map(mapCompatPageListItem)
      };
    case "pageBlockDetails":
      return {
        ...obj,
        blocks: obj.blocks.map(mapCompatPageBlock)
      };
    case "pageBlockEmbedPost":
      return {
        ...obj,
        blocks: obj.blocks.map(mapCompatPageBlock)
      };
    case "pageBlockCollage":
    case "pageBlockSlideshow":
      return {
        ...obj,
        items: obj.items.map(mapCompatPageBlock)
      };
    case "pageBlockCover":
      return {
        ...obj,
        cover: mapCompatPageBlock(obj.cover)
      };
    default:
      return obj;
  }
}
function mapCompatMessageMedia(obj) {
  switch (obj._) {
    case "messageMediaDocument_layer197":
      return replaceType(obj, "messageMediaDocument");
    case "messageMediaDice_layer220":
      return replaceType(obj, "messageMediaDice");
    case "messageMediaPhoto_layer223":
      return replaceType(obj, "messageMediaPhoto");
    case "messageMediaPoll_layer223":
      return {
        _: "messageMediaPoll",
        poll: mapCompatPoll(obj.poll),
        results: obj.results
      };
    default:
      return obj;
  }
}
function mapCompatMessageAction(obj) {
  switch (obj._) {
    case "messageActionStarGiftUnique_layer197":
    case "messageActionStarGiftUnique_layer202":
    case "messageActionStarGiftUnique_layer214":
    case "messageActionStarGiftUnique_layer221":
      return {
        ...obj,
        _: "messageActionStarGiftUnique",
        gift: mapCompatStarGift(obj.gift)
      };
    case "messageActionStarGiftUnique_layer210":
      return {
        ...obj,
        _: "messageActionStarGiftUnique",
        gift: mapCompatStarGift(obj.gift),
        resaleAmount: obj.resaleStars ? {
          _: "starsAmount",
          amount: obj.resaleStars,
          nanos: 0
        } : void 0
      };
    case "messageActionStarGift_layer197":
    case "messageActionStarGift_layer211":
    case "messageActionStarGift_layer216":
    case "messageActionStarGift_layer218":
      return {
        ...obj,
        _: "messageActionStarGift",
        gift: mapCompatStarGift(obj.gift)
      };
    case "messageActionPaidMessagesPrice_layer203":
      return replaceType(obj, "messageActionPaidMessagesPrice");
    case "messageActionSetChatTheme_layer211":
      return {
        _: "messageActionSetChatTheme",
        theme: {
          _: "chatTheme",
          emoticon: obj.emoticon
        }
      };
    case "messageActionGiftCode_layer216":
      return {
        ...obj,
        _: "messageActionGiftCode",
        days: obj.months / 30
      };
    case "messageActionGiftPremium_layer216":
      return {
        ...obj,
        _: "messageActionGiftPremium",
        days: obj.months / 30
      };
    default:
      return obj;
  }
}
function mapMessageReplyHeader(obj) {
  switch (obj._) {
    case "messageReplyHeader_layer206":
      return {
        ...obj,
        _: "messageReplyHeader",
        replyMedia: obj.replyMedia ? mapCompatMessageMedia(obj.replyMedia) : void 0
      };
    case "messageReplyHeader_layer223":
      return {
        ...obj,
        _: "messageReplyHeader",
        replyMedia: obj.replyMedia ? mapCompatMessageMedia(obj.replyMedia) : void 0
      };
    default:
      return obj;
  }
}
function mapCompatMessage(obj) {
  switch (obj._) {
    case "message_layer199":
    case "message_layer204":
    case "message_layer216":
    case "message_layer220":
    case "message_layer222":
    case "message_layer224":
    case "message_layer225":
      return {
        ...obj,
        _: "message",
        media: obj.media ? mapCompatMessageMedia(obj.media) : void 0,
        replyTo: obj.replyTo ? mapMessageReplyHeader(obj.replyTo) : void 0
      };
    case "messageService_layer204":
      return {
        ...obj,
        _: "messageService",
        action: mapCompatMessageAction(obj.action),
        replyTo: obj.replyTo ? mapMessageReplyHeader(obj.replyTo) : void 0
      };
    default:
      return obj;
  }
}
function mapCompatKeyboardButton(obj) {
  switch (obj._) {
    case "keyboardButton_layer221":
      return replaceType(obj, "keyboardButton");
    case "keyboardButtonUrl_layer221":
      return replaceType(obj, "keyboardButtonUrl");
    case "keyboardButtonCallback_layer221":
      return replaceType(obj, "keyboardButtonCallback");
    case "keyboardButtonRequestPhone_layer221":
      return replaceType(obj, "keyboardButtonRequestPhone");
    case "keyboardButtonRequestGeoLocation_layer221":
      return replaceType(obj, "keyboardButtonRequestGeoLocation");
    case "keyboardButtonSwitchInline_layer221":
      return replaceType(obj, "keyboardButtonSwitchInline");
    case "keyboardButtonGame_layer221":
      return replaceType(obj, "keyboardButtonGame");
    case "keyboardButtonBuy_layer221":
      return replaceType(obj, "keyboardButtonBuy");
    case "keyboardButtonUrlAuth_layer221":
      return replaceType(obj, "keyboardButtonUrlAuth");
    case "inputKeyboardButtonUrlAuth_layer221":
      return replaceType(obj, "inputKeyboardButtonUrlAuth");
    case "keyboardButtonRequestPoll_layer221":
      return replaceType(obj, "keyboardButtonRequestPoll");
    case "inputKeyboardButtonUserProfile_layer221":
      return replaceType(obj, "inputKeyboardButtonUserProfile");
    case "keyboardButtonUserProfile_layer221":
      return replaceType(obj, "keyboardButtonUserProfile");
    case "keyboardButtonWebView_layer221":
      return replaceType(obj, "keyboardButtonWebView");
    case "keyboardButtonSimpleWebView_layer221":
      return replaceType(obj, "keyboardButtonSimpleWebView");
    case "keyboardButtonRequestPeer_layer221":
      return replaceType(obj, "keyboardButtonRequestPeer");
    case "inputKeyboardButtonRequestPeer_layer221":
      return replaceType(obj, "inputKeyboardButtonRequestPeer");
    case "keyboardButtonCopy_layer221":
      return replaceType(obj, "keyboardButtonCopy");
    default:
      return obj;
  }
}
function mapCompatObject(obj) {
  switch (obj._) {
    case "starGiftUnique_layer197":
    case "starGiftUnique_layer198":
    case "starGiftUnique_layer202":
    case "starGiftUnique_layer206":
    case "starGiftUnique_layer210":
    case "starGiftUnique_layer211":
    case "starGiftUnique_layer214":
    case "starGiftUnique_layer218":
    case "starGiftUnique_layer221":
    case "starGift_layer202":
    case "starGift_layer206":
    case "starGift_layer209":
    case "starGift_layer211":
    case "starGift_layer216":
    case "starGift_layer218":
      return mapCompatStarGift(obj);
    case "emojiStatus_layer197":
    case "emojiStatusUntil_layer214":
      return mapCompatEmojiStatus(obj);
    case "messageMediaDocument_layer197":
    case "messageMediaDice_layer220":
    case "messageMediaPhoto_layer223":
    case "messageMediaPoll_layer223":
      return mapCompatMessageMedia(obj);
    case "channelFull_layer197":
    case "channelFull_layer204":
    case "channelFull_layer225":
      return replaceType(obj, "channelFull");
    case "messageActionStarGift_layer197":
    case "messageActionStarGift_layer211":
    case "messageActionStarGiftUnique_layer197":
    case "messageActionStarGiftUnique_layer202":
    case "messageActionStarGiftUnique_layer210":
    case "messageActionPaidMessagesPrice_layer203":
    case "messageActionSetChatTheme_layer211":
    case "messageActionStarGiftUnique_layer214":
    case "messageActionGiftCode_layer216":
    case "messageActionGiftPremium_layer216":
    case "messageActionStarGift_layer216":
    case "messageActionStarGift_layer218":
    case "messageActionStarGiftUnique_layer221":
      return mapCompatMessageAction(obj);
    case "userFull_layer199":
      return replaceType(dropFields(obj, ["premiumGifts"]), "userFull");
    case "userFull_layer200":
    case "userFull_layer209":
    case "userFull_layer210":
    case "userFull_layer211":
    case "userFull_layer214":
    case "userFull_layer223":
      return replaceType(obj, "userFull");
    case "user_layer199":
    case "user_layer216":
      return {
        ...obj,
        _: "user",
        storiesMaxId: obj.storiesMaxId ? { _: "recentStory", maxId: obj.storiesMaxId } : void 0,
        emojiStatus: obj.emojiStatus ? mapCompatEmojiStatus(obj.emojiStatus) : void 0
      };
    case "channel_layer199":
    case "channel_layer203":
    case "channel_layer216":
      return {
        ...obj,
        _: "channel",
        storiesMaxId: obj.storiesMaxId ? { _: "recentStory", maxId: obj.storiesMaxId } : void 0,
        emojiStatus: obj.emojiStatus ? mapCompatEmojiStatus(obj.emojiStatus) : void 0
      };
    case "channelFull_layer211":
      return replaceType(obj, "channelFull");
    case "phoneCallDiscardReasonAllowGroupCall_layer202":
      return { _: "phoneCallDiscardReasonMissed" };
    case "message_layer199":
    case "message_layer204":
    case "message_layer216":
    case "message_layer220":
    case "message_layer222":
    case "message_layer224":
    case "message_layer225":
    case "messageService_layer204":
      return mapCompatMessage(obj);
    case "messageReplyHeader_layer206":
    case "messageReplyHeader_layer223":
      return mapMessageReplyHeader(obj);
    case "storyItem_layer210":
    case "storyItem_layer223":
      return {
        ...obj,
        _: "storyItem",
        media: mapCompatMessageMedia(obj.media)
      };
    case "todoCompletion_layer216":
      return {
        _: "todoCompletion",
        date: obj.date,
        completedBy: { _: "peerUser", userId: obj.completedBy },
        id: obj.id
      };
    case "premiumGiftOption_layer199":
      return null;
    case "webPageAttributeStarGiftAuction_layer218":
      return {
        _: "webPageAttributeStarGiftAuction",
        gift: mapCompatStarGift(obj.gift),
        endDate: obj.endDate
      };
    case "starGiftAttributeBackdrop_layer221":
    case "starGiftAttributePattern_layer221":
    case "starGiftAttributeModel_layer221":
      return mapCompatStarGiftAttribute(obj);
    case "chatParticipant_layer222":
      return replaceType(obj, "chatParticipant");
    case "chatParticipantCreator_layer222":
      return replaceType(obj, "chatParticipantCreator");
    case "chatParticipantAdmin_layer222":
      return replaceType(obj, "chatParticipantAdmin");
    case "pollResults_layer223":
      return replaceType(obj, "pollResults");
    case "pollAnswerVoters_layer223":
      return replaceType(obj, "pollAnswerVoters");
    case "poll_layer223":
      return mapCompatPoll(obj);
    case "pollAnswer_layer223":
      return replaceType(obj, "pollAnswer");
    case "poll_layer224":
      return mapCompatPoll(obj);
    case "keyboardButton_layer221":
    case "keyboardButtonUrl_layer221":
    case "keyboardButtonCallback_layer221":
    case "keyboardButtonRequestPhone_layer221":
    case "keyboardButtonRequestGeoLocation_layer221":
    case "keyboardButtonSwitchInline_layer221":
    case "keyboardButtonGame_layer221":
    case "keyboardButtonBuy_layer221":
    case "keyboardButtonUrlAuth_layer221":
    case "inputKeyboardButtonUrlAuth_layer221":
    case "keyboardButtonRequestPoll_layer221":
    case "inputKeyboardButtonUserProfile_layer221":
    case "keyboardButtonUserProfile_layer221":
    case "keyboardButtonWebView_layer221":
    case "keyboardButtonSimpleWebView_layer221":
    case "keyboardButtonRequestPeer_layer221":
    case "inputKeyboardButtonRequestPeer_layer221":
    case "keyboardButtonCopy_layer221":
      return mapCompatKeyboardButton(obj);
    case "pageBlockOrderedList_layer225":
      return mapCompatPageBlock(obj);
    case "pageListOrderedItemBlocks_layer225":
      return mapCompatPageListOrderedItem(obj);
    case "pageListItemBlocks_layer225":
      return mapCompatPageListItem(obj);
    case "pageListItemText_layer225":
      return mapCompatPageListItem(obj);
    case "pageListOrderedItemText_layer225":
      return mapCompatPageListOrderedItem(obj);
    default:
      return obj;
  }
}
function wrapReader(reader2) {
  return (r) => mapCompatObject(reader2(r));
}
function getCombinedReaderMap() {
  const ret = {
    ...reader.__tlReaderMap
  };
  for (const [id, reader2] of utils.objectEntries(reader$1.__tlReaderMapCompat)) {
    ret[id] = wrapReader(reader2);
  }
  return ret;
}
const _combinedReaderMap = /* @__PURE__ */ getCombinedReaderMap();
function deserializeObjectWithCompat(data) {
  return tlRuntime.TlBinaryReader.deserializeObject(_combinedReaderMap, data);
}
function deserializePeersIndexWithCompat(data) {
  const res = new peersIndex.PeersIndex();
  const bytes = io.Bytes.from(data);
  const userCount = io.read.int32le(bytes);
  for (let i = 0; i < userCount; i++) {
    const len = io.read.int32le(bytes);
    const obj = deserializeObjectWithCompat(io.read.exactly(bytes, len));
    utils.assert(innerTl.isAnyUser(obj));
    res.users.set(obj.id, obj);
  }
  const chatCount = io.read.int32le(bytes);
  for (let i = 0; i < chatCount; i++) {
    const len = io.read.int32le(bytes);
    const obj = deserializeObjectWithCompat(io.read.exactly(bytes, len));
    utils.assert(innerTl.isAnyChat(obj));
    res.chats.set(obj.id, obj);
  }
  return res;
}
exports.__tlReaderMapWithCompat = _combinedReaderMap;
exports.deserializeObjectWithCompat = deserializeObjectWithCompat;
exports.deserializePeersIndexWithCompat = deserializePeersIndexWithCompat;
