// @mtcute/core/highlevel/storage/storage.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const currentUser = require("./service/current-user.cjs");
const peers = require("./service/peers.cjs");
const refMessages = require("./service/ref-messages.cjs");
const updates = require("./service/updates.cjs");
class TelegramStorageManager {
  constructor(mt, options) {
    this.mt = mt;
    this.options = options;
    this.provider = this.options.provider;
    const serviceOptions = {
      driver: this.mt.driver,
      readerMap: this.mt.options.readerMap,
      writerMap: this.mt.options.writerMap,
      log: this.mt.log
    };
    this.updates = new updates.UpdatesStateService(this.provider.kv, serviceOptions);
    this.self = new currentUser.CurrentUserService(this.provider.kv, serviceOptions);
    this.refMsgs = new refMessages.RefMessagesService(
      this.options.refMessages ?? {},
      this.provider.refMessages,
      serviceOptions
    );
    this.peers = new peers.PeersService(
      this.options.peers ?? {},
      this.provider.peers,
      this.refMsgs,
      serviceOptions
    );
  }
  provider;
  updates;
  self;
  refMsgs;
  peers;
  async close() {
    await this.peers.close();
  }
  async clear(withAuthKeys = false) {
    await this.provider.peers.deleteAll();
    await this.provider.refMessages.deleteAll();
    await this.mt.clear(withAuthKeys);
  }
}
exports.TelegramStorageManager = TelegramStorageManager;
