// @mtcute/core/highlevel/storage/storage.js
import { CurrentUserService } from "./service/current-user.js";
import { PeersService } from "./service/peers.js";
import { RefMessagesService } from "./service/ref-messages.js";
import { UpdatesStateService } from "./service/updates.js";
class TelegramStorageManager {
  constructor(mt, options) {
    this.mt = mt;
    this.options = options;
    this.provider = this.options.provider;
    const serviceOptions = {
      driver: this.mt.driver,
      readerMap: this.mt.options.readerMap,
      writerMap: this.mt.options.writerMap,
      log: this.mt.log
    };
    this.updates = new UpdatesStateService(this.provider.kv, serviceOptions);
    this.self = new CurrentUserService(this.provider.kv, serviceOptions);
    this.refMsgs = new RefMessagesService(
      this.options.refMessages ?? {},
      this.provider.refMessages,
      serviceOptions
    );
    this.peers = new PeersService(
      this.options.peers ?? {},
      this.provider.peers,
      this.refMsgs,
      serviceOptions
    );
  }
  provider;
  updates;
  self;
  refMsgs;
  peers;
  async close() {
    await this.peers.close();
  }
  async clear(withAuthKeys = false) {
    await this.provider.peers.deleteAll();
    await this.provider.refMessages.deleteAll();
    await this.mt.clear(withAuthKeys);
  }
}
export {
  TelegramStorageManager
};
