// @fuman/net/ip/v6.js
import { read } from "@fuman/io";
import { u8, typed } from "@fuman/utils";
const IPV6_PARTS = 8;
function parseV6(string) {
  const parts = new Uint16Array(8);
  let parsedParts = 0;
  let afterDoubleColon;
  let zoneId;
  let currentPartStart = 0;
  let ipv4Part = false;
  let hadSquareBracket = false;
  let pos = 0;
  function addPart(str, part) {
    if (str.length > 4 || Number.isNaN(part) || part < 0 || part > 65535) {
      throw new Error(`Invalid IPv6 part: ${str} (in ${string})`);
    }
    if (afterDoubleColon) {
      afterDoubleColon.push(part);
    } else {
      parts[parsedParts] = part;
      parsedParts += 1;
      if (parsedParts > IPV6_PARTS) {
        throw new Error(`Invalid IPv6 address (too many parts): ${string}`);
      }
    }
  }
  function handleCurrentPart() {
    if (currentPartStart === pos) return;
    const str = string.slice(currentPartStart, pos);
    if (ipv4Part) {
      const ipv4 = str.split(".");
      if (ipv4.length !== 4) {
        throw new Error(`Invalid IPv4 part: ${str} (in ${string})`);
      }
      addPart("", Number.parseInt(ipv4[0], 10) << 8 | Number.parseInt(ipv4[1], 10));
      addPart("", Number.parseInt(ipv4[2], 10) << 8 | Number.parseInt(ipv4[3], 10));
      return;
    }
    const part = Number.parseInt(str, 16);
    addPart(str, part);
  }
  while (pos < string.length) {
    const c = string[pos];
    if (c === "%") {
      handleCurrentPart();
      pos += 1;
      zoneId = string.slice(pos);
      if (zoneId[zoneId.length - 1] === "]") {
        zoneId = zoneId.slice(0, -1);
      }
      break;
    } else if (c === "[") {
      hadSquareBracket = true;
      currentPartStart = pos + 1;
    } else if (c === "]") {
      if (!hadSquareBracket || pos !== string.length - 1 && string[pos + 1] !== "%") {
        throw new Error(`Invalid IPv6 address (unexpected closing bracket): ${string}`);
      }
      handleCurrentPart();
      currentPartStart = pos + 1;
    } else if (c === ":") {
      if (ipv4Part) {
        throw new Error(`Invalid IPv6 address (colon after IPv4 part): ${string}`);
      }
      if (pos !== 0) {
        handleCurrentPart();
      }
      if (string[pos + 1] === ":") {
        if (afterDoubleColon) {
          throw new Error(`Invalid IPv6 address (multiple double-colons): ${string}`);
        }
        afterDoubleColon = [];
        pos += 1;
      } else if (pos === 0) {
        throw new Error(`Invalid IPv6 address (colon at the beginning): ${string}`);
      }
      currentPartStart = pos + 1;
    } else if (c === ".") {
      ipv4Part = true;
    }
    pos += 1;
  }
  if (!hadSquareBracket && zoneId === void 0) {
    handleCurrentPart();
  }
  if (afterDoubleColon) {
    const missing = IPV6_PARTS - parsedParts - afterDoubleColon.length;
    if (missing < 0) {
      throw new Error(`Invalid IPv6 address (too many parts): ${string}`);
    }
    for (let i = 0; i < missing; i++) {
      parts[parsedParts + i] = 0;
    }
    for (let i = 0; i < afterDoubleColon.length; i++) {
      parts[parsedParts + missing + i] = afterDoubleColon[i];
    }
  } else if (parsedParts < IPV6_PARTS) {
    throw new Error(`Invalid IPv6 address (too few parts): ${string}`);
  }
  return {
    type: "ipv6",
    parts,
    zoneId
  };
}
function stringifyV6(parsed, options) {
  const {
    zeroCompression = true,
    fixedLength = false
  } = options || {};
  let result = "";
  let compressStart = -1;
  let compressEnd = 0;
  if (zeroCompression) {
    let maxLength = 0;
    let maxStart = -1;
    let start = -1;
    let length = 0;
    for (let i2 = 0; i2 < parsed.parts.length; i2++) {
      if (parsed.parts[i2] === 0) {
        if (start === -1) {
          start = i2;
        }
        length++;
      } else {
        if (length > maxLength) {
          maxLength = length;
          maxStart = start;
        }
        start = -1;
        length = 0;
      }
    }
    if (length > maxLength) {
      maxLength = length;
      maxStart = start;
    }
    if (maxLength > 1) {
      compressStart = maxStart;
      compressEnd = maxStart + maxLength;
    }
  }
  let i = 0;
  let writeColon = false;
  while (i < parsed.parts.length) {
    const part = parsed.parts[i];
    if (part < 0 || part > 65535) {
      throw new Error(`Invalid IPv6 part: ${part.toString(16)}`);
    }
    if (i === compressStart) {
      result += "::";
      i = compressEnd;
      writeColon = false;
      continue;
    }
    if (writeColon) result += ":";
    else writeColon = true;
    if (fixedLength) {
      if (part < 16) result += "000";
      else if (part < 256) result += "00";
      else if (part < 4096) result += "0";
    }
    result += part.toString(16);
    i += 1;
  }
  if (parsed.zoneId != null) {
    result += `%${parsed.zoneId}`;
  }
  return result;
}
function expandV6(string) {
  return stringifyV6(parseV6(string), { zeroCompression: false });
}
function _fromBytesInternal(bytes) {
  const parts = new Uint16Array(bytes.buffer, bytes.byteOffset, 8);
  if (typed.getPlatformByteOrder() === "little") {
    u8.swap16(bytes);
  }
  return {
    type: "ipv6",
    parts
  };
}
function fromBytesV6(bytes) {
  if (bytes.length !== 16) {
    throw new Error(`Invalid IPv6 address buffer: ${bytes.length} ≠ 16`);
  }
  return _fromBytesInternal(u8.clone(bytes));
}
function readV6(reader) {
  return _fromBytesInternal(read.exactly(reader, 16));
}
function toBytesV6(parsed) {
  const copy = new Uint16Array(parsed.parts);
  const bytes = new Uint8Array(copy.buffer);
  if (typed.getPlatformByteOrder() === "little") {
    u8.swap16(bytes);
  }
  return bytes;
}
function writeV6(parsed, writer) {
  const buf = writer.writeSync(16);
  buf.set(new Uint8Array(parsed.parts.buffer, parsed.parts.byteOffset, 16));
  if (typed.getPlatformByteOrder() === "little") {
    u8.swap16(buf);
  }
  writer.disposeWriteSync();
}
export {
  expandV6,
  fromBytesV6,
  parseV6,
  readV6,
  stringifyV6,
  toBytesV6,
  writeV6
};
