// @fuman/utils/encodings/hex.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const pool = require("../arrays/u8/pool.cjs");
const hexSliceLookupTable = function() {
  const alphabet = "0123456789abcdef";
  const table = Array.from({ length: 256 });
  for (let i = 0; i < 16; ++i) {
    const i16 = i * 16;
    for (let j = 0; j < 16; ++j) {
      table[i16 + j] = alphabet[i] + alphabet[j];
    }
  }
  return table;
}();
const hexCharValueTable = {
  0: 0,
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  a: 10,
  b: 11,
  c: 12,
  d: 13,
  e: 14,
  f: 15,
  A: 10,
  B: 11,
  C: 12,
  D: 13,
  E: 14,
  F: 15
};
const HAS_TO_HEX = typeof Uint8Array.prototype.toHex === "function";
const HAS_FROM_HEX = typeof Uint8Array.fromHex === "function";
function encode(buf) {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(buf).toString("hex");
  }
  if (HAS_TO_HEX) {
    return buf.toHex();
  }
  let out = "";
  for (let i = 0; i < buf.byteLength; ++i) {
    out += hexSliceLookupTable[buf[i]];
  }
  return out;
}
function decode(string) {
  if (typeof Buffer !== "undefined") {
    const buf2 = Buffer.from(string, "hex");
    return new Uint8Array(buf2.buffer, buf2.byteOffset, buf2.byteLength);
  }
  if (HAS_FROM_HEX) {
    return Uint8Array.fromHex(string);
  }
  const buf = pool.alloc(Math.ceil(string.length / 2));
  const strLen = string.length;
  const length = Math.min(buf.length, strLen / 2);
  let i;
  for (i = 0; i < length; ++i) {
    const a = hexCharValueTable[string[i * 2]];
    const b = hexCharValueTable[string[i * 2 + 1]];
    if (a === void 0 || b === void 0) {
      throw new Error("Invalid hex string");
    }
    buf[i] = a << 4 | b;
  }
  return buf;
}
function encodedLength(n) {
  return n * 2;
}
function decodedLength(n) {
  return Math.ceil(n / 2);
}
exports.decode = decode;
exports.decodedLength = decodedLength;
exports.encode = encode;
exports.encodedLength = encodedLength;
