// @fuman/utils/async/async-resource.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const deferred = require("./deferred.cjs");
const emitter = require("./emitter.cjs");
const timers = require("./timers.cjs");
class AsyncResource {
  constructor(params) {
    this.params = params;
    this.#ctx = {
      current: null,
      currentFetchedAt: 0,
      currentExpiresAt: 0,
      isBackground: false,
      // eslint-disable-next-line ts/no-non-null-assertion
      abort: null
    };
  }
  #abort;
  #ctx;
  #updating;
  #timeout;
  onUpdated = new emitter.Emitter();
  get isStale() {
    return this.#ctx.current === null || this.#ctx.currentExpiresAt <= performance.now();
  }
  setData(data, expiresIn) {
    const now = performance.now();
    this.#ctx.current = data;
    this.#ctx.currentExpiresAt = now + expiresIn;
    this.#ctx.currentFetchedAt = now;
    this.onUpdated.emit(this.#ctx);
    if (this.params.autoReload) {
      if (this.#timeout) {
        timers.clearTimeout(this.#timeout);
      }
      this.#timeout = timers.setTimeout(() => {
        this.#ctx.isBackground = true;
        this.update().catch(() => {
        });
      }, expiresIn + (this.params.autoReloadAfter ?? 0));
    }
  }
  /**
   * update the resource
   *
   * @param force  whether to force the update even if the resource hasn't expired yet
   */
  async update(force = false) {
    if (this.#updating) {
      await this.#updating.promise;
      return;
    }
    if (!force && !this.isStale) {
      return;
    }
    this.#abort?.abort();
    this.#abort = new AbortController();
    this.#ctx.abort = this.#abort.signal;
    this.#updating = new deferred.Deferred();
    let result;
    try {
      result = await this.params.fetcher(this.#ctx);
    } catch (e) {
      if (this.params.onError) {
        this.params.onError(e, this.#ctx);
      } else {
        console.error(e);
      }
      this.#updating.resolve();
      this.#updating = void 0;
      return;
    }
    this.#updating.resolve();
    this.#updating = void 0;
    this.setData(result.data, result.expiresIn);
  }
  /**
   * get the resource value, refreshing it if needed
   *
   * note: this method still *might* return `null` if the resource wasn't ever cached yet,
   * and the underlying call to `fetcher` failed
   */
  async get() {
    if (this.params.swr === true && this.#ctx.current !== null) {
      const validator = this.params.swrValidator;
      if (!validator || validator(this.#ctx)) {
        this.#ctx.isBackground = true;
        this.update(true).catch(() => {
        });
        return this.#ctx.current;
      }
    }
    this.#ctx.isBackground = false;
    await this.update();
    return this.#ctx.current;
  }
  /**
   * get the cached resource value immediately (if any)
   * note that it may be stale, which you should check separately
   */
  getCached() {
    return this.#ctx.current;
  }
  destroy() {
    if (this.#timeout) timers.clearTimeout(this.#timeout);
    this.onUpdated.clear();
    this.#abort?.abort();
  }
}
exports.AsyncResource = AsyncResource;
