// @fuman/utils/async/async-queue.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const deque = require("../structures/deque.cjs");
const deferred = require("./deferred.cjs");
class AsyncQueue {
  constructor(from) {
    if (from) {
      if (from instanceof deque.Deque) {
        this.queue = from;
      } else {
        this.queue = new deque.Deque(from);
      }
    } else {
      this.queue = new deque.Deque();
    }
  }
  /**
   * Underlying deque.
   *
   * Use with care and don't modify it directly,
   * as it may break the async queue's internal state.
   */
  queue;
  #waiters = new deque.Deque();
  #ended = false;
  get length() {
    return this.queue.length;
  }
  enqueue(item) {
    if (this.#ended) {
      throw new Error(".end() has already been called");
    }
    if (this.#waiters.length > 0) {
      const waiter = this.#waiters.popFront();
      waiter.resolve(item);
      return;
    }
    this.queue.pushBack(item);
  }
  end() {
    if (this.#ended) {
      throw new Error(".end() has already been called");
    }
    this.#ended = true;
    for (const waiter of this.#waiters) {
      waiter.resolve(void 0);
    }
    this.#waiters.clear();
  }
  get ended() {
    return this.#ended;
  }
  peek() {
    return this.queue.peekFront();
  }
  next() {
    return this.queue.popFront();
  }
  async nextOrWait() {
    if (this.queue.length > 0 || this.#ended) {
      return this.queue.popFront();
    }
    const waiter = new deferred.Deferred();
    this.#waiters.pushBack(waiter);
    return waiter.promise;
  }
  [Symbol.asyncIterator]() {
    const iterator = {
      [Symbol.asyncIterator]: () => iterator,
      next: async () => {
        const item = await this.nextOrWait();
        if (item === void 0) return { done: true, value: void 0 };
        return { value: item, done: false };
      }
    };
    return iterator;
  }
}
exports.AsyncQueue = AsyncQueue;
