// @fuman/utils/async/pool.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class ErrorInfo {
  constructor(item, index, error) {
    this.item = item;
    this.index = index;
    this.error = error;
  }
}
class AggregateError extends Error {
  constructor(errors) {
    super(`AggregateError: ${errors.length} errors`);
    this.errors = errors;
  }
}
async function asyncPool(iterable, executor, options = {}) {
  const {
    limit = 16,
    signal,
    onError
  } = options;
  if (limit <= 0) throw new Error("Pool limit must be a positive integer");
  const iteratorFactory = iterable[Symbol.asyncIterator] ?? iterable[Symbol.iterator];
  if (iteratorFactory === void 0) throw new Error("`iterable` must be an iterable!");
  const iterator = iteratorFactory.call(iterable);
  let idx = 0;
  let errored = false;
  let ended = false;
  const errors = [];
  if (signal) {
    if (signal.aborted) {
      throw signal.reason;
    }
  }
  async function worker() {
    while (true) {
      if (signal?.aborted) throw signal.reason;
      if (errored || ended) break;
      const result = await iterator.next();
      if (signal?.aborted) throw signal.reason;
      if (result.done) {
        ended = true;
        break;
      }
      if (errored) break;
      const thisIdx = idx++;
      const item = result.value;
      try {
        await executor(item, thisIdx);
      } catch (err) {
        const action = onError?.(item, thisIdx, err) ?? "throw";
        if (action === "ignore") {
          continue;
        }
        if (action === "collect") {
          errors.push(new ErrorInfo(item, thisIdx, err));
          continue;
        }
        if (action === "throw") {
          errored = true;
          throw err;
        }
      }
    }
  }
  await Promise.all(Array.from({ length: limit }, worker));
  if (signal?.aborted) throw signal.reason;
  if (errors.length > 0) {
    throw new AggregateError(errors);
  }
}
async function parallelMap(iterable, executor, options = {}) {
  const result = [];
  if (Array.isArray(iterable)) {
    result.length = iterable.length;
  }
  await asyncPool(iterable, async (item, index) => {
    const res = await executor(item, index);
    result[index] = res;
  }, options);
  return result;
}
exports.AggregateError = AggregateError;
exports.asyncPool = asyncPool;
exports.parallelMap = parallelMap;
