// @mtcute/core/highlevel/types/bots/inline-result/factories.js
import { MtArgumentError } from "../../../../types/errors.js";
import { fileIdToInputPhoto, fileIdToInputDocument } from "../../../utils/convert-file-id.js";
import { extractFileName } from "../../../utils/file-utils.js";
import { _convertToTl as _convertToTl$1 } from "../inline-message/factories.js";
function article(id, params) {
  const ret = params;
  ret.id = id;
  ret.type = "article";
  return ret;
}
function gif(id, media, params = {}) {
  const ret = params;
  ret.id = id;
  ret.type = "gif";
  ret.media = media;
  return ret;
}
function video(id, media, params) {
  const ret = params;
  ret.id = id;
  ret.type = "video";
  ret.media = media;
  return ret;
}
function audio(id, media, params) {
  const ret = params;
  ret.id = id;
  ret.type = "audio";
  ret.media = media;
  return ret;
}
function voice(id, media, params) {
  const ret = params;
  ret.id = id;
  ret.type = "voice";
  ret.media = media;
  return ret;
}
function photo(id, media, params = {}) {
  const ret = params;
  ret.id = id;
  ret.type = "photo";
  ret.media = media;
  return ret;
}
function sticker(id, media) {
  return {
    id,
    type: "sticker",
    media
  };
}
function file(id, media, params) {
  const ret = params;
  ret.id = id;
  ret.type = "file";
  ret.media = media;
  return ret;
}
function geo(id, params) {
  const ret = params;
  ret.id = id;
  ret.type = "geo";
  return ret;
}
function venue(id, params) {
  const ret = params;
  ret.id = id;
  ret.type = "venue";
  return ret;
}
function contact(id, params) {
  const ret = params;
  ret.id = id;
  ret.type = "contact";
  return ret;
}
function game(id, shortName, params = {}) {
  const ret = params;
  ret.id = id;
  ret.type = "game";
  ret.shortName = shortName;
  return ret;
}
async function _convertToTl(client, results) {
  const normalizeThumb = (obj, fallback) => {
    if (obj.type !== "voice" && obj.type !== "audio" && obj.type !== "sticker" && obj.type !== "game") {
      if (!obj.thumb || typeof obj.thumb === "string") {
        if (!obj.thumb && !fallback) {
          return void 0;
        }
        return {
          _: "inputWebDocument",
          size: 0,
          url: obj.thumb || fallback,
          mimeType: obj.type === "gif" ? obj.thumbMime ?? obj.mime ?? "video/mp4" : "image/jpeg",
          attributes: []
        };
      }
      return obj.thumb;
    }
  };
  const items = [];
  let isGallery = false;
  let forceVertical = false;
  for (const obj of results) {
    switch (obj.type) {
      case "article": {
        forceVertical = true;
        let sendMessage2;
        if (obj.message) {
          sendMessage2 = await _convertToTl$1(client, obj.message);
        } else {
          let message = obj.title;
          const entities = [
            {
              _: "messageEntityBold",
              offset: 0,
              length: message.length
            }
          ];
          if (obj.url) {
            entities.push({
              _: "messageEntityTextUrl",
              url: obj.url,
              offset: 0,
              length: message.length
            });
          }
          if (obj.description) {
            message += `
${obj.description}`;
          }
          sendMessage2 = {
            _: "inputBotInlineMessageText",
            message,
            entities
          };
        }
        items.push({
          _: "inputBotInlineResult",
          id: obj.id,
          type: obj.type,
          title: obj.title,
          description: obj.description,
          url: obj.hideUrl ? void 0 : obj.url,
          content: obj.url && obj.hideUrl ? {
            _: "inputWebDocument",
            url: obj.url,
            mimeType: "text/html",
            size: 0,
            attributes: []
          } : void 0,
          thumb: typeof obj.thumb === "string" ? normalizeThumb(obj) : obj.thumb,
          sendMessage: sendMessage2
        });
        continue;
      }
      case "game": {
        let sendMessage2;
        if (obj.message) {
          sendMessage2 = await _convertToTl$1(client, obj.message);
          if (sendMessage2._ !== "inputBotInlineMessageGame") {
            throw new MtArgumentError("game inline result must contain a game inline message");
          }
        } else {
          sendMessage2 = {
            _: "inputBotInlineMessageGame"
          };
        }
        items.push({
          _: "inputBotInlineResultGame",
          id: obj.id,
          shortName: obj.shortName,
          sendMessage: sendMessage2
        });
        continue;
      }
      case "gif":
      case "photo":
      case "sticker":
        isGallery = true;
        break;
      case "audio":
      case "contact":
      case "voice":
        forceVertical = true;
    }
    let sendMessage;
    if (obj.message) {
      sendMessage = await _convertToTl$1(client, obj.message);
    } else if (obj.type === "venue") {
      if (obj.latitude && obj.longitude) {
        sendMessage = {
          _: "inputBotInlineMessageMediaVenue",
          title: obj.title,
          address: obj.address,
          geoPoint: {
            _: "inputGeoPoint",
            lat: obj.latitude,
            long: obj.longitude
          },
          provider: "",
          venueId: "",
          venueType: ""
        };
      } else {
        throw new MtArgumentError("message or location (lat&lon) bust be supplied for venue inline result");
      }
    } else if (obj.type === "video" && obj.isEmbed && typeof obj.media === "string") {
      sendMessage = {
        _: "inputBotInlineMessageText",
        message: obj.media
      };
    } else if (obj.type === "geo") {
      sendMessage = {
        _: "inputBotInlineMessageMediaGeo",
        geoPoint: {
          _: "inputGeoPoint",
          lat: obj.latitude,
          long: obj.longitude
        }
      };
    } else if (obj.type === "contact") {
      sendMessage = {
        _: "inputBotInlineMessageMediaContact",
        phoneNumber: obj.phone,
        firstName: obj.firstName,
        lastName: obj.lastName ?? "",
        vcard: ""
      };
    } else {
      sendMessage = {
        _: "inputBotInlineMessageMediaAuto",
        message: ""
      };
    }
    let media;
    if (obj.type !== "geo" && obj.type !== "venue" && obj.type !== "contact") {
      if (typeof obj.media === "string") {
        if (obj.media.match(/^https?:\/\//)) {
          if (obj.type === "sticker") {
            throw new MtArgumentError("sticker inline result cannot contain a URL");
          }
          let mime;
          if (obj.type === "video") {
            mime = "video/mp4";
          } else if (obj.type === "audio") {
            mime = obj.mime ?? "audio/mpeg";
          } else if (obj.type === "gif") {
            mime = obj.mime ?? "video/mp4";
          } else if (obj.type === "voice") {
            mime = "audio/ogg";
          } else if (obj.type === "file") {
            if (!obj.mime) {
              throw new MtArgumentError("MIME type must be specified for file inline result");
            }
            mime = obj.mime;
          } else {
            mime = "image/jpeg";
          }
          const attributes = [];
          if ((obj.type === "video" || obj.type === "gif" || obj.type === "photo") && obj.width && obj.height) {
            if (obj.type !== "photo" && obj.duration) {
              attributes.push({
                _: "documentAttributeVideo",
                w: obj.width,
                h: obj.height,
                duration: obj.duration
              });
            } else {
              attributes.push({
                _: "documentAttributeImageSize",
                w: obj.width,
                h: obj.height
              });
            }
          } else if (obj.type === "audio" || obj.type === "voice") {
            attributes.push({
              _: "documentAttributeAudio",
              voice: obj.type === "voice",
              duration: obj.duration ?? 0,
              title: obj.type === "audio" ? obj.title : "",
              performer: obj.type === "audio" ? obj.performer : ""
            });
          }
          attributes.push({
            _: "documentAttributeFilename",
            fileName: extractFileName(obj.media)
          });
          media = {
            _: "inputWebDocument",
            url: obj.media,
            mimeType: mime,
            size: 0,
            attributes
          };
        } else if (obj.type === "photo") {
          media = fileIdToInputPhoto(obj.media);
        } else {
          media = fileIdToInputDocument(obj.media);
        }
      } else {
        media = obj.media;
      }
    }
    let title;
    let description;
    if (obj.type === "contact") {
      title = obj.lastName?.length ? `${obj.firstName} ${obj.lastName}` : obj.firstName;
    } else if (obj.type !== "sticker") {
      title = obj.title;
    }
    if (obj.type === "audio") {
      description = obj.performer;
    } else if (obj.type === "geo") {
      description = `${obj.latitude} ${obj.longitude}`;
    } else if (obj.type === "venue") {
      description = obj.address;
    } else if (obj.type === "contact") {
      description = obj.phone;
    } else if (obj.type !== "voice" && obj.type !== "sticker") {
      description = obj.description;
    }
    if (!media || media._ === "inputWebDocument") {
      items.push({
        _: "inputBotInlineResult",
        id: obj.id,
        type: obj.type,
        title,
        description,
        content: media,
        thumb: normalizeThumb(obj, media?.url),
        sendMessage
      });
      continue;
    }
    if (media._ === "inputPhoto" || media._ === "inputPhotoEmpty") {
      items.push({
        _: "inputBotInlineResultPhoto",
        id: obj.id,
        type: obj.type,
        photo: media,
        sendMessage
      });
      continue;
    }
    items.push({
      _: "inputBotInlineResultDocument",
      id: obj.id,
      type: obj.type,
      title,
      description,
      document: media,
      sendMessage
    });
  }
  return [isGallery && !forceVertical, items];
}
export {
  _convertToTl,
  article,
  audio,
  contact,
  file,
  game,
  geo,
  gif,
  photo,
  sticker,
  venue,
  video,
  voice
};
