// @mtcute/core/storage/memory/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const driver = require("./driver.cjs");
const authKeys = require("./repository/auth-keys.cjs");
const kv = require("./repository/kv.cjs");
const peers = require("./repository/peers.cjs");
const refMessages = require("./repository/ref-messages.cjs");
class MemoryStorage {
  driver = new driver.MemoryStorageDriver();
  kv = new kv.MemoryKeyValueRepository(this.driver);
  authKeys = new authKeys.MemoryAuthKeysRepository(this.driver);
  peers = new peers.MemoryPeersRepository(this.driver);
  refMessages = new refMessages.MemoryRefMessagesRepository(this.driver);
}
exports.MemoryStorageDriver = driver.MemoryStorageDriver;
exports.MemoryAuthKeysRepository = authKeys.MemoryAuthKeysRepository;
exports.MemoryKeyValueRepository = kv.MemoryKeyValueRepository;
exports.MemoryPeersRepository = peers.MemoryPeersRepository;
exports.MemoryRefMessagesRepository = refMessages.MemoryRefMessagesRepository;
exports.MemoryStorage = MemoryStorage;
