// @mtcute/core/network/client.js
import { Emitter } from "@fuman/utils";
import { StorageManager } from "../storage/storage.js";
import { __tlReaderMap } from "../tl/binary/reader.js";
import { __tlWriterMap } from "../tl/binary/writer.js";
import { defaultTestIpv6Dc, defaultTestDc, defaultProductionIpv6Dc, defaultProductionDc } from "../utils/dcs.js";
import { asyncResettable } from "../utils/function-utils.js";
import { LogManager } from "../utils/logger.js";
import { dropUndefined } from "../utils/misc-utils.js";
import { isTlRpcError } from "../utils/type-assertions.js";
import "@mtcute/tl-runtime";
import { ConfigManager } from "./config-manager.js";
import { NetworkManager } from "./network-manager.js";
import { LAYER } from "../tl/inner-tl.js";
class MtClient {
  constructor(params) {
    this.params = params;
    this.log = params.logger ?? new LogManager(void 0, params.platform);
    if (params.logLevel !== void 0) {
      this.log.mgr.level = params.logLevel;
    }
    this.crypto = params.crypto;
    this._testMode = Boolean(params.testMode);
    let dc = params.defaultDcs;
    if (!dc) {
      if (params.testMode) {
        dc = params.useIpv6 ? defaultTestIpv6Dc : defaultTestDc;
      } else {
        dc = params.useIpv6 ? defaultProductionIpv6Dc : defaultProductionDc;
      }
    }
    this._defaultDcs = dc;
    this._layer = params.overrideLayer ?? LAYER;
    this._readerMap = params.readerMap ?? __tlReaderMap;
    this._writerMap = params.writerMap ?? __tlWriterMap;
    this._abortController = new AbortController();
    this.stopSignal = this._abortController.signal;
    this.storage = new StorageManager({
      provider: params.storage,
      log: this.log,
      readerMap: this._readerMap,
      writerMap: this._writerMap,
      platform: params.platform,
      ...dropUndefined(params.storageOptions)
    });
    this.network = new NetworkManager(
      {
        apiId: params.apiId,
        crypto: this.crypto,
        disableUpdates: params.disableUpdates ?? false,
        initConnectionOptions: params.initConnectionOptions,
        layer: this._layer,
        log: this.log,
        readerMap: this._readerMap,
        writerMap: this._writerMap,
        reconnectionStrategy: params.reconnectionStrategy,
        storage: this.storage,
        testMode: Boolean(params.testMode),
        transport: params.transport,
        emitError: params.onError,
        isPremium: false,
        useIpv6: Boolean(params.useIpv6),
        onUsable: this.onUsable.emit.bind(this.onUsable),
        onConnecting: this.onConnecting.emit.bind(this.onConnecting),
        onNetworkChanged: this.onNetworkChanged.emit.bind(this.onNetworkChanged),
        onUpdate: this.onUpdate.emit.bind(this.onUpdate),
        stopSignal: this.stopSignal,
        platform: params.platform,
        ...dropUndefined(params.network)
      },
      this._config
    );
  }
  /**
   * Crypto provider taken from {@link MtClientOptions.crypto}
   */
  crypto;
  /** Storage manager */
  storage;
  /**
   * "Test mode" taken from {@link MtClientOptions.testMode}
   */
  _testMode;
  /**
   * Primary DCs taken from {@link MtClientOptions.defaultDcs},
   * loaded from session or changed by other means (like redirecting).
   */
  _defaultDcs;
  /** TL layer used by the client */
  _layer;
  /** TL readers map used by the client */
  _readerMap;
  /** TL writers map used by the client */
  _writerMap;
  _config = new ConfigManager(async () => {
    const res = await this.call({ _: "help.getConfig" });
    if (isTlRpcError(res)) throw new Error(`Failed to get config: ${res.errorMessage}`);
    return res;
  });
  log;
  network;
  _abortController;
  stopSignal;
  onUsable = new Emitter();
  onConnecting = new Emitter();
  onNetworkChanged = new Emitter();
  onUpdate = new Emitter();
  _prepare = asyncResettable(async () => {
    await this.crypto.initialize?.();
    await this.storage.load();
    const primaryDc = await this.storage.dcs.fetch();
    if (primaryDc !== null) this._defaultDcs = primaryDc;
  });
  /**
   * **ADVANCED**
   *
   * Do all the preparations, but don't connect just yet.
   * Useful when you want to do some preparations before
   * connecting, like setting up session.
   *
   * Call {@link connect} to actually connect.
   */
  prepare() {
    return this._prepare.run();
  }
  _connect = asyncResettable(async () => {
    await this._prepare.run();
    await this.network.connect(this._defaultDcs);
  });
  /**
   * Initialize the connection to the primary DC.
   *
   * You shouldn't usually call this method directly as it is called
   * implicitly the first time you call {@link call}.
   */
  async connect() {
    return this._connect.run();
  }
  async disconnect() {
    await this.network.destroy();
    await this.storage.save();
    this._config.destroy();
    this._prepare.reset();
    this._connect.reset();
    this._abortController.abort();
    this.log.debug("client disconnected");
  }
  /**
   * Close all connections and finalize the client.
   *
   * **Note**: must be called *after* {@link disconnect}
   */
  async destroy() {
    await this.storage.destroy();
    this.log.debug("client destroyed");
  }
  /**
   * Make an RPC call.
   *
   * The connection must have been {@link connect}-ed
   * before calling this method.
   *
   * This method is still quite low-level and you shouldn't use this
   * when using high-level API provided by `@mtcute/client`.
   *
   * @param message  RPC method to call
   * @param params  Additional call parameters
   */
  async call(message, params) {
    return this.network.call(message, params);
  }
}
export {
  MtClient
};
