// @mtcute/core/network/config-manager.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
class ConfigManager extends utils.AsyncResource {
  constructor(update) {
    super({
      fetcher: async () => {
        const res = await update();
        return {
          data: res,
          expiresIn: res.expires * 1e3 - Date.now()
        };
      },
      autoReload: true
    });
  }
  async findOption(params) {
    if (this.isStale) await this.update();
    const data = this.getCached();
    const options = data.dcOptions.filter((opt) => {
      if (opt.tcpoOnly) return false;
      if (opt.ipv6 && !params.allowIpv6) return false;
      if (opt.mediaOnly && !params.allowMedia) return false;
      if (opt.cdn && !params.cdn) return false;
      return opt.id === params.dcId;
    });
    if (params.preferMedia && params.preferIpv6) {
      const r = options.find((opt) => opt.mediaOnly && opt.ipv6);
      if (r) return r;
    }
    if (params.preferMedia) {
      const r = options.find((opt) => opt.mediaOnly);
      if (r) return r;
    }
    if (params.preferIpv6) {
      const r = options.find((opt) => opt.ipv6);
      if (r) return r;
    }
    return options[0];
  }
}
exports.ConfigManager = ConfigManager;
