// @mtcute/core/network/config-manager.js
import { AsyncResource } from "@fuman/utils";
class ConfigManager extends AsyncResource {
  constructor(update) {
    super({
      fetcher: async () => {
        const res = await update();
        return {
          data: res,
          expiresIn: res.expires * 1e3 - Date.now()
        };
      },
      autoReload: true
    });
  }
  async findOption(params) {
    if (this.isStale) await this.update();
    const data = this.getCached();
    const options = data.dcOptions.filter((opt) => {
      if (opt.tcpoOnly) return false;
      if (opt.ipv6 && !params.allowIpv6) return false;
      if (opt.mediaOnly && !params.allowMedia) return false;
      if (opt.cdn && !params.cdn) return false;
      return opt.id === params.dcId;
    });
    if (params.preferMedia && params.preferIpv6) {
      const r = options.find((opt) => opt.mediaOnly && opt.ipv6);
      if (r) return r;
    }
    if (params.preferMedia) {
      const r = options.find((opt) => opt.mediaOnly);
      if (r) return r;
    }
    if (params.preferIpv6) {
      const r = options.find((opt) => opt.ipv6);
      if (r) return r;
    }
    return options[0];
  }
}
export {
  ConfigManager
};
