// @mtcute/core/network/flood-control.js
import { timers } from "@fuman/utils";
class FloodControl {
  _rules;
  _maxWindow;
  _events = [];
  constructor(rules) {
    if (rules.length === 0) throw new Error("FloodControl needs at least one rule");
    this._rules = rules;
    this._maxWindow = rules.reduce((m, r) => Math.max(m, r.windowMs), 0);
  }
  /** Earliest absolute perf-now ms at which a new event is permitted. */
  getWakeupAt(now) {
    let wakeup = 0;
    for (const { count, windowMs } of this._rules) {
      const cutoff = now - windowMs;
      let inWindow = 0;
      let oldestInWindowIdx = -1;
      for (let i = this._events.length - 1; i >= 0; i--) {
        if (this._events[i] <= cutoff) break;
        inWindow++;
        oldestInWindowIdx = i;
      }
      if (inWindow >= count) {
        const t = this._events[oldestInWindowIdx] + windowMs;
        if (t > wakeup) wakeup = t;
      }
    }
    return wakeup;
  }
  addEvent(now) {
    this._events.push(now);
    this._prune(now);
  }
  _prune(now) {
    const cutoff = now - this._maxWindow;
    let i = 0;
    while (i < this._events.length && this._events[i] < cutoff) i++;
    if (i > 0) this._events.splice(0, i);
  }
  reset() {
    this._events = [];
  }
}
const DEFAULT_SANITY_LIMITS = [
  { count: 5, windowMs: 1e4 }
];
const DEFAULT_FLOOD_LIMITS = [
  { count: 1, windowMs: 1e3 },
  { count: 4, windowMs: 2e3 },
  { count: 8, windowMs: 3e3 }
];
const DEFAULT_MTPROTO_ERROR_LIMITS = [
  { count: 1, windowMs: 1e3 },
  { count: 4, windowMs: 2e3 },
  { count: 8, windowMs: 3e3 }
];
class ConnectionFloodController {
  sanity;
  flood;
  mtprotoError;
  _log;
  constructor(limits = {}) {
    this.sanity = new FloodControl(limits.sanity ?? DEFAULT_SANITY_LIMITS);
    this.flood = new FloodControl(limits.flood ?? DEFAULT_FLOOD_LIMITS);
    this.mtprotoError = new FloodControl(limits.mtprotoError ?? DEFAULT_MTPROTO_ERROR_LIMITS);
  }
  setLogger(log) {
    this._log = log;
  }
  /**
   * Notify a network-up transition.
   *
   * TDLib clears sanity + online-regime flood + backoff on
   * `ConnectionCreator::on_network(true)`. We don't model the online regime,
   * and our transport layer owns its own reconnect backoff, so only the
   * sanity limiter is reset here.
   */
  notifyNetworkUp() {
    this._log?.debug("network up, resetting sanity limiter");
    this.sanity.reset();
  }
  /** Record an MTProto-level error (transport 429; per TDLib, other codes don't tick this). */
  addMtprotoError(now = performance.now()) {
    this._log?.debug("recording mtproto error");
    this.mtprotoError.addEvent(now);
  }
  /** Earliest perf-now ms at which a new connection attempt is permitted. */
  getWakeupAt(now) {
    return Math.max(
      this.sanity.getWakeupAt(now),
      this.flood.getWakeupAt(now),
      this.mtprotoError.getWakeupAt(now)
    );
  }
  /**
   * Block until a connection attempt is permitted under all applicable
   * limiters, then record the attempt. Concurrent waiters serialize fairly
   * because the first to wake records its event before the next checks state.
   *
   * If `signal` aborts, the pending timer is cleared and the promise rejects
   * with the signal's reason.
   */
  async wait(signal) {
    let waited = false;
    while (true) {
      signal?.throwIfAborted();
      const now = performance.now();
      const wakeup = this.getWakeupAt(now);
      if (wakeup <= now) {
        this.sanity.addEvent(now);
        this.flood.addEvent(now);
        if (waited) this._log?.debug("admitted connect after waiting");
        return;
      }
      const delay = wakeup - now;
      this._log?.debug("connect throttled (waiting %dms)", Math.round(delay));
      waited = true;
      await new Promise((resolve, reject) => {
        const handle = timers.setTimeout(() => {
          signal?.removeEventListener("abort", onAbort);
          resolve();
        }, delay);
        const onAbort = () => {
          timers.clearTimeout(handle);
          reject(signal.reason);
        };
        signal?.addEventListener("abort", onAbort, { once: true });
      });
    }
  }
  /** Reset all internal state. Mainly for tests. */
  reset() {
    this.sanity.reset();
    this.flood.reset();
    this.mtprotoError.reset();
  }
}
export {
  ConnectionFloodController,
  DEFAULT_FLOOD_LIMITS,
  DEFAULT_MTPROTO_ERROR_LIMITS,
  DEFAULT_SANITY_LIMITS,
  FloodControl
};
