// @mtcute/core/network/mtproto-session.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const tlRuntime = require("@mtcute/tl-runtime");
const Long = require("long");
const longUtils = require("../utils/long-utils.cjs");
const miscUtils = require("../utils/misc-utils.cjs");
const sortedArray = require("../utils/sorted-array.cjs");
const authKey = require("./auth-key.cjs");
const errors = require("../types/errors.cjs");
class MtprotoSession {
  constructor(_crypto, log, _readerMap, _writerMap, _salts) {
    this._crypto = _crypto;
    this.log = log;
    this._readerMap = _readerMap;
    this._writerMap = _writerMap;
    this._salts = _salts;
    this.log.prefix = `[SESSION ${this._sessionId.toString(16)}] `;
    this._authKey = new authKey.AuthKey(_crypto, log, _readerMap);
    this._authKeyTemp = new authKey.AuthKey(_crypto, log, _readerMap);
    this._authKeyTempSecondary = new authKey.AuthKey(_crypto, log, _readerMap);
  }
  _sessionId = longUtils.randomLong();
  _authKey;
  _authKeyTemp;
  _authKeyTempSecondary;
  _timeOffset = Math.floor(performance.timeOrigin / 1e3);
  _lastMessageId = Long.ZERO;
  _seqNo = 0;
  /// state ///
  // recent msg ids
  recentOutgoingMsgIds = new utils.LruSet(1e3, longUtils.LongSet);
  recentIncomingMsgIds = new utils.LruSet(1e3, longUtils.LongSet);
  // <deno-tsignore>
  recentStateRequests = new utils.LruMap(1e3, longUtils.LongMap);
  // queues
  queuedRpc = new utils.Deque();
  queuedAcks = [];
  queuedStateReq = [];
  queuedResendReq = [];
  queuedCancelReq = [];
  getStateSchedule = new sortedArray.SortedArray([], (a, b) => {
    const diff = a.getState - b.getState;
    if (diff !== 0) return diff;
    if (a.msgId && b.msgId) return a.msgId.compare(b.msgId);
    return 0;
  });
  pendingGetStateTimeouts = new longUtils.LongMap();
  chains = /* @__PURE__ */ new Map();
  chainsPendingFails = /* @__PURE__ */ new Map();
  // requests info
  pendingMessages = new longUtils.LongMap();
  lastPingRtt = Number.NaN;
  lastPingTime = 0;
  lastPingMsgId = Long.ZERO;
  lastSessionCreatedUid = Long.ZERO;
  initConnectionCalled = false;
  authorizationPending = false;
  get hasPendingMessages() {
    return Boolean(
      this.queuedRpc.length || this.queuedAcks.length || this.queuedStateReq.length || this.queuedResendReq.length || this.queuedCancelReq.length
    );
  }
  /**
   * Reset session by resetting auth key(s) and session state
   */
  reset(withAuthKey = false) {
    if (withAuthKey) {
      this.resetAuthKey();
    }
    this.resetState(withAuthKey);
  }
  resetAuthKey() {
    this._authKey.reset();
    this._authKeyTemp.reset();
    this._authKeyTempSecondary.reset();
  }
  updateTimeOffset(offset) {
    this.log.debug("time offset updated: %d", offset);
    this._timeOffset = offset;
    this._lastMessageId = Long.ZERO;
  }
  /**
   * Reset session state and generate a new session ID.
   *
   * By default, also cancels any pending RPC requests.
   * If `keepPending` is set to `true`, pending requests will be kept
   */
  resetState(keepPending = false) {
    this._lastMessageId = Long.ZERO;
    this._seqNo = 0;
    this._sessionId = longUtils.randomLong();
    this.log.debug("session reset, new sid = %h", this._sessionId);
    this.log.prefix = `[SESSION ${this._sessionId.toString(16)}] `;
    if (!keepPending) {
      for (const info of this.pendingMessages.values()) {
        if (info._ === "rpc") {
          this.log.debug("rejecting pending rpc %s", info.rpc.method);
          info.rpc.promise.reject(new errors.MtcuteError("Session is reset"));
        }
      }
      this.pendingMessages.clear();
    }
    this.recentOutgoingMsgIds.clear();
    this.recentIncomingMsgIds.clear();
    this.recentStateRequests.clear();
    if (!keepPending) {
      while (this.queuedRpc.length) {
        const rpc = this.queuedRpc.popFront();
        if (rpc.sent === false) {
          this.log.debug("rejecting pending rpc %s", rpc.method);
          rpc.promise.reject(new errors.MtcuteError("Session is reset"));
        }
      }
    }
    this.queuedAcks.length = 0;
    this.queuedStateReq.length = 0;
    this.queuedResendReq.length = 0;
    this.queuedCancelReq.length = 0;
    this.getStateSchedule.clear();
    this.chains.clear();
    this.chainsPendingFails.clear();
    this.resetLastPing(true);
  }
  enqueueRpc(rpc, force) {
    if (!force && !rpc.sent || rpc.cancelled) return false;
    rpc.sent = false;
    rpc.containerId = void 0;
    this.log.debug("enqueued %s for sending (msg_id = %s)", rpc.method, rpc.msgId || "n/a");
    this.queuedRpc.pushBack(rpc);
    return true;
  }
  getMessageId() {
    const timeTicks = performance.now();
    const timeSec = Math.floor(timeTicks / 1e3) + this._timeOffset;
    const timeMSec = timeTicks % 1e3;
    const random = miscUtils.getRandomInt(65535);
    let messageId = new Long(timeMSec << 21 | random << 3 | 4, timeSec);
    if (this._lastMessageId.ge(messageId)) {
      messageId = this._lastMessageId.add(4);
    }
    this._lastMessageId = messageId;
    return messageId;
  }
  getSeqNo(isContentRelated = true) {
    let seqNo = this._seqNo;
    if (isContentRelated) {
      seqNo += 1;
      this._seqNo += 2;
    }
    return seqNo;
  }
  /** Encrypt a single MTProto message using session's keys */
  encryptMessage(message) {
    const key = this._authKeyTemp.ready ? this._authKeyTemp : this._authKey;
    return key.encryptMessage(message, this._salts.currentSalt, this._sessionId);
  }
  /** Decrypt a single MTProto message using session's keys */
  decryptMessage(data, callback) {
    if (!this._authKey.ready) throw new errors.MtcuteError("Keys are not set up!");
    const authKeyId = data.subarray(0, 8);
    let key;
    if (this._authKey.match(authKeyId)) {
      key = this._authKey;
    } else if (this._authKeyTemp.match(authKeyId)) {
      key = this._authKeyTemp;
    } else if (this._authKeyTempSecondary.match(authKeyId)) {
      key = this._authKeyTempSecondary;
    } else {
      this.log.warn(
        "received message with unknown authKey = %h (expected %h or %h or %h)",
        authKeyId,
        this._authKey.id,
        this._authKeyTemp.id,
        this._authKeyTempSecondary.id
      );
      return;
    }
    return key.decryptMessage(data, this._sessionId, callback);
  }
  writeMessage(writer, content, isContentRelated = true) {
    const messageId = this.getMessageId();
    const seqNo = this.getSeqNo(isContentRelated);
    const length = ArrayBuffer.isView(content) ? content.length : tlRuntime.TlSerializationCounter.countNeededBytes(writer.objectMap, content);
    writer.long(messageId);
    writer.int(seqNo);
    writer.uint(length);
    if (ArrayBuffer.isView(content)) writer.raw(content);
    else writer.object(content);
    return messageId;
  }
  resetLastPing(withTime = false) {
    if (withTime) this.lastPingTime = 0;
    if (!this.lastPingMsgId.isZero()) {
      this.pendingMessages.delete(this.lastPingMsgId);
    }
    this.lastPingMsgId = Long.ZERO;
  }
  addToChain(chainId, msgId) {
    const prevMsgId = this.chains.get(chainId);
    this.chains.set(chainId, msgId);
    this.log.debug("added message %l to chain %s (prev: %l)", msgId, chainId, prevMsgId);
    return prevMsgId;
  }
  removeFromChain(chainId, msgId) {
    const lastMsgId = this.chains.get(chainId);
    if (!lastMsgId) {
      this.log.warn("tried to remove message %l from empty chain %s", msgId, chainId);
      return;
    }
    if (lastMsgId.eq(msgId)) {
      this.log.debug("chain %s: exhausted, last message %l", chainId, msgId);
      this.chains.delete(chainId);
    }
  }
  getPendingChainedFails(chainId) {
    let arr = this.chainsPendingFails.get(chainId);
    if (!arr) {
      arr = new sortedArray.SortedArray([], (a, b) => longUtils.compareLongs(a.invokeAfter, b.invokeAfter));
      this.chainsPendingFails.set(chainId, arr);
    }
    return arr;
  }
}
exports.MtprotoSession = MtprotoSession;
