// @mtcute/core/network/multi-session-connection.js
import { Emitter, Deferred, unknownToError } from "@fuman/utils";
import { SessionConnection } from "./session-connection.js";
class MultiSessionConnection {
  constructor(params, _count, log, logPrefix = "") {
    this.params = params;
    this._count = _count;
    this._log = log.create("multi");
    if (logPrefix) this._log.prefix = `[${logPrefix}] `;
    this._enforcePfs = _count > 1 && params.isMainConnection;
    this._updateConnections();
  }
  static _pickConnection(loads, ready) {
    let min = Infinity;
    let minIdx = 0;
    let foundReady = false;
    for (let i = 0; i < loads.length; i++) {
      const isReady = ready[i];
      if (foundReady && !isReady) continue;
      if (isReady && !foundReady) {
        foundReady = true;
        min = Infinity;
      }
      if (loads[i] < min) {
        min = loads[i];
        minIdx = i;
      }
    }
    return minIdx;
  }
  _log;
  _enforcePfs = false;
  // NB: dont forget to update .reset()
  onRequestKeys = new Emitter();
  onError = new Emitter();
  onUpdate = new Emitter();
  onKeyChange = new Emitter();
  onTmpKeyChange = new Emitter();
  onFutureSalts = new Emitter();
  onAuthBegin = new Emitter();
  onUsable = new Emitter();
  onWait = new Emitter();
  onRequestAuth = new Emitter();
  _connections = [];
  setCount(count, connect = this.params.isMainConnection) {
    this._count = count;
    this._updateConnections(connect);
  }
  getCount() {
    return this._count;
  }
  _updateConnections(connect = false) {
    if (this._connections.length === this._count) return;
    this._log.debug("updating connections count: %d -> %d", this._connections.length, this._count);
    const newEnforcePfs = this._count > 1 && this.params.isMainConnection;
    const enforcePfsChanged = newEnforcePfs !== this._enforcePfs;
    if (enforcePfsChanged) {
      this._log.debug("enforcePfs changed: %s -> %s", this._enforcePfs, newEnforcePfs);
      this._enforcePfs = newEnforcePfs;
    }
    if (this._connections.length > this._count) {
      for (let i = this._connections.length - 1; i >= this._count; i--) {
        this._connections[i].destroy().catch((err) => {
          this._log.warn("error destroying connection: %e", err);
        });
      }
      this._connections.splice(this._count);
      return;
    }
    if (enforcePfsChanged) {
      const promise = new Deferred();
      this.onRequestKeys.emit(promise);
      promise.promise.then(() => {
        this._connections.forEach((conn) => {
          conn.setUsePfs(this.params.usePfs || this._enforcePfs);
          if (connect) conn.connect();
        });
      }).catch((err) => {
        this.onError.emit(unknownToError(err));
      });
    }
    for (let i = this._connections.length; i < this._count; i++) {
      const conn = new SessionConnection(
        {
          ...this.params,
          usePfs: this.params.usePfs || this._enforcePfs,
          isMainConnection: this.params.isMainConnection && i === 0,
          withUpdates: this.params.isMainConnection && this.params.isMainDcConnection && !this.params.disableUpdates
        },
        this._log
      );
      if (this.params.isMainConnection && this.params.isMainDcConnection) {
        conn.onUpdate.add((update) => this.onUpdate.emit(update));
      }
      conn.onError.add((err) => this.onError.emit(err));
      conn.onKeyChange.add((key) => {
        this.onKeyChange.emit([i, key]);
        for (const conn_ of this._connections) {
          if (conn_ === conn) continue;
          conn_.onConnected();
        }
      });
      conn.onTmpKeyChange.add(
        (event) => this.onTmpKeyChange.emit(event === null ? [i, null, 0] : [i, event[0], event[1]])
      );
      conn.onFutureSalts.add((salts) => this.onFutureSalts.emit(salts));
      conn.onAuthBegin.add(() => {
        this._log.debug("received auth-begin from connection %d", i);
        this.onAuthBegin.emit(i);
        this._connections.forEach((conn_) => {
          conn_._session._authKeyTemp.reset();
          if (conn_ !== conn) conn_.reconnect();
        });
      });
      conn.onUsable.add(() => this.onUsable.emit(i));
      conn.onWait.add(() => this.onWait.emit(i));
      conn.onRequestAuth.add(() => this.onRequestAuth.emit(i));
      this._connections.push(conn);
      if (this._authKey !== null) conn._session._authKey.setup(this._authKey);
      if (connect && !enforcePfsChanged) conn.connect();
    }
  }
  _destroyed = false;
  async destroy() {
    await Promise.all(this._connections.map((conn) => conn.destroy()));
    this.onRequestKeys.clear();
    this.onError.clear();
    this.onUpdate.clear();
    this.onKeyChange.clear();
    this.onTmpKeyChange.clear();
    this.onFutureSalts.clear();
    this.onAuthBegin.clear();
    this.onUsable.clear();
    this.onWait.clear();
    this.onRequestAuth.clear();
    this._destroyed = true;
  }
  sendRpc(request, timeout, abortSignal, chainId) {
    const loads = this._connections.map((c) => c._session.queuedRpc.length + c._session.pendingMessages.size);
    const ready = this._connections.map((c) => c.isConnected);
    return this._connections[MultiSessionConnection._pickConnection(loads, ready)].sendRpc(
      request,
      timeout,
      abortSignal,
      chainId
    );
  }
  connect() {
    for (const conn of this._connections) {
      conn.connect();
    }
  }
  ensureConnected() {
    if (this._connections[0].isConnected) return;
    this.connect();
  }
  _authKey = null;
  setAuthKey(authKey) {
    this._authKey = authKey;
    for (const conn of this._connections) {
      conn._session._authKey.setup(authKey);
    }
  }
  setTempAuthKey(authKey, idx) {
    this._connections[idx]?._session._authKeyTemp.setup(authKey);
  }
  resetAuthKeys() {
    for (const conn of this._connections) {
      conn._session.reset(true);
    }
    this.notifyKeyChange();
  }
  setInactivityTimeout(timeout) {
    this._log.debug("setting inactivity timeout to %s", timeout);
    this.params.inactivityTimeout = timeout;
    for (const conn of this._connections) {
      conn.setInactivityTimeout(timeout);
    }
  }
  notifyKeyChange() {
    for (const conn of this._connections) {
      const session = conn._session;
      if (this.params.usePfs && !session._authKeyTemp.ready) {
        this._log.debug("temp auth key needed but not ready, ignoring key change");
        continue;
      }
      if (session.queuedRpc.length) {
        this._log.debug("notifying key change on the connection due to queued rpc");
        this._connections.forEach((conn2) => conn2.onConnected());
      }
    }
  }
  notifyNetworkChanged(connected) {
    for (const conn of this._connections) {
      conn.notifyNetworkChanged(connected);
    }
  }
  requestAuth() {
    this._connections[0]._authorize();
  }
  resetSessions() {
    if (this.params.isMainConnection) {
      for (const conn of this._connections) {
        conn._resetSession();
      }
    } else {
      this._connections[0]._resetSession();
    }
  }
  async changeTransport(factory) {
    await Promise.all(this._connections.map((conn) => conn.changeTransport(factory)));
  }
  getPoolSize() {
    return this._connections.length;
  }
}
export {
  MultiSessionConnection
};
