// @mtcute/core/network/persistent-connection.js
import { FramedReader, FramedWriter } from "@fuman/io";
import { PersistentConnection as PersistentConnection$1, ip, ConnectionClosedError } from "@fuman/net";
import { Emitter, timers } from "@fuman/utils";
import { ConnectionFloodController } from "./flood-control.js";
let nextConnectionUid = 0;
class PersistentConnection {
  constructor(params, log) {
    this.log = log;
    this.params = params;
    this.params.transport.setup?.(this.params.crypto, log);
    this._codec = this.params.transport.packetCodec(params.dc);
    this._codec.setup?.(this.params.crypto, this.log);
    if (params.floodControl) {
      this._floodControl = new ConnectionFloodController(params.floodControl);
      this._floodControl.setLogger(log.create("flood"));
    }
    this._onInactivityTimeout = this._onInactivityTimeout.bind(this);
    this._fuman = new PersistentConnection$1({
      connect: async (dc) => {
        if (this._floodControl) {
          if (!this._connectAbort || this._connectAbort.signal.aborted) {
            this._connectAbort = new AbortController();
          }
          this._updateLogPrefix();
          await this._floodControl.wait(this._connectAbort.signal);
        }
        this._updateLogPrefix();
        this.log.debug("connecting to %j", dc);
        return params.transport.connect(dc);
      },
      strategy: params.reconnectionStrategy,
      onOpen: this._onOpen.bind(this),
      onClose: this._onClose.bind(this),
      onError: this._onError.bind(this),
      onWait: (wait) => {
        this._updateLogPrefix();
        this.log.debug("waiting for %d ms before reconnecting", wait);
        this.onWait.emit(wait);
      }
    });
    this._updateLogPrefix();
  }
  _uid = nextConnectionUid++;
  params;
  _sendOnceConnected = [];
  _codec;
  _fuman;
  // reconnection
  _disconnectedManually = false;
  // inactivity timeout
  _inactivityTimeout = null;
  _inactive = true;
  _destroyed = false;
  _usable = false;
  // each connection slot throttles its own connect attempts, mirroring
  // TDLib's per-SessionProxy flood control. a fresh controller has empty
  // limiters, so the first connect always passes immediately.
  _floodControl;
  // aborts a pending flood-control wait when the connection is being
  // closed (manually or for good) — otherwise close() would block until
  // the wait expires, and the connect would still proceed afterwards
  _connectAbort;
  onWait = new Emitter();
  onUsable = new Emitter();
  onError = new Emitter();
  _updateLogPrefix() {
    const uidPrefix = `[UID ${this._uid}] `;
    if (this._fuman.isConnected) {
      const dc = this.params.dc;
      const prettifiedIp = ip.prettify(dc.ipAddress, { encloseIpv6: true });
      this.log.prefix = `${uidPrefix}[DC ${dc.id} @ ${prettifiedIp}:${dc.port}] `;
    } else if (this._fuman.isConnecting) {
      this.log.prefix = `${uidPrefix}[connecting] `;
    } else {
      this.log.prefix = `${uidPrefix}[disconnected] `;
    }
  }
  get isConnected() {
    return this._fuman.isConnected;
  }
  _writer;
  _mtproxyInfo;
  async _onOpen(conn) {
    this._updateLogPrefix();
    this.log.debug("connected");
    const tag = await this._codec.tag();
    if (tag) {
      await conn.write(tag);
    }
    this._mtproxyInfo = conn.getMtproxyInfo?.();
    const reader = new FramedReader(conn, this._codec);
    this._writer = new FramedWriter(conn, this._codec);
    while (this._sendOnceConnected.length) {
      const data = this._sendOnceConnected.shift();
      try {
        await this._writer.write(data);
      } catch (e) {
        this._sendOnceConnected.unshift(data);
        throw e;
      }
    }
    this._rescheduleInactivity();
    this.onUsable.emit();
    this.onConnected();
    while (true) {
      const msg = await reader.read();
      if (msg) {
        this.onMessage(msg);
      }
    }
  }
  _onClose() {
    this.log.debug("connection closed");
    this._updateLogPrefix();
    this._writer = void 0;
    this._codec.reset();
    this.onClosed();
  }
  _onError(err) {
    this._updateLogPrefix();
    if (!(err instanceof ConnectionClosedError)) {
      this.handleError(err);
    }
    return "reconnect";
  }
  async changeTransport(transport) {
    this._connectAbort?.abort();
    await this._fuman.close();
    this._codec = transport.packetCodec(this.params.dc);
    this._codec.setup?.(this.params.crypto, this.log);
    await this._fuman.changeTransport(transport.connect.bind(transport));
    this._fuman.connect(this.params.dc);
  }
  connect() {
    this._fuman.connect(this.params.dc);
    this._inactive = false;
  }
  reconnect() {
    if (this._disconnectedManually) {
      this._disconnectedManually = false;
      this.connect();
      return;
    }
    this._fuman.reconnect(true);
  }
  async disconnectManual() {
    if (this._inactivityTimeout) {
      timers.clearTimeout(this._inactivityTimeout);
    }
    this._disconnectedManually = true;
    this._connectAbort?.abort();
    await this._fuman.close();
  }
  async destroy() {
    if (this._inactivityTimeout) {
      timers.clearTimeout(this._inactivityTimeout);
    }
    this._destroyed = true;
    this._connectAbort?.abort();
    await this._fuman.close();
  }
  _rescheduleInactivity() {
    if (!this.params.inactivityTimeout) return;
    if (this._inactivityTimeout) timers.clearTimeout(this._inactivityTimeout);
    this._inactivityTimeout = timers.setTimeout(this._onInactivityTimeout, this.params.inactivityTimeout);
  }
  _onInactivityTimeout() {
    this.log.info("disconnected because of inactivity for %d", this.params.inactivityTimeout);
    this._inactive = true;
    this._inactivityTimeout = null;
    this._connectAbort?.abort();
    Promise.resolve(this._fuman.close()).catch((err) => {
      this.log.warn("error closing transport: %e", err);
    });
  }
  setInactivityTimeout(timeout) {
    this.params.inactivityTimeout = timeout;
    if (this._inactivityTimeout) {
      timers.clearTimeout(this._inactivityTimeout);
    }
    if (timeout) {
      this._rescheduleInactivity();
    }
  }
  async send(data) {
    if (this._inactive) {
      this.connect();
    }
    if (this._writer) {
      this._rescheduleInactivity();
      try {
        await this._writer.write(data);
      } catch (e) {
        if (!(e instanceof ConnectionClosedError)) {
          this.log.warn("encountered an error while sending, reconnecting: %e", e);
          this._writer = void 0;
          this._fuman.reconnect(true);
          this._sendOnceConnected.push(data);
        }
      }
    } else {
      this._sendOnceConnected.push(data);
    }
  }
}
export {
  PersistentConnection
};
