// @mtcute/core/highlevel/types/misc/input-privacy-rule/disallow.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const all = { _: "inputPrivacyValueDisallowAll" };
const contacts = { _: "inputPrivacyValueDisallowContacts" };
function users(users2) {
  return {
    allow: false,
    users: Array.isArray(users2) ? users2 : [users2]
  };
}
function chatParticipants(chats) {
  return {
    allow: false,
    chats: Array.isArray(chats) ? chats : [chats]
  };
}
exports.all = all;
exports.chatParticipants = chatParticipants;
exports.contacts = contacts;
exports.users = users;
