// @fuman/io/bytes.js
import { u8 } from "@fuman/utils";
import { nextPowerOfTwo } from "./_utils.js";
class Bytes {
  #buffer;
  #writePos = 0;
  #readPos = 0;
  #preferredCapacity;
  constructor(buf) {
    this.#buffer = buf;
    this.#preferredCapacity = buf.length;
  }
  static alloc(capacity = 1024 * 16) {
    const bytes = new Bytes(u8.alloc(capacity));
    return bytes;
  }
  static from(data) {
    const bytes = new Bytes(data);
    bytes.#writePos = data.length;
    return bytes;
  }
  /** Total number of bytes in the underlying buffer */
  get capacity() {
    return this.#buffer.byteLength;
  }
  /** Number of bytes available to be read */
  get available() {
    return this.#writePos - this.#readPos;
  }
  /** Number of bytes written */
  get written() {
    return this.#writePos;
  }
  #sharedRead = new Uint8Array(1);
  readSync(bytes) {
    if (this.#readPos >= this.#writePos) {
      return u8.empty;
    }
    if (bytes === 1) {
      this.#sharedRead[0] = this.#buffer[this.#readPos++];
      return this.#sharedRead;
    }
    const end = Math.min(this.#writePos, this.#readPos + bytes);
    const result = this.#buffer.subarray(this.#readPos, end);
    this.#readPos = end;
    return result;
  }
  async read(into) {
    const size = Math.min(into.length, this.#writePos - this.#readPos);
    into.set(this.#buffer.subarray(this.#readPos, this.#readPos + size));
    this.#readPos += size;
    return size;
  }
  #lastWriteSize = 0;
  writeSync(size) {
    this.#lastWriteSize = size;
    const newPos = this.#writePos + size;
    if (newPos > this.#buffer.length) {
      const newBuffer = u8.alloc(nextPowerOfTwo(newPos));
      newBuffer.set(this.#buffer);
      this.#buffer = newBuffer;
    }
    const slice = this.#buffer.subarray(this.#writePos, newPos);
    this.#writePos = newPos;
    return slice;
  }
  disposeWriteSync(written) {
    if (written !== void 0) {
      if (written > this.#lastWriteSize) {
        throw new RangeError(`written exceeds last write size: ${written} > ${this.#lastWriteSize}`);
      }
      this.#writePos -= this.#lastWriteSize - written;
    }
  }
  async write(bytes) {
    this.writeSync(bytes.length).set(bytes);
    this.disposeWriteSync();
  }
  /**
   * get the "result" of the buffer, i.e. everything that has been written so far,
   * but not yet read
   *
   * **Note**: this method returns a view into the underlying buffer, and does advance the read cursor
   */
  result() {
    return this.#buffer.subarray(this.#readPos, this.#writePos);
  }
  /** Reclaim memory by only keeping the yet-unread data */
  reclaim() {
    if (this.#readPos === 0) return;
    const remaining = this.#writePos - this.#readPos;
    if (remaining > 0) {
      if (remaining < this.#preferredCapacity && this.capacity > this.#preferredCapacity) {
        const newBuffer = u8.alloc(this.#preferredCapacity);
        newBuffer.set(this.#buffer.subarray(this.#readPos, this.#writePos));
        this.#buffer = newBuffer;
      } else {
        this.#buffer.copyWithin(0, this.#readPos, this.#writePos);
      }
    }
    this.#writePos = remaining;
    this.#readPos = 0;
  }
  /** Mark last n bytes as unread */
  rewind(n) {
    if (n > this.#readPos) {
      throw new RangeError(`rewind: ${n} > ${this.#readPos}`);
    }
    this.#readPos -= n;
  }
  /** reset the read/write cursors */
  reset() {
    this.#readPos = 0;
    this.#writePos = 0;
  }
}
export {
  Bytes
};
