// @fuman/io/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const index = require("./read/index.cjs");
const index$1 = require("./write/index.cjs");
const bufReader = require("./buf-reader.cjs");
const bytes = require("./bytes.cjs");
const errors = require("./errors.cjs");
const adapters = require("./read/adapters.cjs");
const readerWithFinal = require("./reader-with-final.cjs");
const adapters$1 = require("./write/adapters.cjs");
const reader = require("./bits/reader.cjs");
const utils = require("./bits/utils.cjs");
const delimiter = require("./codec/delimiter.cjs");
const lengthDelimited = require("./codec/length-delimited.cjs");
const reader$1 = require("./codec/reader.cjs");
const textDelimiter = require("./codec/text-delimiter.cjs");
const writer = require("./codec/writer.cjs");
exports.read = index;
exports.write = index$1;
exports.BufReader = bufReader.BufReader;
exports.Bytes = bytes.Bytes;
exports.PartialReadError = errors.PartialReadError;
exports.fumanReadableToWeb = adapters.fumanReadableToWeb;
exports.fumanSyncReadableToAsync = adapters.fumanSyncReadableToAsync;
exports.webReadableToFuman = adapters.webReadableToFuman;
exports.ReaderWithFinal = readerWithFinal.ReaderWithFinal;
exports.fumanSyncWritableToAsync = adapters$1.fumanSyncWritableToAsync;
exports.fumanWritableToWeb = adapters$1.fumanWritableToWeb;
exports.webWritableToFuman = adapters$1.webWritableToFuman;
exports.BitReader = reader.BitReader;
exports.reverse8Bits = utils.reverse8Bits;
exports.reverseBits = utils.reverseBits;
exports.reverseBitsAll = utils.reverseBitsAll;
exports.reverseBitsBig = utils.reverseBitsBig;
exports.DelimiterCodec = delimiter.DelimiterCodec;
exports.LengthDelimitedCodec = lengthDelimited.LengthDelimitedCodec;
exports.FramedReader = reader$1.FramedReader;
exports.TextDelimiterCodec = textDelimiter.TextDelimiterCodec;
exports.FramedWriter = writer.FramedWriter;
