// @mtcute/core/highlevel/methods/gifts/get-resale-star-gifts.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const Long = require("long");
const longUtils = require("../../../utils/long-utils.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const starGiftUnique = require("../../types/premium/star-gift-unique.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
function _normalizeResaleAttributes(attrs) {
  if (Array.isArray(attrs)) {
    return attrs;
  }
  const res = [];
  if (attrs.model) {
    if (Array.isArray(attrs.model)) {
      for (const id of attrs.model) {
        res.push({
          _: "starGiftAttributeIdModel",
          documentId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdModel",
        documentId: attrs.model
      });
    }
  }
  if (attrs.pattern) {
    if (Array.isArray(attrs.pattern)) {
      for (const id of attrs.pattern) {
        res.push({
          _: "starGiftAttributeIdPattern",
          documentId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdPattern",
        documentId: attrs.pattern
      });
    }
  }
  if (attrs.backdrop) {
    if (Array.isArray(attrs.backdrop)) {
      for (const id of attrs.backdrop) {
        res.push({
          _: "starGiftAttributeIdBackdrop",
          backdropId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdBackdrop",
        backdropId: attrs.backdrop
      });
    }
  }
  return res;
}
async function getStarGiftResaleOptions(client, params) {
  const res = await client.call({
    _: "payments.getResaleStarGifts",
    sortByNum: params.sort === "num",
    sortByPrice: params.sort === "price",
    attributesHash: params.attributesHash,
    giftId: params.giftId,
    attributes: params.attributes ? _normalizeResaleAttributes(params.attributes) : void 0,
    offset: params.offset ?? "",
    limit: params.limit ?? 100,
    forCraft: params.forCraft
  });
  const peers = peersIndex.PeersIndex.from(res);
  const attributes = {
    model: new longUtils.LongMap(),
    pattern: new longUtils.LongMap(),
    backdrop: /* @__PURE__ */ new Map()
  };
  for (const attr of res.attributes ?? []) {
    if (attr._ === "starGiftAttributeModel") {
      attributes.model.set(attr.document.id, new starGiftUnique.StarGiftUniqueAttribute(attr));
    } else if (attr._ === "starGiftAttributePattern") {
      attributes.pattern.set(attr.document.id, new starGiftUnique.StarGiftUniqueAttribute(attr));
    } else if (attr._ === "starGiftAttributeBackdrop") {
      attributes.backdrop.set(attr.backdropId, new starGiftUnique.StarGiftUniqueBackdrop(attr));
    }
  }
  return miscUtils.makeArrayPaginatedWithMeta(
    res.gifts.map((it) => {
      utils.assert(it._ === "starGiftUnique");
      return new starGiftUnique.StarGiftUnique(it, peers);
    }),
    res.count,
    {
      attributes,
      attributesHash: res.attributesHash ?? params.attributesHash ?? Long.ZERO,
      counters: res.counters ?? []
    },
    res.nextOffset
  );
}
exports._normalizeResaleAttributes = _normalizeResaleAttributes;
exports.getStarGiftResaleOptions = getStarGiftResaleOptions;
