// @mtcute/core/highlevel/methods/gifts/get-saved-star-gifts.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const savedStarGift = require("../../types/premium/saved-star-gift.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
async function getSavedStarGifts(client, params) {
  const res = await client.call({
    _: "payments.getSavedStarGifts",
    peer: await resolvePeer.resolvePeer(client, params.owner),
    excludeUnsaved: params.excludeHidden,
    excludeSaved: params.excludePublic,
    excludeUnlimited: params.excludeUnlimited,
    excludeUnupgradable: params.excludeUnupgradable,
    excludeUpgradable: params.excludeUpgradable,
    excludeUnique: params.excludeUnique,
    peerColorAvailable: params.peerColorAvailable,
    sortByValue: params.sortByValue,
    collectionId: params.collectionId,
    offset: params.offset ?? "",
    limit: params.limit ?? 100
  });
  const peers = peersIndex.PeersIndex.from(res);
  return miscUtils.makeArrayPaginated(
    res.gifts.map((it) => new savedStarGift.SavedStarGift(it, peers)),
    res.count,
    res.nextOffset
  );
}
exports.getSavedStarGifts = getSavedStarGifts;
