// @mtcute/core/highlevel/methods/gifts/get-saved-star-gifts.js
import { SavedStarGift as SavedStarGift$2 } from "../../types/premium/saved-star-gift.js";
import { makeArrayPaginated } from "../../utils/misc-utils.js";
import { resolvePeer } from "../users/resolve-peer.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
async function getSavedStarGifts(client, params) {
  const res = await client.call({
    _: "payments.getSavedStarGifts",
    peer: await resolvePeer(client, params.owner),
    excludeUnsaved: params.excludeHidden,
    excludeSaved: params.excludePublic,
    excludeUnlimited: params.excludeUnlimited,
    excludeUnupgradable: params.excludeUnupgradable,
    excludeUpgradable: params.excludeUpgradable,
    excludeUnique: params.excludeUnique,
    peerColorAvailable: params.peerColorAvailable,
    sortByValue: params.sortByValue,
    collectionId: params.collectionId,
    offset: params.offset ?? "",
    limit: params.limit ?? 100
  });
  const peers = PeersIndex.from(res);
  return makeArrayPaginated(
    res.gifts.map((it) => new SavedStarGift$2(it, peers)),
    res.count,
    res.nextOffset
  );
}
export {
  getSavedStarGifts
};
