// @mtcute/core/highlevel/methods/gifts/get-resale-star-gifts.js
import { assert } from "@fuman/utils";
import Long from "long";
import { LongMap } from "../../../utils/long-utils.js";
import { makeArrayPaginatedWithMeta } from "../../utils/misc-utils.js";
import { StarGiftUniqueAttribute as StarGiftUniqueAttribute$2, StarGiftUniqueBackdrop as StarGiftUniqueBackdrop$1, StarGiftUnique as StarGiftUnique$2 } from "../../types/premium/star-gift-unique.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
function _normalizeResaleAttributes(attrs) {
  if (Array.isArray(attrs)) {
    return attrs;
  }
  const res = [];
  if (attrs.model) {
    if (Array.isArray(attrs.model)) {
      for (const id of attrs.model) {
        res.push({
          _: "starGiftAttributeIdModel",
          documentId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdModel",
        documentId: attrs.model
      });
    }
  }
  if (attrs.pattern) {
    if (Array.isArray(attrs.pattern)) {
      for (const id of attrs.pattern) {
        res.push({
          _: "starGiftAttributeIdPattern",
          documentId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdPattern",
        documentId: attrs.pattern
      });
    }
  }
  if (attrs.backdrop) {
    if (Array.isArray(attrs.backdrop)) {
      for (const id of attrs.backdrop) {
        res.push({
          _: "starGiftAttributeIdBackdrop",
          backdropId: id
        });
      }
    } else {
      res.push({
        _: "starGiftAttributeIdBackdrop",
        backdropId: attrs.backdrop
      });
    }
  }
  return res;
}
async function getStarGiftResaleOptions(client, params) {
  const res = await client.call({
    _: "payments.getResaleStarGifts",
    sortByNum: params.sort === "num",
    sortByPrice: params.sort === "price",
    attributesHash: params.attributesHash,
    giftId: params.giftId,
    attributes: params.attributes ? _normalizeResaleAttributes(params.attributes) : void 0,
    offset: params.offset ?? "",
    limit: params.limit ?? 100,
    forCraft: params.forCraft
  });
  const peers = PeersIndex.from(res);
  const attributes = {
    model: new LongMap(),
    pattern: new LongMap(),
    backdrop: /* @__PURE__ */ new Map()
  };
  for (const attr of res.attributes ?? []) {
    if (attr._ === "starGiftAttributeModel") {
      attributes.model.set(attr.document.id, new StarGiftUniqueAttribute$2(attr));
    } else if (attr._ === "starGiftAttributePattern") {
      attributes.pattern.set(attr.document.id, new StarGiftUniqueAttribute$2(attr));
    } else if (attr._ === "starGiftAttributeBackdrop") {
      attributes.backdrop.set(attr.backdropId, new StarGiftUniqueBackdrop$1(attr));
    }
  }
  return makeArrayPaginatedWithMeta(
    res.gifts.map((it) => {
      assert(it._ === "starGiftUnique");
      return new StarGiftUnique$2(it, peers);
    }),
    res.count,
    {
      attributes,
      attributesHash: res.attributesHash ?? params.attributesHash ?? Long.ZERO,
      counters: res.counters ?? []
    },
    res.nextOffset
  );
}
export {
  _normalizeResaleAttributes,
  getStarGiftResaleOptions
};
