// @mtcute/core/highlevel/methods/gifts/send-star-gift.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const entities = require("../../types/misc/entities.cjs");
const findInUpdate = require("../messages/find-in-update.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
async function sendStarGift(client, params) {
  const {
    peerId,
    gift,
    anonymous,
    message,
    shouldDispatch,
    withUpgrade
  } = params;
  const invoice = {
    _: "inputInvoiceStarGift",
    hideName: anonymous,
    peer: await resolvePeer.resolvePeer(client, peerId),
    giftId: Long.isLong(gift) ? gift : gift.id,
    message: message ? entities.inputTextToTl(message) : void 0,
    includeUpgrade: withUpgrade
  };
  const form = await client.call({
    _: "payments.getPaymentForm",
    invoice
  });
  const res = await client.call({
    _: "payments.sendStarsForm",
    invoice,
    formId: form.formId
  });
  typeAssertions.assertTypeIs("payments.sendStarsForm", res, "payments.paymentResult");
  return findInUpdate._findMessageInUpdate(client, res.updates, false, !shouldDispatch, true);
}
exports.sendStarGift = sendStarGift;
