// @mtcute/core/highlevel/methods/gifts/send-star-gift.js
import Long from "long";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { inputTextToTl } from "../../types/misc/entities.js";
import { _findMessageInUpdate } from "../messages/find-in-update.js";
import { resolvePeer } from "../users/resolve-peer.js";
async function sendStarGift(client, params) {
  const {
    peerId,
    gift,
    anonymous,
    message,
    shouldDispatch,
    withUpgrade
  } = params;
  const invoice = {
    _: "inputInvoiceStarGift",
    hideName: anonymous,
    peer: await resolvePeer(client, peerId),
    giftId: Long.isLong(gift) ? gift : gift.id,
    message: message ? inputTextToTl(message) : void 0,
    includeUpgrade: withUpgrade
  };
  const form = await client.call({
    _: "payments.getPaymentForm",
    invoice
  });
  const res = await client.call({
    _: "payments.sendStarsForm",
    invoice,
    formId: form.formId
  });
  assertTypeIs("payments.sendStarsForm", res, "payments.paymentResult");
  return _findMessageInUpdate(client, res.updates, false, !shouldDispatch, true);
}
export {
  sendStarGift
};
