// @mtcute/core/highlevel/types/peers/bot-info.js
import { asNonNull } from "@fuman/utils";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parseDocument } from "../media/document-utils.js";
import { Photo as Photo$2 } from "../media/photo.js";
class BotInfo {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether the bot has preview medias available */
  get hasPreviewMedia() {
    return this.raw.hasPreviewMedias;
  }
  /** ID of the bot */
  get id() {
    return asNonNull(this.raw.userId);
  }
  get description() {
    return this.raw.description ?? "";
  }
  get descriptionPhoto() {
    if (!this.raw.descriptionPhoto || this.raw.descriptionPhoto._ !== "photo") return null;
    return new Photo$2(this.raw.descriptionPhoto);
  }
  get descriptionVideo() {
    if (!this.raw.descriptionDocument || this.raw.descriptionDocument._ !== "document") return null;
    const parsed = parseDocument(this.raw.descriptionDocument);
    if (parsed.type !== "video") return null;
    return parsed;
  }
  /** List of the bot's registered commands */
  get commands() {
    return this.raw.commands ?? [];
  }
  /** Action to be performed when the bot's menu button is pressed */
  get menuButton() {
    return this.raw.menuButton ?? null;
  }
  /** URL of the bot's privacy policy */
  get privacyPolicyUrl() {
    return this.raw.privacyPolicyUrl ?? null;
  }
  get verifierSettings() {
    return this.raw.verifierSettings ?? null;
  }
}
const BotInfo$1 = /* @__PURE__ */ makeInspectable(BotInfo);
const BotInfo$2 = /* @__PURE__ */ memoizeGetters(BotInfo$1, ["descriptionPhoto", "descriptionVideo"]);
export {
  BotInfo$2 as BotInfo
};
