// @mtcute/core/highlevel/types/peers/bot-info.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const documentUtils = require("../media/document-utils.cjs");
const photo = require("../media/photo.cjs");
class BotInfo {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether the bot has preview medias available */
  get hasPreviewMedia() {
    return this.raw.hasPreviewMedias;
  }
  /** ID of the bot */
  get id() {
    return utils.asNonNull(this.raw.userId);
  }
  get description() {
    return this.raw.description ?? "";
  }
  get descriptionPhoto() {
    if (!this.raw.descriptionPhoto || this.raw.descriptionPhoto._ !== "photo") return null;
    return new photo.Photo(this.raw.descriptionPhoto);
  }
  get descriptionVideo() {
    if (!this.raw.descriptionDocument || this.raw.descriptionDocument._ !== "document") return null;
    const parsed = documentUtils.parseDocument(this.raw.descriptionDocument);
    if (parsed.type !== "video") return null;
    return parsed;
  }
  /** List of the bot's registered commands */
  get commands() {
    return this.raw.commands ?? [];
  }
  /** Action to be performed when the bot's menu button is pressed */
  get menuButton() {
    return this.raw.menuButton ?? null;
  }
  /** URL of the bot's privacy policy */
  get privacyPolicyUrl() {
    return this.raw.privacyPolicyUrl ?? null;
  }
  get verifierSettings() {
    return this.raw.verifierSettings ?? null;
  }
}
const BotInfo$1 = /* @__PURE__ */ inspectable.makeInspectable(BotInfo);
const BotInfo$2 = /* @__PURE__ */ memoize.memoizeGetters(BotInfo$1, ["descriptionPhoto", "descriptionVideo"]);
exports.BotInfo = BotInfo$2;
