// @mtcute/core/highlevel/types/peers/chat-permissions.js
import { makeInspectable } from "../../utils/inspectable.js";
class ChatPermissions {
  raw;
  constructor(bannedRights) {
    this.raw = bannedRights;
  }
  /**
   * Whether users can view messages
   */
  get canViewMessages() {
    return !this.raw.viewMessages;
  }
  /**
   * Whether users can send text messages,
   * contacts, locations and venues
   */
  get canSendMessages() {
    return !this.raw.sendMessages;
  }
  /**
   * Whether users can send media messages,
   * including documents, photos, videos, video notes and voice notes.
   *
   * Implies {@link canSendMessages}
   */
  get canSendMedia() {
    return !this.raw.sendMedia;
  }
  /**
   * Whether users can send stickers.
   *
   * Implies {@link canSendMedia}
   */
  get canSendStickers() {
    return !this.raw.sendStickers;
  }
  /**
   * Whether users can send GIFs.
   *
   * Implies {@link canSendMedia}
   */
  get canSendGifs() {
    return !this.raw.sendGifs;
  }
  /**
   * Whether users can send photos.
   *
   * Implies {@link canSendMedia}
   */
  get canSendPhotos() {
    return !this.raw.sendPhotos;
  }
  /**
   * Whether users can send videos.
   *
   * Implies {@link canSendMedia}
   */
  get canSendVideos() {
    return !this.raw.sendVideos;
  }
  /**
   * Whether users can send round videos (i.e. video notes).
   *
   * Implies {@link canSendMedia}
   */
  get canSendRoundVideos() {
    return !this.raw.sendRoundvideos;
  }
  /**
   * Whether users can send audio files.
   *
   * Implies {@link canSendMedia}
   */
  get canSendAudios() {
    return !this.raw.sendAudios;
  }
  /**
   * Whether users can send voice notes.
   *
   * Implies {@link canSendMedia}
   */
  get canSendVoices() {
    return !this.raw.sendVoices;
  }
  /**
   * Whether users can send files.
   *
   * Implies {@link canSendMedia}
   */
  get canSendFiles() {
    return !this.raw.sendDocs;
  }
  /**
   * Whether users can send games.
   *
   * Implies {@link canSendMedia}
   */
  get canSendGames() {
    return !this.raw.sendGames;
  }
  /**
   * Whether users can use inline bots.
   *
   * Implies {@link canSendMedia}
   */
  get canUseInline() {
    return !this.raw.sendInline;
  }
  /**
   * Whether users can use inline bots.
   *
   * Implies {@link canSendMedia}
   */
  get canAddWebPreviews() {
    return !this.raw.embedLinks;
  }
  /**
   * Whether users can send text messages.
   *
   * Implies {@link canSendMessages}
   */
  get canSendText() {
    return !this.raw.sendPlain;
  }
  /**
   * Whether users can send polls.
   *
   * Implies {@link canSendMessages}
   */
  get canSendPolls() {
    return !this.raw.sendPolls;
  }
  /**
   * Whether users can react to messages.
   */
  get canSendReactions() {
    return !this.raw.sendReactions;
  }
  /**
   * Whether users can change the chat title,
   * photo and other settings.
   */
  get canChangeInfo() {
    return !this.raw.changeInfo;
  }
  /**
   * Whether users can invite other users to the chat
   */
  get canInviteUsers() {
    return !this.raw.inviteUsers;
  }
  /**
   * Whether users can pin messages
   */
  get canPinMessages() {
    return !this.raw.pinMessages;
  }
  /**
   * Whether users can pin messages
   */
  get canManageTopics() {
    return !this.raw.manageTopics;
  }
  /**
   * UNIX date until which these permissions are valid,
   * or `null` if forever.
   *
   * For example, represents the time when the restrictions
   * will be lifted from a {@link ChatMember}
   */
  get untilDate() {
    return this.raw.untilDate === 0 ? null : new Date(this.raw.untilDate * 1e3);
  }
}
const ChatPermissions$1 = /* @__PURE__ */ makeInspectable(ChatPermissions);
export {
  ChatPermissions$1 as ChatPermissions
};
