// @mtcute/core/highlevel/types/peers/chat-member.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const memoize = require("../../utils/memoize.cjs");
const chatPermissions = require("./chat-permissions.cjs");
const user = require("./user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class ChatMember {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Information about the user
   */
  get user() {
    switch (this.raw._) {
      case "channelParticipantBanned":
      case "channelParticipantLeft":
        typeAssertions.assertTypeIs("ChatMember#user (raw.peer)", this.raw.peer, "peerUser");
        return new user.User(this._peers.user(this.raw.peer.userId));
      default:
        return new user.User(this._peers.user(this.raw.userId));
    }
  }
  /**
   * Get the chat member status
   */
  get status() {
    switch (this.raw._) {
      case "channelParticipant":
      case "channelParticipantSelf":
      case "chatParticipant":
        return "member";
      case "channelParticipantCreator":
      case "chatParticipantCreator":
        return "creator";
      case "channelParticipantAdmin":
      case "chatParticipantAdmin":
        return "admin";
      case "channelParticipantLeft":
        return "left";
      case "channelParticipantBanned":
        return this.raw.bannedRights.viewMessages ? "banned" : "restricted";
    }
  }
  /**
   * Custom title (rank) of the chat member.
   *
   * `null` in case custom title (rank) is not set.
   */
  get title() {
    if ("rank" in this.raw) {
      return this.raw.rank ?? null;
    }
    return null;
  }
  /**
   * Date when the user has joined the chat.
   *
   * Not available for creators and left members
   */
  get joinedDate() {
    switch (this.raw._) {
      case "channelParticipantCreator":
      case "chatParticipantCreator":
      case "channelParticipantLeft":
        return null;
      default:
        return new Date(this.raw.date * 1e3);
    }
  }
  /**
   * Information about whoever invited this member to the chat.
   *
   * Only available in the following cases:
   *  - `user` is yourself
   *  - `chat` is a legacy group
   *  - `chat` is a supergroup/channel, and `user` is an admin
   */
  get invitedBy() {
    if ("inviterId" in this.raw && this.raw.inviterId) {
      return new user.User(this._peers.user(this.raw.inviterId));
    }
    return null;
  }
  /**
   * Information about whoever promoted this admin.
   *
   * Only available if `status = admin`.
   */
  get promotedBy() {
    if (this.raw._ === "channelParticipantAdmin") {
      return new user.User(this._peers.user(this.raw.promotedBy));
    }
    return null;
  }
  /**
   * Information about whoever restricted this user.
   *
   * Only available if `status = restricted or status = banned`
   */
  get restrictedBy() {
    if (this.raw._ === "channelParticipantBanned") {
      return new user.User(this._peers.user(this.raw.kickedBy));
    }
    return null;
  }
  /**
   * For restricted and banned users,
   * information about the restrictions
   */
  get restrictions() {
    if (this.raw._ !== "channelParticipantBanned") return null;
    return new chatPermissions.ChatPermissions(this.raw.bannedRights);
  }
  /**
   * Whether this member is a part of the chat now.
   *
   * Makes sense only when `status = restricted or staus = banned`
   */
  get isMember() {
    return this.raw._ === "channelParticipantBanned" ? !this.raw.left : this.raw._ !== "channelParticipantLeft";
  }
  /**
   * For admins and creator of supergroup/channels,
   * list of their admin permissions.
   *
   * Also contains whether this admin is anonymous.
   */
  get permissions() {
    switch (this.raw._) {
      case "channelParticipantAdmin":
      case "channelParticipantCreator":
        return this.raw.adminRights;
      default:
        return null;
    }
  }
  /**
   * If this member subscribed to a channel using Telegram Stars,
   * this field will contain the date when the subscription will expire
   */
  get subscriptionUntilDate() {
    switch (this.raw._) {
      case "channelParticipant":
      case "channelParticipantSelf":
        if (!this.raw.subscriptionUntilDate) return null;
        return new Date(this.raw.subscriptionUntilDate * 1e3);
    }
    return null;
  }
}
const ChatMember$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatMember, ["user", "invitedBy", "promotedBy", "restrictedBy", "restrictions"]);
const ChatMember$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatMember$1);
exports.ChatMember = ChatMember$2;
