// @mtcute/core/highlevel/types/peers/chatlist-preview.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("./peer.cjs");
const peersIndex = require("./peers-index.cjs");
class ChatlistPreview {
  constructor(raw) {
    this.raw = raw;
    this.peers = peersIndex.PeersIndex.from(raw);
  }
  peers;
  /** Whether the current user has already joined this chatlist */
  get isJoined() {
    return this.raw._ === "chatlists.chatlistInviteAlready";
  }
  /** If we joined the chatlist, ID of the folder representing it */
  get folderId() {
    return this.raw._ === "chatlists.chatlistInviteAlready" ? this.raw.filterId : null;
  }
  /** Title of the chatlist (only available for non-joined chatlists) */
  get title() {
    return this.raw._ === "chatlists.chatlistInvite" ? this.raw.title : { text: "", entities: [] };
  }
  /** Emoji representing an icon of the chatlist (may only be available for non-joined chatlists) */
  get emoji() {
    return this.raw._ === "chatlists.chatlistInvite" ? this.raw.emoticon ?? null : null;
  }
  /** List of all chats contained in the chatlist */
  get chats() {
    let peers;
    if (this.raw._ === "chatlists.chatlistInvite") {
      peers = this.raw.peers;
    } else {
      peers = [...this.raw.alreadyPeers, ...this.raw.missingPeers];
    }
    return peers.map((peer$1) => peer.parsePeer(peer$1, this.peers));
  }
}
const ChatlistPreview$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatlistPreview, ["chats"]);
const ChatlistPreview$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatlistPreview$1);
exports.ChatlistPreview = ChatlistPreview$2;
