// @mtcute/core/highlevel/types/peers/chat-photo.js
import { toFileId, tdFileId, toUniqueFileId } from "@mtcute/file-id";
import Long from "long";
import { MtArgumentError } from "../../../types/errors.js";
import { toggleChannelIdMark } from "../../../utils/peer-utils.js";
import { strippedPhotoToJpg } from "../../utils/file-utils.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { FileLocation as FileLocation$1 } from "../files/file-location.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ChatPhotoSize extends FileLocation$1 {
  constructor(peer, obj, big) {
    super(
      {
        _: "inputPeerPhotoFileLocation",
        peer,
        photoId: obj.photoId,
        big
      },
      void 0,
      obj.dcId
    );
    this.peer = peer;
    this.obj = obj;
    this.big = big;
    this.peer = peer;
    this.obj = obj;
    this.big = big;
  }
  /**
   * TDLib and Bot API compatible File ID representing this size
   */
  get fileId() {
    const peer = this.peer;
    let id;
    let hash;
    switch (peer._) {
      case "inputPeerUser":
        id = peer.userId;
        hash = peer.accessHash;
        break;
      case "inputPeerChat":
        id = -peer.chatId;
        hash = Long.ZERO;
        break;
      case "inputPeerChannel":
        id = toggleChannelIdMark(peer.channelId);
        hash = peer.accessHash;
        break;
      default:
        throw new MtArgumentError("Input peer was invalid");
    }
    return toFileId({
      dcId: this.obj.dcId,
      type: tdFileId.FileType.ProfilePhoto,
      fileReference: null,
      location: {
        _: "photo",
        id: this.obj.photoId,
        accessHash: Long.ZERO,
        source: {
          _: "dialogPhoto",
          big: this.big,
          id,
          accessHash: hash
        }
      }
    });
  }
  /**
   * TDLib and Bot API compatible unique File ID representing this size
   */
  get uniqueFileId() {
    return toUniqueFileId(tdFileId.FileType.ProfilePhoto, {
      _: "photo",
      id: this.obj.photoId,
      // eslint-disable-next-line ts/no-unsafe-assignment
      source: {
        _: "dialogPhoto",
        big: this.big
      }
    });
  }
}
const ChatPhotoSize$1 = /* @__PURE__ */ memoizeGetters(ChatPhotoSize, ["fileId", "uniqueFileId"]);
const ChatPhotoSize$2 = /* @__PURE__ */ makeInspectable(ChatPhotoSize$1, ["dcId", "big"]);
class ChatPhoto {
  constructor(peer, raw) {
    this.peer = peer;
    this.raw = raw;
  }
  /**
   * Whether this photo is personal (i.e. you set it yourself for this peer)
   */
  get isPersonal() {
    return this.raw._ === "userProfilePhoto" && this.raw.personal;
  }
  /** Chat photo file location in small resolution (160x160) */
  get small() {
    return new ChatPhotoSize(this.peer, this.raw, false);
  }
  /** Chat photo file location in big resolution (640x640) */
  get big() {
    return new ChatPhotoSize(this.peer, this.raw, true);
  }
  /**
   * Chat photo preview in *very* small resolution, if available
   */
  get thumb() {
    if (!this.raw.strippedThumb) return null;
    return strippedPhotoToJpg(this.raw.strippedThumb);
  }
}
const ChatPhoto$1 = /* @__PURE__ */ memoizeGetters(ChatPhoto, ["small", "big", "thumb"]);
const ChatPhoto$2 = /* @__PURE__ */ makeInspectable(ChatPhoto$1, [], ["thumb"]);
export {
  ChatPhoto$2 as ChatPhoto,
  ChatPhotoSize$2 as ChatPhotoSize
};
