// @mtcute/core/highlevel/types/peers/chat-preview.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const photo = require("../media/photo.cjs");
const botVerification = require("./bot-verification.cjs");
const user = require("./user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class ChatPreview {
  constructor(invite, link) {
    this.invite = invite;
    this.link = link;
  }
  /**
   * Title of the chat
   */
  get title() {
    return this.invite.title;
  }
  /**
   * Type of the chat
   */
  get type() {
    if (this.invite.broadcast) return "channel";
    if (this.invite.megagroup || this.invite.channel) return "supergroup";
    return "group";
  }
  /**
   * Whether this chat is public
   */
  get public() {
    return this.invite.public;
  }
  /**
   * Total chat member count
   */
  get memberCount() {
    return this.invite.participantsCount;
  }
  /**
   * Chat photo
   */
  get photo() {
    if (this.invite.photo._ === "photoEmpty") return null;
    return new photo.Photo(this.invite.photo);
  }
  /**
   * Preview of some of the chat members.
   *
   * This usually contains around 10 members,
   * and members that are inside your contacts list are
   * ordered before others.
   */
  get someMembers() {
    return this.invite.participants?.map((it) => new user.User(it)) ?? [];
  }
  /**
   * Whether by using this link you'll also need
   * to wait for admin approval.
   */
  get withApproval() {
    return this.invite.requestNeeded;
  }
  /**
   * If non-null, this user was verified by a bot, and this field contains
   * the ID of the custom emoji to display as the verification icon.
   */
  get customVerification() {
    return this.invite.botVerification ? new botVerification.BotVerification(this.invite.botVerification) : null;
  }
}
const ChatPreview$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatPreview, ["photo", "someMembers"]);
const ChatPreview$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatPreview$1, ["link"]);
exports.ChatPreview = ChatPreview$2;
