// @mtcute/core/highlevel/types/peers/chat-invite-link-member.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const user = require("./user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class ChatInviteLinkMember {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * User who joined the chat
   */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * Date when the user joined the chat
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this user currently has a pending join request
   * (and is actually not a member yet)
   */
  get isPendingRequest() {
    return this.raw.requested;
  }
  /**
   * Whether the participant joined by importing a chat folder deep link
   */
  get isViaChatlist() {
    return this.raw.requested;
  }
  /**
   * For users with pending requests, contains bio of the user that requested to join
   */
  get bio() {
    return this.raw.about ?? null;
  }
  /**
   * The administrator that approved the join request of the user
   */
  get approvedBy() {
    if (!this.raw.approvedBy) return null;
    return new user.User(this._peers.user(this.raw.approvedBy));
  }
}
const ChatInviteLinkMember$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatInviteLinkMember, ["user", "approvedBy"]);
const ChatInviteLinkMember$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatInviteLinkMember$1);
exports.ChatInviteLinkMember = ChatInviteLinkMember$2;
