// @mtcute/core/highlevel/types/peers/chat-invite-link-member.js
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "./user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ChatInviteLinkMember {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * User who joined the chat
   */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * Date when the user joined the chat
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this user currently has a pending join request
   * (and is actually not a member yet)
   */
  get isPendingRequest() {
    return this.raw.requested;
  }
  /**
   * Whether the participant joined by importing a chat folder deep link
   */
  get isViaChatlist() {
    return this.raw.requested;
  }
  /**
   * For users with pending requests, contains bio of the user that requested to join
   */
  get bio() {
    return this.raw.about ?? null;
  }
  /**
   * The administrator that approved the join request of the user
   */
  get approvedBy() {
    if (!this.raw.approvedBy) return null;
    return new User$2(this._peers.user(this.raw.approvedBy));
  }
}
const ChatInviteLinkMember$1 = /* @__PURE__ */ memoizeGetters(ChatInviteLinkMember, ["user", "approvedBy"]);
const ChatInviteLinkMember$2 = /* @__PURE__ */ makeInspectable(ChatInviteLinkMember$1);
export {
  ChatInviteLinkMember$2 as ChatInviteLinkMember
};
