// @mtcute/core/highlevel/types/peers/chat-invite-link.js
import { assertTypeIsNot } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "./user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ChatInviteLink {
  constructor(raw, _peers) {
    this._peers = _peers;
    assertTypeIsNot("ChatInviteLink", raw, "chatInvitePublicJoinRequests");
    this.raw = raw;
  }
  raw;
  /**
   * The invite link as a `t.me/joinchat/` string.
   *
   * If the link was created by another administrator, the second
   * part of the link will be censored with `...` (e.g. `https://t.me/joinchat/BGxxHIg4...`
   *
   * See also: {@link isMyLink}
   */
  get link() {
    return this.raw.link;
  }
  /**
   * Whether this invite link was created by the current user.
   *
   * If so, {@link link} will be a full invite link.
   */
  get isMyLink() {
    return this.creator?.isSelf ?? !this.raw.link.endsWith("...");
  }
  /**
   * Creator of the invite link, if available
   */
  get creator() {
    if (!this._peers) return null;
    return new User$2(this._peers.user(this.raw.adminId));
  }
  /**
   * Creation date of the link
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Whether this link is primary (i.e. "permanent")
   */
  get isPrimary() {
    return this.raw.permanent;
  }
  /**
   * Whether this link was revoked and can't be used anymore
   */
  get isRevoked() {
    return this.raw.revoked;
  }
  /**
   * The date since which the link will be valid (if any)
   */
  get startDate() {
    return this.raw.startDate ? new Date(this.raw.startDate * 1e3) : null;
  }
  /**
   * The date until which the link will be valid (if any)
   */
  get endDate() {
    return this.raw.expireDate ? new Date(this.raw.expireDate * 1e3) : null;
  }
  /**
   * Maximum number of users that can be members of this chat
   * at the same time after joining using this link.
   *
   * Integer in range `[1, 99999]` or `Infinity` if unspecified
   */
  get usageLimit() {
    return this.raw.usageLimit ?? Infinity;
  }
  /**
   * Number of users currently in the chat that joined using this link
   */
  get usage() {
    return this.raw.usage ?? 0;
  }
  /**
   * Whether users joined by this link need to be
   * approved by the group administrator before they can join
   */
  get approvalNeeded() {
    return this.raw.requestNeeded;
  }
  /**
   * Number of users currently awaiting admin approval.
   *
   * 0 in case the link is not using approvals
   */
  get pendingApprovals() {
    return this.raw.requested ?? 0;
  }
}
const ChatInviteLink$1 = /* @__PURE__ */ memoizeGetters(ChatInviteLink, ["creator"]);
const ChatInviteLink$2 = /* @__PURE__ */ makeInspectable(ChatInviteLink$1);
export {
  ChatInviteLink$2 as ChatInviteLink
};
