// @mtcute/core/highlevel/types/peers/chatlist-preview.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parsePeer } from "./peer.js";
import { PeersIndex } from "./peers-index.js";
class ChatlistPreview {
  constructor(raw) {
    this.raw = raw;
    this.peers = PeersIndex.from(raw);
  }
  peers;
  /** Whether the current user has already joined this chatlist */
  get isJoined() {
    return this.raw._ === "chatlists.chatlistInviteAlready";
  }
  /** If we joined the chatlist, ID of the folder representing it */
  get folderId() {
    return this.raw._ === "chatlists.chatlistInviteAlready" ? this.raw.filterId : null;
  }
  /** Title of the chatlist (only available for non-joined chatlists) */
  get title() {
    return this.raw._ === "chatlists.chatlistInvite" ? this.raw.title : { text: "", entities: [] };
  }
  /** Emoji representing an icon of the chatlist (may only be available for non-joined chatlists) */
  get emoji() {
    return this.raw._ === "chatlists.chatlistInvite" ? this.raw.emoticon ?? null : null;
  }
  /** List of all chats contained in the chatlist */
  get chats() {
    let peers;
    if (this.raw._ === "chatlists.chatlistInvite") {
      peers = this.raw.peers;
    } else {
      peers = [...this.raw.alreadyPeers, ...this.raw.missingPeers];
    }
    return peers.map((peer) => parsePeer(peer, this.peers));
  }
}
const ChatlistPreview$1 = /* @__PURE__ */ memoizeGetters(ChatlistPreview, ["chats"]);
const ChatlistPreview$2 = /* @__PURE__ */ makeInspectable(ChatlistPreview$1);
export {
  ChatlistPreview$2 as ChatlistPreview
};
