// @mtcute/core/highlevel/types/peers/chat-member.js
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { ChatPermissions as ChatPermissions$1 } from "./chat-permissions.js";
import { User as User$2 } from "./user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ChatMember {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Information about the user
   */
  get user() {
    switch (this.raw._) {
      case "channelParticipantBanned":
      case "channelParticipantLeft":
        assertTypeIs("ChatMember#user (raw.peer)", this.raw.peer, "peerUser");
        return new User$2(this._peers.user(this.raw.peer.userId));
      default:
        return new User$2(this._peers.user(this.raw.userId));
    }
  }
  /**
   * Get the chat member status
   */
  get status() {
    switch (this.raw._) {
      case "channelParticipant":
      case "channelParticipantSelf":
      case "chatParticipant":
        return "member";
      case "channelParticipantCreator":
      case "chatParticipantCreator":
        return "creator";
      case "channelParticipantAdmin":
      case "chatParticipantAdmin":
        return "admin";
      case "channelParticipantLeft":
        return "left";
      case "channelParticipantBanned":
        return this.raw.bannedRights.viewMessages ? "banned" : "restricted";
    }
  }
  /**
   * Custom title (rank) of the chat member.
   *
   * `null` in case custom title (rank) is not set.
   */
  get title() {
    if ("rank" in this.raw) {
      return this.raw.rank ?? null;
    }
    return null;
  }
  /**
   * Date when the user has joined the chat.
   *
   * Not available for creators and left members
   */
  get joinedDate() {
    switch (this.raw._) {
      case "channelParticipantCreator":
      case "chatParticipantCreator":
      case "channelParticipantLeft":
        return null;
      default:
        return new Date(this.raw.date * 1e3);
    }
  }
  /**
   * Information about whoever invited this member to the chat.
   *
   * Only available in the following cases:
   *  - `user` is yourself
   *  - `chat` is a legacy group
   *  - `chat` is a supergroup/channel, and `user` is an admin
   */
  get invitedBy() {
    if ("inviterId" in this.raw && this.raw.inviterId) {
      return new User$2(this._peers.user(this.raw.inviterId));
    }
    return null;
  }
  /**
   * Information about whoever promoted this admin.
   *
   * Only available if `status = admin`.
   */
  get promotedBy() {
    if (this.raw._ === "channelParticipantAdmin") {
      return new User$2(this._peers.user(this.raw.promotedBy));
    }
    return null;
  }
  /**
   * Information about whoever restricted this user.
   *
   * Only available if `status = restricted or status = banned`
   */
  get restrictedBy() {
    if (this.raw._ === "channelParticipantBanned") {
      return new User$2(this._peers.user(this.raw.kickedBy));
    }
    return null;
  }
  /**
   * For restricted and banned users,
   * information about the restrictions
   */
  get restrictions() {
    if (this.raw._ !== "channelParticipantBanned") return null;
    return new ChatPermissions$1(this.raw.bannedRights);
  }
  /**
   * Whether this member is a part of the chat now.
   *
   * Makes sense only when `status = restricted or staus = banned`
   */
  get isMember() {
    return this.raw._ === "channelParticipantBanned" ? !this.raw.left : this.raw._ !== "channelParticipantLeft";
  }
  /**
   * For admins and creator of supergroup/channels,
   * list of their admin permissions.
   *
   * Also contains whether this admin is anonymous.
   */
  get permissions() {
    switch (this.raw._) {
      case "channelParticipantAdmin":
      case "channelParticipantCreator":
        return this.raw.adminRights;
      default:
        return null;
    }
  }
  /**
   * If this member subscribed to a channel using Telegram Stars,
   * this field will contain the date when the subscription will expire
   */
  get subscriptionUntilDate() {
    switch (this.raw._) {
      case "channelParticipant":
      case "channelParticipantSelf":
        if (!this.raw.subscriptionUntilDate) return null;
        return new Date(this.raw.subscriptionUntilDate * 1e3);
    }
    return null;
  }
}
const ChatMember$1 = /* @__PURE__ */ memoizeGetters(ChatMember, ["user", "invitedBy", "promotedBy", "restrictedBy", "restrictions"]);
const ChatMember$2 = /* @__PURE__ */ makeInspectable(ChatMember$1);
export {
  ChatMember$2 as ChatMember
};
