// @mtcute/core/highlevel/types/peers/chat-preview.js
import { memoizeGetters } from "../../utils/memoize.js";
import { Photo as Photo$2 } from "../media/photo.js";
import { BotVerification as BotVerification$1 } from "./bot-verification.js";
import { User as User$2 } from "./user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ChatPreview {
  constructor(invite, link) {
    this.invite = invite;
    this.link = link;
  }
  /**
   * Title of the chat
   */
  get title() {
    return this.invite.title;
  }
  /**
   * Type of the chat
   */
  get type() {
    if (this.invite.broadcast) return "channel";
    if (this.invite.megagroup || this.invite.channel) return "supergroup";
    return "group";
  }
  /**
   * Whether this chat is public
   */
  get public() {
    return this.invite.public;
  }
  /**
   * Total chat member count
   */
  get memberCount() {
    return this.invite.participantsCount;
  }
  /**
   * Chat photo
   */
  get photo() {
    if (this.invite.photo._ === "photoEmpty") return null;
    return new Photo$2(this.invite.photo);
  }
  /**
   * Preview of some of the chat members.
   *
   * This usually contains around 10 members,
   * and members that are inside your contacts list are
   * ordered before others.
   */
  get someMembers() {
    return this.invite.participants?.map((it) => new User$2(it)) ?? [];
  }
  /**
   * Whether by using this link you'll also need
   * to wait for admin approval.
   */
  get withApproval() {
    return this.invite.requestNeeded;
  }
  /**
   * If non-null, this user was verified by a bot, and this field contains
   * the ID of the custom emoji to display as the verification icon.
   */
  get customVerification() {
    return this.invite.botVerification ? new BotVerification$1(this.invite.botVerification) : null;
  }
}
const ChatPreview$1 = /* @__PURE__ */ memoizeGetters(ChatPreview, ["photo", "someMembers"]);
const ChatPreview$2 = /* @__PURE__ */ makeInspectable(ChatPreview$1, ["link"]);
export {
  ChatPreview$2 as ChatPreview
};
