// @mtcute/core/highlevel/types/peers/chat-photo.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fileId = require("@mtcute/file-id");
const Long = require("long");
const errors = require("../../../types/errors.cjs");
const peerUtils = require("../../../utils/peer-utils.cjs");
const fileUtils = require("../../utils/file-utils.cjs");
const memoize = require("../../utils/memoize.cjs");
const fileLocation = require("../files/file-location.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class ChatPhotoSize extends fileLocation.FileLocation {
  constructor(peer, obj, big) {
    super(
      {
        _: "inputPeerPhotoFileLocation",
        peer,
        photoId: obj.photoId,
        big
      },
      void 0,
      obj.dcId
    );
    this.peer = peer;
    this.obj = obj;
    this.big = big;
    this.peer = peer;
    this.obj = obj;
    this.big = big;
  }
  /**
   * TDLib and Bot API compatible File ID representing this size
   */
  get fileId() {
    const peer = this.peer;
    let id;
    let hash;
    switch (peer._) {
      case "inputPeerUser":
        id = peer.userId;
        hash = peer.accessHash;
        break;
      case "inputPeerChat":
        id = -peer.chatId;
        hash = Long.ZERO;
        break;
      case "inputPeerChannel":
        id = peerUtils.toggleChannelIdMark(peer.channelId);
        hash = peer.accessHash;
        break;
      default:
        throw new errors.MtArgumentError("Input peer was invalid");
    }
    return fileId.toFileId({
      dcId: this.obj.dcId,
      type: fileId.tdFileId.FileType.ProfilePhoto,
      fileReference: null,
      location: {
        _: "photo",
        id: this.obj.photoId,
        accessHash: Long.ZERO,
        source: {
          _: "dialogPhoto",
          big: this.big,
          id,
          accessHash: hash
        }
      }
    });
  }
  /**
   * TDLib and Bot API compatible unique File ID representing this size
   */
  get uniqueFileId() {
    return fileId.toUniqueFileId(fileId.tdFileId.FileType.ProfilePhoto, {
      _: "photo",
      id: this.obj.photoId,
      // eslint-disable-next-line ts/no-unsafe-assignment
      source: {
        _: "dialogPhoto",
        big: this.big
      }
    });
  }
}
const ChatPhotoSize$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatPhotoSize, ["fileId", "uniqueFileId"]);
const ChatPhotoSize$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatPhotoSize$1, ["dcId", "big"]);
class ChatPhoto {
  constructor(peer, raw) {
    this.peer = peer;
    this.raw = raw;
  }
  /**
   * Whether this photo is personal (i.e. you set it yourself for this peer)
   */
  get isPersonal() {
    return this.raw._ === "userProfilePhoto" && this.raw.personal;
  }
  /** Chat photo file location in small resolution (160x160) */
  get small() {
    return new ChatPhotoSize(this.peer, this.raw, false);
  }
  /** Chat photo file location in big resolution (640x640) */
  get big() {
    return new ChatPhotoSize(this.peer, this.raw, true);
  }
  /**
   * Chat photo preview in *very* small resolution, if available
   */
  get thumb() {
    if (!this.raw.strippedThumb) return null;
    return fileUtils.strippedPhotoToJpg(this.raw.strippedThumb);
  }
}
const ChatPhoto$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatPhoto, ["small", "big", "thumb"]);
const ChatPhoto$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatPhoto$1, [], ["thumb"]);
exports.ChatPhoto = ChatPhoto$2;
exports.ChatPhotoSize = ChatPhotoSize$2;
