// @mtcute/core/highlevel/types/peers/forum-topic.js
import { hasValueAtKey } from "../../../utils/type-assertions.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { MtMessageNotFoundError } from "../errors.js";
import { parsePeer } from "./peer.js";
import { PeersIndex } from "./peers-index.js";
import { Message as Message$2 } from "../messages/message.js";
import { DraftMessage as DraftMessage$2 } from "../messages/draft-message.js";
import { makeInspectable } from "../../utils/inspectable.js";
class ForumTopic {
  constructor(raw, _peers, _messages) {
    this.raw = raw;
    this._peers = _peers;
    this._messages = _messages;
  }
  static COLOR_BLUE = 7322096;
  static COLOR_YELLOW = 16766590;
  static COLOR_PURPLE = 13338331;
  static COLOR_GREEN = 9367192;
  static COLOR_PINK = 16749490;
  static COLOR_RED = 16478047;
  static parseTlForumTopics(topics) {
    const peers = PeersIndex.from(topics);
    const messages = /* @__PURE__ */ new Map();
    topics.messages.forEach((msg) => {
      if (!msg.peerId) return;
      messages.set(msg.id, msg);
    });
    return topics.topics.filter(hasValueAtKey("_", "forumTopic")).map((it) => new ForumTopic(it, peers, messages));
  }
  /**
   * Whether the topic was created by the current user
   */
  get isMy() {
    return this.raw.my;
  }
  /**
   * Whether the topic is closed
   */
  get isClosed() {
    return this.raw.closed;
  }
  /**
   * Whether the topic is pinned
   */
  get isPinned() {
    return this.raw.pinned;
  }
  /**
   * Whether this constructor is a reduced version of the full topic information.
   *
   * If `true`, only {@link isMy}, {@link isClosed}, {@link id}, {@link date},
   * {@link title}, {@link iconColor}, {@link iconCustomEmoji} and {@link creator}
   * parameters will contain valid information.
   */
  get isShort() {
    return this.raw.short;
  }
  /**
   * ID of the topic
   */
  get id() {
    return this.raw.id;
  }
  /**
   * Date when the topic was created
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Title of the topic
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Whether the title is not set yet
   */
  get titleMissing() {
    return this.raw.titleMissing;
  }
  /**
   * Color of the topic's icon, used as a fallback
   * in case {@link iconEmoji} is not set.
   *
   * One of the static `COLOR_*` fields.
   */
  get iconColor() {
    return this.raw.iconColor ?? null;
  }
  /**
   * Emoji used as the topic's icon.
   */
  get iconCustomEmoji() {
    return this.raw.iconEmojiId ?? null;
  }
  /**
   * Creator of the topic
   */
  get creator() {
    return parsePeer(this.raw.fromId, this._peers);
  }
  // idk???
  get peer() {
    return parsePeer(this.raw.peer, this._peers);
  }
  /**
   * The latest message sent in this topic
   */
  get lastMessage() {
    const id = this.raw.topMessage;
    if (this._messages?.has(id)) {
      return new Message$2(this._messages.get(id), this._peers);
    }
    throw new MtMessageNotFoundError(0, id);
  }
  /**
   * ID of the last read outgoing message in this topic
   */
  get lastReadIngoing() {
    return this.raw.readInboxMaxId;
  }
  /**
   * ID of the last read ingoing message in this topic
   */
  get lastReadOutgoing() {
    return this.raw.readOutboxMaxId;
  }
  /**
   * ID of the last read message in this topic
   */
  get lastRead() {
    return Math.max(this.raw.readOutboxMaxId, this.raw.readInboxMaxId);
  }
  /**
   * Number of unread messages in the topic
   */
  get unreadCount() {
    return this.raw.unreadCount;
  }
  /**
   * Number of unread mentions in the topic
   */
  get unreadMentionsCount() {
    return this.raw.unreadMentionsCount;
  }
  /**
   * Number of unread reactions in the topic
   */
  get unreadReactionsCount() {
    return this.raw.unreadReactionsCount;
  }
  /** Number of unread poll votes in the topic */
  get unreadPollVotesCount() {
    return this.raw.unreadPollVotesCount;
  }
  /**
   * Draft message in the topic
   */
  get draftMessage() {
    if (!this.raw.draft || this.raw.draft._ === "draftMessageEmpty") return null;
    return new DraftMessage$2(this.raw.draft);
  }
}
const ForumTopic$1 = /* @__PURE__ */ memoizeGetters(ForumTopic, ["creator", "lastMessage", "draftMessage"]);
const ForumTopic$2 = /* @__PURE__ */ makeInspectable(ForumTopic$1);
export {
  ForumTopic$2 as ForumTopic
};
