// @mtcute/core/highlevel/types/peers/full-user.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const documentUtils = require("../media/document-utils.cjs");
const photo = require("../media/photo.cjs");
const businessAccount = require("../premium/business-account.cjs");
const botVerification = require("./bot-verification.cjs");
const chat = require("./chat.cjs");
const peersIndex = require("./peers-index.cjs");
const user = require("./user.cjs");
class FullUser extends user.User {
  peers;
  full;
  constructor(obj) {
    const peers = peersIndex.PeersIndex.from(obj);
    super(peers.user(obj.fullUser.id));
    this.peers = peers;
    this.full = obj.fullUser;
  }
  /** Whether this chat is archived */
  get isArchived() {
    return this.full.folderId === 1;
  }
  /**
   * Whether this user has restricted sending them voice/video messages.
   */
  get hasBlockedVoices() {
    return this.full.voiceMessagesForbidden;
  }
  /** Whether the user is blocked by the current user */
  get isBlocked() {
    return this.full.blocked;
  }
  /** Whether the user can make/accept calls */
  get hasCallsAvailable() {
    return this.full.phoneCallsAvailable;
  }
  /** Whether the user can make/accept video calls */
  get hasVideoCallsAvailable() {
    return this.full.videoCallsAvailable;
  }
  /** Whether the current user can call this user */
  get canCall() {
    return !this.full.phoneCallsPrivate;
  }
  /** Whether the chat with this user has some scheduled messages */
  get hasScheduled() {
    return this.full.hasScheduled;
  }
  /** Whether the real-time translations popup should not be shown for this channel */
  get hasTranslationsDisabled() {
    return this.full.translationsDisabled;
  }
  /** Whether this chat has pinned stories */
  get hasPinnedStories() {
    return this.full.storiesPinnedAvailable;
  }
  /** Whether the current user has blocked this user from seeing our stories */
  get blockedMyStoriesFrom() {
    return this.full.blockedMyStoriesFrom;
  }
  /** Whether the chat with this user has a custom wallpaper */
  get hasCustomWallpaper() {
    return this.full.wallpaperOverridden;
  }
  /** Whether this user has hidden the exact read dates */
  get hasHiddenReadDate() {
    return this.full.readDatesPrivate;
  }
  /**
   * (Only for current user) Whether we have re-enabled sponsored messages,
   * despite having Telegram Premium
   */
  get hasSponsoredEnabled() {
    return this.full.sponsoredEnabled;
  }
  /** Whether the current user can view ad revenue for this bot */
  get canViewAdRevenue() {
    return this.full.canViewRevenue;
  }
  /**
   * Full information about this chat's photo, if any.
   *
   * Unlike {@link Chat.photo}, this field contains additional information
   * about the photo, such as its date, more sizes, and is the only
   * way to get the animated profile photo.
   *
   * This field takes into account any personal/fallback photo
   * that the user may have set
   */
  get fullPhoto() {
    if (!this.full) return null;
    const photo$1 = this.full.personalPhoto ?? this.full.profilePhoto ?? this.full.fallbackPhoto;
    if (photo$1?._ !== "photo") return null;
    return new photo.Photo(photo$1);
  }
  /**
   * A custom photo (set by the current user) that should be displayed
   * instead of the actual chat photo.
   *
   * Currently only available for users.
   */
  get personalPhoto() {
    if (!this.full || this.full._ !== "userFull") return null;
    if (this.full.personalPhoto?._ !== "photo") return null;
    return new photo.Photo(this.full.personalPhoto);
  }
  /**
   * Actual profile photo of the user, bypassing the custom one.
   */
  get realPhoto() {
    if (!this.full) return null;
    if (this.full._ !== "userFull") return this.fullPhoto;
    if (this.full.personalPhoto?._ !== "photo") return null;
    return new photo.Photo(this.full.personalPhoto);
  }
  /**
   * A photo that the user has set to be shown
   * in case their actual profile photo is not available
   * due to privacy settings.
   *
   * Currently only available for users.
   */
  get publicPhoto() {
    if (!this.full || this.full._ !== "userFull") return null;
    if (this.full.fallbackPhoto?._ !== "photo") return null;
    return new photo.Photo(this.full.fallbackPhoto);
  }
  /**
   * Bio of the other party in a private chat, or description of a
   * group, supergroup or channel.
   */
  get bio() {
    return this.full.about ?? "";
  }
  /** Note about the user, only visible to you */
  get note() {
    return this.full.note ?? null;
  }
  /** Number of star gifts the user has chosen to display on their profile */
  get starGiftsCount() {
    return this.full.stargiftsCount ?? 0;
  }
  /** If this chat is a bot, whether the current user can manage its emoji status */
  get canManageBotEmojiStatus() {
    return this.full.botCanManageEmojiStatus;
  }
  get peerSettings() {
    return this.full.settings ?? null;
  }
  /** Notification settings for this chat */
  get notificationsSettings() {
    return this.full.notifySettings ?? null;
  }
  /** ID of the last pinned message in the chat with this user */
  get pinnedMsgId() {
    return this.full.pinnedMsgId ?? null;
  }
  /** Number of common chats with this user */
  get commonChatsCount() {
    return this.full.commonChatsCount;
  }
  /** Anonymized text to be shown instead of the user's name when forwarding their messages */
  get privateForwardName() {
    return this.full.privateForwardName ?? null;
  }
  /** Suggested admin rights for groups for this bot */
  get botGroupAdminRights() {
    return this.full.botGroupAdminRights ?? null;
  }
  /** Suggested admin rights for channels for this bot */
  get botBroadcastAdminRights() {
    return this.full.botBroadcastAdminRights ?? null;
  }
  /** Information about the user's birthday */
  get birthday() {
    return this.full.birthday ?? null;
  }
  /**
   * Information about the user's personal channel.
   */
  get personalChannel() {
    if (this.full.personalChannelId == null) return null;
    return new chat.Chat(this.peers.chat(this.full.personalChannelId));
  }
  /** ID of the last message in {@link personalChannel} */
  get personalChannelMessageId() {
    return this.full.personalChannelMessage ?? null;
  }
  /**
   * TTL of all messages in this chat, in seconds
   */
  get ttlPeriod() {
    return this.full?.ttlPeriod ?? null;
  }
  /**
   * Whether the current user has enabled content protection for own messages in DMs with this user,
   * i.e. our outgoing messages to this user will be `noforwards`.
   */
  get contentProtectionOutgoing() {
    return this.full.noforwardsMyEnabled;
  }
  /**
   * Whether this user has enabled content protection for their messages in DMs with the current user,
   * i.e. our incoming messages from this user will be `noforwards`.
   */
  get contentProtectionIncoming() {
    return this.full.noforwardsPeerEnabled;
  }
  /** Whether this user has been flagged by the server as a security risk due to using an unofficial app */
  get unofficialSecurityRisk() {
    return this.full.unofficialSecurityRisk;
  }
  /** Bot that manages this bot, if any */
  get botManager() {
    return this.full.botManagerId ? new user.User(this.peers.user(this.full.botManagerId)) : null;
  }
  /** Whether the "Gifts" tab should be shown */
  get showGifts() {
    return this.full.displayGiftsButton;
  }
  /** Information about the user's gift receive settings */
  get disallowedGifts() {
    return this.full.disallowedGifts ?? null;
  }
  /**
   * If this is a business account, information about the business.
   */
  get business() {
    if (this.full.businessWorkHours != null || this.full.businessLocation != null || this.full.businessGreetingMessage != null || this.full.businessAwayMessage != null || this.full.businessIntro != null) {
      return new businessAccount.BusinessAccount(this.full);
    }
    return null;
  }
  /** Information about a bot-issued verification (if any) */
  get botVerification() {
    if (!this.full.botVerification) return null;
    return new botVerification.BotVerification(this.full.botVerification);
  }
  /**
   * Information about the user's stars rating
   */
  get starsRating() {
    return this.full.starsRating ?? null;
  }
  /**
   * Information about the user's stars rating
   *
   * Only available for the current user
   */
  get starsRatingPending() {
    if (!this.full.starsMyPendingRating) return null;
    return {
      rating: this.full.starsMyPendingRating,
      date: new Date(this.full.starsMyPendingRatingDate * 1e3)
    };
  }
  /** User's last saved audio file, to be displayed in the profile under their name */
  get lastSavedMusic() {
    if (this.full.savedMusic?._ !== "document") return null;
    const doc = documentUtils.parseDocument(this.full.savedMusic);
    if (doc.type !== "audio") return null;
    return doc;
  }
}
const FullUser$1 = /* @__PURE__ */ memoize.memoizeGetters(FullUser, [
  "fullPhoto",
  "personalPhoto",
  "realPhoto",
  "publicPhoto",
  "personalChannel",
  "business",
  "botVerification",
  "starsRatingPending",
  "lastSavedMusic"
]);
const FullUser$2 = /* @__PURE__ */ inspectable.makeInspectable(FullUser$1);
exports.FullUser = FullUser$2;
