// @mtcute/core/highlevel/types/peers/full-chat.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Photo as Photo$2 } from "../media/photo.js";
import { StickerSet as StickerSet$2 } from "../misc/sticker-set.js";
import { PeerStories as PeerStories$2 } from "../stories/peer-stories.js";
import { BotInfo as BotInfo$2 } from "./bot-info.js";
import { BotVerification as BotVerification$1 } from "./bot-verification.js";
import { ChatInviteLink as ChatInviteLink$2 } from "./chat-invite-link.js";
import { Chat as Chat$2 } from "./chat.js";
import { PeersIndex } from "./peers-index.js";
import { User as User$2 } from "./user.js";
class FullChat extends Chat$2 {
  peers;
  full;
  constructor(obj) {
    const peers = PeersIndex.from(obj);
    super(peers.chat(obj.fullChat.id));
    this.peers = peers;
    this.full = obj.fullChat;
  }
  /** Whether this chat is archived */
  get isArchived() {
    return this.full._ === "channelFull" && this.full.folderId === 1;
  }
  /** Whether paid reactions are enabled for this channel */
  get hasPaidReactions() {
    return this.full?._ === "channelFull" && this.full.paidReactionsAvailable;
  }
  /** Whether this channel has hidden participants */
  get hasHiddenParticipants() {
    return this.full._ === "channelFull" && this.full.participantsHidden;
  }
  /** Whether the current user can change the chat's username */
  get canSetUsername() {
    return this.full.canSetUsername;
  }
  /** Whether the current user can change the chat's sticker set */
  get canSetStickers() {
    return this.full._ === "channelFull" && this.full.canSetStickers;
  }
  /** Whether the history before we joined the chat is hidden */
  get hasHiddenHistory() {
    return this.full._ === "channelFull" && this.full.hiddenPrehistory;
  }
  /** Whether there are scheduled messages in this chat */
  get hasScheduledMessages() {
    return this.full._ === "channelFull" && this.full.hasScheduled;
  }
  /** Whether the current user can view the chat's statistics */
  get canViewStats() {
    return this.full._ === "channelFull" && this.full.canViewStats;
  }
  /** Whether the current user can view the chat's participants list */
  get canViewParticipants() {
    return this.full._ === "channelFull" && this.full.canViewParticipants;
  }
  /** Whether the current user can delete the chat */
  get canDeleteChat() {
    switch (this.full._) {
      case "channelFull":
        return this.full.canDeleteChannel;
      case "chatFull":
        return this.raw._ === "chat" && this.raw.creator;
    }
  }
  /**
   * Whether the current user has blocked any anonumous admin of the supergroup.
   * If set, you won't receive mentions from them, nor their replies in `@replies`
   */
  get isBlocked() {
    return this.full._ === "channelFull" && this.full.blocked;
  }
  /** Whether the native antispam is enabled for this channel */
  get hasNativeAntispam() {
    return this.full._ === "channelFull" && this.full.antispam;
  }
  /** Whether the real-time translations popup should not be shown for this channel */
  get hasTranslationsDisabled() {
    return this.full._ === "channelFull" && this.full.translationsDisabled;
  }
  /** Whether this chat has pinned stories */
  get hasPinnedStories() {
    return this.full._ === "channelFull" && this.full.storiesPinnedAvailable;
  }
  /**
   * Whether ads on this channels were disabled
   * (this flag is only visible to channel owner)
   */
  get hasAdsDisabled() {
    return this.full._ === "channelFull" && this.full.restrictedSponsored;
  }
  /** Whether sending paid media is available in this channel */
  get isPaidMediaAvailable() {
    return this.full._ === "channelFull" && this.full.paidMediaAllowed;
  }
  /**
   * Full information about this chat's photo, if any.
   *
   * Unlike {@link Chat.photo}, this field contains additional information
   * about the photo, such as its date, more sizes, and is the only
   * way to get the animated profile photo.
   *
   * This field takes into account any personal/fallback photo
   * that the user may have set
   */
  get fullPhoto() {
    if (!this.full) return null;
    const photo = this.full.chatPhoto;
    if (photo?._ !== "photo") return null;
    return new Photo$2(photo);
  }
  /**
   * Bio of the other party in a private chat, or description of a
   * group, supergroup or channel.
   */
  get bio() {
    return this.full.about;
  }
  /**
   * Chat's primary invite link, if available
   */
  get inviteLink() {
    switch (this.full.exportedInvite?._) {
      case "chatInvitePublicJoinRequests":
        return null;
      case "chatInviteExported":
        return new ChatInviteLink$2(this.full.exportedInvite);
    }
    return null;
  }
  /**
   * For supergroups, information about the group sticker set.
   */
  get stickerSet() {
    if (this.full?._ !== "channelFull" || !this.full.stickerset) return null;
    return new StickerSet$2(this.full.stickerset);
  }
  /**
   * For supergroups, information about the group emoji set.
   */
  get emojiSet() {
    if (this.full?._ !== "channelFull" || !this.full.emojiset) return null;
    return new StickerSet$2(this.full.emojiset);
  }
  /**
   * Whether the group sticker set can be changed by you.
   */
  get canSetStickerSet() {
    return this.full && this.full._ === "channelFull" ? this.full.canSetStickers ?? null : null;
  }
  /**
   * Number of boosts applied by the current user to this chat.
   */
  get boostsApplied() {
    if (!this.full || this.full._ !== "channelFull") return 0;
    return this.full?.boostsApplied ?? 0;
  }
  /**
   * Number of boosts required for the user to be unrestricted in this chat.
   */
  get boostsForUnrestrict() {
    if (!this.full || this.full._ !== "channelFull") return 0;
    return this.full?.boostsUnrestrict ?? 0;
  }
  /** Whether the current user can view Telegram Stars revenue for this chat */
  get canViewStarsRevenue() {
    return this.full._ === "channelFull" && this.full.canViewStarsRevenue;
  }
  /** Whether the current user can view ad revenue for this chat */
  get canViewAdRevenue() {
    return this.full._ === "channelFull" && this.full.canViewRevenue;
  }
  /** Number of members in the chat (0 if not available) */
  get membersCount() {
    return this.full._ === "channelFull" ? this.full.participantsCount ?? 0 : 0;
  }
  /** Number of admins in the chat (0 if not available) */
  get adminsCount() {
    return this.full._ === "channelFull" ? this.full.adminsCount ?? 0 : 0;
  }
  /** Number of users kicked from the chat (if available) */
  get kickedCount() {
    return this.full._ === "channelFull" ? this.full.kickedCount ?? null : null;
  }
  /** Number of users kicked from the chat (if available) */
  get bannedCount() {
    return this.full._ === "channelFull" ? this.full.bannedCount ?? null : null;
  }
  /** Number of members of the chat that are currently online */
  get onlineCount() {
    return this.full._ === "channelFull" ? this.full.onlineCount ?? 0 : 0;
  }
  /** ID of the last read incoming message in the chat */
  get readInboxMaxId() {
    return this.full._ === "channelFull" ? this.full.readInboxMaxId : 0;
  }
  /** ID of the last read outgoing message in the chat */
  get readOutboxMaxId() {
    return this.full._ === "channelFull" ? this.full.readOutboxMaxId : 0;
  }
  /** Number of unread messages in the chat */
  get unreadCount() {
    return this.full._ === "channelFull" ? this.full.unreadCount : 0;
  }
  /** Notification settings for this chat */
  get notificationsSettings() {
    return this.full._ === "channelFull" ? this.full.notifySettings ?? null : null;
  }
  /** Information about bots in this chat */
  get botInfo() {
    return this.full._ === "channelFull" ? this.full.botInfo.map((it) => new BotInfo$2(it)) : [];
  }
  /** Bot that approves join requests in this chat, if any */
  get guardBot() {
    if (this.full._ !== "channelFull" || !this.full.guardBotId) return null;
    return new User$2(this.peers.user(this.full.guardBotId));
  }
  /**
   * For supergroups, ID of the basic group from which this supergroup was upgraded,
   * and the identifier of the last message from the original group.
   */
  get migratedFrom() {
    if (this.full._ !== "channelFull") return null;
    if (!this.full.migratedFromChatId || !this.full.migratedFromMaxId) return null;
    return {
      chatId: this.full.migratedFromChatId.toNumber(),
      msgId: this.full.migratedFromMaxId
    };
  }
  /** For supergroups with hidden history, ID of the first message visible to the current user */
  get historyMinId() {
    return this.full._ === "channelFull" ? this.full.availableMinId ?? 0 : 0;
  }
  /** ID of the last pinned message in the chat */
  get pinnedMsgId() {
    return this.full._ === "channelFull" ? this.full.pinnedMsgId ?? null : null;
  }
  /**
   * Information about a linked chat:
   * - for channels: the discussion group
   * - for supergroups: the linked channel
   */
  get linkedChat() {
    if (this.full._ !== "channelFull") return null;
    if (!this.full.linkedChatId) return null;
    return new Chat$2(this.peers.chat(this.full.linkedChatId));
  }
  /**
   * Information about a linked monoforum chat:
   * - for channels: the linked monoforum
   * - for monoforums: the channel this monoforum is linked to
   */
  get monoforumLinkedChat() {
    if (this.raw._ !== "channel") return null;
    if (!this.raw.linkedMonoforumId) return null;
    return new Chat$2(this.peers.chat(this.raw.linkedMonoforumId));
  }
  /**
   * TTL of all messages in this chat, in seconds
   */
  get ttlPeriod() {
    return this.full?.ttlPeriod ?? null;
  }
  /** Slowmode delay for this chat, if any */
  get slowmodeSeconds() {
    return this.full._ === "channelFull" ? this.full.slowmodeSeconds ?? null : null;
  }
  /**
   * For supergroups with slow mode, date of the next time the current user
   * will be able to send a message. `null` if they can send messages now.
   */
  get slowmodeNextSendDate() {
    if (this.full._ !== "channelFull" || !this.full.slowmodeNextSendDate) return null;
    return new Date(this.full.slowmodeNextSendDate * 1e3);
  }
  /** Number of pending join requests in the chat (only visible to admins) */
  get pendingJoinRequests() {
    return this.full._ === "channelFull" ? this.full.requestsPending ?? 0 : 0;
  }
  /** Users who have recently requested to join the chat (only visible to admins) */
  get recentRequesters() {
    if (this.full._ !== "channelFull" || !this.full.recentRequesters) return [];
    return this.full.recentRequesters.map((it) => new User$2(this.peers.user(it)));
  }
  /** Reactions available in this chat */
  get availableReactions() {
    return this.full._ === "channelFull" ? this.full.availableReactions ?? null : null;
  }
  /** Maximum number of unique reactions on a single message in this chat, if any */
  get maxReactions() {
    return this.full._ === "channelFull" ? this.full.reactionsLimit ?? null : null;
  }
  /** Channel stories */
  get stories() {
    if (this.full._ !== "channelFull" || !this.full.stories) return null;
    return new PeerStories$2(this.full.stories, this.peers);
  }
  /** Information about a bot-issued verification (if any) */
  get botVerification() {
    if (this.full._ !== "channelFull" || !this.full.botVerification) return null;
    return new BotVerification$1(this.full.botVerification);
  }
  /** Whether star gifts are available in this chat */
  get starGiftsAvailable() {
    return this.full._ === "channelFull" && this.full.stargiftsAvailable;
  }
  /** Number of star gifts available in this chat */
  get starGiftsCount() {
    return this.full._ === "channelFull" ? this.full.stargiftsCount ?? 0 : 0;
  }
  /** Whether you can enable paid messages in this chat */
  get paidMessagesAvailable() {
    return this.full._ === "channelFull" && this.full.paidMessagesAvailable;
  }
}
const FullChat$1 = /* @__PURE__ */ memoizeGetters(FullChat, [
  "fullPhoto",
  "inviteLink",
  "stickerSet",
  "emojiSet",
  "botInfo",
  "guardBot",
  "linkedChat",
  "recentRequesters",
  "stories",
  "botVerification"
]);
const FullChat$2 = /* @__PURE__ */ makeInspectable(FullChat$1);
export {
  FullChat$2 as FullChat
};
