// @mtcute/core/highlevel/types/peers/peer-settings.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "./user.js";
class PeerSettings {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Action bar that should be displayed in the chat with the user */
  get actionBar() {
    if (this.raw.requestChatTitle) {
      return {
        type: "join_request",
        chatTitle: this.raw.requestChatTitle,
        isBroadcast: this.raw.requestChatBroadcast,
        requestDate: new Date(this.raw.requestChatDate ?? 0)
      };
    }
    if (this.raw.reportGeo) {
      return null;
    }
    if (this.raw.inviteMembers) {
      return { type: "invite_members" };
    }
    if (this.raw.shareContact) {
      return { type: "share_phone_number" };
    }
    if (this.raw.blockContact || this.raw.reportSpam) {
      return {
        type: "add_or_report",
        canReportSpam: this.raw.reportSpam,
        canUnarchive: this.raw.autoarchived
      };
    }
    if (this.raw.addContact) {
      return {
        type: "add_contact",
        needPhoneNumberPrivacyException: this.raw.needContactsException
      };
    }
    if (this.raw.businessBotId && this._peers) {
      return {
        type: "business_bot",
        bot: new User$2(this._peers.user(this.raw.businessBotId)),
        manageUrl: this.raw.businessBotManageUrl ?? ""
      };
    }
    return null;
  }
  /** Distance in meters between us and the user */
  get geoDistance() {
    return this.raw.geoDistance ?? null;
  }
  /**
   * Price of paid messages to this peer for the current user
   * (note that this might differ from {@link User.paidMessagePrice},
   * because the user might have added us to their contacts or wrote first,
   * in which case the messages are free regardless)
   */
  get paidMessagePrice() {
    return this.raw.chargePaidMessageStars ?? null;
  }
  /**
   * Month and year when the user registered.
   * Only available for users who contacted us first.
   */
  get registrationDate() {
    return this.raw.registrationMonth ?? null;
  }
  /**
   * ISO country code of the user's phone number.
   * Only available for users who contacted us first.
   */
  get phoneCountry() {
    return this.raw.phoneCountry ?? null;
  }
  /**
   * Date when the user has last changed their name.
   * Only available for users who contacted us first.
   */
  get nameChangeDate() {
    return this.raw.nameChangeDate ? new Date(this.raw.nameChangeDate * 1e3) : null;
  }
  /**
   * Date when the user has last changed their photo.
   * Only available for users who contacted us first.
   */
  get photoChangeDate() {
    return this.raw.photoChangeDate ? new Date(this.raw.photoChangeDate * 1e3) : null;
  }
}
const PeerSettings$1 = /* @__PURE__ */ makeInspectable(PeerSettings);
const PeerSettings$2 = /* @__PURE__ */ memoizeGetters(PeerSettings$1, ["actionBar"]);
export {
  PeerSettings$2 as PeerSettings
};
