// @mtcute/core/highlevel/types/peers/peers-index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const ERROR_MSG = "Given peer is not available in this index. This is most likely an internal library error.";
class PeersIndex {
  constructor(users = /* @__PURE__ */ new Map(), chats = /* @__PURE__ */ new Map()) {
    this.users = users;
    this.chats = chats;
  }
  hasMin = false;
  static from(obj) {
    const index = new PeersIndex();
    obj.users?.forEach((user) => {
      if (user._ === "userEmpty") return;
      index.users.set(user.id, user);
      if (user.min) {
        index.hasMin = true;
      }
    });
    obj.chats?.forEach((chat) => {
      if (chat._ === "chatEmpty") return;
      index.chats.set(chat.id, chat);
      if ("min" in chat && chat.min) {
        index.hasMin = true;
      }
    });
    return index;
  }
  get empty() {
    return this.users.size === 0 && this.chats.size === 0;
  }
  user(id) {
    const r = this.users.get(id);
    if (!r) {
      throw new errors.MtArgumentError(ERROR_MSG);
    }
    return r;
  }
  chat(id) {
    const r = this.chats.get(id);
    if (!r) {
      throw new errors.MtArgumentError(ERROR_MSG);
    }
    return r;
  }
  get(peer) {
    switch (peer._) {
      case "peerUser":
        return this.user(peer.userId);
      case "peerChat":
        return this.chat(peer.chatId);
      case "peerChannel":
        return this.chat(peer.channelId);
    }
  }
  has(peer) {
    switch (peer._) {
      case "peerUser":
        return this.users.has(peer.userId);
      case "peerChat":
        return this.chats.has(peer.chatId);
      case "peerChannel":
        return this.chats.has(peer.channelId);
    }
  }
}
exports.PeersIndex = PeersIndex;
