// @mtcute/core/highlevel/types/peers/peer-settings.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const user = require("./user.cjs");
class PeerSettings {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Action bar that should be displayed in the chat with the user */
  get actionBar() {
    if (this.raw.requestChatTitle) {
      return {
        type: "join_request",
        chatTitle: this.raw.requestChatTitle,
        isBroadcast: this.raw.requestChatBroadcast,
        requestDate: new Date(this.raw.requestChatDate ?? 0)
      };
    }
    if (this.raw.reportGeo) {
      return null;
    }
    if (this.raw.inviteMembers) {
      return { type: "invite_members" };
    }
    if (this.raw.shareContact) {
      return { type: "share_phone_number" };
    }
    if (this.raw.blockContact || this.raw.reportSpam) {
      return {
        type: "add_or_report",
        canReportSpam: this.raw.reportSpam,
        canUnarchive: this.raw.autoarchived
      };
    }
    if (this.raw.addContact) {
      return {
        type: "add_contact",
        needPhoneNumberPrivacyException: this.raw.needContactsException
      };
    }
    if (this.raw.businessBotId && this._peers) {
      return {
        type: "business_bot",
        bot: new user.User(this._peers.user(this.raw.businessBotId)),
        manageUrl: this.raw.businessBotManageUrl ?? ""
      };
    }
    return null;
  }
  /** Distance in meters between us and the user */
  get geoDistance() {
    return this.raw.geoDistance ?? null;
  }
  /**
   * Price of paid messages to this peer for the current user
   * (note that this might differ from {@link User.paidMessagePrice},
   * because the user might have added us to their contacts or wrote first,
   * in which case the messages are free regardless)
   */
  get paidMessagePrice() {
    return this.raw.chargePaidMessageStars ?? null;
  }
  /**
   * Month and year when the user registered.
   * Only available for users who contacted us first.
   */
  get registrationDate() {
    return this.raw.registrationMonth ?? null;
  }
  /**
   * ISO country code of the user's phone number.
   * Only available for users who contacted us first.
   */
  get phoneCountry() {
    return this.raw.phoneCountry ?? null;
  }
  /**
   * Date when the user has last changed their name.
   * Only available for users who contacted us first.
   */
  get nameChangeDate() {
    return this.raw.nameChangeDate ? new Date(this.raw.nameChangeDate * 1e3) : null;
  }
  /**
   * Date when the user has last changed their photo.
   * Only available for users who contacted us first.
   */
  get photoChangeDate() {
    return this.raw.photoChangeDate ? new Date(this.raw.photoChangeDate * 1e3) : null;
  }
}
const PeerSettings$1 = /* @__PURE__ */ inspectable.makeInspectable(PeerSettings);
const PeerSettings$2 = /* @__PURE__ */ memoize.memoizeGetters(PeerSettings$1, ["actionBar"]);
exports.PeerSettings = PeerSettings$2;
