// @fuman/io/codec/delimiter.js
import { typed } from "@fuman/utils";
class DelimiterCodec {
  /**
   * @param delimiter  delimiter to use
   * @param options  options
   */
  constructor(delimiter, options) {
    this.delimiter = delimiter;
    this.options = options;
    this.#strategy = options?.strategy ?? "discard";
  }
  #strategy;
  decode(buf, eof) {
    const { delimiter } = this;
    const data = buf.readSync(buf.available);
    if (eof && data.length === 0) {
      return null;
    }
    const delimiterIdx = typed.indexOfArray(data, delimiter);
    if (delimiterIdx === -1) {
      if (eof) {
        return data;
      }
      buf.rewind(data.length);
      return null;
    }
    buf.rewind(data.length - delimiterIdx - delimiter.length);
    const frameEnd = this.#strategy === "keep" ? delimiterIdx + delimiter.length : delimiterIdx;
    return data.slice(0, frameEnd);
  }
  encode(data, into) {
    const buf = into.writeSync(data.length + this.delimiter.length);
    buf.set(data);
    buf.set(this.delimiter, data.length);
    into.disposeWriteSync();
  }
  reset() {
  }
}
export {
  DelimiterCodec
};
