// @mtcute/core/highlevel/methods/invite-links/create-invite-link.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const miscUtils = require("../../utils/misc-utils.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
const chatInviteLink = require("../../types/peers/chat-invite-link.cjs");
async function createInviteLink(client, chatId, params) {
  if (!params) params = {};
  const res = await client.call({
    _: "messages.exportChatInvite",
    peer: await resolvePeer.resolvePeer(client, chatId),
    expireDate: miscUtils.normalizeDate(params.expires),
    usageLimit: params.usageLimit,
    requestNeeded: params.withApproval,
    subscriptionPricing: params.subscriptionPricing
  });
  return new chatInviteLink.ChatInviteLink(res);
}
exports.createInviteLink = createInviteLink;
