// @mtcute/core/highlevel/methods/invite-links/get-invite-links.js
import { resolveUser, resolvePeer } from "../users/resolve-peer.js";
import { PeersIndex } from "../../types/peers/peers-index.js";
import { ChatInviteLink as ChatInviteLink$2 } from "../../types/peers/chat-invite-link.js";
import { makeArrayPaginated } from "../../utils/misc-utils.js";
async function getInviteLinks(client, chatId, params) {
  if (!params) params = {};
  const { revoked = false, limit = Infinity, admin, offset } = params;
  const res = await client.call({
    _: "messages.getExportedChatInvites",
    peer: await resolvePeer(client, chatId),
    revoked,
    adminId: admin ? await resolveUser(client, admin) : { _: "inputUserSelf" },
    limit,
    offsetDate: offset?.date,
    offsetLink: offset?.link
  });
  const peers = PeersIndex.from(res);
  const links = res.invites.map((it) => new ChatInviteLink$2(it, peers));
  const last = links[links.length - 1];
  const nextOffset = last ? {
    date: last.raw.date,
    link: last.raw.link
  } : void 0;
  return makeArrayPaginated(links, res.count, nextOffset);
}
export {
  getInviteLinks
};
