// @mtcute/core/highlevel/methods/invite-links/get-invite-links.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const resolvePeer = require("../users/resolve-peer.cjs");
const peersIndex = require("../../types/peers/peers-index.cjs");
const chatInviteLink = require("../../types/peers/chat-invite-link.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
async function getInviteLinks(client, chatId, params) {
  if (!params) params = {};
  const { revoked = false, limit = Infinity, admin, offset } = params;
  const res = await client.call({
    _: "messages.getExportedChatInvites",
    peer: await resolvePeer.resolvePeer(client, chatId),
    revoked,
    adminId: admin ? await resolvePeer.resolveUser(client, admin) : { _: "inputUserSelf" },
    limit,
    offsetDate: offset?.date,
    offsetLink: offset?.link
  });
  const peers = peersIndex.PeersIndex.from(res);
  const links = res.invites.map((it) => new chatInviteLink.ChatInviteLink(it, peers));
  const last = links[links.length - 1];
  const nextOffset = last ? {
    date: last.raw.date,
    link: last.raw.link
  } : void 0;
  return miscUtils.makeArrayPaginated(links, res.count, nextOffset);
}
exports.getInviteLinks = getInviteLinks;
