// @mtcute/core/storage/sqlite/repository/peers.js
import { MtcuteError } from "../../../types/errors.js";
function mapPeerDto(dto) {
  return {
    id: dto.id,
    accessHash: dto.hash,
    isMin: dto.isMin === 1,
    usernames: JSON.parse(dto.usernames),
    updated: dto.updated,
    phone: dto.phone || void 0,
    complete: dto.complete
  };
}
class SqlitePeersRepository {
  constructor(_driver) {
    this._driver = _driver;
    _driver.registerMigration("peers", 1, (db) => {
      db.exec(`
                create table peers (
                    id integer primary key,
                    hash text not null,
                    usernames json not null,
                    updated integer not null,
                    phone text,
                    complete blob
                );
                create index idx_peers_usernames on peers (usernames);
                create index idx_peers_phone on peers (phone);
            `);
    });
    _driver.registerMigration("peers", 2, (db) => {
      db.exec("alter table peers add column isMin integer not null default false;");
    });
    _driver.onLoad((db) => {
      this._loaded = true;
      this._store = db.prepare(
        "insert or replace into peers (id, hash, isMin, usernames, updated, phone, complete) values (?, ?, ?, ?, ?, ?, ?)"
      );
      this._getById = db.prepare("select * from peers where id = ?");
      this._getByUsername = db.prepare(
        "select * from peers where exists (select 1 from json_each(usernames) where value = ?) and isMin = false"
      );
      this._getByPhone = db.prepare("select * from peers where phone = ? and isMin = false");
      this._delAll = db.prepare("delete from peers");
    });
  }
  _loaded = false;
  _ensureLoaded() {
    if (!this._loaded) {
      throw new MtcuteError("Peers repository is not loaded. Have you called client.start() (or similar)?");
    }
  }
  _store;
  store(peer) {
    this._driver._writeLater(this._store, [
      peer.id,
      peer.accessHash,
      peer.isMin ? 1 : 0,
      JSON.stringify(peer.usernames),
      peer.updated,
      peer.phone ?? null,
      peer.complete
    ]);
  }
  _getById;
  getById(id) {
    this._ensureLoaded();
    const row = this._getById.get(id);
    if (!row) return null;
    return mapPeerDto(row);
  }
  _getByUsername;
  getByUsername(username) {
    this._ensureLoaded();
    const row = this._getByUsername.get(username);
    if (!row) return null;
    return mapPeerDto(row);
  }
  _getByPhone;
  getByPhone(phone) {
    this._ensureLoaded();
    const row = this._getByPhone.get(phone);
    if (!row) return null;
    return mapPeerDto(row);
  }
  _delAll;
  deleteAll() {
    this._delAll.run();
  }
}
export {
  SqlitePeersRepository
};
