// @mtcute/core/storage/sqlite/repository/ref-messages.js
class SqliteRefMessagesRepository {
  constructor(_driver) {
    this._driver = _driver;
    _driver.registerMigration("ref_messages", 1, (db) => {
      db.exec(`
                create table if not exists message_refs (
                    peer_id integer not null,
                    chat_id integer not null,
                    msg_id integer not null
                );
                create index if not exists idx_message_refs_peer on message_refs (peer_id);
                create index if not exists idx_message_refs on message_refs (chat_id, msg_id);
            `);
    });
    _driver.onLoad(() => {
      this._store = this._driver.db.prepare(
        "insert or replace into message_refs (peer_id, chat_id, msg_id) values (?, ?, ?)"
      );
      this._getByPeer = this._driver.db.prepare("select chat_id, msg_id from message_refs where peer_id = ?");
      this._del = this._driver.db.prepare("delete from message_refs where chat_id = ? and msg_id = ?");
      this._delByPeer = this._driver.db.prepare("delete from message_refs where peer_id = ?");
      this._delAll = this._driver.db.prepare("delete from message_refs");
    });
  }
  _store;
  store(peerId, chatId, msgId) {
    this._store.run(peerId, chatId, msgId);
  }
  _getByPeer;
  getByPeer(peerId) {
    const res = this._getByPeer.get(peerId);
    if (!res) return null;
    const res_ = res;
    return [res_.chat_id, res_.msg_id];
  }
  _del;
  delete(chatId, msgIds) {
    for (const msgId of msgIds) {
      this._driver._writeLater(this._del, [chatId, msgId]);
    }
  }
  _delByPeer;
  deleteByPeer(peerId) {
    this._delByPeer.run(peerId);
  }
  _delAll;
  deleteAll() {
    this._delAll.run();
  }
}
export {
  SqliteRefMessagesRepository
};
