// @mtcute/core/utils/dcs.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const tlRuntime = require("@mtcute/tl-runtime");
function serializeBasicDcOption(dc) {
  const writer = tlRuntime.TlBinaryWriter.manual(64);
  const flags = (dc.ipv6 ? 1 : 0) | (dc.mediaOnly ? 2 : 0) | (dc.testMode ? 4 : 0);
  writer.raw(
    new Uint8Array([
      2,
      // version
      dc.id,
      flags
    ])
  );
  writer.string(dc.ipAddress);
  writer.int(dc.port);
  return writer.result();
}
function parseBasicDcOption(data) {
  const reader = tlRuntime.TlBinaryReader.manual(data);
  const [version, id, flags] = reader.raw(3);
  if (version !== 1 && version !== 2) return null;
  const ipAddress = reader.string();
  const port = reader.int();
  return {
    id,
    ipAddress,
    port,
    ipv6: (flags & 1) !== 0,
    mediaOnly: (flags & 2) !== 0,
    testMode: version === 2 && (flags & 4) !== 0
  };
}
const defaultProductionDc = {
  main: {
    ipAddress: "149.154.167.50",
    port: 443,
    id: 2
  },
  media: {
    ipAddress: "149.154.167.222",
    port: 443,
    id: 2,
    mediaOnly: true
  }
};
const defaultProductionIpv6Dc = {
  main: {
    ipAddress: "2001:067c:04e8:f002:0000:0000:0000:000a",
    ipv6: true,
    port: 443,
    id: 2
  },
  media: {
    ipAddress: "2001:067c:04e8:f002:0000:0000:0000:000b",
    ipv6: true,
    port: 443,
    id: 2,
    mediaOnly: true
  }
};
const defaultTestDc = {
  main: {
    ipAddress: "149.154.167.40",
    port: 443,
    id: 2,
    testMode: true
  },
  media: {
    ipAddress: "149.154.167.40",
    port: 443,
    id: 2,
    mediaOnly: true,
    testMode: true
  }
};
const defaultTestIpv6Dc = {
  main: {
    ipAddress: "2001:67c:4e8:f002::e",
    port: 443,
    ipv6: true,
    id: 2,
    testMode: true
  },
  media: {
    ipAddress: "2001:67c:4e8:f002::e",
    port: 443,
    ipv6: true,
    id: 2,
    mediaOnly: true,
    testMode: true
  }
};
exports.defaultProductionDc = defaultProductionDc;
exports.defaultProductionIpv6Dc = defaultProductionIpv6Dc;
exports.defaultTestDc = defaultTestDc;
exports.defaultTestIpv6Dc = defaultTestIpv6Dc;
exports.parseBasicDcOption = parseBasicDcOption;
exports.serializeBasicDcOption = serializeBasicDcOption;
