// @mtcute/core/utils/dcs.js
import { TlBinaryWriter, TlBinaryReader } from "@mtcute/tl-runtime";
function serializeBasicDcOption(dc) {
  const writer = TlBinaryWriter.manual(64);
  const flags = (dc.ipv6 ? 1 : 0) | (dc.mediaOnly ? 2 : 0) | (dc.testMode ? 4 : 0);
  writer.raw(
    new Uint8Array([
      2,
      // version
      dc.id,
      flags
    ])
  );
  writer.string(dc.ipAddress);
  writer.int(dc.port);
  return writer.result();
}
function parseBasicDcOption(data) {
  const reader = TlBinaryReader.manual(data);
  const [version, id, flags] = reader.raw(3);
  if (version !== 1 && version !== 2) return null;
  const ipAddress = reader.string();
  const port = reader.int();
  return {
    id,
    ipAddress,
    port,
    ipv6: (flags & 1) !== 0,
    mediaOnly: (flags & 2) !== 0,
    testMode: version === 2 && (flags & 4) !== 0
  };
}
const defaultProductionDc = {
  main: {
    ipAddress: "149.154.167.50",
    port: 443,
    id: 2
  },
  media: {
    ipAddress: "149.154.167.222",
    port: 443,
    id: 2,
    mediaOnly: true
  }
};
const defaultProductionIpv6Dc = {
  main: {
    ipAddress: "2001:067c:04e8:f002:0000:0000:0000:000a",
    ipv6: true,
    port: 443,
    id: 2
  },
  media: {
    ipAddress: "2001:067c:04e8:f002:0000:0000:0000:000b",
    ipv6: true,
    port: 443,
    id: 2,
    mediaOnly: true
  }
};
const defaultTestDc = {
  main: {
    ipAddress: "149.154.167.40",
    port: 443,
    id: 2,
    testMode: true
  },
  media: {
    ipAddress: "149.154.167.40",
    port: 443,
    id: 2,
    mediaOnly: true,
    testMode: true
  }
};
const defaultTestIpv6Dc = {
  main: {
    ipAddress: "2001:67c:4e8:f002::e",
    port: 443,
    ipv6: true,
    id: 2,
    testMode: true
  },
  media: {
    ipAddress: "2001:67c:4e8:f002::e",
    port: 443,
    ipv6: true,
    id: 2,
    mediaOnly: true,
    testMode: true
  }
};
export {
  defaultProductionDc,
  defaultProductionIpv6Dc,
  defaultTestDc,
  defaultTestIpv6Dc,
  parseBasicDcOption,
  serializeBasicDcOption
};
