// @mtcute/core/utils/early-timer.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
class EarlyTimer {
  _timeout;
  _timeoutTs;
  _handler = () => {
  };
  constructor() {
    this.emitNow = this.emitNow.bind(this);
  }
  /**
   * Emit the timer when the event loop is idle
   * (basically `queueMicrotask()`)
   */
  emitWhenIdle() {
    utils.timers.clearTimeout(this._timeout);
    this._timeoutTs = performance.now();
    if (typeof queueMicrotask !== "undefined") {
      queueMicrotask(this.emitNow);
    } else {
      this._timeout = utils.timers.setTimeout(this.emitNow, 0);
    }
  }
  /**
   * Emit the timer before the next given milliseconds
   *
   * Shorthand for `emitBefore(performance.now() + ms)`
   *
   * @param ms  Milliseconds to schedule for
   */
  emitBeforeNext(ms) {
    return this.emitBefore(performance.now() + ms);
  }
  /**
   * Emit the timer before the given time
   *
   * @param ts  timestamp in ms relative to `performance.timeOrigin`
   */
  emitBefore(ts) {
    if (!this._timeoutTs || ts < this._timeoutTs) {
      this.reset();
      const diff = ts - performance.now();
      if (diff > 0) {
        this._timeout = utils.timers.setTimeout(this.emitNow, diff);
        this._timeoutTs = ts;
      } else {
        this._handler();
      }
    }
  }
  /**
   * Emit the timer right now
   */
  emitNow() {
    this.reset();
    this._handler();
  }
  /**
   * Cancel the timer
   */
  reset() {
    utils.timers.clearTimeout(this._timeout);
    this._timeoutTs = void 0;
  }
  /**
   * Set timeout handler
   */
  onTimeout(handler) {
    this._handler = handler;
  }
}
exports.EarlyTimer = EarlyTimer;
