// @mtcute/core/utils/logger.js
import { hex } from "@fuman/utils";
import { isTlRpcError } from "./type-assertions.js";
import { RpcError } from "../tl/inner-tl.js";
const DEFAULT_LOG_LEVEL = 2;
const FORMATTER_RE = /%[a-z]/gi;
class Logger {
  constructor(mgr, tag, parent = mgr) {
    this.mgr = mgr;
    this.tag = tag;
    this.parent = parent;
    let hash = 0;
    for (let i = 0; i < tag.length; i++) {
      hash = (hash << 5) - hash + tag.charCodeAt(i);
      hash |= 0;
    }
    this.color = Math.abs(hash) % 6;
  }
  color;
  prefix = "";
  getPrefix() {
    let s = "";
    let obj = this;
    while (obj) {
      if (obj.prefix) s = obj.prefix + s;
      obj = obj.parent;
    }
    return s;
  }
  log(level, fmt, ...args) {
    if (level > this.mgr.level) return;
    if (!this.mgr["_filter"](this.tag)) return;
    if (fmt.includes("%h") || fmt.includes("%b") || fmt.includes("%j") || fmt.includes("%J") || fmt.includes("%l") || fmt.includes("%L") || fmt.includes("%e")) {
      let idx = 0;
      fmt = fmt.replace(FORMATTER_RE, (m) => {
        if (m === "%h" || m === "%b" || m === "%j" || m === "%J" || m === "%l" || m === "%L" || m === "%e") {
          let val = args[idx];
          args.splice(idx, 1);
          if (m === "%h") {
            if (ArrayBuffer.isView(val)) return hex.encode(val);
            if (typeof val === "number" || typeof val === "bigint") return val.toString(16);
            return String(val);
          }
          if (m === "%b") return String(Boolean(val));
          if (m === "%j" || m === "%J") {
            if (m === "%J") {
              val = [...val];
            }
            return JSON.stringify(val, (k, v) => {
              if (ArrayBuffer.isView(v) || typeof v === "object" && v.type === "Buffer" && Array.isArray(v.data)) {
                let str = v.data ? Buffer.from(v.data).toString("hex") : hex.encode(v);
                if (str.length > 300) {
                  str = `${str.slice(0, 300)}...`;
                }
                return str;
              }
              return v;
            });
          }
          if (m === "%l") return String(val);
          if (m === "%L") {
            if (!Array.isArray(val)) return "n/a";
            return `[${val.map(String).join(", ")}]`;
          }
          if (m === "%e") {
            if (isTlRpcError(val)) {
              return `${val.errorCode} ${val.errorMessage}`;
            }
            if (RpcError.is(val)) {
              return `${val.code} ${val.text}`;
            }
            return val && typeof val === "object" ? val.stack || val.message : String(val);
          }
        }
        idx++;
        return m;
      });
    }
    this.mgr.handler(this.color, level, this.tag, this.getPrefix() + fmt, args);
  }
  error = this.log.bind(this, LogManager.ERROR);
  warn = this.log.bind(this, LogManager.WARN);
  info = this.log.bind(this, LogManager.INFO);
  debug = this.log.bind(this, LogManager.DEBUG);
  verbose = this.log.bind(this, LogManager.VERBOSE);
  /**
   * Create a {@link Logger} with the given tag
   * from the same {@link LogManager} as the current
   * Logger.
   *
   * @param tag  Logger tag
   */
  create(tag) {
    return new Logger(this.mgr, tag, this);
  }
}
const defaultFilter = () => true;
class LogManager extends Logger {
  static OFF = 0;
  static ERROR = 1;
  static WARN = 2;
  static INFO = 3;
  static DEBUG = 4;
  static VERBOSE = 5;
  level;
  handler;
  constructor(tag = "base", platform) {
    super(null, tag);
    this.mgr = this;
    this.level = platform.getDefaultLogLevel() ?? DEFAULT_LOG_LEVEL;
    this.handler = platform.log.bind(platform);
  }
  _filter = defaultFilter;
  /**
   * Create a {@link Logger} with the given tag
   *
   * @param tag  Logger tag
   */
  create(tag) {
    return new Logger(this, tag);
  }
  /**
   * Filter logging by tags.
   *
   * @param cb
   */
  filter(cb) {
    this._filter = cb ?? defaultFilter;
  }
}
export {
  LogManager,
  Logger
};
