// @mtcute/core/utils/long-utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const Long = require("long");
const inspectable = require("../highlevel/utils/inspectable.cjs");
const miscUtils = require("./misc-utils.cjs");
function randomLong(unsigned = false) {
  const lo = miscUtils.getRandomInt(4294967295);
  const hi = miscUtils.getRandomInt(4294967295);
  return new Long(lo, hi, unsigned);
}
function longFromBuffer(buf, unsigned = false, le = true) {
  const dv = utils.typed.toDataView(buf);
  if (le) {
    return new Long(dv.getInt32(0, true), dv.getInt32(4, true), unsigned);
  }
  return new Long(dv.getInt32(4, false), dv.getInt32(0, false), unsigned);
}
function removeFromLongArray(arr, val) {
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i];
    if (v === val || v.eq(val)) {
      arr.splice(i, 1);
      return true;
    }
  }
  return false;
}
function compareLongs(a, b) {
  if (a.eq(b)) return 0;
  if (a.gt(b)) return 1;
  return -1;
}
function longToFastString(val) {
  return `${val.low}|${val.high}`;
}
function longFromFastString(val, unsigned = false) {
  const parts = val.split("|");
  if (parts.length !== 2) throw new Error(`Invalid long fast string: ${val}`);
  const low = Number.parseInt(parts[0]);
  const high = Number.parseInt(parts[1]);
  if (Number.isNaN(low) || Number.isNaN(high)) {
    throw new TypeError(`Invalid long fast string: ${val}`);
  }
  return new Long(low, high, unsigned);
}
class LongMap extends utils.CustomMap {
  constructor() {
    super(longToFastString, longFromFastString);
  }
}
const proto = class LongMap2 {
}.prototype;
LongMap.prototype[inspectable.customInspectSymbol] = function() {
  const res = Object.create(proto);
  this.forEach((v, k) => {
    res[k.toString()] = v;
  });
  return res;
};
class LongSet extends utils.CustomSet {
  constructor() {
    super(longToFastString, longFromFastString);
  }
}
exports.LongMap = LongMap;
exports.LongSet = LongSet;
exports.compareLongs = compareLongs;
exports.longFromBuffer = longFromBuffer;
exports.longFromFastString = longFromFastString;
exports.longToFastString = longToFastString;
exports.randomLong = randomLong;
exports.removeFromLongArray = removeFromLongArray;
