// @mtcute/core/utils/long-utils.js
import { typed, CustomMap, CustomSet } from "@fuman/utils";
import Long from "long";
import { customInspectSymbol } from "../highlevel/utils/inspectable.js";
import { getRandomInt } from "./misc-utils.js";
function randomLong(unsigned = false) {
  const lo = getRandomInt(4294967295);
  const hi = getRandomInt(4294967295);
  return new Long(lo, hi, unsigned);
}
function longFromBuffer(buf, unsigned = false, le = true) {
  const dv = typed.toDataView(buf);
  if (le) {
    return new Long(dv.getInt32(0, true), dv.getInt32(4, true), unsigned);
  }
  return new Long(dv.getInt32(4, false), dv.getInt32(0, false), unsigned);
}
function removeFromLongArray(arr, val) {
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i];
    if (v === val || v.eq(val)) {
      arr.splice(i, 1);
      return true;
    }
  }
  return false;
}
function compareLongs(a, b) {
  if (a.eq(b)) return 0;
  if (a.gt(b)) return 1;
  return -1;
}
function longToFastString(val) {
  return `${val.low}|${val.high}`;
}
function longFromFastString(val, unsigned = false) {
  const parts = val.split("|");
  if (parts.length !== 2) throw new Error(`Invalid long fast string: ${val}`);
  const low = Number.parseInt(parts[0]);
  const high = Number.parseInt(parts[1]);
  if (Number.isNaN(low) || Number.isNaN(high)) {
    throw new TypeError(`Invalid long fast string: ${val}`);
  }
  return new Long(low, high, unsigned);
}
class LongMap extends CustomMap {
  constructor() {
    super(longToFastString, longFromFastString);
  }
}
const proto = class LongMap2 {
}.prototype;
LongMap.prototype[customInspectSymbol] = function() {
  const res = Object.create(proto);
  this.forEach((v, k) => {
    res[k.toString()] = v;
  });
  return res;
};
class LongSet extends CustomSet {
  constructor() {
    super(longToFastString, longFromFastString);
  }
}
export {
  LongMap,
  LongSet,
  compareLongs,
  longFromBuffer,
  longFromFastString,
  longToFastString,
  randomLong,
  removeFromLongArray
};
