// @mtcute/core/utils/peer-utils.js
import { MtArgumentError } from "../types/errors.js";
import { assertNever } from "../types/utils.js";
const ZERO_CHANNEL_ID = -1e12;
const MIN_MARKED_CHAT_ID = -999999999999;
const MAX_USER_ID = 1099511627775;
function toggleChannelIdMark(id) {
  return ZERO_CHANNEL_ID - id;
}
function getBarePeerId(peer) {
  switch (peer._) {
    case "peerUser":
      return peer.userId;
    case "peerChat":
      return peer.chatId;
    case "peerChannel":
      return peer.channelId;
  }
}
function getMarkedPeerId(peer, peerType) {
  if (typeof peer === "number") {
    switch (peerType) {
      case "user":
        return peer;
      case "chat":
        return -peer;
      case "channel":
        return ZERO_CHANNEL_ID - peer;
    }
    throw new MtArgumentError("Invalid peer type");
  }
  switch (peer._) {
    case "mtcute.dummyInputPeerMinUser":
    case "peerUser":
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
    case "inputUser":
    case "inputUserFromMessage":
      return peer.userId;
    case "peerChat":
    case "inputPeerChat":
      return -peer.chatId;
    case "mtcute.dummyInputPeerMinChannel":
    case "peerChannel":
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
    case "inputChannel":
    case "inputChannelFromMessage":
      return ZERO_CHANNEL_ID - peer.channelId;
  }
  throw new MtArgumentError(`Invalid peer: ${peer._}`);
}
function parseMarkedPeerId(id) {
  if (id < 0) {
    if (MIN_MARKED_CHAT_ID <= id) {
      return ["chat", -id];
    }
    if (id !== ZERO_CHANNEL_ID) {
      return ["channel", ZERO_CHANNEL_ID - id];
    }
  } else if (id > 0 && id <= MAX_USER_ID) {
    return ["user", id];
  }
  throw new MtArgumentError(`Invalid marked peer id: ${id}`);
}
function* getAllPeersFrom(obj) {
  if (typeof obj !== "object") return;
  if (Array.isArray(obj)) {
    for (const it of obj) {
      yield* getAllPeersFrom(it);
    }
    return;
  }
  switch (obj._) {
    case "user":
    case "chat":
    case "channel":
    case "chatForbidden":
    case "channelForbidden":
      yield obj;
      return;
  }
  if ("user" in obj && typeof obj.user === "object" && obj.user._ === "user") {
    yield obj.user;
  }
  if ("chat" in obj && typeof obj.chat === "object") {
    switch (obj.chat._) {
      case "chat":
      case "channel":
      case "chatForbidden":
      case "channelForbidden":
        yield obj.chat;
        break;
    }
  }
  if ("channel" in obj && typeof obj.channel === "object") {
    switch (obj.channel._) {
      case "chat":
      case "channel":
      case "chatForbidden":
      case "channelForbidden":
        yield obj.channel;
        break;
    }
  }
  if ("users" in obj && Array.isArray(obj.users) && obj.users.length) {
    for (const user of obj.users) {
      if (typeof user === "object" && user._ === "user") {
        yield user;
      }
    }
  }
  if ("chats" in obj && Array.isArray(obj.chats) && obj.chats.length) {
    for (const chat of obj.chats) {
      if (typeof chat === "object") {
        switch (chat._) {
          case "chat":
          case "channel":
          case "chatForbidden":
          case "channelForbidden":
            yield chat;
            break;
        }
      }
    }
  }
}
function getBasicPeerType(id) {
  if (typeof id === "number") {
    if (id < 0) {
      if (MIN_MARKED_CHAT_ID <= id) {
        return "chat";
      }
      if (id !== ZERO_CHANNEL_ID) {
        return "channel";
      }
    } else if (id > 0 && id <= MAX_USER_ID) {
      return "user";
    }
    throw new Error(`Invalid marked peer id: ${id}`);
  }
  switch (id._) {
    case "peerUser":
    case "inputUser":
    case "inputUserSelf":
    case "inputPeerSelf":
    case "inputUserFromMessage":
    case "inputPeerUser":
    case "inputPeerUserFromMessage":
    case "inputUserEmpty":
    case "mtcute.dummyInputPeerMinUser":
      return "user";
    case "peerChat":
    case "inputPeerChat":
      return "chat";
    case "peerChannel":
    case "inputChannel":
    case "inputChannelEmpty":
    case "mtcute.dummyInputPeerMinChannel":
    case "inputChannelFromMessage":
    case "inputPeerChannel":
    case "inputPeerChannelFromMessage":
      return "channel";
    case "inputPeerEmpty":
      throw new Error("Empty peer");
    default:
      assertNever();
  }
}
export {
  getAllPeersFrom,
  getBarePeerId,
  getBasicPeerType,
  getMarkedPeerId,
  parseMarkedPeerId,
  toggleChannelIdMark
};
