// @mtcute/core/utils/sorted-array.js
class SortedArray {
  constructor(array = [], comparator) {
    this.comparator = comparator;
    this.raw = array.sort(comparator);
    this.comparator = comparator;
  }
  raw;
  get length() {
    return this.raw.length;
  }
  insert(item) {
    if (Array.isArray(item)) {
      let ind = -1;
      item.forEach((it) => {
        ind = this.insert(it);
      });
      return ind;
    }
    if (this.raw.length === 0) {
      this.raw.push(item);
      return 0;
    }
    let lo = 0;
    let hi = this.raw.length;
    while (lo < hi) {
      const mid = Math.floor((lo + hi) / 2);
      if (this.comparator(this.raw[mid], item) > 0) {
        hi = mid;
      } else {
        lo = mid + 1;
      }
    }
    this.raw.splice(lo, 0, item);
    return lo;
  }
  // closest: return the closest value (right-hand side)
  // meaning that raw[idx - 1] <= item <= raw[idx]
  // in other words, smallest idx such that raw[idx] >= item
  index(item, closest = false) {
    let lo = 0;
    let hi = this.raw.length;
    while (lo < hi) {
      const mid = Math.floor((lo + hi) / 2);
      const cmp = this.comparator(this.raw[mid], item);
      if (cmp === 0) {
        return mid;
      }
      if (cmp > 0) {
        hi = mid;
      } else {
        lo = mid + 1;
      }
    }
    return closest ? lo : -1;
  }
  remove(item) {
    const idx = this.raw.indexOf(item);
    if (idx === -1) return;
    this.raw.splice(idx, 1);
  }
  removeIndex(idx) {
    this.raw.splice(idx, 1);
  }
  includes(item) {
    return this.index(item) !== -1;
  }
  find(item) {
    const ind = this.index(item);
    return ind === -1 ? null : this.raw[ind];
  }
  clear() {
    this.raw.length = 0;
  }
  popFront() {
    return this.raw.shift();
  }
}
export {
  SortedArray
};
