// @mtcute/core/utils/type-assertions.js
import { MtTypeAssertionError } from "../types/errors.js";
function isPresent(t) {
  return t !== void 0 && t !== null;
}
function hasValueAtKey(k, v) {
  return function(a) {
    return a[k] === v;
  };
}
function assertTypeIs(context, obj, expected) {
  if (obj._ !== expected) {
    throw new MtTypeAssertionError(context, expected, obj._);
  }
}
function assertTypeIsNot(context, obj, expectedNot) {
  if (obj._ === expectedNot) {
    throw new MtTypeAssertionError(context, `not ${expectedNot}`, obj._);
  }
}
function mtpAssertTypeIs(context, obj, expected) {
  if (obj._ !== expected) {
    throw new MtTypeAssertionError(context, expected, obj._);
  }
}
function assertTrue(context, cond) {
  if (!cond) {
    throw new MtTypeAssertionError(context, "true", "false");
  }
}
function isTlRpcError(obj) {
  return typeof obj === "object" && obj !== null && obj._ === "mt_rpc_error";
}
export {
  assertTrue,
  assertTypeIs,
  assertTypeIsNot,
  hasValueAtKey,
  isPresent,
  isTlRpcError,
  mtpAssertTypeIs
};
