// @mtcute/core/utils/type-assertions.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../types/errors.cjs");
function isPresent(t) {
  return t !== void 0 && t !== null;
}
function hasValueAtKey(k, v) {
  return function(a) {
    return a[k] === v;
  };
}
function assertTypeIs(context, obj, expected) {
  if (obj._ !== expected) {
    throw new errors.MtTypeAssertionError(context, expected, obj._);
  }
}
function assertTypeIsNot(context, obj, expectedNot) {
  if (obj._ === expectedNot) {
    throw new errors.MtTypeAssertionError(context, `not ${expectedNot}`, obj._);
  }
}
function mtpAssertTypeIs(context, obj, expected) {
  if (obj._ !== expected) {
    throw new errors.MtTypeAssertionError(context, expected, obj._);
  }
}
function assertTrue(context, cond) {
  if (!cond) {
    throw new errors.MtTypeAssertionError(context, "true", "false");
  }
}
function isTlRpcError(obj) {
  return typeof obj === "object" && obj !== null && obj._ === "mt_rpc_error";
}
exports.assertTrue = assertTrue;
exports.assertTypeIs = assertTypeIs;
exports.assertTypeIsNot = assertTypeIsNot;
exports.hasValueAtKey = hasValueAtKey;
exports.isPresent = isPresent;
exports.isTlRpcError = isTlRpcError;
exports.mtpAssertTypeIs = mtpAssertTypeIs;
