// @mtcute/core/highlevel/types/bots/command-scope/inner.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const default_ = {
  _: "botCommandScopeDefault"
};
const allPrivate = {
  _: "botCommandScopeUsers"
};
const allGroups = {
  _: "botCommandScopeChats"
};
const allGroupAdmins = {
  _: "botCommandScopeChatAdmins"
};
function peer(peer2) {
  return {
    type: "peer",
    peer: peer2
  };
}
function groupAdmins(peer2) {
  return {
    type: "peer_admins",
    peer: peer2
  };
}
function groupMember(chat, user) {
  return {
    type: "member",
    chat,
    user
  };
}
function cmd(command, description) {
  return {
    _: "botCommand",
    command,
    description
  };
}
exports.allGroupAdmins = allGroupAdmins;
exports.allGroups = allGroups;
exports.allPrivate = allPrivate;
exports.cmd = cmd;
exports.default_ = default_;
exports.groupAdmins = groupAdmins;
exports.groupMember = groupMember;
exports.peer = peer;
