// @mtcute/core/highlevel/types/messages/rich/inner.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const entities = require("../../../utils/entities.cjs");
const RICH_EMPTY = { _: "textEmpty" };
function normalizeRichText(text) {
  if (typeof text === "string") return { _: "textPlain", text };
  if ("_" in text) return text;
  return entities.textWithEntitiesToRichText(text);
}
function normalizeCaption(input) {
  if (input == null) return { _: "pageCaption", text: RICH_EMPTY, credit: RICH_EMPTY };
  if (typeof input === "string" || "_" in input || "entities" in input) {
    return { _: "pageCaption", text: normalizeRichText(input), credit: RICH_EMPTY };
  }
  return {
    _: "pageCaption",
    text: input.text != null ? normalizeRichText(input.text) : RICH_EMPTY,
    credit: "credit" in input && input.credit != null ? normalizeRichText(input.credit) : RICH_EMPTY
  };
}
function footer(text) {
  return { _: "pageBlockFooter", text: normalizeRichText(text) };
}
function paragraph(text) {
  return { _: "pageBlockParagraph", text: normalizeRichText(text) };
}
function thinking(text) {
  return { _: "pageBlockThinking", text: normalizeRichText(text) };
}
function heading(level, text) {
  return { _: `pageBlockHeading${level}`, text: normalizeRichText(text) };
}
function divider() {
  return { _: "pageBlockDivider" };
}
function anchor(name) {
  return { _: "pageBlockAnchor", name };
}
function math(source) {
  return { _: "pageBlockMath", source };
}
function preformatted(text, language = "") {
  return { _: "pageBlockPreformatted", text: normalizeRichText(text), language };
}
function blockquote(content, caption) {
  const normalizedCaption = caption != null ? normalizeRichText(caption) : RICH_EMPTY;
  if (Array.isArray(content)) {
    return { _: "pageBlockBlockquoteBlocks", blocks: content, caption: normalizedCaption };
  }
  return { _: "pageBlockBlockquote", text: normalizeRichText(content), caption: normalizedCaption };
}
function pullquote(text, caption) {
  return {
    _: "pageBlockPullquote",
    text: normalizeRichText(text),
    caption: caption != null ? normalizeRichText(caption) : RICH_EMPTY
  };
}
function map(params) {
  return {
    _: "inputPageBlockMap",
    geo: { _: "inputGeoPoint", lat: params.latitude, long: params.longitude },
    zoom: params.zoom,
    w: params.width,
    h: params.height,
    caption: normalizeCaption(params.caption)
  };
}
function listItem(content, params = {}) {
  if (Array.isArray(content)) {
    return {
      _: "pageListItemBlocks",
      blocks: content,
      checkbox: params.checkbox,
      checked: params.checked
    };
  }
  return {
    _: "pageListItemText",
    text: normalizeRichText(content),
    checkbox: params.checkbox,
    checked: params.checked
  };
}
function list(items) {
  return { _: "pageBlockList", items };
}
function orderedListItem(content, params = {}) {
  if (Array.isArray(content)) {
    return {
      _: "pageListOrderedItemBlocks",
      blocks: content,
      checkbox: params.checkbox,
      checked: params.checked,
      num: params.num,
      value: params.value,
      type: params.type
    };
  }
  return {
    _: "pageListOrderedItemText",
    text: normalizeRichText(content),
    checkbox: params.checkbox,
    checked: params.checked,
    num: params.num,
    value: params.value,
    type: params.type
  };
}
function orderedList(items, params = {}) {
  return {
    _: "pageBlockOrderedList",
    items,
    reversed: params.reversed,
    start: params.start,
    type: params.type
  };
}
function tableCell(text, params = {}) {
  return {
    _: "pageTableCell",
    text: text != null ? normalizeRichText(text) : void 0,
    header: params.header,
    alignCenter: params.alignCenter,
    alignRight: params.alignRight,
    valignMiddle: params.valignMiddle,
    valignBottom: params.valignBottom,
    colspan: params.colspan,
    rowspan: params.rowspan
  };
}
function tableRow(cells) {
  return { _: "pageTableRow", cells };
}
function table(rows, params = {}) {
  return {
    _: "pageBlockTable",
    title: params.title != null ? normalizeRichText(params.title) : RICH_EMPTY,
    rows,
    bordered: params.bordered,
    striped: params.striped
  };
}
function collage(items, caption) {
  return { _: "pageBlockCollage", items, caption: normalizeCaption(caption) };
}
function slideshow(items, caption) {
  return { _: "pageBlockSlideshow", items, caption: normalizeCaption(caption) };
}
function details(title, blocks, params = {}) {
  return { _: "pageBlockDetails", blocks, title: normalizeRichText(title), open: params.open };
}
function photo(file, params = {}) {
  return {
    // the default `auto` mode would upload it as a document, but we already know it should be a photo
    file: typeof file === "string" ? { type: "photo", file } : file,
    block: {
      _: "pageBlockPhoto",
      photoId: Long.ZERO,
      caption: normalizeCaption(params.caption),
      url: params.url,
      webpageId: params.webpageId,
      spoiler: params.spoiler
    }
  };
}
function video(file, params = {}) {
  return {
    file,
    block: {
      _: "pageBlockVideo",
      videoId: Long.ZERO,
      caption: normalizeCaption(params.caption),
      autoplay: params.autoplay,
      loop: params.loop,
      spoiler: params.spoiler
    }
  };
}
function audio(file, params = {}) {
  return {
    file,
    block: {
      _: "pageBlockAudio",
      audioId: Long.ZERO,
      caption: normalizeCaption(params.caption)
    }
  };
}
exports.anchor = anchor;
exports.audio = audio;
exports.blockquote = blockquote;
exports.collage = collage;
exports.details = details;
exports.divider = divider;
exports.footer = footer;
exports.heading = heading;
exports.list = list;
exports.listItem = listItem;
exports.map = map;
exports.math = math;
exports.orderedList = orderedList;
exports.orderedListItem = orderedListItem;
exports.paragraph = paragraph;
exports.photo = photo;
exports.preformatted = preformatted;
exports.pullquote = pullquote;
exports.slideshow = slideshow;
exports.table = table;
exports.tableCell = tableCell;
exports.tableRow = tableRow;
exports.thinking = thinking;
exports.video = video;
