// @mtcute/core/highlevel/types/messages/rich/rich-message.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../../utils/long-utils.cjs");
const inspectable = require("../../../utils/inspectable.cjs");
const memoize = require("../../../utils/memoize.cjs");
const documentUtils = require("../../media/document-utils.cjs");
const photo = require("../../media/photo.cjs");
class RichMessage {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether this message is RTL */
  get isRtl() {
    return this.raw.rtl;
  }
  /** Whether this message is partial */
  get isPart() {
    return this.raw.part;
  }
  /** Photos in this message */
  get photos() {
    const map = new longUtils.LongMap();
    for (const photo$1 of this.raw.photos) {
      if (photo$1._ === "photoEmpty") continue;
      map.set(photo$1.id, new photo.Photo(photo$1));
    }
    return map;
  }
  /** Documents (usually videos and audios) in this message */
  get documents() {
    const map = new longUtils.LongMap();
    for (const photo2 of this.raw.documents) {
      if (photo2._ === "documentEmpty") continue;
      map.set(photo2.id, documentUtils.parseDocument(photo2));
    }
    return map;
  }
  /** Blocks in this message */
  get blocks() {
    return this.raw.blocks;
  }
}
const RichMessage$1 = /* @__PURE__ */ inspectable.makeInspectable(RichMessage);
const RichMessage$2 = /* @__PURE__ */ memoize.memoizeGetters(RichMessage$1, ["documents", "blocks"]);
exports.RichMessage = RichMessage$2;
