// @mtcute/core/highlevel/types/messages/rich/rich-message.js
import { LongMap } from "../../../../utils/long-utils.js";
import { makeInspectable } from "../../../utils/inspectable.js";
import { memoizeGetters } from "../../../utils/memoize.js";
import { parseDocument } from "../../media/document-utils.js";
import { Photo as Photo$2 } from "../../media/photo.js";
class RichMessage {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether this message is RTL */
  get isRtl() {
    return this.raw.rtl;
  }
  /** Whether this message is partial */
  get isPart() {
    return this.raw.part;
  }
  /** Photos in this message */
  get photos() {
    const map = new LongMap();
    for (const photo of this.raw.photos) {
      if (photo._ === "photoEmpty") continue;
      map.set(photo.id, new Photo$2(photo));
    }
    return map;
  }
  /** Documents (usually videos and audios) in this message */
  get documents() {
    const map = new LongMap();
    for (const photo of this.raw.documents) {
      if (photo._ === "documentEmpty") continue;
      map.set(photo.id, parseDocument(photo));
    }
    return map;
  }
  /** Blocks in this message */
  get blocks() {
    return this.raw.blocks;
  }
}
const RichMessage$1 = /* @__PURE__ */ makeInspectable(RichMessage);
const RichMessage$2 = /* @__PURE__ */ memoizeGetters(RichMessage$1, ["documents", "blocks"]);
export {
  RichMessage$2 as RichMessage
};
