// @fuman/node/streams/convert-fuman.js
import { Readable, Writable } from "node:stream";
import { Bytes } from "@fuman/io";
import { Deque, Deferred } from "@fuman/utils";
function nodeReadableToFuman(stream) {
  const waiters = new Deque();
  stream.on("end", () => {
    while (waiters.length > 0) {
      waiters.popFront().resolve();
    }
  });
  stream.on("error", (err) => {
    while (waiters.length > 0) {
      waiters.popFront().reject(err);
    }
  });
  stream.on("readable", () => {
    waiters.popFront()?.resolve();
  });
  const readBuffer = Bytes.alloc();
  return {
    async read(into) {
      if (stream.readableEnded) return 0;
      if (readBuffer.available > 0) {
        const size = Math.min(readBuffer.available, into.length);
        into.set(readBuffer.readSync(size));
        return size;
      }
      let buf = null;
      while (buf == null) {
        buf = stream.read(into.length);
        if (buf === null) {
          if (stream.readableEnded) return 0;
          const deferred = new Deferred();
          waiters.pushBack(deferred);
          await deferred.promise;
        }
      }
      if (into.length < buf.length) {
        const toReadBuffer = buf.length - into.length;
        readBuffer.writeSync(toReadBuffer).set(buf.subarray(toReadBuffer));
        buf = buf.subarray(0, into.length);
      }
      into.set(buf);
      return buf.length;
    },
    close() {
      stream.destroy();
    }
  };
}
function fumanReadableToNode(stream) {
  return new Readable({
    read(size) {
      void (async () => {
        try {
          const buf = new Uint8Array(size);
          const nread = await stream.read(buf);
          if (nread === 0) {
            this.push(null);
          } else {
            this.push(Buffer.from(buf.subarray(0, nread)));
          }
        } catch (err) {
          this.destroy(err);
        }
      })();
    }
  });
}
function nodeWritableToFuman(writable) {
  let ended = false;
  const waiters = new Deque();
  writable.on("finish", () => {
    ended = true;
    while (waiters.length > 0) {
      waiters.popFront().resolve();
    }
  });
  let full = false;
  writable.on("drain", () => {
    full = false;
    waiters.popFront()?.resolve();
  });
  writable.on("error", (err) => {
    while (waiters.length > 0) {
      waiters.popFront().reject(err);
    }
  });
  return {
    async write(bytes) {
      if (ended) return;
      if (full) {
        const deferred = new Deferred();
        waiters.pushBack(deferred);
        await deferred.promise;
      }
      await new Promise((resolve, reject) => {
        const written = writable.write(bytes, (err) => {
          if (err) reject(err);
          else resolve();
        });
        if (!written) {
          full = true;
          const deferred = new Deferred();
          waiters.pushBack(deferred);
          resolve(deferred.promise);
        }
      });
    },
    close() {
      ended = true;
      writable.end();
    }
  };
}
function fumanWritableToNode(writable) {
  return new Writable({
    write(chunk, encoding, callback) {
      void (async () => {
        try {
          await writable.write(chunk);
          callback();
        } catch (err) {
          callback(err);
        }
      })();
    },
    final(callback) {
      if (!("close" in writable)) {
        callback();
        return;
      }
      void (async () => {
        try {
          writable.close();
          callback();
        } catch (err) {
          callback(err);
        }
      })();
    }
  });
}
export {
  fumanReadableToNode,
  fumanWritableToNode,
  nodeReadableToFuman,
  nodeWritableToFuman
};
