// @fuman/node/streams/convert-web.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_stream = require("node:stream");
const version = require("../version.cjs");
function nodeReadableToWeb(stream) {
  if (typeof node_stream.Readable.toWeb === "function") {
    return node_stream.Readable.toWeb(stream);
  }
  stream.pause();
  return new ReadableStream({
    start(c) {
      stream.on("data", (chunk) => {
        c.enqueue(chunk);
      });
      stream.on("end", () => {
        c.close();
      });
      stream.on("error", (err) => {
        c.error(err);
      });
    },
    pull() {
      stream.resume();
    }
  });
}
function nodeWritableToWeb(writable) {
  if (typeof node_stream.Writable.toWeb === "function") {
    return node_stream.Writable.toWeb(writable);
  }
  return new WritableStream({
    write(chunk) {
      return new Promise((resolve, reject) => {
        writable.write(chunk, (err) => {
          if (err) reject(err);
          else resolve();
        });
      });
    },
    close() {
      return new Promise((resolve, reject) => {
        writable.end((err) => {
          if (err != null) {
            reject(err);
          } else {
            resolve();
            writable.destroy();
          }
        });
      });
    }
  });
}
function webReadableToNode(stream) {
  if (typeof node_stream.Readable.fromWeb === "function" && version.isNodeVersionAfter(18, 13, 0)) {
    return node_stream.Readable.fromWeb(stream);
  }
  const reader = stream.getReader();
  let ended = false;
  const readable = new node_stream.Readable({
    // eslint-disable-next-line ts/no-misused-promises
    async read() {
      try {
        const { done, value } = await reader.read();
        if (done) {
          this.push(null);
        } else {
          this.push(Buffer.from(value.buffer, value.byteOffset, value.byteLength));
        }
      } catch (err) {
        this.destroy(err);
      }
    },
    destroy(error, cb) {
      if (!ended) {
        void reader.cancel(error).catch(() => {
        }).then(() => {
          cb(error);
        });
        return;
      }
      cb(error);
    }
  });
  reader.closed.then(() => {
    ended = true;
  }).catch((err) => {
    readable.destroy(err);
  });
  return readable;
}
function webWritableToNode(writable) {
  if (typeof node_stream.Writable.fromWeb === "function") {
    return node_stream.Writable.fromWeb(writable);
  }
  const writer = writable.getWriter();
  return new node_stream.Writable({
    write(chunk, encoding, callback) {
      writer.write(chunk).then(() => callback()).catch(callback);
    },
    final(callback) {
      writer.close().then(() => callback()).catch(callback);
    }
  });
}
exports.nodeReadableToWeb = nodeReadableToWeb;
exports.nodeWritableToWeb = nodeWritableToWeb;
exports.webReadableToNode = webReadableToNode;
exports.webWritableToNode = webWritableToNode;
