// @mtcute/core/highlevel/worker/errors.js
import { MtUnsupportedError, MtSecurityError, MtArgumentError, MtTimeoutError, MtTypeAssertionError, MtcuteError } from "../../types/errors.js";
import { MtEmptyError, MtInvalidPeerTypeError, MtPeerNotFoundError, MtMessageNotFoundError } from "../types/errors.js";
import { RpcError } from "../../tl/inner-tl.js";
function serializeError(error) {
  if (!(error instanceof Error)) return { name: "Error", message: String(error) };
  const res = {
    name: "Error",
    message: error.message,
    stack: error.stack
  };
  const ctor = error.constructor;
  if (ctor === MtTypeAssertionError) {
    const _error = error;
    res.name = "MtTypeAssertionError";
    res.custom = {
      context: _error.context,
      expected: _error.expected,
      actual: _error.actual
    };
  } else if (ctor === MtTimeoutError) {
    const _error = error;
    res.name = "MtTimeoutError";
    res.custom = {
      timeout: _error.timeout
    };
  } else if (ctor === MtMessageNotFoundError) {
    const _error = error;
    res.name = "MtMessageNotFoundError";
    res.custom = {
      peerId: _error.peerId,
      messageId: _error.messageId,
      context: _error.context
    };
  } else if (ctor === RpcError) {
    res.name = "RpcError";
    res.custom = { ...error };
  } else if (ctor === MtArgumentError) {
    res.name = "MtArgumentError";
  } else if (ctor === MtSecurityError) {
    res.name = "MtSecurityError";
  } else if (ctor === MtUnsupportedError) {
    res.name = "MtUnsupportedError";
  } else if (ctor === MtPeerNotFoundError) {
    res.name = "MtPeerNotFoundError";
  } else if (ctor === MtInvalidPeerTypeError) {
    res.name = "MtInvalidPeerTypeError";
  } else if (ctor === MtEmptyError) {
    res.name = "MtEmptyError";
  } else if (ctor instanceof MtcuteError) {
    res.name = "MtcuteError";
  }
  return res;
}
function deserializeError(error) {
  let err2;
  switch (error.name) {
    case "MtTypeAssertionError": {
      const custom = error.custom;
      err2 = new MtTypeAssertionError(custom.context, custom.expected, custom.actual);
      break;
    }
    case "MtTimeoutError": {
      const custom = error.custom;
      err2 = new MtTimeoutError(custom.timeout);
      break;
    }
    case "MtMessageNotFoundError": {
      const custom = error.custom;
      err2 = new MtMessageNotFoundError(custom.peerId, custom.messageId, custom.context);
      break;
    }
    case "RpcError": {
      const custom = error.custom;
      err2 = new RpcError(custom.code, custom.text);
      err2.message = error.message;
      for (const key in custom) {
        if (key === "code" || key === "text") continue;
        err2[key] = custom[key];
      }
      break;
    }
    case "MtArgumentError":
      err2 = new MtArgumentError();
      break;
    case "MtSecurityError":
      err2 = new MtSecurityError();
      break;
    case "MtUnsupportedError":
      err2 = new MtUnsupportedError();
      break;
    case "MtPeerNotFoundError":
      err2 = new MtPeerNotFoundError();
      break;
    case "MtInvalidPeerTypeError":
      err2 = new MtInvalidPeerTypeError("", "");
      err2.message = error.message;
      break;
    case "MtEmptyError":
      err2 = new MtEmptyError();
      break;
    case "MtcuteError":
      err2 = new MtcuteError();
      break;
    default:
      err2 = new Error(error.message);
  }
  err2.stack = error.stack;
  return err2;
}
export {
  deserializeError,
  serializeError
};
