// @mtcute/core/highlevel/worker/errors.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors$1 = require("../../types/errors.cjs");
const errors = require("../types/errors.cjs");
const innerTl = require("../../tl/inner-tl.cjs");
function serializeError(error) {
  if (!(error instanceof Error)) return { name: "Error", message: String(error) };
  const res = {
    name: "Error",
    message: error.message,
    stack: error.stack
  };
  const ctor = error.constructor;
  if (ctor === errors$1.MtTypeAssertionError) {
    const _error = error;
    res.name = "MtTypeAssertionError";
    res.custom = {
      context: _error.context,
      expected: _error.expected,
      actual: _error.actual
    };
  } else if (ctor === errors$1.MtTimeoutError) {
    const _error = error;
    res.name = "MtTimeoutError";
    res.custom = {
      timeout: _error.timeout
    };
  } else if (ctor === errors.MtMessageNotFoundError) {
    const _error = error;
    res.name = "MtMessageNotFoundError";
    res.custom = {
      peerId: _error.peerId,
      messageId: _error.messageId,
      context: _error.context
    };
  } else if (ctor === innerTl.RpcError) {
    res.name = "RpcError";
    res.custom = { ...error };
  } else if (ctor === errors$1.MtArgumentError) {
    res.name = "MtArgumentError";
  } else if (ctor === errors$1.MtSecurityError) {
    res.name = "MtSecurityError";
  } else if (ctor === errors$1.MtUnsupportedError) {
    res.name = "MtUnsupportedError";
  } else if (ctor === errors.MtPeerNotFoundError) {
    res.name = "MtPeerNotFoundError";
  } else if (ctor === errors.MtInvalidPeerTypeError) {
    res.name = "MtInvalidPeerTypeError";
  } else if (ctor === errors.MtEmptyError) {
    res.name = "MtEmptyError";
  } else if (ctor instanceof errors$1.MtcuteError) {
    res.name = "MtcuteError";
  }
  return res;
}
function deserializeError(error) {
  let err2;
  switch (error.name) {
    case "MtTypeAssertionError": {
      const custom = error.custom;
      err2 = new errors$1.MtTypeAssertionError(custom.context, custom.expected, custom.actual);
      break;
    }
    case "MtTimeoutError": {
      const custom = error.custom;
      err2 = new errors$1.MtTimeoutError(custom.timeout);
      break;
    }
    case "MtMessageNotFoundError": {
      const custom = error.custom;
      err2 = new errors.MtMessageNotFoundError(custom.peerId, custom.messageId, custom.context);
      break;
    }
    case "RpcError": {
      const custom = error.custom;
      err2 = new innerTl.RpcError(custom.code, custom.text);
      err2.message = error.message;
      for (const key in custom) {
        if (key === "code" || key === "text") continue;
        err2[key] = custom[key];
      }
      break;
    }
    case "MtArgumentError":
      err2 = new errors$1.MtArgumentError();
      break;
    case "MtSecurityError":
      err2 = new errors$1.MtSecurityError();
      break;
    case "MtUnsupportedError":
      err2 = new errors$1.MtUnsupportedError();
      break;
    case "MtPeerNotFoundError":
      err2 = new errors.MtPeerNotFoundError();
      break;
    case "MtInvalidPeerTypeError":
      err2 = new errors.MtInvalidPeerTypeError("", "");
      err2.message = error.message;
      break;
    case "MtEmptyError":
      err2 = new errors.MtEmptyError();
      break;
    case "MtcuteError":
      err2 = new errors$1.MtcuteError();
      break;
    default:
      err2 = new Error(error.message);
  }
  err2.stack = error.stack;
  return err2;
}
exports.deserializeError = deserializeError;
exports.serializeError = serializeError;
