// @mtcute/core/highlevel/worker/invoker.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("./errors.cjs");
const protocol = require("./protocol.cjs");
class WorkerInvoker {
  constructor(send, workerId, connectionId) {
    this.send = send;
    this.workerId = workerId;
    this.connectionId = connectionId;
  }
  _nextId = 0;
  _released = false;
  _releasedError = new Error("Worker connection closed");
  _pending = /* @__PURE__ */ new Map();
  _close(error, shouldRelease) {
    if (this._released) return;
    this._released = true;
    this._releasedError = error;
    if (shouldRelease) {
      this.send({
        _mtcuteWorkerId: this.workerId,
        connectionId: this.connectionId,
        type: "release"
      });
    }
    for (const pending of this._pending.values()) {
      pending.reject(error);
    }
    this._pending.clear();
  }
  connect() {
    this.send({
      _mtcuteWorkerId: this.workerId,
      connectionId: this.connectionId,
      type: "connect"
    });
  }
  heartbeat() {
    if (this._released) return;
    this.send({
      _mtcuteWorkerId: this.workerId,
      connectionId: this.connectionId,
      type: "heartbeat"
    });
  }
  release() {
    this._close(new Error("Worker connection closed"), true);
  }
  expire() {
    this._close(new Error("Worker connection expired"), false);
  }
  _invoke(target, method, args, isVoid, abortSignal) {
    if (this._released) {
      if (isVoid) throw this._releasedError;
      return Promise.reject(this._releasedError);
    }
    const id = this._nextId++;
    this.send({
      _mtcuteWorkerId: this.workerId,
      connectionId: this.connectionId,
      type: "invoke",
      id,
      target,
      method,
      args: protocol.serializeResult(args),
      void: isVoid,
      withAbort: Boolean(abortSignal)
    });
    if (abortSignal) {
      const onAbort = () => {
        if (this._released) return;
        this.send({
          _mtcuteWorkerId: this.workerId,
          connectionId: this.connectionId,
          type: "abort",
          id
        });
      };
      abortSignal.addEventListener("abort", onAbort, { once: true });
      if (!isVoid) {
        const promise = new utils.Deferred();
        this._pending.set(id, promise);
        return promise.promise.finally(() => {
          abortSignal.removeEventListener("abort", onAbort);
        });
      }
      return;
    }
    if (!isVoid) {
      const promise = new utils.Deferred();
      this._pending.set(id, promise);
      return promise.promise;
    }
  }
  invoke(target, method, args) {
    return this._invoke(target, method, args, false);
  }
  invokeVoid(target, method, args) {
    void this._invoke(target, method, args, true);
  }
  invokeWithAbort(target, method, args, abortSignal) {
    if (abortSignal.aborted) return Promise.reject(abortSignal.reason);
    return this._invoke(target, method, args, false, abortSignal);
  }
  handleResult(msg) {
    const promise = this._pending.get(msg.id);
    if (!promise) return;
    this._pending.delete(msg.id);
    if (msg.error) {
      promise.reject(errors.deserializeError(msg.error));
    } else {
      promise.resolve(protocol.deserializeResult(msg.result));
    }
  }
  makeBinder(target) {
    return (method, isVoid = false) => {
      let fn;
      if (isVoid) {
        fn = (...args) => this.invokeVoid(target, method, args);
      } else {
        fn = (...args) => this.invoke(target, method, args);
      }
      return fn;
    };
  }
}
exports.WorkerInvoker = WorkerInvoker;
