// @mtcute/core/highlevel/worker/port.js
import { Emitter } from "@fuman/utils";
import { LogManager } from "../../utils/logger.js";
import { PeersIndex } from "../types/peers/peers-index.js";
import { AppConfigManagerProxy } from "./app-config.js";
import { deserializeError } from "./errors.js";
import { WorkerInvoker } from "./invoker.js";
import { DEFAULT_WORKER_ID, deserializeResult } from "./protocol.js";
import { TelegramStorageProxy } from "./storage.js";
import { WorkerTimersManager } from "./timers.js";
const WORKER_HEARTBEAT_INTERVAL = 1e4;
let _nextConnectionId = 0;
function makeConnectionId() {
  return `${Date.now().toString(36)}-${(_nextConnectionId++).toString(36)}-${Math.random().toString(36).slice(2)}`;
}
class TelegramWorkerPort {
  constructor(options) {
    this.options = options;
    this.log = new LogManager("worker", options.platform);
    this.platform = options.platform;
    this.workerId = options.workerId ?? DEFAULT_WORKER_ID;
    this.connectionId = makeConnectionId();
    this._connection = this.connectToWorker(this.options.worker, this._onMessage);
    this._invoker = new WorkerInvoker(this._connection[0], this.workerId, this.connectionId);
    this._invoker.connect();
    this._heartbeatTimer = setInterval(() => {
      this._invoker.heartbeat();
    }, WORKER_HEARTBEAT_INTERVAL);
    this._cancelBeforeExit = this.platform.beforeExit(() => {
      void this.destroy();
    });
    this.storage = new TelegramStorageProxy(this._invoker);
    this.appConfig = new AppConfigManagerProxy(this._invoker);
    const bind = this._invoker.makeBinder("client");
    this.prepare = bind("prepare");
    this._connect = bind("connect");
    this.disconnect = bind("disconnect");
    this.unsafeForceDestroy = bind("destroy");
    this.notifyLoggedIn = bind("notifyLoggedIn");
    this.notifyLoggedOut = bind("notifyLoggedOut");
    this.notifyChannelOpened = bind("notifyChannelOpened");
    this.notifyChannelClosed = bind("notifyChannelClosed");
    this.importSession = bind("importSession");
    this.exportSession = bind("exportSession");
    this.handleClientUpdate = bind("handleClientUpdate", true);
    this.getApiCredentials = bind("getApiCredentials");
    this.getPoolSize = bind("getPoolSize");
    this.getPrimaryDcId = bind("getPrimaryDcId");
    this.changePrimaryDc = bind("changePrimaryDc");
    this.computeSrpParams = bind("computeSrpParams");
    this.computeNewPasswordHash = bind("computeNewPasswordHash");
    this.startUpdatesLoop = bind("startUpdatesLoop");
    this.stopUpdatesLoop = bind("stopUpdatesLoop");
    this.getMtprotoMessageId = bind("getMtprotoMessageId");
    this.recreateDc = bind("recreateDc");
    this.timers = new WorkerTimersManager(this._invoker);
  }
  log;
  platform;
  _connection;
  _invoker;
  storage;
  appConfig;
  timers;
  // bound methods
  prepare;
  _connect;
  disconnect;
  unsafeForceDestroy;
  notifyLoggedIn;
  notifyLoggedOut;
  notifyChannelOpened;
  notifyChannelClosed;
  importSession;
  exportSession;
  handleClientUpdate;
  getApiCredentials;
  getPoolSize;
  getPrimaryDcId;
  changePrimaryDc;
  computeSrpParams;
  computeNewPasswordHash;
  startUpdatesLoop;
  stopUpdatesLoop;
  getMtprotoMessageId;
  recreateDc;
  _abortController = new AbortController();
  stopSignal = this._abortController.signal;
  workerId;
  connectionId;
  _cancelBeforeExit;
  _heartbeatTimer;
  call(message, params) {
    if (params?.abortSignal) {
      const { abortSignal, ...rest } = params;
      return this._invoker.invokeWithAbort("client", "call", [message, rest], abortSignal);
    }
    return this._invoker.invoke("client", "call", [message, params]);
  }
  onServerUpdate = new Emitter();
  onRawUpdate = new Emitter();
  onConnectionState = new Emitter();
  onError = new Emitter();
  _onMessage = (message) => {
    if (this._closed) return;
    if (message._mtcuteWorkerId !== this.workerId) return;
    if ("connectionId" in message && message.connectionId !== this.connectionId) return;
    switch (message.type) {
      case "log":
        this.log.handler(message.color, message.level, message.tag, message.fmt, message.args);
        break;
      case "server_update":
        this.onServerUpdate.emit(deserializeResult(message.update));
        break;
      case "self_sync":
        this.storage.self.setCached(deserializeResult(message.self));
        break;
      case "connection_expired": {
        const error = new Error("Worker connection expired");
        this.onError.emit(error);
        this._close(false);
        break;
      }
      case "conn_state":
        this.onConnectionState.emit(message.state);
        break;
      case "update": {
        const peers = new PeersIndex(deserializeResult(message.users), deserializeResult(message.chats));
        peers.hasMin = message.hasMin;
        this.onRawUpdate.emit({ update: deserializeResult(message.update), peers });
        break;
      }
      case "result":
        this._invoker.handleResult(message);
        break;
      case "error":
        this.onError.emit(deserializeError(message.error));
        break;
      case "stop":
        this._close(false);
        break;
    }
  };
  _closed = false;
  _close(sendRelease) {
    if (this._closed) return;
    this._closed = true;
    if (this._heartbeatTimer !== void 0) {
      clearInterval(this._heartbeatTimer);
    }
    this._cancelBeforeExit?.();
    if (sendRelease) {
      this._invoker.release();
    } else {
      this._invoker.expire();
    }
    this._abortController.abort();
    queueMicrotask(() => this._connection[1]());
  }
  async destroy(terminate = false) {
    this._close(true);
    if (terminate && "terminate" in this.options.worker) {
      await Promise.resolve(this.options.worker.terminate());
    }
  }
  invokeCustom(method, ...args) {
    return this._invoker.invoke("custom", method, args);
  }
  async connect() {
    await this._connect();
    await this.storage.self.fetch();
  }
}
export {
  TelegramWorkerPort
};
