// @mtcute/core/highlevel/worker/protocol.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const DEFAULT_WORKER_ID = "default";
function serializeLong(value) {
  return {
    __type: "long",
    low: value.low,
    high: value.high,
    unsigned: value.unsigned
  };
}
function isSerializedLong(value) {
  return Boolean(value && typeof value === "object" && value.__type === "long");
}
function serializeResult(result) {
  if (Long.isLong(result)) {
    return serializeLong(result);
  }
  if (ArrayBuffer.isView(result)) return result;
  if (Array.isArray(result)) {
    return result.map(serializeResult);
  }
  if (result instanceof Map) {
    for (const [key, value] of result.entries()) {
      result.set(key, serializeResult(value));
    }
    return result;
  }
  if (result && typeof result === "object") {
    const newResult = {};
    for (const [key, value] of Object.entries(result)) {
      if (Long.isLong(value)) {
        newResult[key] = serializeLong(value);
      } else if (typeof value === "object") {
        newResult[key] = serializeResult(value);
      } else {
        newResult[key] = value;
      }
    }
    return newResult;
  }
  return result;
}
function deserializeResult(result) {
  if (isSerializedLong(result)) {
    return Long.fromValue(result);
  }
  if (ArrayBuffer.isView(result)) return result;
  if (Array.isArray(result)) {
    return result.map(deserializeResult);
  }
  if (result instanceof Map) {
    for (const [key, value] of result.entries()) {
      result.set(key, deserializeResult(value));
    }
    return result;
  }
  if (result && typeof result === "object") {
    for (const [key, value] of Object.entries(result)) {
      if (isSerializedLong(value)) {
        result[key] = Long.fromValue(value);
      } else if (typeof value === "object") {
        result[key] = deserializeResult(value);
      }
    }
  }
  return result;
}
exports.DEFAULT_WORKER_ID = DEFAULT_WORKER_ID;
exports.deserializeResult = deserializeResult;
exports.serializeResult = serializeResult;
