// @mtcute/core/highlevel/worker/storage.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../types/errors.cjs");
class CurrentUserServiceProxy {
  _store;
  _storeFrom;
  _fetch;
  _update;
  constructor(invoker) {
    const bind = invoker.makeBinder("storage-self");
    this._store = bind("store");
    this._storeFrom = bind("storeFrom");
    this._fetch = bind("fetch");
    this._update = bind("update");
  }
  _cached;
  async store(info) {
    await this._store(info);
    this._cached = info;
  }
  async storeFrom(user) {
    this._cached = await this._storeFrom(user);
    return this._cached;
  }
  async fetch() {
    if (this._cached !== void 0) return this._cached;
    this._cached = await this._fetch();
    return this._cached;
  }
  getCached(safe) {
    if (this._cached === void 0) {
      if (safe) return null;
      throw new errors.MtArgumentError("User info is not cached yet");
    }
    return this._cached;
  }
  async update(params) {
    await this._update(params);
    this._cached = await this._fetch();
  }
  setCached(info) {
    this._cached = info;
  }
}
class PeersServiceProxy {
  constructor(_invoker) {
    this._invoker = _invoker;
    const bind = this._invoker.makeBinder("storage-peers");
    this.updatePeersFrom = bind("updatePeersFrom");
    this.store = bind("store");
    this.getById = bind("getById");
    this.getByPhone = bind("getByPhone");
    this.getByUsername = bind("getByUsername");
    this.getCompleteById = bind("getCompleteById");
    this.getMinAccessHash = bind("getMinAccessHash");
    this.close = bind("close");
  }
  updatePeersFrom;
  store;
  getById;
  getByPhone;
  getByUsername;
  getCompleteById;
  getMinAccessHash;
  close;
}
class TelegramStorageProxy {
  constructor(_invoker) {
    this._invoker = _invoker;
    const bind = this._invoker.makeBinder("storage");
    this.self = new CurrentUserServiceProxy(this._invoker);
    this.peers = new PeersServiceProxy(this._invoker);
    this.clear = bind("clear");
    this.close = bind("close");
  }
  self;
  peers;
  clear;
  close;
  // todo - remove once we move these to updates manager
  updates = null;
  refMsgs = null;
}
exports.TelegramStorageProxy = TelegramStorageProxy;
