// @mtcute/core/highlevel/worker/worker.js
import { serializeError } from "./errors.js";
import { DEFAULT_WORKER_ID, deserializeResult, serializeResult } from "./protocol.js";
const WORKER_HEARTBEAT_TIMEOUT = 6e4;
const WORKER_HEARTBEAT_SWEEP_INTERVAL = 1e4;
class TelegramWorker {
  constructor(params) {
    this.params = params;
    this.workerId = params.workerId ?? DEFAULT_WORKER_ID;
    this.client = params.client;
  }
  client;
  broadcast = () => {
  };
  _cleanup = [];
  _heartbeatSweep;
  _mounted = false;
  activeConnections = /* @__PURE__ */ new Set();
  connectionLastSeen = /* @__PURE__ */ new Map();
  pendingAborts = /* @__PURE__ */ new Map();
  workerId;
  _disconnecting;
  _destroying;
  mount() {
    if (this._mounted) return this;
    const [broadcast, cleanup] = this.registerWorker((message, respond) => {
      if (message._mtcuteWorkerId !== this.workerId) return;
      switch (message.type) {
        case "connect":
          this.touchConnection(message.connectionId, true);
          break;
        case "heartbeat":
          if (!this.activeConnections.has(message.connectionId)) {
            this.respondConnectionExpired(respond, message.connectionId);
            break;
          }
          this.touchConnection(message.connectionId);
          break;
        case "release":
          this.onRelease(message.connectionId);
          break;
        case "invoke":
          if (!this.activeConnections.has(message.connectionId)) {
            this.respondConnectionExpired(respond, message.connectionId);
            break;
          }
          this.touchConnection(message.connectionId);
          this.onInvoke(message, respond);
          break;
        case "abort": {
          if (!this.activeConnections.has(message.connectionId)) {
            this.respondConnectionExpired(respond, message.connectionId);
            break;
          }
          this.touchConnection(message.connectionId);
          const pending = this.pendingAborts.get(message.connectionId);
          const abort = pending?.get(message.id);
          if (!abort) break;
          abort.abort();
          pending.delete(message.id);
          if (!pending.size) {
            this.pendingAborts.delete(message.connectionId);
          }
          break;
        }
      }
    });
    this.broadcast = broadcast;
    this._cleanup.push(cleanup);
    this._heartbeatSweep = setInterval(this.sweepExpiredConnections, WORKER_HEARTBEAT_SWEEP_INTERVAL);
    this._mounted = true;
    const client = this.client;
    const prevLogHandler = client.log.mgr.handler;
    const onLog = (color, level, tag, fmt, args) => this.broadcast({
      _mtcuteWorkerId: this.workerId,
      type: "log",
      color,
      level,
      tag,
      fmt,
      args
    });
    const onError = (err) => this.broadcast({
      _mtcuteWorkerId: this.workerId,
      type: "error",
      error: serializeError(err)
    });
    const onConnectionState = (state) => this.broadcast({
      _mtcuteWorkerId: this.workerId,
      type: "conn_state",
      state
    });
    const onStop = () => this.broadcast({
      _mtcuteWorkerId: this.workerId,
      type: "stop"
    });
    client.log.mgr.handler = onLog;
    client.onError.add(onError);
    client.onConnectionState.add(onConnectionState);
    client.stopSignal.addEventListener("abort", onStop);
    this._cleanup.push(
      () => {
        client.log.mgr.handler = prevLogHandler;
      },
      () => client.onError.remove(onError),
      () => client.onConnectionState.remove(onConnectionState),
      () => client.stopSignal.removeEventListener("abort", onStop)
    );
    if (client.updates) {
      const onRawUpdate = ({ update, peers }) => this.broadcast({
        _mtcuteWorkerId: this.workerId,
        type: "update",
        update: serializeResult(update),
        users: serializeResult(peers.users),
        chats: serializeResult(peers.chats),
        hasMin: peers.hasMin
      });
      client.onRawUpdate.add(onRawUpdate);
      this._cleanup.push(() => client.onRawUpdate.remove(onRawUpdate));
    } else {
      const onServerUpdate = (update) => this.broadcast({
        _mtcuteWorkerId: this.workerId,
        type: "server_update",
        update: serializeResult(update)
      });
      client.onServerUpdate.add(onServerUpdate);
      this._cleanup.push(() => client.onServerUpdate.remove(onServerUpdate));
    }
    return this;
  }
  getConnectionAborts(connectionId) {
    const existing = this.pendingAborts.get(connectionId);
    if (existing) return existing;
    const created = /* @__PURE__ */ new Map();
    this.pendingAborts.set(connectionId, created);
    return created;
  }
  deletePendingAbort(connectionId, id) {
    const pending = this.pendingAborts.get(connectionId);
    if (!pending) return;
    pending.delete(id);
    if (!pending.size) {
      this.pendingAborts.delete(connectionId);
    }
  }
  touchConnection(connectionId, activate = false) {
    if (activate) {
      this.activeConnections.add(connectionId);
    } else if (!this.activeConnections.has(connectionId)) {
      return;
    }
    this.connectionLastSeen.set(connectionId, Date.now());
  }
  respondConnectionExpired(respond, connectionId) {
    respond({
      _mtcuteWorkerId: this.workerId,
      connectionId,
      type: "connection_expired"
    });
  }
  sweepExpiredConnections = () => {
    const now = Date.now();
    for (const [connectionId, lastSeen] of this.connectionLastSeen) {
      if (now - lastSeen < WORKER_HEARTBEAT_TIMEOUT) continue;
      this.respondConnectionExpired(this.broadcast, connectionId);
      this.onRelease(connectionId);
    }
  };
  onRelease(connectionId) {
    const wasActive = this.activeConnections.delete(connectionId);
    const hadLastSeen = this.connectionLastSeen.delete(connectionId);
    this.client.timers.clearOwner(connectionId);
    const pending = this.pendingAborts.get(connectionId);
    if (pending) {
      for (const abort of pending.values()) {
        abort.abort();
      }
      this.pendingAborts.delete(connectionId);
    }
    if (!wasActive && !hadLastSeen && !pending) return;
    if (this.activeConnections.size !== 0) return;
    let onLastDisconnected = this.params.onLastDisconnected ?? "nothing";
    if (typeof onLastDisconnected === "function") {
      onLastDisconnected = onLastDisconnected();
      return;
    }
    switch (onLastDisconnected) {
      case "nothing":
        return;
      case "disconnect":
        if (this._disconnecting || this._destroying) return;
        void this.disconnectSharedClient().catch(() => {
        });
        return;
      case "destroy":
        if (this._destroying) return;
        void this.forceDestroy().catch(() => {
        });
    }
  }
  disconnectSharedClient() {
    if (this._destroying) return this._destroying;
    if (this._disconnecting) return this._disconnecting;
    this._disconnecting = this.client.disconnect().finally(() => {
      this._disconnecting = void 0;
    });
    return this._disconnecting;
  }
  forceDestroy() {
    this.activeConnections.clear();
    this.connectionLastSeen.clear();
    for (const pending of this.pendingAborts.values()) {
      for (const abort of pending.values()) {
        abort.abort();
      }
    }
    this.pendingAborts.clear();
    if (!this._destroying) {
      this._destroying = (this._disconnecting ?? Promise.resolve()).catch(() => {
      }).then(() => this.client.destroy());
    }
    return this._destroying;
  }
  shouldSyncCurrentUser(msg) {
    if (msg.target === "storage-self") {
      return msg.method === "store" || msg.method === "storeFrom" || msg.method === "update";
    }
    return msg.target === "client" && (msg.method === "notifyLoggedIn" || msg.method === "notifyLoggedOut" || msg.method === "importSession");
  }
  onInvoke(msg, respond) {
    if (msg.target === "client" && msg.method === "destroy") {
      this.forceDestroy().then(() => {
        if (msg.void) return;
        respond({
          _mtcuteWorkerId: this.workerId,
          connectionId: msg.connectionId,
          type: "result",
          id: msg.id
        });
      }).catch((err) => {
        respond({
          _mtcuteWorkerId: this.workerId,
          connectionId: msg.connectionId,
          type: "result",
          id: msg.id,
          error: serializeError(err)
        });
      });
      return;
    }
    if (msg.target === "client" && msg.method === "disconnect") {
      this.disconnectSharedClient().then(() => {
        if (msg.void) return;
        respond({
          _mtcuteWorkerId: this.workerId,
          connectionId: msg.connectionId,
          type: "result",
          id: msg.id
        });
      }).catch((err) => {
        respond({
          _mtcuteWorkerId: this.workerId,
          connectionId: msg.connectionId,
          type: "result",
          id: msg.id,
          error: serializeError(err)
        });
      });
      return;
    }
    let target;
    switch (msg.target) {
      case "custom":
        target = this.params.customMethods;
        break;
      case "timers":
        target = {
          upsert: (spec) => Promise.resolve(this.client.timers.upsertOwned(msg.connectionId, spec)),
          cancel: (key) => this.client.timers.cancel(key),
          exists: (key) => this.client.timers.exists(key)
        };
        break;
      case "client":
        target = this.client;
        break;
      case "storage":
        target = this.client.storage;
        break;
      case "storage-self":
        target = this.client.storage.self;
        break;
      case "storage-peers":
        target = this.client.storage.peers;
        break;
      case "app-config":
        target = this.client.appConfig;
        break;
      default: {
        respond({
          _mtcuteWorkerId: this.workerId,
          connectionId: msg.connectionId,
          type: "result",
          id: msg.id,
          error: serializeError(new Error(`Unknown target ${msg.target}`))
        });
        return;
      }
    }
    const method = target[msg.method];
    if (!method) {
      respond({
        _mtcuteWorkerId: this.workerId,
        connectionId: msg.connectionId,
        type: "result",
        id: msg.id,
        error: serializeError(new Error(`Method ${msg.method} not found on ${msg.target}`))
      });
      return;
    }
    let args;
    if (msg.target === "client" && msg.method === "call" && msg.withAbort) {
      const abort = new AbortController();
      this.getConnectionAborts(msg.connectionId).set(msg.id, abort);
      args = [
        deserializeResult(msg.args[0]),
        {
          ...msg.args[1],
          abortSignal: abort.signal
        }
      ];
    } else {
      args = deserializeResult(msg.args);
    }
    const syncCurrentUser = this.shouldSyncCurrentUser(msg);
    Promise.resolve(method.apply(target, args)).then(async (res) => {
      if (msg.withAbort) {
        this.deletePendingAbort(msg.connectionId, msg.id);
      }
      if (syncCurrentUser) {
        try {
          this.broadcast({
            _mtcuteWorkerId: this.workerId,
            type: "self_sync",
            self: serializeResult(await this.client.storage.self.fetch())
          });
        } catch {
        }
      }
      if (msg.void) return;
      respond({
        _mtcuteWorkerId: this.workerId,
        connectionId: msg.connectionId,
        type: "result",
        id: msg.id,
        result: serializeResult(res)
      });
    }).catch((err) => {
      if (msg.withAbort) {
        this.deletePendingAbort(msg.connectionId, msg.id);
      }
      respond({
        _mtcuteWorkerId: this.workerId,
        connectionId: msg.connectionId,
        type: "result",
        id: msg.id,
        error: serializeError(err)
      });
    });
  }
  destroy() {
    if (this._heartbeatSweep !== void 0) {
      clearInterval(this._heartbeatSweep);
    }
    for (const cleanup of this._cleanup) {
      cleanup();
    }
    this._mounted = false;
  }
}
export {
  TelegramWorker
};
