// @fuman/node/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const version = require("./version.cjs");
const connection = require("./net/connection.cjs");
const functions = require("./net/functions.cjs");
const listener = require("./net/listener.cjs");
const convertFuman = require("./streams/convert-fuman.cjs");
const convertWeb = require("./streams/convert-web.cjs");
exports.NODE_VERSION = version.NODE_VERSION;
exports.NODE_VERSION_TUPLE = version.NODE_VERSION_TUPLE;
exports.isNodeVersionAfter = version.isNodeVersionAfter;
exports.TcpConnection = connection.TcpConnection;
exports.TlsConnection = connection.TlsConnection;
exports.connectTcp = functions.connectTcp;
exports.connectTls = functions.connectTls;
exports.listenTcp = functions.listenTcp;
exports.listenTls = functions.listenTls;
exports.upgradeTls = functions.upgradeTls;
exports.TcpListener = listener.TcpListener;
exports.TlsListener = listener.TlsListener;
exports.fumanReadableToNode = convertFuman.fumanReadableToNode;
exports.fumanWritableToNode = convertFuman.fumanWritableToNode;
exports.nodeReadableToFuman = convertFuman.nodeReadableToFuman;
exports.nodeWritableToFuman = convertFuman.nodeWritableToFuman;
exports.nodeReadableToWeb = convertWeb.nodeReadableToWeb;
exports.nodeWritableToWeb = convertWeb.nodeWritableToWeb;
exports.webReadableToNode = convertWeb.webReadableToNode;
exports.webWritableToNode = convertWeb.webWritableToNode;
