// @mtcute/core/highlevel/storage/service/current-user.js
import { u8 } from "@fuman/utils";
import { TlSerializationCounter, TlBinaryWriter, TlBinaryReader } from "@mtcute/tl-runtime";
import { BaseService } from "../../../storage/service/base.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { extractUsernames } from "../../utils/peer-utils.js";
import { MtArgumentError } from "../../../types/errors.js";
const KV_CURRENT_USER = "current_user";
function serialize(info) {
  if (!info) return u8.alloc(0);
  const hasUsernames = info.usernames.length > 0;
  let usernamesOverhead = hasUsernames ? 4 : 0;
  for (const username of info.usernames) {
    usernamesOverhead += TlSerializationCounter.countBytesOverhead(username.length) + username.length;
  }
  const writer = TlBinaryWriter.manual(16 + usernamesOverhead);
  writer.int(1);
  let flags = 0;
  if (info.isBot) flags |= 1;
  if (hasUsernames) flags |= 2;
  if (info.isPremium) flags |= 4;
  writer.int(flags);
  writer.int53(info.userId);
  if (hasUsernames) {
    writer.int(info.usernames.length);
    for (const username of info.usernames) {
      writer.string(username);
    }
  }
  return writer.result();
}
function parse(data) {
  if (data.length === 0) return null;
  const reader = TlBinaryReader.manual(data);
  if (reader.int() !== 1) return null;
  const flags = reader.int();
  const userId = reader.int53();
  let usernames = [];
  if (flags & 2) {
    const len = reader.int();
    usernames = new Array(len);
    for (let i = 0; i < len; i++) {
      usernames[i] = reader.string();
    }
  }
  return {
    userId,
    isBot: (flags & 1) !== 0,
    isPremium: (flags & 4) !== 0,
    usernames
  };
}
class CurrentUserService extends BaseService {
  constructor(_kv, opts) {
    super(opts);
    this._kv = _kv;
  }
  _cached;
  async store(info) {
    if (info && this._cached) {
      this._cached.userId = info.userId;
      this._cached.isBot = info.isBot;
      this._cached.isPremium = info.isPremium;
      this._cached.usernames = info.usernames;
    } else {
      this._cached = info;
    }
    await this._kv.set(KV_CURRENT_USER, serialize(info));
    await this._driver.save?.();
  }
  async storeFrom(user) {
    assertTypeIs("storeFrom", user, "user");
    const obj = {
      userId: user.id,
      isBot: user.bot,
      isPremium: user.premium,
      usernames: extractUsernames(user)
    };
    await this.store(obj);
    return obj;
  }
  async fetch() {
    if (this._cached !== void 0) {
      return this._cached;
    }
    const data = await this._kv.get(KV_CURRENT_USER);
    if (!data) return null;
    const info = parse(data);
    this._cached = info;
    return info;
  }
  getCached(safe = false) {
    if (this._cached === void 0) {
      if (safe) return null;
      throw new MtArgumentError("User info is not cached yet");
    }
    return this._cached;
  }
  async update(params) {
    const info = await this.fetch();
    if (!info) return;
    const { username, usernames, isPremium } = params;
    if (isPremium !== void 0) info.isPremium = isPremium;
    if (username !== void 0) {
      info.usernames[0] = username;
    } else if (usernames !== void 0) {
      info.usernames = usernames;
    }
    return this.store(info);
  }
}
export {
  CurrentUserService
};
