// @mtcute/core/highlevel/storage/service/peers.js
import { LruMap, timers, asyncPool } from "@fuman/utils";
import Long from "long";
import { BaseService } from "../../../storage/service/base.js";
import { longToFastString, longFromFastString } from "../../../utils/long-utils.js";
import { getAllPeersFrom, toggleChannelIdMark, parseMarkedPeerId } from "../../../utils/peer-utils.js";
import { extractUsernames } from "../../utils/peer-utils.js";
const USERNAME_TTL = 24 * 60 * 60 * 1e3;
function getInputPeer(dto) {
  const [type, id] = parseMarkedPeerId(dto.id);
  switch (type) {
    case "user":
      return {
        _: "inputPeerUser",
        userId: id,
        accessHash: longFromFastString(dto.accessHash)
      };
    case "chat":
      return {
        _: "inputPeerChat",
        chatId: id
      };
    case "channel":
      return {
        _: "inputPeerChannel",
        channelId: id,
        accessHash: longFromFastString(dto.accessHash)
      };
  }
}
class PeersService extends BaseService {
  constructor(options, _peers, _refs, common) {
    super(common);
    this.options = options;
    this._peers = _peers;
    this._refs = _refs;
    this._cache = new LruMap(options.cacheSize ?? 100);
    this._writePending = this._writePending.bind(this);
  }
  _cache;
  _pendingWrites = /* @__PURE__ */ new Map();
  _pendingWritesTimer = null;
  async close() {
    if (this._pendingWritesTimer) {
      timers.clearTimeout(this._pendingWritesTimer);
      this._pendingWritesTimer = null;
    }
    await this._writePending();
  }
  _writePendingLater() {
    if (this._pendingWritesTimer) return;
    this._pendingWritesTimer = timers.setTimeout(
      this._writePending,
      this.options.updatesWriteInterval ?? 3e4
    );
  }
  async updatePeersFrom(obj) {
    let count = 0;
    let minCount = 0;
    for (const peer of getAllPeersFrom(obj)) {
      if (peer.min) {
        minCount += 1;
      }
      count += 1;
      await this.store(peer);
    }
    if (count > 0) {
      await this._driver.save?.();
      this._log.debug("cached %d peers (%d min)", count, minCount);
      return true;
    }
    return false;
  }
  async store(peer) {
    let dto;
    let accessHash;
    switch (peer._) {
      case "user": {
        if (!peer.accessHash) {
          this._log.warn("received user without access hash: %j", peer);
          return;
        }
        dto = {
          id: peer.id,
          accessHash: longToFastString(peer.accessHash),
          isMin: peer.min && !(peer.phone !== void 0 && peer.phone.length === 0),
          phone: peer.phone,
          usernames: extractUsernames(peer),
          updated: Date.now(),
          complete: this._serializeTl(peer)
        };
        accessHash = peer.accessHash;
        break;
      }
      case "chat":
      case "chatForbidden": {
        dto = {
          id: -peer.id,
          accessHash: "",
          isMin: false,
          // chats can't be "min"
          updated: Date.now(),
          complete: this._serializeTl(peer),
          usernames: []
        };
        accessHash = Long.ZERO;
        break;
      }
      case "channel":
      case "channelForbidden": {
        if (!peer.accessHash) {
          this._log.warn("received channel without access hash: %j", peer);
          return;
        }
        dto = {
          id: toggleChannelIdMark(peer.id),
          accessHash: longToFastString(peer.accessHash),
          isMin: peer._ === "channel" ? peer.min : false,
          usernames: extractUsernames(peer),
          updated: Date.now(),
          complete: this._serializeTl(peer)
        };
        accessHash = peer.accessHash;
        break;
      }
      default:
        return;
    }
    const cached = this._cache.get(peer.id);
    let newComplete = peer;
    if (peer.min) {
      const existing = cached?.complete ?? await this.getCompleteById(peer.id);
      if (existing && !existing.min) {
        if (existing._ === "channel" && peer._ === "channel") {
          newComplete = {
            ...existing,
            title: peer.title,
            megagroup: peer.megagroup,
            color: peer.color,
            photo: peer.photo,
            username: peer.username,
            usernames: peer.usernames,
            hasGeo: peer.hasGeo,
            noforwards: peer.noforwards,
            emojiStatus: peer.emojiStatus,
            hasLink: peer.hasLink,
            slowmodeEnabled: peer.slowmodeEnabled,
            scam: peer.scam,
            fake: peer.fake,
            gigagroup: peer.gigagroup,
            forum: peer.forum,
            level: peer.level,
            restricted: peer.restricted,
            restrictionReason: peer.restrictionReason,
            joinToSend: peer.joinToSend,
            joinRequest: peer.joinRequest,
            verified: peer.verified,
            defaultBannedRights: peer.defaultBannedRights
          };
        } else if (existing._ === "user" && peer._ === "user") {
          newComplete = {
            ...existing,
            deleted: peer.deleted,
            bot: peer.bot,
            botChatHistory: peer.botChatHistory,
            botNochats: peer.botNochats,
            verified: peer.verified,
            restricted: peer.restricted,
            botInlineGeo: peer.botInlineGeo,
            support: peer.support,
            scam: peer.scam,
            fake: peer.fake,
            botAttachMenu: peer.botAttachMenu,
            premium: peer.premium,
            storiesUnavailable: peer.storiesUnavailable,
            contactRequirePremium: peer.contactRequirePremium,
            botBusiness: peer.botBusiness,
            botGuard: peer.botGuard,
            botHasMainApp: peer.botHasMainApp,
            botCanManageBots: peer.botCanManageBots,
            botGuestchat: peer.botGuestchat,
            photo: peer.applyMinPhoto ? peer.photo : existing.photo,
            status: !existing.status || existing.status._ === "userStatusEmpty" ? peer.status : existing.status,
            botInfoVersion: peer.botInfoVersion,
            restrictionReason: peer.restrictionReason,
            botInlinePlaceholder: peer.botInlinePlaceholder,
            langCode: peer.langCode,
            emojiStatus: peer.emojiStatus,
            color: peer.color,
            profileColor: peer.profileColor,
            botActiveUsers: peer.botActiveUsers
          };
        }
      }
    }
    if (cached && this.options.updatesWriteInterval !== 0) {
      const oldAccessHash = cached.peer.accessHash;
      if (oldAccessHash?.eq(accessHash)) {
        this._pendingWrites.set(peer.id, dto);
        this._writePendingLater();
        cached.complete = newComplete;
        return;
      }
    }
    await this._peers.store(dto);
    this._cache.set(peer.id, {
      peer: getInputPeer(dto),
      complete: newComplete
    });
    await this._refs.deleteByPeer(peer.id);
  }
  async _writePending() {
    try {
      await asyncPool(this._pendingWrites.values(), async (dto) => {
        await this._peers.store(dto);
        await this._refs.deleteByPeer(dto.id);
      });
      this._pendingWrites.clear();
    } catch (err) {
      this._log.warn("failed to write pending updates: %e", err);
    }
  }
  _returnCaching(id, dto) {
    const peer = getInputPeer(dto);
    const complete = this._deserializeTl(dto.complete);
    this._cache.set(id, {
      peer,
      complete
    });
    return peer;
  }
  async getById(id, allowRefs = true) {
    const cached = this._cache.get(id);
    if (cached) return cached.peer;
    const dto = await this._peers.getById(id);
    if (dto && !dto.isMin) {
      return this._returnCaching(id, dto);
    }
    if ((!dto || dto.isMin) && allowRefs) {
      const ref = await this._refs.getForPeer(id);
      if (!ref) return null;
      const [chatId, msgId] = ref;
      const chat = await this.getById(chatId, false);
      if (!chat) return null;
      if (id > 0) {
        return {
          _: "inputPeerUserFromMessage",
          peer: chat,
          msgId,
          userId: id
        };
      }
      return {
        _: "inputPeerChannelFromMessage",
        peer: chat,
        msgId,
        channelId: toggleChannelIdMark(id)
      };
    }
    return null;
  }
  async getMinAccessHash(id) {
    const dto = await this._peers.getById(id);
    if (!dto) return null;
    return longFromFastString(dto.accessHash);
  }
  async getByPhone(phone) {
    const dto = await this._peers.getByPhone(phone);
    if (!dto) return null;
    return this._returnCaching(dto.id, dto);
  }
  async getByUsername(username) {
    const dto = await this._peers.getByUsername(username.toLowerCase());
    if (!dto) return null;
    if (Date.now() - dto.updated > USERNAME_TTL) {
      return null;
    }
    return this._returnCaching(dto.id, dto);
  }
  async getCompleteById(id, allowMin = false) {
    const cached = this._cache.get(id);
    if (cached) return cached.complete;
    const dto = await this._peers.getById(id);
    if (!dto) return null;
    if (!allowMin && dto.isMin) return null;
    const cacheItem = {
      peer: getInputPeer(dto),
      complete: this._deserializeTl(dto.complete)
    };
    this._cache.set(id, cacheItem);
    return cacheItem.complete;
  }
}
export {
  PeersService
};
