// @mtcute/core/highlevel/storage/service/ref-messages.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const base = require("../../../storage/service/base.cjs");
class RefMessagesService extends base.BaseService {
  constructor(options, _refs, common) {
    super(common);
    this.options = options;
    this._refs = _refs;
    this._cache = new utils.LruMap(options.cacheSize ?? 1e3);
  }
  _cache;
  async store(peerId, chatId, msgId) {
    await this._refs.store(peerId, chatId, msgId);
    this._cache.set(peerId, [chatId, msgId]);
  }
  async getForPeer(peerId) {
    const cached = this._cache.get(peerId);
    if (cached) return cached;
    const ref = await this._refs.getByPeer(peerId);
    if (ref) this._cache.set(peerId, ref);
    return ref;
  }
  async delete(chatId, msgIds) {
    await this._refs.delete(chatId, msgIds);
    this._cache.clear();
  }
  async deleteByPeer(peerId) {
    await this._refs.deleteByPeer(peerId);
    this._cache.delete(peerId);
  }
}
exports.RefMessagesService = RefMessagesService;
