// @mtcute/core/highlevel/methods/users/resolve-peer.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const errors = require("../../../types/errors.cjs");
const peerUtils = require("../../../utils/peer-utils.cjs");
const errors$1 = require("../../types/errors.cjs");
const peerUtils$1 = require("../../utils/peer-utils.cjs");
const batchedQueries = require("../chats/batched-queries.cjs");
const _resolveCommon = require("./_resolve-common.cjs");
const resolvePhoneNumber = require("./resolve-phone-number.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
async function resolvePeer(client, peerId, force = false) {
  peerId = _resolveCommon._normalizePeerId(peerId);
  if (typeof peerId === "object") {
    return peerId;
  }
  if (typeof peerId === "number" && !force) {
    const fromStorage = await client.storage.peers.getById(peerId);
    if (fromStorage) return fromStorage;
  }
  if (typeof peerId === "string") {
    if (peerId === "self" || peerId === "me") return { _: "inputPeerSelf" };
    peerId = peerId.replace(/^@/, "");
    if (peerId.match(/^\d/)) {
      throw new errors.MtArgumentError(
        `Invalid username: ${peerId}. If you wanted to resolve by user ID, pass a number instead. If you wanted to resolve by phone number, use resolvePhoneNumber instead.`
      );
    }
    if (!force) {
      const fromStorage = await client.storage.peers.getByUsername(peerId);
      if (fromStorage) return fromStorage;
    }
    try {
      const res = await client.call({
        _: "contacts.resolveUsername",
        username: peerId
      });
      return _resolveCommon._handleContactsResolvedPeer(peerId, res);
    } catch (e) {
      if (innerTl.RpcError.is(e, "USERNAME_NOT_OCCUPIED") || innerTl.RpcError.is(e, "USERNAME_INVALID")) {
        throw new errors$1.MtPeerNotFoundError(`Peer with username ${peerId} was not found`);
      } else {
        throw e;
      }
    }
  }
  const [peerType, bareId] = peerUtils.parseMarkedPeerId(peerId);
  if (peerType === "chat") {
    return {
      _: "inputPeerChat",
      chatId: bareId
    };
  }
  const cached = await client.storage.peers.getCompleteById(peerId, true);
  if (cached && (cached._ === "channel" || cached._ === "user")) {
    const [username] = peerUtils$1.extractUsernames(cached);
    if (username) {
      const resolved = await resolvePeer(client, username, true).catch(() => null);
      if (resolved && peerUtils.getMarkedPeerId(resolved) === peerId) {
        return resolved;
      }
    }
    if (cached._ === "user" && cached.phone) {
      const resolved = await resolvePhoneNumber.resolvePhoneNumber(client, cached.phone, true).catch(() => null);
      if (resolved && peerUtils.getMarkedPeerId(resolved) === peerId) {
        return resolved;
      }
    }
  }
  switch (peerType) {
    case "user": {
      try {
        const res = await batchedQueries._getUsersBatched(client, {
          _: "inputUser",
          userId: bareId,
          accessHash: Long.ZERO
        });
        if (res != null && res._ === "user" && res.accessHash != null) {
          return {
            _: "inputPeerUser",
            userId: bareId,
            accessHash: res.accessHash
          };
        }
      } catch (e) {
        if (!innerTl.RpcError.is(e, "USER_INVALID")) throw e;
      }
      break;
    }
    case "channel": {
      try {
        const res = await batchedQueries._getChannelsBatched(client, {
          _: "inputChannel",
          channelId: bareId,
          accessHash: Long.ZERO
        });
        if (res != null && res._ === "channel" && res.accessHash != null) {
          return {
            _: "inputPeerChannel",
            channelId: bareId,
            accessHash: res.accessHash
          };
        }
      } catch (e) {
        if (!innerTl.RpcError.is(e, "CHANNEL_INVALID")) throw e;
      }
      break;
    }
  }
  throw new errors$1.MtPeerNotFoundError(`Peer ${peerId} is not found in local cache`);
}
async function resolveUser(client, peerId, force = false) {
  return peerUtils$1.toInputUser(await resolvePeer(client, peerId, force), peerId);
}
async function resolveChannel(client, peerId, force = false) {
  return peerUtils$1.toInputChannel(await resolvePeer(client, peerId, force), peerId);
}
exports.resolveChannel = resolveChannel;
exports.resolvePeer = resolvePeer;
exports.resolveUser = resolveUser;
