// @mtcute/core/highlevel/methods/users/resolve-peer.js
import Long from "long";
import { MtArgumentError } from "../../../types/errors.js";
import { parseMarkedPeerId, getMarkedPeerId } from "../../../utils/peer-utils.js";
import { MtPeerNotFoundError } from "../../types/errors.js";
import { extractUsernames, toInputUser, toInputChannel } from "../../utils/peer-utils.js";
import { _getChannelsBatched, _getUsersBatched } from "../chats/batched-queries.js";
import { _normalizePeerId, _handleContactsResolvedPeer } from "./_resolve-common.js";
import { resolvePhoneNumber } from "./resolve-phone-number.js";
import { RpcError } from "../../../tl/inner-tl.js";
async function resolvePeer(client, peerId, force = false) {
  peerId = _normalizePeerId(peerId);
  if (typeof peerId === "object") {
    return peerId;
  }
  if (typeof peerId === "number" && !force) {
    const fromStorage = await client.storage.peers.getById(peerId);
    if (fromStorage) return fromStorage;
  }
  if (typeof peerId === "string") {
    if (peerId === "self" || peerId === "me") return { _: "inputPeerSelf" };
    peerId = peerId.replace(/^@/, "");
    if (peerId.match(/^\d/)) {
      throw new MtArgumentError(
        `Invalid username: ${peerId}. If you wanted to resolve by user ID, pass a number instead. If you wanted to resolve by phone number, use resolvePhoneNumber instead.`
      );
    }
    if (!force) {
      const fromStorage = await client.storage.peers.getByUsername(peerId);
      if (fromStorage) return fromStorage;
    }
    try {
      const res = await client.call({
        _: "contacts.resolveUsername",
        username: peerId
      });
      return _handleContactsResolvedPeer(peerId, res);
    } catch (e) {
      if (RpcError.is(e, "USERNAME_NOT_OCCUPIED") || RpcError.is(e, "USERNAME_INVALID")) {
        throw new MtPeerNotFoundError(`Peer with username ${peerId} was not found`);
      } else {
        throw e;
      }
    }
  }
  const [peerType, bareId] = parseMarkedPeerId(peerId);
  if (peerType === "chat") {
    return {
      _: "inputPeerChat",
      chatId: bareId
    };
  }
  const cached = await client.storage.peers.getCompleteById(peerId, true);
  if (cached && (cached._ === "channel" || cached._ === "user")) {
    const [username] = extractUsernames(cached);
    if (username) {
      const resolved = await resolvePeer(client, username, true).catch(() => null);
      if (resolved && getMarkedPeerId(resolved) === peerId) {
        return resolved;
      }
    }
    if (cached._ === "user" && cached.phone) {
      const resolved = await resolvePhoneNumber(client, cached.phone, true).catch(() => null);
      if (resolved && getMarkedPeerId(resolved) === peerId) {
        return resolved;
      }
    }
  }
  switch (peerType) {
    case "user": {
      try {
        const res = await _getUsersBatched(client, {
          _: "inputUser",
          userId: bareId,
          accessHash: Long.ZERO
        });
        if (res != null && res._ === "user" && res.accessHash != null) {
          return {
            _: "inputPeerUser",
            userId: bareId,
            accessHash: res.accessHash
          };
        }
      } catch (e) {
        if (!RpcError.is(e, "USER_INVALID")) throw e;
      }
      break;
    }
    case "channel": {
      try {
        const res = await _getChannelsBatched(client, {
          _: "inputChannel",
          channelId: bareId,
          accessHash: Long.ZERO
        });
        if (res != null && res._ === "channel" && res.accessHash != null) {
          return {
            _: "inputPeerChannel",
            channelId: bareId,
            accessHash: res.accessHash
          };
        }
      } catch (e) {
        if (!RpcError.is(e, "CHANNEL_INVALID")) throw e;
      }
      break;
    }
  }
  throw new MtPeerNotFoundError(`Peer ${peerId} is not found in local cache`);
}
async function resolveUser(client, peerId, force = false) {
  return toInputUser(await resolvePeer(client, peerId, force), peerId);
}
async function resolveChannel(client, peerId, force = false) {
  return toInputChannel(await resolvePeer(client, peerId, force), peerId);
}
export {
  resolveChannel,
  resolvePeer,
  resolveUser
};
