// @mtcute/core/highlevel/methods/users/save-music-to-profile.js
import { assert } from "@fuman/utils";
import { tdFileId } from "@mtcute/file-id";
import { assertTrue } from "../../../utils/type-assertions.js";
import { fileIdToInputDocument } from "../../utils/convert-file-id.js";
import { _normalizeInputMedia } from "../files/normalize-input-media.js";
import { isAnyInputDocument } from "../../../tl/inner-tl.js";
async function saveMusicToProfile(client, params) {
  let after = params.after;
  if (tdFileId.isFileIdLike(after)) {
    after = fileIdToInputDocument(after);
  }
  let audio = params.audio;
  if (tdFileId.isFileIdLike(audio)) {
    audio = {
      type: "audio",
      file: audio
    };
  }
  let audioId;
  if (isAnyInputDocument(audio)) {
    audioId = audio;
  } else {
    const media = await _normalizeInputMedia(client, audio, {
      progressCallback: params.progressCallback,
      uploadPeer: { _: "inputPeerSelf" }
    }, true);
    assert(media._ === "inputMediaDocument");
    audioId = media.id;
  }
  const res = await client.call({
    _: "account.saveMusic",
    id: audioId,
    afterId: after
  });
  assertTrue("account.saveMusic", res);
}
async function unsaveMusicFromProfile(client, params) {
  let audio = params.audio;
  if (tdFileId.isFileIdLike(audio)) {
    audio = fileIdToInputDocument(audio);
  }
  const res = await client.call({
    _: "account.saveMusic",
    id: audio,
    unsave: true
  });
  assertTrue("account.saveMusic", res);
}
export {
  saveMusicToProfile,
  unsaveMusicFromProfile
};
