// @mtcute/core/highlevel/methods/users/save-music-to-profile.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const fileId = require("@mtcute/file-id");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const convertFileId = require("../../utils/convert-file-id.cjs");
const normalizeInputMedia = require("../files/normalize-input-media.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
async function saveMusicToProfile(client, params) {
  let after = params.after;
  if (fileId.tdFileId.isFileIdLike(after)) {
    after = convertFileId.fileIdToInputDocument(after);
  }
  let audio = params.audio;
  if (fileId.tdFileId.isFileIdLike(audio)) {
    audio = {
      type: "audio",
      file: audio
    };
  }
  let audioId;
  if (innerTl.isAnyInputDocument(audio)) {
    audioId = audio;
  } else {
    const media = await normalizeInputMedia._normalizeInputMedia(client, audio, {
      progressCallback: params.progressCallback,
      uploadPeer: { _: "inputPeerSelf" }
    }, true);
    utils.assert(media._ === "inputMediaDocument");
    audioId = media.id;
  }
  const res = await client.call({
    _: "account.saveMusic",
    id: audioId,
    afterId: after
  });
  typeAssertions.assertTrue("account.saveMusic", res);
}
async function unsaveMusicFromProfile(client, params) {
  let audio = params.audio;
  if (fileId.tdFileId.isFileIdLike(audio)) {
    audio = convertFileId.fileIdToInputDocument(audio);
  }
  const res = await client.call({
    _: "account.saveMusic",
    id: audio,
    unsave: true
  });
  typeAssertions.assertTrue("account.saveMusic", res);
}
exports.saveMusicToProfile = saveMusicToProfile;
exports.unsaveMusicFromProfile = unsaveMusicFromProfile;
