// @mtcute/core/highlevel/methods/users/set-online.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
async function sendOnline(client, online) {
  await client.call({
    _: "account.updateStatus",
    offline: !online
  });
}
const TIMER_ID = "online";
const TIMER_INTERVAL = 24e4;
async function setOnline(client, online = true) {
  if (online) {
    await client.timers.upsert({
      kind: "rpc",
      key: TIMER_ID,
      interval: TIMER_INTERVAL,
      startNow: true,
      request: {
        _: "account.updateStatus",
        offline: false
      }
    });
  } else {
    await client.timers.cancel(TIMER_ID);
    await client.call({
      _: "account.updateStatus",
      offline: true
    });
  }
}
exports.sendOnline = sendOnline;
exports.setOnline = setOnline;
