// @mtcute/core/highlevel/types/peers/chat-event/actions.js
import { toggleChannelIdMark } from "../../../../utils/peer-utils.js";
import { assertTypeIs } from "../../../../utils/type-assertions.js";
import { Photo as Photo$2 } from "../../media/photo.js";
import { Message as Message$2 } from "../../messages/message.js";
import { EmojiStatus as EmojiStatus$1 } from "../../reactions/emoji-status.js";
import { ChatColors as ChatColors$1 } from "../chat-colors.js";
import { ChatInviteLink as ChatInviteLink$2 } from "../chat-invite-link.js";
import { ChatMember as ChatMember$2 } from "../chat-member.js";
import { ChatPermissions as ChatPermissions$1 } from "../chat-permissions.js";
import { ForumTopic as ForumTopic$2 } from "../forum-topic.js";
import { User as User$2 } from "../user.js";
function _actionFromTl(e, peers, peerId) {
  switch (e._) {
    case "channelAdminLogEventActionParticipantJoin":
      return { type: "user_joined" };
    case "channelAdminLogEventActionChangeTitle":
      return {
        type: "title_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeAbout":
      return {
        type: "description_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeUsername":
      return {
        type: "username_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeUsernames":
      return {
        type: "usernames_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangePhoto":
      return {
        type: "photo_changed",
        old: e.prevPhoto._ === "photo" ? new Photo$2(e.prevPhoto) : null,
        new: e.newPhoto._ === "photo" ? new Photo$2(e.newPhoto) : null
      };
    case "channelAdminLogEventActionToggleInvites":
      return {
        type: "invites_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionToggleSignatures":
      return {
        type: "signatures_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionUpdatePinned":
      return {
        type: "msg_pinned",
        message: new Message$2(e.message, peers)
      };
    case "channelAdminLogEventActionEditMessage":
      return {
        type: "msg_edited",
        old: new Message$2(e.prevMessage, peers),
        new: new Message$2(e.newMessage, peers)
      };
    case "channelAdminLogEventActionDeleteMessage":
      return {
        type: "msg_deleted",
        message: new Message$2(e.message, peers)
      };
    case "channelAdminLogEventActionParticipantLeave":
      return { type: "user_left" };
    case "channelAdminLogEventActionParticipantInvite":
      return {
        type: "user_invited",
        member: new ChatMember$2(e.participant, peers)
      };
    case "channelAdminLogEventActionParticipantToggleBan":
      return {
        type: "user_perms_changed",
        old: new ChatMember$2(e.prevParticipant, peers),
        new: new ChatMember$2(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionParticipantToggleAdmin":
      return {
        type: "user_admin_perms_changed",
        old: new ChatMember$2(e.prevParticipant, peers),
        new: new ChatMember$2(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionChangeStickerSet":
      return {
        type: "stickerset_changed",
        old: e.prevStickerset,
        new: e.newStickerset
      };
    case "channelAdminLogEventActionTogglePreHistoryHidden":
      return {
        type: "history_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionDefaultBannedRights":
      return {
        type: "def_perms_changed",
        old: new ChatPermissions$1(e.prevBannedRights),
        new: new ChatPermissions$1(e.newBannedRights)
      };
    case "channelAdminLogEventActionStopPoll":
      return {
        type: "poll_stopped",
        message: new Message$2(e.message, peers)
      };
    case "channelAdminLogEventActionChangeLinkedChat":
      return {
        type: "linked_chat_changed",
        old: toggleChannelIdMark(e.prevValue),
        new: toggleChannelIdMark(e.newValue)
      };
    case "channelAdminLogEventActionChangeLocation":
      return {
        type: "location_changed",
        old: e.prevValue._ === "channelLocationEmpty" ? null : e.prevValue,
        new: e.newValue._ === "channelLocationEmpty" ? null : e.newValue
      };
    case "channelAdminLogEventActionToggleSlowMode":
      return {
        type: "slow_mode_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionStartGroupCall":
      return {
        type: "call_started",
        call: e.call
      };
    case "channelAdminLogEventActionDiscardGroupCall":
      return {
        type: "call_ended",
        call: e.call
      };
    case "channelAdminLogEventActionParticipantMute":
    case "channelAdminLogEventActionParticipantUnmute":
    case "channelAdminLogEventActionParticipantVolume":
      return null;
    case "channelAdminLogEventActionToggleGroupCallSetting":
      return {
        type: "call_setting_changed",
        joinMuted: e.joinMuted
      };
    case "channelAdminLogEventActionParticipantJoinByInvite":
      return {
        type: "user_joined_invite",
        link: new ChatInviteLink$2(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteDelete":
      return {
        type: "invite_deleted",
        link: new ChatInviteLink$2(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteRevoke":
      return {
        type: "invite_revoked",
        link: new ChatInviteLink$2(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteEdit":
      return {
        type: "invite_edited",
        old: new ChatInviteLink$2(e.prevInvite, peers),
        new: new ChatInviteLink$2(e.newInvite, peers)
      };
    case "channelAdminLogEventActionChangeHistoryTTL":
      return {
        type: "ttl_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionParticipantJoinByRequest":
      return {
        type: "user_joined_approved",
        link: new ChatInviteLink$2(e.invite, peers),
        approvedBy: new User$2(peers.user(e.approvedBy))
      };
    // channelAdminLogEventActionCreateTopic#58707d28 topic:ForumTopic = ChannelAdminLogEventAction;
    // channelAdminLogEventActionEditTopic#f06fe208 prev_topic:ForumTopic new_topic:ForumTopic
    // channelAdminLogEventActionDeleteTopic#ae168909 topic:ForumTopic = ChannelAdminLogEventAction;
    case "channelAdminLogEventActionToggleForum":
      return {
        type: "forum_toggled",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionCreateTopic":
      assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.topic, "forumTopic");
      return {
        type: "topic_created",
        topic: new ForumTopic$2(e.topic, peers)
      };
    case "channelAdminLogEventActionEditTopic":
      assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.prevTopic, "forumTopic");
      assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.newTopic, "forumTopic");
      return {
        type: "topic_edited",
        old: new ForumTopic$2(e.prevTopic, peers),
        new: new ForumTopic$2(e.newTopic, peers)
      };
    case "channelAdminLogEventActionDeleteTopic":
      assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.topic, "forumTopic");
      return {
        type: "topic_deleted",
        topic: new ForumTopic$2(e.topic, peers)
      };
    case "channelAdminLogEventActionToggleNoForwards":
      return {
        type: "no_forwards_toggled",
        enabled: e.newValue
      };
    // case 'channelAdminLogEventActionPinTopic'
    // ^ looks like it is not used, and pinned topics are not at all presented in the event log
    case "channelAdminLogEventActionToggleSignatureProfiles":
      return {
        type: "signature_profiles_toggled",
        new: e.newValue
      };
    case "channelAdminLogEventActionParticipantSubExtend":
      return {
        type: "sub_extend",
        prev: new ChatMember$2(e.prevParticipant, peers),
        new: new ChatMember$2(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionChangeAvailableReactions":
      return {
        type: "available_reactions_changed",
        prev: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeEmojiStatus":
      return {
        type: "emoji_status_changed",
        prev: EmojiStatus$1.fromTl(e.prevValue),
        new: EmojiStatus$1.fromTl(e.newValue)
      };
    case "channelAdminLogEventActionChangeEmojiStickerSet":
      return {
        type: "emoji_stickerset_changed",
        old: e.prevStickerset,
        new: e.newStickerset
      };
    case "channelAdminLogEventActionChangePeerColor":
    case "channelAdminLogEventActionChangeProfilePeerColor":
      if (e.prevValue._ !== "peerColor" || e.newValue._ !== "peerColor") return null;
      return {
        type: e._ === "channelAdminLogEventActionChangePeerColor" ? "peer_color_changed" : "profile_peer_color_changed",
        prev: new ChatColors$1(peerId, e.prevValue),
        new: new ChatColors$1(peerId, e.newValue)
      };
    case "channelAdminLogEventActionChangeWallpaper":
      return {
        type: "wallpaper_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionSendMessage":
      return {
        type: "msg_sent",
        message: new Message$2(e.message, peers)
      };
    case "channelAdminLogEventActionToggleAntiSpam":
      return {
        type: "toggle_anti_spam",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionToggleAutotranslation":
      return {
        type: "toggle_autotranslation",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionParticipantEditRank":
      return {
        type: "participant_rank_edited",
        user: new User$2(peers.user(e.userId)),
        prevRank: e.prevRank,
        newRank: e.newRank
      };
    default:
      return null;
  }
}
export {
  _actionFromTl
};
