// @mtcute/core/highlevel/types/peers/chat-event/actions.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const peerUtils = require("../../../../utils/peer-utils.cjs");
const typeAssertions = require("../../../../utils/type-assertions.cjs");
const photo = require("../../media/photo.cjs");
const message = require("../../messages/message.cjs");
const emojiStatus = require("../../reactions/emoji-status.cjs");
const chatColors = require("../chat-colors.cjs");
const chatInviteLink = require("../chat-invite-link.cjs");
const chatMember = require("../chat-member.cjs");
const chatPermissions = require("../chat-permissions.cjs");
const forumTopic = require("../forum-topic.cjs");
const user = require("../user.cjs");
function _actionFromTl(e, peers, peerId) {
  switch (e._) {
    case "channelAdminLogEventActionParticipantJoin":
      return { type: "user_joined" };
    case "channelAdminLogEventActionChangeTitle":
      return {
        type: "title_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeAbout":
      return {
        type: "description_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeUsername":
      return {
        type: "username_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeUsernames":
      return {
        type: "usernames_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangePhoto":
      return {
        type: "photo_changed",
        old: e.prevPhoto._ === "photo" ? new photo.Photo(e.prevPhoto) : null,
        new: e.newPhoto._ === "photo" ? new photo.Photo(e.newPhoto) : null
      };
    case "channelAdminLogEventActionToggleInvites":
      return {
        type: "invites_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionToggleSignatures":
      return {
        type: "signatures_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionUpdatePinned":
      return {
        type: "msg_pinned",
        message: new message.Message(e.message, peers)
      };
    case "channelAdminLogEventActionEditMessage":
      return {
        type: "msg_edited",
        old: new message.Message(e.prevMessage, peers),
        new: new message.Message(e.newMessage, peers)
      };
    case "channelAdminLogEventActionDeleteMessage":
      return {
        type: "msg_deleted",
        message: new message.Message(e.message, peers)
      };
    case "channelAdminLogEventActionParticipantLeave":
      return { type: "user_left" };
    case "channelAdminLogEventActionParticipantInvite":
      return {
        type: "user_invited",
        member: new chatMember.ChatMember(e.participant, peers)
      };
    case "channelAdminLogEventActionParticipantToggleBan":
      return {
        type: "user_perms_changed",
        old: new chatMember.ChatMember(e.prevParticipant, peers),
        new: new chatMember.ChatMember(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionParticipantToggleAdmin":
      return {
        type: "user_admin_perms_changed",
        old: new chatMember.ChatMember(e.prevParticipant, peers),
        new: new chatMember.ChatMember(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionChangeStickerSet":
      return {
        type: "stickerset_changed",
        old: e.prevStickerset,
        new: e.newStickerset
      };
    case "channelAdminLogEventActionTogglePreHistoryHidden":
      return {
        type: "history_toggled",
        old: !e.newValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionDefaultBannedRights":
      return {
        type: "def_perms_changed",
        old: new chatPermissions.ChatPermissions(e.prevBannedRights),
        new: new chatPermissions.ChatPermissions(e.newBannedRights)
      };
    case "channelAdminLogEventActionStopPoll":
      return {
        type: "poll_stopped",
        message: new message.Message(e.message, peers)
      };
    case "channelAdminLogEventActionChangeLinkedChat":
      return {
        type: "linked_chat_changed",
        old: peerUtils.toggleChannelIdMark(e.prevValue),
        new: peerUtils.toggleChannelIdMark(e.newValue)
      };
    case "channelAdminLogEventActionChangeLocation":
      return {
        type: "location_changed",
        old: e.prevValue._ === "channelLocationEmpty" ? null : e.prevValue,
        new: e.newValue._ === "channelLocationEmpty" ? null : e.newValue
      };
    case "channelAdminLogEventActionToggleSlowMode":
      return {
        type: "slow_mode_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionStartGroupCall":
      return {
        type: "call_started",
        call: e.call
      };
    case "channelAdminLogEventActionDiscardGroupCall":
      return {
        type: "call_ended",
        call: e.call
      };
    case "channelAdminLogEventActionParticipantMute":
    case "channelAdminLogEventActionParticipantUnmute":
    case "channelAdminLogEventActionParticipantVolume":
      return null;
    case "channelAdminLogEventActionToggleGroupCallSetting":
      return {
        type: "call_setting_changed",
        joinMuted: e.joinMuted
      };
    case "channelAdminLogEventActionParticipantJoinByInvite":
      return {
        type: "user_joined_invite",
        link: new chatInviteLink.ChatInviteLink(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteDelete":
      return {
        type: "invite_deleted",
        link: new chatInviteLink.ChatInviteLink(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteRevoke":
      return {
        type: "invite_revoked",
        link: new chatInviteLink.ChatInviteLink(e.invite, peers)
      };
    case "channelAdminLogEventActionExportedInviteEdit":
      return {
        type: "invite_edited",
        old: new chatInviteLink.ChatInviteLink(e.prevInvite, peers),
        new: new chatInviteLink.ChatInviteLink(e.newInvite, peers)
      };
    case "channelAdminLogEventActionChangeHistoryTTL":
      return {
        type: "ttl_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionParticipantJoinByRequest":
      return {
        type: "user_joined_approved",
        link: new chatInviteLink.ChatInviteLink(e.invite, peers),
        approvedBy: new user.User(peers.user(e.approvedBy))
      };
    // channelAdminLogEventActionCreateTopic#58707d28 topic:ForumTopic = ChannelAdminLogEventAction;
    // channelAdminLogEventActionEditTopic#f06fe208 prev_topic:ForumTopic new_topic:ForumTopic
    // channelAdminLogEventActionDeleteTopic#ae168909 topic:ForumTopic = ChannelAdminLogEventAction;
    case "channelAdminLogEventActionToggleForum":
      return {
        type: "forum_toggled",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionCreateTopic":
      typeAssertions.assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.topic, "forumTopic");
      return {
        type: "topic_created",
        topic: new forumTopic.ForumTopic(e.topic, peers)
      };
    case "channelAdminLogEventActionEditTopic":
      typeAssertions.assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.prevTopic, "forumTopic");
      typeAssertions.assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.newTopic, "forumTopic");
      return {
        type: "topic_edited",
        old: new forumTopic.ForumTopic(e.prevTopic, peers),
        new: new forumTopic.ForumTopic(e.newTopic, peers)
      };
    case "channelAdminLogEventActionDeleteTopic":
      typeAssertions.assertTypeIs("ChannelAdminLogEventActionCreateTopic#topic", e.topic, "forumTopic");
      return {
        type: "topic_deleted",
        topic: new forumTopic.ForumTopic(e.topic, peers)
      };
    case "channelAdminLogEventActionToggleNoForwards":
      return {
        type: "no_forwards_toggled",
        enabled: e.newValue
      };
    // case 'channelAdminLogEventActionPinTopic'
    // ^ looks like it is not used, and pinned topics are not at all presented in the event log
    case "channelAdminLogEventActionToggleSignatureProfiles":
      return {
        type: "signature_profiles_toggled",
        new: e.newValue
      };
    case "channelAdminLogEventActionParticipantSubExtend":
      return {
        type: "sub_extend",
        prev: new chatMember.ChatMember(e.prevParticipant, peers),
        new: new chatMember.ChatMember(e.newParticipant, peers)
      };
    case "channelAdminLogEventActionChangeAvailableReactions":
      return {
        type: "available_reactions_changed",
        prev: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionChangeEmojiStatus":
      return {
        type: "emoji_status_changed",
        prev: emojiStatus.EmojiStatus.fromTl(e.prevValue),
        new: emojiStatus.EmojiStatus.fromTl(e.newValue)
      };
    case "channelAdminLogEventActionChangeEmojiStickerSet":
      return {
        type: "emoji_stickerset_changed",
        old: e.prevStickerset,
        new: e.newStickerset
      };
    case "channelAdminLogEventActionChangePeerColor":
    case "channelAdminLogEventActionChangeProfilePeerColor":
      if (e.prevValue._ !== "peerColor" || e.newValue._ !== "peerColor") return null;
      return {
        type: e._ === "channelAdminLogEventActionChangePeerColor" ? "peer_color_changed" : "profile_peer_color_changed",
        prev: new chatColors.ChatColors(peerId, e.prevValue),
        new: new chatColors.ChatColors(peerId, e.newValue)
      };
    case "channelAdminLogEventActionChangeWallpaper":
      return {
        type: "wallpaper_changed",
        old: e.prevValue,
        new: e.newValue
      };
    case "channelAdminLogEventActionSendMessage":
      return {
        type: "msg_sent",
        message: new message.Message(e.message, peers)
      };
    case "channelAdminLogEventActionToggleAntiSpam":
      return {
        type: "toggle_anti_spam",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionToggleAutotranslation":
      return {
        type: "toggle_autotranslation",
        enabled: e.newValue
      };
    case "channelAdminLogEventActionParticipantEditRank":
      return {
        type: "participant_rank_edited",
        user: new user.User(peers.user(e.userId)),
        prevRank: e.prevRank,
        newRank: e.newRank
      };
    default:
      return null;
  }
}
exports._actionFromTl = _actionFromTl;
