// @mtcute/core/highlevel/types/peers/chat-event/index.js
import { memoizeGetters } from "../../../utils/memoize.js";
import { User as User$2 } from "../user.js";
import { _actionFromTl } from "./actions.js";
import { makeInspectable } from "../../../utils/inspectable.js";
class ChatEvent {
  constructor(raw, _peers, _chatId) {
    this.raw = raw;
    this._peers = _peers;
    this._chatId = _chatId;
  }
  /**
   * Event ID.
   *
   * Event IDs are generated in direct chronological order
   * (i.e. newer events have bigger event ID)
   */
  get id() {
    return this.raw.id;
  }
  /**
   * Date of the event
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Actor of the event
   */
  get actor() {
    return new User$2(this._peers.user(this.raw.userId));
  }
  get action() {
    return _actionFromTl(this.raw.action, this._peers, this._chatId);
  }
}
const ChatEvent$1 = /* @__PURE__ */ memoizeGetters(ChatEvent, ["actor", "action"]);
const ChatEvent$2 = /* @__PURE__ */ makeInspectable(ChatEvent$1);
export {
  ChatEvent$2 as ChatEvent,
  _actionFromTl
};
