// @mtcute/core/highlevel/types/peers/chat-event/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../../utils/memoize.cjs");
const user = require("../user.cjs");
const actions = require("./actions.cjs");
const inspectable = require("../../../utils/inspectable.cjs");
class ChatEvent {
  constructor(raw, _peers, _chatId) {
    this.raw = raw;
    this._peers = _peers;
    this._chatId = _chatId;
  }
  /**
   * Event ID.
   *
   * Event IDs are generated in direct chronological order
   * (i.e. newer events have bigger event ID)
   */
  get id() {
    return this.raw.id;
  }
  /**
   * Date of the event
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Actor of the event
   */
  get actor() {
    return new user.User(this._peers.user(this.raw.userId));
  }
  get action() {
    return actions._actionFromTl(this.raw.action, this._peers, this._chatId);
  }
}
const ChatEvent$1 = /* @__PURE__ */ memoize.memoizeGetters(ChatEvent, ["actor", "action"]);
const ChatEvent$2 = /* @__PURE__ */ inspectable.makeInspectable(ChatEvent$1);
exports._actionFromTl = actions._actionFromTl;
exports.ChatEvent = ChatEvent$2;
