// @mtcute/core/highlevel/types/bots/keyboards/builder.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class BotKeyboardBuilder {
  constructor(maxRowWidth = 3) {
    this.maxRowWidth = maxRowWidth;
  }
  _buttons = [];
  /**
   * Add buttons, wrapping them once {@link maxRowWidth} is reached
   *
   * @param buttons  Buttons to add
   */
  push(...buttons) {
    if (!buttons.length) return this;
    let row = [];
    buttons.forEach((btn) => {
      if (typeof btn === "function") btn = btn();
      if (!btn) return;
      row.push(btn);
      if (row.length === this.maxRowWidth) {
        this._buttons.push(row);
        row = [];
      }
    });
    if (row.length) {
      this._buttons.push(row);
    }
    return this;
  }
  /**
   * Add a row of buttons. Will not be wrapped.
   *
   * @param row  Row or a function that will populate it
   */
  row(row) {
    if (typeof row === "function") {
      const fn = row;
      row = [];
      fn(row);
    }
    const normal = row.filter(Boolean);
    if (normal.length) this._buttons.push(normal);
    return this;
  }
  /**
   * Append a button to the last row, wrapping if needed.
   *
   * @param btn  Button to add
   * @param force  Whether to forcefully add the button (i.e. do not wrap)
   */
  append(btn, force = false) {
    if (typeof btn === "function") btn = btn();
    if (!btn) return this;
    if (this._buttons.length && (this.maxRowWidth === null || force || this._buttons[this._buttons.length - 1].length < this.maxRowWidth)) {
      this._buttons[this._buttons.length - 1].push(btn);
    } else {
      this._buttons.push([btn]);
    }
    return this;
  }
  /**
   * Return contents of this builder as an inline keyboard
   */
  asInline() {
    return {
      type: "inline",
      buttons: this._buttons
    };
  }
  /**
   * Return contents of this builder as a reply keyboard
   */
  asReply(params = {}) {
    const ret = params;
    ret.type = "reply";
    ret.buttons = this._buttons;
    return ret;
  }
}
exports.BotKeyboardBuilder = BotKeyboardBuilder;
