// @mtcute/core/highlevel/types/bots/keyboards/factories.js
import { utf8 } from "@fuman/utils";
import { assertNever } from "../../../../types/utils.js";
import { toInputUser } from "../../../utils/peer-utils.js";
import { BotKeyboardBuilder } from "./builder.js";
import { isAnyReplyMarkup } from "../../../../tl/inner-tl.js";
function builder(maxRowWidth) {
  return new BotKeyboardBuilder(maxRowWidth);
}
function inline(buttons) {
  return {
    type: "inline",
    buttons
  };
}
function reply(buttons, params = {}) {
  const ret = params;
  ret.type = "reply";
  ret.buttons = buttons;
  return ret;
}
function hideReply(selective) {
  return {
    type: "reply_hide",
    selective
  };
}
function forceReply(params = {}) {
  const ret = params;
  ret.type = "force_reply";
  return ret;
}
function text(text2) {
  return {
    _: "keyboardButton",
    text: text2
  };
}
function requestContact(text2) {
  return {
    _: "keyboardButtonRequestPhone",
    text: text2
  };
}
function requestGeo(text2) {
  return {
    _: "keyboardButtonRequestGeoLocation",
    text: text2
  };
}
function requestPoll(text2, quiz) {
  return {
    _: "keyboardButtonRequestPoll",
    text: text2,
    quiz
  };
}
function url(text2, url2, options) {
  return {
    _: "keyboardButtonUrl",
    text: text2,
    url: url2,
    style: options?.style
  };
}
function callback(text2, data, options) {
  return {
    _: "keyboardButtonCallback",
    text: text2,
    requiresPassword: options?.requiresPassword,
    style: options?.style,
    data: typeof data === "string" ? utf8.encoder.encode(data) : data
  };
}
function switchInline(text2, options) {
  return {
    _: "keyboardButtonSwitchInline",
    samePeer: options?.currentChat,
    style: options?.style,
    text: text2,
    query: options?.query ?? ""
  };
}
function game(text2, options) {
  return { _: "keyboardButtonGame", text: text2, style: options?.style };
}
function pay(text2, options) {
  return { _: "keyboardButtonBuy", text: text2, style: options?.style };
}
function urlAuth(text2, url2, params = {}) {
  return {
    _: "inputKeyboardButtonUrlAuth",
    text: text2,
    url: url2,
    bot: params.bot ?? {
      _: "inputUserSelf"
    },
    fwdText: params.fwdText,
    requestWriteAccess: params.requestWriteAccess,
    style: params.style
  };
}
function webView(text2, url2, options) {
  return {
    _: "keyboardButtonWebView",
    text: text2,
    url: url2,
    style: options?.style
  };
}
function userProfile(text2, user, options) {
  return {
    _: "inputKeyboardButtonUserProfile",
    text: text2,
    userId: toInputUser(user),
    style: options?.style
  };
}
function requestPeer(text2, buttonId, params) {
  return {
    _: "keyboardButtonRequestPeer",
    text: text2,
    buttonId,
    peerType: params.peerType,
    maxQuantity: params.count ?? 1,
    style: params.style
  };
}
function copy(params) {
  return {
    _: "keyboardButtonCopy",
    text: params.text,
    copyText: params?.copyText ?? params.text,
    style: params.style
  };
}
function findButton(buttons, predicate) {
  if (typeof predicate === "string") {
    const text2 = predicate;
    predicate = (btn) => {
      return "text" in btn && btn.text === text2;
    };
  }
  for (const row of buttons) {
    for (const btn of row) {
      if (predicate(btn)) {
        return btn;
      }
    }
  }
  return null;
}
function _rowsTo2d(rows) {
  return rows.map((it) => it.buttons);
}
function _2dToRows(arr, inline2) {
  return arr.map((row) => {
    if (!inline2) {
      row = row.map(
        (btn) => btn._ === "keyboardButtonWebView" ? {
          ...btn,
          _: "keyboardButtonSimpleWebView"
        } : btn
      );
    }
    return {
      _: "keyboardButtonRow",
      buttons: row
    };
  });
}
function _convertToTl(obj) {
  if (!obj) return obj;
  if (isAnyReplyMarkup(obj)) return obj;
  switch (obj.type) {
    case "reply":
      return {
        _: "replyKeyboardMarkup",
        resize: obj.resize,
        singleUse: obj.singleUse,
        selective: obj.selective,
        persistent: obj.persistent,
        placeholder: obj.placeholder,
        rows: _2dToRows(obj.buttons, false)
      };
    case "reply_hide":
      return {
        _: "replyKeyboardHide",
        selective: obj.selective
      };
    case "force_reply":
      return {
        _: "replyKeyboardForceReply",
        singleUse: obj.singleUse,
        selective: obj.selective,
        placeholder: obj.placeholder
      };
    case "inline":
      return {
        _: "replyInlineMarkup",
        rows: _2dToRows(obj.buttons, true)
      };
    default:
      assertNever();
  }
}
export {
  _2dToRows,
  _convertToTl,
  _rowsTo2d,
  builder,
  callback,
  copy,
  findButton,
  forceReply,
  game,
  hideReply,
  inline,
  pay,
  reply,
  requestContact,
  requestGeo,
  requestPeer,
  requestPoll,
  switchInline,
  text,
  url,
  urlAuth,
  userProfile,
  webView
};
