// @mtcute/core/highlevel/methods/auth/send-code.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const sentCode = require("../../types/auth/sent-code.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const user = require("../../types/peers/user.cjs");
async function sendCode(client, params) {
  const phone = miscUtils.normalizePhoneNumber(params.phone);
  const { id, hash } = await client.getApiCredentials();
  const res = await client.call(
    {
      _: "auth.sendCode",
      phoneNumber: phone,
      apiId: id,
      apiHash: hash,
      settings: {
        _: "codeSettings",
        logoutTokens: params.futureAuthTokens,
        ...params.codeSettings
      }
    },
    { abortSignal: params.abortSignal }
  );
  if (res._ === "auth.sentCodeSuccess") {
    return new user.User(await client.notifyLoggedIn(res.authorization));
  }
  if (res._ === "auth.sentCodePaymentRequired") {
    throw new errors.MtArgumentError("Payment is required to sign in, please log in with a first-party client first");
  }
  return new sentCode.SentCode(res);
}
exports.sendCode = sendCode;
