// @mtcute/core/highlevel/methods/auth/send-code.js
import { MtArgumentError } from "../../../types/errors.js";
import { SentCode as SentCode$1 } from "../../types/auth/sent-code.js";
import { normalizePhoneNumber } from "../../utils/misc-utils.js";
import { User as User$2 } from "../../types/peers/user.js";
async function sendCode(client, params) {
  const phone = normalizePhoneNumber(params.phone);
  const { id, hash } = await client.getApiCredentials();
  const res = await client.call(
    {
      _: "auth.sendCode",
      phoneNumber: phone,
      apiId: id,
      apiHash: hash,
      settings: {
        _: "codeSettings",
        logoutTokens: params.futureAuthTokens,
        ...params.codeSettings
      }
    },
    { abortSignal: params.abortSignal }
  );
  if (res._ === "auth.sentCodeSuccess") {
    return new User$2(await client.notifyLoggedIn(res.authorization));
  }
  if (res._ === "auth.sentCodePaymentRequired") {
    throw new MtArgumentError("Payment is required to sign in, please log in with a first-party client first");
  }
  return new SentCode$1(res);
}
export {
  sendCode
};
