// @mtcute/core/highlevel/methods/auth/sign-in-qr.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const miscUtils = require("../../../utils/misc-utils.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const miscUtils$1 = require("../../utils/misc-utils.cjs");
const checkPassword = require("./check-password.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
const user = require("../../types/peers/user.cjs");
async function signInQr(client, params) {
  const { onUrlUpdated, abortSignal, onQrScanned } = params;
  let waiter;
  const onUpdate = (update) => {
    if (update._ === "updateShort" && update.update._ === "updateLoginToken") {
      onQrScanned?.();
      waiter?.resolve();
      client.onServerUpdate.remove(onUpdate);
    }
  };
  client.onServerUpdate.add(onUpdate);
  abortSignal?.addEventListener("abort", () => {
    client.onServerUpdate.remove(onUpdate);
    waiter?.reject(abortSignal.reason);
  });
  async function handle2fa(input) {
    const isDynamic = typeof input === "function";
    while (true) {
      const password = await miscUtils$1.resolveMaybeDynamic(input);
      try {
        return await checkPassword.checkPassword(client, {
          password,
          abortSignal
        });
      } catch (e) {
        if (innerTl.RpcError.is(e, "PASSWORD_HASH_INVALID")) {
          if (!isDynamic) {
            throw e;
          }
          if (params.invalidPasswordCallback) {
            await params.invalidPasswordCallback?.();
          } else {
            console.log("Invalid password. Please try again");
          }
          continue;
        }
        throw e;
      }
    }
  }
  try {
    const { id, hash } = await client.getApiCredentials();
    loop: while (true) {
      let res;
      try {
        res = await client.call(
          {
            _: "auth.exportLoginToken",
            apiId: id,
            apiHash: hash,
            exceptIds: []
          },
          { abortSignal }
        );
      } catch (e) {
        if (innerTl.RpcError.is(e, "SESSION_PASSWORD_NEEDED") && params.password) {
          return await handle2fa(params.password);
        }
        throw e;
      }
      switch (res._) {
        case "auth.loginToken":
          onUrlUpdated(
            `tg://login?token=${utils.base64.encode(res.token, true)}`,
            new Date(res.expires * 1e3)
          );
          waiter = new utils.Deferred();
          await Promise.race([
            waiter.promise,
            miscUtils.sleepWithAbort(res.expires * 1e3 - Date.now(), client.stopSignal)
          ]);
          break;
        case "auth.loginTokenMigrateTo": {
          await client.changePrimaryDc(res.dcId);
          let res2;
          try {
            res2 = await client.call(
              {
                _: "auth.importLoginToken",
                token: res.token
              },
              { abortSignal }
            );
          } catch (e) {
            if (innerTl.RpcError.is(e, "SESSION_PASSWORD_NEEDED") && params.password) {
              return await handle2fa(params.password);
            }
            throw e;
          }
          typeAssertions.assertTypeIs("auth.importLoginToken", res2, "auth.loginTokenSuccess");
          break loop;
        }
        case "auth.loginTokenSuccess":
          break loop;
      }
    }
    const [self] = await client.call(
      {
        _: "users.getUsers",
        id: [{ _: "inputUserSelf" }]
      },
      { abortSignal }
    );
    typeAssertions.assertTypeIs("users.getUsers", self, "user");
    await client.notifyLoggedIn(self);
    return new user.User(self);
  } finally {
    client.onServerUpdate.remove(onUpdate);
  }
}
exports.signInQr = signInQr;
