// @mtcute/core/highlevel/methods/auth/sign-in-qr.js
import { base64, Deferred } from "@fuman/utils";
import { sleepWithAbort } from "../../../utils/misc-utils.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { resolveMaybeDynamic } from "../../utils/misc-utils.js";
import { checkPassword } from "./check-password.js";
import { RpcError } from "../../../tl/inner-tl.js";
import { User as User$2 } from "../../types/peers/user.js";
async function signInQr(client, params) {
  const { onUrlUpdated, abortSignal, onQrScanned } = params;
  let waiter;
  const onUpdate = (update) => {
    if (update._ === "updateShort" && update.update._ === "updateLoginToken") {
      onQrScanned?.();
      waiter?.resolve();
      client.onServerUpdate.remove(onUpdate);
    }
  };
  client.onServerUpdate.add(onUpdate);
  abortSignal?.addEventListener("abort", () => {
    client.onServerUpdate.remove(onUpdate);
    waiter?.reject(abortSignal.reason);
  });
  async function handle2fa(input) {
    const isDynamic = typeof input === "function";
    while (true) {
      const password = await resolveMaybeDynamic(input);
      try {
        return await checkPassword(client, {
          password,
          abortSignal
        });
      } catch (e) {
        if (RpcError.is(e, "PASSWORD_HASH_INVALID")) {
          if (!isDynamic) {
            throw e;
          }
          if (params.invalidPasswordCallback) {
            await params.invalidPasswordCallback?.();
          } else {
            console.log("Invalid password. Please try again");
          }
          continue;
        }
        throw e;
      }
    }
  }
  try {
    const { id, hash } = await client.getApiCredentials();
    loop: while (true) {
      let res;
      try {
        res = await client.call(
          {
            _: "auth.exportLoginToken",
            apiId: id,
            apiHash: hash,
            exceptIds: []
          },
          { abortSignal }
        );
      } catch (e) {
        if (RpcError.is(e, "SESSION_PASSWORD_NEEDED") && params.password) {
          return await handle2fa(params.password);
        }
        throw e;
      }
      switch (res._) {
        case "auth.loginToken":
          onUrlUpdated(
            `tg://login?token=${base64.encode(res.token, true)}`,
            new Date(res.expires * 1e3)
          );
          waiter = new Deferred();
          await Promise.race([
            waiter.promise,
            sleepWithAbort(res.expires * 1e3 - Date.now(), client.stopSignal)
          ]);
          break;
        case "auth.loginTokenMigrateTo": {
          await client.changePrimaryDc(res.dcId);
          let res2;
          try {
            res2 = await client.call(
              {
                _: "auth.importLoginToken",
                token: res.token
              },
              { abortSignal }
            );
          } catch (e) {
            if (RpcError.is(e, "SESSION_PASSWORD_NEEDED") && params.password) {
              return await handle2fa(params.password);
            }
            throw e;
          }
          assertTypeIs("auth.importLoginToken", res2, "auth.loginTokenSuccess");
          break loop;
        }
        case "auth.loginTokenSuccess":
          break loop;
      }
    }
    const [self] = await client.call(
      {
        _: "users.getUsers",
        id: [{ _: "inputUserSelf" }]
      },
      { abortSignal }
    );
    assertTypeIs("users.getUsers", self, "user");
    await client.notifyLoggedIn(self);
    return new User$2(self);
  } finally {
    client.onServerUpdate.remove(onUpdate);
  }
}
export {
  signInQr
};
