// @mtcute/core/highlevel/methods/auth/start.js
import { MtArgumentError, MtcuteError } from "../../../types/errors.js";
import { User as User$2 } from "../../types/peers/user.js";
import { resolveMaybeDynamic, normalizePhoneNumber } from "../../utils/misc-utils.js";
import { getMe } from "../users/get-me.js";
import { checkPassword } from "./check-password.js";
import { logOut } from "./log-out.js";
import { resendCode } from "./resend-code.js";
import { sendCode } from "./send-code.js";
import { signInBot } from "./sign-in-bot.js";
import { signInQr } from "./sign-in-qr.js";
import { signIn } from "./sign-in.js";
import { RpcError } from "../../../tl/inner-tl.js";
async function start(client, params) {
  if (params.session) {
    await client.importSession(params.session, params.sessionForce);
  }
  const { abortSignal } = params;
  let has2fa = false;
  let sentCode;
  let phone = null;
  try {
    const me = await getMe(client);
    client.log.info("Logged in as %s (ID: %s, username: %s, bot: %s)", me.displayName, me.id, me.username, me.isBot);
    await client.notifyLoggedIn(me.raw);
    return me;
  } catch (e) {
    if (RpcError.is(e)) {
      if (e.text === "SESSION_PASSWORD_NEEDED") {
        has2fa = true;
      } else if (e.text === "SESSION_REVOKED" || e.text === "USER_DEACTIVATED" || e.text === "USER_DEACTIVATED_BAN") {
        await logOut(client).catch((err) => {
          client.log.warn("failed to log out: %e", err);
        });
      } else if (e.text !== "AUTH_KEY_UNREGISTERED") {
        throw e;
      }
    } else {
      throw e;
    }
  }
  if (!has2fa && !params.qrCodeHandler) {
    if (!params.phone && !params.botToken) {
      throw new MtArgumentError("Neither phone nor bot token were provided");
    }
    phone = params.phone ? await resolveMaybeDynamic(params.phone) : null;
    if (phone) {
      phone = normalizePhoneNumber(phone);
      if (!params.code) {
        throw new MtArgumentError("You must pass `code` to use `phone`");
      }
    } else {
      const botToken = params.botToken ? await resolveMaybeDynamic(params.botToken) : null;
      if (!botToken) {
        throw new MtArgumentError("Either bot token or phone number must be provided");
      }
      return signInBot(client, botToken);
    }
    try {
      const res = await sendCode(client, {
        phone,
        futureAuthTokens: params.futureAuthTokens,
        codeSettings: params.codeSettings,
        abortSignal
      });
      if (res instanceof User$2) {
        return res;
      }
      sentCode = res;
    } catch (e) {
      if (RpcError.is(e, "SESSION_PASSWORD_NEEDED")) {
        has2fa = true;
      } else {
        throw e;
      }
    }
  }
  if (sentCode) {
    if (params.forceSms && (sentCode.type === "app" || sentCode.type === "email")) {
      sentCode = await resendCode(client, {
        phone,
        phoneCodeHash: sentCode.phoneCodeHash,
        abortSignal
      });
    }
    if (params.codeSentCallback) {
      await params.codeSentCallback(sentCode);
    } else {
      if (sentCode.type === "email_required") {
        throw new MtcuteError("Email login setup is required to sign in");
      }
      console.log(`The confirmation code has been sent via ${sentCode.type}.`);
    }
    for (; ; ) {
      const code = await resolveMaybeDynamic(params.code);
      if (!code) throw new RpcError(400, "PHONE_CODE_EMPTY");
      try {
        return await signIn(client, {
          phone,
          phoneCodeHash: sentCode.phoneCodeHash,
          phoneCode: code,
          abortSignal
        });
      } catch (e) {
        if (!RpcError.is(e)) throw e;
        if (e.is("SESSION_PASSWORD_NEEDED")) {
          has2fa = true;
          break;
        } else if (e.is("PHONE_CODE_EMPTY") || e.is("PHONE_CODE_EXPIRED") || e.is("PHONE_CODE_INVALID") || e.is("PHONE_CODE_HASH_EMPTY")) {
          if (typeof params.code !== "function") {
            throw new MtArgumentError("Provided code was invalid");
          }
          if (params.invalidCodeCallback) {
            await params.invalidCodeCallback("code");
          } else {
            console.log("Invalid code. Please try again");
          }
          continue;
        } else {
          throw e;
        }
      }
      break;
    }
  }
  if (has2fa) {
    if (!params.password) {
      throw new MtArgumentError("2FA is enabled, but `password` was not provided.");
    }
    for (; ; ) {
      const password = await resolveMaybeDynamic(params.password);
      try {
        return await checkPassword(client, {
          password,
          abortSignal
        });
      } catch (e) {
        if (typeof params.password !== "function") {
          throw new MtArgumentError("Provided password was invalid");
        }
        if (RpcError.is(e, "PASSWORD_HASH_INVALID")) {
          if (params.invalidCodeCallback) {
            await params.invalidCodeCallback("password");
          } else {
            console.log("Invalid password. Please try again");
          }
          continue;
        } else {
          throw e;
        }
      }
    }
  }
  if (params.qrCodeHandler) {
    return signInQr(client, {
      onUrlUpdated: params.qrCodeHandler,
      password: params.password,
      invalidPasswordCallback: params.invalidCodeCallback ? () => params.invalidCodeCallback("password") : void 0,
      abortSignal
    });
  }
  throw new MtArgumentError("Failed to log in with provided credentials");
}
export {
  start
};
