// @mtcute/core/highlevel/methods/auth/start.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const user = require("../../types/peers/user.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const getMe = require("../users/get-me.cjs");
const checkPassword = require("./check-password.cjs");
const logOut = require("./log-out.cjs");
const resendCode = require("./resend-code.cjs");
const sendCode = require("./send-code.cjs");
const signInBot = require("./sign-in-bot.cjs");
const signInQr = require("./sign-in-qr.cjs");
const signIn = require("./sign-in.cjs");
const innerTl = require("../../../tl/inner-tl.cjs");
async function start(client, params) {
  if (params.session) {
    await client.importSession(params.session, params.sessionForce);
  }
  const { abortSignal } = params;
  let has2fa = false;
  let sentCode;
  let phone = null;
  try {
    const me = await getMe.getMe(client);
    client.log.info("Logged in as %s (ID: %s, username: %s, bot: %s)", me.displayName, me.id, me.username, me.isBot);
    await client.notifyLoggedIn(me.raw);
    return me;
  } catch (e) {
    if (innerTl.RpcError.is(e)) {
      if (e.text === "SESSION_PASSWORD_NEEDED") {
        has2fa = true;
      } else if (e.text === "SESSION_REVOKED" || e.text === "USER_DEACTIVATED" || e.text === "USER_DEACTIVATED_BAN") {
        await logOut.logOut(client).catch((err) => {
          client.log.warn("failed to log out: %e", err);
        });
      } else if (e.text !== "AUTH_KEY_UNREGISTERED") {
        throw e;
      }
    } else {
      throw e;
    }
  }
  if (!has2fa && !params.qrCodeHandler) {
    if (!params.phone && !params.botToken) {
      throw new errors.MtArgumentError("Neither phone nor bot token were provided");
    }
    phone = params.phone ? await miscUtils.resolveMaybeDynamic(params.phone) : null;
    if (phone) {
      phone = miscUtils.normalizePhoneNumber(phone);
      if (!params.code) {
        throw new errors.MtArgumentError("You must pass `code` to use `phone`");
      }
    } else {
      const botToken = params.botToken ? await miscUtils.resolveMaybeDynamic(params.botToken) : null;
      if (!botToken) {
        throw new errors.MtArgumentError("Either bot token or phone number must be provided");
      }
      return signInBot.signInBot(client, botToken);
    }
    try {
      const res = await sendCode.sendCode(client, {
        phone,
        futureAuthTokens: params.futureAuthTokens,
        codeSettings: params.codeSettings,
        abortSignal
      });
      if (res instanceof user.User) {
        return res;
      }
      sentCode = res;
    } catch (e) {
      if (innerTl.RpcError.is(e, "SESSION_PASSWORD_NEEDED")) {
        has2fa = true;
      } else {
        throw e;
      }
    }
  }
  if (sentCode) {
    if (params.forceSms && (sentCode.type === "app" || sentCode.type === "email")) {
      sentCode = await resendCode.resendCode(client, {
        phone,
        phoneCodeHash: sentCode.phoneCodeHash,
        abortSignal
      });
    }
    if (params.codeSentCallback) {
      await params.codeSentCallback(sentCode);
    } else {
      if (sentCode.type === "email_required") {
        throw new errors.MtcuteError("Email login setup is required to sign in");
      }
      console.log(`The confirmation code has been sent via ${sentCode.type}.`);
    }
    for (; ; ) {
      const code = await miscUtils.resolveMaybeDynamic(params.code);
      if (!code) throw new innerTl.RpcError(400, "PHONE_CODE_EMPTY");
      try {
        return await signIn.signIn(client, {
          phone,
          phoneCodeHash: sentCode.phoneCodeHash,
          phoneCode: code,
          abortSignal
        });
      } catch (e) {
        if (!innerTl.RpcError.is(e)) throw e;
        if (e.is("SESSION_PASSWORD_NEEDED")) {
          has2fa = true;
          break;
        } else if (e.is("PHONE_CODE_EMPTY") || e.is("PHONE_CODE_EXPIRED") || e.is("PHONE_CODE_INVALID") || e.is("PHONE_CODE_HASH_EMPTY")) {
          if (typeof params.code !== "function") {
            throw new errors.MtArgumentError("Provided code was invalid");
          }
          if (params.invalidCodeCallback) {
            await params.invalidCodeCallback("code");
          } else {
            console.log("Invalid code. Please try again");
          }
          continue;
        } else {
          throw e;
        }
      }
      break;
    }
  }
  if (has2fa) {
    if (!params.password) {
      throw new errors.MtArgumentError("2FA is enabled, but `password` was not provided.");
    }
    for (; ; ) {
      const password = await miscUtils.resolveMaybeDynamic(params.password);
      try {
        return await checkPassword.checkPassword(client, {
          password,
          abortSignal
        });
      } catch (e) {
        if (typeof params.password !== "function") {
          throw new errors.MtArgumentError("Provided password was invalid");
        }
        if (innerTl.RpcError.is(e, "PASSWORD_HASH_INVALID")) {
          if (params.invalidCodeCallback) {
            await params.invalidCodeCallback("password");
          } else {
            console.log("Invalid password. Please try again");
          }
          continue;
        } else {
          throw e;
        }
      }
    }
  }
  if (params.qrCodeHandler) {
    return signInQr.signInQr(client, {
      onUrlUpdated: params.qrCodeHandler,
      password: params.password,
      invalidPasswordCallback: params.invalidCodeCallback ? () => params.invalidCodeCallback("password") : void 0,
      abortSignal
    });
  }
  throw new errors.MtArgumentError("Failed to log in with provided credentials");
}
exports.start = start;
