// @mtcute/core/highlevel/types/stories/all-stories.js
import { memoizeGetters } from "../../utils/memoize.js";
import { PeerStories as PeerStories$2 } from "./peer-stories.js";
import { StoriesStealthMode as StoriesStealthMode$1 } from "./stealth-mode.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { PeersIndex } from "../peers/peers-index.js";
class AllStories {
  constructor(raw) {
    this.raw = raw;
    this._peers = PeersIndex.from(this.raw);
  }
  /** Peers index */
  _peers;
  /** Whether there are more stories to fetch */
  get hasMore() {
    return this.raw.hasMore;
  }
  /** Next offset for pagination */
  get next() {
    return this.raw.state;
  }
  /** Total number of {@link PeerStories} available */
  get total() {
    return this.raw.count;
  }
  /** Peers with their stories */
  get peerStories() {
    return this.raw.peerStories.map((it) => new PeerStories$2(it, this._peers));
  }
  /** Stealth mode info */
  get stealthMode() {
    return new StoriesStealthMode$1(this.raw.stealthMode);
  }
}
const AllStories$1 = /* @__PURE__ */ memoizeGetters(AllStories, ["peerStories", "stealthMode"]);
const AllStories$2 = /* @__PURE__ */ makeInspectable(AllStories$1);
export {
  AllStories$2 as AllStories
};
