// @mtcute/core/highlevel/types/stories/all-stories.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const peerStories = require("./peer-stories.cjs");
const stealthMode = require("./stealth-mode.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const peersIndex = require("../peers/peers-index.cjs");
class AllStories {
  constructor(raw) {
    this.raw = raw;
    this._peers = peersIndex.PeersIndex.from(this.raw);
  }
  /** Peers index */
  _peers;
  /** Whether there are more stories to fetch */
  get hasMore() {
    return this.raw.hasMore;
  }
  /** Next offset for pagination */
  get next() {
    return this.raw.state;
  }
  /** Total number of {@link PeerStories} available */
  get total() {
    return this.raw.count;
  }
  /** Peers with their stories */
  get peerStories() {
    return this.raw.peerStories.map((it) => new peerStories.PeerStories(it, this._peers));
  }
  /** Stealth mode info */
  get stealthMode() {
    return new stealthMode.StoriesStealthMode(this.raw.stealthMode);
  }
}
const AllStories$1 = /* @__PURE__ */ memoize.memoizeGetters(AllStories, ["peerStories", "stealthMode"]);
const AllStories$2 = /* @__PURE__ */ inspectable.makeInspectable(AllStories$1);
exports.AllStories = AllStories$2;
