// @mtcute/core/highlevel/types/stories/story.js
import { MtUnsupportedError } from "../../../types/errors.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parseDocument } from "../media/document-utils.js";
import { Photo as Photo$2 } from "../media/photo.js";
import { MessageEntity as MessageEntity$2 } from "../messages/message-entity.js";
import { _storyInteractiveElementFromTl } from "./interactive/index.js";
import { StoryInteractions as StoryInteractions$2 } from "./story-interactions.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { toReactionEmoji } from "../reactions/types.js";
class Story {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this story is pinned */
  get isPinned() {
    return this.raw.pinned;
  }
  /**
   * Whether this object contains reduced set of fields.
   *
   * When `true`, these field will not contain correct data:
   * {@link privacyRules}, {@link interactiveAreas}
   *
   */
  get isShort() {
    return this.raw.min;
  }
  /** Whether this story is content-protected, i.e. can't be forwarded */
  get isContentProtected() {
    return this.raw.noforwards;
  }
  /** Whether this story has been edited */
  get isEdited() {
    return this.raw.edited;
  }
  /** Whether this story has been posted by the current user */
  get isMy() {
    return this.raw.out;
  }
  /** ID of the story */
  get id() {
    return this.raw.id;
  }
  /** Date when this story was posted */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Date when this story will expire */
  get expireDate() {
    return new Date(this.raw.expireDate * 1e3);
  }
  /** Whether the story is active (i.e. not expired yet) */
  get isActive() {
    return Date.now() < this.expireDate.getTime();
  }
  /** Story visibility */
  get visibility() {
    if (this.raw.public) return "public";
    if (this.raw.contacts) return "contacts";
    if (this.raw.closeFriends) return "close_friends";
    if (this.raw.selectedContacts) return "selected_contacts";
    throw new MtUnsupportedError("Unknown story visibility");
  }
  /** Caption of the story */
  get caption() {
    return this.raw.caption ?? null;
  }
  /**
   * Caption entities (may be empty)
   */
  get entities() {
    const entities = [];
    if (this.raw.entities?.length) {
      for (const ent of this.raw.entities) {
        entities.push(new MessageEntity$2(ent, this.raw.caption));
      }
    }
    return entities;
  }
  /**
   * Story media.
   *
   * Currently, can only be {@link Photo} or {@link Video}.
   */
  get media() {
    switch (this.raw.media._) {
      case "messageMediaPhoto":
        if (this.raw.media.photo?._ !== "photo") throw new MtUnsupportedError("Unsupported story media type");
        return new Photo$2(this.raw.media.photo, this.raw.media);
      case "messageMediaDocument": {
        if (this.raw.media.document?._ !== "document") {
          throw new MtUnsupportedError("Unsupported story media type");
        }
        const doc = parseDocument(this.raw.media.document, this.raw.media);
        if (doc.type === "video") return doc;
      }
    }
    throw new MtUnsupportedError("Unsupported story media type");
  }
  /**
   * Interactive elements of the story
   */
  get interactiveElements() {
    if (!this.raw.mediaAreas) return [];
    return this.raw.mediaAreas.map((it) => _storyInteractiveElementFromTl(it, this._peers));
  }
  /**
   * Privacy rules of the story.
   *
   * Only available when {@link isMy} is `true`.
   */
  get privacyRules() {
    if (!this.raw.privacy) return null;
    return this.raw.privacy;
  }
  /**
   * Information about story interactions
   */
  get interactions() {
    if (!this.raw.views) return null;
    return new StoryInteractions$2(this.raw.views, this._peers);
  }
  /**
   * Emoji representing a reaction sent by the current user, if any
   */
  get sentReactionEmoji() {
    if (!this.raw.sentReaction) return null;
    return toReactionEmoji(this.raw.sentReaction, true);
  }
  /**
   * Albums this story belongs to
   */
  get albums() {
    return this.raw.albums ?? [];
  }
}
const Story$1 = /* @__PURE__ */ memoizeGetters(Story, ["entities", "media", "interactiveElements", "interactions"]);
const Story$2 = /* @__PURE__ */ makeInspectable(Story$1);
export {
  Story$2 as Story
};
