// @mtcute/core/highlevel/types/stories/story-viewer.js
import { memoizeGetters } from "../../utils/memoize.js";
import { Story as Story$2 } from "./story.js";
import { parsePeer } from "../peers/peer.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { Message as Message$2 } from "../messages/message.js";
import { toReactionEmoji } from "../reactions/types.js";
import { User as User$2 } from "../peers/user.js";
import { PeersIndex } from "../peers/peers-index.js";
class StoryViewer {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this user is in current user's global blacklist */
  get isBlocked() {
    return this.raw.blocked;
  }
  /** Whether current user's stories are hidden from this user */
  get isStoriesBlocked() {
    return this.raw.blockedMyStoriesFrom;
  }
  /** Date when the view has occurred */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Reaction this user has left, if any */
  get reactionEmoji() {
    if (!this.raw.reaction) return null;
    return toReactionEmoji(this.raw.reaction, true);
  }
  /** Information about the user */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
}
const StoryViewer$1 = /* @__PURE__ */ memoizeGetters(StoryViewer, ["user"]);
const StoryViewer$2 = /* @__PURE__ */ makeInspectable(StoryViewer$1);
class StoryRepost {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this peer is in current user's global blacklist */
  get isBlocked() {
    return this.raw.blocked;
  }
  /** Whether current user's stories are hidden from this peer */
  get isStoriesBlocked() {
    return this.raw.blockedMyStoriesFrom;
  }
  /** Kind of the repost */
  get kind() {
    return this.raw._ === "storyViewPublicForward" ? "forward" : "repost";
  }
  /** Date when the repost has happened, if available */
  get date() {
    if (this.raw._ === "storyViewPublicForward" && this.raw.message._ === "message") {
      return new Date(this.raw.message.date * 1e3);
    }
    if (this.raw._ === "storyViewPublicRepost" && this.raw.story._ === "storyItem") {
      return new Date(this.raw.story.date * 1e3);
    }
    return null;
  }
  /**
   * Message that has been forwarded/reposted.
   *
   * Only available if {@link kind} is `forward`.
   */
  get message() {
    if (this.raw._ === "storyViewPublicRepost") return null;
    return new Message$2(this.raw.message, this._peers);
  }
  /**
   * Story that has been reposted.
   *
   * Only available if {@link kind} is `repost`.
   */
  get story() {
    if (this.raw._ === "storyViewPublicForward") return null;
    if (this.raw.story._ !== "storyItem") return null;
    return new Story$2(this.raw.story, this._peers);
  }
  /** Information about the peer who has made the reposted */
  get peer() {
    if (this.raw._ === "storyViewPublicForward") {
      return this.message.sender;
    }
    return parsePeer(this.raw.peerId, this._peers);
  }
}
class StoryViewersList {
  constructor(raw) {
    this.raw = raw;
    this._peers = PeersIndex.from(this.raw);
  }
  _peers;
  /** Next offset for pagination */
  get next() {
    return this.raw.nextOffset;
  }
  /** Total number of views this story has */
  get total() {
    return this.raw.count;
  }
  /** Total number of reactions this story has */
  get reactionsTotal() {
    return this.raw.reactionsCount;
  }
  /** List of viewers */
  get viewers() {
    const res = [];
    for (const view of this.raw.views) {
      if (view._ === "storyView") {
        res.push(new StoryViewer(view, this._peers));
      }
    }
    return res;
  }
}
const StoryViewersList$1 = /* @__PURE__ */ memoizeGetters(StoryViewersList, ["viewers"]);
const StoryViewersList$2 = /* @__PURE__ */ makeInspectable(StoryViewersList$1);
export {
  StoryRepost,
  StoryViewer$2 as StoryViewer,
  StoryViewersList$2 as StoryViewersList
};
