// @mtcute/core/highlevel/types/stories/story-interactions.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const user = require("../peers/user.cjs");
const reactionCount = require("../reactions/reaction-count.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class StoryInteractions {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether information about viewers is available.
   *
   * When `true`, you can use {@link TelegarmClient.getStoryViewers}
   * to get the full list of viewers, and also {@link recentViewers}
   * will be available.
   */
  get hasViewers() {
    return this.raw.hasViewers;
  }
  /** Number of views */
  get viewsCount() {
    return this.raw.viewsCount;
  }
  /** Number of forwards (if available) */
  get forwardsCount() {
    return this.raw.forwardsCount ?? null;
  }
  /** Total number of reactions */
  get reactionsCount() {
    return this.raw.reactionsCount ?? 0;
  }
  /**
   * Reactions on the message, along with their counts
   */
  get reactions() {
    if (!this.raw.reactions) return [];
    return this.raw.reactions.map((it) => new reactionCount.ReactionCount(it));
  }
  /**
   * List of users who have recently viewed this story.
   */
  get recentViewers() {
    return this.raw.recentViewers?.map((it) => new user.User(this._peers.user(it))) ?? [];
  }
}
const StoryInteractions$1 = /* @__PURE__ */ memoize.memoizeGetters(StoryInteractions, ["reactions", "recentViewers"]);
const StoryInteractions$2 = /* @__PURE__ */ inspectable.makeInspectable(StoryInteractions$1);
exports.StoryInteractions = StoryInteractions$2;
