// @mtcute/core/highlevel/types/stories/story-interactions.js
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "../peers/user.js";
import { ReactionCount as ReactionCount$1 } from "../reactions/reaction-count.js";
import { makeInspectable } from "../../utils/inspectable.js";
class StoryInteractions {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Whether information about viewers is available.
   *
   * When `true`, you can use {@link TelegarmClient.getStoryViewers}
   * to get the full list of viewers, and also {@link recentViewers}
   * will be available.
   */
  get hasViewers() {
    return this.raw.hasViewers;
  }
  /** Number of views */
  get viewsCount() {
    return this.raw.viewsCount;
  }
  /** Number of forwards (if available) */
  get forwardsCount() {
    return this.raw.forwardsCount ?? null;
  }
  /** Total number of reactions */
  get reactionsCount() {
    return this.raw.reactionsCount ?? 0;
  }
  /**
   * Reactions on the message, along with their counts
   */
  get reactions() {
    if (!this.raw.reactions) return [];
    return this.raw.reactions.map((it) => new ReactionCount$1(it));
  }
  /**
   * List of users who have recently viewed this story.
   */
  get recentViewers() {
    return this.raw.recentViewers?.map((it) => new User$2(this._peers.user(it))) ?? [];
  }
}
const StoryInteractions$1 = /* @__PURE__ */ memoizeGetters(StoryInteractions, ["reactions", "recentViewers"]);
const StoryInteractions$2 = /* @__PURE__ */ makeInspectable(StoryInteractions$1);
export {
  StoryInteractions$2 as StoryInteractions
};
