// @mtcute/core/highlevel/types/stories/story-viewer.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const memoize = require("../../utils/memoize.cjs");
const story = require("./story.cjs");
const peer = require("../peers/peer.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const message = require("../messages/message.cjs");
const types = require("../reactions/types.cjs");
const user = require("../peers/user.cjs");
const peersIndex = require("../peers/peers-index.cjs");
class StoryViewer {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this user is in current user's global blacklist */
  get isBlocked() {
    return this.raw.blocked;
  }
  /** Whether current user's stories are hidden from this user */
  get isStoriesBlocked() {
    return this.raw.blockedMyStoriesFrom;
  }
  /** Date when the view has occurred */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Reaction this user has left, if any */
  get reactionEmoji() {
    if (!this.raw.reaction) return null;
    return types.toReactionEmoji(this.raw.reaction, true);
  }
  /** Information about the user */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
}
const StoryViewer$1 = /* @__PURE__ */ memoize.memoizeGetters(StoryViewer, ["user"]);
const StoryViewer$2 = /* @__PURE__ */ inspectable.makeInspectable(StoryViewer$1);
class StoryRepost {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether this peer is in current user's global blacklist */
  get isBlocked() {
    return this.raw.blocked;
  }
  /** Whether current user's stories are hidden from this peer */
  get isStoriesBlocked() {
    return this.raw.blockedMyStoriesFrom;
  }
  /** Kind of the repost */
  get kind() {
    return this.raw._ === "storyViewPublicForward" ? "forward" : "repost";
  }
  /** Date when the repost has happened, if available */
  get date() {
    if (this.raw._ === "storyViewPublicForward" && this.raw.message._ === "message") {
      return new Date(this.raw.message.date * 1e3);
    }
    if (this.raw._ === "storyViewPublicRepost" && this.raw.story._ === "storyItem") {
      return new Date(this.raw.story.date * 1e3);
    }
    return null;
  }
  /**
   * Message that has been forwarded/reposted.
   *
   * Only available if {@link kind} is `forward`.
   */
  get message() {
    if (this.raw._ === "storyViewPublicRepost") return null;
    return new message.Message(this.raw.message, this._peers);
  }
  /**
   * Story that has been reposted.
   *
   * Only available if {@link kind} is `repost`.
   */
  get story() {
    if (this.raw._ === "storyViewPublicForward") return null;
    if (this.raw.story._ !== "storyItem") return null;
    return new story.Story(this.raw.story, this._peers);
  }
  /** Information about the peer who has made the reposted */
  get peer() {
    if (this.raw._ === "storyViewPublicForward") {
      return this.message.sender;
    }
    return peer.parsePeer(this.raw.peerId, this._peers);
  }
}
class StoryViewersList {
  constructor(raw) {
    this.raw = raw;
    this._peers = peersIndex.PeersIndex.from(this.raw);
  }
  _peers;
  /** Next offset for pagination */
  get next() {
    return this.raw.nextOffset;
  }
  /** Total number of views this story has */
  get total() {
    return this.raw.count;
  }
  /** Total number of reactions this story has */
  get reactionsTotal() {
    return this.raw.reactionsCount;
  }
  /** List of viewers */
  get viewers() {
    const res = [];
    for (const view of this.raw.views) {
      if (view._ === "storyView") {
        res.push(new StoryViewer(view, this._peers));
      }
    }
    return res;
  }
}
const StoryViewersList$1 = /* @__PURE__ */ memoize.memoizeGetters(StoryViewersList, ["viewers"]);
const StoryViewersList$2 = /* @__PURE__ */ inspectable.makeInspectable(StoryViewersList$1);
exports.StoryRepost = StoryRepost;
exports.StoryViewer = StoryViewer$2;
exports.StoryViewersList = StoryViewersList$2;
