// @mtcute/core/storage/driver.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class BaseStorageDriver {
  _loadedTimes = 0;
  _destroyed = false;
  _log;
  _platform;
  setup(log, platform) {
    this._log = log.create("sqlite");
    this._platform = platform;
  }
  get loaded() {
    return this._loadedTimes > 0;
  }
  async load() {
    if (this._loadedTimes === 0) {
      await this._load();
      this._destroyed = false;
    }
    this._loadedTimes++;
  }
  async destroy() {
    if (this._destroyed) {
      return;
    }
    this._loadedTimes--;
    if (this._loadedTimes === 0) {
      await this._destroy();
      this._destroyed = true;
    }
  }
  save() {
    return this._save?.();
  }
}
exports.BaseStorageDriver = BaseStorageDriver;
