// @mtcute/core/storage/storage.js
import { asyncResettable } from "../utils/function-utils.js";
import "@mtcute/tl-runtime";
import { AuthKeysService } from "./service/auth-keys.js";
import { DefaultDcsService } from "./service/default-dcs.js";
import { FutureSaltsService } from "./service/future-salts.js";
class StorageManager {
  constructor(options) {
    this.options = options;
    this.provider = this.options.provider;
    this.platform = this.options.platform;
    this.driver = this.provider.driver;
    this.log = this.options.log.create("storage");
    const serviceOptions = {
      driver: this.driver,
      readerMap: this.options.readerMap,
      writerMap: this.options.writerMap,
      log: this.log
    };
    this.dcs = new DefaultDcsService(this.provider.kv, serviceOptions);
    this.salts = new FutureSaltsService(this.provider.kv, serviceOptions);
    this.keys = new AuthKeysService(this.provider.authKeys, this.salts, serviceOptions);
  }
  provider;
  driver;
  platform;
  log;
  dcs;
  salts;
  keys;
  _cleanupRestore;
  _load = asyncResettable(async () => {
    this.driver.setup?.(this.log, this.platform);
    if (this.options.cleanup ?? true) {
      this._cleanupRestore = this.platform.beforeExit(() => {
        this._destroy().catch((err) => this.log.error("cleanup error: %e", err));
      });
    }
    await this.driver.load?.();
  });
  load() {
    return this._load.run();
  }
  async save() {
    await this.driver.save?.();
  }
  async clear(withAuthKeys = false) {
    if (withAuthKeys) {
      await this.provider.authKeys.deleteAll();
    }
    await this.provider.kv.deleteAll();
    await this.save();
  }
  async _destroy() {
    if (!this._load.finished()) return;
    await this._load.wait();
    await this.driver.destroy?.();
    this._load.reset();
  }
  async destroy() {
    if (this._cleanupRestore) {
      this._cleanupRestore();
      this._cleanupRestore = void 0;
    }
    await this._destroy();
  }
}
export {
  StorageManager
};
