// @mtcute/core/storage/storage.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const functionUtils = require("../utils/function-utils.cjs");
require("@mtcute/tl-runtime");
const authKeys = require("./service/auth-keys.cjs");
const defaultDcs = require("./service/default-dcs.cjs");
const futureSalts = require("./service/future-salts.cjs");
class StorageManager {
  constructor(options) {
    this.options = options;
    this.provider = this.options.provider;
    this.platform = this.options.platform;
    this.driver = this.provider.driver;
    this.log = this.options.log.create("storage");
    const serviceOptions = {
      driver: this.driver,
      readerMap: this.options.readerMap,
      writerMap: this.options.writerMap,
      log: this.log
    };
    this.dcs = new defaultDcs.DefaultDcsService(this.provider.kv, serviceOptions);
    this.salts = new futureSalts.FutureSaltsService(this.provider.kv, serviceOptions);
    this.keys = new authKeys.AuthKeysService(this.provider.authKeys, this.salts, serviceOptions);
  }
  provider;
  driver;
  platform;
  log;
  dcs;
  salts;
  keys;
  _cleanupRestore;
  _load = functionUtils.asyncResettable(async () => {
    this.driver.setup?.(this.log, this.platform);
    if (this.options.cleanup ?? true) {
      this._cleanupRestore = this.platform.beforeExit(() => {
        this._destroy().catch((err) => this.log.error("cleanup error: %e", err));
      });
    }
    await this.driver.load?.();
  });
  load() {
    return this._load.run();
  }
  async save() {
    await this.driver.save?.();
  }
  async clear(withAuthKeys = false) {
    if (withAuthKeys) {
      await this.provider.authKeys.deleteAll();
    }
    await this.provider.kv.deleteAll();
    await this.save();
  }
  async _destroy() {
    if (!this._load.finished()) return;
    await this._load.wait();
    await this.driver.destroy?.();
    this._load.reset();
  }
  async destroy() {
    if (this._cleanupRestore) {
      this._cleanupRestore();
      this._cleanupRestore = void 0;
    }
    await this._destroy();
  }
}
exports.StorageManager = StorageManager;
