// @mtcute/core/highlevel/managers/app-config-manager.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("../../types/errors.cjs");
const tlJson = require("../../utils/tl-json.cjs");
class AppConfigManager {
  constructor(client) {
    this.client = client;
    this._resource = new utils.AsyncResource({
      fetcher: async ({ current }) => {
        const res = await this.client.call({
          _: "help.getAppConfig",
          hash: current?.hash ?? 0
        });
        if (res._ === "help.appConfigNotModified") {
          return {
            data: utils.asNonNull(current),
            expiresIn: 36e5
          };
        }
        return {
          data: res,
          expiresIn: 36e5
        };
      }
    });
  }
  _resource;
  _object;
  async get() {
    if (!this._resource.isStale && this._object) return this._object;
    const obj = tlJson.tlJsonToJson((await this._resource.get()).config);
    if (!obj || typeof obj !== "object") {
      throw new errors.MtTypeAssertionError("appConfig", "object", typeof obj);
    }
    this._object = obj;
    return this._object;
  }
  async getField(field, fallback) {
    const obj = await this.get();
    return obj[field] ?? fallback;
  }
}
exports.AppConfigManager = AppConfigManager;
