// @mtcute/core/highlevel/managers/app-config-manager.js
import { AsyncResource, asNonNull } from "@fuman/utils";
import { MtTypeAssertionError } from "../../types/errors.js";
import { tlJsonToJson } from "../../utils/tl-json.js";
class AppConfigManager {
  constructor(client) {
    this.client = client;
    this._resource = new AsyncResource({
      fetcher: async ({ current }) => {
        const res = await this.client.call({
          _: "help.getAppConfig",
          hash: current?.hash ?? 0
        });
        if (res._ === "help.appConfigNotModified") {
          return {
            data: asNonNull(current),
            expiresIn: 36e5
          };
        }
        return {
          data: res,
          expiresIn: 36e5
        };
      }
    });
  }
  _resource;
  _object;
  async get() {
    if (!this._resource.isStale && this._object) return this._object;
    const obj = tlJsonToJson((await this._resource.get()).config);
    if (!obj || typeof obj !== "object") {
      throw new MtTypeAssertionError("appConfig", "object", typeof obj);
    }
    this._object = obj;
    return this._object;
  }
  async getField(field, fallback) {
    const obj = await this.get();
    return obj[field] ?? fallback;
  }
}
export {
  AppConfigManager
};
