// @mtcute/core/highlevel/managers/timers.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
class TimersManager {
  constructor(_executeRpc) {
    this._executeRpc = _executeRpc;
  }
  _timers = /* @__PURE__ */ new Map();
  _errorHandler;
  // These methods are async for API parity with worker-backed clients,
  // where keepalive state is owned by the shared client/worker and accessed over RPC.
  async exists(key) {
    return this._timers.has(key);
  }
  async upsert(spec) {
    this.upsertOwned("local", spec);
  }
  async cancel(key) {
    this._cancel(key);
  }
  upsertOwned(ownerId, spec) {
    this._cancel(spec.key);
    const timer = new utils.AsyncInterval(
      (abortSignal) => this._executeRpc(spec, abortSignal),
      spec.interval
    );
    if (this._errorHandler) {
      timer.onError(this._errorHandler);
    }
    this._timers.set(spec.key, {
      ownerId,
      timer
    });
    if (spec.startNow) timer.startNow();
    else timer.start();
  }
  clearOwner(ownerId) {
    for (const [key, timer] of Array.from(this._timers.entries())) {
      if (timer.ownerId !== ownerId) continue;
      this._cancel(key);
    }
  }
  onError(handler) {
    this._errorHandler = handler;
  }
  destroy() {
    for (const key of Array.from(this._timers.keys())) {
      this._cancel(key);
    }
  }
  _cancel(key) {
    const timer = this._timers.get(key);
    if (!timer) return;
    timer.timer.stop();
    this._timers.delete(key);
  }
}
exports.TimersManager = TimersManager;
