// @mtcute/core/highlevel/managers/timers.js
import { AsyncInterval } from "@fuman/utils";
class TimersManager {
  constructor(_executeRpc) {
    this._executeRpc = _executeRpc;
  }
  _timers = /* @__PURE__ */ new Map();
  _errorHandler;
  // These methods are async for API parity with worker-backed clients,
  // where keepalive state is owned by the shared client/worker and accessed over RPC.
  async exists(key) {
    return this._timers.has(key);
  }
  async upsert(spec) {
    this.upsertOwned("local", spec);
  }
  async cancel(key) {
    this._cancel(key);
  }
  upsertOwned(ownerId, spec) {
    this._cancel(spec.key);
    const timer = new AsyncInterval(
      (abortSignal) => this._executeRpc(spec, abortSignal),
      spec.interval
    );
    if (this._errorHandler) {
      timer.onError(this._errorHandler);
    }
    this._timers.set(spec.key, {
      ownerId,
      timer
    });
    if (spec.startNow) timer.startNow();
    else timer.start();
  }
  clearOwner(ownerId) {
    for (const [key, timer] of Array.from(this._timers.entries())) {
      if (timer.ownerId !== ownerId) continue;
      this._cancel(key);
    }
  }
  onError(handler) {
    this._errorHandler = handler;
  }
  destroy() {
    for (const key of Array.from(this._timers.keys())) {
      this._cancel(key);
    }
  }
  _cancel(key) {
    const timer = this._timers.get(key);
    if (!timer) return;
    timer.timer.stop();
    this._timers.delete(key);
  }
}
export {
  TimersManager
};
