// @mtcute/markdown-parser/index.js
import { dateEntityFormatToString, parseDateEntityFormat } from "@mtcute/core/utils.js";
import Long from "long";
const MENTION_REGEX = /^tg:\/\/user\?id=(\d+)(?:&hash=(-?[0-9a-fA-F]+)(?:&|$)|&|$)/;
const EMOJI_REGEX = /^tg:\/\/emoji\?id=(-?\d+)/;
const TIME_REGEX = /^tg:\/\/time\?unix=(\d+)(?:&format=([^&)]+))?/;
const TAG_BOLD = "**";
const TAG_ITALIC = "__";
const TAG_UNDERLINE = "--";
const TAG_STRIKE = "~~";
const TAG_SPOILER = "||";
const TAG_CODE = "`";
const TAG_PRE = "```";
const TO_BE_ESCAPED = /[*_\-~`[\\\]|]/g;
function escape(str) {
  return str.replace(TO_BE_ESCAPED, (s) => `\\${s}`);
}
function unparse(input) {
  if (typeof input === "string") return escape(input);
  let text = input.text;
  const entities = input.entities ?? [];
  const escaped = [];
  text = text.replace(TO_BE_ESCAPED, (s, pos) => {
    escaped.push(pos);
    return `\\${s}`;
  });
  const hasEscaped = escaped.length > 0;
  const insert = [];
  for (const entity of entities) {
    const type = entity._;
    let start = entity.offset;
    let end = start + entity.length;
    if (start > text.length) continue;
    if (start < 0) start = 0;
    if (end > text.length) end = text.length;
    if (hasEscaped) {
      let escapedPos = 0;
      while (escapedPos < escaped.length && escaped[escapedPos] < start) {
        escapedPos += 1;
      }
      start += escapedPos;
      while (escapedPos < escaped.length && escaped[escapedPos] <= end) {
        escapedPos += 1;
      }
      end += escapedPos;
    }
    let startTag;
    let endTag;
    switch (type) {
      case "messageEntityBold":
        startTag = endTag = TAG_BOLD;
        break;
      case "messageEntityItalic":
        startTag = endTag = TAG_ITALIC;
        break;
      case "messageEntityUnderline":
        startTag = endTag = TAG_UNDERLINE;
        break;
      case "messageEntityStrike":
        startTag = endTag = TAG_STRIKE;
        break;
      case "messageEntitySpoiler":
        startTag = endTag = TAG_SPOILER;
        break;
      case "messageEntityCode":
        startTag = endTag = TAG_CODE;
        break;
      case "messageEntityPre":
        startTag = TAG_PRE;
        if (entity.language) {
          startTag += entity.language;
        }
        startTag += "\n";
        endTag = `
${TAG_PRE}`;
        break;
      case "messageEntityTextUrl":
        startTag = "[";
        endTag = `](${entity.url})`;
        break;
      case "messageEntityMentionName":
        startTag = "[";
        endTag = `](tg://user?id=${entity.userId})`;
        break;
      case "messageEntityCustomEmoji":
        startTag = "[";
        endTag = `](tg://emoji?id=${entity.documentId.toString()})`;
        break;
      case "messageEntityFormattedDate": {
        const fmt = dateEntityFormatToString(entity);
        startTag = "[";
        endTag = `](tg://time?unix=${entity.date}${fmt ? `&format=${fmt}` : ""})`;
        break;
      }
      default:
        continue;
    }
    insert.push([start, startTag]);
    insert.push([end, endTag]);
  }
  insert.sort((a, b) => b[0] - a[0]);
  for (const [offset, tag] of insert) {
    text = text.substr(0, offset) + tag + text.substr(offset);
  }
  return text;
}
function parse(strings, ...sub) {
  const entities = [];
  let result = "";
  const stacks = {};
  let insideCode = false;
  let insidePre = false;
  let insideLink = false;
  let currentBlockquoteStart = null;
  let prevBlockquote = null;
  let insideLinkUrl = false;
  let pendingLinkUrl = "";
  function feed(text) {
    const len = text.length;
    let pos = 0;
    while (pos < len) {
      const c = text[pos];
      if (c === "\\") {
        if (insideLinkUrl) {
          pendingLinkUrl += text[pos + 1];
        } else {
          result += text[pos + 1];
        }
        pos += 2;
        continue;
      }
      if (insideCode) {
        if (c === "`") {
          const ent = stacks.code.pop();
          ent.length = result.length - ent.offset;
          entities.push(ent);
          insideCode = false;
          pos += 1;
        } else {
          pos += 1;
          result += c;
        }
        continue;
      }
      if (insidePre) {
        if (c === "`" || c === "\n" && text[pos + 1] === "`") {
          if (c === "\n") pos += 1;
          if (text[pos + 1] === "`" && text[pos + 2] === "`") {
            const ent = stacks.pre.pop();
            ent.length = result.length - ent.offset;
            entities.push(ent);
            insidePre = false;
            pos += 3;
            continue;
          } else if (c === "\n") {
            pos -= 1;
          }
        }
        pos += 1;
        result += c;
        continue;
      }
      if (insideLink && c === "]") {
        const ent = stacks.link.pop();
        if (text[pos + 1] !== "(") {
          result = `${result.substr(0, ent.offset)}[${result.substr(ent.offset)}]`;
          pos += 1;
          insideLink = false;
          continue;
        }
        pos += 2;
        insideLink = false;
        insideLinkUrl = true;
        stacks.link.push(ent);
        continue;
      }
      if (insideLinkUrl) {
        pos += 1;
        if (c !== ")") {
          pendingLinkUrl += c;
          continue;
        }
        const ent = stacks.link.pop();
        let url = pendingLinkUrl;
        pendingLinkUrl = "";
        insideLinkUrl = false;
        if (!url.length) continue;
        ent.length = result.length - ent.offset;
        let m = url.match(MENTION_REGEX);
        if (m) {
          const userId = Number.parseInt(m[1]);
          const accessHash = m[2];
          if (accessHash) {
            ent._ = "inputMessageEntityMentionName";
            ent.userId = {
              _: "inputUser",
              userId,
              accessHash: Long.fromString(accessHash, false, 16)
            };
          } else {
            ent._ = "messageEntityMentionName";
            ent.userId = userId;
          }
        } else if (m = EMOJI_REGEX.exec(url)) {
          ent._ = "messageEntityCustomEmoji";
          ent.documentId = Long.fromString(m[1]);
        } else if (m = TIME_REGEX.exec(url)) {
          const flags = parseDateEntityFormat(m[2] ?? "");
          ent._ = "messageEntityFormattedDate";
          ent.date = Number.parseInt(m[1]);
          Object.assign(ent, flags);
        } else {
          if (url.match(/^\/\//)) url = `http:${url}`;
          ent._ = "messageEntityTextUrl";
          ent.url = url;
        }
        entities.push(ent);
        continue;
      }
      if (c === "[" && !insideLink) {
        pos += 1;
        insideLink = true;
        if (!("link" in stacks)) stacks.link = [];
        stacks.link.push({
          offset: result.length,
          length: 0
        });
        continue;
      }
      if (c === "`") {
        const isPre = text[pos + 1] === "`" && text[pos + 2] === "`";
        if (isPre) {
          pos += 3;
          let language = "";
          while (pos < text.length && text[pos] !== "\n") {
            language += text[pos++];
          }
          pos += 1;
          if (pos > text.length) {
            result += "```";
            feed(language);
          } else {
            if (!("pre" in stacks)) stacks.pre = [];
            stacks.pre.push({
              _: "messageEntityPre",
              offset: result.length,
              length: 0,
              language
            });
            insidePre = true;
          }
        } else {
          pos += 1;
          if (!("code" in stacks)) stacks.code = [];
          stacks.code.push({
            _: "messageEntityCode",
            offset: result.length,
            length: 0
          });
          insideCode = true;
        }
        continue;
      }
      if (c === text[pos + 1]) {
        let type = null;
        switch (c) {
          case "_":
            type = "Italic";
            break;
          case "*":
            type = "Bold";
            break;
          case "-":
            type = "Underline";
            break;
          case "~":
            type = "Strike";
            break;
          case "|":
            type = "Spoiler";
            break;
        }
        if (type) {
          if (!(type in stacks)) stacks[type] = [];
          const isBegin = stacks[type].length === 0;
          if (isBegin) {
            stacks[type].push({
              _: `messageEntity${type}`,
              offset: result.length,
              length: 0
            });
          } else {
            const ent = stacks[type].pop();
            ent.length = result.length - ent.offset;
            entities.push(ent);
          }
          pos += 2;
          continue;
        }
      }
      if (c === "\n") {
        if (result.length !== 0) {
          result += "\n";
        }
        const nonWhitespace = text.slice(pos + 1).search(/[^ \t]/);
        if (nonWhitespace !== -1) {
          pos += nonWhitespace + 1;
        } else {
          pos = len;
        }
        if (currentBlockquoteStart != null) {
          const prevEnd = prevBlockquote ? prevBlockquote.offset + prevBlockquote.length : Number.NaN;
          if (currentBlockquoteStart - 1 === prevEnd) {
            prevBlockquote.length += result.length - currentBlockquoteStart;
          } else {
            if (prevBlockquote) entities.push(prevBlockquote);
            prevBlockquote = {
              _: "messageEntityBlockquote",
              offset: currentBlockquoteStart,
              length: result.length - currentBlockquoteStart - 1
            };
          }
          currentBlockquoteStart = null;
        }
        continue;
      }
      if (c === ">" && (result.length === 0 || result[result.length - 1] === "\n")) {
        currentBlockquoteStart = result.length;
        pos += 1;
        continue;
      }
      if (c === " " && currentBlockquoteStart === result.length) {
        pos += 1;
        continue;
      }
      result += c;
      pos += 1;
    }
  }
  if (typeof strings === "string") strings = [strings];
  sub.forEach((it, idx) => {
    feed(strings[idx]);
    if (typeof it === "boolean" || !it) return;
    if (insideLinkUrl) {
      if (typeof it === "string" || typeof it === "number") {
        pendingLinkUrl += it;
      } else if (Long.isLong(it)) {
        pendingLinkUrl += it.toString(10);
      } else {
        pendingLinkUrl += it.text;
      }
      return;
    }
    if (typeof it === "string" || typeof it === "number") {
      result += it;
    } else if (Long.isLong(it)) {
      result += it.toString(10);
    } else {
      const text = it.text;
      const innerEntities = "raw" in it ? [it.raw] : it.entities;
      const baseOffset = result.length;
      result += text;
      if (innerEntities) {
        for (const ent of innerEntities) {
          entities.push({ ...ent, offset: ent.offset + baseOffset });
        }
      }
    }
  });
  feed(strings[strings.length - 1]);
  if (prevBlockquote) entities.push(prevBlockquote);
  function adjustOffsets(from, by) {
    for (const ent of entities) {
      if (ent.offset < from) continue;
      if (by >= 0) {
        ent.offset += by;
        continue;
      }
      const adjustTotal = Math.min(ent.offset, Math.abs(by));
      const adjustInternal = Math.max(Math.abs(by) - ent.offset, 0);
      ent.offset -= adjustTotal;
      ent.length -= adjustInternal;
    }
    for (const stack of Object.values(stacks)) {
      for (const ent of stack) {
        if (ent.offset >= from) ent.offset += by;
      }
    }
  }
  for (const [name, stack] of Object.entries(stacks)) {
    if (stack.length > 1) {
      throw new Error(`Malformed ${name} entity`);
    }
    if (stack.length) {
      switch (name) {
        case "link": {
          const startOffset = stack.pop().offset;
          insideLink = false;
          insideLinkUrl = false;
          const url = pendingLinkUrl;
          pendingLinkUrl = "";
          result = `${result.substring(0, startOffset)}[${result.substring(startOffset)}](`;
          adjustOffsets(startOffset, 1);
          feed(url);
          break;
        }
        default: {
          const startOffset = stack.pop().offset;
          const tag = {
            Bold: TAG_BOLD,
            Italic: TAG_ITALIC,
            Underline: TAG_UNDERLINE,
            Strike: TAG_STRIKE,
            Spoiler: TAG_SPOILER,
            code: TAG_CODE
          }[name];
          if (!tag) throw new Error(`invalid tag ${name}`);
          const remaining = result.substring(startOffset);
          result = `${result.substring(0, startOffset)}${tag}`;
          adjustOffsets(startOffset, tag.length);
          feed(remaining);
          break;
        }
      }
    }
  }
  const resultTrimmed = result.trimStart();
  const numCharsTrimmed = result.length - resultTrimmed.length;
  result = resultTrimmed;
  adjustOffsets(0, -numCharsTrimmed);
  result = result.trimEnd();
  const finalEntities = [];
  for (const ent of entities) {
    const end = ent.offset + ent.length;
    if (end > result.length) ent.length = result.length - ent.offset;
    if (ent.length > 0) finalEntities.push(ent);
  }
  return {
    text: result,
    entities: finalEntities
  };
}
const md = Object.assign(parse, {
  escape,
  unparse
});
export {
  md
};
