// @fuman/utils/arrays/typed/find.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
function indexOf(haystack, needle, start = 0) {
  for (let i = start; i < haystack.length; i++) {
    if (haystack[i] === needle) {
      return i;
    }
  }
  return -1;
}
function lastIndexOf(haystack, needle, start = haystack.length - 1) {
  for (let i = start; i >= 0; i--) {
    if (haystack[i] === needle) {
      return i;
    }
  }
  return -1;
}
function indexOfArray(haystack, needle, start = 0) {
  if (needle.length === 0) {
    return start;
  }
  if (needle.length === 1) {
    return indexOf(haystack, needle[0], start);
  }
  const max = haystack.length - needle.length;
  for (let i = start; i <= max; i++) {
    if (haystack[i] === needle[0]) {
      let j = 1;
      for (; j < needle.length; j++) {
        if (haystack[i + j] !== needle[j]) {
          break;
        }
      }
      if (j === needle.length) {
        return i;
      }
    }
  }
  return -1;
}
function lastIndexOfArray(haystack, needle, start = haystack.length - 1) {
  if (needle.length === 0) {
    return start;
  }
  if (needle.length === 1) {
    return lastIndexOf(haystack, needle[0], start);
  }
  const min = needle.length - 1;
  for (let i = start; i >= min; i--) {
    if (haystack[i] === needle[needle.length - 1]) {
      let j = needle.length - 2;
      for (; j >= 0; j--) {
        if (haystack[i - needle.length + j + 1] !== needle[j]) {
          break;
        }
      }
      if (j < 0) {
        return i - needle.length + 1;
      }
    }
  }
  return -1;
}
function includes(haystack, needle) {
  return indexOf(haystack, needle) !== -1;
}
function includesArray(haystack, needle) {
  return indexOfArray(haystack, needle) !== -1;
}
exports.includes = includes;
exports.includesArray = includesArray;
exports.indexOf = indexOf;
exports.indexOfArray = indexOfArray;
exports.lastIndexOf = lastIndexOf;
exports.lastIndexOfArray = lastIndexOfArray;
