// @fuman/utils/misc/assert.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
function assert(condition, message) {
  if (!condition) {
    throw new Error(typeof message == "string" ? message : "Assertion failed", {
      cause: message
    });
  }
}
function assertHasKey(obj, key) {
  if (!(key in obj)) {
    throw new Error(`key ${JSON.stringify(key)} not found in object`, {
      cause: { obj, key }
    });
  }
}
function unsafeCastType(value) {
}
function assertNonNull(value) {
  if (value === null || value === void 0) {
    throw new Error(`value is ${value}`);
  }
}
function asNonNull(value) {
  assertNonNull(value);
  return value;
}
function assertMatches(str, regex) {
  const match = str.match(regex);
  if (!match) {
    throw new Error(`${JSON.stringify(str)} does not match ${regex}`, {
      cause: { regex, str }
    });
  }
  return match;
}
exports.asNonNull = asNonNull;
exports.assert = assert;
exports.assertHasKey = assertHasKey;
exports.assertMatches = assertMatches;
exports.assertNonNull = assertNonNull;
exports.unsafeCastType = unsafeCastType;
