// @fuman/utils/misc/objects.js
function objectKeys(obj) {
  return Object.keys(obj);
}
function objectEntries(obj) {
  return Object.entries(obj);
}
function clearUndefined(obj) {
  for (const key in obj) {
    if (!Object.hasOwn(obj, key)) continue;
    if (obj[key] === void 0) {
      delete obj[key];
    }
  }
  return obj;
}
function deepMerge(into, from, options) {
  if (!Array.isArray(from)) from = [from];
  if (from.length === 0) return into;
  const {
    undefined: onUndefined = "ignore",
    properties: onProperties = "replace",
    arrays: onArrays = "replace",
    objects: onObjects = "merge"
  } = options ?? {};
  for (let i = 0; i < from.length; i++) {
    const fromItem = from[i];
    for (const key in fromItem) {
      if (!Object.hasOwn(fromItem, key)) continue;
      const value = fromItem[key];
      const existing = into[key];
      if (value === void 0) {
        if (onUndefined === "replace") {
          into[key] = void 0;
        }
        continue;
      }
      if (Array.isArray(value)) {
        if (onArrays === "merge") {
          if (Array.isArray(existing)) {
            for (let j = 0; j < value.length; j++) {
              existing.push(value[j]);
            }
          } else {
            into[key] = value;
          }
          continue;
        }
        if (onArrays === "ignore" && existing !== void 0) continue;
        into[key] = value;
        continue;
      }
      if (typeof value === "object" && value !== null) {
        if (onObjects === "merge") {
          if (typeof existing === "object" && existing !== null) {
            into[key] = deepMerge(existing, value);
            continue;
          } else {
            into[key] = value;
          }
          continue;
        }
        if (onObjects === "ignore" && existing !== void 0) continue;
        into[key] = value;
        continue;
      }
      if (onProperties === "ignore" && existing !== void 0) continue;
      into[key] = value;
    }
  }
  return into;
}
export {
  clearUndefined,
  deepMerge,
  objectEntries,
  objectKeys
};
