// @fuman/net/proxy/socks/_protocol.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const parse = require("../../ip/parse.cjs");
const v6 = require("../../ip/v6.cjs");
const SOCKS4_ERRORS = {
  91: "Request rejected or failed",
  92: "Request failed because client is not running identd",
  93: "Request failed because client's identd could not confirm the user ID in the request"
};
const SOCKS5_ERRORS = {
  1: "General failure",
  2: "Connection not allowed by ruleset",
  3: "Network unreachable",
  4: "Host unreachable",
  5: "Connection refused by destination host",
  6: "TTL expired",
  7: "Command not supported / protocol error",
  8: "Address type not supported"
};
function buildSocks4Connect(dest, username = "") {
  let addr;
  try {
    addr = parse.parse(dest.address);
  } catch {
  }
  let isHostname = false;
  if (!addr) {
    isHostname = true;
    addr = { type: "ipv4", parts: utils.u8.allocWith([0, 0, 0, 42]) };
  } else if (addr.type !== "ipv4") {
    throw new TypeError("SOCKS4 only supports IPv4");
  }
  const hostnameLength = isHostname ? utils.utf8.encodedLength(dest.address) + 1 : 0;
  const userIdLength = utils.utf8.encodedLength(username);
  const buf = io.Bytes.alloc(8 + userIdLength + hostnameLength);
  io.write.uint8(buf, 4);
  io.write.uint8(buf, 1);
  io.write.uint16be(buf, dest.port);
  io.write.bytes(buf, addr.parts);
  io.write.cUtf8String(buf, username);
  if (isHostname) {
    io.write.cUtf8String(buf, dest.address);
  }
  return buf.result();
}
function buildSocks5Greeting(authAvailable) {
  const buf = utils.u8.alloc(authAvailable ? 4 : 3);
  buf[0] = 5;
  if (authAvailable) {
    buf[1] = 2;
    buf[2] = 0;
    buf[3] = 2;
  } else {
    buf[1] = 1;
    buf[2] = 0;
  }
  return buf;
}
function buildSocks5Auth(username, password) {
  const usernameLen = utils.utf8.encodedLength(username);
  const passwordLen = utils.utf8.encodedLength(password);
  if (usernameLen > 255) {
    throw new TypeError(`Too long username (${usernameLen} > 255)`);
  }
  if (passwordLen > 255) {
    throw new TypeError(`Too long password (${passwordLen} > 255)`);
  }
  const buf = utils.u8.alloc(3 + usernameLen + passwordLen);
  buf[0] = 1;
  buf[1] = usernameLen;
  utils.utf8.encoder.encodeInto(username, buf.subarray(2));
  buf[2 + usernameLen] = passwordLen;
  utils.utf8.encoder.encodeInto(password, buf.subarray(3 + usernameLen));
  return buf;
}
function buildSocks5Connect(dest) {
  let addr;
  try {
    addr = parse.parse(dest.address);
  } catch {
  }
  const buf = io.Bytes.alloc(32);
  io.write.uint8(buf, 5);
  io.write.uint8(buf, 1);
  io.write.uint8(buf, 0);
  if (!addr) {
    const addrLen = utils.utf8.encodedLength(dest.address);
    if (addrLen > 255) {
      throw new TypeError(`Too long hostname (${addrLen} > 255)`);
    }
    io.write.uint8(buf, 3);
    io.write.uint8(buf, addrLen);
    io.write.utf8String(buf, dest.address);
  } else if (addr.type === "ipv6") {
    io.write.uint8(buf, 4);
    v6.writeV6(addr, buf);
  } else {
    io.write.uint8(buf, 1);
    io.write.bytes(buf, addr.parts);
  }
  io.write.uint16be(buf, dest.port);
  return buf.result();
}
exports.SOCKS4_ERRORS = SOCKS4_ERRORS;
exports.SOCKS5_ERRORS = SOCKS5_ERRORS;
exports.buildSocks4Connect = buildSocks4Connect;
exports.buildSocks5Auth = buildSocks5Auth;
exports.buildSocks5Connect = buildSocks5Connect;
exports.buildSocks5Greeting = buildSocks5Greeting;
