// @fuman/net/proxy/socks/_protocol.js
import { Bytes, write } from "@fuman/io";
import { u8, utf8 } from "@fuman/utils";
import { parse } from "../../ip/parse.js";
import { writeV6 } from "../../ip/v6.js";
const SOCKS4_ERRORS = {
  91: "Request rejected or failed",
  92: "Request failed because client is not running identd",
  93: "Request failed because client's identd could not confirm the user ID in the request"
};
const SOCKS5_ERRORS = {
  1: "General failure",
  2: "Connection not allowed by ruleset",
  3: "Network unreachable",
  4: "Host unreachable",
  5: "Connection refused by destination host",
  6: "TTL expired",
  7: "Command not supported / protocol error",
  8: "Address type not supported"
};
function buildSocks4Connect(dest, username = "") {
  let addr;
  try {
    addr = parse(dest.address);
  } catch {
  }
  let isHostname = false;
  if (!addr) {
    isHostname = true;
    addr = { type: "ipv4", parts: u8.allocWith([0, 0, 0, 42]) };
  } else if (addr.type !== "ipv4") {
    throw new TypeError("SOCKS4 only supports IPv4");
  }
  const hostnameLength = isHostname ? utf8.encodedLength(dest.address) + 1 : 0;
  const userIdLength = utf8.encodedLength(username);
  const buf = Bytes.alloc(8 + userIdLength + hostnameLength);
  write.uint8(buf, 4);
  write.uint8(buf, 1);
  write.uint16be(buf, dest.port);
  write.bytes(buf, addr.parts);
  write.cUtf8String(buf, username);
  if (isHostname) {
    write.cUtf8String(buf, dest.address);
  }
  return buf.result();
}
function buildSocks5Greeting(authAvailable) {
  const buf = u8.alloc(authAvailable ? 4 : 3);
  buf[0] = 5;
  if (authAvailable) {
    buf[1] = 2;
    buf[2] = 0;
    buf[3] = 2;
  } else {
    buf[1] = 1;
    buf[2] = 0;
  }
  return buf;
}
function buildSocks5Auth(username, password) {
  const usernameLen = utf8.encodedLength(username);
  const passwordLen = utf8.encodedLength(password);
  if (usernameLen > 255) {
    throw new TypeError(`Too long username (${usernameLen} > 255)`);
  }
  if (passwordLen > 255) {
    throw new TypeError(`Too long password (${passwordLen} > 255)`);
  }
  const buf = u8.alloc(3 + usernameLen + passwordLen);
  buf[0] = 1;
  buf[1] = usernameLen;
  utf8.encoder.encodeInto(username, buf.subarray(2));
  buf[2 + usernameLen] = passwordLen;
  utf8.encoder.encodeInto(password, buf.subarray(3 + usernameLen));
  return buf;
}
function buildSocks5Connect(dest) {
  let addr;
  try {
    addr = parse(dest.address);
  } catch {
  }
  const buf = Bytes.alloc(32);
  write.uint8(buf, 5);
  write.uint8(buf, 1);
  write.uint8(buf, 0);
  if (!addr) {
    const addrLen = utf8.encodedLength(dest.address);
    if (addrLen > 255) {
      throw new TypeError(`Too long hostname (${addrLen} > 255)`);
    }
    write.uint8(buf, 3);
    write.uint8(buf, addrLen);
    write.utf8String(buf, dest.address);
  } else if (addr.type === "ipv6") {
    write.uint8(buf, 4);
    writeV6(addr, buf);
  } else {
    write.uint8(buf, 1);
    write.bytes(buf, addr.parts);
  }
  write.uint16be(buf, dest.port);
  return buf.result();
}
export {
  SOCKS4_ERRORS,
  SOCKS5_ERRORS,
  buildSocks4Connect,
  buildSocks5Auth,
  buildSocks5Connect,
  buildSocks5Greeting
};
