// @mtcute/core/highlevel/methods/forums/create-forum-topic.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const longUtils = require("../../../utils/long-utils.cjs");
const findInUpdate = require("../messages/find-in-update.cjs");
const resolvePeer = require("../users/resolve-peer.cjs");
async function createForumTopic(client, params) {
  const { chatId, title, icon, sendAs, shouldDispatch } = params;
  const res = await client.call({
    _: "messages.createForumTopic",
    peer: await resolvePeer.resolvePeer(client, chatId),
    title,
    iconColor: typeof icon === "number" ? icon : void 0,
    iconEmojiId: typeof icon !== "number" ? icon : void 0,
    sendAs: sendAs ? await resolvePeer.resolvePeer(client, sendAs) : void 0,
    randomId: longUtils.randomLong()
  });
  return findInUpdate._findMessageInUpdate(client, res, false, !shouldDispatch);
}
exports.createForumTopic = createForumTopic;
