// @mtcute/core/highlevel/methods/forums/get-forum-topics.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const resolvePeer = require("../users/resolve-peer.cjs");
const miscUtils = require("../../utils/misc-utils.cjs");
const forumTopic = require("../../types/peers/forum-topic.cjs");
const defaultOffset = {
  date: 0,
  id: 0,
  topic: 0
};
async function getForumTopics(client, chatId, params) {
  if (!params) params = {};
  const {
    query,
    offset: { date: offsetDate, id: offsetId, topic: offsetTopic } = defaultOffset,
    limit = 100
  } = params;
  const res = await client.call({
    _: "messages.getForumTopics",
    peer: await resolvePeer.resolvePeer(client, chatId),
    q: query,
    offsetDate,
    offsetId,
    offsetTopic,
    limit
  });
  const topics = forumTopic.ForumTopic.parseTlForumTopics(res);
  const last = topics[topics.length - 1];
  const next = last ? {
    date: res.orderByCreateDate ? last.raw.date : last.lastMessage.raw.date,
    id: last.raw.topMessage,
    topic: last.raw.id
  } : void 0;
  return miscUtils.makeArrayPaginated(topics, res.count, next);
}
exports.getForumTopics = getForumTopics;
