// @mtcute/core/highlevel/methods/forums/get-forum-topics.js
import { resolvePeer } from "../users/resolve-peer.js";
import { makeArrayPaginated } from "../../utils/misc-utils.js";
import { ForumTopic as ForumTopic$2 } from "../../types/peers/forum-topic.js";
const defaultOffset = {
  date: 0,
  id: 0,
  topic: 0
};
async function getForumTopics(client, chatId, params) {
  if (!params) params = {};
  const {
    query,
    offset: { date: offsetDate, id: offsetId, topic: offsetTopic } = defaultOffset,
    limit = 100
  } = params;
  const res = await client.call({
    _: "messages.getForumTopics",
    peer: await resolvePeer(client, chatId),
    q: query,
    offsetDate,
    offsetId,
    offsetTopic,
    limit
  });
  const topics = ForumTopic$2.parseTlForumTopics(res);
  const last = topics[topics.length - 1];
  const next = last ? {
    date: res.orderByCreateDate ? last.raw.date : last.lastMessage.raw.date,
    id: last.raw.topMessage,
    topic: last.raw.id
  } : void 0;
  return makeArrayPaginated(topics, res.count, next);
}
export {
  getForumTopics
};
