// @mtcute/core/network/transports/intermediate.js
import { read, write } from "@fuman/io";
import { typed } from "@fuman/utils";
import { getRandomInt } from "../../utils/misc-utils.js";
import "@mtcute/tl-runtime";
import { TransportError } from "./abstract.js";
const TAG = new Uint8Array([238, 238, 238, 238]);
const PADDED_TAG = new Uint8Array([221, 221, 221, 221]);
class IntermediatePacketCodec {
  tag() {
    return TAG;
  }
  decode(reader, eof) {
    if (eof) return null;
    if (reader.available < 8) return null;
    const length = read.uint32le(reader);
    if (length === 4) {
      const code = read.int32le(reader);
      throw new TransportError(-code);
    }
    if (reader.available < length) {
      reader.rewind(4);
      return null;
    }
    return new Uint8Array(read.exactly(reader, length));
  }
  encode(frame, into) {
    write.uint32le(into, frame.length);
    write.bytes(into, frame);
  }
  reset() {
  }
}
class PaddedIntermediatePacketCodec extends IntermediatePacketCodec {
  tag() {
    return PADDED_TAG;
  }
  _crypto;
  setup(crypto) {
    this._crypto = crypto;
  }
  encode(frame, into) {
    const padSize = getRandomInt(16);
    const ret = into.writeSync(frame.length + 4 + padSize);
    const dv = typed.toDataView(ret);
    dv.setUint32(0, frame.length + padSize, true);
    ret.set(frame, 4);
    this._crypto.randomFill(ret.subarray(4 + frame.length));
    into.disposeWriteSync();
  }
}
export {
  IntermediatePacketCodec,
  PaddedIntermediatePacketCodec
};
