// @mtcute/core/network/transports/intermediate.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const io = require("@fuman/io");
const utils = require("@fuman/utils");
const miscUtils = require("../../utils/misc-utils.cjs");
require("@mtcute/tl-runtime");
const abstract = require("./abstract.cjs");
const TAG = new Uint8Array([238, 238, 238, 238]);
const PADDED_TAG = new Uint8Array([221, 221, 221, 221]);
class IntermediatePacketCodec {
  tag() {
    return TAG;
  }
  decode(reader, eof) {
    if (eof) return null;
    if (reader.available < 8) return null;
    const length = io.read.uint32le(reader);
    if (length === 4) {
      const code = io.read.int32le(reader);
      throw new abstract.TransportError(-code);
    }
    if (reader.available < length) {
      reader.rewind(4);
      return null;
    }
    return new Uint8Array(io.read.exactly(reader, length));
  }
  encode(frame, into) {
    io.write.uint32le(into, frame.length);
    io.write.bytes(into, frame);
  }
  reset() {
  }
}
class PaddedIntermediatePacketCodec extends IntermediatePacketCodec {
  tag() {
    return PADDED_TAG;
  }
  _crypto;
  setup(crypto) {
    this._crypto = crypto;
  }
  encode(frame, into) {
    const padSize = miscUtils.getRandomInt(16);
    const ret = into.writeSync(frame.length + 4 + padSize);
    const dv = utils.typed.toDataView(ret);
    dv.setUint32(0, frame.length + padSize, true);
    ret.set(frame, 4);
    this._crypto.randomFill(ret.subarray(4 + frame.length));
    into.disposeWriteSync();
  }
}
exports.IntermediatePacketCodec = IntermediatePacketCodec;
exports.PaddedIntermediatePacketCodec = PaddedIntermediatePacketCodec;
