// @mtcute/core/network/transports/obfuscated.js
import { Bytes, write } from "@fuman/io";
import { typed, u8 } from "@fuman/utils";
class ObfuscatedPacketCodec {
  _encryptor;
  _decryptor;
  _crypto;
  _inner;
  _proxy;
  setup(crypto, log) {
    this._crypto = crypto;
    this._inner.setup?.(crypto, log);
  }
  constructor(inner, proxy) {
    this._inner = inner;
    this._proxy = proxy;
  }
  async tag() {
    let random;
    let dv;
    for (; ; ) {
      random = this._crypto.randomBytes(64);
      if (random[0] === 239) continue;
      dv = typed.toDataView(random);
      const firstInt = dv.getUint32(0, true);
      if (firstInt === 1145128264 || firstInt === 1414745936 || firstInt === 542393671 || firstInt === 1230262351 || firstInt === 3722304989 || firstInt === 4008636142 || firstInt === 33620758) {
        continue;
      }
      if (dv.getInt32(4, true) === 0) continue;
      break;
    }
    let innerTag = await this._inner.tag();
    if (innerTag.length !== 4) {
      const b = innerTag[0];
      innerTag = new Uint8Array([b, b, b, b]);
    }
    random.set(innerTag, 56);
    if (this._proxy) {
      let dcId = this._proxy.dcId;
      if (this._proxy.test) dcId += 1e4;
      if (this._proxy.media) dcId = -dcId;
      dv.setInt16(60, dcId, true);
    }
    const randomRev = u8.toReversed(random.subarray(8, 56));
    let encryptKey = random.subarray(8, 40);
    const encryptIv = random.subarray(40, 56);
    let decryptKey = randomRev.subarray(0, 32);
    const decryptIv = randomRev.subarray(32, 48);
    if (this._proxy) {
      encryptKey = this._crypto.sha256(u8.concat2(encryptKey, this._proxy.secret));
      decryptKey = this._crypto.sha256(u8.concat2(decryptKey, this._proxy.secret));
    }
    this._encryptor = this._crypto.createAesCtr(encryptKey, encryptIv, true);
    this._decryptor = this._crypto.createAesCtr(decryptKey, decryptIv, false);
    const encrypted = this._encryptor.process(random);
    random.set(encrypted.subarray(56, 64), 56);
    return random;
  }
  async encode(packet, into) {
    const temp = Bytes.alloc(packet.length);
    await this._inner.encode(packet, temp);
    write.bytes(into, this._encryptor.process(temp.result()));
  }
  _decodeBuf = Bytes.alloc();
  async decode(reader, eof) {
    if (eof) return null;
    if (reader.available > 0) {
      const into = this._decodeBuf.writeSync(reader.available);
      into.set(this._decryptor.process(reader.readSync(reader.available)));
    }
    return this._inner.decode(this._decodeBuf, eof);
  }
  reset() {
    this._inner.reset();
    this._decodeBuf.reset();
    this._encryptor?.close?.();
    this._decryptor?.close?.();
    this._encryptor = void 0;
    this._decryptor = void 0;
  }
}
export {
  ObfuscatedPacketCodec
};
