// @mtcute/core/network/transports/proxy.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../types/errors.cjs");
const proxy = require("../../utils/links/proxy.cjs");
function createProxyTransportFactory(options) {
  const { SocksProxyTcpTransport, MtProxyTcpTransport, HttpProxyTcpTransport } = options;
  return (url) => {
    const obj = new URL(url);
    if (SocksProxyTcpTransport != null && (obj.protocol === "socks4:" || obj.protocol === "socks5:")) {
      return new SocksProxyTcpTransport({
        version: obj.protocol === "socks4:" ? 4 : 5,
        host: obj.hostname,
        port: Number(obj.port) || 1080,
        user: obj.username,
        password: obj.password
      });
    }
    if (MtProxyTcpTransport != null) {
      const mtproxy = proxy.mtproxy.parse(url);
      if (mtproxy) {
        return new MtProxyTcpTransport({
          host: mtproxy.server,
          port: mtproxy.port,
          secret: mtproxy.secret
        });
      }
    }
    if (HttpProxyTcpTransport && (obj.protocol === "http:" || obj.protocol === "https:")) {
      return new HttpProxyTcpTransport({
        tls: obj.protocol === "https:",
        host: obj.hostname,
        port: Number(obj.port) || (obj.protocol === "https:" ? 443 : 80),
        user: obj.username,
        password: obj.password
      });
    } else {
      throw new errors.MtArgumentError(`Unsupported proxy protocol: ${obj.protocol}`);
    }
  };
}
exports.createProxyTransportFactory = createProxyTransportFactory;
