// @mtcute/core/network/transports/proxy.js
import { MtArgumentError } from "../../types/errors.js";
import { mtproxy } from "../../utils/links/proxy.js";
function createProxyTransportFactory(options) {
  const { SocksProxyTcpTransport, MtProxyTcpTransport, HttpProxyTcpTransport } = options;
  return (url) => {
    const obj = new URL(url);
    if (SocksProxyTcpTransport != null && (obj.protocol === "socks4:" || obj.protocol === "socks5:")) {
      return new SocksProxyTcpTransport({
        version: obj.protocol === "socks4:" ? 4 : 5,
        host: obj.hostname,
        port: Number(obj.port) || 1080,
        user: obj.username,
        password: obj.password
      });
    }
    if (MtProxyTcpTransport != null) {
      const mtproxy$1 = mtproxy.parse(url);
      if (mtproxy$1) {
        return new MtProxyTcpTransport({
          host: mtproxy$1.server,
          port: mtproxy$1.port,
          secret: mtproxy$1.secret
        });
      }
    }
    if (HttpProxyTcpTransport && (obj.protocol === "http:" || obj.protocol === "https:")) {
      return new HttpProxyTcpTransport({
        tls: obj.protocol === "https:",
        host: obj.hostname,
        port: Number(obj.port) || (obj.protocol === "https:" ? 443 : 80),
        user: obj.username,
        password: obj.password
      });
    } else {
      throw new MtArgumentError(`Unsupported proxy protocol: ${obj.protocol}`);
    }
  };
}
export {
  createProxyTransportFactory
};
