// @mtcute/core/highlevel/types/files/web-document.js
import { MtArgumentError } from "../../../types/errors.js";
import { FileLocation as FileLocation$1 } from "./file-location.js";
import { makeInspectable } from "../../utils/inspectable.js";
function STUB_LOCATION() {
  throw new MtArgumentError("This web document is not downloadable through Telegram");
}
class WebDocument extends FileLocation$1 {
  constructor(raw) {
    super(
      raw._ === "webDocument" ? {
        _: "inputWebFileLocation",
        url: raw.url,
        accessHash: raw.accessHash
      } : STUB_LOCATION,
      raw.size
    );
    this.raw = raw;
    this.raw = raw;
  }
  /**
   * URL to the file
   */
  get url() {
    return this.raw.url;
  }
  /**
   * MIME type of the file
   */
  get mimeType() {
    return this.raw.mimeType;
  }
  /**
   * Whether this file can be downloaded through Telegram.
   *
   * If `false`, you should use {@link url} to manually
   * fetch data via HTTP(s), and trying to use `download*` methods
   * will result in an error
   */
  get isDownloadable() {
    return this.raw._ === "webDocument";
  }
}
const WebDocument$1 = /* @__PURE__ */ makeInspectable(WebDocument);
export {
  WebDocument$1 as WebDocument
};
