// @mtcute/core/highlevel/types/files/web-document.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const fileLocation = require("./file-location.cjs");
const inspectable = require("../../utils/inspectable.cjs");
function STUB_LOCATION() {
  throw new errors.MtArgumentError("This web document is not downloadable through Telegram");
}
class WebDocument extends fileLocation.FileLocation {
  constructor(raw) {
    super(
      raw._ === "webDocument" ? {
        _: "inputWebFileLocation",
        url: raw.url,
        accessHash: raw.accessHash
      } : STUB_LOCATION,
      raw.size
    );
    this.raw = raw;
    this.raw = raw;
  }
  /**
   * URL to the file
   */
  get url() {
    return this.raw.url;
  }
  /**
   * MIME type of the file
   */
  get mimeType() {
    return this.raw.mimeType;
  }
  /**
   * Whether this file can be downloaded through Telegram.
   *
   * If `false`, you should use {@link url} to manually
   * fetch data via HTTP(s), and trying to use `download*` methods
   * will result in an error
   */
  get isDownloadable() {
    return this.raw._ === "webDocument";
  }
}
const WebDocument$1 = /* @__PURE__ */ inspectable.makeInspectable(WebDocument);
exports.WebDocument = WebDocument$1;
