// @mtcute/core/highlevel/types/conversation.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const errors = require("../../types/errors.cjs");
const peerUtils = require("../../utils/peer-utils.cjs");
const getPeerDialogs = require("../methods/dialogs/get-peer-dialogs.cjs");
const readHistory = require("../methods/messages/read-history.cjs");
const sendMediaGroup = require("../methods/messages/send-media-group.cjs");
const sendMedia = require("../methods/messages/send-media.cjs");
const sendText = require("../methods/messages/send-text.cjs");
const resolvePeer = require("../methods/users/resolve-peer.cjs");
const CONVERSATION_SYMBOL = Symbol("conversation");
class Conversation {
  constructor(client, chat) {
    this.client = client;
    this.chat = chat;
    if (!(CONVERSATION_SYMBOL in client)) {
      client[CONVERSATION_SYMBOL] = {
        pendingConversations: /* @__PURE__ */ new Map(),
        hasConversations: false
      };
    }
  }
  _inputPeer;
  _chatId;
  _started = false;
  _lastMessage;
  _lastReceivedMessage;
  _queuedNewMessage = new utils.Deque();
  _pendingNewMessages = new utils.Deque();
  _lock = new utils.AsyncLock();
  _pendingEditMessage = /* @__PURE__ */ new Map();
  _recentEdits = new utils.Deque(void 0, { capacity: 10 });
  _pendingRead = /* @__PURE__ */ new Map();
  static _getState(client) {
    return client[CONVERSATION_SYMBOL];
  }
  /**
   * Pass the update to the conversation manager and all registered
   * conversations on this client.
   *
   * @returns `true` if the update was handled by some conversation
   */
  static handleUpdate(client, update) {
    const state = Conversation._getState(client);
    if (!state?.hasConversations) return false;
    let chatId;
    switch (update.name) {
      case "new_message":
      case "edit_message":
        chatId = peerUtils.getMarkedPeerId(update.data.raw.peerId);
        break;
      case "history_read":
        chatId = update.data.chatId;
        break;
      default:
        return false;
    }
    const conv = state.pendingConversations.get(chatId);
    if (!conv) return false;
    for (const c of conv) {
      switch (update.name) {
        case "new_message":
          c._onNewMessage(update.data);
          return true;
        case "edit_message":
          c._onEditMessage(update.data);
          return true;
        case "history_read":
          c._onHistoryRead(update.data);
          return true;
      }
    }
    return false;
  }
  /**
   * Get the input peer that this conversation is with
   */
  get inputPeer() {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    return this._inputPeer;
  }
  /**
   * ID of the very last message in this conversation.
   */
  get lastMessage() {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    return this._lastMessage;
  }
  /**
   * ID of the last incoming message in this conversation.
   *
   * Note that before any messages were received since the {@link start}
   * of the conversation, this will equal to {@link lastMessage}
   */
  get lastReceivedMessage() {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    return this._lastReceivedMessage;
  }
  /**
   * Start the conversation
   */
  async start() {
    if (this._started) return;
    this._started = true;
    this._inputPeer = await resolvePeer.resolvePeer(this.client, this.chat);
    this._chatId = peerUtils.getMarkedPeerId(this._inputPeer);
    const [dialog] = await getPeerDialogs.getPeerDialogs(this.client, this._inputPeer);
    const lastMessage = dialog?.lastMessage;
    if (lastMessage) {
      this._lastMessage = this._lastReceivedMessage = lastMessage.id;
    } else {
      this._lastMessage = this._lastReceivedMessage = 0;
    }
    const state = Conversation._getState(this.client);
    if (!state.pendingConversations.has(this._chatId)) {
      state.pendingConversations.set(this._chatId, []);
    }
    state.pendingConversations.get(this._chatId).push(this);
    state.hasConversations = true;
  }
  /**
   * Stop the conversation
   */
  stop() {
    if (!this._started) return;
    const state = Conversation._getState(this.client);
    const pending = state.pendingConversations.get(this._chatId);
    const pendingIdx = pending?.indexOf(this) ?? -1;
    if (pendingIdx > -1) {
      pending.splice(pendingIdx, 1);
    }
    if (pending && !pending.length) {
      state.pendingConversations.delete(this._chatId);
    }
    state.hasConversations = Boolean(state.pendingConversations.size);
    this._queuedNewMessage.clear();
    this._pendingNewMessages.clear();
    this._pendingEditMessage.clear();
    this._recentEdits.clear();
    this._pendingRead.clear();
    this._started = false;
  }
  _recordMessage(msg, incoming = false) {
    this._lastMessage = msg.id;
    if (incoming) this._lastReceivedMessage = msg.id;
    return msg;
  }
  /**
   * Send a text message to this conversation.
   *
   * Wrapper over {@link sendText}
   */
  async sendText(...params) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    return this._recordMessage(await sendText.sendText(this.client, this._inputPeer, ...params));
  }
  /**
   * Send a media to this conversation.
   *
   * Wrapper over {@link sendMedia}
   */
  async sendMedia(...params) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    return this._recordMessage(await sendMedia.sendMedia(this.client, this._inputPeer, ...params));
  }
  /**
   * Send a media group to this conversation.
   *
   * Wrapper over {@link sendMediaGroup}
   */
  async sendMediaGroup(...params) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    const msgs = await sendMediaGroup.sendMediaGroup(this.client, this._inputPeer, ...params);
    this._recordMessage(msgs[msgs.length - 1]);
    return msgs;
  }
  /**
   * Mark the conversation as read up to a certain point.
   *
   * By default, reads until the last message.
   * You can pass `message=null` to read the entire conversation,
   * or pass message ID to read up until that ID.
   */
  markRead({
    message,
    clearMentions = true
  } = {}) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    if (message === null) {
      message = 0;
    } else if (message === void 0) {
      message = this._lastMessage ?? 0;
    }
    return readHistory.readHistory(this.client, this._inputPeer, { maxId: message, clearMentions });
  }
  /**
   * Helper method that calls {@link start},
   * the provided function and then {@link stop}.
   *
   * It is preferred that you use this function rather than
   * manually starting and stopping the conversation.
   *
   * If you don't stop the conversation when you're done,
   * it *will* lead to memory leaks.
   *
   * @param handler
   */
  async with(handler) {
    await this.start();
    let err;
    let res;
    try {
      res = await handler();
    } catch (e) {
      err = e;
    }
    this.stop();
    if (err) throw err;
    return res;
  }
  /**
   * Wait for a new message in the conversation
   *
   * @param filter  Filter for the handler. You can use any filter you can use for dispatcher
   * @param [timeout]  Timeout for the handler in ms. Pass `null` to disable.
   *   When the timeout is reached, `MtTimeoutError` is thrown.
   */
  waitForNewMessage(filter, timeout = 15e3) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    const promise = new utils.Deferred();
    let timer;
    if (timeout !== null) {
      timer = utils.timers.setTimeout(() => {
        promise.reject(new errors.MtTimeoutError(timeout));
        this._queuedNewMessage.removeBy((it) => it.promise === promise);
      }, timeout);
    }
    this._queuedNewMessage.pushBack({
      promise,
      check: filter,
      timeout: timer
    });
    this._processPendingNewMessages();
    return promise.promise;
  }
  /**
   * Wait for a response for the given message
   * (def. the last one) in the conversation.
   *
   * A message is considered to be a response if
   * it was sent after the given one.
   *
   * @param filter  Filter for the handler. You can use any filter you can use for dispatcher
   * @param params
   */
  waitForResponse(filter, params) {
    const msgId = params?.message ?? this._lastMessage ?? 0;
    const pred = filter ? (msg) => msg.id > msgId ? filter(msg) : false : (msg) => msg.id > msgId;
    return this.waitForNewMessage(pred, params?.timeout);
  }
  /**
   * Wait for the reply for the given message'
   * (def. the last one) in the conversation.
   *
   * @param filter  Filter for the handler. You can use any filter you can use for dispatcher
   * @param params
   */
  waitForReply(filter, params) {
    const msgId = params?.message ?? this._lastMessage;
    if (!msgId) {
      throw new errors.MtArgumentError("Provide message for which to wait for reply for");
    }
    const pred = filter ? (msg) => msg.replyToMessage?.id === msgId ? filter(msg) : false : (msg) => msg.replyToMessage?.id === msgId;
    return this.waitForNewMessage(pred, params?.timeout);
  }
  /**
   * Wait for a message to be edited in the conversation.
   * By defaults wait for the last message sent by the other party
   * (at the moment) to be edited.
   *
   * Returns the edited message.
   *
   * @param filter  Filter for the handler. You can use any filter you can use for dispatcher
   * @param params
   */
  async waitForEdit(filter, params) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    const msgId = params?.message ?? this._lastReceivedMessage;
    if (!msgId) {
      throw new errors.MtArgumentError("Provide message for which to wait for edit for");
    }
    const promise = new utils.Deferred();
    let timer;
    const timeout = params?.timeout;
    if (timeout) {
      timer = utils.timers.setTimeout(() => {
        promise.reject(new errors.MtTimeoutError(timeout));
        this._pendingEditMessage.delete(msgId);
      }, timeout);
    }
    this._pendingEditMessage.set(msgId, {
      promise,
      check: filter,
      timeout: timer
    });
    this._processRecentEdits();
    return promise.promise;
  }
  /**
   * Wait for the message to be read by the other party.
   *
   * Note that reading the message doesn't mean the response was sent,
   * and if the response was sent, it doesn't mean that the message was read.
   *
   * @param message  Message for which to wait for read for. Defaults to last message.
   * @param timeout  Timeout for the handler in ms, def. 15 sec. Pass `null` to disable.
   *   When the timeout is reached, `TimeoutError` is thrown.
   */
  async waitForRead(message, timeout = 15e3) {
    if (!this._started) {
      throw new errors.MtArgumentError("Conversation hasn't started yet");
    }
    const msgId = message ?? this._lastMessage;
    if (!msgId) {
      throw new errors.MtArgumentError("Provide message for which to wait for read for");
    }
    const [dialog] = await getPeerDialogs.getPeerDialogs(this.client, this._inputPeer);
    if (!dialog || dialog.lastRead >= msgId) return;
    const promise = new utils.Deferred();
    let timer;
    if (timeout !== null) {
      timer = utils.timers.setTimeout(() => {
        promise.reject(new errors.MtTimeoutError(timeout));
        this._pendingRead.delete(msgId);
      }, timeout);
    }
    this._pendingRead.set(msgId, {
      promise,
      timeout: timer
    });
    return promise.promise;
  }
  _onNewMessage(msg) {
    if (msg.chat.id !== this._chatId) return;
    if (!this._queuedNewMessage.length) {
      this._pendingNewMessages.pushBack(msg);
      return;
    }
    const it = this._queuedNewMessage.peekFront();
    void this._lock.acquire().then(async () => {
      try {
        if (!it.check || await it.check(msg)) {
          if (it.timeout) utils.timers.clearTimeout(it.timeout);
          it.promise.resolve(msg);
          this._queuedNewMessage.popFront();
        }
      } catch (e) {
        this.client.onError.emit(utils.unknownToError(e));
      }
      this._lastMessage = this._lastReceivedMessage = msg.id;
      this._lock.release();
    });
  }
  _onEditMessage(msg, fromRecent = false) {
    if (msg.chat.id !== this._chatId) return;
    const it = this._pendingEditMessage.get(msg.id);
    if (!it) {
      if (!fromRecent) {
        this._recentEdits.pushBack(msg);
      }
      return;
    }
    (async () => {
      if (!it.check || await it.check(msg)) {
        if (it.timeout) utils.timers.clearTimeout(it.timeout);
        it.promise.resolve(msg);
        this._pendingEditMessage.delete(msg.id);
      }
    })().catch((e) => {
      this.client.onError.emit(utils.unknownToError(e));
    });
  }
  _onHistoryRead(upd) {
    if (upd.chatId !== this._chatId) return;
    const lastRead = upd.maxReadId;
    for (const msgId of this._pendingRead.keys()) {
      if (msgId <= lastRead) {
        const it = this._pendingRead.get(msgId);
        if (it.timeout) utils.timers.clearTimeout(it.timeout);
        it.promise.resolve();
        this._pendingRead.delete(msgId);
      }
    }
  }
  _processPendingNewMessages() {
    if (!this._pendingNewMessages.length) return;
    let it;
    while (it = this._pendingNewMessages.popFront()) {
      this._onNewMessage(it);
    }
  }
  _processRecentEdits() {
    if (!this._recentEdits.length) return;
    let it;
    while (it = this._recentEdits.popFront()) {
      this._onEditMessage(it, true);
    }
  }
}
exports.Conversation = Conversation;
