// @mtcute/core/highlevel/types/errors.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../types/errors.cjs");
class MtPeerNotFoundError extends errors.MtcuteError {
}
class MtMessageNotFoundError extends errors.MtcuteError {
  constructor(peerId, messageId, context) {
    super(`Message${context ? ` ${context}` : ""} ${messageId} not found in ${peerId}`);
    this.peerId = peerId;
    this.messageId = messageId;
    this.context = context;
  }
}
class MtInvalidPeerTypeError extends errors.MtcuteError {
  constructor(peer, expected) {
    super(`Provided identifier ${JSON.stringify(peer)} is not a ${expected}`);
  }
}
class MtEmptyError extends errors.MtcuteError {
  constructor() {
    super("Property is not available on an empty object");
  }
}
exports.MtEmptyError = MtEmptyError;
exports.MtInvalidPeerTypeError = MtInvalidPeerTypeError;
exports.MtMessageNotFoundError = MtMessageNotFoundError;
exports.MtPeerNotFoundError = MtPeerNotFoundError;
