// @mtcute/core/types/errors.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class MtcuteError extends Error {
}
class MtArgumentError extends MtcuteError {
}
class MtSecurityError extends MtcuteError {
}
class MtUnsupportedError extends MtcuteError {
}
class MtTypeAssertionError extends MtcuteError {
  /**
   * Context at which the error occurred.
   * Usually a user-friendly string containing name
   * of the high-level API method, name of the TL
   * RPC method, and path of the entity,
   * like this: `signIn (@ auth.signIn -> user)`
   */
  context;
  /** Expected TL type */
  expected;
  /** Actual TL type */
  actual;
  constructor(context, expected, actual) {
    super(`Type assertion error at ${context}: expected ${expected}, but got ${actual}`);
    this.context = context;
    this.expected = expected;
    this.actual = actual;
  }
}
class MtTimeoutError extends MtcuteError {
  constructor(timeout) {
    super(`Request timed out${timeout ? ` after ${timeout}ms` : ""}`);
    this.timeout = timeout;
  }
}
exports.MtArgumentError = MtArgumentError;
exports.MtSecurityError = MtSecurityError;
exports.MtTimeoutError = MtTimeoutError;
exports.MtTypeAssertionError = MtTypeAssertionError;
exports.MtUnsupportedError = MtUnsupportedError;
exports.MtcuteError = MtcuteError;
