// @mtcute/core/types/errors.js
class MtcuteError extends Error {
}
class MtArgumentError extends MtcuteError {
}
class MtSecurityError extends MtcuteError {
}
class MtUnsupportedError extends MtcuteError {
}
class MtTypeAssertionError extends MtcuteError {
  /**
   * Context at which the error occurred.
   * Usually a user-friendly string containing name
   * of the high-level API method, name of the TL
   * RPC method, and path of the entity,
   * like this: `signIn (@ auth.signIn -> user)`
   */
  context;
  /** Expected TL type */
  expected;
  /** Actual TL type */
  actual;
  constructor(context, expected, actual) {
    super(`Type assertion error at ${context}: expected ${expected}, but got ${actual}`);
    this.context = context;
    this.expected = expected;
    this.actual = actual;
  }
}
class MtTimeoutError extends MtcuteError {
  constructor(timeout) {
    super(`Request timed out${timeout ? ` after ${timeout}ms` : ""}`);
    this.timeout = timeout;
  }
}
export {
  MtArgumentError,
  MtSecurityError,
  MtTimeoutError,
  MtTypeAssertionError,
  MtUnsupportedError,
  MtcuteError
};
