// @mtcute/wasm/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/wasm] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/wasm] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
var _documentCurrentScript = typeof document !== "undefined" ? document.currentScript : null;
const SIMD_AVAILABLE = /* @__PURE__ */ WebAssembly.validate(new Uint8Array(
  [0, 97, 115, 109, 1, 0, 0, 0, 1, 5, 1, 96, 0, 1, 123, 3, 2, 1, 0, 10, 10, 1, 8, 0, 65, 0, 253, 15, 253, 98, 11]
));
function getWasmUrl() {
  if (SIMD_AVAILABLE) {
    return new URL(
      "./mtcute-simd.wasm",
      typeof document === "undefined" ? require("url").pathToFileURL(__filename).href : _documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === "SCRIPT" && _documentCurrentScript.src || new URL("index.cjs", document.baseURI).href
    );
  }
  return new URL(
    "./mtcute.wasm",
    typeof document === "undefined" ? require("url").pathToFileURL(__filename).href : _documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === "SCRIPT" && _documentCurrentScript.src || new URL("index.cjs", document.baseURI).href
  );
}
let wasm;
let compressor;
let decompressor;
let sharedOutPtr;
let sharedKeyPtr;
let sharedIvPtr;
let cachedUint8Memory = null;
function initCommon() {
  compressor = wasm.libdeflate_alloc_compressor(6);
  decompressor = wasm.libdeflate_alloc_decompressor();
  sharedOutPtr = wasm.__get_shared_out();
  sharedKeyPtr = wasm.__get_shared_key_buffer();
  sharedIvPtr = wasm.__get_shared_iv_buffer();
}
function getUint8Memory() {
  if (cachedUint8Memory === null || cachedUint8Memory.byteLength === 0) {
    cachedUint8Memory = new Uint8Array(wasm.memory.buffer);
  }
  return cachedUint8Memory;
}
function initSync(module2) {
  if (wasm !== void 0) return;
  if (!(module2 instanceof WebAssembly.Instance)) {
    if (!(module2 instanceof WebAssembly.Module)) {
      module2 = new WebAssembly.Module(module2);
    }
    module2 = new WebAssembly.Instance(module2);
  }
  wasm = module2.exports;
  initCommon();
}
function deflateMaxSize(bytes, size) {
  const outputPtr = wasm.__malloc(size);
  const inputPtr = wasm.__malloc(bytes.length);
  const mem = getUint8Memory();
  mem.set(bytes, inputPtr);
  const written = wasm.libdeflate_zlib_compress(compressor, inputPtr, bytes.length, outputPtr, size);
  wasm.__free(inputPtr);
  if (written === 0) {
    wasm.__free(outputPtr);
    return null;
  }
  const result = mem.slice(outputPtr, outputPtr + written);
  wasm.__free(outputPtr);
  return result;
}
function gunzip(bytes) {
  const inputPtr = wasm.__malloc(bytes.length);
  getUint8Memory().set(bytes, inputPtr);
  const size = wasm.libdeflate_gzip_get_output_size(inputPtr, bytes.length);
  const outputPtr = wasm.__malloc(size);
  const ret = wasm.libdeflate_gzip_decompress(decompressor, inputPtr, bytes.length, outputPtr, size);
  if (ret === -1) throw new Error("gunzip error -- bad data");
  if (ret === -2) throw new Error("gunzip error -- short output");
  if (ret === -3) throw new Error("gunzip error -- short input");
  const result = getUint8Memory().slice(outputPtr, outputPtr + size);
  wasm.__free(inputPtr);
  wasm.__free(outputPtr);
  return result;
}
function ige256Encrypt(data, key, iv) {
  const ptr = wasm.__malloc(data.length + data.length);
  const inputPtr = ptr;
  const outputPtr = inputPtr + data.length;
  const mem = getUint8Memory();
  mem.set(data, inputPtr);
  mem.set(key, sharedKeyPtr);
  mem.set(iv, sharedIvPtr);
  wasm.ige256_encrypt(inputPtr, data.length, outputPtr);
  const result = mem.slice(outputPtr, outputPtr + data.length);
  wasm.__free(ptr);
  return result;
}
function ige256Decrypt(data, key, iv) {
  const ptr = wasm.__malloc(data.length + data.length);
  const inputPtr = ptr;
  const outputPtr = inputPtr + data.length;
  const mem = getUint8Memory();
  mem.set(data, inputPtr);
  mem.set(key, sharedKeyPtr);
  mem.set(iv, sharedIvPtr);
  wasm.ige256_decrypt(inputPtr, data.length, outputPtr);
  const result = mem.slice(outputPtr, outputPtr + data.length);
  wasm.__free(ptr);
  return result;
}
function createCtr256(key, iv) {
  getUint8Memory().set(key, sharedKeyPtr);
  getUint8Memory().set(iv, sharedIvPtr);
  return wasm.ctr256_alloc();
}
function freeCtr256(ctx) {
  wasm.ctr256_free(ctx);
}
function ctr256(ctx, data) {
  const { __malloc, __free } = wasm;
  const inputPtr = __malloc(data.length);
  const outputPtr = __malloc(data.length);
  const mem = getUint8Memory();
  mem.set(data, inputPtr);
  wasm.ctr256(ctx, inputPtr, data.length, outputPtr);
  const result = mem.slice(outputPtr, outputPtr + data.length);
  __free(outputPtr);
  return result;
}
function sha256(data) {
  const { __malloc, __free } = wasm;
  const inputPtr = __malloc(data.length);
  const mem = getUint8Memory();
  mem.set(data, inputPtr);
  wasm.sha256(inputPtr, data.length);
  __free(inputPtr);
  return mem.slice(sharedOutPtr, sharedOutPtr + 32);
}
function sha1(data) {
  const { __malloc, __free } = wasm;
  const inputPtr = __malloc(data.length);
  const mem = getUint8Memory();
  mem.set(data, inputPtr);
  wasm.sha1(inputPtr, data.length);
  __free(inputPtr);
  return mem.slice(sharedOutPtr, sharedOutPtr + 20);
}
function __getWasm() {
  return wasm;
}
exports.SIMD_AVAILABLE = SIMD_AVAILABLE;
exports.__getWasm = __getWasm;
exports.createCtr256 = createCtr256;
exports.ctr256 = ctr256;
exports.deflateMaxSize = deflateMaxSize;
exports.freeCtr256 = freeCtr256;
exports.getWasmUrl = getWasmUrl;
exports.gunzip = gunzip;
exports.ige256Decrypt = ige256Decrypt;
exports.ige256Encrypt = ige256Encrypt;
exports.initSync = initSync;
exports.sha1 = sha1;
exports.sha256 = sha256;
