// @mtcute/core/highlevel/updates/utils.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../types/errors.cjs");
function messageToUpdate(message) {
  switch (message.peerId._) {
    case "peerUser":
    case "peerChat":
      return {
        _: "updateNewMessage",
        message,
        pts: 0,
        ptsCount: 0
      };
    case "peerChannel":
      return {
        _: "updateNewChannelMessage",
        message,
        pts: 0,
        ptsCount: 0
      };
  }
}
function extractChannelIdFromUpdate(upd) {
  let res = 0;
  if ("channelId" in upd) {
    res = upd.channelId;
  } else if ("message" in upd && typeof upd.message !== "string" && "peerId" in upd.message && upd.message.peerId && "channelId" in upd.message.peerId) {
    res = upd.message.peerId.channelId;
  }
  if (res === 0) return void 0;
  return res;
}
function toPendingUpdate(upd, peers, fromDifference = false) {
  const channelId = extractChannelIdFromUpdate(upd) || 0;
  const pts = "pts" in upd ? upd.pts : void 0;
  const ptsCount = "ptsCount" in upd ? upd.ptsCount : pts ? 0 : void 0;
  const qts = "qts" in upd ? upd.qts : void 0;
  return {
    update: upd,
    channelId,
    pts,
    ptsBefore: pts ? pts - ptsCount : void 0,
    qts,
    qtsBefore: qts ? qts - 1 : void 0,
    peers,
    fromDifference
  };
}
function isMessageEmpty(upd) {
  return upd.message?._ === "messageEmpty";
}
function createDummyUpdatesContainer(updates, seq = 0) {
  return {
    _: "updates",
    seq,
    date: 0,
    chats: [],
    users: [],
    updates
  };
}
function createDummyUpdate(pts, ptsCount, channelId = 0) {
  return createDummyUpdatesContainer([
    {
      _: "mtcute.dummyUpdate",
      channelId,
      pts,
      ptsCount
    }
  ]);
}
function assertIsUpdatesGroup(ctx, upd) {
  switch (upd._) {
    case "updates":
    case "updatesCombined":
      return;
  }
  throw new errors.MtTypeAssertionError(ctx, "updates | updatesCombined", upd._);
}
exports.assertIsUpdatesGroup = assertIsUpdatesGroup;
exports.createDummyUpdate = createDummyUpdate;
exports.createDummyUpdatesContainer = createDummyUpdatesContainer;
exports.extractChannelIdFromUpdate = extractChannelIdFromUpdate;
exports.isMessageEmpty = isMessageEmpty;
exports.messageToUpdate = messageToUpdate;
exports.toPendingUpdate = toPendingUpdate;
