// @mtcute/core/highlevel/types/stories/interactive/base.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class StoryInteractiveArea {
  constructor(raw) {
    this.raw = raw;
    this.raw = raw;
  }
  /** X coordinate of the top-left corner of the area */
  get x() {
    return this.raw.coordinates.x;
  }
  /** Y coordinate of the top-left corner of the area */
  get y() {
    return this.raw.coordinates.y;
  }
  /** Width of the area */
  get width() {
    return this.raw.coordinates.w;
  }
  /** Height of the area */
  get height() {
    return this.raw.coordinates.h;
  }
  /** Rotation of the area */
  get rotation() {
    return this.raw.coordinates.rotation;
  }
}
exports.StoryInteractiveArea = StoryInteractiveArea;
