// @mtcute/core/highlevel/types/stories/interactive/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../../types/errors.cjs");
const channelPost = require("./channel-post.cjs");
const location = require("./location.cjs");
const reaction = require("./reaction.cjs");
const starGift = require("./star-gift.cjs");
const url = require("./url.cjs");
const venue = require("./venue.cjs");
const weather = require("./weather.cjs");
function _storyInteractiveElementFromTl(raw, peers) {
  switch (raw._) {
    case "mediaAreaSuggestedReaction":
      return new reaction.StoryInteractiveReaction(raw);
    case "mediaAreaGeoPoint":
      return new location.StoryInteractiveLocation(raw);
    case "mediaAreaVenue":
      return new venue.StoryInteractiveVenue(raw);
    case "mediaAreaChannelPost":
      return new channelPost.StoryInteractiveChannelPost(raw, peers);
    case "mediaAreaUrl":
      return new url.StoryInteractiveUrl(raw);
    case "mediaAreaWeather":
      return new weather.StoryInteractiveWeather(raw);
    case "mediaAreaStarGift":
      return new starGift.StoryInteractiveStarGift(raw);
    case "inputMediaAreaVenue":
    case "inputMediaAreaChannelPost":
      throw new errors.MtTypeAssertionError("StoryInteractiveElement", "!input*", raw._);
  }
}
exports.StoryInteractiveChannelPost = channelPost.StoryInteractiveChannelPost;
exports.StoryInteractiveLocation = location.StoryInteractiveLocation;
exports.StoryInteractiveReaction = reaction.StoryInteractiveReaction;
exports.StoryInteractiveStarGift = starGift.StoryInteractiveStarGift;
exports.StoryInteractiveUrl = url.StoryInteractiveUrl;
exports.StoryInteractiveVenue = venue.StoryInteractiveVenue;
exports.StoryInteractiveWeather = weather.StoryInteractiveWeather;
exports._storyInteractiveElementFromTl = _storyInteractiveElementFromTl;
