// @mtcute/core/highlevel/types/stories/interactive/index.js
import { MtTypeAssertionError } from "../../../../types/errors.js";
import { StoryInteractiveChannelPost as StoryInteractiveChannelPost$2 } from "./channel-post.js";
import { StoryInteractiveLocation as StoryInteractiveLocation$2 } from "./location.js";
import { StoryInteractiveReaction as StoryInteractiveReaction$1 } from "./reaction.js";
import { StoryInteractiveStarGift as StoryInteractiveStarGift$1 } from "./star-gift.js";
import { StoryInteractiveUrl } from "./url.js";
import { StoryInteractiveVenue as StoryInteractiveVenue$2 } from "./venue.js";
import { StoryInteractiveWeather as StoryInteractiveWeather$1 } from "./weather.js";
function _storyInteractiveElementFromTl(raw, peers) {
  switch (raw._) {
    case "mediaAreaSuggestedReaction":
      return new StoryInteractiveReaction$1(raw);
    case "mediaAreaGeoPoint":
      return new StoryInteractiveLocation$2(raw);
    case "mediaAreaVenue":
      return new StoryInteractiveVenue$2(raw);
    case "mediaAreaChannelPost":
      return new StoryInteractiveChannelPost$2(raw, peers);
    case "mediaAreaUrl":
      return new StoryInteractiveUrl(raw);
    case "mediaAreaWeather":
      return new StoryInteractiveWeather$1(raw);
    case "mediaAreaStarGift":
      return new StoryInteractiveStarGift$1(raw);
    case "inputMediaAreaVenue":
    case "inputMediaAreaChannelPost":
      throw new MtTypeAssertionError("StoryInteractiveElement", "!input*", raw._);
  }
}
export {
  StoryInteractiveChannelPost$2 as StoryInteractiveChannelPost,
  StoryInteractiveLocation$2 as StoryInteractiveLocation,
  StoryInteractiveReaction$1 as StoryInteractiveReaction,
  StoryInteractiveStarGift$1 as StoryInteractiveStarGift,
  StoryInteractiveUrl,
  StoryInteractiveVenue$2 as StoryInteractiveVenue,
  StoryInteractiveWeather$1 as StoryInteractiveWeather,
  _storyInteractiveElementFromTl
};
