// @mtcute/core/highlevel/types/stories/interactive/input.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const types = require("../../reactions/types.cjs");
class StoryElement {
  constructor(_position) {
    this._position = _position;
  }
  static at(params) {
    return new StoryElement({
      _: "mediaAreaCoordinates",
      x: params.x,
      y: params.y,
      w: params.width,
      h: params.height,
      rotation: params.rotation ?? 0
    });
  }
  venue(params) {
    return {
      _: "mediaAreaVenue",
      coordinates: this._position,
      geo: {
        _: "geoPoint",
        lat: params.latitude,
        long: params.longitude,
        accessHash: Long.ZERO
      },
      title: params.title,
      address: params.address,
      provider: params.source.provider ?? "foursquare",
      venueId: params.source.id,
      venueType: params.source.type
    };
  }
  venueFromInline(queryId, resultId) {
    return {
      _: "inputMediaAreaVenue",
      coordinates: this._position,
      queryId,
      resultId
    };
  }
  location(params) {
    return {
      _: "mediaAreaGeoPoint",
      coordinates: this._position,
      geo: {
        _: "geoPoint",
        lat: params.latitude,
        long: params.longitude,
        accessHash: Long.ZERO
      }
    };
  }
  reaction(reaction, params = {}) {
    this._position.h *= 9 / 16;
    return {
      _: "mediaAreaSuggestedReaction",
      coordinates: this._position,
      reaction: types.normalizeInputReaction(reaction),
      dark: params.dark,
      flipped: params.flipped
    };
  }
  url(url) {
    return {
      _: "mediaAreaUrl",
      coordinates: this._position,
      url
    };
  }
}
exports.StoryElement = StoryElement;
