// @mtcute/core/highlevel/types/stories/interactive/input.js
import Long from "long";
import { normalizeInputReaction } from "../../reactions/types.js";
class StoryElement {
  constructor(_position) {
    this._position = _position;
  }
  static at(params) {
    return new StoryElement({
      _: "mediaAreaCoordinates",
      x: params.x,
      y: params.y,
      w: params.width,
      h: params.height,
      rotation: params.rotation ?? 0
    });
  }
  venue(params) {
    return {
      _: "mediaAreaVenue",
      coordinates: this._position,
      geo: {
        _: "geoPoint",
        lat: params.latitude,
        long: params.longitude,
        accessHash: Long.ZERO
      },
      title: params.title,
      address: params.address,
      provider: params.source.provider ?? "foursquare",
      venueId: params.source.id,
      venueType: params.source.type
    };
  }
  venueFromInline(queryId, resultId) {
    return {
      _: "inputMediaAreaVenue",
      coordinates: this._position,
      queryId,
      resultId
    };
  }
  location(params) {
    return {
      _: "mediaAreaGeoPoint",
      coordinates: this._position,
      geo: {
        _: "geoPoint",
        lat: params.latitude,
        long: params.longitude,
        accessHash: Long.ZERO
      }
    };
  }
  reaction(reaction, params = {}) {
    this._position.h *= 9 / 16;
    return {
      _: "mediaAreaSuggestedReaction",
      coordinates: this._position,
      reaction: normalizeInputReaction(reaction),
      dark: params.dark,
      flipped: params.flipped
    };
  }
  url(url) {
    return {
      _: "mediaAreaUrl",
      coordinates: this._position,
      url
    };
  }
}
export {
  StoryElement
};
