// @mtcute/core/highlevel/types/stories/interactive/venue.js
import { assertTypeIs } from "../../../../utils/type-assertions.js";
import { memoizeGetters } from "../../../utils/memoize.js";
import { Location as Location$1 } from "../../media/location.js";
import { StoryInteractiveArea } from "./base.js";
import { makeInspectable } from "../../../utils/inspectable.js";
class StoryInteractiveVenue extends StoryInteractiveArea {
  constructor(raw) {
    super(raw);
    this.raw = raw;
  }
  type = "venue";
  /**
   * Geolocation of the venue
   */
  get location() {
    assertTypeIs("StoryInteractiveVenue#location", this.raw.geo, "geoPoint");
    return new Location$1(this.raw.geo);
  }
  /**
   * Venue name
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Venue address
   */
  get address() {
    return this.raw.address;
  }
  /**
   * When available, source from where this venue was acquired
   */
  get source() {
    if (!this.raw.provider) return null;
    return {
      provider: this.raw.provider,
      id: this.raw.venueId,
      type: this.raw.venueType
    };
  }
}
const StoryInteractiveVenue$1 = /* @__PURE__ */ memoizeGetters(StoryInteractiveVenue, ["location"]);
const StoryInteractiveVenue$2 = /* @__PURE__ */ makeInspectable(StoryInteractiveVenue$1);
export {
  StoryInteractiveVenue$2 as StoryInteractiveVenue
};
