// @mtcute/core/highlevel/types/stories/interactive/venue.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../../utils/type-assertions.cjs");
const memoize = require("../../../utils/memoize.cjs");
const location = require("../../media/location.cjs");
const base = require("./base.cjs");
const inspectable = require("../../../utils/inspectable.cjs");
class StoryInteractiveVenue extends base.StoryInteractiveArea {
  constructor(raw) {
    super(raw);
    this.raw = raw;
  }
  type = "venue";
  /**
   * Geolocation of the venue
   */
  get location() {
    typeAssertions.assertTypeIs("StoryInteractiveVenue#location", this.raw.geo, "geoPoint");
    return new location.Location(this.raw.geo);
  }
  /**
   * Venue name
   */
  get title() {
    return this.raw.title;
  }
  /**
   * Venue address
   */
  get address() {
    return this.raw.address;
  }
  /**
   * When available, source from where this venue was acquired
   */
  get source() {
    if (!this.raw.provider) return null;
    return {
      provider: this.raw.provider,
      id: this.raw.venueId,
      type: this.raw.venueType
    };
  }
}
const StoryInteractiveVenue$1 = /* @__PURE__ */ memoize.memoizeGetters(StoryInteractiveVenue, ["location"]);
const StoryInteractiveVenue$2 = /* @__PURE__ */ inspectable.makeInspectable(StoryInteractiveVenue$1);
exports.StoryInteractiveVenue = StoryInteractiveVenue$2;
