// @mtcute/core/highlevel/types/premium/boost.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const memoize = require("../../utils/memoize.cjs");
const user = require("../peers/user.cjs");
const inspectable = require("../../utils/inspectable.cjs");
class Boost {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Unique ID of this boost */
  get id() {
    return this.raw.id;
  }
  /** Number of boosts this boost is actually representing */
  get count() {
    return this.raw.multiplier ?? 1;
  }
  /** Date when this boost was applied */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Amount of Telegram Stars that were awarded along with this boost */
  get stars() {
    return this.raw.stars ?? null;
  }
  /**
   * Date when this boost will automatically expire.
   *
   * > **Note**: User can still manually cancel the boost before that date
   */
  get expireDate() {
    return new Date(this.raw.expires * 1e3);
  }
  /**
   * Whether this boost was applied because the channel
   * directly gifted a subscription to the user
   */
  get origin() {
    if (this.raw.unclaimed) return "unclaimed_gift";
    if (this.raw.gift) return "gift";
    if (this.raw.giveaway) return "giveaway";
    if (this.raw.userId) return "user";
    throw new errors.MtUnsupportedError("Unknown boost origin");
  }
  /**
   * User who is boosting the channel.
   *
   * Only available for some origins
   */
  get user() {
    if (!this.raw.userId) return null;
    return new user.User(this._peers.user(this.raw.userId));
  }
  /**
   * ID of the message containing the giveaway where this
   * user has won
   */
  get giveawayMessageId() {
    return this.raw.giveawayMsgId ?? null;
  }
  /**
   * The created Telegram Premium gift code, only set if `origin` is not `user`,
   * AND it is either a gift code for the currently logged in user,
   * or if it was already claimed
   */
  get usedGiftSlug() {
    return this.raw.usedGiftSlug ?? null;
  }
}
const Boost$1 = /* @__PURE__ */ memoize.memoizeGetters(Boost, ["user"]);
const Boost$2 = /* @__PURE__ */ inspectable.makeInspectable(Boost$1);
exports.Boost = Boost$2;
