// @mtcute/core/highlevel/types/premium/boost-slot.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const typeAssertions = require("../../../utils/type-assertions.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const chat = require("../peers/chat.cjs");
class BoostSlot {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** ID of this slot */
  get id() {
    return this.raw.slot;
  }
  /**
   * Whether this slot is occupied
   */
  get occupied() {
    return this.raw.peer !== void 0;
  }
  /**
   * Channel that is occupying this slot, if any
   */
  get chat() {
    if (!this.raw.peer) return null;
    typeAssertions.assertTypeIs("BoostSlot.chat", this.raw.peer, "peerChannel");
    return new chat.Chat(this._peers.chat(this.raw.peer.channelId));
  }
  /**
   * Date when we started boosting this channel
   *
   * If this slot is not occupied, will be `0`
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Date when this boost will automatically expire.
   */
  get expireDate() {
    return new Date(this.raw.expires * 1e3);
  }
  /**
   * If this slot is occupied, returns the date when
   * we can reassing this slot to another channel.
   *
   * If `null`, we can reassign it immediately.
   */
  get cooldownUntil() {
    if (!this.raw.cooldownUntilDate) return null;
    return new Date(this.raw.cooldownUntilDate * 1e3);
  }
}
const BoostSlot$1 = /* @__PURE__ */ memoize.memoizeGetters(BoostSlot, ["chat"]);
const BoostSlot$2 = /* @__PURE__ */ inspectable.makeInspectable(BoostSlot$1);
exports.BoostSlot = BoostSlot$2;
