// @mtcute/core/highlevel/types/premium/boost.js
import { MtUnsupportedError } from "../../../types/errors.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "../peers/user.js";
import { makeInspectable } from "../../utils/inspectable.js";
class Boost {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Unique ID of this boost */
  get id() {
    return this.raw.id;
  }
  /** Number of boosts this boost is actually representing */
  get count() {
    return this.raw.multiplier ?? 1;
  }
  /** Date when this boost was applied */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Amount of Telegram Stars that were awarded along with this boost */
  get stars() {
    return this.raw.stars ?? null;
  }
  /**
   * Date when this boost will automatically expire.
   *
   * > **Note**: User can still manually cancel the boost before that date
   */
  get expireDate() {
    return new Date(this.raw.expires * 1e3);
  }
  /**
   * Whether this boost was applied because the channel
   * directly gifted a subscription to the user
   */
  get origin() {
    if (this.raw.unclaimed) return "unclaimed_gift";
    if (this.raw.gift) return "gift";
    if (this.raw.giveaway) return "giveaway";
    if (this.raw.userId) return "user";
    throw new MtUnsupportedError("Unknown boost origin");
  }
  /**
   * User who is boosting the channel.
   *
   * Only available for some origins
   */
  get user() {
    if (!this.raw.userId) return null;
    return new User$2(this._peers.user(this.raw.userId));
  }
  /**
   * ID of the message containing the giveaway where this
   * user has won
   */
  get giveawayMessageId() {
    return this.raw.giveawayMsgId ?? null;
  }
  /**
   * The created Telegram Premium gift code, only set if `origin` is not `user`,
   * AND it is either a gift code for the currently logged in user,
   * or if it was already claimed
   */
  get usedGiftSlug() {
    return this.raw.usedGiftSlug ?? null;
  }
}
const Boost$1 = /* @__PURE__ */ memoizeGetters(Boost, ["user"]);
const Boost$2 = /* @__PURE__ */ makeInspectable(Boost$1);
export {
  Boost$2 as Boost
};
