// @mtcute/core/highlevel/types/premium/boost-slot.js
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { Chat as Chat$2 } from "../peers/chat.js";
class BoostSlot {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** ID of this slot */
  get id() {
    return this.raw.slot;
  }
  /**
   * Whether this slot is occupied
   */
  get occupied() {
    return this.raw.peer !== void 0;
  }
  /**
   * Channel that is occupying this slot, if any
   */
  get chat() {
    if (!this.raw.peer) return null;
    assertTypeIs("BoostSlot.chat", this.raw.peer, "peerChannel");
    return new Chat$2(this._peers.chat(this.raw.peer.channelId));
  }
  /**
   * Date when we started boosting this channel
   *
   * If this slot is not occupied, will be `0`
   */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /**
   * Date when this boost will automatically expire.
   */
  get expireDate() {
    return new Date(this.raw.expires * 1e3);
  }
  /**
   * If this slot is occupied, returns the date when
   * we can reassing this slot to another channel.
   *
   * If `null`, we can reassign it immediately.
   */
  get cooldownUntil() {
    if (!this.raw.cooldownUntilDate) return null;
    return new Date(this.raw.cooldownUntilDate * 1e3);
  }
}
const BoostSlot$1 = /* @__PURE__ */ memoizeGetters(BoostSlot, ["chat"]);
const BoostSlot$2 = /* @__PURE__ */ makeInspectable(BoostSlot$1);
export {
  BoostSlot$2 as BoostSlot
};
