// @mtcute/core/highlevel/types/premium/business-account.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const businessIntro = require("./business-intro.cjs");
const businessLocation = require("./business-location.cjs");
const businessWorkHours = require("./business-work-hours.cjs");
class BusinessAccount {
  constructor(info) {
    this.info = info;
  }
  /** Introduction of the business account */
  get intro() {
    if (!this.info.businessIntro) return null;
    return new businessIntro.BusinessIntro(this.info.businessIntro);
  }
  /** Work hours of the business */
  get workHours() {
    if (!this.info.businessWorkHours) return null;
    return new businessWorkHours.BusinessWorkHours(this.info.businessWorkHours);
  }
  /** Location of the business */
  get location() {
    if (!this.info.businessLocation) return null;
    return new businessLocation.BusinessLocation(this.info.businessLocation);
  }
  /** Information about a greeting message */
  get greetingMessage() {
    return this.info.businessGreetingMessage ?? null;
  }
  /** Information about an "away" message */
  get awayMessage() {
    return this.info.businessAwayMessage ?? null;
  }
}
const BusinessAccount$1 = /* @__PURE__ */ memoize.memoizeGetters(BusinessAccount, ["intro", "workHours", "location"]);
const BusinessAccount$2 = /* @__PURE__ */ inspectable.makeInspectable(BusinessAccount$1);
exports.BusinessAccount = BusinessAccount$2;
