// @mtcute/core/highlevel/types/premium/business-account.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { BusinessIntro as BusinessIntro$2 } from "./business-intro.js";
import { BusinessLocation as BusinessLocation$2 } from "./business-location.js";
import { BusinessWorkHours as BusinessWorkHours$2 } from "./business-work-hours.js";
class BusinessAccount {
  constructor(info) {
    this.info = info;
  }
  /** Introduction of the business account */
  get intro() {
    if (!this.info.businessIntro) return null;
    return new BusinessIntro$2(this.info.businessIntro);
  }
  /** Work hours of the business */
  get workHours() {
    if (!this.info.businessWorkHours) return null;
    return new BusinessWorkHours$2(this.info.businessWorkHours);
  }
  /** Location of the business */
  get location() {
    if (!this.info.businessLocation) return null;
    return new BusinessLocation$2(this.info.businessLocation);
  }
  /** Information about a greeting message */
  get greetingMessage() {
    return this.info.businessGreetingMessage ?? null;
  }
  /** Information about an "away" message */
  get awayMessage() {
    return this.info.businessAwayMessage ?? null;
  }
}
const BusinessAccount$1 = /* @__PURE__ */ memoizeGetters(BusinessAccount, ["intro", "workHours", "location"]);
const BusinessAccount$2 = /* @__PURE__ */ makeInspectable(BusinessAccount$1);
export {
  BusinessAccount$2 as BusinessAccount
};
