// @mtcute/core/highlevel/types/premium/business-connection.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const user = require("../peers/user.cjs");
class BusinessConnection {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether the connection was removed by the user */
  get isRemoved() {
    return this.raw.disabled;
  }
  /** ID of the connection */
  get id() {
    return this.raw.connectionId;
  }
  /** Datacenter ID of the connected user */
  get dcId() {
    return this.raw.dcId;
  }
  /** Date when the connection was created */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** What can the connected bot do on behalf of the user */
  get rights() {
    return this.raw.rights ?? null;
  }
  /** Business account user that created the business connection */
  get user() {
    return new user.User(this._peers.user(this.raw.userId));
  }
}
const BusinessConnection$1 = /* @__PURE__ */ inspectable.makeInspectable(BusinessConnection);
const BusinessConnection$2 = /* @__PURE__ */ memoize.memoizeGetters(BusinessConnection$1, ["user"]);
exports.BusinessConnection = BusinessConnection$2;
