// @mtcute/core/highlevel/types/premium/business-work-hours.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const errors = require("../../../types/errors.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const DAYS_IN_WEEK = 7;
const MINUTES_IN_DAY = 24 * 60;
function businessWorkHoursDaysToRaw(day) {
  const res = [];
  for (const d of day) {
    const dayStart = d.day * MINUTES_IN_DAY;
    if (d.is24h) {
      res.push({
        _: "businessWeeklyOpen",
        startMinute: dayStart,
        endMinute: dayStart + MINUTES_IN_DAY
      });
      continue;
    }
    for (const interval of d.intervals) {
      const start = dayStart + interval.startHour * 60 + interval.startMinute;
      const end = dayStart + interval.endHour * 60 + interval.endMinute;
      if (start >= end) {
        throw new errors.MtArgumentError("startMinute >= endMinute");
      }
      res.push({
        _: "businessWeeklyOpen",
        startMinute: start,
        endMinute: end
      });
    }
  }
  return res;
}
class BusinessWorkHours {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether the business is open right now */
  get isOpenNow() {
    return this.raw.openNow;
  }
  /**
   * Identifier of the time zone in which the {@link hours} are defined,
   * in the IANA format.
   */
  get timezoneId() {
    return this.raw.timezoneId;
  }
  /** Raw "open" intervals */
  get intervals() {
    return this.raw.weeklyOpen;
  }
  /**
   * Parsed business hours intervals per week day.
   *
   * @returns  Array of 7 elements, each representing a day of the week (starting from Monday = 0)
   */
  get days() {
    const days = Array.from({ length: DAYS_IN_WEEK }, (_, i) => ({
      day: i,
      is24h: false,
      intervals: []
    }));
    const sorted = [...this.raw.weeklyOpen].sort((a, b) => a.startMinute - b.startMinute);
    for (let i = 1; i < sorted.length; i++) {
      const prev = sorted[i - 1];
      const cur = sorted[i];
      if (prev.endMinute >= cur.startMinute) {
        prev.endMinute = cur.endMinute;
        sorted.splice(i, 1);
        i--;
      }
    }
    const mondayPrepend = [];
    for (const interval of sorted) {
      if (interval.startMinute > interval.endMinute) {
        throw new errors.MtArgumentError("startMinute is greater than endMinute");
      }
      const startDay = Math.floor(interval.startMinute / MINUTES_IN_DAY);
      const endDay = Math.floor(interval.endMinute / MINUTES_IN_DAY);
      if (endDay > DAYS_IN_WEEK + 1) {
        throw new errors.MtArgumentError("interval spans more than a week");
      }
      for (let day = startDay, dayStart = startDay * MINUTES_IN_DAY; day <= endDay; day++, dayStart += MINUTES_IN_DAY) {
        const startWithin = Math.max(interval.startMinute, dayStart) - dayStart;
        const endWithin = Math.min(interval.endMinute, dayStart + MINUTES_IN_DAY) - dayStart;
        const startHour = Math.floor(startWithin / 60);
        const startMinute = startWithin % 60;
        const endHour = Math.floor(endWithin / 60);
        const endMinute = endWithin % 60;
        if (startHour === 0 && startMinute === 0 && endHour === 0 && endMinute === 0) {
          continue;
        }
        const obj = {
          startHour,
          startMinute,
          endHour,
          endMinute
        };
        if (day === DAYS_IN_WEEK) {
          mondayPrepend.push(obj);
        } else {
          days[day].intervals.push(obj);
        }
      }
    }
    if (mondayPrepend.length > 0) {
      days[0].intervals.unshift(...mondayPrepend);
    }
    for (const day of days) {
      if (day.intervals.length !== 1) continue;
      const interval = day.intervals[0];
      if (interval.startHour === 0 && interval.startMinute === 0 && (interval.endHour === 24 && interval.endMinute === 0 || interval.endHour === 23 && interval.endMinute === 59)) {
        day.is24h = true;
      }
    }
    return days;
  }
}
const BusinessWorkHours$1 = /* @__PURE__ */ inspectable.makeInspectable(BusinessWorkHours, void 0, ["intervals"]);
const BusinessWorkHours$2 = /* @__PURE__ */ memoize.memoizeGetters(BusinessWorkHours$1, ["days"]);
exports.BusinessWorkHours = BusinessWorkHours$2;
exports.businessWorkHoursDaysToRaw = businessWorkHoursDaysToRaw;
