// @mtcute/core/highlevel/types/premium/business-connection.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { User as User$2 } from "../peers/user.js";
class BusinessConnection {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether the connection was removed by the user */
  get isRemoved() {
    return this.raw.disabled;
  }
  /** ID of the connection */
  get id() {
    return this.raw.connectionId;
  }
  /** Datacenter ID of the connected user */
  get dcId() {
    return this.raw.dcId;
  }
  /** Date when the connection was created */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** What can the connected bot do on behalf of the user */
  get rights() {
    return this.raw.rights ?? null;
  }
  /** Business account user that created the business connection */
  get user() {
    return new User$2(this._peers.user(this.raw.userId));
  }
}
const BusinessConnection$1 = /* @__PURE__ */ makeInspectable(BusinessConnection);
const BusinessConnection$2 = /* @__PURE__ */ memoizeGetters(BusinessConnection$1, ["user"]);
export {
  BusinessConnection$2 as BusinessConnection
};
