// @mtcute/core/highlevel/types/premium/business-chat-link.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const messageEntity = require("../messages/message-entity.cjs");
class BusinessChatLink {
  constructor(raw) {
    this.raw = raw;
  }
  /** The link itself */
  get link() {
    return this.raw.link;
  }
  /** Text to be inserted into the message input */
  get text() {
    return this.raw.message;
  }
  /** Entities for the text */
  get entities() {
    return this.raw.entities?.map((x) => new messageEntity.MessageEntity(x)) ?? [];
  }
  /** Custom title for the link */
  get title() {
    return this.raw.title ?? null;
  }
  /** Number of clicks on the link */
  get clicks() {
    return this.raw.views;
  }
}
const BusinessChatLink$1 = /* @__PURE__ */ inspectable.makeInspectable(BusinessChatLink);
const BusinessChatLink$2 = /* @__PURE__ */ memoize.memoizeGetters(BusinessChatLink$1, ["entities"]);
exports.BusinessChatLink = BusinessChatLink$2;
