// @mtcute/core/highlevel/types/premium/saved-star-gift.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peer = require("../peers/peer.cjs");
const starGiftUnique = require("./star-gift-unique.cjs");
const starGift = require("./star-gift.cjs");
class SavedStarGift {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** Whether the sender name is hidden */
  get nameHidden() {
    return this.raw.nameHidden;
  }
  /** Whether this gift was not saved yet */
  get unsaved() {
    return this.raw.unsaved;
  }
  /** Whether this gift was refunded */
  get refunded() {
    return this.raw.refunded;
  }
  /** Whether this gift is pinned to the top of the list */
  get pinned() {
    return this.raw.pinnedToTop;
  }
  /** Whether this gift can be upgraded to a unique gift */
  get canUpgrade() {
    return this.raw.canUpgrade;
  }
  /** Sender of the gift */
  get sender() {
    return this.raw.fromId ? peer.parsePeer(this.raw.fromId, this._peers) : null;
  }
  /** Date when the gift was sent */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** The gift itself */
  get gift() {
    if (this.raw.gift._ === "starGift") {
      return new starGift.StarGift(this.raw.gift, this._peers);
    }
    return new starGiftUnique.StarGiftUnique(this.raw.gift, this._peers);
  }
  /** The attached message */
  get message() {
    return this.raw.message ?? null;
  }
  /** ID of the message where the gift was originally sent (if available) */
  get messageId() {
    return this.raw.msgId ?? null;
  }
  /** ID of the saved gift */
  get savedId() {
    return this.raw.savedId ?? null;
  }
  /** Amount of stars the gift can be converted to */
  get convertStars() {
    return this.raw.convertStars ?? null;
  }
  /** Amount of stars **already paid** to upgrade the gift (e.g. by the sender) */
  get upgradeStars() {
    return this.raw.upgradeStars ?? null;
  }
  /** Amount of stars needed to transfer the gift */
  get transferStars() {
    return this.raw.transferStars ?? null;
  }
  /** Amount of stars needed to drop the original details */
  get dropDetailsStars() {
    return this.raw.dropOriginalDetailsStars ?? null;
  }
  /** Date when the gift can be exported to blockchain */
  get canExportAt() {
    return this.raw.canExportAt ? new Date(this.raw.canExportAt * 1e3) : null;
  }
  /** Date when the gift can be transferred */
  get canTransferAt() {
    return this.raw.canTransferAt ? new Date(this.raw.canTransferAt * 1e3) : null;
  }
  /** Date when the gift can be resold */
  get canResellAt() {
    return this.raw.canResellAt ? new Date(this.raw.canResellAt * 1e3) : null;
  }
  /** Date when the gift can be crafted */
  get canCraftAt() {
    return this.raw.canCraftAt ? new Date(this.raw.canCraftAt * 1e3) : null;
  }
  /** IDs of the collections this gift belongs to */
  get collectionIds() {
    return this.raw.collectionId ?? [];
  }
  /**
   * If available, you can pay for this gift's upgrade by passing this hash to `prepayStarGiftUpgrade`
   */
  get prepaidUpgradeHash() {
    return this.raw.prepaidUpgradeHash ?? null;
  }
  /** Number of the gift for limited-offer gifts */
  get giftNum() {
    return this.raw.giftNum ?? null;
  }
}
const SavedStarGift$1 = /* @__PURE__ */ inspectable.makeInspectable(SavedStarGift);
const SavedStarGift$2 = /* @__PURE__ */ memoize.memoizeGetters(SavedStarGift$1, ["sender", "gift"]);
exports.SavedStarGift = SavedStarGift$2;
