// @mtcute/core/highlevel/types/premium/stars-transaction.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const webDocument = require("../files/web-document.cjs");
const messageMedia = require("../messages/message-media.cjs");
const peer = require("../peers/peer.cjs");
const starGiftUnique = require("./star-gift-unique.cjs");
const starGift = require("./star-gift.cjs");
class StarsTransaction {
  constructor(raw, peers) {
    this.raw = raw;
    this.peers = peers;
  }
  /** ID of the transaction */
  get id() {
    return this.raw.id;
  }
  /** Whether this transaction is a refund */
  get isRefund() {
    return this.raw.refund;
  }
  /**
   * Whether this transaction is outgoing or incoming
   */
  get direction() {
    let isNegative = this.raw.amount.amount.isNegative();
    if (this.raw.refund) isNegative = !isNegative;
    return isNegative ? "outgoing" : "incoming";
  }
  /** Absolute amount of stars/TON in the transaction */
  get amount() {
    let res = this.raw.amount;
    if (res.amount.isNegative()) {
      res = res._ === "starsAmount" ? {
        ...res,
        nanos: Math.abs(res.nanos),
        amount: res.amount.negate()
      } : {
        _: "starsTonAmount",
        amount: res.amount.negate()
      };
    }
    return res;
  }
  /** Date of the transaction */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Type of this transaction */
  get type() {
    switch (this.raw.peer._) {
      case "starsTransactionPeerAppStore":
        return { type: "app_store" };
      case "starsTransactionPeerPlayMarket":
        return { type: "play_market" };
      case "starsTransactionPeerPremiumBot":
        return { type: "premium_bot" };
      case "starsTransactionPeerFragment": {
        if (this.raw.gift) {
          return { type: "fragment" };
        }
        let status;
        if (this.raw.pending) {
          status = "pending";
        } else if (this.raw.failed) {
          status = "failed";
        } else {
          status = "success";
        }
        return {
          type: "fragment_withdraw",
          status,
          date: this.raw.transactionDate ? new Date(this.raw.transactionDate * 1e3) : void 0,
          url: this.raw.transactionUrl
        };
      }
      case "starsTransactionPeerAds":
        return { type: "ads" };
      case "starsTransactionPeerAPI":
        if (this.raw.floodskipNumber != null) {
          return {
            type: "api_floodskip",
            count: this.raw.floodskipNumber
          };
        }
        return { type: "unsupported" };
      case "starsTransactionPeer": {
        const peer$1 = peer.parsePeer(this.raw.peer.peer, this.peers);
        if (this.raw.postsSearch) {
          return { type: "paid_search" };
        }
        if (this.raw.giveawayPostId) {
          return {
            type: "giveaway",
            peer: peer$1,
            messageId: this.raw.giveawayPostId
          };
        }
        if (this.raw.paidMessages) {
          return {
            type: "paid_message",
            peer: peer$1,
            count: this.raw.paidMessages
          };
        }
        if (this.raw.premiumGiftMonths) {
          return {
            type: "premium_gift",
            peer: peer$1,
            months: this.raw.premiumGiftMonths
          };
        }
        if (this.raw.stargift) {
          if (this.raw.stargift._ === "starGift") {
            return {
              type: "star_gift",
              peer: peer$1,
              gift: new starGift.StarGift(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftUpgrade) {
            return {
              type: "star_gift_upgrade",
              peer: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftResale) {
            return {
              type: "star_gift_resale",
              peer: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftPrepaidUpgrade) {
            return {
              peer: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers),
              type: "star_gift_prepaid_upgrade"
            };
          } else if (this.raw.stargiftDropOriginalDetails) {
            return {
              type: "star_gift_drop_details",
              peer: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers)
            };
          } else if (this.raw.offer) {
            return {
              type: "star_gift_offer",
              peer: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers)
            };
          } else {
            return {
              type: "star_gift_transfer",
              recipient: peer$1,
              gift: new starGiftUnique.StarGiftUnique(this.raw.stargift, this.peers)
            };
          }
        }
        if (this.raw.stargiftUpgrade) {
          if (this.raw.msgId) {
            if (this.raw.reaction) {
              return {
                type: "reaction",
                peer: peer$1,
                messageId: this.raw.msgId
              };
            }
            return {
              type: "media_purchase",
              peer: peer$1,
              messageId: this.raw.msgId,
              media: this.raw.extendedMedia ? this.raw.extendedMedia.map((it) => messageMedia._messageMediaFromTl(this.peers, it)) : void 0
            };
          }
        }
        if (this.raw.subscriptionPeriod) {
          return {
            type: "channel_subscription",
            peer: peer$1,
            period: this.raw.subscriptionPeriod
          };
        }
        if (this.raw.starrefCommissionPermille) {
          return {
            type: "bot_referral",
            peer: peer$1,
            commission: this.raw.starrefCommissionPermille
          };
        }
        if (this.raw.adsProceedsFromDate && this.raw.adsProceedsToDate) {
          return {
            type: "ads_proceeds",
            peer: peer$1,
            fromDate: new Date(this.raw.adsProceedsFromDate * 1e3),
            toDate: new Date(this.raw.adsProceedsToDate * 1e3)
          };
        }
        if (peer$1.type === "user") {
          if (this.raw.gift && !peer$1.isBot) {
            return { type: "gift", user: peer$1 };
          }
          if (this.raw.title || this.raw.description || this.raw.photo || this.raw.botPayload) {
            return {
              type: "bot_purchase",
              user: peer$1,
              title: this.raw.title ?? "",
              description: this.raw.description,
              payload: this.raw.botPayload,
              photo: this.raw.photo ? new webDocument.WebDocument(this.raw.photo) : void 0
            };
          }
          return { type: "unsupported" };
        }
        return { type: "unsupported" };
      }
      default:
        return { type: "unsupported" };
    }
  }
  /** Whether this transaction was made by a business bot */
  get viaBusinessBot() {
    return this.raw.businessTransfer;
  }
}
const StarsTransaction$1 = /* @__PURE__ */ inspectable.makeInspectable(StarsTransaction);
const StarsTransaction$2 = /* @__PURE__ */ memoize.memoizeGetters(StarsTransaction$1, ["amount", "type"]);
exports.StarsTransaction = StarsTransaction$2;
