// @mtcute/core/highlevel/types/premium/stars-status.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const peersIndex = require("../peers/peers-index.cjs");
const starsTransaction = require("./stars-transaction.cjs");
class StarsStatus {
  constructor(raw) {
    this.raw = raw;
    this.peers = peersIndex.PeersIndex.from(raw);
  }
  peers;
  /** Current Telegram Stars balance */
  get balance() {
    return this.raw.balance;
  }
  /**
   * History of Telegram Stars transactions
   */
  get transactions() {
    return this.raw.history?.map((it) => new starsTransaction.StarsTransaction(it, this.peers)) ?? [];
  }
  /** Next offset of {@link transactions} for pagination */
  get transactionsNextOffset() {
    return this.raw.nextOffset ?? null;
  }
}
const StarsStatus$1 = /* @__PURE__ */ inspectable.makeInspectable(StarsStatus);
const StarsStatus$2 = /* @__PURE__ */ memoize.memoizeGetters(StarsStatus$1, ["transactions"]);
exports.StarsStatus = StarsStatus$2;
