// @mtcute/core/highlevel/types/premium/star-gift-value.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
class StarGiftValue {
  constructor(raw) {
    this.raw = raw;
  }
  /** Currency of values in this object */
  get currency() {
    return this.raw.currency;
  }
  /** Information about the initial sale of the gift */
  get initialSale() {
    return {
      date: new Date(this.raw.initialSaleDate * 1e3),
      stars: this.raw.initialSaleStars,
      price: this.raw.initialSalePrice
    };
  }
  /** Information about the last sale of the gift on the secondary market (Telegram and Fragment) */
  get lastSale() {
    if (!this.raw.lastSaleDate || !this.raw.lastSalePrice) return null;
    return {
      date: new Date(this.raw.lastSaleDate * 1e3),
      price: this.raw.lastSalePrice,
      fragment: this.raw.lastSaleOnFragment
    };
  }
  /** Whether {@link value} is calculated as an after-market average over the past month */
  get isValueAverage() {
    return this.raw.valueIsAverage;
  }
  /** Estimated value of the gift (in {@link currency}) */
  get value() {
    return this.raw.value;
  }
  /** Floor price for the gift on the secondary market (Telegram and Fragment), if available */
  get floorPrice() {
    return this.raw.floorPrice ?? null;
  }
  /** Average price for the gift on the secondary market (Telegram and Fragment), if available */
  get averagePrice() {
    return this.raw.averagePrice ?? null;
  }
  /** Number of such gifts listed on the secondary market (Telegram and Fragment) */
  get listedCount() {
    return {
      telegram: this.raw.listedCount ?? 0,
      fragment: this.raw.fragmentListedCount ?? 0
    };
  }
  /** URL to the collectible on Fragment */
  get fragmentUrl() {
    return this.raw.fragmentListedUrl ?? null;
  }
}
const StarGiftValue$1 = /* @__PURE__ */ memoize.memoizeGetters(StarGiftValue, ["initialSale", "lastSale"]);
const StarGiftValue$2 = /* @__PURE__ */ inspectable.makeInspectable(StarGiftValue$1);
exports.StarGiftValue = StarGiftValue$2;
