// @mtcute/core/highlevel/types/premium/star-gift-unique.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const peerUtils = require("../../../utils/peer-utils.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const documentUtils = require("../media/document-utils.cjs");
const peer = require("../peers/peer.cjs");
function mapRarity(rarity) {
  switch (rarity._) {
    case "starGiftAttributeRarityUncommon":
      return "uncommon";
    case "starGiftAttributeRarityRare":
      return "rare";
    case "starGiftAttributeRarityEpic":
      return "epic";
    case "starGiftAttributeRarityLegendary":
      return "legendary";
    default:
      return null;
  }
}
class StarGiftUniqueAttribute {
  constructor(raw) {
    this.raw = raw;
  }
  /** Whether this attribute is specific to this craft */
  get crafted() {
    return this.raw._ === "starGiftAttributeModel" && this.raw.crafted;
  }
  /** Rarity permille of the attribute, if available */
  get permille() {
    return this.raw.rarity._ === "starGiftAttributeRarity" ? this.raw.rarity.permille : null;
  }
  /** Rarity type of the attribute, if available. Mutually exclusive with {@link permille} */
  get rarity() {
    return mapRarity(this.raw.rarity);
  }
  /** Display name of the attribute */
  get name() {
    return this.raw.name;
  }
  /** Sticker of the attribute */
  get sticker() {
    utils.assert(this.raw.document._ === "document");
    const parsed = documentUtils.parseDocument(this.raw.document);
    utils.assert(parsed.type === "sticker");
    return parsed;
  }
}
const StarGiftUniqueAttribute$1 = /* @__PURE__ */ inspectable.makeInspectable(StarGiftUniqueAttribute);
const StarGiftUniqueAttribute$2 = /* @__PURE__ */ memoize.memoizeGetters(StarGiftUniqueAttribute$1, ["sticker"]);
class StarGiftUniqueBackdrop {
  constructor(raw) {
    this.raw = raw;
  }
  /** ID of the backdrop */
  get id() {
    return this.raw.backdropId;
  }
  /** Rarity permille of the attribute, if available */
  get permille() {
    return this.raw.rarity._ === "starGiftAttributeRarity" ? this.raw.rarity.permille : null;
  }
  /** Rarity type of the attribute, if available. Mutually exclusive with {@link permille} */
  get rarity() {
    return mapRarity(this.raw.rarity);
  }
  /** Display name of the attribute */
  get name() {
    return this.raw.name;
  }
  get centerColor() {
    return this.raw.centerColor;
  }
  get edgeColor() {
    return this.raw.edgeColor;
  }
  get patternColor() {
    return this.raw.patternColor;
  }
  get textColor() {
    return this.raw.textColor;
  }
}
const StarGiftUniqueBackdrop$1 = /* @__PURE__ */ inspectable.makeInspectable(StarGiftUniqueBackdrop);
class StarGiftUniqueOriginalDetails {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /**
   * Peer who sent the original star gift, if available
   *
   * > Note: in some cases, {@link senderId} might be available, but not this field.
   * > In such cases, you should try fetching the peer manually using {@link getPeer})
   */
  get sender() {
    if (!this.raw.senderId) return null;
    if (!this._peers.has(this.raw.senderId)) return null;
    return peer.parsePeer(this.raw.senderId, this._peers);
  }
  /** ID of the peer who sent the original star gift */
  get senderId() {
    return this.raw.senderId ? peerUtils.getMarkedPeerId(this.raw.senderId) : null;
  }
  /**
   * Peer who received the original star gift, if available
   *
   * > Note: {@link recipientId} is always available. If ID is available, but this field is null,
   * > you should try fetching the peer manually using {@link getPeer})
   */
  get recipient() {
    if (!this._peers.has(this.raw.recipientId)) return null;
    return peer.parsePeer(this.raw.recipientId, this._peers);
  }
  /** ID of the peer who received the original star gift */
  get recipientId() {
    return peerUtils.getMarkedPeerId(this.raw.recipientId);
  }
  /** Date when the original star gift was sent */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Message attached to the original star gift */
  get message() {
    return this.raw.message ?? null;
  }
}
const StarGiftUniqueOriginalDetails$1 = /* @__PURE__ */ inspectable.makeInspectable(StarGiftUniqueOriginalDetails);
class StarGiftUnique {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
    for (const attr of this.raw.attributes) {
      if (attr._ === "starGiftAttributeModel") {
        this._model = attr;
      } else if (attr._ === "starGiftAttributePattern") {
        this._pattern = attr;
      } else if (attr._ === "starGiftAttributeBackdrop") {
        this._backdrop = attr;
      } else if (attr._ === "starGiftAttributeOriginalDetails") {
        this._originalDetails = attr;
      }
    }
    utils.assert(this._model !== void 0);
    utils.assert(this._pattern !== void 0);
    utils.assert(this._backdrop !== void 0);
  }
  _model;
  _pattern;
  _backdrop;
  _originalDetails;
  /** Whether this gift is a unique gift */
  isUnique = true;
  /** Whether this gift is available only to premium users */
  get isPremiumOnly() {
    return this.raw.requirePremium;
  }
  /** Whether this gift was burned to craft a new one */
  get isBurned() {
    return this.raw.burned;
  }
  /** Whether this gift was crafted */
  get isCrafted() {
    return this.raw.crafted;
  }
  /** If this gift can be used in a craft, the chance of success, in permille */
  get craftChancePermille() {
    return this.raw.craftChancePermille ?? null;
  }
  /** Number of the NFT */
  get num() {
    return this.raw.num;
  }
  /** Title of the NFT */
  get title() {
    return this.raw.title;
  }
  /** Slug of the gift */
  get slug() {
    return this.raw.slug;
  }
  /**
   * ID of the peer who owns this gift, if available
   * > Note: in some cases, only {@link ownerId} is available, and not {@link owner},
   * > in which cases please use `tg.getPeer(ownerId)` manually
   */
  get owner() {
    if (!this.raw.ownerId) return null;
    if (!this._peers.has(this.raw.ownerId)) return null;
    return peer.parsePeer(this.raw.ownerId, this._peers);
  }
  /** ID of the peer who owns this gift, if available */
  get ownerId() {
    return this.raw.ownerId ? peerUtils.getMarkedPeerId(this.raw.ownerId) : null;
  }
  /** Name of the user who owns this gift, if available */
  get ownerName() {
    return this.raw.ownerName ?? null;
  }
  // todo: merge into one property
  get availabilityIssued() {
    return this.raw.availabilityIssued;
  }
  get availabilityTotal() {
    return this.raw.availabilityTotal;
  }
  /** Number of stars this gift is up for resell for */
  get resellPriceStars() {
    const resellPriceStars = this.raw.resellAmount?.find((it) => it._ === "starsAmount");
    return resellPriceStars?.amount ?? null;
  }
  /** Number of nanoton this gift is up for resell for */
  get resellPriceTon() {
    const resellPriceTon = this.raw.resellAmount?.find((it) => it._ === "starsTonAmount");
    return resellPriceTon?.amount ?? null;
  }
  /** Profile colors associated with this gift */
  get peerColor() {
    if (this.raw.peerColor?._ !== "peerColorCollectible") return null;
    return this.raw.peerColor;
  }
  /**
   * ID of the user who "hosts" this gift (i.e. this gift is actually stored on the TON blockchain,
   * but is displayed as a gift on this peer's profile)
   */
  get hostId() {
    if (!this.raw.hostId) return null;
    return peerUtils.getMarkedPeerId(this.raw.hostId);
  }
  /**
   * ID of the peer who "hosts" this gift (i.e. this gift is actually stored on the TON blockchain,
   * but is displayed as a gift on this peer's profile), if available
   * > Note: in some cases, only {@link hostId} is available, and not {@link host},
   * > in which cases please use `tg.getPeer(ownerId)` manually
   */
  get host() {
    if (!this.raw.hostId) return null;
    if (!this._peers.has(this.raw.hostId)) return null;
    return peer.parsePeer(this.raw.hostId, this._peers);
  }
  /** Model (i.e. the gift itself) of the unique star gift */
  get model() {
    return new StarGiftUniqueAttribute(this._model);
  }
  /** Pattern of the unique star gift */
  get pattern() {
    return new StarGiftUniqueAttribute(this._pattern);
  }
  /** Backdrop of the unique star gift */
  get backdrop() {
    return new StarGiftUniqueBackdrop(this._backdrop);
  }
  /** Details about the original star gift (if they were retained when minting) */
  get originalDetails() {
    if (!this._originalDetails) return null;
    return new StarGiftUniqueOriginalDetails(this._originalDetails, this._peers);
  }
  /** TON address of the owner of the unique star gift */
  get ownerAddress() {
    return this.raw.ownerAddress ?? null;
  }
  /** TON address of the gift NFT */
  get giftAddress() {
    return this.raw.giftAddress ?? null;
  }
  /**
   * User/channel who released this collection
   */
  get releasedBy() {
    if (!this.raw.releasedBy) return null;
    return peer.parsePeer(this.raw.releasedBy, this._peers);
  }
  /** Value of the gift in USD */
  get valueUsd() {
    return this.raw.valueUsdAmount ?? null;
  }
  /** Minimum number of stars required to create a purchase offer for the gift */
  get offerMinStars() {
    return this.raw.offerMinStars ?? null;
  }
}
const StarGiftUnique$1 = /* @__PURE__ */ inspectable.makeInspectable(StarGiftUnique);
const StarGiftUnique$2 = /* @__PURE__ */ memoize.memoizeGetters(StarGiftUnique$1, [
  "model",
  "pattern",
  "backdrop",
  "originalDetails",
  "resellPriceStars",
  "resellPriceTon",
  "releasedBy"
]);
exports.StarGiftUnique = StarGiftUnique$2;
exports.StarGiftUniqueAttribute = StarGiftUniqueAttribute$2;
exports.StarGiftUniqueBackdrop = StarGiftUniqueBackdrop$1;
exports.StarGiftUniqueOriginalDetails = StarGiftUniqueOriginalDetails$1;
