// @mtcute/core/highlevel/types/premium/star-gift.js
import Long from "long";
import { MtTypeAssertionError } from "../../../types/errors.js";
import { assertTypeIs } from "../../../utils/type-assertions.js";
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { parseDocument } from "../media/document-utils.js";
import { parsePeer } from "../peers/peer.js";
class StarGift {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** ID of the gift */
  get id() {
    return this.raw.id;
  }
  /** Whether this gift sold out and cannot be bought anymore */
  get isSoldOut() {
    return this.raw.soldOut;
  }
  /** Whether this gift has limited availability */
  get isLimited() {
    return this.raw.limited;
  }
  /** Whether this gift is available only to premium users */
  get isPremiumOnly() {
    return this.raw.requirePremium;
  }
  /** If this gift is currently sold through an auction, info about that auction. `null` otherwise */
  get auction() {
    if (!this.raw.auctionSlug || this.raw.giftsPerRound == null || !this.raw.auctionStartDate) return null;
    return {
      slug: this.raw.auctionSlug,
      giftsPerRound: this.raw.giftsPerRound,
      startDate: new Date(this.raw.auctionStartDate * 1e3)
    };
  }
  /** Whether this gift includes a peer color that can be applied to user's profile */
  get hasPeerColor() {
    return this.raw.peerColorAvailable;
  }
  /** Whether this gift is a unique gift */
  isUnique = false;
  /** Additional information for sold-out gifts */
  get soldOutInfo() {
    if (this.raw.firstSaleDate == null || this.raw.lastSaleDate == null) {
      return null;
    }
    return {
      firstSale: new Date(this.raw.firstSaleDate * 1e3),
      lastSale: new Date(this.raw.lastSaleDate * 1e3)
    };
  }
  /** Sticker associated with the gift */
  get sticker() {
    assertTypeIs("StarGift#sticker", this.raw.sticker, "document");
    const parsed = parseDocument(this.raw.sticker);
    if (parsed.type !== "sticker") {
      throw new MtTypeAssertionError("StarGift#sticker", "sticker", parsed.type);
    }
    return parsed;
  }
  /** Amount of stars the gift was purchased for */
  get purchaseStars() {
    return this.raw.stars;
  }
  /**
   * Amount of stars the gift can be converted to by the recipient
   */
  get convertStars() {
    return this.raw.convertStars;
  }
  /**
   * Amount of stars the gift can be upgraded for
   */
  get upgradeStars() {
    return this.raw.upgradeStars ?? null;
  }
  /**
   * For limited availability gifts,
   * the number of remaining and total gifts available
   */
  get availability() {
    if (this.raw.availabilityRemains == null || this.raw.availabilityTotal == null) {
      return null;
    }
    return {
      remains: this.raw.availabilityRemains,
      total: this.raw.availabilityTotal,
      resale: this.raw.availabilityResale ?? Long.ZERO
    };
  }
  /**
   * For gifts with limited availability per user,
   * the number of remaining and total gifts available
   */
  get perUserAvailability() {
    if (!this.raw.limitedPerUser) return null;
    return {
      // they share the same flag so they must be present
      remains: this.raw.perUserRemains,
      total: this.raw.perUserTotal
    };
  }
  /** Floor price for the gift on the secondary market */
  get resaleFloorPrice() {
    return this.raw.resellMinStars ?? null;
  }
  /** Title of the gift */
  get title() {
    return this.raw.title ?? null;
  }
  /**
   * User/channel who released this collection
   */
  get releasedBy() {
    if (!this.raw.releasedBy) return null;
    return parsePeer(this.raw.releasedBy, this._peers);
  }
  /** Number of unique upgrade variants that will be available for the gift */
  get upgradeVariants() {
    return this.raw.upgradeVariants ?? 0;
  }
  /** Background color of the gift */
  get background() {
    return this.raw.background ?? null;
  }
}
const StarGift$1 = /* @__PURE__ */ makeInspectable(StarGift);
const StarGift$2 = /* @__PURE__ */ memoizeGetters(StarGift$1, ["sticker", "auction"]);
export {
  StarGift$2 as StarGift
};
