// @mtcute/core/highlevel/types/premium/star-gift.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/core] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/core] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Long = require("long");
const errors = require("../../../types/errors.cjs");
const typeAssertions = require("../../../utils/type-assertions.cjs");
const inspectable = require("../../utils/inspectable.cjs");
const memoize = require("../../utils/memoize.cjs");
const documentUtils = require("../media/document-utils.cjs");
const peer = require("../peers/peer.cjs");
class StarGift {
  constructor(raw, _peers) {
    this.raw = raw;
    this._peers = _peers;
  }
  /** ID of the gift */
  get id() {
    return this.raw.id;
  }
  /** Whether this gift sold out and cannot be bought anymore */
  get isSoldOut() {
    return this.raw.soldOut;
  }
  /** Whether this gift has limited availability */
  get isLimited() {
    return this.raw.limited;
  }
  /** Whether this gift is available only to premium users */
  get isPremiumOnly() {
    return this.raw.requirePremium;
  }
  /** If this gift is currently sold through an auction, info about that auction. `null` otherwise */
  get auction() {
    if (!this.raw.auctionSlug || this.raw.giftsPerRound == null || !this.raw.auctionStartDate) return null;
    return {
      slug: this.raw.auctionSlug,
      giftsPerRound: this.raw.giftsPerRound,
      startDate: new Date(this.raw.auctionStartDate * 1e3)
    };
  }
  /** Whether this gift includes a peer color that can be applied to user's profile */
  get hasPeerColor() {
    return this.raw.peerColorAvailable;
  }
  /** Whether this gift is a unique gift */
  isUnique = false;
  /** Additional information for sold-out gifts */
  get soldOutInfo() {
    if (this.raw.firstSaleDate == null || this.raw.lastSaleDate == null) {
      return null;
    }
    return {
      firstSale: new Date(this.raw.firstSaleDate * 1e3),
      lastSale: new Date(this.raw.lastSaleDate * 1e3)
    };
  }
  /** Sticker associated with the gift */
  get sticker() {
    typeAssertions.assertTypeIs("StarGift#sticker", this.raw.sticker, "document");
    const parsed = documentUtils.parseDocument(this.raw.sticker);
    if (parsed.type !== "sticker") {
      throw new errors.MtTypeAssertionError("StarGift#sticker", "sticker", parsed.type);
    }
    return parsed;
  }
  /** Amount of stars the gift was purchased for */
  get purchaseStars() {
    return this.raw.stars;
  }
  /**
   * Amount of stars the gift can be converted to by the recipient
   */
  get convertStars() {
    return this.raw.convertStars;
  }
  /**
   * Amount of stars the gift can be upgraded for
   */
  get upgradeStars() {
    return this.raw.upgradeStars ?? null;
  }
  /**
   * For limited availability gifts,
   * the number of remaining and total gifts available
   */
  get availability() {
    if (this.raw.availabilityRemains == null || this.raw.availabilityTotal == null) {
      return null;
    }
    return {
      remains: this.raw.availabilityRemains,
      total: this.raw.availabilityTotal,
      resale: this.raw.availabilityResale ?? Long.ZERO
    };
  }
  /**
   * For gifts with limited availability per user,
   * the number of remaining and total gifts available
   */
  get perUserAvailability() {
    if (!this.raw.limitedPerUser) return null;
    return {
      // they share the same flag so they must be present
      remains: this.raw.perUserRemains,
      total: this.raw.perUserTotal
    };
  }
  /** Floor price for the gift on the secondary market */
  get resaleFloorPrice() {
    return this.raw.resellMinStars ?? null;
  }
  /** Title of the gift */
  get title() {
    return this.raw.title ?? null;
  }
  /**
   * User/channel who released this collection
   */
  get releasedBy() {
    if (!this.raw.releasedBy) return null;
    return peer.parsePeer(this.raw.releasedBy, this._peers);
  }
  /** Number of unique upgrade variants that will be available for the gift */
  get upgradeVariants() {
    return this.raw.upgradeVariants ?? 0;
  }
  /** Background color of the gift */
  get background() {
    return this.raw.background ?? null;
  }
}
const StarGift$1 = /* @__PURE__ */ inspectable.makeInspectable(StarGift);
const StarGift$2 = /* @__PURE__ */ memoize.memoizeGetters(StarGift$1, ["sticker", "auction"]);
exports.StarGift = StarGift$2;
