// @mtcute/core/highlevel/types/premium/stars-transaction.js
import { makeInspectable } from "../../utils/inspectable.js";
import { memoizeGetters } from "../../utils/memoize.js";
import { WebDocument as WebDocument$1 } from "../files/web-document.js";
import { _messageMediaFromTl } from "../messages/message-media.js";
import { parsePeer } from "../peers/peer.js";
import { StarGiftUnique as StarGiftUnique$2 } from "./star-gift-unique.js";
import { StarGift as StarGift$2 } from "./star-gift.js";
class StarsTransaction {
  constructor(raw, peers) {
    this.raw = raw;
    this.peers = peers;
  }
  /** ID of the transaction */
  get id() {
    return this.raw.id;
  }
  /** Whether this transaction is a refund */
  get isRefund() {
    return this.raw.refund;
  }
  /**
   * Whether this transaction is outgoing or incoming
   */
  get direction() {
    let isNegative = this.raw.amount.amount.isNegative();
    if (this.raw.refund) isNegative = !isNegative;
    return isNegative ? "outgoing" : "incoming";
  }
  /** Absolute amount of stars/TON in the transaction */
  get amount() {
    let res = this.raw.amount;
    if (res.amount.isNegative()) {
      res = res._ === "starsAmount" ? {
        ...res,
        nanos: Math.abs(res.nanos),
        amount: res.amount.negate()
      } : {
        _: "starsTonAmount",
        amount: res.amount.negate()
      };
    }
    return res;
  }
  /** Date of the transaction */
  get date() {
    return new Date(this.raw.date * 1e3);
  }
  /** Type of this transaction */
  get type() {
    switch (this.raw.peer._) {
      case "starsTransactionPeerAppStore":
        return { type: "app_store" };
      case "starsTransactionPeerPlayMarket":
        return { type: "play_market" };
      case "starsTransactionPeerPremiumBot":
        return { type: "premium_bot" };
      case "starsTransactionPeerFragment": {
        if (this.raw.gift) {
          return { type: "fragment" };
        }
        let status;
        if (this.raw.pending) {
          status = "pending";
        } else if (this.raw.failed) {
          status = "failed";
        } else {
          status = "success";
        }
        return {
          type: "fragment_withdraw",
          status,
          date: this.raw.transactionDate ? new Date(this.raw.transactionDate * 1e3) : void 0,
          url: this.raw.transactionUrl
        };
      }
      case "starsTransactionPeerAds":
        return { type: "ads" };
      case "starsTransactionPeerAPI":
        if (this.raw.floodskipNumber != null) {
          return {
            type: "api_floodskip",
            count: this.raw.floodskipNumber
          };
        }
        return { type: "unsupported" };
      case "starsTransactionPeer": {
        const peer = parsePeer(this.raw.peer.peer, this.peers);
        if (this.raw.postsSearch) {
          return { type: "paid_search" };
        }
        if (this.raw.giveawayPostId) {
          return {
            type: "giveaway",
            peer,
            messageId: this.raw.giveawayPostId
          };
        }
        if (this.raw.paidMessages) {
          return {
            type: "paid_message",
            peer,
            count: this.raw.paidMessages
          };
        }
        if (this.raw.premiumGiftMonths) {
          return {
            type: "premium_gift",
            peer,
            months: this.raw.premiumGiftMonths
          };
        }
        if (this.raw.stargift) {
          if (this.raw.stargift._ === "starGift") {
            return {
              type: "star_gift",
              peer,
              gift: new StarGift$2(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftUpgrade) {
            return {
              type: "star_gift_upgrade",
              peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftResale) {
            return {
              type: "star_gift_resale",
              peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers)
            };
          } else if (this.raw.stargiftPrepaidUpgrade) {
            return {
              peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers),
              type: "star_gift_prepaid_upgrade"
            };
          } else if (this.raw.stargiftDropOriginalDetails) {
            return {
              type: "star_gift_drop_details",
              peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers)
            };
          } else if (this.raw.offer) {
            return {
              type: "star_gift_offer",
              peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers)
            };
          } else {
            return {
              type: "star_gift_transfer",
              recipient: peer,
              gift: new StarGiftUnique$2(this.raw.stargift, this.peers)
            };
          }
        }
        if (this.raw.stargiftUpgrade) {
          if (this.raw.msgId) {
            if (this.raw.reaction) {
              return {
                type: "reaction",
                peer,
                messageId: this.raw.msgId
              };
            }
            return {
              type: "media_purchase",
              peer,
              messageId: this.raw.msgId,
              media: this.raw.extendedMedia ? this.raw.extendedMedia.map((it) => _messageMediaFromTl(this.peers, it)) : void 0
            };
          }
        }
        if (this.raw.subscriptionPeriod) {
          return {
            type: "channel_subscription",
            peer,
            period: this.raw.subscriptionPeriod
          };
        }
        if (this.raw.starrefCommissionPermille) {
          return {
            type: "bot_referral",
            peer,
            commission: this.raw.starrefCommissionPermille
          };
        }
        if (this.raw.adsProceedsFromDate && this.raw.adsProceedsToDate) {
          return {
            type: "ads_proceeds",
            peer,
            fromDate: new Date(this.raw.adsProceedsFromDate * 1e3),
            toDate: new Date(this.raw.adsProceedsToDate * 1e3)
          };
        }
        if (peer.type === "user") {
          if (this.raw.gift && !peer.isBot) {
            return { type: "gift", user: peer };
          }
          if (this.raw.title || this.raw.description || this.raw.photo || this.raw.botPayload) {
            return {
              type: "bot_purchase",
              user: peer,
              title: this.raw.title ?? "",
              description: this.raw.description,
              payload: this.raw.botPayload,
              photo: this.raw.photo ? new WebDocument$1(this.raw.photo) : void 0
            };
          }
          return { type: "unsupported" };
        }
        return { type: "unsupported" };
      }
      default:
        return { type: "unsupported" };
    }
  }
  /** Whether this transaction was made by a business bot */
  get viaBusinessBot() {
    return this.raw.businessTransfer;
  }
}
const StarsTransaction$1 = /* @__PURE__ */ makeInspectable(StarsTransaction);
const StarsTransaction$2 = /* @__PURE__ */ memoizeGetters(StarsTransaction$1, ["amount", "type"]);
export {
  StarsTransaction$2 as StarsTransaction
};
