// @fuman/utils/bigint/math.js
function bitLength(n) {
  if (n === 0n) return 0;
  const i = (n.toString(16).length - 1) * 4;
  return i + 32 - Math.clz32(Number(n >> BigInt(i)));
}
function twoMultiplicity(n) {
  if (n === 0n) return 0n;
  let m = 0n;
  let pow = 1n;
  while (true) {
    if ((n & pow) !== 0n) return m;
    m += 1n;
    pow <<= 1n;
  }
}
function min2(a, b) {
  return a < b ? a : b;
}
function min(...args) {
  let m = args[0];
  for (let i = 1; i < args.length; i++) {
    if (args[i] < m) m = args[i];
  }
  return m;
}
function max2(a, b) {
  return a > b ? a : b;
}
function max(...args) {
  let m = args[0];
  for (let i = 1; i < args.length; i++) {
    if (args[i] > m) m = args[i];
  }
  return m;
}
function abs(a) {
  return a < 0n ? -a : a;
}
function euclideanGcd(a, b) {
  while (b !== 0n) {
    const t = b;
    b = a % b;
    a = t;
  }
  return a;
}
function modPowBinary(base, exp, mod) {
  base %= mod;
  let result = 1n;
  while (exp > 0n) {
    if (exp % 2n === 1n) {
      result = result * base % mod;
    }
    exp >>= 1n;
    base = base ** 2n % mod;
  }
  return result;
}
function eGcd(a, b) {
  let x = 0n;
  let y = 1n;
  let u = 1n;
  let v = 0n;
  while (a !== 0n) {
    const q = b / a;
    const r = b % a;
    const m = x - u * q;
    const n = y - v * q;
    b = a;
    a = r;
    x = u;
    y = v;
    u = m;
    v = n;
  }
  return [b, x, y];
}
function toZn(a, n) {
  if (typeof a === "number") a = BigInt(a);
  if (typeof n === "number") n = BigInt(n);
  if (n <= 0n) {
    throw new RangeError("n must be > 0");
  }
  const aZn = a % n;
  return aZn < 0n ? aZn + n : aZn;
}
function modInv(a, n) {
  const [g, x] = eGcd(toZn(a, n), n);
  if (g !== 1n) {
    throw new RangeError(`${a.toString()} does not have inverse modulo ${n.toString()}`);
  } else {
    return toZn(x, n);
  }
}
export {
  abs,
  bitLength,
  euclideanGcd,
  max,
  max2,
  min,
  min2,
  modInv,
  modPowBinary,
  twoMultiplicity
};
